<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Graph Type Tutorials · Mycelia</title><meta name="title" content="Graph Type Tutorials · Mycelia"/><meta property="og:title" content="Graph Type Tutorials · Mycelia"/><meta property="twitter:title" content="Graph Type Tutorials · Mycelia"/><meta name="description" content="Documentation for Mycelia."/><meta property="og:description" content="Documentation for Mycelia."/><meta property="twitter:description" content="Documentation for Mycelia."/><script data-outdated-warner src="../../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../search_index.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../../">Mycelia</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../../">Home</a></li><li><a class="tocitem" href="../../../getting-started/">Getting Started</a></li><li><a class="tocitem" href="../../../installation/">Installation</a></li><li><a class="tocitem" href="../../../concepts/">Concepts</a></li><li><a class="tocitem" href="../../../workflow-map/">Workflow Map</a></li><li><a class="tocitem" href="../../../metagenomic-workflow/">Metagenomic Workflow</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../../../tutorials/">Overview</a></li><li><a class="tocitem" href="../01_data_acquisition/">Step 1: Data Acquisition</a></li><li><a class="tocitem" href="../02_quality_control/">Step 2: Quality Control</a></li><li><a class="tocitem" href="../03_kmer_analysis/">Step 3: K-mer Analysis</a></li><li><a class="tocitem" href="../04_genome_assembly/">Step 4: Genome Assembly</a></li><li><a class="tocitem" href="../05_assembly_validation/">Step 5: Assembly Validation</a></li><li><a class="tocitem" href="../06_gene_annotation/">Step 6: Gene Annotation</a></li><li><a class="tocitem" href="../07_comparative_genomics/">Step 7: Comparative Genomics</a></li><li><a class="tocitem" href="../08_tool_integration/">Step 8: Tool Integration</a></li><li><a class="tocitem" href="../14_binning_workflow/">Step 9: Binning Workflow</a></li><li class="is-active"><a class="tocitem" href>Graph Type Tutorials</a></li><li><a class="tocitem" href="../09_round_trip_01_string_graphs/">Round Trip 01 String Graphs</a></li><li><a class="tocitem" href="../09_round_trip_02_ngram_to_string/">Round Trip 02 Ngram To String</a></li><li><a class="tocitem" href="../09_round_trip_03_fasta_sequences/">Round Trip 03 Fasta Sequences</a></li><li><a class="tocitem" href="../09_round_trip_04_kmer_to_sequence/">Round Trip 04 Kmer To Sequence</a></li><li><a class="tocitem" href="../09_round_trip_05_fastq_graphs/">Round Trip 05 Fastq Graphs</a></li><li><a class="tocitem" href="../09_round_trip_06_qualmer_graphs/">Round Trip 06 Qualmer Graphs</a></li><li><a class="tocitem" href="../10_viroid_assembly_workflow/">Viroid Assembly Workflow</a></li><li><a class="tocitem" href="../11_reduced_amino_acid_alphabets/">Reduced Amino Acid Alphabets</a></li><li><a class="tocitem" href="../12_coverm_coverage/">Coverm Coverage</a></li><li><a class="tocitem" href="../13_rhizomorph_assembly/">Rhizomorph Assembly</a></li><li><a class="tocitem" href="../14_mash_classification/">Mash Classification</a></li><li><a class="tocitem" href="../15_round_trip_benchmarking/">Round Trip Benchmarking</a></li><li><a class="tocitem" href="../advanced-assembly-theory-and-practice/">Advanced-Assembly-Theory-And-Practice</a></li><li><a class="tocitem" href="../run_all_tutorials/">Run All Tutorials</a></li></ul></li><li><span class="tocitem">API Reference</span><ul><li><a class="tocitem" href="../../../api/all-functions/">Complete API Surface</a></li></ul></li><li><a class="tocitem" href="../../../benchmarks/">Benchmarks</a></li><li><a class="tocitem" href="../../../contributing/">Contributing</a></li><li><a class="tocitem" href="../../../related-projects/">Related Projects</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorials</a></li><li class="is-active"><a href>Graph Type Tutorials</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Graph Type Tutorials</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/cjprybol/Mycelia" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/cjprybol/Mycelia/blob/master/tutorials/04_graph_type_tutorials.jl" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><pre><code class="language-julia hljs">&quot;&quot;&quot;
Rhizomorph Graph Type Tutorials

This tutorial demonstrates the usage of all 6 graph types in Mycelia&#39;s Rhizomorph submodule:
1. N-gram Graphs - For unicode text analysis
2. K-mer Graphs - For DNA/RNA/protein sequence analysis
3. Qualmer Graphs - For quality-aware sequence analysis
4. String Graphs - For variable-length string overlaps (OLC)
5. FASTA Graphs - For variable-length BioSequence overlaps (OLC)
6. FASTQ Graphs - For quality-aware BioSequence overlaps (OLC)

Each section shows practical examples with real data using Rhizomorph APIs.
&quot;&quot;&quot;

import Mycelia
import FASTX
import MetaGraphsNext
import Statistics

println(&quot;=&quot;^80)
println(&quot;RHIZOMORPH GRAPH TYPE TUTORIALS&quot;)
println(&quot;=&quot;^80)</code></pre><p>============================================================================= Tutorial 1: N-gram Graphs - Unicode Text Analysis =============================================================================</p><pre><code class="language-julia hljs">println(&quot;\n1. N-GRAM GRAPHS - Unicode Text Analysis&quot;)
println(&quot;-&quot;^50)</code></pre><p>Example: Analyzing text patterns</p><pre><code class="language-julia hljs">text = &quot;The quick brown fox jumps over the lazy dog&quot;
println(&quot;Input text: \&quot;$text\&quot;&quot;)</code></pre><p>Create N-gram graph</p><pre><code class="language-julia hljs">ngram_graph = Mycelia.Rhizomorph.build_ngram_graph([text], 3; dataset_id=&quot;text_demo&quot;)
ngram_labels = collect(MetaGraphsNext.labels(ngram_graph))
println(&quot;Number of 3-grams: $(length(ngram_labels))&quot;)</code></pre><p>Show some N-grams</p><pre><code class="language-julia hljs">ngrams = ngram_labels
println(&quot;Example 3-grams: $(ngrams[1:min(5, length(ngrams))])&quot;)</code></pre><p>Use case: Text compression, pattern detection, linguistic analysis</p><pre><code class="language-julia hljs">println(&quot;Use cases: Text compression, pattern detection, linguistic analysis&quot;)</code></pre><p>============================================================================= Tutorial 2: K-mer Graphs - BioSequence Analysis =============================================================================</p><pre><code class="language-julia hljs">println(&quot;\n2. K-MER GRAPHS - BioSequence Analysis&quot;)
println(&quot;-&quot;^50)</code></pre><p>Example: DNA sequence analysis</p><pre><code class="language-julia hljs">dna_sequence = &quot;ATCGATCGATCGATCGTAGCTAGCTAGCT&quot;
println(&quot;DNA sequence: $dna_sequence&quot;)</code></pre><p>Create FASTA record</p><pre><code class="language-julia hljs">dna_record = FASTX.FASTA.Record(&quot;sample_dna&quot;, dna_sequence)
dna_records = [dna_record]</code></pre><p>Create k-mer graph</p><pre><code class="language-julia hljs">dna_dataset_id = &quot;dna_demo&quot;
dna_kmer_graph = Mycelia.Rhizomorph.build_kmer_graph(dna_records, 5; dataset_id=dna_dataset_id)
dna_kmers = collect(MetaGraphsNext.labels(dna_kmer_graph))
println(&quot;DNA k-mers (k=5): $(length(dna_kmers)) unique k-mers&quot;)
println(&quot;Example k-mers: $(dna_kmers[1:min(3, length(dna_kmers))])&quot;)</code></pre><p>Get vertex data to see evidence counts</p><pre><code class="language-julia hljs">first_kmer = first(dna_kmers)
vertex_data = dna_kmer_graph[first_kmer]
coverage = Mycelia.Rhizomorph.count_evidence_entries(vertex_data)
println(&quot;Evidence entries for $(first_kmer): $(coverage) observations&quot;)</code></pre><p>Example: RNA sequence analysis</p><pre><code class="language-julia hljs">rna_sequence = &quot;AUCGAUCGAUCGAUCGUAGCUAGCUAGCU&quot;
println(&quot;\\nRNA sequence: $rna_sequence&quot;)

rna_record = FASTX.FASTA.Record(&quot;sample_rna&quot;, rna_sequence)
rna_records = [rna_record]</code></pre><p>Create RNA k-mer graph</p><pre><code class="language-julia hljs">rna_kmer_graph = Mycelia.Rhizomorph.build_kmer_graph(rna_records, 4; dataset_id=&quot;rna_demo&quot;)
rna_kmers = collect(MetaGraphsNext.labels(rna_kmer_graph))
println(&quot;RNA k-mers (k=4): $(length(rna_kmers)) unique k-mers&quot;)
println(&quot;Example k-mers: $(rna_kmers[1:min(3, length(rna_kmers))])&quot;)</code></pre><p>Example: Protein sequence analysis</p><pre><code class="language-julia hljs">protein_sequence = &quot;ALAVALINEGLUTAMINELEUCINE&quot;
println(&quot;\\nProtein sequence: $protein_sequence&quot;)

protein_record = FASTX.FASTA.Record(&quot;sample_protein&quot;, protein_sequence)
protein_records = [protein_record]</code></pre><p>Create protein k-mer graph</p><pre><code class="language-julia hljs">protein_kmer_graph = Mycelia.Rhizomorph.build_kmer_graph(protein_records, 3; dataset_id=&quot;protein_demo&quot;)
protein_kmers = collect(MetaGraphsNext.labels(protein_kmer_graph))
println(&quot;Protein k-mers (k=3): $(length(protein_kmers)) unique k-mers&quot;)
println(&quot;Example k-mers: $(protein_kmers[1:min(3, length(protein_kmers))])&quot;)</code></pre><p>Use cases</p><pre><code class="language-julia hljs">println(&quot;Use cases: Genome assembly, repeat detection, sequence comparison&quot;)</code></pre><p>============================================================================= Tutorial 3: Qualmer Graphs - Quality-Aware Sequence Analysis =============================================================================</p><pre><code class="language-julia hljs">println(&quot;\n3. QUALMER GRAPHS - Quality-Aware Sequence Analysis&quot;)
println(&quot;-&quot;^50)</code></pre><p>Example: High-quality sequencing data</p><pre><code class="language-julia hljs">high_quality_seq = &quot;ATCGATCGATCGATCGTAGCTAGCT&quot;
high_quality_scores = &quot;HHHHHHHHHHHHHHHHHHHHHHHHH&quot;  ## PHRED 39
println(&quot;High-quality sequence: $high_quality_seq&quot;)</code></pre><p>Create FASTQ record</p><pre><code class="language-julia hljs">hq_record = FASTX.FASTQ.Record(&quot;high_quality&quot;, high_quality_seq, high_quality_scores)
hq_records = [hq_record]</code></pre><p>Create qualmer graph</p><pre><code class="language-julia hljs">hq_dataset_id = &quot;high_quality&quot;
hq_qualmer_graph = Mycelia.Rhizomorph.build_qualmer_graph(hq_records, 5; dataset_id=hq_dataset_id)
hq_kmers = collect(MetaGraphsNext.labels(hq_qualmer_graph))
println(&quot;High-quality k-mers: $(length(hq_kmers)) unique k-mers&quot;)</code></pre><p>Show quality information</p><pre><code class="language-julia hljs">first_hq_kmer = first(hq_kmers)
hq_vertex_data = hq_qualmer_graph[first_hq_kmer]
hq_joint_quality = Mycelia.Rhizomorph.get_vertex_joint_quality(hq_vertex_data, hq_dataset_id)
hq_mean_quality = Mycelia.Rhizomorph.get_vertex_mean_quality(hq_vertex_data, hq_dataset_id)
hq_joint_mean = Statistics.mean(hq_joint_quality)
hq_mean_mean = Statistics.mean(hq_mean_quality)
println(&quot;$(first_hq_kmer): mean joint Q = $(round(hq_joint_mean, digits=2)), mean Q = $(round(hq_mean_mean, digits=2))&quot;)</code></pre><p>Example: Medium-quality sequencing data</p><pre><code class="language-julia hljs">medium_quality_seq = &quot;ATCGATCGATCGATCGTAGCTAGCT&quot;
medium_quality_scores = &quot;?????????????????????????&quot;  ## PHRED 30
println(&quot;\\nMedium-quality sequence: $medium_quality_seq&quot;)

mq_record = FASTX.FASTQ.Record(&quot;medium_quality&quot;, medium_quality_seq, medium_quality_scores)
mq_records = [mq_record]</code></pre><p>Create qualmer graph</p><pre><code class="language-julia hljs">mq_dataset_id = &quot;medium_quality&quot;
mq_qualmer_graph = Mycelia.Rhizomorph.build_qualmer_graph(mq_records, 5; dataset_id=mq_dataset_id)
mq_kmers = collect(MetaGraphsNext.labels(mq_qualmer_graph))
println(&quot;Medium-quality k-mers: $(length(mq_kmers)) unique k-mers&quot;)

first_mq_kmer = first(mq_kmers)
mq_vertex_data = mq_qualmer_graph[first_mq_kmer]
mq_joint_quality = Mycelia.Rhizomorph.get_vertex_joint_quality(mq_vertex_data, mq_dataset_id)
mq_mean_quality = Mycelia.Rhizomorph.get_vertex_mean_quality(mq_vertex_data, mq_dataset_id)
mq_joint_mean = Statistics.mean(mq_joint_quality)
mq_mean_mean = Statistics.mean(mq_mean_quality)
println(&quot;$(first_mq_kmer): mean joint Q = $(round(mq_joint_mean, digits=2)), mean Q = $(round(mq_mean_mean, digits=2))&quot;)</code></pre><p>Quality comparison</p><pre><code class="language-julia hljs">println(&quot;\\nQuality comparison:&quot;)
println(&quot;High-quality mean joint Q: $(round(hq_joint_mean, digits=2))&quot;)
println(&quot;Medium-quality mean joint Q: $(round(mq_joint_mean, digits=2))&quot;)
println(&quot;Quality-aware assembly can make better decisions using this information!&quot;)</code></pre><p>Use cases</p><pre><code class="language-julia hljs">println(&quot;Use cases: Error correction, quality-aware assembly, variant calling&quot;)</code></pre><p>============================================================================= Tutorial 4: String Graphs - Variable-Length Overlap Analysis =============================================================================</p><pre><code class="language-julia hljs">println(&quot;\n4. STRING GRAPHS - Variable-Length Overlap Analysis&quot;)
println(&quot;-&quot;^50)</code></pre><p>Example: Overlap-aware text fragments</p><pre><code class="language-julia hljs">string_fragments = [&quot;ABCDEF&quot;, &quot;CDEFGH&quot;, &quot;EFGHIJ&quot;, &quot;GHIJKL&quot;]
println(&quot;Input fragments: $(string_fragments)&quot;)</code></pre><p>Create variable-length string graph (OLC)</p><pre><code class="language-julia hljs">string_graph = Mycelia.Rhizomorph.build_string_graph(string_fragments; dataset_id=&quot;string_demo&quot;, min_overlap=3)
string_stats = Mycelia.Rhizomorph.get_string_graph_statistics(string_graph)
println(&quot;String graph: $(string_stats[:num_vertices]) strings, $(string_stats[:num_edges]) overlaps&quot;)</code></pre><p>Show example overlaps if present</p><pre><code class="language-julia hljs">if string_stats[:num_edges] &gt; 0
    example_edge = first(MetaGraphsNext.edge_labels(string_graph))
    src, dst = example_edge
    overlap_len = string_graph[src, dst].overlap_length
    println(&quot;Example overlap: \&quot;$src\&quot; → \&quot;$dst\&quot; (overlap length: $overlap_len)&quot;)
end</code></pre><p>Use cases</p><pre><code class="language-julia hljs">println(&quot;Use cases: Text reconstruction, fragment assembly, overlap analysis&quot;)</code></pre><p>============================================================================= Tutorial 5: FASTA Graphs - Variable-Length BioSequence Analysis =============================================================================</p><pre><code class="language-julia hljs">println(&quot;\n5. FASTA GRAPHS - Variable-Length BioSequence Analysis&quot;)
println(&quot;-&quot;^50)</code></pre><p>Example: Direct FASTA graph construction</p><pre><code class="language-julia hljs">fasta_sequence = &quot;ATCGATCGATCGATCGTAGCTAGCTAGCTGCGCGCGC&quot;
println(&quot;FASTA sequence: $fasta_sequence&quot;)

fasta_record = FASTX.FASTA.Record(&quot;sample&quot;, fasta_sequence)
fasta_records = [fasta_record]</code></pre><p>Create FASTA overlap graph directly (OLC)</p><pre><code class="language-julia hljs">bio_graph = Mycelia.Rhizomorph.build_fasta_graph(fasta_records; dataset_id=&quot;fasta_demo&quot;, min_overlap=5)
bio_sequences = collect(MetaGraphsNext.labels(bio_graph))
println(&quot;FASTA graph: $(length(bio_sequences)) sequences&quot;)

if !isempty(bio_sequences)
    first_seq = first(bio_sequences)
    println(&quot;First sequence: $first_seq (length: $(length(first_seq)))&quot;)
    println(&quot;Sequence type: $(typeof(first_seq))&quot;)
end</code></pre><p>Example: K-mer to FASTA graph conversion</p><pre><code class="language-julia hljs">println(&quot;\\nK-mer to FASTA graph conversion:&quot;)
kmer_graph = Mycelia.Rhizomorph.build_kmer_graph(fasta_records, 5; dataset_id=&quot;kmer_demo&quot;)
bio_converted = Mycelia.Rhizomorph.convert_fixed_to_variable(kmer_graph)
converted_sequences = collect(MetaGraphsNext.labels(bio_converted))
kmer_labels = collect(MetaGraphsNext.labels(kmer_graph))
println(&quot;Converted $(length(kmer_labels)) k-mers to $(length(converted_sequences)) FASTA sequences&quot;)</code></pre><p>Use cases</p><pre><code class="language-julia hljs">println(&quot;Use cases: Genome assembly, contig construction, sequence alignment&quot;)</code></pre><p>============================================================================= Tutorial 6: FASTQ Graphs - Quality-Aware BioSequence Analysis =============================================================================</p><pre><code class="language-julia hljs">println(&quot;\n6. FASTQ GRAPHS - Quality-Aware BioSequence Analysis&quot;)
println(&quot;-&quot;^50)</code></pre><p>Example: Quality-aware FASTQ graph</p><pre><code class="language-julia hljs">fastq_sequence = &quot;ATCGATCGATCGATCGTAGCTAGCTAGCTGCGCGCGC&quot;
fastq_quality = repeat(&quot;H&quot;, length(fastq_sequence))  ## High quality
println(&quot;FASTQ sequence: $fastq_sequence&quot;)
println(&quot;Quality string: $fastq_quality&quot;)

fastq_record = FASTX.FASTQ.Record(&quot;quality_sample&quot;, fastq_sequence, fastq_quality)
fastq_records = [fastq_record]</code></pre><p>Create quality-aware FASTQ graph</p><pre><code class="language-julia hljs">fastq_dataset_id = &quot;fastq_demo&quot;
quality_bio_graph = Mycelia.Rhizomorph.build_fastq_graph(fastq_records; dataset_id=fastq_dataset_id, min_overlap=5)
quality_sequences = collect(MetaGraphsNext.labels(quality_bio_graph))
println(&quot;FASTQ graph: $(length(quality_sequences)) sequences&quot;)

if !isempty(quality_sequences)
    first_quality_seq = first(quality_sequences)
    quality_vertex_data = quality_bio_graph[first_quality_seq]
    println(&quot;First sequence: $first_quality_seq&quot;)
    joint_quality = Mycelia.Rhizomorph.get_vertex_joint_quality(quality_vertex_data, fastq_dataset_id)
    println(&quot;Joint quality (first 10): $(joint_quality[1:min(10, length(joint_quality))])&quot;)
end</code></pre><p>Example: Qualmer to FASTQ graph conversion</p><pre><code class="language-julia hljs">println(&quot;\\nQualmer to FASTQ graph conversion:&quot;)
qualmer_graph = Mycelia.Rhizomorph.build_qualmer_graph(fastq_records, 5; dataset_id=fastq_dataset_id)
quality_converted = Mycelia.Rhizomorph.convert_fixed_to_variable(qualmer_graph)
quality_converted_sequences = collect(MetaGraphsNext.labels(quality_converted))
qualmer_labels = collect(MetaGraphsNext.labels(qualmer_graph))
println(&quot;Converted $(length(qualmer_labels)) qualmers to $(length(quality_converted_sequences)) FASTQ sequences&quot;)</code></pre><p>Use cases</p><pre><code class="language-julia hljs">println(&quot;Use cases: Quality-aware assembly, error correction, variant calling&quot;)</code></pre><p>============================================================================= Tutorial 7: Graph Type Hierarchy and Conversions =============================================================================</p><pre><code class="language-julia hljs">println(&quot;\n7. GRAPH TYPE HIERARCHY AND CONVERSIONS&quot;)
println(&quot;-&quot;^50)</code></pre><p>Demonstrate the hierarchy: Fixed-length → Variable-length</p><pre><code class="language-julia hljs">sequence_for_hierarchy = &quot;ATCGATCGATCGATCGTAGCTAGCTAGCTGCGCGCGC&quot;
records_for_hierarchy = [FASTX.FASTA.Record(&quot;hierarchy_demo&quot;, sequence_for_hierarchy)]
fastq_records_for_hierarchy = [FASTX.FASTQ.Record(&quot;hierarchy_demo&quot;, sequence_for_hierarchy, repeat(&quot;H&quot;, length(sequence_for_hierarchy)))]

println(&quot;Original sequence: $sequence_for_hierarchy&quot;)</code></pre><p>Fixed-length graphs</p><pre><code class="language-julia hljs">ngram_result = Mycelia.Rhizomorph.build_ngram_graph([sequence_for_hierarchy], 3; dataset_id=&quot;hierarchy&quot;)
kmer_result = Mycelia.Rhizomorph.build_kmer_graph(records_for_hierarchy, 5; dataset_id=&quot;hierarchy&quot;)
qualmer_result = Mycelia.Rhizomorph.build_qualmer_graph(fastq_records_for_hierarchy, 5; dataset_id=&quot;hierarchy&quot;)
ngram_labels = collect(MetaGraphsNext.labels(ngram_result))
kmer_labels = collect(MetaGraphsNext.labels(kmer_result))
qualmer_labels = collect(MetaGraphsNext.labels(qualmer_result))

println(&quot;\\nFixed-length graphs:&quot;)
println(&quot;  N-grams (n=3): $(length(ngram_labels)) vertices&quot;)
println(&quot;  K-mers (k=5): $(length(kmer_labels)) vertices&quot;)
println(&quot;  Qualmers (k=5): $(length(qualmer_labels)) vertices&quot;)</code></pre><p>Variable-length graphs</p><pre><code class="language-julia hljs">bio_result = Mycelia.Rhizomorph.build_fasta_graph(records_for_hierarchy; dataset_id=&quot;hierarchy&quot;, min_overlap=5)
quality_bio_result = Mycelia.Rhizomorph.build_fastq_graph(fastq_records_for_hierarchy; dataset_id=&quot;hierarchy&quot;, min_overlap=5)
bio_labels = collect(MetaGraphsNext.labels(bio_result))
quality_bio_labels = collect(MetaGraphsNext.labels(quality_bio_result))

println(&quot;\\nVariable-length graphs:&quot;)
println(&quot;  FASTA graph: $(length(bio_labels)) vertices&quot;)
println(&quot;  FASTQ graph: $(length(quality_bio_labels)) vertices&quot;)</code></pre><p>Conversions</p><pre><code class="language-julia hljs">kmer_to_bio = Mycelia.Rhizomorph.convert_fixed_to_variable(kmer_result)
qualmer_to_quality_bio = Mycelia.Rhizomorph.convert_fixed_to_variable(qualmer_result)
kmer_to_bio_labels = collect(MetaGraphsNext.labels(kmer_to_bio))
qualmer_to_quality_labels = collect(MetaGraphsNext.labels(qualmer_to_quality_bio))

println(&quot;\\nConversions:&quot;)
println(&quot;  K-mer → FASTA graph: $(length(kmer_labels)) → $(length(kmer_to_bio_labels))&quot;)
println(&quot;  Qualmer → FASTQ graph: $(length(qualmer_labels)) → $(length(qualmer_to_quality_labels))&quot;)</code></pre><p>============================================================================= Tutorial 8: Practical Assembly Workflow =============================================================================</p><pre><code class="language-julia hljs">println(&quot;\n8. PRACTICAL ASSEMBLY WORKFLOW&quot;)
println(&quot;-&quot;^50)</code></pre><p>Simulate a realistic assembly scenario</p><pre><code class="language-julia hljs">simulated_reads = [
    &quot;ATCGATCGATCGATCGTAGC&quot;,
    &quot;GATCGATCGATCGTAGCTAG&quot;,
    &quot;TCGATCGATCGTAGCTAGCT&quot;,
    &quot;GATCGATCGTAGCTAGCTGC&quot;,
    &quot;TCGATCGTAGCTAGCTGCGC&quot;
]

println(&quot;Simulated sequencing reads:&quot;)
for (i, read) in enumerate(simulated_reads)
    println(&quot;  Read $i: $read&quot;)
end</code></pre><p>Create records</p><pre><code class="language-julia hljs">fasta_reads = [FASTX.FASTA.Record(&quot;read_$i&quot;, read) for (i, read) in enumerate(simulated_reads)]
fastq_reads = [FASTX.FASTQ.Record(&quot;read_$i&quot;, read, repeat(&quot;H&quot;, length(read))) for (i, read) in enumerate(simulated_reads)]</code></pre><p>Method 1: K-mer graph assembly</p><pre><code class="language-julia hljs">println(&quot;\\nMethod 1: K-mer graph assembly&quot;)
assembly_kmer_graph = Mycelia.Rhizomorph.build_kmer_graph(fasta_reads, 5; dataset_id=&quot;assembly_kmer&quot;)
assembly_kmers = collect(MetaGraphsNext.labels(assembly_kmer_graph))
println(&quot;  K-mer graph: $(length(assembly_kmers)) unique k-mers&quot;)</code></pre><p>Method 2: Quality-aware Qualmer assembly</p><pre><code class="language-julia hljs">println(&quot;\\nMethod 2: Quality-aware Qualmer assembly&quot;)
assembly_qualmer_graph = Mycelia.Rhizomorph.build_qualmer_graph(fastq_reads, 5; dataset_id=&quot;assembly_qualmer&quot;)
assembly_qualmers = collect(MetaGraphsNext.labels(assembly_qualmer_graph))
println(&quot;  Qualmer graph: $(length(assembly_qualmers)) unique qualmers&quot;)</code></pre><p>Show quality information</p><pre><code class="language-julia hljs">if !isempty(assembly_qualmers)
    first_qualmer = first(assembly_qualmers)
    qualmer_vertex_data = assembly_qualmer_graph[first_qualmer]
    assembly_joint_quality = Mycelia.Rhizomorph.get_vertex_joint_quality(qualmer_vertex_data, &quot;assembly_qualmer&quot;)
    joint_mean = Statistics.mean(assembly_joint_quality)
    println(&quot;  Quality example: $(first_qualmer) with mean joint Q = $(round(joint_mean, digits=2))&quot;)
end</code></pre><p>Method 3: FASTA graph assembly</p><pre><code class="language-julia hljs">println(&quot;\\nMethod 3: FASTA graph assembly&quot;)
assembly_bio_graph = Mycelia.Rhizomorph.build_fasta_graph(fasta_reads; dataset_id=&quot;assembly_fasta&quot;, min_overlap=5)
assembly_sequences = collect(MetaGraphsNext.labels(assembly_bio_graph))
println(&quot;  FASTA graph: $(length(assembly_sequences)) sequences&quot;)

println(&quot;\\nWorkflow summary:&quot;)
println(&quot;  1. Choose graph type based on data: FASTA → K-mer/FASTA graph, FASTQ → Qualmer/FASTQ graph&quot;)
println(&quot;  2. Consider quality: Use Qualmer graphs for quality-aware assembly&quot;)
println(&quot;  3. Use appropriate k-mer size: Smaller k for higher sensitivity, larger k for specificity&quot;)
println(&quot;  4. Convert between graph types as needed for different analysis steps&quot;)</code></pre><p>============================================================================= Summary =============================================================================</p><pre><code class="language-julia hljs">println(&quot;\n&quot; * &quot;=&quot;^80)
println(&quot;TUTORIAL SUMMARY&quot;)
println(&quot;=&quot;^80)
println(&quot;You&#39;ve learned about all 6 graph types in Mycelia Rhizomorph:&quot;)
println()
println(&quot;FIXED-LENGTH GRAPHS (Assembly Foundation):&quot;)
println(&quot;  1. N-gram Graphs - Unicode text analysis&quot;)
println(&quot;  2. K-mer Graphs - DNA/RNA/protein sequence analysis&quot;)
println(&quot;  3. Qualmer Graphs - Quality-aware sequence analysis&quot;)
println()
println(&quot;VARIABLE-LENGTH GRAPHS (OLC Products):&quot;)
println(&quot;  4. String Graphs - Variable-length string overlaps&quot;)
println(&quot;  5. FASTA Graphs - Variable-length BioSequence overlaps&quot;)
println(&quot;  6. FASTQ Graphs - Quality-aware BioSequence overlaps&quot;)
println()
println(&quot;KEY FEATURES:&quot;)
println(&quot;  ✓ Type-stable implementation with storage parameters&quot;)
println(&quot;  ✓ Quality-aware assembly with per-base quality scores&quot;)
println(&quot;  ✓ Multi-alphabet support (DNA, RNA, amino acids)&quot;)
println(&quot;  ✓ Strand-aware graph construction&quot;)
println(&quot;  ✓ Graph type conversions and hierarchy&quot;)
println(&quot;  ✓ MetaGraphsNext.jl foundation for performance&quot;)
println()
println(&quot;Choose the right graph type for your analysis:&quot;)
println(&quot;  - Use N-gram graphs for text analysis&quot;)
println(&quot;  - Use K-mer graphs for basic sequence analysis&quot;)
println(&quot;  - Use Qualmer graphs for quality-aware assembly&quot;)
println(&quot;  - Use FASTA graphs for variable-length analysis&quot;)
println(&quot;  - Use FASTQ graphs for quality-aware variable-length analysis&quot;)
println(&quot;=&quot;^80)</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../14_binning_workflow/">« Step 9: Binning Workflow</a><a class="docs-footer-nextpage" href="../09_round_trip_01_string_graphs/">Round Trip 01 String Graphs »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Tuesday 13 January 2026 21:22">Tuesday 13 January 2026</span>. Using Julia version 1.10.10.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
