{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Interactive Test Development\n",
    "\n",
    "Make sure that we have the kernel installed\n",
    "```bash\n",
    "# Cloud\n",
    "julia -e 'using Pkg; Pkg.add(\"IJulia\"); import IJulia; IJulia.installkernel(\"julia\", env=Dict(\"JULIA_NUM_THREADS\"=>\"$(Sys.CPU_THREADS)\"))'\n",
    "\n",
    "# HPC\n",
    "export LD_LIBRARY_PATH=\"\" && julia -e 'import Pkg; Pkg.add([\"IJulia\"]); using IJulia; IJulia.installkernel(\"julia\", env=Dict(\"JULIA_NUM_THREADS\"=>\"$(Sys.CPU_THREADS)\", \"LD_LIBRARY_PATH\" => \"\"))'\n",
    "```\n",
    "\n",
    "Install pip & jupyter tools if not present\n",
    "```bash\n",
    "sudo apt-get update && sudo apt-get install python3-pip && pip install nbconvert jupyter\n",
    "```\n",
    "\n",
    "Convert notebook to Julia file and run tests\n",
    "```bash\n",
    "jupyter nbconvert --to script test/test-development.ipynb && cp test/test-development.jl test/runtests.jl && julia --project=. -e 'using Pkg; Pkg.test(\"Mycelia\")'\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Pkg\n",
    "if isinteractive()\n",
    "    Pkg.activate(\"..\")\n",
    "else\n",
    "    # this should be set via the `--project=` flag\n",
    "end\n",
    "\n",
    "using Revise\n",
    "using Test\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import FASTX\n",
    "import DocStringExtensions\n",
    "import BioSequences\n",
    "import Dates\n",
    "import Random\n",
    "import SHA\n",
    "import CSV\n",
    "import uCSV\n",
    "import DataFrames\n",
    "import Arrow\n",
    "import StatsBase\n",
    "\n",
    "const SEED = 42\n",
    "const phiX174_accession_id = \"NC_001422.1\"\n",
    "const phiX174_assembly_id = \"GCF_000819615.1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Threads.nthreads()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ENV[\"BLASTDB\"] = Mycelia.DEFAULT_BLASTDB_PATH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@testset \"blastdatabases\" begin\n",
    "    blast_db_search_paths = split(first(readlines(`$(Mycelia.CONDA_RUNNER) run --live-stream -n blast blastdbcmd -show_blastdb_search_path`)), ':')\n",
    "    @test blast_db_search_paths == [realpath(pwd()), realpath(Mycelia.DEFAULT_BLASTDB_PATH), \"\"]\n",
    "    \n",
    "    # blast_core_nt_path = Mycelia.download_blast_db(db = \"core_nt\", source=\"ncbi\")\n",
    "    # 30 minutes to generate?\n",
    "    # blastdb_core_nt_tax_info = Mycelia.get_blastdb_tax_info(blastdb=blast_core_nt_path)\n",
    "    # StatsBase.countmap(blastdb_core_nt_tax_info[!, \"taxonomic super kingdom\"])\n",
    "    # Dict{InlineStrings.String15, Int64} with 6 entries:\n",
    "    #   \"Bacteria\"  => 454778\n",
    "    #   \"-\"         => 19619\n",
    "    #   \"Eukaryota\" => 1431644\n",
    "    #   \"N/A\"       => 196\n",
    "    #   \"Archaea\"   => 10761\n",
    "    #   \"Viruses\"   => 236449\n",
    "    # https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi\n",
    "    \n",
    "    # 149Gb - too big\n",
    "    # 6573.781228 seconds (395.81 k allocations: 22.089 MiB, 0.00% gc time, 0.01% compilation time: 17% of which was recompilation)\n",
    "    # sans_eukaryota_taxids = [Mycelia.NCBI_SUPERKINGDOMS[x] for x in (\"Bacteria\", \"Archaea\", \"Viruses\", \"Other sequences (-)\")]\n",
    "    # @time generated_fasta = Mycelia.blastdb_to_fasta(blastdb=blast_core_nt_path, taxids = sans_eukaryota_taxids)\n",
    "    # sans_eukaryota_fna_path = replace(generated_fasta, Mycelia.FASTA_REGEX => \".sans_eukaryota.fna.gz\")\n",
    "    # mv(generated_fasta, sans_eukaryota_fna_path)\n",
    "    \n",
    "    # 83.4Gb\n",
    "    # eukaryota_taxids = [Mycelia.NCBI_SUPERKINGDOMS[\"Eukaryota\"]]\n",
    "    # @time generated_fasta = Mycelia.blastdb_to_fasta(blastdb=blast_core_nt_path, taxids = eukaryota_taxids)\n",
    "    # eukaryota_fna_path = replace(generated_fasta, Mycelia.FASTA_REGEX => \".eukaryota.fna.gz\")\n",
    "    # mv(generated_fasta, eukaryota_fna_path)\n",
    "    \n",
    "    # 82.7Gb\n",
    "    # ~ 3k seconds\n",
    "    # bacteria_taxids = [Mycelia.NCBI_SUPERKINGDOMS[\"Bacteria\"]]\n",
    "    # @time generated_fasta = Mycelia.blastdb_to_fasta(blastdb=blast_core_nt_path, taxids = bacteria_taxids)\n",
    "    # bacteria_fna_path = replace(generated_fasta, Mycelia.FASTA_REGEX => \".bacteria.fna.gz\")\n",
    "    # mv(generated_fasta, bacteria_fna_path)\n",
    "    \n",
    "    # 64.1Gb\n",
    "    # Archaea, viruses, and other\n",
    "    # 2787.466275 seconds (66.90 k allocations: 2.779 MiB)\n",
    "    # archaea_viruses_other_taxids = [Mycelia.NCBI_SUPERKINGDOMS[x] for x in (\"Archaea\", \"Viruses\", \"Other sequences (-)\")]\n",
    "    # @time generated_fasta = Mycelia.blastdb_to_fasta(blastdb=blast_core_nt_path, taxids = archaea_viruses_other_taxids)\n",
    "    # archaea_viruses_other_fna_path = replace(generated_fasta, Mycelia.FASTA_REGEX => \".archaea_viruses_other.fna.gz\")\n",
    "    # mv(generated_fasta, archaea_viruses_other_fna_path)\n",
    "    \n",
    "    \n",
    "    # # Processing 51 records from Blast DB /global/homes/c/cjprybol/workspace/blastdb/ref_viroids_rep_genomes: 100% Time: 0:00:04\n",
    "    # # \"/global/homes/c/cjprybol/workspace/blastdb/ref_viroids_rep_genomes.fna.tsv.gz\"\n",
    "    blast_viroids_path = Mycelia.download_blast_db(db = \"ref_viroids_rep_genomes\", source=\"ncbi\")\n",
    "    \n",
    "    blastdb_metadata = Mycelia.get_blastdb_metadata(blastdb=blast_viroids_path)\n",
    "    \n",
    "    @test sort(collect(keys(blastdb_metadata))) == [\n",
    "        \"bytes-to-cache\",\n",
    "        \"bytes-total\",\n",
    "        \"db-version\",\n",
    "        \"dbname\",\n",
    "        \"dbtype\",\n",
    "        \"description\",\n",
    "        \"files\",\n",
    "        \"last-updated\",\n",
    "        \"number-of-letters\",\n",
    "        \"number-of-sequences\",\n",
    "        \"number-of-taxids\",\n",
    "        \"number-of-volumes\",\n",
    "        \"version\",\n",
    "    ]\n",
    "    \n",
    "    blast_viroids_arrow_path = Mycelia.blastdb2table(blastdb = blast_viroids_path)\n",
    "    blast_viroids_arrow_table = DataFrames.DataFrame(Arrow.Table(blast_viroids_arrow_path))\n",
    "    @test DataFrames.nrow(blast_viroids_arrow_table) == blastdb_metadata[\"number-of-sequences\"]\n",
    "    rm(blast_viroids_arrow_path)\n",
    "    \n",
    "    ref_viroids_blast_fasta = Mycelia.blastdb_to_fasta(blastdb=blast_viroids_path, force=true)\n",
    "    @test length(collect(Mycelia.open_fastx(ref_viroids_blast_fasta))) == blastdb_metadata[\"number-of-sequences\"]\n",
    "    \n",
    "    ref_viroids_blast_fasta_taxid_filter = Mycelia.blastdb_to_fasta(blastdb=blast_viroids_path, taxids = [12892], force=true)\n",
    "    ref_viroids_blast_fasta_entry_filter = Mycelia.blastdb_to_fasta(blastdb=blast_viroids_path, entries = [\"NC_002030.1\"], force=true)\n",
    "    @test collect(Mycelia.open_fastx(ref_viroids_blast_fasta_entry_filter)) == collect(Mycelia.open_fastx(ref_viroids_blast_fasta_taxid_filter))\n",
    "    \n",
    "    @test Mycelia.get_blastdb_tax_info(blastdb=blast_viroids_path) isa DataFrames.DataFrame\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@testset \"statistical kmer analyses\" begin\n",
    "    @test Mycelia.optimal_subsequence_length(error_rate=0.001, sequence_length=100, threshold=0.99) == 10\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@testset \"tool integration\" begin\n",
    "    @testset \"padloc\" begin\n",
    "        ecoli_k12_accession = \"GCF_000005845.2\"\n",
    "        result = Mycelia.ncbi_genome_download_accession(accession=ecoli_k12_accession, include_string=\"genome\")\n",
    "        # @show result.genome\n",
    "        padloc_result = Mycelia.run_padloc(fasta_file = result.genome)\n",
    "        # @show padloc_result\n",
    "        @test isfile(padloc_result.csv)\n",
    "        rm(ecoli_k12_accession, recursive=true)\n",
    "    end\n",
    "\n",
    "    @testset \"ncbi-blast\" begin\n",
    "        blast_database_table = Mycelia.list_blastdbs()\n",
    "        @test blast_database_table isa DataFrames.DataFrame\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@testset \"Constants\" begin\n",
    "    @testset \"FASTQ regex\" begin\n",
    "        hypothethical_fastq_files = [\n",
    "            \"sample1.fastq\",\n",
    "            \"sample2.fq\",\n",
    "        ]\n",
    "        for hypothethical_fastq_file in hypothethical_fastq_files\n",
    "            @test occursin(Mycelia.FASTQ_REGEX, hypothethical_fastq_file)\n",
    "            @test occursin(Mycelia.FASTQ_REGEX, hypothethical_fastq_file * \".gz\")\n",
    "        end\n",
    "    end\n",
    "    @testset \"FASTA regex\" begin\n",
    "        hypothethical_fasta_files = [\n",
    "            \"genome1.fasta\",\n",
    "            \"fasta-sequences.fas\",\n",
    "            \"fasta.fa\",\n",
    "            \"fasta-nucleic-acid.fna\",\n",
    "            \"open-reading-frames.ffn\",\n",
    "            \"fasta-amino-acid.faa\",\n",
    "            \"multi-protein-fasta.mpfa\",\n",
    "            \"transcriptome8.frn\",\n",
    "        ]\n",
    "        for hypothethical_fasta_file in hypothethical_fasta_files\n",
    "            @test occursin(Mycelia.FASTA_REGEX, hypothethical_fasta_file)\n",
    "            @test occursin(Mycelia.FASTA_REGEX, hypothethical_fasta_file * \".gz\")\n",
    "        end\n",
    "    end\n",
    "    @testset \"VCF regex\" begin\n",
    "        hypothethical_vcf_files = [\n",
    "            \"variants1.vcf\",\n",
    "            \"variants2.vcf.gz\",\n",
    "        ]\n",
    "        for hypothethical_vcf_file in hypothethical_vcf_files\n",
    "            @test occursin(Mycelia.VCF_REGEX, hypothethical_vcf_file)\n",
    "        end\n",
    "    end\n",
    "    @testset \"XAM regex\" begin\n",
    "        hypothetical_xam_files = [\n",
    "            \"alignment1.sam\",\n",
    "            \"alignment2.sam.gz\",\n",
    "            \"alignment3.bam\",\n",
    "            \"alignment4.cram\"\n",
    "        ]\n",
    "        for hypothetical_xam_file in hypothetical_xam_files\n",
    "            @test occursin(Mycelia.XAM_REGEX, hypothetical_xam_file)\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Simulation: for multi-omics data generation for benchmarking or testing.\n",
    "@testset \"FASTA simulation and acquisition\" begin\n",
    "    @testset \"dna record\" begin\n",
    "        dna_record = Mycelia.random_fasta_record(moltype=:DNA, seed=SEED, L = 10)\n",
    "        # @test FASTX.identifier(dna_record) == \"e6614e938383857305089158322604fb9c07d0aac05567844d01f1517ce82ddc\"\n",
    "        @test FASTX.sequence(dna_record) == \"CCGCCGCTCA\"\n",
    "    end\n",
    "\n",
    "    @testset \"rna record\" begin\n",
    "        rna_record = Mycelia.random_fasta_record(moltype=:RNA, seed=SEED, L = 10)\n",
    "        # @test FASTX.identifier(rna_record) == \"a4ebfc1dd4051dd9db816c42c5fb8f528ae0b5387d7a8ccad67995e2c31ee48d\"\n",
    "        @test FASTX.sequence(rna_record) == \"CCGCCGCUCA\"\n",
    "    end\n",
    "\n",
    "    @testset \"aa record\" begin\n",
    "        aa_record = Mycelia.random_fasta_record(moltype=:AA, seed=SEED, L = 10)\n",
    "        # @test FASTX.identifier(aa_record) == \"cdb40a05b841c44e3757de6a54a1fd011536fce187c7e8c9c1ed36c415db8344\"\n",
    "        @test FASTX.sequence(aa_record) == \"VATAGWWITI\"\n",
    "    end\n",
    "    \n",
    "    @testset \"virus phiX174\" begin\n",
    "        genome_result = Mycelia.download_genome_by_accession(accession=phiX174_accession_id)\n",
    "        @test basename(genome_result) == phiX174_accession_id * \".fna.gz\"\n",
    "        @test Mycelia.get_base_extension(genome_result) == \".fna.gz\"\n",
    "        rm(genome_result)\n",
    "        # @test Mycelia.sha256_file(genome_result) == \"765354d42319e4a350c93b09260bf911864263653f828ad97b5c35eed950591a\"\n",
    "        \n",
    "        phiX174_assembly_dataset = Mycelia.ncbi_genome_download_accession(accession=phiX174_assembly_id, include_string=\"gff3,rna,cds,protein,genome,seq-report\")\n",
    "        @test basename(phiX174_assembly_dataset.genome) == phiX174_assembly_id * \"_ViralProj14015_genomic.fna\"\n",
    "        #TODO add tests for remaining items? do that elsewhere?\n",
    "        @test Mycelia.get_base_extension(phiX174_assembly_dataset.genome) == \".fna\"\n",
    "        @test Mycelia.get_base_extension(phiX174_assembly_dataset.protein) == \".faa\"\n",
    "        rm(phiX174_assembly_id, recursive=true)\n",
    "    end\n",
    "\n",
    "    @testset \"bacteria-like\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"protist-like\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"fungi-like\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"plant-like\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"animal-like\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"microbiome\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@testset \"FASTQ simulation\" begin\n",
    "    @testset \"Illumina\" begin\n",
    "        if isdir(phiX174_assembly_id)\n",
    "            rm(phiX174_assembly_id, recursive=true)\n",
    "        end\n",
    "        phiX174_assembly_dataset = Mycelia.ncbi_genome_download_accession(accession=phiX174_assembly_id, include_string=\"genome\")\n",
    "        read_simulation_result = Mycelia.simulate_illumina_paired_reads(in_fasta = phiX174_assembly_dataset.genome, coverage=10)\n",
    "        @test isfile(read_simulation_result.forward_reads)\n",
    "        @test isfile(read_simulation_result.reverse_reads)\n",
    "        @test isfile(read_simulation_result.sam)\n",
    "        @test isfile(read_simulation_result.error_free_sam)\n",
    "        rm(phiX174_assembly_id, recursive=true)\n",
    "    end\n",
    "\n",
    "    @testset \"Ultima\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"Nanopore\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"PacBio\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"multi-entity, even coverage\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"multi-entity, log-distributed coverage\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@testset \"sequence IO\" begin\n",
    "    @testset \"detect alphabet type\" begin\n",
    "        @test Mycelia.detect_alphabet(FASTX.sequence(Mycelia.random_fasta_record(L=100, moltype=:DNA))) == :DNA\n",
    "        @test Mycelia.detect_alphabet(FASTX.sequence(Mycelia.random_fasta_record(L=100, moltype=:RNA))) == :RNA\n",
    "        @test Mycelia.detect_alphabet(FASTX.sequence(Mycelia.random_fasta_record(L=100, moltype=:AA))) == :AA\n",
    "    end\n",
    "\n",
    "    @testset \"autoconvert sequences\" begin\n",
    "        @test typeof(Mycelia.convert_sequence(FASTX.sequence(Mycelia.random_fasta_record(L=100, moltype=:DNA)))) <: BioSequences.LongDNA{4}\n",
    "        @test typeof(Mycelia.convert_sequence(FASTX.sequence(Mycelia.random_fasta_record(L=100, moltype=:RNA)))) <: BioSequences.LongRNA{4}\n",
    "        @test typeof(Mycelia.convert_sequence(FASTX.sequence(Mycelia.random_fasta_record(L=100, moltype=:AA)))) <: BioSequences.LongAA\n",
    "    end\n",
    "\n",
    "    @testset \"detect sequence extension\" begin\n",
    "        @test Mycelia.detect_sequence_extension(FASTX.sequence(Mycelia.random_fasta_record(L=100, moltype=:DNA))) == \".fna\"\n",
    "        @test Mycelia.detect_sequence_extension(FASTX.sequence(Mycelia.random_fasta_record(L=100, moltype=:RNA))) == \".frn\"\n",
    "        @test Mycelia.detect_sequence_extension(FASTX.sequence(Mycelia.random_fasta_record(L=100, moltype=:AA))) == \".faa\"\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@testset \"SRA downloading\" begin\n",
    "    outdir = mkpath(\"test-sra-tools\")\n",
    "    @testset \"download SRA short read data\" begin\n",
    "        # https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR31271002&display=metadata\n",
    "        srr_identifier = \"SRR31271002\"\n",
    "        prefetch_results = Mycelia.prefetch(SRR=srr_identifier, outdir=outdir)\n",
    "        @test prefetch_results.directory == \"test-sra-tools/$(srr_identifier)\"\n",
    "        @test prefetch_results.archive == \"test-sra-tools/$(srr_identifier)/$(srr_identifier).sra\"\n",
    "        @test filesize(prefetch_results.archive) == 18295489\n",
    "        \n",
    "        fasterq_dump_result = Mycelia.fasterq_dump(outdir=outdir, srr_identifier=srr_identifier)\n",
    "        @test fasterq_dump_result.forward_reads == \"$(outdir)/$(srr_identifier)/$(srr_identifier)_1.fastq.gz\"\n",
    "        @test fasterq_dump_result.reverse_reads == \"$(outdir)/$(srr_identifier)/$(srr_identifier)_2.fastq.gz\"\n",
    "        @test ismissing(fasterq_dump_result.unpaired_reads)\n",
    "        @test filesize(fasterq_dump_result.forward_reads) == 11637534\n",
    "        @test filesize(fasterq_dump_result.reverse_reads) == 11908008\n",
    "    end\n",
    "    \n",
    "    @testset \"download SRA long read data\" begin\n",
    "        # https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR31812976&display=metadata\n",
    "        srr_identifier = \"SRR31812976\"\n",
    "        prefetch_results = Mycelia.prefetch(SRR=srr_identifier, outdir=outdir)\n",
    "        @test prefetch_results.directory == \"$(outdir)/$(srr_identifier)\"\n",
    "        @test prefetch_results.archive == \"$(outdir)/$(srr_identifier)/$(srr_identifier).sra\"\n",
    "        @test filesize(prefetch_results.archive) == 33392268\n",
    "\n",
    "        fasterq_dump_result = Mycelia.fasterq_dump(outdir=outdir, srr_identifier=srr_identifier)\n",
    "        @test ismissing(fasterq_dump_result.forward_reads)\n",
    "        @test ismissing(fasterq_dump_result.reverse_reads)\n",
    "        @test fasterq_dump_result.unpaired_reads == \"$(outdir)/$(srr_identifier)/$(srr_identifier).fastq.gz\"\n",
    "        @test filesize(fasterq_dump_result.unpaired_reads) == 34353450\n",
    "    end\n",
    "    rm(outdir, recursive=true)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PRE-PROCESSING & READ QC Tests\n",
    "@testset \"Preprocessing\" begin\n",
    "    @testset \"FASTX stats\" begin\n",
    "        # https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR31812976&display=metadata\n",
    "        srr_identifier = \"SRR31812976\"\n",
    "        outdir = mkpath(\"fastx-stats-test\")\n",
    "        fasterq_dump_result = Mycelia.fasterq_dump(outdir=outdir, srr_identifier=srr_identifier)\n",
    "        @test fasterq_dump_result.unpaired_reads == \"$(outdir)/$(srr_identifier)/$(srr_identifier).fastq.gz\"\n",
    "        table = Mycelia.fastx_stats(fasterq_dump_result.unpaired_reads)\n",
    "        io = IOBuffer()\n",
    "        CSV.write(io, table)\n",
    "        @test String(take!(io)) == \n",
    "        \"\"\"\n",
    "        file,format,type,num_seqs,sum_len,min_len,avg_len,max_len,Q1,Q2,Q3,sum_gap,N50,N50_num,Q20(%),Q30(%),AvgQual,GC(%),sum_n,N90\n",
    "        fastx-stats-test/SRR31812976/SRR31812976.fastq.gz,FASTQ,DNA,58982,34921275,80,592.1,2315,341.0,544.0,811.0,0,743,608,50.26,9.75,11.4,39.62,0,329\n",
    "        \"\"\"\n",
    "        rm(outdir, recursive=true)\n",
    "    end\n",
    "\n",
    "    @testset \"fastx2normalized_table\" begin\n",
    "        genome_result = Mycelia.download_genome_by_accession(accession=phiX174_accession_id)\n",
    "        fastx_table = Mycelia.fastx2normalized_table(genome_result)\n",
    "        io = IOBuffer()\n",
    "        CSV.write(io, fastx_table, delim='\\t')\n",
    "        @test String(take!(io)) == \n",
    "        \"\"\"\n",
    "        fastx_path\tfastx_sha256\trecord_identifier\trecord_description\trecord_sha256\trecord_quality\trecord_alphabet\trecord_type\tmean_record_quality\tmedian_record_quality\trecord_length\trecord_sequence\n",
    "        NC_001422.1.fna.gz\tae3c70dc4d180aab0aa8d6ab81c0048998575c05e402639ffe6b99e82815f953\tNC_001422.1\tNC_001422.1 Escherichia phage phiX174, complete genome\t97038c7e1edea2297667d7f0426ba942b322c74cb30e072ec66ba47f9c0448d0\t\tACGT\tDNA\t\t\t5386\tGAGTTTTATCGCTTCCATGACGCAGAAGTTAACACTTTCGGATATTTCTGATGAGTCGAAAAATTATCTTGATAAAGCAGGAATTACTACTGCTTGTTTACGAATTAAATCGAAGTGGACTGCTGGCGGAAAATGAGAAAATTCGACCTATCCTTGCGCAGCTCGAGAAGCTCTTACTTTGCGACCTTTCGCCATCAACTAACGATTCTGTCAAAAACTGACGCGTTGGATGAGGAGAAGTGGCTTAATATGCTTGGCACGTTCGTCAAGGACTGGTTTAGATATGAGTCACATTTTGTTCATGGTAGAGATTCTCTTGTTGACATTTTAAAAGAGCGTGGATTACTATCTGAGTCCGATGCTGTTCAACCACTAATAGGTAAGAAATCATGAGTCAAGTTACTGAACAATCCGTACGTTTCCAGACCGCTTTGGCCTCTATTAAGCTCATTCAGGCTTCTGCCGTTTTGGATTTAACCGAAGATGATTTCGATTTTCTGACGAGTAACAAAGTTTGGATTGCTACTGACCGCTCTCGTGCTCGTCGCTGCGTTGAGGCTTGCGTTTATGGTACGCTGGACTTTGTGGGATACCCTCGCTTTCCTGCTCCTGTTGAGTTTATTGCTGCCGTCATTGCTTATTATGTTCATCCCGTCAACATTCAAACGGCCTGTCTCATCATGGAAGGCGCTGAATTTACGGAAAACATTATTAATGGCGTCGAGCGTCCGGTTAAAGCCGCTGAATTGTTCGCGTTTACCTTGCGTGTACGCGCAGGAAACACTGACGTTCTTACTGACGCAGAAGAAAACGTGCGTCAAAAATTACGTGCGGAAGGAGTGATGTAATGTCTAAAGGTAAAAAACGTTCTGGCGCTCGCCCTGGTCGTCCGCAGCCGTTGCGAGGTACTAAAGGCAAGCGTAAAGGCGCTCGTCTTTGGTATGTAGGTGGTCAACAATTTTAATTGCAGGGGCTTCGGCCCCTTACTTGAGGATAAATTATGTCTAATATTCAAACTGGCGCCGAGCGTATGCCGCATGACCTTTCCCATCTTGGCTTCCTTGCTGGTCAGATTGGTCGTCTTATTACCATTTCAACTACTCCGGTTATCGCTGGCGACTCCTTCGAGATGGACGCCGTTGGCGCTCTCCGTCTTTCTCCATTGCGTCGTGGCCTTGCTATTGACTCTACTGTAGACATTTTTACTTTTTATGTCCCTCATCGTCACGTTTATGGTGAACAGTGGATTAAGTTCATGAAGGATGGTGTTAATGCCACTCCTCTCCCGACTGTTAACACTACTGGTTATATTGACCATGCCGCTTTTCTTGGCACGATTAACCCTGATACCAATAAAATCCCTAAGCATTTGTTTCAGGGTTATTTGAATATCTATAACAACTATTTTAAAGCGCCGTGGATGCCTGACCGTACCGAGGCTAACCCTAATGAGCTTAATCAAGATGATGCTCGTTATGGTTTCCGTTGCTGCCATCTCAAAAACATTTGGACTGCTCCGCTTCCTCCTGAGACTGAGCTTTCTCGCCAAATGACGACTTCTACCACATCTATTGACATTATGGGTCTGCAAGCTGCTTATGCTAATTTGCATACTGACCAAGAACGTGATTACTTCATGCAGCGTTACCATGATGTTATTTCTTCATTTGGAGGTAAAACCTCTTATGACGCTGACAACCGTCCTTTACTTGTCATGCGCTCTAATCTCTGGGCATCTGGCTATGATGTTGATGGAACTGACCAAACGTCGTTAGGCCAGTTTTCTGGTCGTGTTCAACAGACCTATAAACATTCTGTGCCGCGTTTCTTTGTTCCTGAGCATGGCACTATGTTTACTCTTGCGCTTGTTCGTTTTCCGCCTACTGCGACTAAAGAGATTCAGTACCTTAACGCTAAAGGTGCTTTGACTTATACCGATATTGCTGGCGACCCTGTTTTGTATGGCAACTTGCCGCCGCGTGAAATTTCTATGAAGGATGTTTTCCGTTCTGGTGATTCGTCTAAGAAGTTTAAGATTGCTGAGGGTCAGTGGTATCGTTATGCGCCTTCGTATGTTTCTCCTGCTTATCACCTTCTTGAAGGCTTCCCATTCATTCAGGAACCGCCTTCTGGTGATTTGCAAGAACGCGTACTTATTCGCCACCATGATTATGACCAGTGTTTCCAGTCCGTTCAGTTGTTGCAGTGGAATAGTCAGGTTAAATTTAATGTGACCGTTTATCGCAATCTGCCGACCACTCGCGATTCAATCATGACTTCGTGATAAAAGATTGAGTGTGAGGTTATAACGCCGAAGCGGTAAAAATTTTAATTTTTGCCGCTGAGGGGTTGACCAAGCGAAGCGCGGTAGGTTTTCTGCTTAGGAGTTTAATCATGTTTCAGACTTTTATTTCTCGCCATAATTCAAACTTTTTTTCTGATAAGCTGGTTCTCACTTCTGTTACTCCAGCTTCTTCGGCACCTGTTTTACAGACACCTAAAGCTACATCGTCAACGTTATATTTTGATAGTTTGACGGTTAATGCTGGTAATGGTGGTTTTCTTCATTGCATTCAGATGGATACATCTGTCAACGCCGCTAATCAGGTTGTTTCTGTTGGTGCTGATATTGCTTTTGATGCCGACCCTAAATTTTTTGCCTGTTTGGTTCGCTTTGAGTCTTCTTCGGTTCCGACTACCCTCCCGACTGCCTATGATGTTTATCCTTTGAATGGTCGCCATGATGGTGGTTATTATACCGTCAAGGACTGTGTGACTATTGACGTCCTTCCCCGTACGCCGGGCAATAACGTTTATGTTGGTTTCATGGTTTGGTCTAACTTTACCGCTACTAAATGCCGCGGATTGGTTTCGCTGAATCAGGTTATTAAAGAGATTATTTGTCTCCAGCCACTTAAGTGAGGTGATTTATGTTTGGTGCTATTGCTGGCGGTATTGCTTCTGCTCTTGCTGGTGGCGCCATGTCTAAATTGTTTGGAGGCGGTCAAAAAGCCGCCTCCGGTGGCATTCAAGGTGATGTGCTTGCTACCGATAACAATACTGTAGGCATGGGTGATGCTGGTATTAAATCTGCCATTCAAGGCTCTAATGTTCCTAACCCTGATGAGGCCGCCCCTAGTTTTGTTTCTGGTGCTATGGCTAAAGCTGGTAAAGGACTTCTTGAAGGTACGTTGCAGGCTGGCACTTCTGCCGTTTCTGATAAGTTGCTTGATTTGGTTGGACTTGGTGGCAAGTCTGCCGCTGATAAAGGAAAGGATACTCGTGATTATCTTGCTGCTGCATTTCCTGAGCTTAATGCTTGGGAGCGTGCTGGTGCTGATGCTTCCTCTGCTGGTATGGTTGACGCCGGATTTGAGAATCAAAAAGAGCTTACTAAAATGCAACTGGACAATCAGAAAGAGATTGCCGAGATGCAAAATGAGACTCAAAAAGAGATTGCTGGCATTCAGTCGGCGACTTCACGCCAGAATACGAAAGACCAGGTATATGCACAAAATGAGATGCTTGCTTATCAACAGAAGGAGTCTACTGCTCGCGTTGCGTCTATTATGGAAAACACCAATCTTTCCAAGCAACAGCAGGTTTCCGAGATTATGCGCCAAATGCTTACTCAAGCTCAAACGGCTGGTCAGTATTTTACCAATGACCAAATCAAAGAAATGACTCGCAAGGTTAGTGCTGAGGTTGACTTAGTTCATCAGCAAACGCAGAATCAGCGGTATGGCTCTTCTCATATTGGCGCTACTGCAAAGGATATTTCTAATGTCGTCACTGATGCTGCTTCTGGTGTGGTTGATATTTTTCATGGTATTGATAAAGCTGTTGCCGATACTTGGAACAATTTCTGGAAAGACGGTAAAGCTGATGGTATTGGCTCTAATTTGTCTAGGAAATAACCGTCAGGATTGACACCCTCCCAATTGTATGTTTTCATGCCTCCAAATCTTGGAGGCTTTTTTATGGTTCGTTCTTATTACCCTTCTGAATGTCACGCTGATTATTTTGACTTTGAGCGTATCGAGGCTCTTAAACCTGCTATTGAGGCTTGTGGCATTTCTACTCTTTCTCAATCCCCAATGCTTGGCTTCCATAAGCAGATGGATAACCGCATCAAGCTCTTGGAAGAGATTCTGTCTTTTCGTATGCAGGGCGTTGAGTTCGATAATGGTGATATGTATGTTGACGGCCATAAGGCTGCTTCTGACGTTCGTGATGAGTTTGTATCTGTTACTGAGAAGTTAATGGATGAATTGGCACAATGCTACAATGTGCTCCCCCAACTTGATATTAATAACACTATAGACCACCGCCCCGAAGGGGACGAAAAATGGTTTTTAGAGAACGAGAAGACGGTTACGCAGTTTTGCCGCAAGCTGGCTGCTGAACGCCCTCTTAAGGATATTCGCGATGAGTATAATTACCCCAAAAAGAAAGGTATTAAGGATGAGTGTTCAAGATTGCTGGAGGCCTCCACTATGAAATCGCGTAGAGGCTTTGCTATTCAGCGTTTGATGAATGCAATGCGACAGGCTCATGCTGATGGTTGGTTTATCGTTTTTGACACTCTCACGTTGGCTGACGACCGATTAGAGGCGTTTTATGATAATCCCAATGCTTTGCGTGACTATTTTCGTGATATTGGTCGTATGGTTCTTGCTGCCGAGGGTCGCAAGGCTAATGATTCACACGCCGACTGCTATCAGTATTTTTGTGTGCCTGAGTATGGTACAGCTAATGGCCGTCTTCATTTCCATGCGGTGCACTTTATGCGGACACTTCCTACAGGTAGCGTTGACCCTAATTTTGGTCGTCGGGTACGCAATCGCCGCCAGTTAAATAGCTTGCAAAATACGTGGCCTTATGGTTACAGTATGCCCATCGCAGTTCGCTACACGCAGGACGCTTTTTCACGTTCTGGTTGGTTGTGGCCTGTTGATGCTAAAGGTGAGCCGCTTAAAGCTACCAGTTATATGGCTGTTGGTTTCTATGTGGCTAAATACGTTAACAAAAAGTCAGATATGGACCTTGCTGCTAAAGGTCTAGGAGCTAAAGAATGGAACAACTCACTAAAAACCAAGCTGTCGCTACTTCCCAAGAAGCTGTTCAGAATCAGAATGAGCCGCAACTTCGGGATGAAAATGCTCACAATGACAAATCTGTCCACGGAGTGCTTAATCCAACTTACCAAGCTGGGTTACGACGCGACGCCGTTCAACCAGATATTGAAGCAGAACGCAAAAAGAGAGATGAGATTGAGGCTGGGAAAAGTTACTGTAGCCGACGTTTTGGCGGCGCAACCTGTGACGACAAATCTGCTCAAATTTATGCGCGCTTCGATAAAAATGATTGGCGTATCCAACCTGCA\n",
    "        \"\"\"\n",
    "        rm(genome_result)\n",
    "    end\n",
    "    \n",
    "    @testset \"Read Quality Control\" begin\n",
    "        @test true # https://github.com/OpenGene/fastp\n",
    "        @test true # https://github.com/FelixKrueger/TrimGalore\n",
    "        @test true # https://github.com/rrwick/Filtlong\n",
    "        @test true # https://github.com/OpenGene/fastplong\n",
    "        @test true # https://github.com/wdecoster/chopper\n",
    "        # Example: test that adapter trimming and quality filtering work.\n",
    "        # result = MyceliaAssembly.preprocess_reads(\"test_data/reads.fastq\")\n",
    "        # @test length(result.filtered_reads) > 0\n",
    "        # @test result.mean_quality ≥ 30\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "    @testset \"Read Statistics\" begin\n",
    "        # Example: test that estimated community composition is within expected bounds.\n",
    "        # comp = MyceliaAssembly.analyze_community(\"test_data/reads.fastq\")\n",
    "        # @test 0.8 <= comp[\"expected_coverage\"] <= 1.2\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data ingestion & normalization: for multi-omics data loading, sanity checks, QC.\n",
    "# Data Ingestion & Normalization\n",
    "# - Loading multi-omics data\n",
    "# - Sanity checks and quality control (QC)\n",
    "# - Subsampling reads (e.g., subsample_reads_seqtk)\n",
    "# @testset \"Data ingestion & normalization\" begin\n",
    "#     @testset \"fasta2normalized_table\" begin\n",
    "#         temp_dir = mktempdir()\n",
    "#         for alphabet in [:DNA, :AA, :RNA]\n",
    "#             fasta_record = Mycelia.random_fasta_record(moltype=alphabet, seed=SEED, L = 100)\n",
    "#             extension = Mycelia.detect_sequence_extension(fasta_record)\n",
    "#             fasta_file = joinpath(temp_dir, \"test\" * extension)\n",
    "#             @test Mycelia.get_base_extension(fasta_file) == extension\n",
    "#             Mycelia.write_fasta(outfile=fasta_file, records=[fasta_record])\n",
    "#             result_file = Mycelia.fasta2normalized_table(fasta_file, normalize_name=true)\n",
    "#             if alphabet == :DNA\n",
    "#                 @test basename(result_file) == \"b58f92d3ad303b329300debba8a8b4e9ebe6fb68fabc1ff7b2aefc1e267b64fb\" * extension * \".tsv.gz\"\n",
    "#             elseif alphabet == :RNA\n",
    "#                 @test basename(result_file) == \"dad9c65109b858e38bdc34463807729136ecc102bfbd0247dc00bd4811098dab\" * extension * \".tsv.gz\"\n",
    "#             elseif alphabet == :AA\n",
    "#                 @test basename(result_file) == \"5abeeb482b1af9097f71b81d53c72f98c0a69716b6fbfda1a8a855c33b3b0888\" * extension * \".tsv.gz\"\n",
    "#             end\n",
    "#         end\n",
    "#         rm(temp_dir, recursive=true)\n",
    "#     end\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assembly modules\n",
    "@testset \"assembly modules\" begin\n",
    "    @testset \"1. Pre‐processing & Read QC\" begin\n",
    "    end\n",
    "    @testset \"2. k‑mer Analysis\" begin\n",
    "    end\n",
    "    @testset \"3. Hybrid Assembly (Mutual Support Strategy)\" begin\n",
    "    end\n",
    "    @testset \"4. Assembly merging\" begin\n",
    "    end\n",
    "    @testset \"5. Polishing & Error Correction\" begin\n",
    "    end\n",
    "    @testset \"6. Strain resolution\" begin\n",
    "    end\n",
    "    @testset \"7. Validation & Quality Control\" begin\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PANGENOME/PANPROTEOME AND K-MER ANALYSIS Tests\n",
    "@testset \"Reference Graph and K-mer Analysis\" begin\n",
    "    @testset \"Pangenome Construction\" begin\n",
    "        # Example: verify that a reference pangenome graph is built\n",
    "        # pg_graph = MyceliaAssembly.build_pangenome([\"ref1.fasta\", \"ref2.fasta\"])\n",
    "        # @test typeof(pg_graph) <: AbstractGraph\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "    @testset \"Optimal K-mer Selection\" begin\n",
    "        # Example: determine the best k-mer length from given reads\n",
    "        # best_k = MyceliaAssembly.select_optimal_k(\"test_data/reads.fastq\")\n",
    "        # @test best_k isa Int\n",
    "        # @test 21 ≤ best_k ≤ 127\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HYBRID ASSEMBLY Tests\n",
    "@testset \"Hybrid Assembly\" begin\n",
    "    @testset \"Assembly Core\" begin\n",
    "        # Example: run hybrid assembly on a small simulated dataset.\n",
    "        # assembly = MyceliaAssembly.hybrid_assemble(\"test_data/reads.fastq\"; long_read=\"test_data/long.fastq\")\n",
    "        # @test length(assembly.contigs) > 0\n",
    "        # @test assembly.N50 > 5000\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "    @testset \"Contig Overlap Graph Integrity\" begin\n",
    "        # Example: check that the overlap graph correctly represents strain variants.\n",
    "        # graph = MyceliaAssembly.get_overlap_graph(assembly)\n",
    "        # @test MyceliaAssembly.validate_overlap_graph(graph) == true\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ASSEMBLY MERGE Tests\n",
    "@testset \"Assembly Merging\" begin\n",
    "    @testset \"Contig Merging\" begin\n",
    "        # Example: test that QuickMerge improves contiguity compared to input assemblies.\n",
    "        # merged = MyceliaAssembly.quick_merge(\"assembly1.fasta\", \"assembly2.fasta\")\n",
    "        # @test merged.N50 > max(assembly1.N50, assembly2.N50)\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# POLISHING Tests\n",
    "@testset \"Assembly Polishing\" begin\n",
    "    @testset \"Error Correction\" begin\n",
    "        # Example: simulate a polishing step and verify base accuracy improvement.\n",
    "        # polished = MyceliaAssembly.polish_assembly(\"raw_assembly.fasta\", reads=\"test_data/reads.fastq\")\n",
    "        # accuracy_before = MyceliaAssembly.evaluate_accuracy(\"raw_assembly.fasta\", \"reference.fasta\")\n",
    "        # accuracy_after = MyceliaAssembly.evaluate_accuracy(polished, \"reference.fasta\")\n",
    "        # @test accuracy_after > accuracy_before\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# STRAIN RESOLUTION Tests\n",
    "@testset \"Strain Resolution\" begin\n",
    "    @testset \"Strain-aware Reassembly\" begin\n",
    "        # Example: test that strain-specific contigs or haplotigs are generated.\n",
    "        # strains = MyceliaAssembly.resolve_strains(\"polished_assembly.fasta\", reads=\"test_data/long.fastq\")\n",
    "        # @test length(strains) >= 2  # expect at least two strains in a mixed sample\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VALIDATION & QUALITY ASSESSMENT Tests\n",
    "@testset \"Assembly Validation\" begin\n",
    "    @testset \"Reference-Free Validation\" begin\n",
    "        # merqury\n",
    "        # ALE\n",
    "        # CGAL\n",
    "        # read mapping stats\n",
    "        # Example: run QUAST analysis and ensure basic quality metric\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "    @testset \"Reference-Based Validation\" begin\n",
    "        # Example: run MetaQUAST analysis and ensure basic quality metrics.\n",
    "        # stats = MyceliaAssembly.validate_with_metaquast(\"final_assembly.fasta\", reference=\"ref_genome.fasta\")\n",
    "        # @test stats.NGA50 > 10000\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "    @testset \"Marker Gene Completeness\" begin\n",
    "        # Example: check completeness with CheckM.\n",
    "        # quality = MyceliaAssembly.check_assembly_quality(\"final_assembly.fasta\")\n",
    "        # @test quality.completeness >= 90\n",
    "        # @test quality.contamination <= 5\n",
    "        @test true  # placeholder\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assembly & consensus generation: for combining reads under diverse coverage patterns.\n",
    "# Assembly & Consensus Generation\n",
    "# - Combining reads under diverse coverage patterns\n",
    "# - Generating consensus sequences\n",
    "# - Polishing assemblies (e.g., polish_fastq)\n",
    "@testset \"probabilistic ensemble assembly\" begin\n",
    "    @testset \"Illumina\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"Ultima\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"Nanopore\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"PacBio\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"multi-entity, even coverage\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"multi-entity, log-distributed coverage\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "\n",
    "    @testset \"multi-platform\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Multi-omics alignment & mapping: for consistent coordinate systems or annotation references.\n",
    "# Multi-omics Alignment & Mapping\n",
    "# - Consistent coordinate systems or annotation references\n",
    "# - Mapping reads to reference genomes (e.g., minimap2)\n",
    "# - Handling different sequencing technologies (e.g., PacBio, ONT)\n",
    "@testset \"Multi-omics alignment & mapping\" begin\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Graph-based integration: for building and merging genomic/transcriptomic/proteomic graphs.\n",
    "# Graph-based Integration\n",
    "# - Building and merging genomic/transcriptomic/proteomic graphs\n",
    "# - Graph traversal and pathfinding (e.g., take_a_walk)\n",
    "# - Graph-based clustering and visualization (e.g., draw_radial_tree)\n",
    "@testset \"pangenome construction\" begin\n",
    "    @testset \"?\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end\n",
    "\n",
    "@testset \"pantranscriptome construction\" begin\n",
    "    @testset \"?\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end\n",
    "\n",
    "@testset \"panproteome construction\" begin\n",
    "    @testset \"?\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotation & feature extraction: for identifying and comparing variants, functional domains, or transcripts.\n",
    "# Annotation & Feature Extraction\n",
    "# - Identifying and comparing variants, functional domains, or transcripts\n",
    "# - Annotating sequences (e.g., annotate_fasta)\n",
    "# - Parsing and handling GFF files (e.g., parse_xam)\n",
    "@testset \"Annotation & feature extraction\" begin\n",
    "    @testset \"?\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comparative analyses: for visualizing similarities/differences across strains or conditions.\n",
    "# Comparative Analyses\n",
    "# - Visualizing similarities/differences across strains or conditions\n",
    "# - Comparative genomics and transcriptomics\n",
    "# - Clustering and dimensionality reduction (e.g., heirarchically_cluster_distance_matrix)\n",
    "@testset \"Comparative Analyses\" begin\n",
    "    @testset \"?\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sequence Classification\n",
    "# - Classifying sequences based on k-mer content\n",
    "# - Using machine learning models for classification\n",
    "# - Handling large-scale sequence data efficiently\n",
    "@testset \"sequence classification\" begin\n",
    "    @testset \"?\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end\n",
    "\n",
    "@testset \"phylogenetic analyses\" begin\n",
    "    @testset \"?\" begin\n",
    "        @test 1 + 1 == 2\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Advanced metadata handling: for storing and retrieving detailed lineage or ontology data.\n",
    "# Advanced Metadata Handling\n",
    "# - Storing and retrieving detailed lineage or ontology data\n",
    "# - Handling taxonomic data (e.g., list_full_taxonomy)\n",
    "# - Integrating external databases and resources (e.g., NCBI, UniProt)\n",
    "\n",
    "# Multi-omics Integration\n",
    "# - Integrating data from genomics, transcriptomics, proteomics, and metabolomics\n",
    "# - Cross-omics correlation and analysis\n",
    "# - Building comprehensive multi-omics models\n",
    "\n",
    "# Visualization & Reporting\n",
    "# - Generating visual reports for analysis results\n",
    "# - Interactive visualization of multi-omics data\n",
    "# - Exporting results in various formats (e.g., PNG, SVG)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.8",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
