{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"/Users/cameronprybol/2021-05-16-phage-k-pangenome\""
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATE = \"2021-05-16\"\n",
    "TASK = \"phage-k-pangenome\"\n",
    "DIR = mkpath(\"$(homedir())/$(DATE)-$(TASK)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import LightGraphs\n",
    "import MetaGraphs\n",
    "import BioSequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import uCSV\n",
    "import DataFrames\n",
    "import FASTX\n",
    "import HTTP\n",
    "import CodecZlib\n",
    "import DataStructures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Pkg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "add_evidence! (generic function with 2 methods)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function add_evidence!(kmer_graph, index::Int, evidence)\n",
    "    if MetaGraphs.has_prop(kmer_graph, index, :evidence)\n",
    "        push!(kmer_graph.vprops[index][:evidence], evidence)\n",
    "    else\n",
    "        MetaGraphs.set_prop!(kmer_graph, index, :evidence, Set([evidence]))\n",
    "    end\n",
    "end\n",
    "\n",
    "function add_evidence!(kmer_graph, edge::LightGraphs.SimpleGraphs.AbstractSimpleEdge, evidence)\n",
    "    if MetaGraphs.has_prop(kmer_graph, edge, :evidence)\n",
    "        push!(kmer_graph.eprops[edge][:evidence], evidence)\n",
    "    else\n",
    "        MetaGraphs.set_prop!(kmer_graph, edge, :evidence, Set([evidence]))\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count_kmers (generic function with 3 methods)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function count_kmers(::Type{KMER_TYPE}, sequence::BioSequences.LongSequence) where KMER_TYPE\n",
    "    kmer_counts = DataStructures.OrderedDict{KMER_TYPE, Int}()\n",
    "    kmer_iterator = (kmer.fw for kmer in BioSequences.each(KMER_TYPE, sequence))\n",
    "    for kmer in kmer_iterator\n",
    "        kmer_counts[kmer] = get(kmer_counts, kmer, 0) + 1\n",
    "    end\n",
    "    return kmer_counts\n",
    "end\n",
    "\n",
    "function count_kmers(::Type{KMER_TYPE}, record::R) where {KMER_TYPE, R <: Union{FASTX.FASTA.Record, FASTX.FASTQ.Record}}\n",
    "    return count_kmers(KMER_TYPE, FASTX.sequence(record))    \n",
    "end\n",
    "\n",
    "function count_kmers(::Type{KMER_TYPE}, sequences) where KMER_TYPE\n",
    "    joint_kmer_counts = DataStructures.OrderedDict{KMER_TYPE, Int}()\n",
    "    for sequence in sequences\n",
    "        sequence_kmer_counts = count_kmers(KMER_TYPE, sequence)\n",
    "        merge!(+, joint_kmer_counts, sequence_kmer_counts)\n",
    "    end\n",
    "    sort!(joint_kmer_counts)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_kmer_index (generic function with 1 method)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function get_kmer_index(kmers, kmer)\n",
    "    index = searchsortedfirst(kmers, kmer)\n",
    "    @assert kmers[index] == kmer \"$kmer\"\n",
    "    return index\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_sequence"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "\tfunction get_sequence(;db=\"\"::String, accession=\"\"::String, ftp=\"\"::String)\n",
    "\n",
    "Get dna (db = \"nuccore\") or protein (db = \"protein\") sequences from NCBI\n",
    "or get fasta directly from FTP site\n",
    "\"\"\"\n",
    "function get_sequence(;db=\"\"::String, accession=\"\"::String, ftp=\"\"::String)\n",
    "    if !isempty(db) && !isempty(accession)\n",
    "        # API will block if we request more than 3 times per second, so set a 1/2 second sleep to set max of 2 requests per second when looping\n",
    "        sleep(0.5)\n",
    "        url = \"https://www.ncbi.nlm.nih.gov/sviewer/viewer.cgi?db=$(db)&report=fasta&id=$(accession)\"\n",
    "        return FASTX.FASTA.Reader(IOBuffer(HTTP.get(url).body))\n",
    "    elseif !isempty(ftp)\n",
    "        return FASTX.FASTA.Reader(CodecZlib.GzipDecompressorStream(IOBuffer(HTTP.get(ftp).body)))\n",
    "    else\n",
    "        @error \"invalid call\"\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "┌ Info: Precompiling Eisenia [453d265d-8292-4a7b-a57c-dce3f9ae6acd]\n",
      "└ @ Base loading.jl:1278\n",
      "┌ Warning: Package Eisenia does not have MetaGraphs in its dependencies:\n",
      "│ - If you have Eisenia checked out for development and have\n",
      "│   added MetaGraphs as a dependency but haven't updated your primary\n",
      "│   environment's manifest file, try `Pkg.resolve()`.\n",
      "│ - Otherwise you may need to report an issue with Eisenia\n",
      "└ Loading MetaGraphs into Eisenia from project dependency, future warnings for Eisenia are suppressed.\n"
     ]
    }
   ],
   "source": [
    "import Eisenia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "graph_to_gfa (generic function with 1 method)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function graph_to_gfa(;graph, outfile)\n",
    "#     outfile = \"test.$(k).simplified.gfa\"\n",
    "    open(outfile, \"w\") do io\n",
    "        println(io, \"H\\tVN:Z:1.0\")\n",
    "        for vertex in LightGraphs.vertices(graph)\n",
    "            sequence = graph.vprops[vertex][:sequence]\n",
    "            if haskey(graph.vprops, vertex) && haskey(graph.vprops[vertex], :evidence)\n",
    "                depth = length(graph.vprops[vertex][:evidence])\n",
    "            else\n",
    "                depth = 1\n",
    "            end\n",
    "            fields = [\"S\", \"$vertex\", sequence, \"RC:i:$(depth)\"]\n",
    "            line = join(fields, '\\t')\n",
    "            println(io, line)\n",
    "        end\n",
    "        for edge in LightGraphs.edges(graph)\n",
    "            overlap = graph.gprops[:k] - 1\n",
    "            link = [\"L\",\n",
    "                        edge.src,\n",
    "                        graph.eprops[edge][:orientations].source_orientation ? '+' : '-',\n",
    "                        edge.dst,\n",
    "                        graph.eprops[edge][:orientations].destination_orientation ? '+' : '-',\n",
    "                        \"$(overlap)M\"]\n",
    "            line = join(link, '\\t')\n",
    "            println(io, line)\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Git_directory = homedir() * \"/\" * first(filter(x -> occursin(r\"^git$\"i, x), readdir(homedir())))\n",
    "# path = \"$(Git_directory)/Eisenia\"\n",
    "# push!(LOAD_PATH, path); Pkg.activate(path); Pkg.update(); Pkg.instantiate(); Pkg.activate()\n",
    "# import Eisenia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th>Accession</th><th>SRA_Accession</th><th>Submitters</th></tr><tr><th></th><th>String</th><th>String</th><th>String</th></tr></thead><tbody><p>354 rows × 21 columns (omitted printing of 18 columns)</p><tr><th>1</th><td>NC_048634</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>2</th><td>NC_048635</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>3</th><td>NC_048636</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>4</th><td>NC_048644</td><td></td><td>Escobar-Perez,J., Reyes,N., Marquez-Ortiz,A., Rebollo,J., Pinzon,H., Tovar,C., Moreno,J.C., Corredor,Z.R., Castro,B.C., Moncada,M.G., Vanegas,N.G.</td></tr><tr><th>5</th><td>NC_048657</td><td></td><td>Sun,Q., Zhang,X., Xing,S., Tong,Y.-G.</td></tr><tr><th>6</th><td>NC_048658</td><td></td><td>Kim,D.</td></tr><tr><th>7</th><td>NC_048681</td><td></td><td>O'Brien,F.G., Baines,S.L., Howden,B.P., Coombs,G.W.</td></tr><tr><th>8</th><td>NC_048710</td><td></td><td>Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V.</td></tr><tr><th>9</th><td>NC_048711</td><td></td><td>Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V.</td></tr><tr><th>10</th><td>NC_048713</td><td></td><td>Dmitrenko,O.A., Balbutskaya,A., Alkhovsky,S.V.</td></tr><tr><th>11</th><td>NC_047720</td><td></td><td>Vandersteegen,K., Mattheus,W., Ceyssens,P.J., Bilocq,F., De Vos,D., Pirnay,J.P., Noben,J.P., Merabishvili,M., Lipinska,U., Hermans,K., Lavigne,R.</td></tr><tr><th>12</th><td>NC_047721</td><td></td><td>Kirby,A.E.</td></tr><tr><th>13</th><td>NC_047722</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>14</th><td>NC_047723</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>15</th><td>NC_047724</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>16</th><td>NC_047725</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>17</th><td>NC_047726</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>18</th><td>NC_047727</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>19</th><td>NC_047728</td><td></td><td>Ulatowska,M.I., Weber-Dabrowska,B., Sadowy,E., Krzyszton-Russjan,J., Gorski,A., Lobocka,M.B.</td></tr><tr><th>20</th><td>NC_047729</td><td></td><td>Zhang,X., Kang,H., Tong,Y.</td></tr><tr><th>21</th><td>NC_047730</td><td></td><td>Zhang,X., Kang,H., Tong,Y.</td></tr><tr><th>22</th><td>NC_047731</td><td></td><td>Zhang,X.</td></tr><tr><th>23</th><td>NC_047732</td><td></td><td>Tong,Y., Zhang,X., Liu,X.</td></tr><tr><th>24</th><td>NC_047788</td><td></td><td>Aleshkin,A.V., Volozhantsev,N.V., Verevkin,V.V., Krasilnikova,V.M., Myakinina,V.P., Popova,A.V., Svetoch,E.A.</td></tr><tr><th>25</th><td>NC_047855</td><td></td><td>Kraushaar,B., Dinh Thanh,M., Reetz,J., Fetsch,A., Hammerl,J.A., Hertwig,S., Hammerl,J.</td></tr><tr><th>26</th><td>NC_047919</td><td></td><td>Gozdek,A., Glowacka-Rutkowska,A., Gawor,J., Empel,J., Gromadka,R., Lobocka,M.B.</td></tr><tr><th>27</th><td>NC_047948</td><td></td><td>Jin,X., Zhang,C., Wang,X., Zhong,J.</td></tr><tr><th>28</th><td>NC_048107</td><td></td><td>Bari,S.M.N., Hernandez,A.C., Hatoum-Aslan,A.</td></tr><tr><th>29</th><td>NC_048159</td><td></td><td>Cha,Y., Chun,J., Son,B., Ryu,S.</td></tr><tr><th>30</th><td>NC_031008</td><td></td><td>Wang,Z., Zheng,P.</td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccc}\n",
       "\t& Accession & SRA\\_Accession & Submitters & \\\\\n",
       "\t\\hline\n",
       "\t& String & String & String & \\\\\n",
       "\t\\hline\n",
       "\t1 & NC\\_048634 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t2 & NC\\_048635 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t3 & NC\\_048636 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t4 & NC\\_048644 &  & Escobar-Perez,J., Reyes,N., Marquez-Ortiz,A., Rebollo,J., Pinzon,H., Tovar,C., Moreno,J.C., Corredor,Z.R., Castro,B.C., Moncada,M.G., Vanegas,N.G. & $\\dots$ \\\\\n",
       "\t5 & NC\\_048657 &  & Sun,Q., Zhang,X., Xing,S., Tong,Y.-G. & $\\dots$ \\\\\n",
       "\t6 & NC\\_048658 &  & Kim,D. & $\\dots$ \\\\\n",
       "\t7 & NC\\_048681 &  & O'Brien,F.G., Baines,S.L., Howden,B.P., Coombs,G.W. & $\\dots$ \\\\\n",
       "\t8 & NC\\_048710 &  & Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V. & $\\dots$ \\\\\n",
       "\t9 & NC\\_048711 &  & Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V. & $\\dots$ \\\\\n",
       "\t10 & NC\\_048713 &  & Dmitrenko,O.A., Balbutskaya,A., Alkhovsky,S.V. & $\\dots$ \\\\\n",
       "\t11 & NC\\_047720 &  & Vandersteegen,K., Mattheus,W., Ceyssens,P.J., Bilocq,F., De Vos,D., Pirnay,J.P., Noben,J.P., Merabishvili,M., Lipinska,U., Hermans,K., Lavigne,R. & $\\dots$ \\\\\n",
       "\t12 & NC\\_047721 &  & Kirby,A.E. & $\\dots$ \\\\\n",
       "\t13 & NC\\_047722 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t14 & NC\\_047723 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t15 & NC\\_047724 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t16 & NC\\_047725 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t17 & NC\\_047726 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t18 & NC\\_047727 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t19 & NC\\_047728 &  & Ulatowska,M.I., Weber-Dabrowska,B., Sadowy,E., Krzyszton-Russjan,J., Gorski,A., Lobocka,M.B. & $\\dots$ \\\\\n",
       "\t20 & NC\\_047729 &  & Zhang,X., Kang,H., Tong,Y. & $\\dots$ \\\\\n",
       "\t21 & NC\\_047730 &  & Zhang,X., Kang,H., Tong,Y. & $\\dots$ \\\\\n",
       "\t22 & NC\\_047731 &  & Zhang,X. & $\\dots$ \\\\\n",
       "\t23 & NC\\_047732 &  & Tong,Y., Zhang,X., Liu,X. & $\\dots$ \\\\\n",
       "\t24 & NC\\_047788 &  & Aleshkin,A.V., Volozhantsev,N.V., Verevkin,V.V., Krasilnikova,V.M., Myakinina,V.P., Popova,A.V., Svetoch,E.A. & $\\dots$ \\\\\n",
       "\t25 & NC\\_047855 &  & Kraushaar,B., Dinh Thanh,M., Reetz,J., Fetsch,A., Hammerl,J.A., Hertwig,S., Hammerl,J. & $\\dots$ \\\\\n",
       "\t26 & NC\\_047919 &  & Gozdek,A., Glowacka-Rutkowska,A., Gawor,J., Empel,J., Gromadka,R., Lobocka,M.B. & $\\dots$ \\\\\n",
       "\t27 & NC\\_047948 &  & Jin,X., Zhang,C., Wang,X., Zhong,J. & $\\dots$ \\\\\n",
       "\t28 & NC\\_048107 &  & Bari,S.M.N., Hernandez,A.C., Hatoum-Aslan,A. & $\\dots$ \\\\\n",
       "\t29 & NC\\_048159 &  & Cha,Y., Chun,J., Son,B., Ryu,S. & $\\dots$ \\\\\n",
       "\t30 & NC\\_031008 &  & Wang,Z., Zheng,P. & $\\dots$ \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ &  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "354×21 DataFrame. Omitted printing of 19 columns\n",
       "│ Row │ Accession │ SRA_Accession │\n",
       "│     │ \u001b[90mString\u001b[39m    │ \u001b[90mString\u001b[39m        │\n",
       "├─────┼───────────┼───────────────┤\n",
       "│ 1   │ NC_048634 │               │\n",
       "│ 2   │ NC_048635 │               │\n",
       "│ 3   │ NC_048636 │               │\n",
       "│ 4   │ NC_048644 │               │\n",
       "│ 5   │ NC_048657 │               │\n",
       "│ 6   │ NC_048658 │               │\n",
       "│ 7   │ NC_048681 │               │\n",
       "│ 8   │ NC_048710 │               │\n",
       "│ 9   │ NC_048711 │               │\n",
       "│ 10  │ NC_048713 │               │\n",
       "⋮\n",
       "│ 344 │ AB009866  │               │\n",
       "│ 345 │ Y07739    │               │\n",
       "│ 346 │ Y07740    │               │\n",
       "│ 347 │ Y12633    │               │\n",
       "│ 348 │ U72397    │               │\n",
       "│ 349 │ A17530    │               │\n",
       "│ 350 │ A17529    │               │\n",
       "│ 351 │ U01872    │               │\n",
       "│ 352 │ U01875    │               │\n",
       "│ 353 │ M27965    │               │\n",
       "│ 354 │ M57455    │               │"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_staph_phage_metadata = DataFrames.DataFrame(uCSV.read(\"$(dirname(pwd()))/metadata/ncbi-staph-phage.csv\", header=1, quotes='\"', typedetectrows=100)...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th>Accession</th><th>SRA_Accession</th><th>Submitters</th></tr><tr><th></th><th>String</th><th>String</th><th>String</th></tr></thead><tbody><p>276 rows × 21 columns (omitted printing of 18 columns)</p><tr><th>1</th><td>NC_048634</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>2</th><td>NC_048635</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>3</th><td>NC_048636</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>4</th><td>NC_048644</td><td></td><td>Escobar-Perez,J., Reyes,N., Marquez-Ortiz,A., Rebollo,J., Pinzon,H., Tovar,C., Moreno,J.C., Corredor,Z.R., Castro,B.C., Moncada,M.G., Vanegas,N.G.</td></tr><tr><th>5</th><td>NC_048657</td><td></td><td>Sun,Q., Zhang,X., Xing,S., Tong,Y.-G.</td></tr><tr><th>6</th><td>NC_048658</td><td></td><td>Kim,D.</td></tr><tr><th>7</th><td>NC_048681</td><td></td><td>O'Brien,F.G., Baines,S.L., Howden,B.P., Coombs,G.W.</td></tr><tr><th>8</th><td>NC_048710</td><td></td><td>Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V.</td></tr><tr><th>9</th><td>NC_048711</td><td></td><td>Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V.</td></tr><tr><th>10</th><td>NC_048713</td><td></td><td>Dmitrenko,O.A., Balbutskaya,A., Alkhovsky,S.V.</td></tr><tr><th>11</th><td>NC_047720</td><td></td><td>Vandersteegen,K., Mattheus,W., Ceyssens,P.J., Bilocq,F., De Vos,D., Pirnay,J.P., Noben,J.P., Merabishvili,M., Lipinska,U., Hermans,K., Lavigne,R.</td></tr><tr><th>12</th><td>NC_047721</td><td></td><td>Kirby,A.E.</td></tr><tr><th>13</th><td>NC_047722</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>14</th><td>NC_047723</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>15</th><td>NC_047724</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>16</th><td>NC_047725</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>17</th><td>NC_047726</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>18</th><td>NC_047727</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>19</th><td>NC_047728</td><td></td><td>Ulatowska,M.I., Weber-Dabrowska,B., Sadowy,E., Krzyszton-Russjan,J., Gorski,A., Lobocka,M.B.</td></tr><tr><th>20</th><td>NC_047729</td><td></td><td>Zhang,X., Kang,H., Tong,Y.</td></tr><tr><th>21</th><td>NC_047730</td><td></td><td>Zhang,X., Kang,H., Tong,Y.</td></tr><tr><th>22</th><td>NC_047731</td><td></td><td>Zhang,X.</td></tr><tr><th>23</th><td>NC_047732</td><td></td><td>Tong,Y., Zhang,X., Liu,X.</td></tr><tr><th>24</th><td>NC_047788</td><td></td><td>Aleshkin,A.V., Volozhantsev,N.V., Verevkin,V.V., Krasilnikova,V.M., Myakinina,V.P., Popova,A.V., Svetoch,E.A.</td></tr><tr><th>25</th><td>NC_047855</td><td></td><td>Kraushaar,B., Dinh Thanh,M., Reetz,J., Fetsch,A., Hammerl,J.A., Hertwig,S., Hammerl,J.</td></tr><tr><th>26</th><td>NC_047919</td><td></td><td>Gozdek,A., Glowacka-Rutkowska,A., Gawor,J., Empel,J., Gromadka,R., Lobocka,M.B.</td></tr><tr><th>27</th><td>NC_047948</td><td></td><td>Jin,X., Zhang,C., Wang,X., Zhong,J.</td></tr><tr><th>28</th><td>NC_048107</td><td></td><td>Bari,S.M.N., Hernandez,A.C., Hatoum-Aslan,A.</td></tr><tr><th>29</th><td>NC_048159</td><td></td><td>Cha,Y., Chun,J., Son,B., Ryu,S.</td></tr><tr><th>30</th><td>NC_031008</td><td></td><td>Wang,Z., Zheng,P.</td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccc}\n",
       "\t& Accession & SRA\\_Accession & Submitters & \\\\\n",
       "\t\\hline\n",
       "\t& String & String & String & \\\\\n",
       "\t\\hline\n",
       "\t1 & NC\\_048634 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t2 & NC\\_048635 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t3 & NC\\_048636 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t4 & NC\\_048644 &  & Escobar-Perez,J., Reyes,N., Marquez-Ortiz,A., Rebollo,J., Pinzon,H., Tovar,C., Moreno,J.C., Corredor,Z.R., Castro,B.C., Moncada,M.G., Vanegas,N.G. & $\\dots$ \\\\\n",
       "\t5 & NC\\_048657 &  & Sun,Q., Zhang,X., Xing,S., Tong,Y.-G. & $\\dots$ \\\\\n",
       "\t6 & NC\\_048658 &  & Kim,D. & $\\dots$ \\\\\n",
       "\t7 & NC\\_048681 &  & O'Brien,F.G., Baines,S.L., Howden,B.P., Coombs,G.W. & $\\dots$ \\\\\n",
       "\t8 & NC\\_048710 &  & Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V. & $\\dots$ \\\\\n",
       "\t9 & NC\\_048711 &  & Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V. & $\\dots$ \\\\\n",
       "\t10 & NC\\_048713 &  & Dmitrenko,O.A., Balbutskaya,A., Alkhovsky,S.V. & $\\dots$ \\\\\n",
       "\t11 & NC\\_047720 &  & Vandersteegen,K., Mattheus,W., Ceyssens,P.J., Bilocq,F., De Vos,D., Pirnay,J.P., Noben,J.P., Merabishvili,M., Lipinska,U., Hermans,K., Lavigne,R. & $\\dots$ \\\\\n",
       "\t12 & NC\\_047721 &  & Kirby,A.E. & $\\dots$ \\\\\n",
       "\t13 & NC\\_047722 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t14 & NC\\_047723 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t15 & NC\\_047724 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t16 & NC\\_047725 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t17 & NC\\_047726 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t18 & NC\\_047727 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t19 & NC\\_047728 &  & Ulatowska,M.I., Weber-Dabrowska,B., Sadowy,E., Krzyszton-Russjan,J., Gorski,A., Lobocka,M.B. & $\\dots$ \\\\\n",
       "\t20 & NC\\_047729 &  & Zhang,X., Kang,H., Tong,Y. & $\\dots$ \\\\\n",
       "\t21 & NC\\_047730 &  & Zhang,X., Kang,H., Tong,Y. & $\\dots$ \\\\\n",
       "\t22 & NC\\_047731 &  & Zhang,X. & $\\dots$ \\\\\n",
       "\t23 & NC\\_047732 &  & Tong,Y., Zhang,X., Liu,X. & $\\dots$ \\\\\n",
       "\t24 & NC\\_047788 &  & Aleshkin,A.V., Volozhantsev,N.V., Verevkin,V.V., Krasilnikova,V.M., Myakinina,V.P., Popova,A.V., Svetoch,E.A. & $\\dots$ \\\\\n",
       "\t25 & NC\\_047855 &  & Kraushaar,B., Dinh Thanh,M., Reetz,J., Fetsch,A., Hammerl,J.A., Hertwig,S., Hammerl,J. & $\\dots$ \\\\\n",
       "\t26 & NC\\_047919 &  & Gozdek,A., Glowacka-Rutkowska,A., Gawor,J., Empel,J., Gromadka,R., Lobocka,M.B. & $\\dots$ \\\\\n",
       "\t27 & NC\\_047948 &  & Jin,X., Zhang,C., Wang,X., Zhong,J. & $\\dots$ \\\\\n",
       "\t28 & NC\\_048107 &  & Bari,S.M.N., Hernandez,A.C., Hatoum-Aslan,A. & $\\dots$ \\\\\n",
       "\t29 & NC\\_048159 &  & Cha,Y., Chun,J., Son,B., Ryu,S. & $\\dots$ \\\\\n",
       "\t30 & NC\\_031008 &  & Wang,Z., Zheng,P. & $\\dots$ \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ &  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "276×21 DataFrame. Omitted printing of 19 columns\n",
       "│ Row │ Accession │ SRA_Accession │\n",
       "│     │ \u001b[90mString\u001b[39m    │ \u001b[90mString\u001b[39m        │\n",
       "├─────┼───────────┼───────────────┤\n",
       "│ 1   │ NC_048634 │               │\n",
       "│ 2   │ NC_048635 │               │\n",
       "│ 3   │ NC_048636 │               │\n",
       "│ 4   │ NC_048644 │               │\n",
       "│ 5   │ NC_048657 │               │\n",
       "│ 6   │ NC_048658 │               │\n",
       "│ 7   │ NC_048681 │               │\n",
       "│ 8   │ NC_048710 │               │\n",
       "│ 9   │ NC_048711 │               │\n",
       "│ 10  │ NC_048713 │               │\n",
       "⋮\n",
       "│ 266 │ EU418428  │               │\n",
       "│ 267 │ GU169904  │               │\n",
       "│ 268 │ GQ398772  │               │\n",
       "│ 269 │ FJ713816  │               │\n",
       "│ 270 │ EU861004  │               │\n",
       "│ 271 │ EU861005  │               │\n",
       "│ 272 │ EU136189  │               │\n",
       "│ 273 │ AY954948  │               │\n",
       "│ 274 │ AF424781  │               │\n",
       "│ 275 │ AF424782  │               │\n",
       "│ 276 │ AF424783  │               │"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_staph_phage_metadata = ncbi_staph_phage_metadata[ncbi_staph_phage_metadata[!, \"Nuc_Completeness\"] .== \"complete\", :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th>Accession</th><th>SRA_Accession</th><th>Submitters</th></tr><tr><th></th><th>String</th><th>String</th><th>String</th></tr></thead><tbody><p>111 rows × 21 columns (omitted printing of 18 columns)</p><tr><th>1</th><td>NC_048634</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>2</th><td>NC_048635</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>3</th><td>NC_048636</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>4</th><td>NC_048644</td><td></td><td>Escobar-Perez,J., Reyes,N., Marquez-Ortiz,A., Rebollo,J., Pinzon,H., Tovar,C., Moreno,J.C., Corredor,Z.R., Castro,B.C., Moncada,M.G., Vanegas,N.G.</td></tr><tr><th>5</th><td>NC_048657</td><td></td><td>Sun,Q., Zhang,X., Xing,S., Tong,Y.-G.</td></tr><tr><th>6</th><td>NC_048658</td><td></td><td>Kim,D.</td></tr><tr><th>7</th><td>NC_048681</td><td></td><td>O'Brien,F.G., Baines,S.L., Howden,B.P., Coombs,G.W.</td></tr><tr><th>8</th><td>NC_048710</td><td></td><td>Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V.</td></tr><tr><th>9</th><td>NC_048711</td><td></td><td>Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V.</td></tr><tr><th>10</th><td>NC_048713</td><td></td><td>Dmitrenko,O.A., Balbutskaya,A., Alkhovsky,S.V.</td></tr><tr><th>11</th><td>NC_047720</td><td></td><td>Vandersteegen,K., Mattheus,W., Ceyssens,P.J., Bilocq,F., De Vos,D., Pirnay,J.P., Noben,J.P., Merabishvili,M., Lipinska,U., Hermans,K., Lavigne,R.</td></tr><tr><th>12</th><td>NC_047721</td><td></td><td>Kirby,A.E.</td></tr><tr><th>13</th><td>NC_047722</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>14</th><td>NC_047723</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>15</th><td>NC_047724</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>16</th><td>NC_047725</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>17</th><td>NC_047726</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>18</th><td>NC_047727</td><td></td><td>Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A.</td></tr><tr><th>19</th><td>NC_047728</td><td></td><td>Ulatowska,M.I., Weber-Dabrowska,B., Sadowy,E., Krzyszton-Russjan,J., Gorski,A., Lobocka,M.B.</td></tr><tr><th>20</th><td>NC_047729</td><td></td><td>Zhang,X., Kang,H., Tong,Y.</td></tr><tr><th>21</th><td>NC_047730</td><td></td><td>Zhang,X., Kang,H., Tong,Y.</td></tr><tr><th>22</th><td>NC_047731</td><td></td><td>Zhang,X.</td></tr><tr><th>23</th><td>NC_047732</td><td></td><td>Tong,Y., Zhang,X., Liu,X.</td></tr><tr><th>24</th><td>NC_047788</td><td></td><td>Aleshkin,A.V., Volozhantsev,N.V., Verevkin,V.V., Krasilnikova,V.M., Myakinina,V.P., Popova,A.V., Svetoch,E.A.</td></tr><tr><th>25</th><td>NC_047855</td><td></td><td>Kraushaar,B., Dinh Thanh,M., Reetz,J., Fetsch,A., Hammerl,J.A., Hertwig,S., Hammerl,J.</td></tr><tr><th>26</th><td>NC_047919</td><td></td><td>Gozdek,A., Glowacka-Rutkowska,A., Gawor,J., Empel,J., Gromadka,R., Lobocka,M.B.</td></tr><tr><th>27</th><td>NC_047948</td><td></td><td>Jin,X., Zhang,C., Wang,X., Zhong,J.</td></tr><tr><th>28</th><td>NC_048107</td><td></td><td>Bari,S.M.N., Hernandez,A.C., Hatoum-Aslan,A.</td></tr><tr><th>29</th><td>NC_048159</td><td></td><td>Cha,Y., Chun,J., Son,B., Ryu,S.</td></tr><tr><th>30</th><td>NC_031008</td><td></td><td>Wang,Z., Zheng,P.</td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccc}\n",
       "\t& Accession & SRA\\_Accession & Submitters & \\\\\n",
       "\t\\hline\n",
       "\t& String & String & String & \\\\\n",
       "\t\\hline\n",
       "\t1 & NC\\_048634 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t2 & NC\\_048635 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t3 & NC\\_048636 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t4 & NC\\_048644 &  & Escobar-Perez,J., Reyes,N., Marquez-Ortiz,A., Rebollo,J., Pinzon,H., Tovar,C., Moreno,J.C., Corredor,Z.R., Castro,B.C., Moncada,M.G., Vanegas,N.G. & $\\dots$ \\\\\n",
       "\t5 & NC\\_048657 &  & Sun,Q., Zhang,X., Xing,S., Tong,Y.-G. & $\\dots$ \\\\\n",
       "\t6 & NC\\_048658 &  & Kim,D. & $\\dots$ \\\\\n",
       "\t7 & NC\\_048681 &  & O'Brien,F.G., Baines,S.L., Howden,B.P., Coombs,G.W. & $\\dots$ \\\\\n",
       "\t8 & NC\\_048710 &  & Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V. & $\\dots$ \\\\\n",
       "\t9 & NC\\_048711 &  & Dmitrenko,O.A., Tikhomirov,T., Balbutskaya,A., Fedorova,N., Alkhovsky,S.V. & $\\dots$ \\\\\n",
       "\t10 & NC\\_048713 &  & Dmitrenko,O.A., Balbutskaya,A., Alkhovsky,S.V. & $\\dots$ \\\\\n",
       "\t11 & NC\\_047720 &  & Vandersteegen,K., Mattheus,W., Ceyssens,P.J., Bilocq,F., De Vos,D., Pirnay,J.P., Noben,J.P., Merabishvili,M., Lipinska,U., Hermans,K., Lavigne,R. & $\\dots$ \\\\\n",
       "\t12 & NC\\_047721 &  & Kirby,A.E. & $\\dots$ \\\\\n",
       "\t13 & NC\\_047722 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t14 & NC\\_047723 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t15 & NC\\_047724 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t16 & NC\\_047725 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t17 & NC\\_047726 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t18 & NC\\_047727 &  & Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. & $\\dots$ \\\\\n",
       "\t19 & NC\\_047728 &  & Ulatowska,M.I., Weber-Dabrowska,B., Sadowy,E., Krzyszton-Russjan,J., Gorski,A., Lobocka,M.B. & $\\dots$ \\\\\n",
       "\t20 & NC\\_047729 &  & Zhang,X., Kang,H., Tong,Y. & $\\dots$ \\\\\n",
       "\t21 & NC\\_047730 &  & Zhang,X., Kang,H., Tong,Y. & $\\dots$ \\\\\n",
       "\t22 & NC\\_047731 &  & Zhang,X. & $\\dots$ \\\\\n",
       "\t23 & NC\\_047732 &  & Tong,Y., Zhang,X., Liu,X. & $\\dots$ \\\\\n",
       "\t24 & NC\\_047788 &  & Aleshkin,A.V., Volozhantsev,N.V., Verevkin,V.V., Krasilnikova,V.M., Myakinina,V.P., Popova,A.V., Svetoch,E.A. & $\\dots$ \\\\\n",
       "\t25 & NC\\_047855 &  & Kraushaar,B., Dinh Thanh,M., Reetz,J., Fetsch,A., Hammerl,J.A., Hertwig,S., Hammerl,J. & $\\dots$ \\\\\n",
       "\t26 & NC\\_047919 &  & Gozdek,A., Glowacka-Rutkowska,A., Gawor,J., Empel,J., Gromadka,R., Lobocka,M.B. & $\\dots$ \\\\\n",
       "\t27 & NC\\_047948 &  & Jin,X., Zhang,C., Wang,X., Zhong,J. & $\\dots$ \\\\\n",
       "\t28 & NC\\_048107 &  & Bari,S.M.N., Hernandez,A.C., Hatoum-Aslan,A. & $\\dots$ \\\\\n",
       "\t29 & NC\\_048159 &  & Cha,Y., Chun,J., Son,B., Ryu,S. & $\\dots$ \\\\\n",
       "\t30 & NC\\_031008 &  & Wang,Z., Zheng,P. & $\\dots$ \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ &  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "111×21 DataFrame. Omitted printing of 19 columns\n",
       "│ Row │ Accession │ SRA_Accession │\n",
       "│     │ \u001b[90mString\u001b[39m    │ \u001b[90mString\u001b[39m        │\n",
       "├─────┼───────────┼───────────────┤\n",
       "│ 1   │ NC_048634 │               │\n",
       "│ 2   │ NC_048635 │               │\n",
       "│ 3   │ NC_048636 │               │\n",
       "│ 4   │ NC_048644 │               │\n",
       "│ 5   │ NC_048657 │               │\n",
       "│ 6   │ NC_048658 │               │\n",
       "│ 7   │ NC_048681 │               │\n",
       "│ 8   │ NC_048710 │               │\n",
       "│ 9   │ NC_048711 │               │\n",
       "│ 10  │ NC_048713 │               │\n",
       "⋮\n",
       "│ 101 │ NC_005356 │               │\n",
       "│ 102 │ NC_004740 │               │\n",
       "│ 103 │ NC_004678 │               │\n",
       "│ 104 │ NC_004679 │               │\n",
       "│ 105 │ NC_004615 │               │\n",
       "│ 106 │ NC_004616 │               │\n",
       "│ 107 │ NC_004617 │               │\n",
       "│ 108 │ NC_002661 │               │\n",
       "│ 109 │ NC_002486 │               │\n",
       "│ 110 │ NC_003288 │               │\n",
       "│ 111 │ NC_002321 │               │"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi_staph_phage_metadata = ncbi_staph_phage_metadata[ncbi_staph_phage_metadata[!, \"Sequence_Type\"] .== \"RefSeq\", :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th>Accession</th><th>SRA_Accession</th><th>Submitters</th></tr><tr><th></th><th>String</th><th>String</th><th>String</th></tr></thead><tbody><p>111 rows × 21 columns (omitted printing of 18 columns)</p><tr><th>1</th><td>NC_007045</td><td></td><td>Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J.</td></tr><tr><th>2</th><td>NC_004678</td><td></td><td>Vybiral,D., Takac,M., Loessner,M., Witte,A., von Ahsen,U., Blasi,U., Blaesi,U.</td></tr><tr><th>3</th><td>NC_048159</td><td></td><td>Cha,Y., Chun,J., Son,B., Ryu,S.</td></tr><tr><th>4</th><td>NC_047855</td><td></td><td>Kraushaar,B., Dinh Thanh,M., Reetz,J., Fetsch,A., Hammerl,J.A., Hertwig,S., Hammerl,J.</td></tr><tr><th>5</th><td>NC_047919</td><td></td><td>Gozdek,A., Glowacka-Rutkowska,A., Gawor,J., Empel,J., Gromadka,R., Lobocka,M.B.</td></tr><tr><th>6</th><td>NC_031046</td><td></td><td>Mandeville,R.</td></tr><tr><th>7</th><td>NC_048107</td><td></td><td>Bari,S.M.N., Hernandez,A.C., Hatoum-Aslan,A.</td></tr><tr><th>8</th><td>NC_031008</td><td></td><td>Wang,Z., Zheng,P.</td></tr><tr><th>9</th><td>NC_023550</td><td></td><td>Swift,S.M., Nelson,D.C.</td></tr><tr><th>10</th><td>NC_009875</td><td></td><td>Son,J.S., Lee,S.J., Jun,S.Y., Yoon,S.J., Kang,S.H., Paik,H.R., Kang,J.O., Choi,Y.J.</td></tr><tr><th>11</th><td>NC_047788</td><td></td><td>Aleshkin,A.V., Volozhantsev,N.V., Verevkin,V.V., Krasilnikova,V.M., Myakinina,V.P., Popova,A.V., Svetoch,E.A.</td></tr><tr><th>12</th><td>NC_016565</td><td></td><td>Uchiyama,J., Takemura,I., Matsuzaki,S., Daibata,M.</td></tr><tr><th>13</th><td>NC_007046</td><td></td><td>Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J.</td></tr><tr><th>14</th><td>NC_004679</td><td></td><td>Vybiral,D., Takac,M., Loessner,M., Witte,A., von Ahsen,U., Blasi,U., Blaesi,U.</td></tr><tr><th>15</th><td>NC_048658</td><td></td><td>Kim,D.</td></tr><tr><th>16</th><td>NC_048681</td><td></td><td>O'Brien,F.G., Baines,S.L., Howden,B.P., Coombs,G.W.</td></tr><tr><th>17</th><td>NC_048635</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>18</th><td>NC_029010</td><td></td><td>Chang,Y., Shin,H., Lee,J.H., Park,C.J., Paik,S.Y., Ryu,S.</td></tr><tr><th>19</th><td>NC_013195</td><td></td><td>Young,R.F., Summers,E.J.</td></tr><tr><th>20</th><td>NC_014460</td><td></td><td>Rahman,M., Kim,S., Kim,S.-M., Kim,J.</td></tr><tr><th>21</th><td>NC_007051</td><td></td><td>Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J.</td></tr><tr><th>22</th><td>NC_021332</td><td></td><td>van der Mee-Marquet,N., Valentin,A.-S., Corvaglia,A.-R., Bertrand,X., Girard,M., Kluytmans,J., Donnio,P.-Y., Quentin,R., Schrenzel,J., Francois,P.</td></tr><tr><th>23</th><td>NC_002321</td><td></td><td>Kaneko,J., Kimura,T., Narita,S., Tomita,T., Kamio,Y., Kawakami,Y.</td></tr><tr><th>24</th><td>NC_007062</td><td></td><td>Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J.</td></tr><tr><th>25</th><td>NC_005356</td><td></td><td>Liu,J., Dehbi,M., Moeck,G., Arhin,F., Bauda,P., Bergeron,D., Callejo,M., Ferretti,V., Ha,N., Kwan,T., McCarty,J., Srikumar,R., Williams,D., Wu,J.J., Gros,P., Pelletier,J., DuBow,M., Dubow,M.</td></tr><tr><th>26</th><td>NC_007060</td><td></td><td>Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J.</td></tr><tr><th>27</th><td>NC_028917</td><td></td><td>Santiago-Rodriguez,T.M., Naidu,M., Jones,M.B., Ly,M., Pride,D.T.</td></tr><tr><th>28</th><td>NC_024391</td><td></td><td>Keary,R., McAuliffe,O., Ross,R.P., Hill,C., O'Mahony,J., Coffey,A.</td></tr><tr><th>29</th><td>NC_048634</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>30</th><td>NC_048636</td><td></td><td>Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S.</td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccc}\n",
       "\t& Accession & SRA\\_Accession & Submitters & \\\\\n",
       "\t\\hline\n",
       "\t& String & String & String & \\\\\n",
       "\t\\hline\n",
       "\t1 & NC\\_007045 &  & Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J. & $\\dots$ \\\\\n",
       "\t2 & NC\\_004678 &  & Vybiral,D., Takac,M., Loessner,M., Witte,A., von Ahsen,U., Blasi,U., Blaesi,U. & $\\dots$ \\\\\n",
       "\t3 & NC\\_048159 &  & Cha,Y., Chun,J., Son,B., Ryu,S. & $\\dots$ \\\\\n",
       "\t4 & NC\\_047855 &  & Kraushaar,B., Dinh Thanh,M., Reetz,J., Fetsch,A., Hammerl,J.A., Hertwig,S., Hammerl,J. & $\\dots$ \\\\\n",
       "\t5 & NC\\_047919 &  & Gozdek,A., Glowacka-Rutkowska,A., Gawor,J., Empel,J., Gromadka,R., Lobocka,M.B. & $\\dots$ \\\\\n",
       "\t6 & NC\\_031046 &  & Mandeville,R. & $\\dots$ \\\\\n",
       "\t7 & NC\\_048107 &  & Bari,S.M.N., Hernandez,A.C., Hatoum-Aslan,A. & $\\dots$ \\\\\n",
       "\t8 & NC\\_031008 &  & Wang,Z., Zheng,P. & $\\dots$ \\\\\n",
       "\t9 & NC\\_023550 &  & Swift,S.M., Nelson,D.C. & $\\dots$ \\\\\n",
       "\t10 & NC\\_009875 &  & Son,J.S., Lee,S.J., Jun,S.Y., Yoon,S.J., Kang,S.H., Paik,H.R., Kang,J.O., Choi,Y.J. & $\\dots$ \\\\\n",
       "\t11 & NC\\_047788 &  & Aleshkin,A.V., Volozhantsev,N.V., Verevkin,V.V., Krasilnikova,V.M., Myakinina,V.P., Popova,A.V., Svetoch,E.A. & $\\dots$ \\\\\n",
       "\t12 & NC\\_016565 &  & Uchiyama,J., Takemura,I., Matsuzaki,S., Daibata,M. & $\\dots$ \\\\\n",
       "\t13 & NC\\_007046 &  & Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J. & $\\dots$ \\\\\n",
       "\t14 & NC\\_004679 &  & Vybiral,D., Takac,M., Loessner,M., Witte,A., von Ahsen,U., Blasi,U., Blaesi,U. & $\\dots$ \\\\\n",
       "\t15 & NC\\_048658 &  & Kim,D. & $\\dots$ \\\\\n",
       "\t16 & NC\\_048681 &  & O'Brien,F.G., Baines,S.L., Howden,B.P., Coombs,G.W. & $\\dots$ \\\\\n",
       "\t17 & NC\\_048635 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t18 & NC\\_029010 &  & Chang,Y., Shin,H., Lee,J.H., Park,C.J., Paik,S.Y., Ryu,S. & $\\dots$ \\\\\n",
       "\t19 & NC\\_013195 &  & Young,R.F., Summers,E.J. & $\\dots$ \\\\\n",
       "\t20 & NC\\_014460 &  & Rahman,M., Kim,S., Kim,S.-M., Kim,J. & $\\dots$ \\\\\n",
       "\t21 & NC\\_007051 &  & Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J. & $\\dots$ \\\\\n",
       "\t22 & NC\\_021332 &  & van der Mee-Marquet,N., Valentin,A.-S., Corvaglia,A.-R., Bertrand,X., Girard,M., Kluytmans,J., Donnio,P.-Y., Quentin,R., Schrenzel,J., Francois,P. & $\\dots$ \\\\\n",
       "\t23 & NC\\_002321 &  & Kaneko,J., Kimura,T., Narita,S., Tomita,T., Kamio,Y., Kawakami,Y. & $\\dots$ \\\\\n",
       "\t24 & NC\\_007062 &  & Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J. & $\\dots$ \\\\\n",
       "\t25 & NC\\_005356 &  & Liu,J., Dehbi,M., Moeck,G., Arhin,F., Bauda,P., Bergeron,D., Callejo,M., Ferretti,V., Ha,N., Kwan,T., McCarty,J., Srikumar,R., Williams,D., Wu,J.J., Gros,P., Pelletier,J., DuBow,M., Dubow,M. & $\\dots$ \\\\\n",
       "\t26 & NC\\_007060 &  & Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J. & $\\dots$ \\\\\n",
       "\t27 & NC\\_028917 &  & Santiago-Rodriguez,T.M., Naidu,M., Jones,M.B., Ly,M., Pride,D.T. & $\\dots$ \\\\\n",
       "\t28 & NC\\_024391 &  & Keary,R., McAuliffe,O., Ross,R.P., Hill,C., O'Mahony,J., Coffey,A. & $\\dots$ \\\\\n",
       "\t29 & NC\\_048634 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t30 & NC\\_048636 &  & Kraushaar,B., Hammerl,J.A., Kienol,M., Heinig,M.L., Sperling,N., Dinh Thanh,M., Reetz,J., Jackel,C., Fetsch,A., Hertwig,S. & $\\dots$ \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ &  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "111×21 DataFrame. Omitted printing of 19 columns\n",
       "│ Row │ Accession │ SRA_Accession │\n",
       "│     │ \u001b[90mString\u001b[39m    │ \u001b[90mString\u001b[39m        │\n",
       "├─────┼───────────┼───────────────┤\n",
       "│ 1   │ NC_007045 │               │\n",
       "│ 2   │ NC_004678 │               │\n",
       "│ 3   │ NC_048159 │               │\n",
       "│ 4   │ NC_047855 │               │\n",
       "│ 5   │ NC_047919 │               │\n",
       "│ 6   │ NC_031046 │               │\n",
       "│ 7   │ NC_048107 │               │\n",
       "│ 8   │ NC_031008 │               │\n",
       "│ 9   │ NC_023550 │               │\n",
       "│ 10  │ NC_009875 │               │\n",
       "⋮\n",
       "│ 101 │ NC_025416 │               │\n",
       "│ 102 │ NC_023573 │               │\n",
       "│ 103 │ NC_047728 │               │\n",
       "│ 104 │ NC_047722 │               │\n",
       "│ 105 │ NC_027991 │               │\n",
       "│ 106 │ NC_047727 │               │\n",
       "│ 107 │ NC_047726 │               │\n",
       "│ 108 │ NC_005880 │               │\n",
       "│ 109 │ NC_047725 │               │\n",
       "│ 110 │ NC_047724 │               │\n",
       "│ 111 │ NC_047948 │               │"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sort!(ncbi_staph_phage_metadata, \"Length\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "111×21 DataFrame\n",
      "│ Row │ Accession │ SRA_Accession │\n",
      "│     │ \u001b[90mString\u001b[39m    │ \u001b[90mString\u001b[39m        │\n",
      "├─────┼───────────┼───────────────┤\n",
      "│ 1   │ NC_007045 │               │\n",
      "│ 2   │ NC_004678 │               │\n",
      "│ 3   │ NC_048159 │               │\n",
      "│ 4   │ NC_047855 │               │\n",
      "│ 5   │ NC_047919 │               │\n",
      "│ 6   │ NC_031046 │               │\n",
      "│ 7   │ NC_048107 │               │\n",
      "│ 8   │ NC_031008 │               │\n",
      "│ 9   │ NC_023550 │               │\n",
      "│ 10  │ NC_009875 │               │\n",
      "⋮\n",
      "│ 101 │ NC_025416 │               │\n",
      "│ 102 │ NC_023573 │               │\n",
      "│ 103 │ NC_047728 │               │\n",
      "│ 104 │ NC_047722 │               │\n",
      "│ 105 │ NC_027991 │               │\n",
      "│ 106 │ NC_047727 │               │\n",
      "│ 107 │ NC_047726 │               │\n",
      "│ 108 │ NC_005880 │               │\n",
      "│ 109 │ NC_047725 │               │\n",
      "│ 110 │ NC_047724 │               │\n",
      "│ 111 │ NC_047948 │               │\n",
      "\n",
      "│ Row │ Submitters                                                                                                                                                                        │\n",
      "│     │ \u001b[90mString\u001b[39m                                                                                                                                                                            │\n",
      "├─────┼───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤\n",
      "│ 1   │ Kwan,T., Liu,J., DuBow,M., Gros,P., Pelletier,J.                                                                                                                                  │\n",
      "│ 2   │ Vybiral,D., Takac,M., Loessner,M., Witte,A., von Ahsen,U., Blasi,U., Blaesi,U.                                                                                                    │\n",
      "│ 3   │ Cha,Y., Chun,J., Son,B., Ryu,S.                                                                                                                                                   │\n",
      "│ 4   │ Kraushaar,B., Dinh Thanh,M., Reetz,J., Fetsch,A., Hammerl,J.A., Hertwig,S., Hammerl,J.                                                                                            │\n",
      "│ 5   │ Gozdek,A., Glowacka-Rutkowska,A., Gawor,J., Empel,J., Gromadka,R., Lobocka,M.B.                                                                                                   │\n",
      "│ 6   │ Mandeville,R.                                                                                                                                                                     │\n",
      "│ 7   │ Bari,S.M.N., Hernandez,A.C., Hatoum-Aslan,A.                                                                                                                                      │\n",
      "│ 8   │ Wang,Z., Zheng,P.                                                                                                                                                                 │\n",
      "│ 9   │ Swift,S.M., Nelson,D.C.                                                                                                                                                           │\n",
      "│ 10  │ Son,J.S., Lee,S.J., Jun,S.Y., Yoon,S.J., Kang,S.H., Paik,H.R., Kang,J.O., Choi,Y.J.                                                                                               │\n",
      "⋮\n",
      "│ 101 │ Alves,D.R., Gaudion,A., Bean,J.E., Perez Esteban,P., Arnot,T.C., Harper,D.R., Kot,W., Hansen,L.H., Enright,M.C., Jenkins,A.T., Perez-Esteban,P., Arnot,T., Hanson,L.H.            │\n",
      "│ 102 │ Takeuchi,I., Osada,K., Asakawa,H., Miyanaga,K., Tanji,Y.                                                                                                                          │\n",
      "│ 103 │ Ulatowska,M.I., Weber-Dabrowska,B., Sadowy,E., Krzyszton-Russjan,J., Gorski,A., Lobocka,M.B.                                                                                      │\n",
      "│ 104 │ Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. │\n",
      "│ 105 │ Lee,Y., Park,J., Chang,H., Ryu,T.                                                                                                                                                 │\n",
      "│ 106 │ Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. │\n",
      "│ 107 │ Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. │\n",
      "│ 108 │ Gill,J.J.                                                                                                                                                                         │\n",
      "│ 109 │ Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. │\n",
      "│ 110 │ Lobocka,M., Hejnowicz,M.S., Dabrowski,K., Gozdek,A., Kosakowski,J., Witkowska,M., Ulatowska,M.I., Weber-Dabrowska,B., Kwiatek,M., Parasion,S., Gawor,J., Kosowska,H., Glowacka,A. │\n",
      "│ 111 │ Jin,X., Zhang,C., Wang,X., Zhong,J.                                                                                                                                               │\n",
      "\n",
      "│ Row │ Release_Date         │ Species                       │ Genus          │\n",
      "│     │ \u001b[90mString\u001b[39m               │ \u001b[90mString\u001b[39m                        │ \u001b[90mString\u001b[39m         │\n",
      "├─────┼──────────────────────┼───────────────────────────────┼────────────────┤\n",
      "│ 1   │ 2005-05-23T00:00:00Z │ Staphylococcus phage PT1028   │                │\n",
      "│ 2   │ 2003-04-05T00:00:00Z │ Staphylococcus virus 44AHJD   │ Rosenblumvirus │\n",
      "│ 3   │ 2020-05-15T00:00:00Z │ Staphylococcus virus CSA13    │ Rosenblumvirus │\n",
      "│ 4   │ 2020-05-15T00:00:00Z │ Staphylococcus virus PSa3     │ Rosenblumvirus │\n",
      "│ 5   │ 2020-05-15T00:00:00Z │ Staphylococcus virus phiAGO13 │ Rosenblumvirus │\n",
      "│ 6   │ 2016-09-23T00:00:00Z │ Staphylococcus virus BP39     │ Rosenblumvirus │\n",
      "│ 7   │ 2020-05-15T00:00:00Z │ Staphylococcus virus Pabna    │ Rosenblumvirus │\n",
      "│ 8   │ 2016-09-23T00:00:00Z │ Staphylococcus virus SLPW     │ Rosenblumvirus │\n",
      "│ 9   │ 2014-03-04T00:00:00Z │ Staphylococcus virus GRCS     │ Rosenblumvirus │\n",
      "│ 10  │ 2007-09-28T00:00:00Z │ Staphylococcus virus SAP2     │ Rosenblumvirus │\n",
      "⋮\n",
      "│ 101 │ 2014-11-25T00:00:00Z │ Staphylococcus virus MCE2014  │ Kayvirus       │\n",
      "│ 102 │ 2014-03-02T00:00:00Z │ Staphylococcus virus SA12     │ Kayvirus       │\n",
      "│ 103 │ 2020-05-15T00:00:00Z │ Staphylococcus virus G1       │ Kayvirus       │\n",
      "│ 104 │ 2020-05-15T00:00:00Z │ Staphylococcus virus G1       │ Kayvirus       │\n",
      "│ 105 │ 2015-10-07T00:00:00Z │ Staphylococcus phage SA1      │                │\n",
      "│ 106 │ 2020-05-15T00:00:00Z │ Staphylococcus virus G1       │ Kayvirus       │\n",
      "│ 107 │ 2020-05-15T00:00:00Z │ Staphylococcus virus G1       │ Kayvirus       │\n",
      "│ 108 │ 2004-06-14T00:00:00Z │ Staphylococcus virus K        │ Kayvirus       │\n",
      "│ 109 │ 2020-05-15T00:00:00Z │ Staphylococcus virus G1       │ Kayvirus       │\n",
      "│ 110 │ 2020-05-15T00:00:00Z │ Staphylococcus virus G1       │ Kayvirus       │\n",
      "│ 111 │ 2020-05-15T00:00:00Z │ Staphylococcus virus BS2      │ Baoshanvirus   │\n",
      "\n",
      "│ Row │ Family         │ Molecule_type │ Length │ Sequence_Type │\n",
      "│     │ \u001b[90mString\u001b[39m         │ \u001b[90mString\u001b[39m        │ \u001b[90mInt64\u001b[39m  │ \u001b[90mString\u001b[39m        │\n",
      "├─────┼────────────────┼───────────────┼────────┼───────────────┤\n",
      "│ 1   │                │ dsDNA         │ 15603  │ RefSeq        │\n",
      "│ 2   │ Podoviridae    │ dsDNA         │ 16784  │ RefSeq        │\n",
      "│ 3   │ Podoviridae    │ dsDNA         │ 17034  │ RefSeq        │\n",
      "│ 4   │ Podoviridae    │ dsDNA         │ 17602  │ RefSeq        │\n",
      "│ 5   │ Podoviridae    │ dsDNA         │ 17603  │ RefSeq        │\n",
      "│ 6   │ Podoviridae    │ dsDNA         │ 17641  │ RefSeq        │\n",
      "│ 7   │ Podoviridae    │ dsDNA         │ 17700  │ RefSeq        │\n",
      "│ 8   │ Podoviridae    │ dsDNA         │ 17861  │ RefSeq        │\n",
      "│ 9   │ Podoviridae    │ dsDNA         │ 17869  │ RefSeq        │\n",
      "│ 10  │ Podoviridae    │ dsDNA         │ 17938  │ RefSeq        │\n",
      "⋮\n",
      "│ 101 │ Herelleviridae │ dsDNA         │ 141907 │ RefSeq        │\n",
      "│ 102 │ Herelleviridae │ dsDNA         │ 142094 │ RefSeq        │\n",
      "│ 103 │ Herelleviridae │ dsDNA         │ 145542 │ RefSeq        │\n",
      "│ 104 │ Herelleviridae │ dsDNA         │ 145647 │ RefSeq        │\n",
      "│ 105 │ Myoviridae     │ dsDNA         │ 147303 │ RefSeq        │\n",
      "│ 106 │ Herelleviridae │ dsDNA         │ 147590 │ RefSeq        │\n",
      "│ 107 │ Herelleviridae │ dsDNA         │ 148243 │ RefSeq        │\n",
      "│ 108 │ Herelleviridae │ dsDNA         │ 148317 │ RefSeq        │\n",
      "│ 109 │ Herelleviridae │ dsDNA         │ 148481 │ RefSeq        │\n",
      "│ 110 │ Herelleviridae │ dsDNA         │ 148564 │ RefSeq        │\n",
      "│ 111 │ Herelleviridae │ dsDNA         │ 149229 │ RefSeq        │\n",
      "\n",
      "│ Row │ Nuc_Completeness │ Genotype │ Segment │ Publications │\n",
      "│     │ \u001b[90mString\u001b[39m           │ \u001b[90mString\u001b[39m   │ \u001b[90mString\u001b[39m  │ \u001b[90mString\u001b[39m       │\n",
      "├─────┼──────────────────┼──────────┼─────────┼──────────────┤\n",
      "│ 1   │ complete         │          │         │ 1            │\n",
      "│ 2   │ complete         │          │         │ 1            │\n",
      "│ 3   │ complete         │          │         │ 1            │\n",
      "│ 4   │ complete         │          │         │              │\n",
      "│ 5   │ complete         │          │         │              │\n",
      "│ 6   │ complete         │          │         │              │\n",
      "│ 7   │ complete         │          │         │              │\n",
      "│ 8   │ complete         │          │         │              │\n",
      "│ 9   │ complete         │          │         │ 1            │\n",
      "│ 10  │ complete         │          │         │ 1            │\n",
      "⋮\n",
      "│ 101 │ complete         │          │         │ 1            │\n",
      "│ 102 │ complete         │          │         │              │\n",
      "│ 103 │ complete         │          │         │              │\n",
      "│ 104 │ complete         │          │         │ 1            │\n",
      "│ 105 │ complete         │          │         │              │\n",
      "│ 106 │ complete         │          │         │ 1            │\n",
      "│ 107 │ complete         │          │         │ 1            │\n",
      "│ 108 │ complete         │          │         │ 1            │\n",
      "│ 109 │ complete         │          │         │ 1            │\n",
      "│ 110 │ complete         │          │         │ 1            │\n",
      "│ 111 │ complete         │          │         │              │\n",
      "\n",
      "│ Row │ Geo_Location                        │ USA    │\n",
      "│     │ \u001b[90mString\u001b[39m                              │ \u001b[90mString\u001b[39m │\n",
      "├─────┼─────────────────────────────────────┼────────┤\n",
      "│ 1   │                                     │        │\n",
      "│ 2   │                                     │        │\n",
      "│ 3   │ South Korea                         │        │\n",
      "│ 4   │                                     │        │\n",
      "│ 5   │                                     │        │\n",
      "│ 6   │ Canada                              │        │\n",
      "│ 7   │ USA                                 │        │\n",
      "│ 8   │ China                               │        │\n",
      "│ 9   │                                     │        │\n",
      "│ 10  │ South Korea                         │        │\n",
      "⋮\n",
      "│ 101 │                                     │        │\n",
      "│ 102 │                                     │        │\n",
      "│ 103 │                                     │        │\n",
      "│ 104 │                                     │        │\n",
      "│ 105 │ South Korea                         │        │\n",
      "│ 106 │                                     │        │\n",
      "│ 107 │                                     │        │\n",
      "│ 108 │                                     │        │\n",
      "│ 109 │                                     │        │\n",
      "│ 110 │                                     │        │\n",
      "│ 111 │ China: Baoshan District of Shanghai │        │\n",
      "\n",
      "│ Row │ Host                                       │ Isolation_Source │\n",
      "│     │ \u001b[90mString\u001b[39m                                     │ \u001b[90mString\u001b[39m           │\n",
      "├─────┼────────────────────────────────────────────┼──────────────────┤\n",
      "│ 1   │ Staphylococcus aureus                      │                  │\n",
      "│ 2   │ Staphylococcus aureus                      │                  │\n",
      "│ 3   │ Staphylococcus aureus                      │                  │\n",
      "│ 4   │ Staphylococcus aureus                      │                  │\n",
      "│ 5   │ Staphylococcus aureus                      │                  │\n",
      "│ 6   │ Staphylococcus aureus                      │                  │\n",
      "│ 7   │ Staphylococcus aureus subsp. aureus RN4220 │                  │\n",
      "│ 8   │ Staphylococcus aureus                      │                  │\n",
      "│ 9   │ Staphylococcus aureus                      │                  │\n",
      "│ 10  │ Staphylococcus aureus                      │                  │\n",
      "⋮\n",
      "│ 101 │ Staphylococcus aureus                      │                  │\n",
      "│ 102 │ Staphylococcus aureus                      │                  │\n",
      "│ 103 │ Staphylococcus aureus                      │                  │\n",
      "│ 104 │ Staphylococcus aureus                      │                  │\n",
      "│ 105 │ Staphylococcus aureus                      │ feces            │\n",
      "│ 106 │ Staphylococcus aureus                      │                  │\n",
      "│ 107 │ Staphylococcus aureus                      │                  │\n",
      "│ 108 │ Staphylococcus aureus                      │                  │\n",
      "│ 109 │ Staphylococcus aureus                      │                  │\n",
      "│ 110 │ Staphylococcus aureus                      │                  │\n",
      "│ 111 │ Staphylococcus aureus                      │                  │\n",
      "\n",
      "│ Row │ Collection_Date │ BioSample │\n",
      "│     │ \u001b[90mString\u001b[39m          │ \u001b[90mString\u001b[39m    │\n",
      "├─────┼─────────────────┼───────────┤\n",
      "│ 1   │                 │           │\n",
      "│ 2   │                 │           │\n",
      "│ 3   │ 2017-07-30      │           │\n",
      "│ 4   │                 │           │\n",
      "│ 5   │                 │           │\n",
      "│ 6   │ 2002            │           │\n",
      "│ 7   │ 2015-08-25      │           │\n",
      "│ 8   │ 2013-10         │           │\n",
      "│ 9   │                 │           │\n",
      "│ 10  │                 │           │\n",
      "⋮\n",
      "│ 101 │                 │           │\n",
      "│ 102 │                 │           │\n",
      "│ 103 │                 │           │\n",
      "│ 104 │                 │           │\n",
      "│ 105 │                 │           │\n",
      "│ 106 │                 │           │\n",
      "│ 107 │                 │           │\n",
      "│ 108 │                 │           │\n",
      "│ 109 │                 │           │\n",
      "│ 110 │                 │           │\n",
      "│ 111 │                 │           │\n",
      "\n",
      "│ Row │ GenBank_Title                                           │\n",
      "│     │ \u001b[90mString\u001b[39m                                                  │\n",
      "├─────┼─────────────────────────────────────────────────────────┤\n",
      "│ 1   │ Staphylococcus phage PT1028, complete genome            │\n",
      "│ 2   │ Staphylococcus phage 44AHJD, complete genome            │\n",
      "│ 3   │ Staphylococcus phage CSA13, complete genome             │\n",
      "│ 4   │ Staphylococcus phage PSa3 complete sequence             │\n",
      "│ 5   │ Staphylococcus phage vB_SauP_phiAGO1.3, complete genome │\n",
      "│ 6   │ Staphylococcus phage BP39, complete genome              │\n",
      "│ 7   │ Staphylococcus phage Pabna, complete genome             │\n",
      "│ 8   │ Staphylococcus phage SLPW, complete genome              │\n",
      "│ 9   │ Staphylococcus phage GRCS, complete genome              │\n",
      "│ 10  │ Staphylococcus phage SAP-2, complete genome             │\n",
      "⋮\n",
      "│ 101 │ Staphylococcus phage MCE-2014, complete genome          │\n",
      "│ 102 │ Staphylococcus phage phiSA012 DNA, complete genome      │\n",
      "│ 103 │ Staphylococcus phage A5W, complete genome               │\n",
      "│ 104 │ Staphylococcus phage Staph1N, complete genome           │\n",
      "│ 105 │ Staphylococcus phage SA1, complete genome               │\n",
      "│ 106 │ Staphylococcus phage P4W, complete genome               │\n",
      "│ 107 │ Staphylococcus phage MSA6, complete genome              │\n",
      "│ 108 │ Staphylococcus phage K, complete genome                 │\n",
      "│ 109 │ Staphylococcus phage Fi200W, complete genome            │\n",
      "│ 110 │ Staphylococcus phage 676Z, complete genome              │\n",
      "│ 111 │ Staphylococcus phage phiSA_BS2, complete genome         │"
     ]
    }
   ],
   "source": [
    "show(ncbi_staph_phage_metadata, allcols=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "find_downstream_vertices (generic function with 1 method)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function find_downstream_vertices(kmer_graph, vertex, orientation)\n",
    "    viable_neighbors = Int[]\n",
    "    for neighbor in LightGraphs.neighbors(kmer_graph, vertex)\n",
    "        not_same_vertex = vertex != neighbor\n",
    "        candidate_edge = LightGraphs.Edge(vertex, neighbor)\n",
    "        edge_src_orientation = kmer_graph.eprops[candidate_edge][:orientations].source_orientation\n",
    "        viable_orientation = edge_src_orientation == orientation\n",
    "        if not_same_vertex && viable_orientation\n",
    "            push!(viable_neighbors, neighbor)\n",
    "        end\n",
    "    end\n",
    "    return viable_neighbors\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "find_unbranched_neighbors (generic function with 1 method)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function find_unbranched_neighbors(kmer_graph, vertex, orientation)\n",
    "    downstream_vertices = find_downstream_vertices(kmer_graph, vertex, orientation)\n",
    "    if length(downstream_vertices) == 1\n",
    "        downstream_vertex = first(downstream_vertices)\n",
    "        destination_orientation = kmer_graph.eprops[LightGraphs.Edge(vertex, downstream_vertex)][:orientations].destination_orientation\n",
    "        backtrack_vertices = find_downstream_vertices(kmer_graph, downstream_vertex, !destination_orientation)\n",
    "        if backtrack_vertices == [vertex]\n",
    "            return downstream_vertices\n",
    "        else\n",
    "            return Int[]\n",
    "        end\n",
    "    else\n",
    "        return Int[]\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "oriented_unbranching_walk (generic function with 1 method)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function oriented_unbranching_walk(kmer_graph, vertex, orientation)\n",
    "    walk = []\n",
    "    viable_neighbors = find_unbranched_neighbors(kmer_graph, vertex, orientation)\n",
    "    while length(viable_neighbors) == 1\n",
    "#         @show \"found a viable neighbor!!\"\n",
    "        viable_neighbor = first(viable_neighbors)\n",
    "        edge = LightGraphs.Edge(vertex, viable_neighbor)\n",
    "        push!(walk, edge)\n",
    "        vertex = edge.dst\n",
    "        orientation = kmer_graph.eprops[edge][:orientations].destination_orientation\n",
    "        viable_neighbors = find_unbranched_neighbors(kmer_graph, vertex, orientation)\n",
    "    end\n",
    "    return walk\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Primes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k = 11\n",
      "k = 13\n",
      "k = 17\n",
      "k = 19\n",
      "k = 23\n",
      "k = 29\n",
      "k = 31\n"
     ]
    }
   ],
   "source": [
    "for k in Primes.primes(11, 31)\n",
    "    @show k\n",
    "    KMER_TYPE = BioSequences.DNAMer{k}\n",
    "\n",
    "    EDGE_MER = BioSequences.DNAMer{k+1}\n",
    "\n",
    "    accession = ncbi_staph_phage_metadata[1, \"Accession\"]\n",
    "    fastx = collect(get_sequence(db = \"nuccore\", accession = accession))\n",
    "    kmers = Set(collect(keys(Eisenia.count_canonical_kmers(KMER_TYPE, fastx))))\n",
    "    # for accession in ncbi_staph_phage_metadata[2:end, \"Accession\"]\n",
    "    #     fastx = collect(get_sequence(db = \"nuccore\", accession = accession))\n",
    "    #     kmers = union!(kmers, collect(keys(Eisenia.count_canonical_kmers(KMER_TYPE, fastx))))\n",
    "    # end\n",
    "    # for kmer in kmers\n",
    "    #     push!(kmers, BioSequences.reverse_complement(kmer))\n",
    "    # end\n",
    "    kmers = unique(sort(collect(kmers)))\n",
    "\n",
    "    kmer_graph = MetaGraphs.MetaDiGraph(length(kmers))\n",
    "    MetaGraphs.set_prop!(kmer_graph, :k, k)\n",
    "\n",
    "    for (vertex, kmer) in enumerate(kmers)\n",
    "        MetaGraphs.set_prop!(kmer_graph, vertex, :sequence, kmer)\n",
    "    end\n",
    "\n",
    "    record = first(fastx)\n",
    "\n",
    "    sequence = FASTX.sequence(record)\n",
    "    record_identifier = FASTX.identifier(record) \n",
    "    edge_iterator = BioSequences.each(EDGE_MER, sequence)\n",
    "\n",
    "    for sequence_edge in edge_iterator\n",
    "\n",
    "        forward_sequence_edge = BioSequences.LongDNASeq(sequence_edge.fw)\n",
    "\n",
    "        observed_source_kmer = BioSequences.DNAMer(forward_sequence_edge[1:end-1])\n",
    "\n",
    "        observed_destination_kmer = BioSequences.DNAMer(forward_sequence_edge[2:end])\n",
    "\n",
    "        oriented_source_kmer = \n",
    "            (canonical_kmer = BioSequences.canonical(observed_source_kmer),\n",
    "             orientation = BioSequences.iscanonical(observed_source_kmer))\n",
    "\n",
    "        oriented_destination_kmer = \n",
    "            (canonical_kmer = BioSequences.canonical(observed_destination_kmer),\n",
    "             orientation = BioSequences.iscanonical(observed_destination_kmer))\n",
    "\n",
    "        oriented_source_vertex = \n",
    "            (vertex = searchsortedfirst(kmers, oriented_source_kmer.canonical_kmer),\n",
    "             orientation = oriented_source_kmer.orientation)\n",
    "\n",
    "        oriented_destination_vertex = \n",
    "            (vertex = searchsortedfirst(kmers, oriented_destination_kmer.canonical_kmer),\n",
    "             orientation = oriented_destination_kmer.orientation)\n",
    "\n",
    "        source_evidence = \n",
    "            (record = record_identifier,\n",
    "             index = sequence_edge.position,\n",
    "             orientation = oriented_source_vertex.orientation)\n",
    "\n",
    "        destination_evidence = \n",
    "            (record = record_identifier,\n",
    "             index = sequence_edge.position + 1,\n",
    "             orientation = oriented_destination_vertex.orientation)\n",
    "\n",
    "        add_evidence!(kmer_graph, oriented_source_vertex.vertex, source_evidence)\n",
    "\n",
    "        add_evidence!(kmer_graph, oriented_destination_vertex.vertex, destination_evidence)\n",
    "\n",
    "        forward_edge = LightGraphs.Edge(oriented_source_vertex.vertex, oriented_destination_vertex.vertex)\n",
    "\n",
    "        LightGraphs.add_edge!(kmer_graph, forward_edge)\n",
    "\n",
    "        forward_edge_orientations = \n",
    "            (source_orientation = oriented_source_vertex.orientation,\n",
    "             destination_orientation = oriented_destination_vertex.orientation)\n",
    "\n",
    "        MetaGraphs.set_prop!(kmer_graph, forward_edge, :orientations, forward_edge_orientations)\n",
    "\n",
    "        forward_edge_evidence = (\n",
    "            record = record_identifier,\n",
    "            index = sequence_edge.position,\n",
    "            orientation = true\n",
    "        )\n",
    "\n",
    "        add_evidence!(kmer_graph, forward_edge, forward_edge_evidence)\n",
    "\n",
    "        reverse_edge = LightGraphs.Edge(oriented_destination_vertex.vertex, oriented_source_vertex.vertex)\n",
    "\n",
    "        LightGraphs.add_edge!(kmer_graph, reverse_edge)\n",
    "\n",
    "        reverse_edge_orientations = \n",
    "            (source_orientation = !oriented_destination_vertex.orientation,\n",
    "             destination_orientation = !oriented_source_vertex.orientation)\n",
    "\n",
    "        MetaGraphs.set_prop!(kmer_graph, reverse_edge, :orientations, reverse_edge_orientations)\n",
    "\n",
    "        reverse_edge_evidence = (\n",
    "            record = record_identifier,\n",
    "            index = sequence_edge.position,\n",
    "            orientation = false\n",
    "        )\n",
    "\n",
    "        add_evidence!(kmer_graph, reverse_edge, reverse_edge_evidence)\n",
    "    end\n",
    "\n",
    "    # NOTE!! here 1 => 1 has both + => + and - => - but only the - is recorded because the value over-wrote the initial\n",
    "    # may need to make orientations a set\n",
    "    graph_to_gfa(graph=kmer_graph, outfile=\"test.$(k).gfa\")\n",
    "\n",
    "    untigs = []\n",
    "\n",
    "    if isempty(untigs)\n",
    "        visited = Int[]\n",
    "    else\n",
    "        visited = sort(unique(reduce(vcat, untigs)))\n",
    "    end\n",
    "    unvisited = setdiff(1:LightGraphs.nv(kmer_graph), visited)\n",
    "    while !isempty(unvisited)\n",
    "#     if !isempty(unvisited)\n",
    "        first_unvisited = first(unvisited)\n",
    "        forward_walk = oriented_unbranching_walk(kmer_graph, first_unvisited, true)\n",
    "        reverse_walk = oriented_unbranching_walk(kmer_graph, first_unvisited, false)\n",
    "        inverted_reverse_walk = [LightGraphs.Edge(e.dst, e.src) for e in reverse(reverse_walk)]\n",
    "        edge_path = vcat(inverted_reverse_walk, forward_walk)\n",
    "        if !isempty(edge_path)\n",
    "            untig = [first(edge_path).src, [e.dst for e in edge_path]...]\n",
    "        else\n",
    "            untig = [first_unvisited]\n",
    "        end\n",
    "        push!(untigs, untig)\n",
    "        visited = sort(unique(reduce(vcat, untigs)))\n",
    "        unvisited = setdiff(1:LightGraphs.nv(kmer_graph), visited)\n",
    "    end\n",
    "\n",
    "    oriented_untigs = []\n",
    "\n",
    "    for path in untigs\n",
    "        sequence = BioSequences.LongDNASeq(kmers[first(path)])\n",
    "        if length(path) == 1\n",
    "            orientations = [true]\n",
    "        elseif length(path) > 1\n",
    "            initial_edge = LightGraphs.Edge(path[1], path[2])\n",
    "            initial_orientation = kmer_graph.eprops[initial_edge][:orientations].source_orientation\n",
    "            orientations = [initial_orientation]\n",
    "            if !initial_orientation\n",
    "                sequence = BioSequences.reverse_complement(sequence)\n",
    "            end\n",
    "\n",
    "            for (src, dst) in zip(path[1:end-1], path[2:end])\n",
    "                edge = LightGraphs.Edge(src, dst)\n",
    "                destination = BioSequences.LongDNASeq(kmers[edge.dst])\n",
    "                destination_orientation = kmer_graph.eprops[edge][:orientations].destination_orientation\n",
    "                push!(orientations, destination_orientation)\n",
    "                if !destination_orientation\n",
    "                    destination = BioSequences.reverse_complement(destination)\n",
    "                end\n",
    "                sequence_suffix = sequence[end-length(destination)+2:end]\n",
    "                destination_prefix = destination[1:end-1]\n",
    "                @assert sequence_suffix == destination_prefix\n",
    "                push!(sequence, destination[end])\n",
    "            end\n",
    "        end\n",
    "\n",
    "        oriented_untig = \n",
    "        (\n",
    "            sequence = BioSequences.canonical(sequence),\n",
    "            path = BioSequences.iscanonical(sequence) ? path : reverse(path),\n",
    "            orientations = BioSequences.iscanonical(sequence) ? orientations : reverse(.!orientations)\n",
    "        )\n",
    "\n",
    "        push!(oriented_untigs, oriented_untig)\n",
    "    end\n",
    "\n",
    "    simplified_graph = MetaGraphs.MetaDiGraph(length(oriented_untigs))\n",
    "\n",
    "    MetaGraphs.set_prop!(simplified_graph, :k, k)\n",
    "\n",
    "    for (vertex, untig) in enumerate(oriented_untigs)\n",
    "        MetaGraphs.set_prop!(simplified_graph, vertex, :sequence, untig.sequence)\n",
    "        MetaGraphs.set_prop!(simplified_graph, vertex, :path, untig.path)\n",
    "        MetaGraphs.set_prop!(simplified_graph, vertex, :orientations, untig.orientations)\n",
    "    end\n",
    "    \n",
    "    simplified_untigs = []\n",
    "\n",
    "    for vertex in LightGraphs.vertices(simplified_graph)\n",
    "        in_kmer = simplified_graph.vprops[vertex][:path][1] => simplified_graph.vprops[vertex][:orientations][1]\n",
    "        out_kmer = simplified_graph.vprops[vertex][:path][end] => simplified_graph.vprops[vertex][:orientations][end]\n",
    "    #     @show vertex, in_kmer, out_kmer\n",
    "        push!(simplified_untigs, in_kmer => out_kmer)\n",
    "    end\n",
    "\n",
    "    for (ui, u) in enumerate(simplified_untigs)\n",
    "        for (vi, v) in enumerate(simplified_untigs)\n",
    "    #         + => +\n",
    "            source_kmer_index, source_orientation = last(u)\n",
    "            destination_kmer_index, destination_orientation = first(v)\n",
    "            edge = LightGraphs.Edge(source_kmer_index, destination_kmer_index)\n",
    "            if LightGraphs.has_edge(kmer_graph, edge)\n",
    "                source_orientation_matches = (kmer_graph.eprops[edge][:orientations].source_orientation == source_orientation)\n",
    "                destination_orientation_matches = (kmer_graph.eprops[edge][:orientations].destination_orientation == destination_orientation)\n",
    "                if source_orientation_matches && destination_orientation_matches\n",
    "#                     @show \"right orientation!! + +\"\n",
    "\n",
    "                    simplified_graph_edge = LightGraphs.Edge(ui, vi)\n",
    "\n",
    "                    LightGraphs.add_edge!(simplified_graph, simplified_graph_edge)\n",
    "                    edge_orientations = (\n",
    "                        source_orientation = source_orientation,\n",
    "                        destination_orientation = destination_orientation\n",
    "                    )\n",
    "                    MetaGraphs.set_prop!(simplified_graph, simplified_graph_edge, :orientations, edge_orientations)\n",
    "                end\n",
    "            end\n",
    "    #         + => -\n",
    "            source_kmer_index, source_orientation = last(u)\n",
    "            destination_kmer_index, destination_orientation = last(v)\n",
    "            destination_orientation = !destination_orientation\n",
    "\n",
    "            edge = LightGraphs.Edge(source_kmer_index, destination_kmer_index)\n",
    "            if LightGraphs.has_edge(kmer_graph, edge)\n",
    "                source_orientation_matches = (kmer_graph.eprops[edge][:orientations].source_orientation == source_orientation)\n",
    "                destination_orientation_matches = (kmer_graph.eprops[edge][:orientations].destination_orientation == destination_orientation)\n",
    "                if source_orientation_matches && destination_orientation_matches\n",
    "#                     @show \"right orientation!! + -\"\n",
    "                    simplified_graph_edge = LightGraphs.Edge(ui, vi)\n",
    "\n",
    "                    LightGraphs.add_edge!(simplified_graph, simplified_graph_edge)\n",
    "                    edge_orientations = (\n",
    "                        source_orientation = source_orientation,\n",
    "                        destination_orientation = destination_orientation\n",
    "                    )\n",
    "                    MetaGraphs.set_prop!(simplified_graph, simplified_graph_edge, :orientations, edge_orientations)\n",
    "                end\n",
    "            end\n",
    "    #         - => +\n",
    "            source_kmer_index, source_orientation = first(u)\n",
    "            source_orientation = !source_orientation\n",
    "            destination_kmer_index, destination_orientation = first(v)\n",
    "\n",
    "            edge = LightGraphs.Edge(source_kmer_index, destination_kmer_index)\n",
    "            if LightGraphs.has_edge(kmer_graph, edge)\n",
    "                source_orientation_matches = (kmer_graph.eprops[edge][:orientations].source_orientation == source_orientation)\n",
    "                destination_orientation_matches = (kmer_graph.eprops[edge][:orientations].destination_orientation == destination_orientation)\n",
    "                if source_orientation_matches && destination_orientation_matches\n",
    "#                     @show \"right orientation!! - +\"\n",
    "                    simplified_graph_edge = LightGraphs.Edge(ui, vi)\n",
    "\n",
    "                    LightGraphs.add_edge!(simplified_graph, simplified_graph_edge)\n",
    "                    edge_orientations = (\n",
    "                        source_orientation = source_orientation,\n",
    "                        destination_orientation = destination_orientation\n",
    "                    )\n",
    "                    MetaGraphs.set_prop!(simplified_graph, simplified_graph_edge, :orientations, edge_orientations)\n",
    "                end\n",
    "            end\n",
    "    #         - => -\n",
    "            source_kmer_index, source_orientation = first(u)\n",
    "            source_orientation = !source_orientation\n",
    "            destination_kmer_index, destination_orientation = last(v)\n",
    "            destination_orientation = !destination_orientation\n",
    "\n",
    "            edge = LightGraphs.Edge(source_kmer_index, destination_kmer_index)\n",
    "            if LightGraphs.has_edge(kmer_graph, edge)\n",
    "                source_orientation_matches = (kmer_graph.eprops[edge][:orientations].source_orientation == source_orientation)\n",
    "                destination_orientation_matches = (kmer_graph.eprops[edge][:orientations].destination_orientation == destination_orientation)\n",
    "                if source_orientation_matches && destination_orientation_matches\n",
    "#                     @show \"right orientation!! - -\"\n",
    "                    simplified_graph_edge = LightGraphs.Edge(ui, vi)\n",
    "\n",
    "                    LightGraphs.add_edge!(simplified_graph, simplified_graph_edge)\n",
    "                    edge_orientations = (\n",
    "                        source_orientation = source_orientation,\n",
    "                        destination_orientation = destination_orientation\n",
    "                    )\n",
    "                    MetaGraphs.set_prop!(simplified_graph, simplified_graph_edge, :orientations, edge_orientations)\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    graph_to_gfa(graph=simplified_graph, outfile=\"test.$(k).simplified.gfa\")\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13-element Array{Int64,1}:\n",
       " 13\n",
       " 17\n",
       " 19\n",
       " 23\n",
       " 29\n",
       " 31\n",
       " 37\n",
       " 41\n",
       " 43\n",
       " 47\n",
       " 53\n",
       " 59\n",
       " 61"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 11 is too small, minimum should be k=13\n",
    "Primes.primes(13, 61)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.3",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
