<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Mash Classification · Mycelia</title><meta name="title" content="Mash Classification · Mycelia"/><meta property="og:title" content="Mash Classification · Mycelia"/><meta property="twitter:title" content="Mash Classification · Mycelia"/><meta name="description" content="Documentation for Mycelia."/><meta property="og:description" content="Documentation for Mycelia."/><meta property="twitter:description" content="Documentation for Mycelia."/><script data-outdated-warner src="../../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../search_index.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../../">Mycelia</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../../">Home</a></li><li><a class="tocitem" href="../../../getting-started/">Getting Started</a></li><li><a class="tocitem" href="../../../installation/">Installation</a></li><li><a class="tocitem" href="../../../concepts/">Concepts</a></li><li><a class="tocitem" href="../../../workflow-map/">Workflow Map</a></li><li><a class="tocitem" href="../../../metagenomic-workflow/">Metagenomic Workflow</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../../../tutorials/">Overview</a></li><li><a class="tocitem" href="../01_data_acquisition/">Step 1: Data Acquisition</a></li><li><a class="tocitem" href="../02_quality_control/">Step 2: Quality Control</a></li><li><a class="tocitem" href="../03_kmer_analysis/">Step 3: K-mer Analysis</a></li><li><a class="tocitem" href="../04_genome_assembly/">Step 4: Genome Assembly</a></li><li><a class="tocitem" href="../05_assembly_validation/">Step 5: Assembly Validation</a></li><li><a class="tocitem" href="../06_gene_annotation/">Step 6: Gene Annotation</a></li><li><a class="tocitem" href="../07_comparative_genomics/">Step 7: Comparative Genomics</a></li><li><a class="tocitem" href="../08_tool_integration/">Step 8: Tool Integration</a></li><li><a class="tocitem" href="../14_binning_workflow/">Step 9: Binning Workflow</a></li><li><a class="tocitem" href="../04_graph_type_tutorials/">Graph Type Tutorials</a></li><li><a class="tocitem" href="../09_round_trip_01_string_graphs/">Round Trip 01 String Graphs</a></li><li><a class="tocitem" href="../09_round_trip_02_ngram_to_string/">Round Trip 02 Ngram To String</a></li><li><a class="tocitem" href="../09_round_trip_03_fasta_sequences/">Round Trip 03 Fasta Sequences</a></li><li><a class="tocitem" href="../09_round_trip_04_kmer_to_sequence/">Round Trip 04 Kmer To Sequence</a></li><li><a class="tocitem" href="../09_round_trip_05_fastq_graphs/">Round Trip 05 Fastq Graphs</a></li><li><a class="tocitem" href="../09_round_trip_06_qualmer_graphs/">Round Trip 06 Qualmer Graphs</a></li><li><a class="tocitem" href="../10_viroid_assembly_workflow/">Viroid Assembly Workflow</a></li><li><a class="tocitem" href="../11_reduced_amino_acid_alphabets/">Reduced Amino Acid Alphabets</a></li><li><a class="tocitem" href="../12_coverm_coverage/">Coverm Coverage</a></li><li><a class="tocitem" href="../13_rhizomorph_assembly/">Rhizomorph Assembly</a></li><li class="is-active"><a class="tocitem" href>Mash Classification</a><ul class="internal"><li><a class="tocitem" href="#Learning-Objectives"><span>Learning Objectives</span></a></li><li><a class="tocitem" href="#Setup"><span>Setup</span></a></li><li><a class="tocitem" href="#Part-1:-Sketching-Reference-Genomes"><span>Part 1: Sketching Reference Genomes</span></a></li><li><a class="tocitem" href="#Part-2:-Sketching-Reads"><span>Part 2: Sketching Reads</span></a></li><li><a class="tocitem" href="#Part-3:-Screening-Reads-Against-the-Reference-Database"><span>Part 3: Screening Reads Against the Reference Database</span></a></li><li><a class="tocitem" href="#Part-4:-Interpretation-Notes"><span>Part 4: Interpretation Notes</span></a></li></ul></li><li><a class="tocitem" href="../15_round_trip_benchmarking/">Round Trip Benchmarking</a></li><li><a class="tocitem" href="../advanced-assembly-theory-and-practice/">Advanced-Assembly-Theory-And-Practice</a></li><li><a class="tocitem" href="../run_all_tutorials/">Run All Tutorials</a></li></ul></li><li><span class="tocitem">API Reference</span><ul><li><a class="tocitem" href="../../../api/all-functions/">Complete API Surface</a></li></ul></li><li><a class="tocitem" href="../../../benchmarks/">Benchmarks</a></li><li><a class="tocitem" href="../../../contributing/">Contributing</a></li><li><a class="tocitem" href="../../../related-projects/">Related Projects</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorials</a></li><li class="is-active"><a href>Mash Classification</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Mash Classification</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/cjprybol/Mycelia" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/cjprybol/Mycelia/blob/master/tutorials/14_mash_classification.jl" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Tutorial-14:-Mash-Classification-and-Screening"><a class="docs-heading-anchor" href="#Tutorial-14:-Mash-Classification-and-Screening">Tutorial 14: Mash Classification and Screening</a><a id="Tutorial-14:-Mash-Classification-and-Screening-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial-14:-Mash-Classification-and-Screening" title="Permalink"></a></h1><p>This tutorial demonstrates how to use Mash sketches for quick containment and distance screening against reference sketches.</p><h2 id="Learning-Objectives"><a class="docs-heading-anchor" href="#Learning-Objectives">Learning Objectives</a><a id="Learning-Objectives-1"></a><a class="docs-heading-anchor-permalink" href="#Learning-Objectives" title="Permalink"></a></h2><p>By the end of this tutorial, you will understand:</p><ul><li>How to sketch reference genomes with Mash</li><li>How to sketch reads with Mash (<code>-r</code>, <code>-m</code>)</li><li>How to screen reads against a sketch database</li><li>How to interpret identity and shared-hash outputs</li></ul><h2 id="Setup"><a class="docs-heading-anchor" href="#Setup">Setup</a><a id="Setup-1"></a><a class="docs-heading-anchor-permalink" href="#Setup" title="Permalink"></a></h2><pre><code class="language-julia hljs">import Pkg
if isinteractive()
    Pkg.activate(&quot;..&quot;)
end

import Mycelia
import FASTX
import Random

Random.seed!(42)

println(&quot;=== Mash Classification Tutorial ===&quot;)</code></pre><h2 id="Part-1:-Sketching-Reference-Genomes"><a class="docs-heading-anchor" href="#Part-1:-Sketching-Reference-Genomes">Part 1: Sketching Reference Genomes</a><a id="Part-1:-Sketching-Reference-Genomes-1"></a><a class="docs-heading-anchor-permalink" href="#Part-1:-Sketching-Reference-Genomes" title="Permalink"></a></h2><pre><code class="language-julia hljs">println(&quot;\n--- Reference Sketching ---&quot;)

ref_records = [
    Mycelia.random_fasta_record(moltype=:DNA, seed=1, L=800),
    Mycelia.random_fasta_record(moltype=:DNA, seed=2, L=800),
    Mycelia.random_fasta_record(moltype=:DNA, seed=3, L=800)
]

ref_files = String[]
for (i, record) in enumerate(ref_records)
    filename = &quot;mash_reference_$(i).fasta&quot;
    Mycelia.write_fasta(outfile=filename, records=[record])
    push!(ref_files, filename)
end

if get(ENV, &quot;MYCELIA_RUN_EXTERNAL&quot;, &quot;false&quot;) == &quot;true&quot;
    reference_sketches = Mycelia.run_mash_sketch(
        input_files=ref_files,
        outdir=joinpath(pwd(), &quot;mash_sketches&quot;),
        k=21,
        s=1000
    ).sketches

    reference_db = Mycelia.run_mash_paste(
        out_file=joinpath(pwd(), &quot;mash_sketches&quot;, &quot;reference_db.msh&quot;),
        in_files=reference_sketches
    )

    println(&quot;Reference sketch database: $(reference_db)&quot;)
else
    println(&quot;Skipping Mash reference sketching; set MYCELIA_RUN_EXTERNAL=true to run.&quot;)
end</code></pre><h2 id="Part-2:-Sketching-Reads"><a class="docs-heading-anchor" href="#Part-2:-Sketching-Reads">Part 2: Sketching Reads</a><a id="Part-2:-Sketching-Reads-1"></a><a class="docs-heading-anchor-permalink" href="#Part-2:-Sketching-Reads" title="Permalink"></a></h2><pre><code class="language-julia hljs">println(&quot;\n--- Read Sketching ---&quot;)

reads = Mycelia.create_test_reads(String(FASTX.sequence(ref_records[1])), 25, 0.01)
reads_fastq = &quot;mash_reads.fastq&quot;
Mycelia.write_fastq(records=reads, filename=reads_fastq)

if get(ENV, &quot;MYCELIA_RUN_EXTERNAL&quot;, &quot;false&quot;) == &quot;true&quot;
    read_sketch = Mycelia.run_mash_sketch(
        input_files=[reads_fastq],
        outdir=joinpath(pwd(), &quot;mash_sketches&quot;),
        k=21,
        s=1000,
        r=true,
        min_copies=2
    ).sketches[1]
    println(&quot;Read sketch: $(read_sketch)&quot;)
else
    println(&quot;Skipping Mash execution; set MYCELIA_RUN_EXTERNAL=true to run sketching.&quot;)
end</code></pre><h2 id="Part-3:-Screening-Reads-Against-the-Reference-Database"><a class="docs-heading-anchor" href="#Part-3:-Screening-Reads-Against-the-Reference-Database">Part 3: Screening Reads Against the Reference Database</a><a id="Part-3:-Screening-Reads-Against-the-Reference-Database-1"></a><a class="docs-heading-anchor-permalink" href="#Part-3:-Screening-Reads-Against-the-Reference-Database" title="Permalink"></a></h2><pre><code class="language-julia hljs">println(&quot;\n--- Screening ---&quot;)

if get(ENV, &quot;MYCELIA_RUN_EXTERNAL&quot;, &quot;false&quot;) == &quot;true&quot;
    reference_db = joinpath(pwd(), &quot;mash_sketches&quot;, &quot;reference_db.msh&quot;)
    screen_result = Mycelia.run_mash_screen(
        reference=reference_db,
        query=reads_fastq,
        outdir=joinpath(pwd(), &quot;mash_screen&quot;),
        winner_takes_all=true
    )

    screen_table = Mycelia.parse_mash_screen_output(screen_result.results_tsv)
    println(screen_table)
else
    println(&quot;Skipping Mash screening; set MYCELIA_RUN_EXTERNAL=true to run screening.&quot;)
end</code></pre><h2 id="Part-4:-Interpretation-Notes"><a class="docs-heading-anchor" href="#Part-4:-Interpretation-Notes">Part 4: Interpretation Notes</a><a id="Part-4:-Interpretation-Notes-1"></a><a class="docs-heading-anchor-permalink" href="#Part-4:-Interpretation-Notes" title="Permalink"></a></h2><ul><li><strong>Identity</strong> is the estimated ANI-like similarity (1.0 is identical).</li><li><strong>Shared hashes</strong> are the number of shared MinHash sketches; more shared hashes generally imply stronger containment.</li><li>Combine Mash screening with sourmash and sylph to triangulate the most supported pangenome contexts before mapping with minimap2.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../13_rhizomorph_assembly/">« Rhizomorph Assembly</a><a class="docs-footer-nextpage" href="../15_round_trip_benchmarking/">Round Trip Benchmarking »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Monday 12 January 2026 14:02">Monday 12 January 2026</span>. Using Julia version 1.10.10.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
