{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d721fa48-b788-4576-9cab-b93b33daf55e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m  Activating\u001b[22m\u001b[39m environment at `~/work/Mycelia/projects/viral-pangenome-discovery/notebooks/Project.toml`\n",
      "\u001b[32m\u001b[1m    Updating\u001b[22m\u001b[39m registry at `/opt/julia/registries/General`\n",
      "\u001b[32m\u001b[1m   Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/work/Mycelia/projects/viral-pangenome-discovery/notebooks/Project.toml`\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/work/Mycelia/projects/viral-pangenome-discovery/notebooks/Manifest.toml`\n",
      "\u001b[32m\u001b[1m   Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/work/Mycelia/projects/viral-pangenome-discovery/notebooks/Project.toml`\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/work/Mycelia/projects/viral-pangenome-discovery/notebooks/Manifest.toml`\n"
     ]
    }
   ],
   "source": [
    "import Pkg\n",
    "Pkg.activate(\".\")\n",
    "\n",
    "# not for the faint of heart!\n",
    "# Pkg.update()\n",
    "\n",
    "pkgs = [\n",
    "\"ProgressMeter\",\n",
    "]\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "\n",
    "Pkg.develop(path=\"../../..\")\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fd121cfc-8a38-48ff-86f1-ef01f895353d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "download_and_filter_reads (generic function with 1 method)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function download_and_filter_reads(;outdir=\"\", srr_identifier=\"\")\n",
    "    \n",
    "    forward_reads = joinpath(outdir, \"$(srr_identifier)_1.fastq\")\n",
    "    reverse_reads = joinpath(outdir, \"$(srr_identifier)_2.fastq\")\n",
    "    forward_reads_gz = forward_reads * \".gz\"\n",
    "    reverse_reads_gz = reverse_reads * \".gz\"\n",
    "    trimmed_forward_reads = joinpath(outdir, \"trim_galore\", \"$(srr_identifier)_1_val_1.fq.gz\")\n",
    "    trimmed_reverse_reads = joinpath(outdir, \"trim_galore\", \"$(srr_identifier)_2_val_2.fq.gz\")\n",
    "\n",
    "    if !(isfile(trimmed_forward_reads) && isfile(trimmed_reverse_reads))\n",
    "        fasterq_dump(outdir=outdir, srr_identifier=srr_identifier)\n",
    "        trim_galore(outdir=outdir, identifier=srr_identifier)\n",
    "    end\n",
    "    isfile(forward_reads_gz) && rm(forward_reads_gz)\n",
    "    isfile(reverse_reads_gz) && rm(reverse_reads_gz)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a0c2663a-7484-41fa-974b-face2309f6d3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "fasterq_dump (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function fasterq_dump(;outdir=\"\", srr_identifier=\"\")\n",
    "    \n",
    "    forward_reads = joinpath(outdir, \"$(srr_identifier)_1.fastq\")\n",
    "    reverse_reads = joinpath(outdir, \"$(srr_identifier)_2.fastq\")\n",
    "    \n",
    "    forward_reads_gz = forward_reads * \".gz\"\n",
    "    reverse_reads_gz = reverse_reads * \".gz\"\n",
    "    \n",
    "    if !isfile(forward_reads_gz) && !isfile(reverse_reads_gz)\n",
    "        fasterq_dump_cmd = `\n",
    "            conda run --live-stream -n viral-pangenome-discovery\n",
    "            fasterq-dump\n",
    "                --outdir $(outdir)\n",
    "                --mem 1G\n",
    "                --split-3\n",
    "                --threads $(Sys.CPU_THREADS)\n",
    "                --skip-technical\n",
    "                $(srr_identifier)`\n",
    "        @time run(fasterq_dump_cmd)\n",
    "        run(`pigz $(forward_reads)`)\n",
    "        run(`pigz $(reverse_reads)`)\n",
    "    else\n",
    "        @info \"$(forward_reads_gz) & $(reverse_reads_gz) already present\"\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "92d4aedc-4861-4d48-9276-4c222799c25e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "trim_galore (generic function with 1 method)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function trim_galore(;outdir=\"\", identifier=\"\")\n",
    "    \n",
    "    trim_galore_dir = joinpath(outdir, \"trim_galore\")\n",
    "    \n",
    "    forward_reads = joinpath(outdir, \"$(identifier)_1.fastq.gz\")\n",
    "    reverse_reads = joinpath(outdir, \"$(identifier)_2.fastq.gz\")\n",
    "    \n",
    "    trimmed_forward_reads = joinpath(trim_galore_dir, \"$(identifier)_1_val_1.fq.gz\")\n",
    "    trimmed_reverse_reads = joinpath(trim_galore_dir, \"$(identifier)_2_val_2.fq.gz\")\n",
    "    \n",
    "    if !isfile(trimmed_forward_reads) && !isfile(trimmed_reverse_reads)\n",
    "        cmd = `conda run --live-stream -n viral-pangenome-discovery trim_galore --cores $(Sys.CPU_THREADS) --output_dir $(trim_galore_dir) --paired $(forward_reads) $(reverse_reads)`\n",
    "        # cmd = `conda run -n viral-pangenome-discovery trim_galore --output_dir $(trim_galore_dir) --paired $(forward_reads) $(reverse_reads)`\n",
    "        run(cmd)\n",
    "    else\n",
    "        @info \"$(trimmed_forward_reads) & $(trimmed_reverse_reads) already present\"\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a16c3f90-5e31-46fa-ae9b-535e84a952bc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "594-element Vector{String}:\n",
       " \"SRR6399459\"\n",
       " \"SRR6399460\"\n",
       " \"SRR6399461\"\n",
       " \"SRR6399462\"\n",
       " \"SRR6399463\"\n",
       " \"SRR6399464\"\n",
       " \"SRR6399465\"\n",
       " \"SRR6399466\"\n",
       " \"SRR6399467\"\n",
       " \"SRR6399468\"\n",
       " \"SRR6399469\"\n",
       " \"SRR6399470\"\n",
       " \"SRR6399471\"\n",
       " ⋮\n",
       " \"SRR7365474\"\n",
       " \"SRR7365475\"\n",
       " \"SRR7365476\"\n",
       " \"SRR7365477\"\n",
       " \"SRR7365478\"\n",
       " \"SRR7365479\"\n",
       " \"SRR7365480\"\n",
       " \"SRR7365481\"\n",
       " \"SRR7365482\"\n",
       " \"SRR7365483\"\n",
       " \"SRR7365484\"\n",
       " \"SRR7365485\""
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srr_list = \"$(dirname(pwd()))/metadata/exposome/SraAccList.txt\"\n",
    "srr_identifiers = readlines(srr_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9767c539-d540-498f-8ff2-55b450287543",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   0%|▏                                        |  ETA: 0:06:38\u001b[39mspots read      : 75,614,307\n",
      "reads read      : 151,228,614\n",
      "reads written   : 151,228,614\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4503.502431 seconds (138.66 k allocations: 3.946 MiB)\n",
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/SRR6399464_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count smallRNA: 0, count Nextera: 0, count Illumina: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/trim_galore/SRR6399464_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/SRR6399464_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399464_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/SRR6399464_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "60000000 sequences processed\n",
      "70000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/SRR6399464_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 266.712 s (3.527 µs/read; 17.01 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              75,614,307\n",
      "Reads with adapters:                23,584,888 (31.2%)\n",
      "Reads written (passing filters):    75,614,307 (100.0%)\n",
      "\n",
      "Total basepairs processed: 11,290,247,809 bp\n",
      "Quality-trimmed:                 107,338 bp (0.0%)\n",
      "Total written (filtered):  11,259,708,333 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 23584888 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 23.6%\n",
      "  C: 37.0%\n",
      "  G: 15.6%\n",
      "  T: 23.8%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t18619680\t18903576.8\t0\t18619680\n",
      "2\t3567034\t4725894.2\t0\t3567034\n",
      "3\t1049208\t1181473.5\t0\t1049208\n",
      "4\t262738\t295368.4\t0\t262738\n",
      "5\t52752\t73842.1\t0\t52752\n",
      "6\t20067\t18460.5\t0\t20067\n",
      "7\t13356\t4615.1\t0\t13356\n",
      "9\t12\t288.4\t0\t1 11\n",
      "10\t8\t72.1\t1\t0 8\n",
      "11\t1\t18.0\t1\t0 1\n",
      "14\t1\t1.1\t1\t0 1\n",
      "17\t1\t1.1\t1\t0 1\n",
      "18\t2\t1.1\t1\t0 2\n",
      "26\t1\t1.1\t1\t0 1\n",
      "30\t4\t1.1\t1\t0 4\n",
      "31\t1\t1.1\t1\t0 1\n",
      "36\t1\t1.1\t1\t0 1\n",
      "39\t1\t1.1\t1\t0 1\n",
      "40\t1\t1.1\t1\t0 1\n",
      "44\t1\t1.1\t1\t0 1\n",
      "52\t1\t1.1\t1\t0 1\n",
      "53\t1\t1.1\t1\t0 1\n",
      "79\t2\t1.1\t1\t0 2\n",
      "80\t1\t1.1\t1\t0 1\n",
      "83\t2\t1.1\t1\t0 2\n",
      "84\t1\t1.1\t1\t0 1\n",
      "85\t1\t1.1\t1\t0 1\n",
      "87\t1\t1.1\t1\t0 1\n",
      "90\t1\t1.1\t1\t0 1\n",
      "91\t2\t1.1\t1\t0 2\n",
      "100\t1\t1.1\t1\t0 1\n",
      "103\t1\t1.1\t1\t0 1\n",
      "104\t1\t1.1\t1\t0 1\n",
      "107\t2\t1.1\t1\t0 2\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/SRR6399464_1.fastq.gz\n",
      "=============================================\n",
      "75614307 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/trim_galore/SRR6399464_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/SRR6399464_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399464_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/SRR6399464_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "60000000 sequences processed\n",
      "70000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/SRR6399464_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 211.060 s (2.791 µs/read; 21.50 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              75,614,307\n",
      "Reads with adapters:                23,095,879 (30.5%)\n",
      "Reads written (passing filters):    75,614,307 (100.0%)\n",
      "\n",
      "Total basepairs processed: 10,778,537,022 bp\n",
      "Quality-trimmed:                 483,513 bp (0.0%)\n",
      "Total written (filtered):  10,748,078,745 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 23095879 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 25.7%\n",
      "  C: 38.8%\n",
      "  G: 14.7%\n",
      "  T: 20.8%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t18096083\t18903576.8\t0\t18096083\n",
      "2\t3673221\t4725894.2\t0\t3673221\n",
      "3\t980000\t1181473.5\t0\t980000\n",
      "4\t228553\t295368.4\t0\t228553\n",
      "5\t61271\t73842.1\t0\t61271\n",
      "6\t29450\t18460.5\t0\t29450\n",
      "7\t27201\t4615.1\t0\t27201\n",
      "8\t1\t1153.8\t0\t1\n",
      "9\t14\t288.4\t0\t0 14\n",
      "10\t17\t72.1\t1\t2 15\n",
      "11\t7\t18.0\t1\t0 7\n",
      "12\t2\t4.5\t1\t0 2\n",
      "13\t1\t1.1\t1\t0 1\n",
      "15\t1\t1.1\t1\t0 1\n",
      "18\t1\t1.1\t1\t0 1\n",
      "19\t1\t1.1\t1\t0 1\n",
      "23\t1\t1.1\t1\t0 1\n",
      "25\t1\t1.1\t1\t0 1\n",
      "27\t1\t1.1\t1\t0 1\n",
      "33\t1\t1.1\t1\t0 1\n",
      "34\t2\t1.1\t1\t0 2\n",
      "35\t3\t1.1\t1\t0 3\n",
      "47\t1\t1.1\t1\t0 1\n",
      "52\t1\t1.1\t1\t0 1\n",
      "57\t1\t1.1\t1\t0 1\n",
      "62\t1\t1.1\t1\t0 1\n",
      "65\t1\t1.1\t1\t0 1\n",
      "66\t2\t1.1\t1\t0 2\n",
      "68\t1\t1.1\t1\t0 1\n",
      "69\t2\t1.1\t1\t0 2\n",
      "73\t2\t1.1\t1\t0 2\n",
      "75\t2\t1.1\t1\t0 2\n",
      "76\t1\t1.1\t1\t0 1\n",
      "77\t2\t1.1\t1\t0 2\n",
      "78\t4\t1.1\t1\t0 4\n",
      "79\t2\t1.1\t1\t0 2\n",
      "80\t1\t1.1\t1\t0 1\n",
      "81\t1\t1.1\t1\t0 1\n",
      "89\t1\t1.1\t1\t0 1\n",
      "90\t3\t1.1\t1\t0 3\n",
      "91\t4\t1.1\t1\t0 4\n",
      "92\t2\t1.1\t1\t0 2\n",
      "94\t2\t1.1\t1\t0 2\n",
      "95\t2\t1.1\t1\t0 2\n",
      "101\t2\t1.1\t1\t0 2\n",
      "102\t1\t1.1\t1\t0 1\n",
      "106\t1\t1.1\t1\t0 1\n",
      "107\t2\t1.1\t1\t0 2\n",
      "127\t1\t1.1\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464/SRR6399464_2.fastq.gz\n",
      "=============================================\n",
      "75614307 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399464_1_trimmed.fq.gz and SRR6399464_2_trimmed.fq.gz\n",
      "file_1: SRR6399464_1_trimmed.fq.gz, file_2: SRR6399464_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399464_1_trimmed.fq.gz and SRR6399464_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399464_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399464_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 75614307\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 6169 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399464_1_trimmed.fq.gz and SRR6399464_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   1%|▍                                        |  ETA: 7 days, 23:49:17\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2531.293895 seconds (77.99 k allocations: 2.220 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 39,660,282\n",
      "reads read      : 79,320,564\n",
      "reads written   : 79,320,564\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/SRR6399465_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count smallRNA: 0, count Illumina: 0, count Nextera: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/trim_galore/SRR6399465_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/SRR6399465_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399465_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/SRR6399465_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/SRR6399465_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 99.590 s (2.511 µs/read; 23.89 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              39,660,282\n",
      "Reads with adapters:                11,969,648 (30.2%)\n",
      "Reads written (passing filters):    39,660,282 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,934,207,089 bp\n",
      "Quality-trimmed:                  29,438 bp (0.0%)\n",
      "Total written (filtered):  5,917,732,753 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 11969648 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 30.7%\n",
      "  C: 24.6%\n",
      "  G: 20.5%\n",
      "  T: 24.2%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t8467190\t9915070.5\t0\t8467190\n",
      "2\t2786932\t2478767.6\t0\t2786932\n",
      "3\t551312\t619691.9\t0\t551312\n",
      "4\t110093\t154923.0\t0\t110093\n",
      "5\t28039\t38730.7\t0\t28039\n",
      "6\t16024\t9682.7\t0\t16024\n",
      "7\t10018\t2420.7\t0\t10018\n",
      "8\t1\t605.2\t0\t1\n",
      "9\t2\t151.3\t0\t1 1\n",
      "10\t1\t37.8\t1\t0 1\n",
      "18\t1\t0.6\t1\t0 1\n",
      "23\t1\t0.6\t1\t0 1\n",
      "30\t5\t0.6\t1\t0 5\n",
      "31\t1\t0.6\t1\t0 1\n",
      "40\t1\t0.6\t1\t0 1\n",
      "46\t1\t0.6\t1\t0 1\n",
      "56\t2\t0.6\t1\t0 2\n",
      "58\t1\t0.6\t1\t0 1\n",
      "63\t2\t0.6\t1\t0 2\n",
      "70\t1\t0.6\t1\t0 1\n",
      "95\t1\t0.6\t1\t0 1\n",
      "102\t1\t0.6\t1\t0 1\n",
      "106\t2\t0.6\t1\t0 2\n",
      "110\t1\t0.6\t1\t0 1\n",
      "112\t2\t0.6\t1\t0 2\n",
      "113\t1\t0.6\t1\t0 1\n",
      "116\t2\t0.6\t1\t0 2\n",
      "119\t1\t0.6\t1\t0 1\n",
      "121\t2\t0.6\t1\t0 2\n",
      "122\t1\t0.6\t1\t0 1\n",
      "126\t1\t0.6\t1\t0 1\n",
      "127\t1\t0.6\t1\t0 1\n",
      "131\t1\t0.6\t1\t0 1\n",
      "133\t1\t0.6\t1\t0 1\n",
      "134\t1\t0.6\t1\t0 1\n",
      "139\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/SRR6399465_1.fastq.gz\n",
      "=============================================\n",
      "39660282 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/trim_galore/SRR6399465_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/SRR6399465_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399465_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/SRR6399465_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/SRR6399465_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 109.077 s (2.750 µs/read; 21.82 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              39,660,282\n",
      "Reads with adapters:                10,623,892 (26.8%)\n",
      "Reads written (passing filters):    39,660,282 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,741,665,019 bp\n",
      "Quality-trimmed:                  90,856 bp (0.0%)\n",
      "Total written (filtered):  5,726,680,511 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 10623892 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.4%\n",
      "  C: 23.3%\n",
      "  G: 19.4%\n",
      "  T: 22.9%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t7244734\t9915070.5\t0\t7244734\n",
      "2\t2756857\t2478767.6\t0\t2756857\n",
      "3\t455462\t619691.9\t0\t455462\n",
      "4\t110288\t154923.0\t0\t110288\n",
      "5\t27246\t38730.7\t0\t27246\n",
      "6\t16280\t9682.7\t0\t16280\n",
      "7\t12991\t2420.7\t0\t12991\n",
      "9\t1\t151.3\t0\t0 1\n",
      "10\t1\t37.8\t1\t0 1\n",
      "12\t2\t2.4\t1\t1 1\n",
      "18\t1\t0.6\t1\t0 1\n",
      "24\t1\t0.6\t1\t0 1\n",
      "33\t1\t0.6\t1\t0 1\n",
      "38\t1\t0.6\t1\t0 1\n",
      "55\t2\t0.6\t1\t0 2\n",
      "67\t1\t0.6\t1\t0 1\n",
      "72\t1\t0.6\t1\t0 1\n",
      "74\t1\t0.6\t1\t0 1\n",
      "88\t1\t0.6\t1\t0 1\n",
      "94\t1\t0.6\t1\t0 1\n",
      "96\t1\t0.6\t1\t0 1\n",
      "97\t1\t0.6\t1\t0 1\n",
      "99\t1\t0.6\t1\t0 1\n",
      "100\t1\t0.6\t1\t0 1\n",
      "102\t1\t0.6\t1\t0 1\n",
      "106\t1\t0.6\t1\t0 1\n",
      "109\t1\t0.6\t1\t0 1\n",
      "112\t2\t0.6\t1\t0 2\n",
      "113\t1\t0.6\t1\t0 1\n",
      "115\t1\t0.6\t1\t0 1\n",
      "116\t1\t0.6\t1\t0 1\n",
      "119\t1\t0.6\t1\t0 1\n",
      "121\t1\t0.6\t1\t0 1\n",
      "122\t1\t0.6\t1\t0 1\n",
      "126\t2\t0.6\t1\t0 2\n",
      "131\t1\t0.6\t1\t0 1\n",
      "136\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465/SRR6399465_2.fastq.gz\n",
      "=============================================\n",
      "39660282 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399465_1_trimmed.fq.gz and SRR6399465_2_trimmed.fq.gz\n",
      "file_1: SRR6399465_1_trimmed.fq.gz, file_2: SRR6399465_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399465_1_trimmed.fq.gz and SRR6399465_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399465_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399465_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 39660282\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 4060 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399465_1_trimmed.fq.gz and SRR6399465_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   1%|▌                                        |  ETA: 10.08 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1684.624913 seconds (51.92 k allocations: 1.478 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 38,672,454\n",
      "reads read      : 77,344,908\n",
      "reads written   : 77,344,908\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/SRR6399466_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count Nextera: 0, count Illumina: 0, count smallRNA: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/trim_galore/SRR6399466_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/SRR6399466_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399466_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/SRR6399466_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/SRR6399466_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 88.844 s (2.297 µs/read; 26.12 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              38,672,454\n",
      "Reads with adapters:                11,919,934 (30.8%)\n",
      "Reads written (passing filters):    38,672,454 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,779,682,664 bp\n",
      "Quality-trimmed:                  26,621 bp (0.0%)\n",
      "Total written (filtered):  5,763,135,681 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 11919934 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 30.7%\n",
      "  C: 25.2%\n",
      "  G: 21.7%\n",
      "  T: 22.4%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t8454828\t9668113.5\t0\t8454828\n",
      "2\t2609943\t2417028.4\t0\t2609943\n",
      "3\t681546\t604257.1\t0\t681546\n",
      "4\t115302\t151064.3\t0\t115302\n",
      "5\t26347\t37766.1\t0\t26347\n",
      "6\t16442\t9441.5\t0\t16442\n",
      "7\t15452\t2360.4\t0\t15452\n",
      "8\t1\t590.1\t0\t1\n",
      "9\t51\t147.5\t0\t1 50\n",
      "10\t12\t36.9\t1\t5 7\n",
      "11\t2\t9.2\t1\t0 2\n",
      "27\t1\t0.6\t1\t0 1\n",
      "30\t1\t0.6\t1\t0 1\n",
      "43\t1\t0.6\t1\t0 1\n",
      "68\t1\t0.6\t1\t0 1\n",
      "90\t1\t0.6\t1\t0 1\n",
      "126\t1\t0.6\t1\t0 1\n",
      "127\t1\t0.6\t1\t0 1\n",
      "131\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/SRR6399466_1.fastq.gz\n",
      "=============================================\n",
      "38672454 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/trim_galore/SRR6399466_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/SRR6399466_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399466_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/SRR6399466_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/SRR6399466_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 109.772 s (2.839 µs/read; 21.14 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              38,672,454\n",
      "Reads with adapters:                11,860,265 (30.7%)\n",
      "Reads written (passing filters):    38,672,454 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,692,457,593 bp\n",
      "Quality-trimmed:                  80,771 bp (0.0%)\n",
      "Total written (filtered):  5,676,021,944 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 11860265 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.2%\n",
      "  C: 24.2%\n",
      "  G: 19.6%\n",
      "  T: 21.9%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t8500880\t9668113.5\t0\t8500880\n",
      "2\t2513899\t2417028.4\t0\t2513899\n",
      "3\t668708\t604257.1\t0\t668708\n",
      "4\t116134\t151064.3\t0\t116134\n",
      "5\t26265\t37766.1\t0\t26265\n",
      "6\t17385\t9441.5\t0\t17385\n",
      "7\t16931\t2360.4\t0\t16931\n",
      "8\t2\t590.1\t0\t2\n",
      "9\t35\t147.5\t0\t1 34\n",
      "10\t7\t36.9\t1\t0 7\n",
      "11\t3\t9.2\t1\t0 3\n",
      "12\t4\t2.3\t1\t0 4\n",
      "14\t1\t0.6\t1\t0 1\n",
      "24\t1\t0.6\t1\t0 1\n",
      "39\t1\t0.6\t1\t0 1\n",
      "49\t1\t0.6\t1\t0 1\n",
      "56\t1\t0.6\t1\t0 1\n",
      "71\t2\t0.6\t1\t0 2\n",
      "90\t1\t0.6\t1\t0 1\n",
      "106\t1\t0.6\t1\t0 1\n",
      "127\t1\t0.6\t1\t0 1\n",
      "128\t1\t0.6\t1\t0 1\n",
      "131\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466/SRR6399466_2.fastq.gz\n",
      "=============================================\n",
      "38672454 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399466_1_trimmed.fq.gz and SRR6399466_2_trimmed.fq.gz\n",
      "file_1: SRR6399466_1_trimmed.fq.gz, file_2: SRR6399466_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399466_1_trimmed.fq.gz and SRR6399466_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399466_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399466_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 38672454\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 2317 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399466_1_trimmed.fq.gz and SRR6399466_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   1%|▌                                        |  ETA: 10.93 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1576.473692 seconds (48.59 k allocations: 1.383 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 34,028,058\n",
      "reads read      : 68,056,116\n",
      "reads written   : 68,056,116\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/SRR6399467_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count Nextera: 0, count Illumina: 0, count smallRNA: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/trim_galore/SRR6399467_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/SRR6399467_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399467_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/SRR6399467_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/SRR6399467_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 69.612 s (2.046 µs/read; 29.33 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              34,028,058\n",
      "Reads with adapters:                11,570,815 (34.0%)\n",
      "Reads written (passing filters):    34,028,058 (100.0%)\n",
      "\n",
      "Total basepairs processed: 4,951,996,474 bp\n",
      "Quality-trimmed:                 158,446 bp (0.0%)\n",
      "Total written (filtered):  4,935,926,771 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 11570815 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.4%\n",
      "  C: 29.7%\n",
      "  G: 15.3%\n",
      "  T: 20.6%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t8306500\t8507014.5\t0\t8306500\n",
      "2\t2479768\t2126753.6\t0\t2479768\n",
      "3\t612691\t531688.4\t0\t612691\n",
      "4\t108216\t132922.1\t0\t108216\n",
      "5\t31692\t33230.5\t0\t31692\n",
      "6\t16244\t8307.6\t0\t16244\n",
      "7\t15581\t2076.9\t0\t15581\n",
      "8\t1\t519.2\t0\t1\n",
      "9\t7\t129.8\t0\t1 6\n",
      "10\t6\t32.5\t1\t2 4\n",
      "11\t5\t8.1\t1\t0 5\n",
      "12\t5\t2.0\t1\t3 2\n",
      "14\t1\t0.5\t1\t0 1\n",
      "18\t1\t0.5\t1\t0 1\n",
      "23\t1\t0.5\t1\t0 1\n",
      "30\t2\t0.5\t1\t0 2\n",
      "31\t5\t0.5\t1\t0 5\n",
      "36\t1\t0.5\t1\t0 1\n",
      "48\t1\t0.5\t1\t0 1\n",
      "62\t1\t0.5\t1\t0 1\n",
      "64\t6\t0.5\t1\t0 6\n",
      "65\t6\t0.5\t1\t0 6\n",
      "66\t1\t0.5\t1\t0 1\n",
      "67\t1\t0.5\t1\t0 1\n",
      "71\t1\t0.5\t1\t0 1\n",
      "72\t1\t0.5\t1\t0 1\n",
      "75\t1\t0.5\t1\t0 1\n",
      "76\t5\t0.5\t1\t0 5\n",
      "77\t1\t0.5\t1\t0 1\n",
      "80\t1\t0.5\t1\t0 1\n",
      "81\t1\t0.5\t1\t0 1\n",
      "82\t1\t0.5\t1\t0 1\n",
      "84\t2\t0.5\t1\t0 2\n",
      "85\t1\t0.5\t1\t0 1\n",
      "86\t1\t0.5\t1\t0 1\n",
      "88\t1\t0.5\t1\t0 1\n",
      "92\t1\t0.5\t1\t0 1\n",
      "93\t2\t0.5\t1\t0 2\n",
      "94\t1\t0.5\t1\t0 1\n",
      "96\t4\t0.5\t1\t0 4\n",
      "97\t3\t0.5\t1\t0 3\n",
      "98\t1\t0.5\t1\t0 1\n",
      "100\t1\t0.5\t1\t0 1\n",
      "103\t2\t0.5\t1\t0 2\n",
      "105\t2\t0.5\t1\t0 2\n",
      "106\t1\t0.5\t1\t0 1\n",
      "107\t1\t0.5\t1\t0 1\n",
      "108\t3\t0.5\t1\t0 3\n",
      "109\t1\t0.5\t1\t0 1\n",
      "112\t3\t0.5\t1\t0 3\n",
      "114\t1\t0.5\t1\t0 1\n",
      "115\t2\t0.5\t1\t0 2\n",
      "121\t2\t0.5\t1\t0 2\n",
      "122\t2\t0.5\t1\t0 2\n",
      "123\t1\t0.5\t1\t0 1\n",
      "124\t1\t0.5\t1\t0 1\n",
      "126\t2\t0.5\t1\t0 2\n",
      "128\t2\t0.5\t1\t0 2\n",
      "129\t3\t0.5\t1\t0 3\n",
      "130\t2\t0.5\t1\t0 2\n",
      "132\t1\t0.5\t1\t0 1\n",
      "133\t1\t0.5\t1\t0 1\n",
      "135\t1\t0.5\t1\t0 1\n",
      "136\t1\t0.5\t1\t0 1\n",
      "137\t1\t0.5\t1\t0 1\n",
      "138\t3\t0.5\t1\t0 3\n",
      "139\t3\t0.5\t1\t0 3\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/SRR6399467_1.fastq.gz\n",
      "=============================================\n",
      "34028058 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/trim_galore/SRR6399467_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/SRR6399467_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399467_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/SRR6399467_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/SRR6399467_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 76.604 s (2.251 µs/read; 26.65 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              34,028,058\n",
      "Reads with adapters:                10,532,685 (31.0%)\n",
      "Reads written (passing filters):    34,028,058 (100.0%)\n",
      "\n",
      "Total basepairs processed: 4,723,460,268 bp\n",
      "Quality-trimmed:                 377,813 bp (0.0%)\n",
      "Total written (filtered):  4,708,580,722 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 10532685 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 38.0%\n",
      "  C: 26.2%\n",
      "  G: 15.9%\n",
      "  T: 19.9%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t7560454\t8507014.5\t0\t7560454\n",
      "2\t2284981\t2126753.6\t0\t2284981\n",
      "3\t515166\t531688.4\t0\t515166\n",
      "4\t100374\t132922.1\t0\t100374\n",
      "5\t31545\t33230.5\t0\t31545\n",
      "6\t19341\t8307.6\t0\t19341\n",
      "7\t20730\t2076.9\t0\t20730\n",
      "8\t10\t519.2\t0\t10\n",
      "9\t1\t129.8\t0\t0 1\n",
      "10\t16\t32.5\t1\t4 12\n",
      "11\t6\t8.1\t1\t2 4\n",
      "12\t6\t2.0\t1\t2 4\n",
      "16\t1\t0.5\t1\t0 1\n",
      "24\t2\t0.5\t1\t0 2\n",
      "28\t1\t0.5\t1\t0 1\n",
      "44\t1\t0.5\t1\t0 1\n",
      "50\t1\t0.5\t1\t0 1\n",
      "62\t1\t0.5\t1\t0 1\n",
      "63\t1\t0.5\t1\t0 1\n",
      "70\t3\t0.5\t1\t0 3\n",
      "71\t1\t0.5\t1\t0 1\n",
      "74\t1\t0.5\t1\t0 1\n",
      "76\t4\t0.5\t1\t0 4\n",
      "77\t1\t0.5\t1\t0 1\n",
      "80\t1\t0.5\t1\t0 1\n",
      "81\t1\t0.5\t1\t0 1\n",
      "84\t3\t0.5\t1\t0 3\n",
      "85\t1\t0.5\t1\t0 1\n",
      "87\t1\t0.5\t1\t0 1\n",
      "88\t1\t0.5\t1\t0 1\n",
      "90\t1\t0.5\t1\t0 1\n",
      "92\t1\t0.5\t1\t0 1\n",
      "93\t1\t0.5\t1\t0 1\n",
      "96\t1\t0.5\t1\t0 1\n",
      "97\t1\t0.5\t1\t0 1\n",
      "100\t1\t0.5\t1\t0 1\n",
      "103\t1\t0.5\t1\t0 1\n",
      "107\t1\t0.5\t1\t0 1\n",
      "108\t3\t0.5\t1\t0 3\n",
      "109\t1\t0.5\t1\t0 1\n",
      "111\t2\t0.5\t1\t0 2\n",
      "115\t1\t0.5\t1\t0 1\n",
      "119\t1\t0.5\t1\t0 1\n",
      "121\t1\t0.5\t1\t0 1\n",
      "122\t2\t0.5\t1\t0 2\n",
      "126\t3\t0.5\t1\t0 3\n",
      "128\t1\t0.5\t1\t0 1\n",
      "129\t2\t0.5\t1\t0 2\n",
      "130\t1\t0.5\t1\t0 1\n",
      "135\t1\t0.5\t1\t0 1\n",
      "137\t2\t0.5\t1\t0 2\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467/SRR6399467_2.fastq.gz\n",
      "=============================================\n",
      "34028058 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399467_1_trimmed.fq.gz and SRR6399467_2_trimmed.fq.gz\n",
      "file_1: SRR6399467_1_trimmed.fq.gz, file_2: SRR6399467_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399467_1_trimmed.fq.gz and SRR6399467_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399467_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399467_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 34028058\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 5593 (0.02%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399467_1_trimmed.fq.gz and SRR6399467_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   2%|▋                                        |  ETA: 11.37 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2063.987612 seconds (63.60 k allocations: 1.811 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 47,822,570\n",
      "reads read      : 95,645,140\n",
      "reads written   : 95,645,140\n",
      "\n",
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/SRR6399468_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t1\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Using smallRNA adapter for trimming (count: 1). Second best hit was Nextera (count: 0)\n",
      "\n",
      "Reducing length cutoff to 18bp for small RNA-Seq reads because a cutoff of 20bp may remove some short species of small RNAs if they had been trimmed by 1,2 or 3bp\n",
      "Setting the Illumina smallRNA 5' adapter as adapter 2: 'GATCGTCGGACT'\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/trim_galore/SRR6399468_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/SRR6399468_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'TGGAATTCTCGG' (Illumina small RNA adapter; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Optional adapter 2 sequence (only used for read 2 of paired-end files): 'GATCGTCGGACT'\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 18 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399468_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'TGGAATTCTCGG' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/SRR6399468_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a TGGAATTCTCGG /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/SRR6399468_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 122.090 s (2.553 µs/read; 23.50 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              47,822,570\n",
      "Reads with adapters:                14,845,092 (31.0%)\n",
      "Reads written (passing filters):    47,822,570 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,097,287,663 bp\n",
      "Quality-trimmed:                  74,388 bp (0.0%)\n",
      "Total written (filtered):  7,073,060,596 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: TGGAATTCTCGG; Type: regular 3'; Length: 12; Trimmed: 14845092 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 20.9%\n",
      "  C: 28.9%\n",
      "  G: 24.2%\n",
      "  T: 26.1%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t10608201\t11955642.5\t0\t10608201\n",
      "2\t2957753\t2988910.6\t0\t2957753\n",
      "3\t971609\t747227.7\t0\t971609\n",
      "4\t164701\t186806.9\t0\t164701\n",
      "5\t47879\t46701.7\t0\t47879\n",
      "6\t17462\t11675.4\t0\t17462\n",
      "7\t11146\t2918.9\t0\t11146\n",
      "8\t4552\t729.7\t0\t4552\n",
      "9\t13152\t182.4\t0\t259 12893\n",
      "10\t11326\t45.6\t1\t6 11320\n",
      "11\t1017\t11.4\t1\t0 1017\n",
      "12\t771\t2.9\t1\t1 770\n",
      "13\t105\t2.9\t1\t0 105\n",
      "14\t83\t2.9\t1\t1 82\n",
      "15\t122\t2.9\t1\t0 122\n",
      "16\t238\t2.9\t1\t0 238\n",
      "17\t145\t2.9\t1\t0 145\n",
      "18\t121\t2.9\t1\t0 121\n",
      "19\t41\t2.9\t1\t0 41\n",
      "20\t156\t2.9\t1\t2 154\n",
      "21\t45\t2.9\t1\t1 44\n",
      "22\t53\t2.9\t1\t0 53\n",
      "23\t59\t2.9\t1\t0 59\n",
      "24\t68\t2.9\t1\t0 68\n",
      "25\t79\t2.9\t1\t0 79\n",
      "26\t62\t2.9\t1\t0 62\n",
      "27\t46\t2.9\t1\t1 45\n",
      "28\t76\t2.9\t1\t0 76\n",
      "29\t76\t2.9\t1\t0 76\n",
      "30\t70\t2.9\t1\t0 70\n",
      "31\t86\t2.9\t1\t0 86\n",
      "32\t131\t2.9\t1\t0 131\n",
      "33\t121\t2.9\t1\t0 121\n",
      "34\t85\t2.9\t1\t0 85\n",
      "35\t141\t2.9\t1\t0 141\n",
      "36\t216\t2.9\t1\t0 216\n",
      "37\t175\t2.9\t1\t1 174\n",
      "38\t358\t2.9\t1\t0 358\n",
      "39\t779\t2.9\t1\t1 778\n",
      "40\t459\t2.9\t1\t0 459\n",
      "41\t458\t2.9\t1\t0 458\n",
      "42\t788\t2.9\t1\t1 787\n",
      "43\t396\t2.9\t1\t1 395\n",
      "44\t353\t2.9\t1\t0 353\n",
      "45\t577\t2.9\t1\t0 577\n",
      "46\t672\t2.9\t1\t1 671\n",
      "47\t436\t2.9\t1\t0 436\n",
      "48\t1098\t2.9\t1\t0 1098\n",
      "49\t166\t2.9\t1\t1 165\n",
      "50\t135\t2.9\t1\t0 135\n",
      "51\t113\t2.9\t1\t0 113\n",
      "52\t169\t2.9\t1\t0 169\n",
      "53\t115\t2.9\t1\t0 115\n",
      "54\t116\t2.9\t1\t2 114\n",
      "55\t66\t2.9\t1\t0 66\n",
      "56\t109\t2.9\t1\t0 109\n",
      "57\t68\t2.9\t1\t0 68\n",
      "58\t72\t2.9\t1\t1 71\n",
      "59\t130\t2.9\t1\t2 128\n",
      "60\t107\t2.9\t1\t0 107\n",
      "61\t65\t2.9\t1\t0 65\n",
      "62\t55\t2.9\t1\t1 54\n",
      "63\t133\t2.9\t1\t0 133\n",
      "64\t104\t2.9\t1\t0 104\n",
      "65\t87\t2.9\t1\t0 87\n",
      "66\t173\t2.9\t1\t0 173\n",
      "67\t334\t2.9\t1\t0 334\n",
      "68\t198\t2.9\t1\t0 198\n",
      "69\t244\t2.9\t1\t0 244\n",
      "70\t474\t2.9\t1\t0 474\n",
      "71\t201\t2.9\t1\t0 201\n",
      "72\t142\t2.9\t1\t0 142\n",
      "73\t219\t2.9\t1\t1 218\n",
      "74\t393\t2.9\t1\t0 393\n",
      "75\t327\t2.9\t1\t0 327\n",
      "76\t547\t2.9\t1\t0 547\n",
      "77\t63\t2.9\t1\t0 63\n",
      "78\t134\t2.9\t1\t0 134\n",
      "79\t80\t2.9\t1\t0 80\n",
      "80\t153\t2.9\t1\t0 153\n",
      "81\t59\t2.9\t1\t0 59\n",
      "82\t130\t2.9\t1\t0 130\n",
      "83\t163\t2.9\t1\t0 163\n",
      "84\t122\t2.9\t1\t0 122\n",
      "85\t178\t2.9\t1\t1 177\n",
      "86\t106\t2.9\t1\t1 105\n",
      "87\t131\t2.9\t1\t0 131\n",
      "88\t75\t2.9\t1\t0 75\n",
      "89\t82\t2.9\t1\t1 81\n",
      "90\t112\t2.9\t1\t0 112\n",
      "91\t71\t2.9\t1\t0 71\n",
      "92\t55\t2.9\t1\t0 55\n",
      "93\t47\t2.9\t1\t0 47\n",
      "94\t73\t2.9\t1\t0 73\n",
      "95\t122\t2.9\t1\t0 122\n",
      "96\t171\t2.9\t1\t0 171\n",
      "97\t110\t2.9\t1\t0 110\n",
      "98\t56\t2.9\t1\t2 54\n",
      "99\t96\t2.9\t1\t0 96\n",
      "100\t55\t2.9\t1\t1 54\n",
      "101\t99\t2.9\t1\t0 99\n",
      "102\t162\t2.9\t1\t0 162\n",
      "103\t82\t2.9\t1\t0 82\n",
      "104\t97\t2.9\t1\t0 97\n",
      "105\t78\t2.9\t1\t0 78\n",
      "106\t71\t2.9\t1\t1 70\n",
      "107\t76\t2.9\t1\t3 73\n",
      "108\t47\t2.9\t1\t12 35\n",
      "109\t54\t2.9\t1\t4 50\n",
      "110\t89\t2.9\t1\t7 82\n",
      "111\t158\t2.9\t1\t15 143\n",
      "112\t164\t2.9\t1\t6 158\n",
      "113\t116\t2.9\t1\t1 115\n",
      "114\t148\t2.9\t1\t1 147\n",
      "115\t159\t2.9\t1\t9 150\n",
      "116\t102\t2.9\t1\t4 98\n",
      "117\t132\t2.9\t1\t12 120\n",
      "118\t203\t2.9\t1\t0 203\n",
      "119\t330\t2.9\t1\t2 328\n",
      "120\t665\t2.9\t1\t0 665\n",
      "121\t336\t2.9\t1\t2 334\n",
      "122\t520\t2.9\t1\t0 520\n",
      "123\t723\t2.9\t1\t1 722\n",
      "124\t413\t2.9\t1\t0 413\n",
      "125\t361\t2.9\t1\t1 360\n",
      "126\t450\t2.9\t1\t0 450\n",
      "127\t1304\t2.9\t1\t0 1304\n",
      "128\t1419\t2.9\t1\t3 1416\n",
      "129\t1465\t2.9\t1\t2 1463\n",
      "130\t747\t2.9\t1\t1 746\n",
      "131\t1102\t2.9\t1\t0 1102\n",
      "132\t688\t2.9\t1\t0 688\n",
      "133\t584\t2.9\t1\t0 584\n",
      "134\t687\t2.9\t1\t0 687\n",
      "135\t1035\t2.9\t1\t1 1034\n",
      "136\t1019\t2.9\t1\t1 1018\n",
      "137\t1135\t2.9\t1\t1 1134\n",
      "138\t234\t2.9\t1\t0 234\n",
      "139\t191\t2.9\t1\t0 191\n",
      "140\t229\t2.9\t1\t0 229\n",
      "141\t134\t2.9\t1\t0 134\n",
      "142\t84\t2.9\t1\t0 84\n",
      "143\t157\t2.9\t1\t2 155\n",
      "144\t94\t2.9\t1\t1 93\n",
      "145\t319\t2.9\t1\t1 318\n",
      "146\t62\t2.9\t1\t0 62\n",
      "147\t119\t2.9\t1\t0 119\n",
      "148\t84\t2.9\t1\t0 84\n",
      "149\t98\t2.9\t1\t1 97\n",
      "150\t171\t2.9\t1\t0 171\n",
      "151\t47\t2.9\t1\t1 46\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/SRR6399468_1.fastq.gz\n",
      "=============================================\n",
      "47822570 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/trim_galore/SRR6399468_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/SRR6399468_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'TGGAATTCTCGG' (Illumina small RNA adapter; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Optional adapter 2 sequence (only used for read 2 of paired-end files): 'GATCGTCGGACT'\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 18 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399468_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'GATCGTCGGACT' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/SRR6399468_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a GATCGTCGGACT /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/SRR6399468_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 132.002 s (2.760 µs/read; 21.74 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              47,822,570\n",
      "Reads with adapters:                15,924,700 (33.3%)\n",
      "Reads written (passing filters):    47,822,570 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,980,389,217 bp\n",
      "Quality-trimmed:                 169,738 bp (0.0%)\n",
      "Total written (filtered):  6,959,609,188 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: GATCGTCGGACT; Type: regular 3'; Length: 12; Trimmed: 15924700 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 23.4%\n",
      "  C: 18.5%\n",
      "  G: 34.7%\n",
      "  T: 23.4%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t12885937\t11955642.5\t0\t12885937\n",
      "2\t2302311\t2988910.6\t0\t2302311\n",
      "3\t546238\t747227.7\t0\t546238\n",
      "4\t119084\t186806.9\t0\t119084\n",
      "5\t53224\t46701.7\t0\t53224\n",
      "6\t4759\t11675.4\t0\t4759\n",
      "7\t1829\t2918.9\t0\t1829\n",
      "8\t147\t729.7\t0\t147\n",
      "9\t379\t182.4\t0\t11 368\n",
      "10\t130\t45.6\t1\t1 129\n",
      "11\t57\t11.4\t1\t1 56\n",
      "12\t19\t2.9\t1\t0 19\n",
      "13\t25\t2.9\t1\t0 25\n",
      "14\t23\t2.9\t1\t0 23\n",
      "15\t20\t2.9\t1\t0 20\n",
      "16\t20\t2.9\t1\t0 20\n",
      "17\t21\t2.9\t1\t0 21\n",
      "18\t35\t2.9\t1\t0 35\n",
      "19\t23\t2.9\t1\t0 23\n",
      "20\t25\t2.9\t1\t0 25\n",
      "21\t22\t2.9\t1\t0 22\n",
      "22\t12\t2.9\t1\t0 12\n",
      "23\t24\t2.9\t1\t0 24\n",
      "24\t27\t2.9\t1\t0 27\n",
      "25\t18\t2.9\t1\t0 18\n",
      "26\t27\t2.9\t1\t0 27\n",
      "27\t26\t2.9\t1\t0 26\n",
      "28\t21\t2.9\t1\t0 21\n",
      "29\t21\t2.9\t1\t0 21\n",
      "30\t27\t2.9\t1\t0 27\n",
      "31\t37\t2.9\t1\t0 37\n",
      "32\t21\t2.9\t1\t0 21\n",
      "33\t16\t2.9\t1\t0 16\n",
      "34\t30\t2.9\t1\t0 30\n",
      "35\t51\t2.9\t1\t0 51\n",
      "36\t37\t2.9\t1\t0 37\n",
      "37\t65\t2.9\t1\t0 65\n",
      "38\t143\t2.9\t1\t0 143\n",
      "39\t323\t2.9\t1\t0 323\n",
      "40\t644\t2.9\t1\t0 644\n",
      "41\t409\t2.9\t1\t0 409\n",
      "42\t407\t2.9\t1\t0 407\n",
      "43\t692\t2.9\t1\t0 692\n",
      "44\t424\t2.9\t1\t0 424\n",
      "45\t386\t2.9\t1\t0 386\n",
      "46\t221\t2.9\t1\t0 221\n",
      "47\t530\t2.9\t1\t0 530\n",
      "48\t291\t2.9\t1\t0 291\n",
      "49\t541\t2.9\t1\t0 541\n",
      "50\t50\t2.9\t1\t0 50\n",
      "51\t34\t2.9\t1\t0 34\n",
      "52\t27\t2.9\t1\t0 27\n",
      "53\t89\t2.9\t1\t0 89\n",
      "54\t37\t2.9\t1\t0 37\n",
      "55\t92\t2.9\t1\t0 92\n",
      "56\t42\t2.9\t1\t0 42\n",
      "57\t45\t2.9\t1\t0 45\n",
      "58\t31\t2.9\t1\t0 31\n",
      "59\t24\t2.9\t1\t0 24\n",
      "60\t26\t2.9\t1\t0 26\n",
      "61\t30\t2.9\t1\t0 30\n",
      "62\t32\t2.9\t1\t0 32\n",
      "63\t120\t2.9\t1\t0 120\n",
      "64\t115\t2.9\t1\t0 115\n",
      "65\t95\t2.9\t1\t0 95\n",
      "66\t80\t2.9\t1\t0 80\n",
      "67\t124\t2.9\t1\t0 124\n",
      "68\t70\t2.9\t1\t0 70\n",
      "69\t56\t2.9\t1\t0 56\n",
      "70\t66\t2.9\t1\t0 66\n",
      "71\t114\t2.9\t1\t0 114\n",
      "72\t73\t2.9\t1\t0 73\n",
      "73\t112\t2.9\t1\t0 112\n",
      "74\t55\t2.9\t1\t0 55\n",
      "75\t30\t2.9\t1\t0 30\n",
      "76\t30\t2.9\t1\t0 30\n",
      "77\t28\t2.9\t1\t0 28\n",
      "78\t35\t2.9\t1\t0 35\n",
      "79\t32\t2.9\t1\t0 32\n",
      "80\t53\t2.9\t1\t0 53\n",
      "81\t26\t2.9\t1\t0 26\n",
      "82\t45\t2.9\t1\t0 45\n",
      "83\t45\t2.9\t1\t0 45\n",
      "84\t45\t2.9\t1\t0 45\n",
      "85\t56\t2.9\t1\t0 56\n",
      "86\t27\t2.9\t1\t0 27\n",
      "87\t59\t2.9\t1\t0 59\n",
      "88\t44\t2.9\t1\t0 44\n",
      "89\t38\t2.9\t1\t0 38\n",
      "90\t32\t2.9\t1\t0 32\n",
      "91\t29\t2.9\t1\t0 29\n",
      "92\t25\t2.9\t1\t0 25\n",
      "93\t39\t2.9\t1\t0 39\n",
      "94\t24\t2.9\t1\t0 24\n",
      "95\t77\t2.9\t1\t0 77\n",
      "96\t47\t2.9\t1\t0 47\n",
      "97\t94\t2.9\t1\t0 94\n",
      "98\t132\t2.9\t1\t0 132\n",
      "99\t92\t2.9\t1\t0 92\n",
      "100\t111\t2.9\t1\t0 111\n",
      "101\t102\t2.9\t1\t0 102\n",
      "102\t100\t2.9\t1\t0 100\n",
      "103\t64\t2.9\t1\t0 64\n",
      "104\t87\t2.9\t1\t0 87\n",
      "105\t81\t2.9\t1\t0 81\n",
      "106\t139\t2.9\t1\t0 139\n",
      "107\t101\t2.9\t1\t0 101\n",
      "108\t54\t2.9\t1\t0 54\n",
      "109\t61\t2.9\t1\t1 60\n",
      "110\t49\t2.9\t1\t0 49\n",
      "111\t39\t2.9\t1\t0 39\n",
      "112\t25\t2.9\t1\t0 25\n",
      "113\t53\t2.9\t1\t0 53\n",
      "114\t57\t2.9\t1\t0 57\n",
      "115\t45\t2.9\t1\t0 45\n",
      "116\t52\t2.9\t1\t0 52\n",
      "117\t36\t2.9\t1\t0 36\n",
      "118\t26\t2.9\t1\t0 26\n",
      "119\t33\t2.9\t1\t0 33\n",
      "120\t23\t2.9\t1\t0 23\n",
      "121\t31\t2.9\t1\t0 31\n",
      "122\t24\t2.9\t1\t0 24\n",
      "123\t77\t2.9\t1\t0 77\n",
      "124\t30\t2.9\t1\t0 30\n",
      "125\t24\t2.9\t1\t0 24\n",
      "126\t23\t2.9\t1\t0 23\n",
      "127\t15\t2.9\t1\t0 15\n",
      "128\t10\t2.9\t1\t0 10\n",
      "129\t24\t2.9\t1\t0 24\n",
      "130\t10\t2.9\t1\t0 10\n",
      "131\t26\t2.9\t1\t0 26\n",
      "132\t24\t2.9\t1\t0 24\n",
      "133\t14\t2.9\t1\t0 14\n",
      "134\t42\t2.9\t1\t0 42\n",
      "135\t20\t2.9\t1\t0 20\n",
      "136\t21\t2.9\t1\t0 21\n",
      "137\t17\t2.9\t1\t0 17\n",
      "138\t34\t2.9\t1\t0 34\n",
      "139\t25\t2.9\t1\t0 25\n",
      "140\t25\t2.9\t1\t0 25\n",
      "141\t42\t2.9\t1\t0 42\n",
      "142\t6\t2.9\t1\t0 6\n",
      "143\t20\t2.9\t1\t0 20\n",
      "144\t23\t2.9\t1\t0 23\n",
      "145\t27\t2.9\t1\t0 27\n",
      "146\t19\t2.9\t1\t0 19\n",
      "147\t27\t2.9\t1\t0 27\n",
      "148\t8\t2.9\t1\t0 8\n",
      "149\t12\t2.9\t1\t0 12\n",
      "150\t19\t2.9\t1\t0 19\n",
      "151\t21\t2.9\t1\t0 21\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468/SRR6399468_2.fastq.gz\n",
      "=============================================\n",
      "47822570 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399468_1_trimmed.fq.gz and SRR6399468_2_trimmed.fq.gz\n",
      "file_1: SRR6399468_1_trimmed.fq.gz, file_2: SRR6399468_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399468_1_trimmed.fq.gz and SRR6399468_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399468_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399468_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 47822570\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (18 bp): 7097 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399468_1_trimmed.fq.gz and SRR6399468_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   2%|▊                                        |  ETA: 12.25 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3441.094812 seconds (106.00 k allocations: 3.017 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 74,751,346\n",
      "reads read      : 149,502,692\n",
      "reads written   : 149,502,692\n",
      "\n",
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/SRR6399469_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count Illumina: 0, count smallRNA: 0, count Nextera: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/trim_galore/SRR6399469_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/SRR6399469_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399469_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/SRR6399469_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "60000000 sequences processed\n",
      "70000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/SRR6399469_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 187.370 s (2.507 µs/read; 23.94 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              74,751,346\n",
      "Reads with adapters:                23,176,227 (31.0%)\n",
      "Reads written (passing filters):    74,751,346 (100.0%)\n",
      "\n",
      "Total basepairs processed: 11,029,952,527 bp\n",
      "Quality-trimmed:                 182,848 bp (0.0%)\n",
      "Total written (filtered):  10,997,763,632 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 23176227 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 32.6%\n",
      "  C: 24.8%\n",
      "  G: 19.5%\n",
      "  T: 23.2%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t16752512\t18687836.5\t0\t16752512\n",
      "2\t4874963\t4671959.1\t0\t4874963\n",
      "3\t1079213\t1167989.8\t0\t1079213\n",
      "4\t267851\t291997.4\t0\t267851\n",
      "5\t81987\t72999.4\t0\t81987\n",
      "6\t60691\t18249.8\t0\t60691\n",
      "7\t58807\t4562.5\t0\t58807\n",
      "8\t16\t1140.6\t0\t16\n",
      "9\t37\t285.2\t0\t3 34\n",
      "10\t25\t71.3\t1\t2 23\n",
      "11\t17\t17.8\t1\t0 17\n",
      "12\t3\t4.5\t1\t0 3\n",
      "13\t1\t1.1\t1\t0 1\n",
      "18\t1\t1.1\t1\t0 1\n",
      "20\t1\t1.1\t1\t0 1\n",
      "21\t1\t1.1\t1\t0 1\n",
      "27\t1\t1.1\t1\t0 1\n",
      "29\t2\t1.1\t1\t0 2\n",
      "30\t5\t1.1\t1\t0 5\n",
      "31\t4\t1.1\t1\t0 4\n",
      "34\t1\t1.1\t1\t0 1\n",
      "35\t5\t1.1\t1\t0 5\n",
      "36\t1\t1.1\t1\t0 1\n",
      "37\t1\t1.1\t1\t0 1\n",
      "40\t1\t1.1\t1\t0 1\n",
      "42\t2\t1.1\t1\t0 2\n",
      "59\t1\t1.1\t1\t0 1\n",
      "62\t2\t1.1\t1\t0 2\n",
      "63\t3\t1.1\t1\t0 3\n",
      "64\t1\t1.1\t1\t0 1\n",
      "66\t2\t1.1\t1\t0 2\n",
      "69\t2\t1.1\t1\t0 2\n",
      "71\t2\t1.1\t1\t0 2\n",
      "73\t1\t1.1\t1\t0 1\n",
      "74\t3\t1.1\t1\t0 3\n",
      "75\t2\t1.1\t1\t0 2\n",
      "76\t1\t1.1\t1\t0 1\n",
      "77\t1\t1.1\t1\t0 1\n",
      "78\t3\t1.1\t1\t0 3\n",
      "79\t1\t1.1\t1\t0 1\n",
      "80\t5\t1.1\t1\t0 5\n",
      "81\t1\t1.1\t1\t0 1\n",
      "82\t2\t1.1\t1\t0 2\n",
      "83\t3\t1.1\t1\t0 3\n",
      "85\t2\t1.1\t1\t0 2\n",
      "86\t1\t1.1\t1\t0 1\n",
      "87\t1\t1.1\t1\t0 1\n",
      "88\t2\t1.1\t1\t0 2\n",
      "90\t3\t1.1\t1\t0 3\n",
      "91\t1\t1.1\t1\t0 1\n",
      "92\t1\t1.1\t1\t0 1\n",
      "93\t2\t1.1\t1\t0 2\n",
      "95\t2\t1.1\t1\t0 2\n",
      "96\t2\t1.1\t1\t0 2\n",
      "97\t2\t1.1\t1\t0 2\n",
      "101\t1\t1.1\t1\t0 1\n",
      "102\t2\t1.1\t1\t0 2\n",
      "104\t1\t1.1\t1\t0 1\n",
      "106\t1\t1.1\t1\t0 1\n",
      "108\t1\t1.1\t1\t0 1\n",
      "109\t1\t1.1\t1\t0 1\n",
      "111\t2\t1.1\t1\t0 2\n",
      "112\t1\t1.1\t1\t0 1\n",
      "114\t1\t1.1\t1\t0 1\n",
      "116\t1\t1.1\t1\t0 1\n",
      "119\t2\t1.1\t1\t0 2\n",
      "120\t1\t1.1\t1\t0 1\n",
      "123\t1\t1.1\t1\t0 1\n",
      "124\t1\t1.1\t1\t0 1\n",
      "125\t2\t1.1\t1\t0 2\n",
      "127\t1\t1.1\t1\t0 1\n",
      "130\t1\t1.1\t1\t0 1\n",
      "132\t1\t1.1\t1\t0 1\n",
      "138\t1\t1.1\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/SRR6399469_1.fastq.gz\n",
      "=============================================\n",
      "74751346 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/trim_galore/SRR6399469_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/SRR6399469_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399469_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/SRR6399469_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "60000000 sequences processed\n",
      "70000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/SRR6399469_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 203.905 s (2.728 µs/read; 22.00 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              74,751,346\n",
      "Reads with adapters:                22,829,346 (30.5%)\n",
      "Reads written (passing filters):    74,751,346 (100.0%)\n",
      "\n",
      "Total basepairs processed: 10,707,635,874 bp\n",
      "Quality-trimmed:                 371,730 bp (0.0%)\n",
      "Total written (filtered):  10,675,746,286 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 22829346 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 36.7%\n",
      "  C: 23.1%\n",
      "  G: 17.7%\n",
      "  T: 22.5%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t16552188\t18687836.5\t0\t16552188\n",
      "2\t4751855\t4671959.1\t0\t4751855\n",
      "3\t1046045\t1167989.8\t0\t1046045\n",
      "4\t269683\t291997.4\t0\t269683\n",
      "5\t82032\t72999.4\t0\t82032\n",
      "6\t63828\t18249.8\t0\t63828\n",
      "7\t63511\t4562.5\t0\t63511\n",
      "8\t41\t1140.6\t0\t41\n",
      "9\t45\t285.2\t0\t1 44\n",
      "10\t23\t71.3\t1\t2 21\n",
      "11\t17\t17.8\t1\t0 17\n",
      "12\t3\t4.5\t1\t1 2\n",
      "23\t2\t1.1\t1\t0 2\n",
      "25\t1\t1.1\t1\t0 1\n",
      "29\t1\t1.1\t1\t0 1\n",
      "30\t6\t1.1\t1\t0 6\n",
      "31\t1\t1.1\t1\t0 1\n",
      "36\t1\t1.1\t1\t0 1\n",
      "40\t1\t1.1\t1\t0 1\n",
      "56\t1\t1.1\t1\t0 1\n",
      "63\t1\t1.1\t1\t0 1\n",
      "64\t1\t1.1\t1\t0 1\n",
      "67\t2\t1.1\t1\t0 2\n",
      "68\t2\t1.1\t1\t0 2\n",
      "69\t1\t1.1\t1\t0 1\n",
      "70\t2\t1.1\t1\t0 2\n",
      "73\t2\t1.1\t1\t0 2\n",
      "74\t2\t1.1\t1\t0 2\n",
      "75\t1\t1.1\t1\t0 1\n",
      "76\t1\t1.1\t1\t0 1\n",
      "78\t1\t1.1\t1\t0 1\n",
      "79\t1\t1.1\t1\t0 1\n",
      "80\t1\t1.1\t1\t0 1\n",
      "81\t2\t1.1\t1\t0 2\n",
      "82\t2\t1.1\t1\t0 2\n",
      "83\t2\t1.1\t1\t0 2\n",
      "85\t2\t1.1\t1\t0 2\n",
      "86\t1\t1.1\t1\t0 1\n",
      "87\t2\t1.1\t1\t0 2\n",
      "88\t2\t1.1\t1\t0 2\n",
      "89\t1\t1.1\t1\t0 1\n",
      "90\t1\t1.1\t1\t0 1\n",
      "91\t1\t1.1\t1\t0 1\n",
      "92\t1\t1.1\t1\t0 1\n",
      "93\t1\t1.1\t1\t0 1\n",
      "94\t1\t1.1\t1\t0 1\n",
      "96\t1\t1.1\t1\t0 1\n",
      "97\t1\t1.1\t1\t0 1\n",
      "99\t1\t1.1\t1\t0 1\n",
      "104\t1\t1.1\t1\t0 1\n",
      "105\t1\t1.1\t1\t0 1\n",
      "108\t1\t1.1\t1\t0 1\n",
      "110\t1\t1.1\t1\t0 1\n",
      "111\t1\t1.1\t1\t0 1\n",
      "112\t1\t1.1\t1\t0 1\n",
      "113\t1\t1.1\t1\t0 1\n",
      "114\t1\t1.1\t1\t0 1\n",
      "117\t1\t1.1\t1\t0 1\n",
      "118\t2\t1.1\t1\t0 2\n",
      "123\t2\t1.1\t1\t0 2\n",
      "124\t1\t1.1\t1\t0 1\n",
      "127\t1\t1.1\t1\t0 1\n",
      "131\t1\t1.1\t1\t0 1\n",
      "132\t2\t1.1\t1\t0 2\n",
      "133\t1\t1.1\t1\t0 1\n",
      "138\t1\t1.1\t1\t0 1\n",
      "139\t1\t1.1\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469/SRR6399469_2.fastq.gz\n",
      "=============================================\n",
      "74751346 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399469_1_trimmed.fq.gz and SRR6399469_2_trimmed.fq.gz\n",
      "file_1: SRR6399469_1_trimmed.fq.gz, file_2: SRR6399469_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399469_1_trimmed.fq.gz and SRR6399469_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399469_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399469_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 74751346\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 11156 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399469_1_trimmed.fq.gz and SRR6399469_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   2%|▊                                        |  ETA: 14.12 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2463.706509 seconds (75.90 k allocations: 2.160 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 55,434,719\n",
      "reads read      : 110,869,438\n",
      "reads written   : 110,869,438\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/SRR6399470_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count smallRNA: 0, count Illumina: 0, count Nextera: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/trim_galore/SRR6399470_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/SRR6399470_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399470_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/SRR6399470_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/SRR6399470_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 141.686 s (2.556 µs/read; 23.48 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              55,434,719\n",
      "Reads with adapters:                17,420,059 (31.4%)\n",
      "Reads written (passing filters):    55,434,719 (100.0%)\n",
      "\n",
      "Total basepairs processed: 8,174,229,731 bp\n",
      "Quality-trimmed:                 143,952 bp (0.0%)\n",
      "Total written (filtered):  8,149,672,508 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 17420059 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 31.5%\n",
      "  C: 25.4%\n",
      "  G: 19.9%\n",
      "  T: 23.2%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t12344693\t13858679.8\t0\t12344693\n",
      "2\t3873381\t3464669.9\t0\t3873381\n",
      "3\t842762\t866167.5\t0\t842762\n",
      "4\t176518\t216541.9\t0\t176518\n",
      "5\t70154\t54135.5\t0\t70154\n",
      "6\t57200\t13533.9\t0\t57200\n",
      "7\t55189\t3383.5\t0\t55189\n",
      "8\t4\t845.9\t0\t4\n",
      "9\t48\t211.5\t0\t3 45\n",
      "10\t34\t52.9\t1\t2 32\n",
      "11\t9\t13.2\t1\t0 9\n",
      "12\t2\t3.3\t1\t0 2\n",
      "25\t1\t0.8\t1\t0 1\n",
      "28\t1\t0.8\t1\t0 1\n",
      "29\t1\t0.8\t1\t0 1\n",
      "30\t3\t0.8\t1\t0 3\n",
      "31\t2\t0.8\t1\t0 2\n",
      "36\t1\t0.8\t1\t0 1\n",
      "37\t1\t0.8\t1\t0 1\n",
      "54\t1\t0.8\t1\t0 1\n",
      "57\t1\t0.8\t1\t0 1\n",
      "59\t1\t0.8\t1\t0 1\n",
      "67\t1\t0.8\t1\t0 1\n",
      "72\t2\t0.8\t1\t0 2\n",
      "74\t1\t0.8\t1\t0 1\n",
      "76\t1\t0.8\t1\t0 1\n",
      "81\t2\t0.8\t1\t0 2\n",
      "84\t2\t0.8\t1\t0 2\n",
      "85\t1\t0.8\t1\t0 1\n",
      "86\t2\t0.8\t1\t0 2\n",
      "90\t1\t0.8\t1\t0 1\n",
      "92\t1\t0.8\t1\t0 1\n",
      "93\t1\t0.8\t1\t0 1\n",
      "95\t2\t0.8\t1\t0 2\n",
      "96\t1\t0.8\t1\t0 1\n",
      "97\t1\t0.8\t1\t0 1\n",
      "98\t2\t0.8\t1\t0 2\n",
      "101\t1\t0.8\t1\t0 1\n",
      "107\t2\t0.8\t1\t0 2\n",
      "125\t2\t0.8\t1\t0 2\n",
      "126\t2\t0.8\t1\t0 2\n",
      "129\t8\t0.8\t1\t0 8\n",
      "130\t3\t0.8\t1\t0 3\n",
      "131\t5\t0.8\t1\t0 5\n",
      "132\t2\t0.8\t1\t0 2\n",
      "133\t3\t0.8\t1\t0 3\n",
      "134\t3\t0.8\t1\t0 3\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/SRR6399470_1.fastq.gz\n",
      "=============================================\n",
      "55434719 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/trim_galore/SRR6399470_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/SRR6399470_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399470_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/SRR6399470_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/SRR6399470_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 144.507 s (2.607 µs/read; 23.02 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              55,434,719\n",
      "Reads with adapters:                15,119,834 (27.3%)\n",
      "Reads written (passing filters):    55,434,719 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,955,051,123 bp\n",
      "Quality-trimmed:                 330,396 bp (0.0%)\n",
      "Total written (filtered):  7,933,022,498 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 15119834 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 36.1%\n",
      "  C: 23.5%\n",
      "  G: 17.8%\n",
      "  T: 22.6%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t10348127\t13858679.8\t0\t10348127\n",
      "2\t3749898\t3464669.9\t0\t3749898\n",
      "3\t652762\t866167.5\t0\t652762\n",
      "4\t166210\t216541.9\t0\t166210\n",
      "5\t68349\t54135.5\t0\t68349\n",
      "6\t59824\t13533.9\t0\t59824\n",
      "7\t74537\t3383.5\t0\t74537\n",
      "8\t5\t845.9\t0\t5\n",
      "9\t41\t211.5\t0\t5 36\n",
      "10\t22\t52.9\t1\t5 17\n",
      "11\t12\t13.2\t1\t0 12\n",
      "12\t3\t3.3\t1\t0 3\n",
      "26\t1\t0.8\t1\t0 1\n",
      "31\t2\t0.8\t1\t0 2\n",
      "35\t1\t0.8\t1\t0 1\n",
      "43\t1\t0.8\t1\t0 1\n",
      "56\t1\t0.8\t1\t0 1\n",
      "60\t1\t0.8\t1\t0 1\n",
      "61\t1\t0.8\t1\t0 1\n",
      "65\t1\t0.8\t1\t0 1\n",
      "66\t1\t0.8\t1\t0 1\n",
      "68\t1\t0.8\t1\t0 1\n",
      "69\t1\t0.8\t1\t0 1\n",
      "70\t1\t0.8\t1\t0 1\n",
      "72\t2\t0.8\t1\t0 2\n",
      "73\t1\t0.8\t1\t0 1\n",
      "74\t1\t0.8\t1\t0 1\n",
      "75\t2\t0.8\t1\t0 2\n",
      "81\t1\t0.8\t1\t0 1\n",
      "83\t1\t0.8\t1\t0 1\n",
      "86\t1\t0.8\t1\t0 1\n",
      "87\t2\t0.8\t1\t0 2\n",
      "90\t1\t0.8\t1\t0 1\n",
      "91\t1\t0.8\t1\t0 1\n",
      "92\t1\t0.8\t1\t0 1\n",
      "95\t1\t0.8\t1\t0 1\n",
      "106\t1\t0.8\t1\t0 1\n",
      "107\t1\t0.8\t1\t0 1\n",
      "121\t1\t0.8\t1\t0 1\n",
      "122\t1\t0.8\t1\t0 1\n",
      "127\t1\t0.8\t1\t0 1\n",
      "128\t1\t0.8\t1\t0 1\n",
      "129\t2\t0.8\t1\t0 2\n",
      "130\t3\t0.8\t1\t0 3\n",
      "131\t2\t0.8\t1\t0 2\n",
      "133\t2\t0.8\t1\t0 2\n",
      "134\t1\t0.8\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470/SRR6399470_2.fastq.gz\n",
      "=============================================\n",
      "55434719 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399470_1_trimmed.fq.gz and SRR6399470_2_trimmed.fq.gz\n",
      "file_1: SRR6399470_1_trimmed.fq.gz, file_2: SRR6399470_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399470_1_trimmed.fq.gz and SRR6399470_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399470_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399470_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 55434719\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 5560 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399470_1_trimmed.fq.gz and SRR6399470_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   2%|▉                                        |  ETA: 14.91 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2428.659272 seconds (74.82 k allocations: 2.130 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 49,317,518\n",
      "reads read      : 98,635,036\n",
      "reads written   : 98,635,036\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/SRR6399471_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count smallRNA: 0, count Nextera: 0, count Illumina: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/trim_galore/SRR6399471_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/SRR6399471_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399471_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/SRR6399471_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/SRR6399471_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 139.381 s (2.826 µs/read; 21.23 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              49,317,518\n",
      "Reads with adapters:                15,228,940 (30.9%)\n",
      "Reads written (passing filters):    49,317,518 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,363,005,422 bp\n",
      "Quality-trimmed:                  57,451 bp (0.0%)\n",
      "Total written (filtered):  7,342,298,599 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 15228940 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 30.8%\n",
      "  C: 25.5%\n",
      "  G: 20.9%\n",
      "  T: 22.8%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t11086100\t12329379.5\t0\t11086100\n",
      "2\t3220163\t3082344.9\t0\t3220163\n",
      "3\t703307\t770586.2\t0\t703307\n",
      "4\t145181\t192646.6\t0\t145181\n",
      "5\t33933\t48161.6\t0\t33933\n",
      "6\t21548\t12040.4\t0\t21548\n",
      "7\t18646\t3010.1\t0\t18646\n",
      "8\t1\t752.5\t0\t1\n",
      "9\t21\t188.1\t0\t1 20\n",
      "10\t4\t47.0\t1\t0 4\n",
      "11\t5\t11.8\t1\t0 5\n",
      "12\t1\t2.9\t1\t0 1\n",
      "25\t1\t0.7\t1\t0 1\n",
      "30\t3\t0.7\t1\t0 3\n",
      "33\t1\t0.7\t1\t0 1\n",
      "34\t1\t0.7\t1\t0 1\n",
      "37\t1\t0.7\t1\t0 1\n",
      "58\t1\t0.7\t1\t0 1\n",
      "60\t1\t0.7\t1\t0 1\n",
      "67\t1\t0.7\t1\t0 1\n",
      "76\t1\t0.7\t1\t0 1\n",
      "78\t1\t0.7\t1\t0 1\n",
      "79\t1\t0.7\t1\t0 1\n",
      "82\t1\t0.7\t1\t0 1\n",
      "83\t1\t0.7\t1\t0 1\n",
      "84\t1\t0.7\t1\t0 1\n",
      "88\t1\t0.7\t1\t0 1\n",
      "99\t1\t0.7\t1\t0 1\n",
      "101\t1\t0.7\t1\t0 1\n",
      "109\t1\t0.7\t1\t0 1\n",
      "111\t1\t0.7\t1\t0 1\n",
      "115\t1\t0.7\t1\t0 1\n",
      "118\t1\t0.7\t1\t0 1\n",
      "119\t1\t0.7\t1\t0 1\n",
      "122\t1\t0.7\t1\t0 1\n",
      "123\t1\t0.7\t1\t0 1\n",
      "129\t2\t0.7\t1\t0 2\n",
      "135\t1\t0.7\t1\t0 1\n",
      "138\t1\t0.7\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/SRR6399471_1.fastq.gz\n",
      "=============================================\n",
      "49317518 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/trim_galore/SRR6399471_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/SRR6399471_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399471_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/SRR6399471_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/SRR6399471_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 151.386 s (3.070 µs/read; 19.55 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              49,317,518\n",
      "Reads with adapters:                14,786,028 (30.0%)\n",
      "Reads written (passing filters):    49,317,518 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,102,584,488 bp\n",
      "Quality-trimmed:                 170,120 bp (0.0%)\n",
      "Total written (filtered):  7,082,274,657 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 14786028 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.3%\n",
      "  C: 24.3%\n",
      "  G: 18.8%\n",
      "  T: 22.6%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t10738412\t12329379.5\t0\t10738412\n",
      "2\t3128022\t3082344.9\t0\t3128022\n",
      "3\t682849\t770586.2\t0\t682849\n",
      "4\t155740\t192646.6\t0\t155740\n",
      "5\t35989\t48161.6\t0\t35989\n",
      "6\t23885\t12040.4\t0\t23885\n",
      "7\t21051\t3010.1\t0\t21051\n",
      "8\t2\t752.5\t0\t2\n",
      "9\t34\t188.1\t0\t0 34\n",
      "10\t8\t47.0\t1\t0 8\n",
      "11\t5\t11.8\t1\t0 5\n",
      "28\t1\t0.7\t1\t0 1\n",
      "31\t1\t0.7\t1\t0 1\n",
      "41\t1\t0.7\t1\t0 1\n",
      "42\t1\t0.7\t1\t0 1\n",
      "46\t1\t0.7\t1\t0 1\n",
      "62\t2\t0.7\t1\t0 2\n",
      "68\t1\t0.7\t1\t0 1\n",
      "71\t1\t0.7\t1\t0 1\n",
      "77\t1\t0.7\t1\t0 1\n",
      "79\t2\t0.7\t1\t0 2\n",
      "82\t2\t0.7\t1\t0 2\n",
      "83\t1\t0.7\t1\t0 1\n",
      "84\t1\t0.7\t1\t0 1\n",
      "85\t1\t0.7\t1\t0 1\n",
      "88\t1\t0.7\t1\t0 1\n",
      "95\t1\t0.7\t1\t0 1\n",
      "99\t2\t0.7\t1\t0 2\n",
      "104\t1\t0.7\t1\t0 1\n",
      "107\t1\t0.7\t1\t0 1\n",
      "112\t1\t0.7\t1\t0 1\n",
      "118\t1\t0.7\t1\t0 1\n",
      "119\t1\t0.7\t1\t0 1\n",
      "120\t1\t0.7\t1\t0 1\n",
      "127\t1\t0.7\t1\t0 1\n",
      "129\t2\t0.7\t1\t0 2\n",
      "131\t1\t0.7\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471/SRR6399471_2.fastq.gz\n",
      "=============================================\n",
      "49317518 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399471_1_trimmed.fq.gz and SRR6399471_2_trimmed.fq.gz\n",
      "file_1: SRR6399471_1_trimmed.fq.gz, file_2: SRR6399471_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399471_1_trimmed.fq.gz and SRR6399471_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399471_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399471_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 49317518\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 6839 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399471_1_trimmed.fq.gz and SRR6399471_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   2%|▉                                        |  ETA: 15.54 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2095.843648 seconds (64.58 k allocations: 1.838 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 41,261,975\n",
      "reads read      : 82,523,950\n",
      "reads written   : 82,523,950\n",
      "\n",
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/SRR6399472_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count Nextera: 0, count smallRNA: 0, count Illumina: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/trim_galore/SRR6399472_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/SRR6399472_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399472_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/SRR6399472_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/SRR6399472_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 122.147 s (2.960 µs/read; 20.27 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              41,261,975\n",
      "Reads with adapters:                12,231,541 (29.6%)\n",
      "Reads written (passing filters):    41,261,975 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,122,836,753 bp\n",
      "Quality-trimmed:                  77,128 bp (0.0%)\n",
      "Total written (filtered):  6,105,907,192 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 12231541 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 30.7%\n",
      "  C: 24.7%\n",
      "  G: 19.4%\n",
      "  T: 25.2%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t8816411\t10315493.8\t0\t8816411\n",
      "2\t2624055\t2578873.4\t0\t2624055\n",
      "3\t561805\t644718.4\t0\t561805\n",
      "4\t137018\t161179.6\t0\t137018\n",
      "5\t39955\t40294.9\t0\t39955\n",
      "6\t26663\t10073.7\t0\t26663\n",
      "7\t25433\t2518.4\t0\t25433\n",
      "9\t17\t157.4\t0\t3 14\n",
      "10\t2\t39.4\t1\t0 2\n",
      "11\t1\t9.8\t1\t0 1\n",
      "12\t3\t2.5\t1\t2 1\n",
      "15\t2\t0.6\t1\t0 2\n",
      "22\t2\t0.6\t1\t0 2\n",
      "24\t1\t0.6\t1\t0 1\n",
      "26\t1\t0.6\t1\t0 1\n",
      "27\t2\t0.6\t1\t0 2\n",
      "29\t2\t0.6\t1\t0 2\n",
      "30\t3\t0.6\t1\t0 3\n",
      "32\t1\t0.6\t1\t0 1\n",
      "34\t2\t0.6\t1\t0 2\n",
      "36\t2\t0.6\t1\t0 2\n",
      "37\t4\t0.6\t1\t0 4\n",
      "38\t2\t0.6\t1\t0 2\n",
      "41\t1\t0.6\t1\t0 1\n",
      "42\t2\t0.6\t1\t0 2\n",
      "44\t1\t0.6\t1\t0 1\n",
      "49\t3\t0.6\t1\t0 3\n",
      "50\t2\t0.6\t1\t0 2\n",
      "51\t2\t0.6\t1\t0 2\n",
      "53\t1\t0.6\t1\t0 1\n",
      "56\t2\t0.6\t1\t0 2\n",
      "59\t2\t0.6\t1\t0 2\n",
      "60\t1\t0.6\t1\t0 1\n",
      "61\t1\t0.6\t1\t0 1\n",
      "63\t8\t0.6\t1\t0 8\n",
      "67\t1\t0.6\t1\t0 1\n",
      "70\t1\t0.6\t1\t0 1\n",
      "71\t1\t0.6\t1\t0 1\n",
      "72\t2\t0.6\t1\t0 2\n",
      "74\t2\t0.6\t1\t0 2\n",
      "75\t1\t0.6\t1\t0 1\n",
      "76\t1\t0.6\t1\t0 1\n",
      "77\t2\t0.6\t1\t0 2\n",
      "81\t2\t0.6\t1\t0 2\n",
      "84\t2\t0.6\t1\t0 2\n",
      "87\t1\t0.6\t1\t0 1\n",
      "88\t2\t0.6\t1\t0 2\n",
      "90\t2\t0.6\t1\t0 2\n",
      "92\t3\t0.6\t1\t0 3\n",
      "93\t1\t0.6\t1\t0 1\n",
      "95\t5\t0.6\t1\t0 5\n",
      "96\t1\t0.6\t1\t0 1\n",
      "97\t1\t0.6\t1\t0 1\n",
      "99\t1\t0.6\t1\t0 1\n",
      "101\t1\t0.6\t1\t0 1\n",
      "102\t4\t0.6\t1\t0 4\n",
      "103\t2\t0.6\t1\t0 2\n",
      "104\t2\t0.6\t1\t0 2\n",
      "105\t2\t0.6\t1\t0 2\n",
      "106\t1\t0.6\t1\t0 1\n",
      "107\t4\t0.6\t1\t0 4\n",
      "108\t2\t0.6\t1\t0 2\n",
      "109\t1\t0.6\t1\t0 1\n",
      "110\t1\t0.6\t1\t0 1\n",
      "111\t1\t0.6\t1\t0 1\n",
      "112\t2\t0.6\t1\t0 2\n",
      "113\t2\t0.6\t1\t0 2\n",
      "114\t4\t0.6\t1\t0 4\n",
      "115\t2\t0.6\t1\t0 2\n",
      "116\t2\t0.6\t1\t0 2\n",
      "117\t3\t0.6\t1\t0 3\n",
      "118\t5\t0.6\t1\t0 5\n",
      "119\t1\t0.6\t1\t0 1\n",
      "120\t5\t0.6\t1\t0 5\n",
      "121\t2\t0.6\t1\t0 2\n",
      "122\t3\t0.6\t1\t0 3\n",
      "123\t2\t0.6\t1\t0 2\n",
      "124\t6\t0.6\t1\t0 6\n",
      "125\t3\t0.6\t1\t0 3\n",
      "126\t7\t0.6\t1\t0 7\n",
      "127\t3\t0.6\t1\t0 3\n",
      "128\t2\t0.6\t1\t0 2\n",
      "129\t1\t0.6\t1\t0 1\n",
      "130\t2\t0.6\t1\t0 2\n",
      "131\t2\t0.6\t1\t0 2\n",
      "132\t2\t0.6\t1\t0 2\n",
      "134\t2\t0.6\t1\t0 2\n",
      "135\t2\t0.6\t1\t0 2\n",
      "136\t2\t0.6\t1\t0 2\n",
      "137\t4\t0.6\t1\t0 4\n",
      "138\t3\t0.6\t1\t0 3\n",
      "139\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/SRR6399472_1.fastq.gz\n",
      "=============================================\n",
      "41261975 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/trim_galore/SRR6399472_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/SRR6399472_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399472_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/SRR6399472_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/SRR6399472_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 111.287 s (2.697 µs/read; 22.25 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              41,261,975\n",
      "Reads with adapters:                13,524,723 (32.8%)\n",
      "Reads written (passing filters):    41,261,975 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,969,706,064 bp\n",
      "Quality-trimmed:                 194,184 bp (0.0%)\n",
      "Total written (filtered):  5,951,363,721 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 13524723 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 41.2%\n",
      "  C: 21.5%\n",
      "  G: 15.2%\n",
      "  T: 22.1%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t10168609\t10315493.8\t0\t10168609\n",
      "2\t2524825\t2578873.4\t0\t2524825\n",
      "3\t587877\t644718.4\t0\t587877\n",
      "4\t147217\t161179.6\t0\t147217\n",
      "5\t40145\t40294.9\t0\t40145\n",
      "6\t28001\t10073.7\t0\t28001\n",
      "7\t27868\t2518.4\t0\t27868\n",
      "8\t2\t629.6\t0\t2\n",
      "9\t11\t157.4\t0\t1 10\n",
      "10\t10\t39.4\t1\t1 9\n",
      "11\t1\t9.8\t1\t0 1\n",
      "12\t2\t2.5\t1\t1 1\n",
      "16\t1\t0.6\t1\t0 1\n",
      "24\t1\t0.6\t1\t0 1\n",
      "26\t1\t0.6\t1\t0 1\n",
      "27\t3\t0.6\t1\t0 3\n",
      "28\t2\t0.6\t1\t0 2\n",
      "29\t2\t0.6\t1\t0 2\n",
      "30\t1\t0.6\t1\t0 1\n",
      "31\t1\t0.6\t1\t0 1\n",
      "34\t1\t0.6\t1\t0 1\n",
      "36\t2\t0.6\t1\t0 2\n",
      "38\t1\t0.6\t1\t0 1\n",
      "39\t1\t0.6\t1\t0 1\n",
      "40\t2\t0.6\t1\t0 2\n",
      "42\t5\t0.6\t1\t0 5\n",
      "43\t3\t0.6\t1\t0 3\n",
      "45\t1\t0.6\t1\t0 1\n",
      "46\t4\t0.6\t1\t0 4\n",
      "47\t5\t0.6\t1\t0 5\n",
      "50\t2\t0.6\t1\t0 2\n",
      "51\t2\t0.6\t1\t0 2\n",
      "57\t1\t0.6\t1\t0 1\n",
      "59\t3\t0.6\t1\t0 3\n",
      "61\t2\t0.6\t1\t0 2\n",
      "62\t1\t0.6\t1\t0 1\n",
      "67\t1\t0.6\t1\t0 1\n",
      "68\t2\t0.6\t1\t0 2\n",
      "69\t1\t0.6\t1\t0 1\n",
      "70\t1\t0.6\t1\t0 1\n",
      "71\t2\t0.6\t1\t0 2\n",
      "72\t3\t0.6\t1\t0 3\n",
      "73\t1\t0.6\t1\t0 1\n",
      "74\t1\t0.6\t1\t0 1\n",
      "75\t1\t0.6\t1\t0 1\n",
      "77\t1\t0.6\t1\t0 1\n",
      "78\t1\t0.6\t1\t0 1\n",
      "80\t1\t0.6\t1\t0 1\n",
      "81\t2\t0.6\t1\t0 2\n",
      "82\t1\t0.6\t1\t0 1\n",
      "83\t1\t0.6\t1\t0 1\n",
      "84\t2\t0.6\t1\t0 2\n",
      "85\t2\t0.6\t1\t0 2\n",
      "88\t2\t0.6\t1\t0 2\n",
      "90\t1\t0.6\t1\t0 1\n",
      "91\t1\t0.6\t1\t0 1\n",
      "92\t1\t0.6\t1\t0 1\n",
      "93\t3\t0.6\t1\t0 3\n",
      "94\t3\t0.6\t1\t0 3\n",
      "95\t2\t0.6\t1\t0 2\n",
      "98\t1\t0.6\t1\t0 1\n",
      "100\t1\t0.6\t1\t0 1\n",
      "101\t1\t0.6\t1\t0 1\n",
      "102\t2\t0.6\t1\t0 2\n",
      "103\t1\t0.6\t1\t0 1\n",
      "104\t2\t0.6\t1\t0 2\n",
      "105\t1\t0.6\t1\t0 1\n",
      "106\t1\t0.6\t1\t0 1\n",
      "107\t4\t0.6\t1\t0 4\n",
      "108\t2\t0.6\t1\t0 2\n",
      "109\t2\t0.6\t1\t0 2\n",
      "110\t1\t0.6\t1\t0 1\n",
      "111\t1\t0.6\t1\t0 1\n",
      "112\t2\t0.6\t1\t0 2\n",
      "114\t2\t0.6\t1\t0 2\n",
      "116\t1\t0.6\t1\t0 1\n",
      "117\t3\t0.6\t1\t0 3\n",
      "118\t3\t0.6\t1\t0 3\n",
      "119\t2\t0.6\t1\t0 2\n",
      "120\t4\t0.6\t1\t0 4\n",
      "121\t3\t0.6\t1\t0 3\n",
      "122\t3\t0.6\t1\t0 3\n",
      "123\t1\t0.6\t1\t0 1\n",
      "124\t4\t0.6\t1\t0 4\n",
      "125\t2\t0.6\t1\t0 2\n",
      "126\t4\t0.6\t1\t0 4\n",
      "127\t2\t0.6\t1\t0 2\n",
      "128\t1\t0.6\t1\t0 1\n",
      "129\t2\t0.6\t1\t0 2\n",
      "130\t2\t0.6\t1\t0 2\n",
      "132\t2\t0.6\t1\t0 2\n",
      "133\t1\t0.6\t1\t0 1\n",
      "134\t1\t0.6\t1\t0 1\n",
      "135\t2\t0.6\t1\t0 2\n",
      "137\t2\t0.6\t1\t0 2\n",
      "138\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399472/SRR6399472_2.fastq.gz\n",
      "=============================================\n",
      "41261975 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399472_1_trimmed.fq.gz and SRR6399472_2_trimmed.fq.gz\n",
      "file_1: SRR6399472_1_trimmed.fq.gz, file_2: SRR6399472_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399472_1_trimmed.fq.gz and SRR6399472_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399472_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399472_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 41261975\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 5813 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399472_1_trimmed.fq.gz and SRR6399472_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   2%|█                                        |  ETA: 15.96 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2644.157787 seconds (81.45 k allocations: 2.318 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 59,568,097\n",
      "reads read      : 119,136,194\n",
      "reads written   : 119,136,194\n",
      "\n",
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/SRR6399473_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Using Nextera adapter for trimming (count: 1). Second best hit was smallRNA (count: 0)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/trim_galore/SRR6399473_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/SRR6399473_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399473_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/SRR6399473_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/SRR6399473_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 142.109 s (2.386 µs/read; 25.15 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              59,568,097\n",
      "Reads with adapters:                16,479,860 (27.7%)\n",
      "Reads written (passing filters):    59,568,097 (100.0%)\n",
      "\n",
      "Total basepairs processed: 8,779,272,514 bp\n",
      "Quality-trimmed:                 286,728 bp (0.0%)\n",
      "Total written (filtered):  8,756,464,029 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 16479860 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 35.7%\n",
      "  C: 26.0%\n",
      "  G: 18.7%\n",
      "  T: 19.6%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t12766231\t14892024.2\t0\t12766231\n",
      "2\t2968640\t3723006.1\t0\t2968640\n",
      "3\t526266\t930751.5\t0\t526266\n",
      "4\t146171\t232687.9\t0\t146171\n",
      "5\t28270\t58172.0\t0\t28270\n",
      "6\t10888\t14543.0\t0\t10888\n",
      "7\t4105\t3635.7\t0\t4105\n",
      "8\t70\t908.9\t0\t70\n",
      "9\t61\t227.2\t0\t7 54\n",
      "10\t445\t56.8\t1\t3 442\n",
      "11\t454\t14.2\t1\t1 453\n",
      "12\t32\t3.6\t1\t0 32\n",
      "13\t25\t3.6\t1\t0 25\n",
      "14\t31\t3.6\t1\t0 31\n",
      "15\t44\t3.6\t1\t0 44\n",
      "16\t42\t3.6\t1\t0 42\n",
      "17\t57\t3.6\t1\t0 57\n",
      "18\t86\t3.6\t1\t0 86\n",
      "19\t59\t3.6\t1\t1 58\n",
      "20\t147\t3.6\t1\t0 147\n",
      "21\t557\t3.6\t1\t0 557\n",
      "22\t1026\t3.6\t1\t0 1026\n",
      "23\t654\t3.6\t1\t0 654\n",
      "24\t733\t3.6\t1\t0 733\n",
      "25\t1627\t3.6\t1\t1 1626\n",
      "26\t1088\t3.6\t1\t0 1088\n",
      "27\t710\t3.6\t1\t1 709\n",
      "28\t1491\t3.6\t1\t0 1491\n",
      "29\t1474\t3.6\t1\t0 1474\n",
      "30\t1499\t3.6\t1\t1 1498\n",
      "31\t3411\t3.6\t1\t3 3408\n",
      "32\t196\t3.6\t1\t0 196\n",
      "33\t199\t3.6\t1\t1 198\n",
      "34\t295\t3.6\t1\t1 294\n",
      "35\t359\t3.6\t1\t1 358\n",
      "36\t147\t3.6\t1\t0 147\n",
      "37\t324\t3.6\t1\t0 324\n",
      "38\t78\t3.6\t1\t0 78\n",
      "39\t146\t3.6\t1\t0 146\n",
      "40\t228\t3.6\t1\t2 226\n",
      "41\t332\t3.6\t1\t0 332\n",
      "42\t164\t3.6\t1\t0 164\n",
      "43\t131\t3.6\t1\t0 131\n",
      "44\t132\t3.6\t1\t0 132\n",
      "45\t293\t3.6\t1\t1 292\n",
      "46\t359\t3.6\t1\t1 358\n",
      "47\t1219\t3.6\t1\t9 1210\n",
      "48\t164\t3.6\t1\t1 163\n",
      "49\t65\t3.6\t1\t1 64\n",
      "50\t99\t3.6\t1\t0 99\n",
      "51\t162\t3.6\t1\t0 162\n",
      "52\t109\t3.6\t1\t0 109\n",
      "53\t120\t3.6\t1\t0 120\n",
      "54\t64\t3.6\t1\t1 63\n",
      "55\t73\t3.6\t1\t0 73\n",
      "56\t56\t3.6\t1\t0 56\n",
      "57\t20\t3.6\t1\t0 20\n",
      "58\t138\t3.6\t1\t0 138\n",
      "59\t167\t3.6\t1\t0 167\n",
      "60\t60\t3.6\t1\t0 60\n",
      "61\t120\t3.6\t1\t0 120\n",
      "62\t86\t3.6\t1\t0 86\n",
      "63\t113\t3.6\t1\t0 113\n",
      "64\t99\t3.6\t1\t0 99\n",
      "65\t222\t3.6\t1\t1 221\n",
      "66\t69\t3.6\t1\t1 68\n",
      "67\t78\t3.6\t1\t0 78\n",
      "68\t60\t3.6\t1\t0 60\n",
      "69\t109\t3.6\t1\t0 109\n",
      "70\t91\t3.6\t1\t1 90\n",
      "71\t40\t3.6\t1\t0 40\n",
      "72\t106\t3.6\t1\t0 106\n",
      "73\t62\t3.6\t1\t0 62\n",
      "74\t155\t3.6\t1\t0 155\n",
      "75\t89\t3.6\t1\t0 89\n",
      "76\t68\t3.6\t1\t0 68\n",
      "77\t73\t3.6\t1\t0 73\n",
      "78\t77\t3.6\t1\t0 77\n",
      "79\t78\t3.6\t1\t0 78\n",
      "80\t150\t3.6\t1\t2 148\n",
      "81\t272\t3.6\t1\t0 272\n",
      "82\t37\t3.6\t1\t0 37\n",
      "83\t78\t3.6\t1\t1 77\n",
      "84\t76\t3.6\t1\t0 76\n",
      "85\t178\t3.6\t1\t0 178\n",
      "86\t130\t3.6\t1\t0 130\n",
      "87\t77\t3.6\t1\t0 77\n",
      "88\t53\t3.6\t1\t0 53\n",
      "89\t171\t3.6\t1\t0 171\n",
      "90\t26\t3.6\t1\t0 26\n",
      "91\t132\t3.6\t1\t0 132\n",
      "92\t109\t3.6\t1\t0 109\n",
      "93\t126\t3.6\t1\t0 126\n",
      "94\t48\t3.6\t1\t1 47\n",
      "95\t50\t3.6\t1\t0 50\n",
      "96\t122\t3.6\t1\t0 122\n",
      "97\t71\t3.6\t1\t0 71\n",
      "98\t187\t3.6\t1\t0 187\n",
      "99\t63\t3.6\t1\t1 62\n",
      "100\t84\t3.6\t1\t0 84\n",
      "101\t76\t3.6\t1\t0 76\n",
      "102\t98\t3.6\t1\t0 98\n",
      "103\t104\t3.6\t1\t0 104\n",
      "104\t133\t3.6\t1\t0 133\n",
      "105\t119\t3.6\t1\t0 119\n",
      "106\t49\t3.6\t1\t0 49\n",
      "107\t75\t3.6\t1\t0 75\n",
      "108\t77\t3.6\t1\t0 77\n",
      "109\t61\t3.6\t1\t0 61\n",
      "110\t160\t3.6\t1\t0 160\n",
      "111\t187\t3.6\t1\t0 187\n",
      "112\t167\t3.6\t1\t1 166\n",
      "113\t103\t3.6\t1\t0 103\n",
      "114\t152\t3.6\t1\t0 152\n",
      "115\t54\t3.6\t1\t0 54\n",
      "116\t103\t3.6\t1\t0 103\n",
      "117\t78\t3.6\t1\t0 78\n",
      "118\t94\t3.6\t1\t0 94\n",
      "119\t40\t3.6\t1\t0 40\n",
      "120\t93\t3.6\t1\t0 93\n",
      "121\t54\t3.6\t1\t0 54\n",
      "122\t93\t3.6\t1\t0 93\n",
      "123\t41\t3.6\t1\t0 41\n",
      "124\t20\t3.6\t1\t0 20\n",
      "125\t60\t3.6\t1\t1 59\n",
      "126\t29\t3.6\t1\t0 29\n",
      "127\t92\t3.6\t1\t0 92\n",
      "128\t55\t3.6\t1\t0 55\n",
      "129\t51\t3.6\t1\t0 51\n",
      "130\t63\t3.6\t1\t0 63\n",
      "131\t52\t3.6\t1\t0 52\n",
      "132\t41\t3.6\t1\t0 41\n",
      "133\t80\t3.6\t1\t0 80\n",
      "134\t76\t3.6\t1\t0 76\n",
      "135\t32\t3.6\t1\t0 32\n",
      "136\t43\t3.6\t1\t0 43\n",
      "137\t91\t3.6\t1\t0 91\n",
      "138\t43\t3.6\t1\t0 43\n",
      "139\t46\t3.6\t1\t0 46\n",
      "140\t40\t3.6\t1\t0 40\n",
      "141\t26\t3.6\t1\t0 26\n",
      "142\t14\t3.6\t1\t0 14\n",
      "143\t60\t3.6\t1\t0 60\n",
      "144\t24\t3.6\t1\t0 24\n",
      "145\t49\t3.6\t1\t0 49\n",
      "146\t35\t3.6\t1\t0 35\n",
      "147\t39\t3.6\t1\t0 39\n",
      "148\t20\t3.6\t1\t0 20\n",
      "149\t61\t3.6\t1\t0 61\n",
      "150\t44\t3.6\t1\t0 44\n",
      "151\t35\t3.6\t1\t0 35\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/SRR6399473_1.fastq.gz\n",
      "=============================================\n",
      "59568097 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/trim_galore/SRR6399473_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/SRR6399473_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399473_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/SRR6399473_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/SRR6399473_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 146.486 s (2.459 µs/read; 24.40 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              59,568,097\n",
      "Reads with adapters:                15,033,076 (25.2%)\n",
      "Reads written (passing filters):    59,568,097 (100.0%)\n",
      "\n",
      "Total basepairs processed: 8,270,010,141 bp\n",
      "Quality-trimmed:                 662,890 bp (0.0%)\n",
      "Total written (filtered):  8,248,805,456 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 15033076 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 32.1%\n",
      "  C: 29.3%\n",
      "  G: 18.0%\n",
      "  T: 20.6%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t11648300\t14892024.2\t0\t11648300\n",
      "2\t2697881\t3723006.1\t0\t2697881\n",
      "3\t486268\t930751.5\t0\t486268\n",
      "4\t134178\t232687.9\t0\t134178\n",
      "5\t26856\t58172.0\t0\t26856\n",
      "6\t8541\t14543.0\t0\t8541\n",
      "7\t3508\t3635.7\t0\t3508\n",
      "8\t128\t908.9\t0\t128\n",
      "9\t124\t227.2\t0\t37 87\n",
      "10\t486\t56.8\t1\t4 482\n",
      "11\t527\t14.2\t1\t1 526\n",
      "12\t156\t3.6\t1\t0 156\n",
      "13\t108\t3.6\t1\t0 108\n",
      "14\t108\t3.6\t1\t0 108\n",
      "15\t325\t3.6\t1\t1 324\n",
      "16\t236\t3.6\t1\t2 234\n",
      "17\t242\t3.6\t1\t0 242\n",
      "18\t219\t3.6\t1\t2 217\n",
      "19\t330\t3.6\t1\t0 330\n",
      "20\t348\t3.6\t1\t0 348\n",
      "21\t684\t3.6\t1\t1 683\n",
      "22\t830\t3.6\t1\t1 829\n",
      "23\t917\t3.6\t1\t2 915\n",
      "24\t559\t3.6\t1\t3 556\n",
      "25\t1437\t3.6\t1\t5 1432\n",
      "26\t1296\t3.6\t1\t3 1293\n",
      "27\t921\t3.6\t1\t4 917\n",
      "28\t1318\t3.6\t1\t0 1318\n",
      "29\t673\t3.6\t1\t0 673\n",
      "30\t1050\t3.6\t1\t5 1045\n",
      "31\t2191\t3.6\t1\t7 2184\n",
      "32\t184\t3.6\t1\t0 184\n",
      "33\t210\t3.6\t1\t0 210\n",
      "34\t355\t3.6\t1\t2 353\n",
      "35\t243\t3.6\t1\t1 242\n",
      "36\t156\t3.6\t1\t0 156\n",
      "37\t243\t3.6\t1\t0 243\n",
      "38\t71\t3.6\t1\t0 71\n",
      "39\t177\t3.6\t1\t1 176\n",
      "40\t171\t3.6\t1\t0 171\n",
      "41\t282\t3.6\t1\t2 280\n",
      "42\t154\t3.6\t1\t1 153\n",
      "43\t187\t3.6\t1\t0 187\n",
      "44\t148\t3.6\t1\t0 148\n",
      "45\t312\t3.6\t1\t2 310\n",
      "46\t506\t3.6\t1\t4 502\n",
      "47\t568\t3.6\t1\t0 568\n",
      "48\t147\t3.6\t1\t0 147\n",
      "49\t68\t3.6\t1\t0 68\n",
      "50\t102\t3.6\t1\t0 102\n",
      "51\t211\t3.6\t1\t0 211\n",
      "52\t123\t3.6\t1\t0 123\n",
      "53\t130\t3.6\t1\t0 130\n",
      "54\t75\t3.6\t1\t1 74\n",
      "55\t82\t3.6\t1\t0 82\n",
      "56\t59\t3.6\t1\t0 59\n",
      "57\t37\t3.6\t1\t0 37\n",
      "58\t136\t3.6\t1\t0 136\n",
      "59\t130\t3.6\t1\t2 128\n",
      "60\t84\t3.6\t1\t0 84\n",
      "61\t100\t3.6\t1\t0 100\n",
      "62\t89\t3.6\t1\t0 89\n",
      "63\t119\t3.6\t1\t0 119\n",
      "64\t105\t3.6\t1\t0 105\n",
      "65\t175\t3.6\t1\t1 174\n",
      "66\t68\t3.6\t1\t1 67\n",
      "67\t69\t3.6\t1\t0 69\n",
      "68\t87\t3.6\t1\t0 87\n",
      "69\t107\t3.6\t1\t0 107\n",
      "70\t77\t3.6\t1\t0 77\n",
      "71\t53\t3.6\t1\t0 53\n",
      "72\t82\t3.6\t1\t0 82\n",
      "73\t60\t3.6\t1\t0 60\n",
      "74\t109\t3.6\t1\t0 109\n",
      "75\t71\t3.6\t1\t1 70\n",
      "76\t82\t3.6\t1\t0 82\n",
      "77\t74\t3.6\t1\t0 74\n",
      "78\t89\t3.6\t1\t0 89\n",
      "79\t106\t3.6\t1\t0 106\n",
      "80\t113\t3.6\t1\t0 113\n",
      "81\t195\t3.6\t1\t1 194\n",
      "82\t34\t3.6\t1\t0 34\n",
      "83\t125\t3.6\t1\t0 125\n",
      "84\t59\t3.6\t1\t0 59\n",
      "85\t162\t3.6\t1\t0 162\n",
      "86\t79\t3.6\t1\t0 79\n",
      "87\t59\t3.6\t1\t0 59\n",
      "88\t71\t3.6\t1\t0 71\n",
      "89\t163\t3.6\t1\t0 163\n",
      "90\t46\t3.6\t1\t0 46\n",
      "91\t115\t3.6\t1\t0 115\n",
      "92\t100\t3.6\t1\t0 100\n",
      "93\t88\t3.6\t1\t1 87\n",
      "94\t53\t3.6\t1\t0 53\n",
      "95\t47\t3.6\t1\t0 47\n",
      "96\t109\t3.6\t1\t0 109\n",
      "97\t88\t3.6\t1\t1 87\n",
      "98\t170\t3.6\t1\t2 168\n",
      "99\t47\t3.6\t1\t0 47\n",
      "100\t105\t3.6\t1\t0 105\n",
      "101\t53\t3.6\t1\t0 53\n",
      "102\t126\t3.6\t1\t0 126\n",
      "103\t84\t3.6\t1\t0 84\n",
      "104\t94\t3.6\t1\t0 94\n",
      "105\t95\t3.6\t1\t0 95\n",
      "106\t53\t3.6\t1\t0 53\n",
      "107\t61\t3.6\t1\t0 61\n",
      "108\t94\t3.6\t1\t0 94\n",
      "109\t73\t3.6\t1\t0 73\n",
      "110\t173\t3.6\t1\t0 173\n",
      "111\t126\t3.6\t1\t0 126\n",
      "112\t145\t3.6\t1\t0 145\n",
      "113\t103\t3.6\t1\t0 103\n",
      "114\t101\t3.6\t1\t2 99\n",
      "115\t46\t3.6\t1\t0 46\n",
      "116\t58\t3.6\t1\t0 58\n",
      "117\t62\t3.6\t1\t0 62\n",
      "118\t126\t3.6\t1\t0 126\n",
      "119\t41\t3.6\t1\t0 41\n",
      "120\t105\t3.6\t1\t0 105\n",
      "121\t53\t3.6\t1\t0 53\n",
      "122\t70\t3.6\t1\t1 69\n",
      "123\t34\t3.6\t1\t0 34\n",
      "124\t32\t3.6\t1\t0 32\n",
      "125\t68\t3.6\t1\t0 68\n",
      "126\t15\t3.6\t1\t0 15\n",
      "127\t86\t3.6\t1\t0 86\n",
      "128\t54\t3.6\t1\t0 54\n",
      "129\t46\t3.6\t1\t0 46\n",
      "130\t53\t3.6\t1\t0 53\n",
      "131\t41\t3.6\t1\t0 41\n",
      "132\t58\t3.6\t1\t0 58\n",
      "133\t75\t3.6\t1\t0 75\n",
      "134\t92\t3.6\t1\t0 92\n",
      "135\t25\t3.6\t1\t0 25\n",
      "136\t53\t3.6\t1\t0 53\n",
      "137\t76\t3.6\t1\t0 76\n",
      "138\t48\t3.6\t1\t0 48\n",
      "139\t19\t3.6\t1\t0 19\n",
      "140\t48\t3.6\t1\t0 48\n",
      "141\t16\t3.6\t1\t0 16\n",
      "142\t22\t3.6\t1\t0 22\n",
      "143\t67\t3.6\t1\t0 67\n",
      "144\t30\t3.6\t1\t1 29\n",
      "145\t36\t3.6\t1\t0 36\n",
      "146\t28\t3.6\t1\t0 28\n",
      "147\t27\t3.6\t1\t0 27\n",
      "148\t39\t3.6\t1\t0 39\n",
      "149\t39\t3.6\t1\t1 38\n",
      "150\t42\t3.6\t1\t0 42\n",
      "151\t19\t3.6\t1\t0 19\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399473/SRR6399473_2.fastq.gz\n",
      "=============================================\n",
      "59568097 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399473_1_trimmed.fq.gz and SRR6399473_2_trimmed.fq.gz\n",
      "file_1: SRR6399473_1_trimmed.fq.gz, file_2: SRR6399473_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399473_1_trimmed.fq.gz and SRR6399473_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399473_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399473_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 59568097\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 13762 (0.02%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399473_1_trimmed.fq.gz and SRR6399473_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   3%|█                                        |  ETA: 16.54 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2141.482186 seconds (65.99 k allocations: 1.879 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 51,102,509\n",
      "reads read      : 102,205,018\n",
      "reads written   : 102,205,018\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/SRR6399474_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count Illumina: 0, count Nextera: 0, count smallRNA: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/trim_galore/SRR6399474_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/SRR6399474_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399474_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/SRR6399474_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/SRR6399474_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 121.927 s (2.386 µs/read; 25.15 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              51,102,509\n",
      "Reads with adapters:                15,470,412 (30.3%)\n",
      "Reads written (passing filters):    51,102,509 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,627,143,712 bp\n",
      "Quality-trimmed:                  39,339 bp (0.0%)\n",
      "Total written (filtered):  7,606,063,054 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 15470412 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 32.4%\n",
      "  C: 24.9%\n",
      "  G: 21.1%\n",
      "  T: 21.6%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t11259729\t12775627.2\t0\t11259729\n",
      "2\t3236152\t3193906.8\t0\t3236152\n",
      "3\t728578\t798476.7\t0\t728578\n",
      "4\t165126\t199619.2\t0\t165126\n",
      "5\t42220\t49904.8\t0\t42220\n",
      "6\t20615\t12476.2\t0\t20615\n",
      "7\t17932\t3119.0\t0\t17932\n",
      "8\t2\t779.8\t0\t2\n",
      "9\t10\t194.9\t0\t0 10\n",
      "10\t8\t48.7\t1\t1 7\n",
      "11\t3\t12.2\t1\t0 3\n",
      "12\t2\t3.0\t1\t0 2\n",
      "21\t1\t0.8\t1\t0 1\n",
      "30\t3\t0.8\t1\t0 3\n",
      "31\t1\t0.8\t1\t0 1\n",
      "36\t1\t0.8\t1\t0 1\n",
      "39\t2\t0.8\t1\t0 2\n",
      "43\t1\t0.8\t1\t0 1\n",
      "51\t2\t0.8\t1\t0 2\n",
      "52\t1\t0.8\t1\t0 1\n",
      "55\t1\t0.8\t1\t0 1\n",
      "57\t1\t0.8\t1\t0 1\n",
      "58\t2\t0.8\t1\t0 2\n",
      "59\t1\t0.8\t1\t0 1\n",
      "61\t1\t0.8\t1\t0 1\n",
      "74\t4\t0.8\t1\t0 4\n",
      "75\t1\t0.8\t1\t0 1\n",
      "77\t1\t0.8\t1\t0 1\n",
      "90\t1\t0.8\t1\t0 1\n",
      "92\t1\t0.8\t1\t0 1\n",
      "97\t1\t0.8\t1\t0 1\n",
      "99\t1\t0.8\t1\t0 1\n",
      "113\t1\t0.8\t1\t0 1\n",
      "115\t1\t0.8\t1\t0 1\n",
      "117\t1\t0.8\t1\t0 1\n",
      "123\t1\t0.8\t1\t0 1\n",
      "129\t1\t0.8\t1\t0 1\n",
      "131\t1\t0.8\t1\t0 1\n",
      "136\t1\t0.8\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/SRR6399474_1.fastq.gz\n",
      "=============================================\n",
      "51102509 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/trim_galore/SRR6399474_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/SRR6399474_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399474_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/SRR6399474_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/SRR6399474_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 137.247 s (2.686 µs/read; 22.34 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              51,102,509\n",
      "Reads with adapters:                16,022,306 (31.4%)\n",
      "Reads written (passing filters):    51,102,509 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,469,669,615 bp\n",
      "Quality-trimmed:                 139,622 bp (0.0%)\n",
      "Total written (filtered):  7,447,873,749 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 16022306 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 38.6%\n",
      "  C: 22.6%\n",
      "  G: 18.3%\n",
      "  T: 20.5%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t11788902\t12775627.2\t0\t11788902\n",
      "2\t3235818\t3193906.8\t0\t3235818\n",
      "3\t747658\t798476.7\t0\t747658\n",
      "4\t164095\t199619.2\t0\t164095\n",
      "5\t42199\t49904.8\t0\t42199\n",
      "6\t22762\t12476.2\t0\t22762\n",
      "7\t20818\t3119.0\t0\t20818\n",
      "9\t3\t194.9\t0\t0 3\n",
      "10\t7\t48.7\t1\t2 5\n",
      "11\t3\t12.2\t1\t0 3\n",
      "12\t2\t3.0\t1\t0 2\n",
      "25\t1\t0.8\t1\t0 1\n",
      "30\t2\t0.8\t1\t0 2\n",
      "34\t1\t0.8\t1\t0 1\n",
      "38\t1\t0.8\t1\t0 1\n",
      "41\t2\t0.8\t1\t0 2\n",
      "43\t2\t0.8\t1\t0 2\n",
      "44\t1\t0.8\t1\t0 1\n",
      "55\t1\t0.8\t1\t0 1\n",
      "60\t2\t0.8\t1\t0 2\n",
      "62\t1\t0.8\t1\t0 1\n",
      "64\t1\t0.8\t1\t0 1\n",
      "66\t1\t0.8\t1\t0 1\n",
      "72\t4\t0.8\t1\t0 4\n",
      "73\t3\t0.8\t1\t0 3\n",
      "74\t1\t0.8\t1\t0 1\n",
      "76\t1\t0.8\t1\t0 1\n",
      "78\t1\t0.8\t1\t0 1\n",
      "85\t1\t0.8\t1\t0 1\n",
      "86\t1\t0.8\t1\t0 1\n",
      "91\t1\t0.8\t1\t0 1\n",
      "100\t1\t0.8\t1\t0 1\n",
      "105\t1\t0.8\t1\t0 1\n",
      "110\t2\t0.8\t1\t0 2\n",
      "113\t1\t0.8\t1\t0 1\n",
      "116\t1\t0.8\t1\t0 1\n",
      "127\t1\t0.8\t1\t0 1\n",
      "128\t1\t0.8\t1\t0 1\n",
      "131\t1\t0.8\t1\t0 1\n",
      "132\t1\t0.8\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399474/SRR6399474_2.fastq.gz\n",
      "=============================================\n",
      "51102509 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399474_1_trimmed.fq.gz and SRR6399474_2_trimmed.fq.gz\n",
      "file_1: SRR6399474_1_trimmed.fq.gz, file_2: SRR6399474_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399474_1_trimmed.fq.gz and SRR6399474_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399474_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399474_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 51102509\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 6109 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399474_1_trimmed.fq.gz and SRR6399474_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   3%|█▏                                       |  ETA: 16.78 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1821.954457 seconds (56.14 k allocations: 1.598 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 42,912,566\n",
      "reads read      : 85,825,132\n",
      "reads written   : 85,825,132\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/SRR6399475_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count Illumina: 0, count smallRNA: 0, count Nextera: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/trim_galore/SRR6399475_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/SRR6399475_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399475_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/SRR6399475_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/SRR6399475_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 115.656 s (2.695 µs/read; 22.26 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              42,912,566\n",
      "Reads with adapters:                13,476,306 (31.4%)\n",
      "Reads written (passing filters):    42,912,566 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,302,514,536 bp\n",
      "Quality-trimmed:                 189,542 bp (0.0%)\n",
      "Total written (filtered):  6,284,109,548 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 13476306 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 30.4%\n",
      "  C: 25.8%\n",
      "  G: 19.5%\n",
      "  T: 24.3%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t9760977\t10728141.5\t0\t9760977\n",
      "2\t2967364\t2682035.4\t0\t2967364\n",
      "3\t579394\t670508.8\t0\t579394\n",
      "4\t108971\t167627.2\t0\t108971\n",
      "5\t29373\t41906.8\t0\t29373\n",
      "6\t16931\t10476.7\t0\t16931\n",
      "7\t13215\t2619.2\t0\t13215\n",
      "8\t2\t654.8\t0\t2\n",
      "9\t8\t163.7\t0\t2 6\n",
      "10\t8\t40.9\t1\t3 5\n",
      "11\t2\t10.2\t1\t0 2\n",
      "12\t4\t2.6\t1\t0 4\n",
      "16\t1\t0.6\t1\t0 1\n",
      "20\t1\t0.6\t1\t0 1\n",
      "27\t1\t0.6\t1\t0 1\n",
      "30\t2\t0.6\t1\t0 2\n",
      "34\t1\t0.6\t1\t0 1\n",
      "38\t1\t0.6\t1\t0 1\n",
      "40\t1\t0.6\t1\t0 1\n",
      "45\t1\t0.6\t1\t0 1\n",
      "46\t2\t0.6\t1\t0 2\n",
      "55\t1\t0.6\t1\t0 1\n",
      "56\t1\t0.6\t1\t0 1\n",
      "62\t1\t0.6\t1\t0 1\n",
      "63\t3\t0.6\t1\t0 3\n",
      "71\t1\t0.6\t1\t0 1\n",
      "72\t1\t0.6\t1\t0 1\n",
      "73\t3\t0.6\t1\t0 3\n",
      "74\t12\t0.6\t1\t0 12\n",
      "75\t1\t0.6\t1\t0 1\n",
      "83\t1\t0.6\t1\t0 1\n",
      "86\t1\t0.6\t1\t0 1\n",
      "87\t1\t0.6\t1\t0 1\n",
      "88\t1\t0.6\t1\t0 1\n",
      "89\t1\t0.6\t1\t0 1\n",
      "99\t2\t0.6\t1\t0 2\n",
      "101\t1\t0.6\t1\t0 1\n",
      "104\t1\t0.6\t1\t0 1\n",
      "105\t1\t0.6\t1\t0 1\n",
      "113\t2\t0.6\t1\t0 2\n",
      "118\t2\t0.6\t1\t0 2\n",
      "120\t1\t0.6\t1\t0 1\n",
      "121\t1\t0.6\t1\t0 1\n",
      "122\t1\t0.6\t1\t0 1\n",
      "123\t1\t0.6\t1\t0 1\n",
      "132\t1\t0.6\t1\t0 1\n",
      "133\t2\t0.6\t1\t0 2\n",
      "135\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/SRR6399475_1.fastq.gz\n",
      "=============================================\n",
      "42912566 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/trim_galore/SRR6399475_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/SRR6399475_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399475_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/SRR6399475_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/SRR6399475_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 114.257 s (2.663 µs/read; 22.53 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              42,912,566\n",
      "Reads with adapters:                11,674,078 (27.2%)\n",
      "Reads written (passing filters):    42,912,566 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,139,442,120 bp\n",
      "Quality-trimmed:                 296,355 bp (0.0%)\n",
      "Total written (filtered):  6,123,150,564 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 11674078 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 37.1%\n",
      "  C: 23.4%\n",
      "  G: 17.1%\n",
      "  T: 22.3%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t8260452\t10728141.5\t0\t8260452\n",
      "2\t2817492\t2682035.4\t0\t2817492\n",
      "3\t414876\t670508.8\t0\t414876\n",
      "4\t112086\t167627.2\t0\t112086\n",
      "5\t31218\t41906.8\t0\t31218\n",
      "6\t19149\t10476.7\t0\t19149\n",
      "7\t18687\t2619.2\t0\t18687\n",
      "8\t25\t654.8\t0\t25\n",
      "9\t10\t163.7\t0\t0 10\n",
      "10\t15\t40.9\t1\t11 4\n",
      "11\t5\t10.2\t1\t0 5\n",
      "12\t6\t2.6\t1\t0 6\n",
      "21\t1\t0.6\t1\t0 1\n",
      "26\t2\t0.6\t1\t0 2\n",
      "30\t1\t0.6\t1\t0 1\n",
      "32\t1\t0.6\t1\t0 1\n",
      "42\t1\t0.6\t1\t0 1\n",
      "46\t1\t0.6\t1\t0 1\n",
      "49\t1\t0.6\t1\t0 1\n",
      "52\t1\t0.6\t1\t0 1\n",
      "60\t1\t0.6\t1\t0 1\n",
      "66\t4\t0.6\t1\t0 4\n",
      "67\t1\t0.6\t1\t0 1\n",
      "68\t4\t0.6\t1\t0 4\n",
      "69\t1\t0.6\t1\t0 1\n",
      "70\t4\t0.6\t1\t0 4\n",
      "72\t2\t0.6\t1\t0 2\n",
      "73\t2\t0.6\t1\t0 2\n",
      "74\t9\t0.6\t1\t0 9\n",
      "75\t1\t0.6\t1\t0 1\n",
      "80\t1\t0.6\t1\t0 1\n",
      "83\t2\t0.6\t1\t0 2\n",
      "85\t1\t0.6\t1\t0 1\n",
      "89\t1\t0.6\t1\t0 1\n",
      "97\t1\t0.6\t1\t0 1\n",
      "99\t1\t0.6\t1\t0 1\n",
      "104\t1\t0.6\t1\t0 1\n",
      "105\t1\t0.6\t1\t0 1\n",
      "113\t1\t0.6\t1\t0 1\n",
      "114\t1\t0.6\t1\t0 1\n",
      "118\t1\t0.6\t1\t0 1\n",
      "121\t1\t0.6\t1\t0 1\n",
      "122\t1\t0.6\t1\t0 1\n",
      "123\t1\t0.6\t1\t0 1\n",
      "132\t2\t0.6\t1\t0 2\n",
      "133\t2\t0.6\t1\t0 2\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399475/SRR6399475_2.fastq.gz\n",
      "=============================================\n",
      "42912566 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399475_1_trimmed.fq.gz and SRR6399475_2_trimmed.fq.gz\n",
      "file_1: SRR6399475_1_trimmed.fq.gz, file_2: SRR6399475_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399475_1_trimmed.fq.gz and SRR6399475_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399475_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399475_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 42912566\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 6222 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399475_1_trimmed.fq.gz and SRR6399475_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   3%|█▏                                       |  ETA: 16.82 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1600.976098 seconds (49.34 k allocations: 1.405 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 34,504,343\n",
      "reads read      : 69,008,686\n",
      "reads written   : 69,008,686\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1764.852746 seconds (54.39 k allocations: 1.548 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 40,840,409\n",
      "reads read      : 81,680,818\n",
      "reads written   : 81,680,818\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/SRR6399478_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count Illumina: 0, count smallRNA: 0, count Nextera: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/trim_galore/SRR6399478_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/SRR6399478_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399478_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/SRR6399478_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/SRR6399478_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 96.087 s (2.353 µs/read; 25.50 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              40,840,409\n",
      "Reads with adapters:                12,839,318 (31.4%)\n",
      "Reads written (passing filters):    40,840,409 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,110,361,621 bp\n",
      "Quality-trimmed:                  34,926 bp (0.0%)\n",
      "Total written (filtered):  6,092,644,802 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 12839318 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 29.9%\n",
      "  C: 25.1%\n",
      "  G: 21.3%\n",
      "  T: 23.7%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t9211376\t10210102.2\t0\t9211376\n",
      "2\t2736384\t2552525.6\t0\t2736384\n",
      "3\t684938\t638131.4\t0\t684938\n",
      "4\t141297\t159532.8\t0\t141297\n",
      "5\t31347\t39883.2\t0\t31347\n",
      "6\t18197\t9970.8\t0\t18197\n",
      "7\t15702\t2492.7\t0\t15702\n",
      "8\t2\t623.2\t0\t2\n",
      "9\t49\t155.8\t0\t1 48\n",
      "10\t5\t38.9\t1\t0 5\n",
      "11\t2\t9.7\t1\t0 2\n",
      "12\t2\t2.4\t1\t0 2\n",
      "23\t1\t0.6\t1\t0 1\n",
      "28\t2\t0.6\t1\t0 2\n",
      "29\t1\t0.6\t1\t0 1\n",
      "40\t1\t0.6\t1\t0 1\n",
      "41\t1\t0.6\t1\t0 1\n",
      "84\t1\t0.6\t1\t0 1\n",
      "89\t1\t0.6\t1\t0 1\n",
      "93\t1\t0.6\t1\t0 1\n",
      "96\t1\t0.6\t1\t0 1\n",
      "97\t1\t0.6\t1\t0 1\n",
      "107\t2\t0.6\t1\t0 2\n",
      "109\t1\t0.6\t1\t0 1\n",
      "128\t1\t0.6\t1\t0 1\n",
      "130\t1\t0.6\t1\t0 1\n",
      "134\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/SRR6399478_1.fastq.gz\n",
      "=============================================\n",
      "40840409 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/trim_galore/SRR6399478_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/SRR6399478_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399478_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/SRR6399478_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/SRR6399478_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 108.050 s (2.646 µs/read; 22.68 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              40,840,409\n",
      "Reads with adapters:                12,936,398 (31.7%)\n",
      "Reads written (passing filters):    40,840,409 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,993,879,008 bp\n",
      "Quality-trimmed:                  89,004 bp (0.0%)\n",
      "Total written (filtered):  5,976,241,986 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 12936398 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.1%\n",
      "  C: 23.5%\n",
      "  G: 19.2%\n",
      "  T: 23.2%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t9548135\t10210102.2\t0\t9548135\n",
      "2\t2486987\t2552525.6\t0\t2486987\n",
      "3\t701161\t638131.4\t0\t701161\n",
      "4\t132754\t159532.8\t0\t132754\n",
      "5\t31198\t39883.2\t0\t31198\n",
      "6\t18943\t9970.8\t0\t18943\n",
      "7\t17127\t2492.7\t0\t17127\n",
      "9\t67\t155.8\t0\t0 67\n",
      "10\t7\t38.9\t1\t0 7\n",
      "11\t2\t9.7\t1\t1 1\n",
      "12\t2\t2.4\t1\t0 2\n",
      "28\t1\t0.6\t1\t0 1\n",
      "37\t1\t0.6\t1\t0 1\n",
      "53\t1\t0.6\t1\t0 1\n",
      "73\t1\t0.6\t1\t0 1\n",
      "74\t2\t0.6\t1\t0 2\n",
      "75\t1\t0.6\t1\t0 1\n",
      "82\t1\t0.6\t1\t0 1\n",
      "83\t1\t0.6\t1\t0 1\n",
      "85\t1\t0.6\t1\t0 1\n",
      "86\t1\t0.6\t1\t0 1\n",
      "94\t1\t0.6\t1\t0 1\n",
      "95\t1\t0.6\t1\t0 1\n",
      "106\t1\t0.6\t1\t0 1\n",
      "109\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399478/SRR6399478_2.fastq.gz\n",
      "=============================================\n",
      "40840409 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399478_1_trimmed.fq.gz and SRR6399478_2_trimmed.fq.gz\n",
      "file_1: SRR6399478_1_trimmed.fq.gz, file_2: SRR6399478_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399478_1_trimmed.fq.gz and SRR6399478_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399478_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399478_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 40840409\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 4184 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399478_1_trimmed.fq.gz and SRR6399478_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   3%|█▍                                       |  ETA: 16.73 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2356.304996 seconds (72.60 k allocations: 2.066 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 57,015,422\n",
      "reads read      : 114,030,844\n",
      "reads written   : 114,030,844\n",
      "\n",
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/SRR6399479_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count smallRNA: 0, count Illumina: 0, count Nextera: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/trim_galore/SRR6399479_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/SRR6399479_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399479_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/SRR6399479_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/SRR6399479_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 132.918 s (2.331 µs/read; 25.74 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              57,015,422\n",
      "Reads with adapters:                15,911,590 (27.9%)\n",
      "Reads written (passing filters):    57,015,422 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,826,208,523 bp\n",
      "Quality-trimmed:                 478,228 bp (0.0%)\n",
      "Total written (filtered):  7,803,345,958 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 15911590 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 32.1%\n",
      "  C: 27.9%\n",
      "  G: 18.2%\n",
      "  T: 21.8%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t11379489\t14253855.5\t0\t11379489\n",
      "2\t3512307\t3563463.9\t0\t3512307\n",
      "3\t623287\t890866.0\t0\t623287\n",
      "4\t182191\t222716.5\t0\t182191\n",
      "5\t81057\t55679.1\t0\t81057\n",
      "6\t65659\t13919.8\t0\t65659\n",
      "7\t65016\t3479.9\t0\t65016\n",
      "8\t3\t870.0\t0\t3\n",
      "9\t129\t217.5\t0\t5 124\n",
      "10\t42\t54.4\t1\t6 36\n",
      "11\t27\t13.6\t1\t0 27\n",
      "12\t14\t3.4\t1\t1 13\n",
      "13\t1\t0.8\t1\t0 1\n",
      "14\t4\t0.8\t1\t0 4\n",
      "16\t2\t0.8\t1\t0 2\n",
      "17\t1\t0.8\t1\t0 1\n",
      "19\t2\t0.8\t1\t0 2\n",
      "20\t9\t0.8\t1\t0 9\n",
      "21\t14\t0.8\t1\t0 14\n",
      "22\t7\t0.8\t1\t0 7\n",
      "23\t12\t0.8\t1\t0 12\n",
      "24\t21\t0.8\t1\t0 21\n",
      "25\t43\t0.8\t1\t0 43\n",
      "26\t42\t0.8\t1\t0 42\n",
      "27\t45\t0.8\t1\t0 45\n",
      "28\t38\t0.8\t1\t0 38\n",
      "29\t19\t0.8\t1\t0 19\n",
      "30\t54\t0.8\t1\t0 54\n",
      "31\t337\t0.8\t1\t0 337\n",
      "32\t165\t0.8\t1\t0 165\n",
      "33\t110\t0.8\t1\t0 110\n",
      "34\t52\t0.8\t1\t0 52\n",
      "35\t59\t0.8\t1\t0 59\n",
      "36\t56\t0.8\t1\t0 56\n",
      "37\t139\t0.8\t1\t0 139\n",
      "38\t62\t0.8\t1\t0 62\n",
      "39\t40\t0.8\t1\t0 40\n",
      "40\t43\t0.8\t1\t0 43\n",
      "41\t31\t0.8\t1\t0 31\n",
      "42\t41\t0.8\t1\t0 41\n",
      "43\t13\t0.8\t1\t0 13\n",
      "44\t10\t0.8\t1\t0 10\n",
      "45\t4\t0.8\t1\t0 4\n",
      "46\t5\t0.8\t1\t0 5\n",
      "48\t3\t0.8\t1\t0 3\n",
      "49\t3\t0.8\t1\t0 3\n",
      "50\t3\t0.8\t1\t0 3\n",
      "51\t5\t0.8\t1\t0 5\n",
      "52\t6\t0.8\t1\t0 6\n",
      "53\t2\t0.8\t1\t0 2\n",
      "54\t2\t0.8\t1\t0 2\n",
      "55\t1\t0.8\t1\t0 1\n",
      "56\t7\t0.8\t1\t0 7\n",
      "57\t5\t0.8\t1\t0 5\n",
      "58\t9\t0.8\t1\t0 9\n",
      "59\t4\t0.8\t1\t0 4\n",
      "60\t4\t0.8\t1\t0 4\n",
      "61\t14\t0.8\t1\t0 14\n",
      "62\t42\t0.8\t1\t0 42\n",
      "63\t32\t0.8\t1\t0 32\n",
      "64\t14\t0.8\t1\t0 14\n",
      "65\t17\t0.8\t1\t0 17\n",
      "66\t18\t0.8\t1\t0 18\n",
      "67\t14\t0.8\t1\t0 14\n",
      "68\t15\t0.8\t1\t0 15\n",
      "69\t18\t0.8\t1\t0 18\n",
      "70\t9\t0.8\t1\t0 9\n",
      "71\t19\t0.8\t1\t0 19\n",
      "72\t22\t0.8\t1\t0 22\n",
      "73\t32\t0.8\t1\t0 32\n",
      "74\t74\t0.8\t1\t0 74\n",
      "75\t11\t0.8\t1\t0 11\n",
      "76\t6\t0.8\t1\t0 6\n",
      "77\t13\t0.8\t1\t0 13\n",
      "78\t18\t0.8\t1\t0 18\n",
      "79\t7\t0.8\t1\t0 7\n",
      "80\t10\t0.8\t1\t0 10\n",
      "81\t16\t0.8\t1\t0 16\n",
      "82\t9\t0.8\t1\t0 9\n",
      "83\t8\t0.8\t1\t0 8\n",
      "84\t6\t0.8\t1\t0 6\n",
      "85\t8\t0.8\t1\t0 8\n",
      "86\t9\t0.8\t1\t0 9\n",
      "87\t8\t0.8\t1\t0 8\n",
      "88\t8\t0.8\t1\t0 8\n",
      "89\t11\t0.8\t1\t0 11\n",
      "90\t13\t0.8\t1\t0 13\n",
      "91\t7\t0.8\t1\t0 7\n",
      "92\t13\t0.8\t1\t0 13\n",
      "93\t7\t0.8\t1\t0 7\n",
      "94\t7\t0.8\t1\t0 7\n",
      "95\t13\t0.8\t1\t0 13\n",
      "96\t9\t0.8\t1\t0 9\n",
      "97\t10\t0.8\t1\t0 10\n",
      "98\t6\t0.8\t1\t0 6\n",
      "99\t7\t0.8\t1\t0 7\n",
      "100\t3\t0.8\t1\t0 3\n",
      "101\t8\t0.8\t1\t0 8\n",
      "102\t9\t0.8\t1\t0 9\n",
      "103\t4\t0.8\t1\t0 4\n",
      "104\t4\t0.8\t1\t0 4\n",
      "105\t10\t0.8\t1\t0 10\n",
      "106\t9\t0.8\t1\t0 9\n",
      "107\t9\t0.8\t1\t0 9\n",
      "108\t11\t0.8\t1\t0 11\n",
      "109\t10\t0.8\t1\t0 10\n",
      "110\t15\t0.8\t1\t0 15\n",
      "111\t8\t0.8\t1\t0 8\n",
      "112\t7\t0.8\t1\t0 7\n",
      "113\t6\t0.8\t1\t0 6\n",
      "114\t10\t0.8\t1\t0 10\n",
      "115\t4\t0.8\t1\t0 4\n",
      "116\t5\t0.8\t1\t0 5\n",
      "117\t10\t0.8\t1\t0 10\n",
      "118\t4\t0.8\t1\t0 4\n",
      "119\t4\t0.8\t1\t0 4\n",
      "120\t9\t0.8\t1\t0 9\n",
      "121\t6\t0.8\t1\t0 6\n",
      "122\t7\t0.8\t1\t0 7\n",
      "123\t6\t0.8\t1\t0 6\n",
      "124\t11\t0.8\t1\t0 11\n",
      "125\t6\t0.8\t1\t0 6\n",
      "126\t2\t0.8\t1\t0 2\n",
      "127\t6\t0.8\t1\t0 6\n",
      "128\t7\t0.8\t1\t0 7\n",
      "129\t8\t0.8\t1\t0 8\n",
      "130\t2\t0.8\t1\t0 2\n",
      "131\t4\t0.8\t1\t0 4\n",
      "132\t2\t0.8\t1\t0 2\n",
      "133\t6\t0.8\t1\t0 6\n",
      "134\t4\t0.8\t1\t0 4\n",
      "135\t5\t0.8\t1\t0 5\n",
      "136\t3\t0.8\t1\t0 3\n",
      "137\t4\t0.8\t1\t0 4\n",
      "138\t4\t0.8\t1\t0 4\n",
      "139\t2\t0.8\t1\t0 2\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/SRR6399479_1.fastq.gz\n",
      "=============================================\n",
      "57015422 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/trim_galore/SRR6399479_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/SRR6399479_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399479_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/SRR6399479_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/SRR6399479_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 142.817 s (2.505 µs/read; 23.95 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              57,015,422\n",
      "Reads with adapters:                21,679,240 (38.0%)\n",
      "Reads written (passing filters):    57,015,422 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,728,091,006 bp\n",
      "Quality-trimmed:                 864,782 bp (0.0%)\n",
      "Total written (filtered):  7,699,462,257 bp (99.6%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 21679240 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 61.9%\n",
      "  C: 15.5%\n",
      "  G: 9.4%\n",
      "  T: 13.1%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t17469392\t14253855.5\t0\t17469392\n",
      "2\t3366906\t3563463.9\t0\t3366906\n",
      "3\t418097\t890866.0\t0\t418097\n",
      "4\t178646\t222716.5\t0\t178646\n",
      "5\t78865\t55679.1\t0\t78865\n",
      "6\t67955\t13919.8\t0\t67955\n",
      "7\t96315\t3479.9\t0\t96315\n",
      "8\t101\t870.0\t0\t101\n",
      "9\t123\t217.5\t0\t2 121\n",
      "10\t66\t54.4\t1\t3 63\n",
      "11\t42\t13.6\t1\t0 42\n",
      "12\t71\t3.4\t1\t1 70\n",
      "13\t6\t0.8\t1\t0 6\n",
      "14\t46\t0.8\t1\t0 46\n",
      "15\t24\t0.8\t1\t0 24\n",
      "16\t21\t0.8\t1\t0 21\n",
      "17\t145\t0.8\t1\t0 145\n",
      "18\t12\t0.8\t1\t0 12\n",
      "19\t71\t0.8\t1\t0 71\n",
      "20\t81\t0.8\t1\t0 81\n",
      "21\t147\t0.8\t1\t0 147\n",
      "22\t91\t0.8\t1\t0 91\n",
      "23\t167\t0.8\t1\t0 167\n",
      "24\t146\t0.8\t1\t0 146\n",
      "25\t101\t0.8\t1\t0 101\n",
      "26\t69\t0.8\t1\t0 69\n",
      "27\t65\t0.8\t1\t0 65\n",
      "28\t69\t0.8\t1\t0 69\n",
      "29\t24\t0.8\t1\t0 24\n",
      "30\t101\t0.8\t1\t0 101\n",
      "31\t91\t0.8\t1\t0 91\n",
      "32\t59\t0.8\t1\t0 59\n",
      "33\t73\t0.8\t1\t0 73\n",
      "34\t50\t0.8\t1\t0 50\n",
      "35\t59\t0.8\t1\t0 59\n",
      "36\t40\t0.8\t1\t0 40\n",
      "37\t46\t0.8\t1\t0 46\n",
      "38\t26\t0.8\t1\t0 26\n",
      "39\t23\t0.8\t1\t0 23\n",
      "40\t24\t0.8\t1\t0 24\n",
      "41\t24\t0.8\t1\t0 24\n",
      "42\t32\t0.8\t1\t0 32\n",
      "43\t11\t0.8\t1\t0 11\n",
      "44\t2\t0.8\t1\t0 2\n",
      "45\t2\t0.8\t1\t0 2\n",
      "46\t2\t0.8\t1\t0 2\n",
      "47\t3\t0.8\t1\t0 3\n",
      "48\t1\t0.8\t1\t0 1\n",
      "49\t1\t0.8\t1\t0 1\n",
      "50\t3\t0.8\t1\t0 3\n",
      "51\t4\t0.8\t1\t0 4\n",
      "52\t3\t0.8\t1\t0 3\n",
      "53\t7\t0.8\t1\t0 7\n",
      "54\t3\t0.8\t1\t0 3\n",
      "55\t4\t0.8\t1\t0 4\n",
      "56\t3\t0.8\t1\t0 3\n",
      "57\t16\t0.8\t1\t0 16\n",
      "58\t12\t0.8\t1\t0 12\n",
      "59\t14\t0.8\t1\t0 14\n",
      "60\t7\t0.8\t1\t0 7\n",
      "61\t8\t0.8\t1\t0 8\n",
      "62\t9\t0.8\t1\t0 9\n",
      "63\t10\t0.8\t1\t0 10\n",
      "64\t12\t0.8\t1\t0 12\n",
      "65\t10\t0.8\t1\t0 10\n",
      "66\t13\t0.8\t1\t0 13\n",
      "67\t24\t0.8\t1\t0 24\n",
      "68\t16\t0.8\t1\t0 16\n",
      "69\t12\t0.8\t1\t0 12\n",
      "70\t14\t0.8\t1\t0 14\n",
      "71\t9\t0.8\t1\t0 9\n",
      "72\t18\t0.8\t1\t0 18\n",
      "73\t22\t0.8\t1\t0 22\n",
      "74\t60\t0.8\t1\t0 60\n",
      "75\t7\t0.8\t1\t0 7\n",
      "76\t5\t0.8\t1\t0 5\n",
      "77\t7\t0.8\t1\t0 7\n",
      "78\t10\t0.8\t1\t0 10\n",
      "79\t12\t0.8\t1\t0 12\n",
      "80\t11\t0.8\t1\t0 11\n",
      "81\t10\t0.8\t1\t0 10\n",
      "82\t10\t0.8\t1\t0 10\n",
      "83\t7\t0.8\t1\t0 7\n",
      "84\t6\t0.8\t1\t0 6\n",
      "85\t10\t0.8\t1\t0 10\n",
      "86\t8\t0.8\t1\t0 8\n",
      "87\t11\t0.8\t1\t0 11\n",
      "88\t7\t0.8\t1\t0 7\n",
      "89\t9\t0.8\t1\t0 9\n",
      "90\t8\t0.8\t1\t0 8\n",
      "91\t4\t0.8\t1\t0 4\n",
      "92\t8\t0.8\t1\t0 8\n",
      "93\t8\t0.8\t1\t0 8\n",
      "94\t6\t0.8\t1\t0 6\n",
      "95\t6\t0.8\t1\t0 6\n",
      "96\t5\t0.8\t1\t0 5\n",
      "97\t8\t0.8\t1\t0 8\n",
      "98\t8\t0.8\t1\t0 8\n",
      "99\t8\t0.8\t1\t0 8\n",
      "100\t3\t0.8\t1\t0 3\n",
      "101\t4\t0.8\t1\t0 4\n",
      "102\t13\t0.8\t1\t0 13\n",
      "103\t4\t0.8\t1\t0 4\n",
      "104\t4\t0.8\t1\t0 4\n",
      "105\t6\t0.8\t1\t0 6\n",
      "106\t8\t0.8\t1\t0 8\n",
      "107\t5\t0.8\t1\t0 5\n",
      "108\t12\t0.8\t1\t0 12\n",
      "109\t10\t0.8\t1\t0 10\n",
      "110\t11\t0.8\t1\t0 11\n",
      "111\t9\t0.8\t1\t0 9\n",
      "112\t4\t0.8\t1\t0 4\n",
      "113\t3\t0.8\t1\t0 3\n",
      "114\t10\t0.8\t1\t0 10\n",
      "115\t3\t0.8\t1\t0 3\n",
      "116\t7\t0.8\t1\t0 7\n",
      "117\t7\t0.8\t1\t0 7\n",
      "118\t3\t0.8\t1\t0 3\n",
      "119\t4\t0.8\t1\t0 4\n",
      "120\t3\t0.8\t1\t0 3\n",
      "121\t3\t0.8\t1\t0 3\n",
      "122\t5\t0.8\t1\t0 5\n",
      "123\t4\t0.8\t1\t0 4\n",
      "124\t7\t0.8\t1\t0 7\n",
      "125\t4\t0.8\t1\t0 4\n",
      "126\t3\t0.8\t1\t0 3\n",
      "127\t3\t0.8\t1\t0 3\n",
      "128\t5\t0.8\t1\t0 5\n",
      "129\t4\t0.8\t1\t0 4\n",
      "130\t4\t0.8\t1\t0 4\n",
      "131\t1\t0.8\t1\t0 1\n",
      "132\t2\t0.8\t1\t0 2\n",
      "133\t2\t0.8\t1\t0 2\n",
      "134\t3\t0.8\t1\t0 3\n",
      "135\t2\t0.8\t1\t0 2\n",
      "136\t3\t0.8\t1\t0 3\n",
      "137\t2\t0.8\t1\t0 2\n",
      "138\t3\t0.8\t1\t0 3\n",
      "139\t1\t0.8\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479/SRR6399479_2.fastq.gz\n",
      "=============================================\n",
      "57015422 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399479_1_trimmed.fq.gz and SRR6399479_2_trimmed.fq.gz\n",
      "file_1: SRR6399479_1_trimmed.fq.gz, file_2: SRR6399479_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399479_1_trimmed.fq.gz and SRR6399479_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399479_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399479_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 57015422\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 21451 (0.04%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399479_1_trimmed.fq.gz and SRR6399479_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   4%|█▌                                       |  ETA: 16.98 days\u001b[39mspots read      : 43,956,634\n",
      "reads read      : 87,913,268\n",
      "reads written   : 87,913,268\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1860.910712 seconds (57.45 k allocations: 1.641 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/SRR6399480_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t127\tTGGAATTCTCGG\t1000000\t0.01\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Using smallRNA adapter for trimming (count: 127). Second best hit was Illumina (count: 0)\n",
      "\n",
      "Reducing length cutoff to 18bp for small RNA-Seq reads because a cutoff of 20bp may remove some short species of small RNAs if they had been trimmed by 1,2 or 3bp\n",
      "Setting the Illumina smallRNA 5' adapter as adapter 2: 'GATCGTCGGACT'\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/trim_galore/SRR6399480_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/SRR6399480_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'TGGAATTCTCGG' (Illumina small RNA adapter; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Optional adapter 2 sequence (only used for read 2 of paired-end files): 'GATCGTCGGACT'\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 18 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399480_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'TGGAATTCTCGG' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/SRR6399480_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a TGGAATTCTCGG /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/SRR6399480_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 102.896 s (2.341 µs/read; 25.63 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              43,956,634\n",
      "Reads with adapters:                14,188,922 (32.3%)\n",
      "Reads written (passing filters):    43,956,634 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,342,680,855 bp\n",
      "Quality-trimmed:                 181,343 bp (0.0%)\n",
      "Total written (filtered):  6,308,344,520 bp (99.5%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: TGGAATTCTCGG; Type: regular 3'; Length: 12; Trimmed: 14188922 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 17.2%\n",
      "  C: 24.1%\n",
      "  G: 23.6%\n",
      "  T: 35.0%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t10602395\t10989158.5\t0\t10602395\n",
      "2\t2461601\t2747289.6\t0\t2461601\n",
      "3\t813195\t686822.4\t0\t813195\n",
      "4\t114515\t171705.6\t0\t114515\n",
      "5\t39183\t42926.4\t0\t39183\n",
      "6\t20091\t10731.6\t0\t20091\n",
      "7\t4776\t2682.9\t0\t4776\n",
      "8\t1537\t670.7\t0\t1537\n",
      "9\t2892\t167.7\t0\t1361 1531\n",
      "10\t1893\t41.9\t1\t1027 866\n",
      "11\t1276\t10.5\t1\t20 1256\n",
      "12\t60\t2.6\t1\t46 14\n",
      "13\t16\t2.6\t1\t8 8\n",
      "14\t40\t2.6\t1\t34 6\n",
      "15\t18\t2.6\t1\t16 2\n",
      "16\t23\t2.6\t1\t22 1\n",
      "17\t37\t2.6\t1\t30 7\n",
      "18\t26\t2.6\t1\t18 8\n",
      "19\t17\t2.6\t1\t12 5\n",
      "20\t66\t2.6\t1\t61 5\n",
      "21\t14\t2.6\t1\t11 3\n",
      "22\t25\t2.6\t1\t22 3\n",
      "23\t28\t2.6\t1\t23 5\n",
      "24\t20\t2.6\t1\t19 1\n",
      "25\t26\t2.6\t1\t21 5\n",
      "26\t30\t2.6\t1\t29 1\n",
      "27\t40\t2.6\t1\t33 7\n",
      "28\t33\t2.6\t1\t24 9\n",
      "29\t48\t2.6\t1\t36 12\n",
      "30\t42\t2.6\t1\t32 10\n",
      "31\t74\t2.6\t1\t56 18\n",
      "32\t53\t2.6\t1\t42 11\n",
      "33\t40\t2.6\t1\t33 7\n",
      "34\t37\t2.6\t1\t25 12\n",
      "35\t51\t2.6\t1\t43 8\n",
      "36\t62\t2.6\t1\t45 17\n",
      "37\t50\t2.6\t1\t40 10\n",
      "38\t74\t2.6\t1\t55 19\n",
      "39\t52\t2.6\t1\t36 16\n",
      "40\t111\t2.6\t1\t59 52\n",
      "41\t80\t2.6\t1\t47 33\n",
      "42\t80\t2.6\t1\t59 21\n",
      "43\t116\t2.6\t1\t92 24\n",
      "44\t107\t2.6\t1\t90 17\n",
      "45\t88\t2.6\t1\t60 28\n",
      "46\t111\t2.6\t1\t93 18\n",
      "47\t120\t2.6\t1\t83 37\n",
      "48\t94\t2.6\t1\t83 11\n",
      "49\t124\t2.6\t1\t115 9\n",
      "50\t127\t2.6\t1\t117 10\n",
      "51\t203\t2.6\t1\t184 19\n",
      "52\t78\t2.6\t1\t60 18\n",
      "53\t141\t2.6\t1\t125 16\n",
      "54\t242\t2.6\t1\t223 19\n",
      "55\t74\t2.6\t1\t49 25\n",
      "56\t147\t2.6\t1\t138 9\n",
      "57\t150\t2.6\t1\t139 11\n",
      "58\t144\t2.6\t1\t133 11\n",
      "59\t135\t2.6\t1\t101 34\n",
      "60\t131\t2.6\t1\t115 16\n",
      "61\t88\t2.6\t1\t73 15\n",
      "62\t101\t2.6\t1\t95 6\n",
      "63\t125\t2.6\t1\t119 6\n",
      "64\t95\t2.6\t1\t90 5\n",
      "65\t69\t2.6\t1\t58 11\n",
      "66\t67\t2.6\t1\t60 7\n",
      "67\t41\t2.6\t1\t39 2\n",
      "68\t57\t2.6\t1\t50 7\n",
      "69\t50\t2.6\t1\t47 3\n",
      "70\t59\t2.6\t1\t54 5\n",
      "71\t70\t2.6\t1\t58 12\n",
      "72\t68\t2.6\t1\t54 14\n",
      "73\t62\t2.6\t1\t53 9\n",
      "74\t84\t2.6\t1\t75 9\n",
      "75\t54\t2.6\t1\t48 6\n",
      "76\t58\t2.6\t1\t53 5\n",
      "77\t36\t2.6\t1\t28 8\n",
      "78\t55\t2.6\t1\t52 3\n",
      "79\t52\t2.6\t1\t47 5\n",
      "80\t45\t2.6\t1\t38 7\n",
      "81\t61\t2.6\t1\t51 10\n",
      "82\t82\t2.6\t1\t77 5\n",
      "83\t80\t2.6\t1\t76 4\n",
      "84\t80\t2.6\t1\t65 15\n",
      "85\t59\t2.6\t1\t50 9\n",
      "86\t58\t2.6\t1\t55 3\n",
      "87\t161\t2.6\t1\t153 8\n",
      "88\t171\t2.6\t1\t156 15\n",
      "89\t212\t2.6\t1\t202 10\n",
      "90\t155\t2.6\t1\t146 9\n",
      "91\t199\t2.6\t1\t185 14\n",
      "92\t137\t2.6\t1\t130 7\n",
      "93\t178\t2.6\t1\t171 7\n",
      "94\t128\t2.6\t1\t117 11\n",
      "95\t130\t2.6\t1\t116 14\n",
      "96\t175\t2.6\t1\t162 13\n",
      "97\t237\t2.6\t1\t213 24\n",
      "98\t386\t2.6\t1\t362 24\n",
      "99\t146\t2.6\t1\t129 17\n",
      "100\t126\t2.6\t1\t102 24\n",
      "101\t97\t2.6\t1\t90 7\n",
      "102\t106\t2.6\t1\t101 5\n",
      "103\t169\t2.6\t1\t149 20\n",
      "104\t154\t2.6\t1\t125 29\n",
      "105\t209\t2.6\t1\t186 23\n",
      "106\t107\t2.6\t1\t65 42\n",
      "107\t114\t2.6\t1\t99 15\n",
      "108\t58\t2.6\t1\t44 14\n",
      "109\t67\t2.6\t1\t54 13\n",
      "110\t67\t2.6\t1\t50 17\n",
      "111\t55\t2.6\t1\t37 18\n",
      "112\t68\t2.6\t1\t54 14\n",
      "113\t86\t2.6\t1\t55 31\n",
      "114\t92\t2.6\t1\t77 15\n",
      "115\t167\t2.6\t1\t138 29\n",
      "116\t258\t2.6\t1\t239 19\n",
      "117\t221\t2.6\t1\t212 9\n",
      "118\t1202\t2.6\t1\t1130 72\n",
      "119\t5086\t2.6\t1\t4817 269\n",
      "120\t10419\t2.6\t1\t9809 610\n",
      "121\t4956\t2.6\t1\t4677 279\n",
      "122\t7308\t2.6\t1\t6820 488\n",
      "123\t16159\t2.6\t1\t14810 1349\n",
      "124\t6363\t2.6\t1\t5747 616\n",
      "125\t5041\t2.6\t1\t4540 501\n",
      "126\t4802\t2.6\t1\t4366 436\n",
      "127\t9625\t2.6\t1\t8769 856\n",
      "128\t11738\t2.6\t1\t10671 1067\n",
      "129\t18822\t2.6\t1\t17012 1810\n",
      "130\t1547\t2.6\t1\t1407 140\n",
      "131\t603\t2.6\t1\t533 70\n",
      "132\t1793\t2.6\t1\t1629 164\n",
      "133\t2594\t2.6\t1\t2338 256\n",
      "134\t379\t2.6\t1\t347 32\n",
      "135\t1745\t2.6\t1\t1595 150\n",
      "136\t246\t2.6\t1\t229 17\n",
      "137\t898\t2.6\t1\t856 42\n",
      "138\t263\t2.6\t1\t255 8\n",
      "139\t1267\t2.6\t1\t1206 61\n",
      "140\t367\t2.6\t1\t351 16\n",
      "141\t213\t2.6\t1\t197 16\n",
      "142\t30\t2.6\t1\t24 6\n",
      "143\t55\t2.6\t1\t51 4\n",
      "144\t22\t2.6\t1\t18 4\n",
      "145\t488\t2.6\t1\t465 23\n",
      "146\t176\t2.6\t1\t172 4\n",
      "147\t82\t2.6\t1\t74 8\n",
      "148\t126\t2.6\t1\t118 8\n",
      "149\t175\t2.6\t1\t167 8\n",
      "150\t376\t2.6\t1\t360 16\n",
      "151\t505\t2.6\t1\t262 243\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/SRR6399480_1.fastq.gz\n",
      "=============================================\n",
      "43956634 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/trim_galore/SRR6399480_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/SRR6399480_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'TGGAATTCTCGG' (Illumina small RNA adapter; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Optional adapter 2 sequence (only used for read 2 of paired-end files): 'GATCGTCGGACT'\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 18 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399480_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'GATCGTCGGACT' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/SRR6399480_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a GATCGTCGGACT /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/SRR6399480_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 109.394 s (2.489 µs/read; 24.11 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              43,956,634\n",
      "Reads with adapters:                13,313,338 (30.3%)\n",
      "Reads written (passing filters):    43,956,634 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,132,985,113 bp\n",
      "Quality-trimmed:                 423,191 bp (0.0%)\n",
      "Total written (filtered):  6,115,342,321 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: GATCGTCGGACT; Type: regular 3'; Length: 12; Trimmed: 13313338 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 29.8%\n",
      "  C: 17.5%\n",
      "  G: 30.4%\n",
      "  T: 22.3%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t10832619\t10989158.5\t0\t10832619\n",
      "2\t1895824\t2747289.6\t0\t1895824\n",
      "3\t406722\t686822.4\t0\t406722\n",
      "4\t107240\t171705.6\t0\t107240\n",
      "5\t59265\t42926.4\t0\t59265\n",
      "6\t2860\t10731.6\t0\t2860\n",
      "7\t1777\t2682.9\t0\t1777\n",
      "8\t230\t670.7\t0\t230\n",
      "9\t332\t167.7\t0\t29 303\n",
      "10\t246\t41.9\t1\t1 245\n",
      "11\t40\t10.5\t1\t0 40\n",
      "12\t30\t2.6\t1\t1 29\n",
      "13\t16\t2.6\t1\t0 16\n",
      "14\t24\t2.6\t1\t0 24\n",
      "15\t12\t2.6\t1\t0 12\n",
      "16\t26\t2.6\t1\t0 26\n",
      "17\t22\t2.6\t1\t0 22\n",
      "18\t35\t2.6\t1\t0 35\n",
      "19\t10\t2.6\t1\t0 10\n",
      "20\t18\t2.6\t1\t0 18\n",
      "21\t12\t2.6\t1\t0 12\n",
      "22\t14\t2.6\t1\t0 14\n",
      "23\t33\t2.6\t1\t0 33\n",
      "24\t19\t2.6\t1\t0 19\n",
      "25\t28\t2.6\t1\t0 28\n",
      "26\t30\t2.6\t1\t0 30\n",
      "27\t21\t2.6\t1\t0 21\n",
      "28\t34\t2.6\t1\t0 34\n",
      "29\t19\t2.6\t1\t0 19\n",
      "30\t13\t2.6\t1\t0 13\n",
      "31\t18\t2.6\t1\t0 18\n",
      "32\t21\t2.6\t1\t1 20\n",
      "33\t26\t2.6\t1\t0 26\n",
      "34\t28\t2.6\t1\t0 28\n",
      "35\t14\t2.6\t1\t0 14\n",
      "36\t24\t2.6\t1\t0 24\n",
      "37\t16\t2.6\t1\t0 16\n",
      "38\t22\t2.6\t1\t0 22\n",
      "39\t26\t2.6\t1\t0 26\n",
      "40\t33\t2.6\t1\t0 33\n",
      "41\t39\t2.6\t1\t0 39\n",
      "42\t20\t2.6\t1\t0 20\n",
      "43\t25\t2.6\t1\t0 25\n",
      "44\t25\t2.6\t1\t0 25\n",
      "45\t26\t2.6\t1\t0 26\n",
      "46\t22\t2.6\t1\t0 22\n",
      "47\t29\t2.6\t1\t0 29\n",
      "48\t12\t2.6\t1\t0 12\n",
      "49\t21\t2.6\t1\t0 21\n",
      "50\t11\t2.6\t1\t0 11\n",
      "51\t26\t2.6\t1\t0 26\n",
      "52\t20\t2.6\t1\t0 20\n",
      "53\t15\t2.6\t1\t0 15\n",
      "54\t32\t2.6\t1\t0 32\n",
      "55\t24\t2.6\t1\t0 24\n",
      "56\t12\t2.6\t1\t0 12\n",
      "57\t11\t2.6\t1\t0 11\n",
      "58\t20\t2.6\t1\t0 20\n",
      "59\t14\t2.6\t1\t0 14\n",
      "60\t12\t2.6\t1\t0 12\n",
      "61\t14\t2.6\t1\t0 14\n",
      "62\t54\t2.6\t1\t0 54\n",
      "63\t35\t2.6\t1\t0 35\n",
      "64\t48\t2.6\t1\t0 48\n",
      "65\t24\t2.6\t1\t0 24\n",
      "66\t25\t2.6\t1\t0 25\n",
      "67\t22\t2.6\t1\t0 22\n",
      "68\t31\t2.6\t1\t0 31\n",
      "69\t27\t2.6\t1\t0 27\n",
      "70\t30\t2.6\t1\t0 30\n",
      "71\t34\t2.6\t1\t0 34\n",
      "72\t19\t2.6\t1\t0 19\n",
      "73\t23\t2.6\t1\t0 23\n",
      "74\t14\t2.6\t1\t0 14\n",
      "75\t14\t2.6\t1\t1 13\n",
      "76\t16\t2.6\t1\t0 16\n",
      "77\t22\t2.6\t1\t0 22\n",
      "78\t13\t2.6\t1\t0 13\n",
      "79\t29\t2.6\t1\t0 29\n",
      "80\t22\t2.6\t1\t0 22\n",
      "81\t48\t2.6\t1\t0 48\n",
      "82\t17\t2.6\t1\t0 17\n",
      "83\t33\t2.6\t1\t0 33\n",
      "84\t11\t2.6\t1\t0 11\n",
      "85\t17\t2.6\t1\t0 17\n",
      "86\t13\t2.6\t1\t0 13\n",
      "87\t13\t2.6\t1\t0 13\n",
      "88\t15\t2.6\t1\t0 15\n",
      "89\t19\t2.6\t1\t0 19\n",
      "90\t6\t2.6\t1\t0 6\n",
      "91\t15\t2.6\t1\t0 15\n",
      "92\t13\t2.6\t1\t0 13\n",
      "93\t17\t2.6\t1\t0 17\n",
      "94\t13\t2.6\t1\t0 13\n",
      "95\t19\t2.6\t1\t0 19\n",
      "96\t10\t2.6\t1\t0 10\n",
      "97\t19\t2.6\t1\t0 19\n",
      "98\t18\t2.6\t1\t0 18\n",
      "99\t13\t2.6\t1\t0 13\n",
      "100\t23\t2.6\t1\t0 23\n",
      "101\t14\t2.6\t1\t0 14\n",
      "102\t14\t2.6\t1\t0 14\n",
      "103\t16\t2.6\t1\t0 16\n",
      "104\t28\t2.6\t1\t0 28\n",
      "105\t14\t2.6\t1\t0 14\n",
      "106\t29\t2.6\t1\t0 29\n",
      "107\t32\t2.6\t1\t0 32\n",
      "108\t24\t2.6\t1\t0 24\n",
      "109\t12\t2.6\t1\t0 12\n",
      "110\t16\t2.6\t1\t0 16\n",
      "111\t69\t2.6\t1\t0 69\n",
      "112\t64\t2.6\t1\t0 64\n",
      "113\t181\t2.6\t1\t0 181\n",
      "114\t551\t2.6\t1\t0 551\n",
      "115\t314\t2.6\t1\t0 314\n",
      "116\t250\t2.6\t1\t0 250\n",
      "117\t362\t2.6\t1\t0 362\n",
      "118\t146\t2.6\t1\t0 146\n",
      "119\t258\t2.6\t1\t0 258\n",
      "120\t352\t2.6\t1\t0 352\n",
      "121\t204\t2.6\t1\t0 204\n",
      "122\t195\t2.6\t1\t0 195\n",
      "123\t629\t2.6\t1\t0 629\n",
      "124\t31\t2.6\t1\t0 31\n",
      "125\t31\t2.6\t1\t0 31\n",
      "126\t24\t2.6\t1\t0 24\n",
      "127\t59\t2.6\t1\t0 59\n",
      "128\t17\t2.6\t1\t0 17\n",
      "129\t15\t2.6\t1\t0 15\n",
      "130\t12\t2.6\t1\t0 12\n",
      "131\t13\t2.6\t1\t0 13\n",
      "132\t2\t2.6\t1\t0 2\n",
      "133\t7\t2.6\t1\t0 7\n",
      "134\t9\t2.6\t1\t0 9\n",
      "135\t17\t2.6\t1\t0 17\n",
      "136\t15\t2.6\t1\t0 15\n",
      "137\t10\t2.6\t1\t0 10\n",
      "138\t24\t2.6\t1\t0 24\n",
      "139\t14\t2.6\t1\t0 14\n",
      "140\t17\t2.6\t1\t0 17\n",
      "141\t19\t2.6\t1\t0 19\n",
      "142\t18\t2.6\t1\t0 18\n",
      "143\t7\t2.6\t1\t0 7\n",
      "144\t10\t2.6\t1\t0 10\n",
      "145\t26\t2.6\t1\t0 26\n",
      "146\t13\t2.6\t1\t0 13\n",
      "147\t37\t2.6\t1\t0 37\n",
      "148\t1\t2.6\t1\t0 1\n",
      "149\t10\t2.6\t1\t0 10\n",
      "150\t13\t2.6\t1\t0 13\n",
      "151\t10\t2.6\t1\t0 10\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480/SRR6399480_2.fastq.gz\n",
      "=============================================\n",
      "43956634 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399480_1_trimmed.fq.gz and SRR6399480_2_trimmed.fq.gz\n",
      "file_1: SRR6399480_1_trimmed.fq.gz, file_2: SRR6399480_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399480_1_trimmed.fq.gz and SRR6399480_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399480_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399480_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 43956634\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (18 bp): 8683 (0.02%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399480_1_trimmed.fq.gz and SRR6399480_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   4%|█▌                                       |  ETA: 16.99 days\u001b[39mspots read      : 42,170,873\n",
      "reads read      : 84,341,746\n",
      "reads written   : 84,341,746\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1867.190175 seconds (57.65 k allocations: 1.647 MiB)\n",
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/SRR6399481_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count smallRNA: 0, count Illumina: 0, count Nextera: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/trim_galore/SRR6399481_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/SRR6399481_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399481_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/SRR6399481_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/SRR6399481_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 104.158 s (2.470 µs/read; 24.29 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              42,170,873\n",
      "Reads with adapters:                13,057,211 (31.0%)\n",
      "Reads written (passing filters):    42,170,873 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,290,115,352 bp\n",
      "Quality-trimmed:                  70,014 bp (0.0%)\n",
      "Total written (filtered):  6,272,162,765 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 13057211 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 29.3%\n",
      "  C: 25.9%\n",
      "  G: 20.2%\n",
      "  T: 24.6%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t9355161\t10542718.2\t0\t9355161\n",
      "2\t2916390\t2635679.6\t0\t2916390\n",
      "3\t577585\t658919.9\t0\t577585\n",
      "4\t135848\t164730.0\t0\t135848\n",
      "5\t35705\t41182.5\t0\t35705\n",
      "6\t18818\t10295.6\t0\t18818\n",
      "7\t17606\t2573.9\t0\t17606\n",
      "8\t3\t643.5\t0\t3\n",
      "9\t34\t160.9\t0\t1 33\n",
      "10\t23\t40.2\t1\t1 22\n",
      "11\t2\t10.1\t1\t0 2\n",
      "26\t1\t0.6\t1\t0 1\n",
      "30\t1\t0.6\t1\t0 1\n",
      "37\t1\t0.6\t1\t0 1\n",
      "55\t1\t0.6\t1\t0 1\n",
      "58\t1\t0.6\t1\t0 1\n",
      "59\t2\t0.6\t1\t0 2\n",
      "62\t1\t0.6\t1\t0 1\n",
      "64\t1\t0.6\t1\t0 1\n",
      "69\t1\t0.6\t1\t0 1\n",
      "72\t1\t0.6\t1\t0 1\n",
      "73\t1\t0.6\t1\t0 1\n",
      "75\t1\t0.6\t1\t0 1\n",
      "76\t2\t0.6\t1\t0 2\n",
      "77\t1\t0.6\t1\t0 1\n",
      "80\t1\t0.6\t1\t0 1\n",
      "83\t1\t0.6\t1\t0 1\n",
      "88\t1\t0.6\t1\t0 1\n",
      "89\t1\t0.6\t1\t0 1\n",
      "101\t2\t0.6\t1\t0 2\n",
      "102\t1\t0.6\t1\t0 1\n",
      "106\t1\t0.6\t1\t0 1\n",
      "111\t1\t0.6\t1\t0 1\n",
      "118\t1\t0.6\t1\t0 1\n",
      "119\t1\t0.6\t1\t0 1\n",
      "121\t1\t0.6\t1\t0 1\n",
      "123\t1\t0.6\t1\t0 1\n",
      "126\t1\t0.6\t1\t0 1\n",
      "128\t1\t0.6\t1\t0 1\n",
      "129\t1\t0.6\t1\t0 1\n",
      "130\t1\t0.6\t1\t0 1\n",
      "133\t1\t0.6\t1\t0 1\n",
      "136\t2\t0.6\t1\t0 2\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/SRR6399481_1.fastq.gz\n",
      "=============================================\n",
      "42170873 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/trim_galore/SRR6399481_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/SRR6399481_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399481_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/SRR6399481_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/SRR6399481_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 114.122 s (2.706 µs/read; 22.17 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              42,170,873\n",
      "Reads with adapters:                10,881,535 (25.8%)\n",
      "Reads written (passing filters):    42,170,873 (100.0%)\n",
      "\n",
      "Total basepairs processed: 6,127,361,975 bp\n",
      "Quality-trimmed:                 207,005 bp (0.0%)\n",
      "Total written (filtered):  6,111,785,815 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 10881535 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 33.9%\n",
      "  C: 24.8%\n",
      "  G: 17.5%\n",
      "  T: 23.8%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t7421691\t10542718.2\t0\t7421691\n",
      "2\t2815256\t2635679.6\t0\t2815256\n",
      "3\t420149\t658919.9\t0\t420149\n",
      "4\t141073\t164730.0\t0\t141073\n",
      "5\t36644\t41182.5\t0\t36644\n",
      "6\t21289\t10295.6\t0\t21289\n",
      "7\t25328\t2573.9\t0\t25328\n",
      "8\t4\t643.5\t0\t4\n",
      "9\t31\t160.9\t0\t2 29\n",
      "10\t27\t40.2\t1\t0 27\n",
      "11\t2\t10.1\t1\t0 2\n",
      "12\t5\t2.5\t1\t0 5\n",
      "31\t1\t0.6\t1\t0 1\n",
      "35\t1\t0.6\t1\t0 1\n",
      "58\t1\t0.6\t1\t0 1\n",
      "66\t2\t0.6\t1\t0 2\n",
      "67\t2\t0.6\t1\t0 2\n",
      "68\t1\t0.6\t1\t0 1\n",
      "69\t1\t0.6\t1\t0 1\n",
      "70\t1\t0.6\t1\t0 1\n",
      "71\t1\t0.6\t1\t0 1\n",
      "72\t2\t0.6\t1\t0 2\n",
      "74\t1\t0.6\t1\t0 1\n",
      "76\t1\t0.6\t1\t0 1\n",
      "77\t1\t0.6\t1\t0 1\n",
      "79\t1\t0.6\t1\t0 1\n",
      "84\t1\t0.6\t1\t0 1\n",
      "89\t1\t0.6\t1\t0 1\n",
      "101\t2\t0.6\t1\t0 2\n",
      "102\t1\t0.6\t1\t0 1\n",
      "104\t2\t0.6\t1\t0 2\n",
      "107\t1\t0.6\t1\t0 1\n",
      "111\t1\t0.6\t1\t0 1\n",
      "118\t1\t0.6\t1\t0 1\n",
      "119\t1\t0.6\t1\t0 1\n",
      "123\t1\t0.6\t1\t0 1\n",
      "126\t2\t0.6\t1\t0 2\n",
      "130\t1\t0.6\t1\t0 1\n",
      "131\t1\t0.6\t1\t0 1\n",
      "133\t1\t0.6\t1\t0 1\n",
      "136\t1\t0.6\t1\t0 1\n",
      "137\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481/SRR6399481_2.fastq.gz\n",
      "=============================================\n",
      "42170873 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399481_1_trimmed.fq.gz and SRR6399481_2_trimmed.fq.gz\n",
      "file_1: SRR6399481_1_trimmed.fq.gz, file_2: SRR6399481_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399481_1_trimmed.fq.gz and SRR6399481_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399481_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399481_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 42170873\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 2999 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399481_1_trimmed.fq.gz and SRR6399481_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   4%|█▋                                       |  ETA: 17.00 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1603.904919 seconds (49.44 k allocations: 1.407 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 38,130,900\n",
      "reads read      : 76,261,800\n",
      "reads written   : 76,261,800\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/SRR6399482_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count smallRNA: 0, count Illumina: 0, count Nextera: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/trim_galore/SRR6399482_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/SRR6399482_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399482_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/SRR6399482_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/SRR6399482_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 95.924 s (2.516 µs/read; 23.85 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              38,130,900\n",
      "Reads with adapters:                12,039,886 (31.6%)\n",
      "Reads written (passing filters):    38,130,900 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,694,526,044 bp\n",
      "Quality-trimmed:                  64,017 bp (0.0%)\n",
      "Total written (filtered):  5,677,997,966 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 12039886 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 30.4%\n",
      "  C: 24.8%\n",
      "  G: 20.2%\n",
      "  T: 24.4%\n",
      "  none/other: 0.2%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t8725922\t9532725.0\t0\t8725922\n",
      "2\t2531670\t2383181.2\t0\t2531670\n",
      "3\t583498\t595795.3\t0\t583498\n",
      "4\t125697\t148948.8\t0\t125697\n",
      "5\t36520\t37237.2\t0\t36520\n",
      "6\t19253\t9309.3\t0\t19253\n",
      "7\t17271\t2327.3\t0\t17271\n",
      "9\t17\t145.5\t0\t0 17\n",
      "10\t7\t36.4\t1\t0 7\n",
      "11\t4\t9.1\t1\t0 4\n",
      "12\t1\t2.3\t1\t0 1\n",
      "24\t1\t0.6\t1\t0 1\n",
      "32\t3\t0.6\t1\t0 3\n",
      "41\t2\t0.6\t1\t0 2\n",
      "50\t1\t0.6\t1\t0 1\n",
      "56\t1\t0.6\t1\t0 1\n",
      "63\t3\t0.6\t1\t0 3\n",
      "74\t1\t0.6\t1\t0 1\n",
      "95\t1\t0.6\t1\t0 1\n",
      "100\t1\t0.6\t1\t0 1\n",
      "104\t2\t0.6\t1\t0 2\n",
      "111\t1\t0.6\t1\t0 1\n",
      "114\t1\t0.6\t1\t0 1\n",
      "118\t1\t0.6\t1\t0 1\n",
      "124\t1\t0.6\t1\t0 1\n",
      "127\t2\t0.6\t1\t0 2\n",
      "130\t1\t0.6\t1\t0 1\n",
      "132\t2\t0.6\t1\t0 2\n",
      "134\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/SRR6399482_1.fastq.gz\n",
      "=============================================\n",
      "38130900 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/trim_galore/SRR6399482_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/SRR6399482_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399482_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/SRR6399482_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/SRR6399482_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 102.165 s (2.679 µs/read; 22.39 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              38,130,900\n",
      "Reads with adapters:                11,965,630 (31.4%)\n",
      "Reads written (passing filters):    38,130,900 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,592,341,410 bp\n",
      "Quality-trimmed:                 153,570 bp (0.0%)\n",
      "Total written (filtered):  5,575,843,999 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 11965630 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 34.1%\n",
      "  C: 23.6%\n",
      "  G: 18.5%\n",
      "  T: 23.8%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t8722007\t9532725.0\t0\t8722007\n",
      "2\t2463992\t2383181.2\t0\t2463992\n",
      "3\t570060\t595795.3\t0\t570060\n",
      "4\t128631\t148948.8\t0\t128631\n",
      "5\t38547\t37237.2\t0\t38547\n",
      "6\t22275\t9309.3\t0\t22275\n",
      "7\t20063\t2327.3\t0\t20063\n",
      "9\t20\t145.5\t0\t0 20\n",
      "10\t3\t36.4\t1\t0 3\n",
      "11\t1\t9.1\t1\t0 1\n",
      "12\t3\t2.3\t1\t0 3\n",
      "17\t1\t0.6\t1\t0 1\n",
      "21\t1\t0.6\t1\t0 1\n",
      "22\t1\t0.6\t1\t0 1\n",
      "34\t1\t0.6\t1\t0 1\n",
      "36\t1\t0.6\t1\t0 1\n",
      "37\t2\t0.6\t1\t0 2\n",
      "38\t2\t0.6\t1\t0 2\n",
      "45\t1\t0.6\t1\t0 1\n",
      "47\t1\t0.6\t1\t0 1\n",
      "67\t1\t0.6\t1\t0 1\n",
      "71\t1\t0.6\t1\t0 1\n",
      "74\t3\t0.6\t1\t0 3\n",
      "76\t1\t0.6\t1\t0 1\n",
      "95\t1\t0.6\t1\t0 1\n",
      "100\t1\t0.6\t1\t0 1\n",
      "104\t1\t0.6\t1\t0 1\n",
      "111\t2\t0.6\t1\t0 2\n",
      "113\t1\t0.6\t1\t0 1\n",
      "117\t1\t0.6\t1\t0 1\n",
      "122\t1\t0.6\t1\t0 1\n",
      "124\t1\t0.6\t1\t0 1\n",
      "127\t1\t0.6\t1\t0 1\n",
      "131\t1\t0.6\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482/SRR6399482_2.fastq.gz\n",
      "=============================================\n",
      "38130900 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399482_1_trimmed.fq.gz and SRR6399482_2_trimmed.fq.gz\n",
      "file_1: SRR6399482_1_trimmed.fq.gz, file_2: SRR6399482_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399482_1_trimmed.fq.gz and SRR6399482_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399482_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399482_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 38130900\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 3643 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399482_1_trimmed.fq.gz and SRR6399482_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   4%|█▋                                       |  ETA: 16.91 days\u001b[39mspots read      : 52,184,124\n",
      "reads read      : 104,368,248\n",
      "reads written   : 104,368,248\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2198.561313 seconds (67.85 k allocations: 1.937 MiB)\n",
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/SRR6399483_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Unable to auto-detect most prominent adapter from the first specified file (count Nextera: 0, count Illumina: 0, count smallRNA: 0)\n",
      "Defaulting to Illumina universal adapter ( AGATCGGAAGAGC ). Specify -a SEQUENCE to avoid this behavior).\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/trim_galore/SRR6399483_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/SRR6399483_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399483_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/SRR6399483_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/SRR6399483_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 127.911 s (2.451 µs/read; 24.48 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              52,184,124\n",
      "Reads with adapters:                16,045,024 (30.7%)\n",
      "Reads written (passing filters):    52,184,124 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,791,682,780 bp\n",
      "Quality-trimmed:                  41,373 bp (0.0%)\n",
      "Total written (filtered):  7,769,756,254 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 16045024 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 30.3%\n",
      "  C: 24.6%\n",
      "  G: 21.4%\n",
      "  T: 23.7%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t11580972\t13046031.0\t0\t11580972\n",
      "2\t3441136\t3261507.8\t0\t3441136\n",
      "3\t794949\t815376.9\t0\t794949\n",
      "4\t158228\t203844.2\t0\t158228\n",
      "5\t35211\t50961.1\t0\t35211\n",
      "6\t18663\t12740.3\t0\t18663\n",
      "7\t15779\t3185.1\t0\t15779\n",
      "8\t3\t796.3\t0\t3\n",
      "9\t17\t199.1\t0\t1 16\n",
      "10\t5\t49.8\t1\t0 5\n",
      "11\t3\t12.4\t1\t0 3\n",
      "12\t1\t3.1\t1\t0 1\n",
      "28\t2\t0.8\t1\t0 2\n",
      "30\t2\t0.8\t1\t0 2\n",
      "32\t1\t0.8\t1\t0 1\n",
      "33\t1\t0.8\t1\t0 1\n",
      "36\t1\t0.8\t1\t0 1\n",
      "62\t4\t0.8\t1\t0 4\n",
      "73\t1\t0.8\t1\t0 1\n",
      "74\t3\t0.8\t1\t0 3\n",
      "75\t2\t0.8\t1\t0 2\n",
      "76\t2\t0.8\t1\t0 2\n",
      "86\t1\t0.8\t1\t0 1\n",
      "87\t1\t0.8\t1\t0 1\n",
      "90\t1\t0.8\t1\t0 1\n",
      "91\t1\t0.8\t1\t0 1\n",
      "93\t1\t0.8\t1\t0 1\n",
      "95\t1\t0.8\t1\t0 1\n",
      "99\t1\t0.8\t1\t0 1\n",
      "107\t3\t0.8\t1\t0 3\n",
      "109\t1\t0.8\t1\t0 1\n",
      "110\t3\t0.8\t1\t0 3\n",
      "111\t3\t0.8\t1\t0 3\n",
      "112\t1\t0.8\t1\t0 1\n",
      "113\t1\t0.8\t1\t0 1\n",
      "115\t3\t0.8\t1\t0 3\n",
      "119\t1\t0.8\t1\t0 1\n",
      "120\t3\t0.8\t1\t0 3\n",
      "121\t2\t0.8\t1\t0 2\n",
      "125\t1\t0.8\t1\t0 1\n",
      "126\t1\t0.8\t1\t0 1\n",
      "127\t2\t0.8\t1\t0 2\n",
      "129\t1\t0.8\t1\t0 1\n",
      "130\t1\t0.8\t1\t0 1\n",
      "131\t1\t0.8\t1\t0 1\n",
      "135\t1\t0.8\t1\t0 1\n",
      "137\t1\t0.8\t1\t0 1\n",
      "138\t1\t0.8\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/SRR6399483_1.fastq.gz\n",
      "=============================================\n",
      "52184124 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/trim_galore/SRR6399483_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/SRR6399483_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'AGATCGGAAGAGC' (Illumina TruSeq, Sanger iPCR; default (inconclusive auto-detection))\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399483_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'AGATCGGAAGAGC' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/SRR6399483_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a AGATCGGAAGAGC /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/SRR6399483_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 141.504 s (2.712 µs/read; 22.13 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              52,184,124\n",
      "Reads with adapters:                15,777,982 (30.2%)\n",
      "Reads written (passing filters):    52,184,124 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,602,173,740 bp\n",
      "Quality-trimmed:                 117,513 bp (0.0%)\n",
      "Total written (filtered):  7,580,677,511 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 15777982 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 36.1%\n",
      "  C: 22.6%\n",
      "  G: 18.6%\n",
      "  T: 22.7%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t11519652\t13046031.0\t0\t11519652\n",
      "2\t3284041\t3261507.8\t0\t3284041\n",
      "3\t745495\t815376.9\t0\t745495\n",
      "4\t150083\t203844.2\t0\t150083\n",
      "5\t40197\t50961.1\t0\t40197\n",
      "6\t20041\t12740.3\t0\t20041\n",
      "7\t18392\t3185.1\t0\t18392\n",
      "8\t2\t796.3\t0\t2\n",
      "9\t29\t199.1\t0\t2 27\n",
      "10\t5\t49.8\t1\t0 5\n",
      "11\t4\t12.4\t1\t0 4\n",
      "12\t2\t3.1\t1\t0 2\n",
      "20\t1\t0.8\t1\t0 1\n",
      "23\t1\t0.8\t1\t0 1\n",
      "40\t1\t0.8\t1\t0 1\n",
      "57\t1\t0.8\t1\t0 1\n",
      "61\t1\t0.8\t1\t0 1\n",
      "62\t1\t0.8\t1\t0 1\n",
      "73\t1\t0.8\t1\t0 1\n",
      "74\t2\t0.8\t1\t0 2\n",
      "75\t1\t0.8\t1\t0 1\n",
      "78\t1\t0.8\t1\t0 1\n",
      "80\t1\t0.8\t1\t0 1\n",
      "86\t1\t0.8\t1\t0 1\n",
      "87\t1\t0.8\t1\t0 1\n",
      "89\t1\t0.8\t1\t0 1\n",
      "97\t1\t0.8\t1\t0 1\n",
      "107\t3\t0.8\t1\t0 3\n",
      "109\t3\t0.8\t1\t0 3\n",
      "110\t1\t0.8\t1\t0 1\n",
      "111\t1\t0.8\t1\t0 1\n",
      "113\t2\t0.8\t1\t0 2\n",
      "115\t2\t0.8\t1\t0 2\n",
      "118\t1\t0.8\t1\t0 1\n",
      "120\t1\t0.8\t1\t0 1\n",
      "121\t2\t0.8\t1\t0 2\n",
      "126\t1\t0.8\t1\t0 1\n",
      "127\t2\t0.8\t1\t0 2\n",
      "130\t1\t0.8\t1\t0 1\n",
      "131\t1\t0.8\t1\t0 1\n",
      "135\t1\t0.8\t1\t0 1\n",
      "138\t1\t0.8\t1\t0 1\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483/SRR6399483_2.fastq.gz\n",
      "=============================================\n",
      "52184124 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399483_1_trimmed.fq.gz and SRR6399483_2_trimmed.fq.gz\n",
      "file_1: SRR6399483_1_trimmed.fq.gz, file_2: SRR6399483_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399483_1_trimmed.fq.gz and SRR6399483_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399483_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399483_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 52184124\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 6299 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399483_1_trimmed.fq.gz and SRR6399483_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n",
      "\u001b[32mProgress:   4%|█▊                                       |  ETA: 17.05 days\u001b[39mspots read      : 37,169,462\n",
      "reads read      : 74,338,924\n",
      "reads written   : 74,338,924\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1568.163173 seconds (48.44 k allocations: 1.385 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/SRR6399484_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t1\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Using smallRNA adapter for trimming (count: 1). Second best hit was Nextera (count: 0)\n",
      "\n",
      "Reducing length cutoff to 18bp for small RNA-Seq reads because a cutoff of 20bp may remove some short species of small RNAs if they had been trimmed by 1,2 or 3bp\n",
      "Setting the Illumina smallRNA 5' adapter as adapter 2: 'GATCGTCGGACT'\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/trim_galore/SRR6399484_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/SRR6399484_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'TGGAATTCTCGG' (Illumina small RNA adapter; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Optional adapter 2 sequence (only used for read 2 of paired-end files): 'GATCGTCGGACT'\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 18 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399484_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'TGGAATTCTCGG' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/SRR6399484_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a TGGAATTCTCGG /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/SRR6399484_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 87.192 s (2.346 µs/read; 25.58 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              37,169,462\n",
      "Reads with adapters:                12,964,900 (34.9%)\n",
      "Reads written (passing filters):    37,169,462 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,552,690,424 bp\n",
      "Quality-trimmed:                  23,816 bp (0.0%)\n",
      "Total written (filtered):  5,534,265,012 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: TGGAATTCTCGG; Type: regular 3'; Length: 12; Trimmed: 12964900 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 18.1%\n",
      "  C: 24.1%\n",
      "  G: 17.9%\n",
      "  T: 39.8%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t10076665\t9292365.5\t0\t10076665\n",
      "2\t2002993\t2323091.4\t0\t2002993\n",
      "3\t681065\t580772.8\t0\t681065\n",
      "4\t119249\t145193.2\t0\t119249\n",
      "5\t48322\t36298.3\t0\t48322\n",
      "6\t15359\t9074.6\t0\t15359\n",
      "7\t3102\t2268.6\t0\t3102\n",
      "8\t2225\t567.2\t0\t2225\n",
      "9\t1158\t141.8\t0\t124 1034\n",
      "10\t739\t35.4\t1\t0 739\n",
      "11\t751\t8.9\t1\t0 751\n",
      "12\t441\t2.2\t1\t1 440\n",
      "13\t128\t2.2\t1\t0 128\n",
      "14\t63\t2.2\t1\t0 63\n",
      "15\t48\t2.2\t1\t0 48\n",
      "16\t33\t2.2\t1\t0 33\n",
      "17\t48\t2.2\t1\t0 48\n",
      "18\t32\t2.2\t1\t1 31\n",
      "19\t35\t2.2\t1\t0 35\n",
      "20\t19\t2.2\t1\t0 19\n",
      "21\t17\t2.2\t1\t0 17\n",
      "22\t28\t2.2\t1\t0 28\n",
      "23\t15\t2.2\t1\t0 15\n",
      "24\t23\t2.2\t1\t0 23\n",
      "25\t15\t2.2\t1\t0 15\n",
      "26\t20\t2.2\t1\t1 19\n",
      "27\t15\t2.2\t1\t0 15\n",
      "28\t17\t2.2\t1\t0 17\n",
      "29\t11\t2.2\t1\t0 11\n",
      "30\t12\t2.2\t1\t0 12\n",
      "31\t13\t2.2\t1\t1 12\n",
      "32\t34\t2.2\t1\t3 31\n",
      "33\t18\t2.2\t1\t1 17\n",
      "34\t11\t2.2\t1\t0 11\n",
      "35\t11\t2.2\t1\t0 11\n",
      "36\t15\t2.2\t1\t0 15\n",
      "37\t9\t2.2\t1\t0 9\n",
      "38\t2\t2.2\t1\t0 2\n",
      "39\t5\t2.2\t1\t0 5\n",
      "40\t25\t2.2\t1\t0 25\n",
      "41\t18\t2.2\t1\t0 18\n",
      "42\t17\t2.2\t1\t0 17\n",
      "43\t21\t2.2\t1\t0 21\n",
      "44\t11\t2.2\t1\t0 11\n",
      "45\t30\t2.2\t1\t0 30\n",
      "46\t19\t2.2\t1\t0 19\n",
      "47\t49\t2.2\t1\t1 48\n",
      "48\t133\t2.2\t1\t0 133\n",
      "49\t280\t2.2\t1\t1 279\n",
      "50\t142\t2.2\t1\t0 142\n",
      "51\t157\t2.2\t1\t2 155\n",
      "52\t254\t2.2\t1\t1 253\n",
      "53\t100\t2.2\t1\t1 99\n",
      "54\t108\t2.2\t1\t3 105\n",
      "55\t118\t2.2\t1\t0 118\n",
      "56\t268\t2.2\t1\t2 266\n",
      "57\t131\t2.2\t1\t0 131\n",
      "58\t330\t2.2\t1\t0 330\n",
      "59\t17\t2.2\t1\t0 17\n",
      "60\t28\t2.2\t1\t0 28\n",
      "61\t27\t2.2\t1\t0 27\n",
      "62\t43\t2.2\t1\t0 43\n",
      "63\t30\t2.2\t1\t1 29\n",
      "64\t35\t2.2\t1\t0 35\n",
      "65\t18\t2.2\t1\t0 18\n",
      "66\t36\t2.2\t1\t1 35\n",
      "67\t21\t2.2\t1\t0 21\n",
      "68\t11\t2.2\t1\t0 11\n",
      "69\t13\t2.2\t1\t0 13\n",
      "70\t15\t2.2\t1\t0 15\n",
      "71\t13\t2.2\t1\t0 13\n",
      "72\t14\t2.2\t1\t0 14\n",
      "73\t7\t2.2\t1\t0 7\n",
      "74\t10\t2.2\t1\t0 10\n",
      "75\t10\t2.2\t1\t1 9\n",
      "76\t23\t2.2\t1\t0 23\n",
      "77\t7\t2.2\t1\t0 7\n",
      "78\t22\t2.2\t1\t1 21\n",
      "79\t5\t2.2\t1\t0 5\n",
      "80\t21\t2.2\t1\t0 21\n",
      "81\t14\t2.2\t1\t0 14\n",
      "82\t18\t2.2\t1\t0 18\n",
      "83\t16\t2.2\t1\t0 16\n",
      "84\t15\t2.2\t1\t0 15\n",
      "85\t31\t2.2\t1\t0 31\n",
      "86\t8\t2.2\t1\t0 8\n",
      "87\t25\t2.2\t1\t0 25\n",
      "88\t11\t2.2\t1\t1 10\n",
      "89\t14\t2.2\t1\t0 14\n",
      "90\t10\t2.2\t1\t1 9\n",
      "91\t12\t2.2\t1\t0 12\n",
      "92\t17\t2.2\t1\t1 16\n",
      "93\t13\t2.2\t1\t0 13\n",
      "94\t15\t2.2\t1\t0 15\n",
      "95\t16\t2.2\t1\t2 14\n",
      "96\t19\t2.2\t1\t0 19\n",
      "97\t26\t2.2\t1\t0 26\n",
      "98\t21\t2.2\t1\t2 19\n",
      "99\t28\t2.2\t1\t1 27\n",
      "100\t30\t2.2\t1\t0 30\n",
      "101\t14\t2.2\t1\t0 14\n",
      "102\t18\t2.2\t1\t0 18\n",
      "103\t15\t2.2\t1\t1 14\n",
      "104\t20\t2.2\t1\t1 19\n",
      "105\t43\t2.2\t1\t0 43\n",
      "106\t16\t2.2\t1\t0 16\n",
      "107\t30\t2.2\t1\t0 30\n",
      "108\t12\t2.2\t1\t1 11\n",
      "109\t25\t2.2\t1\t1 24\n",
      "110\t11\t2.2\t1\t0 11\n",
      "111\t33\t2.2\t1\t7 26\n",
      "112\t43\t2.2\t1\t21 22\n",
      "113\t48\t2.2\t1\t10 38\n",
      "114\t46\t2.2\t1\t22 24\n",
      "115\t50\t2.2\t1\t10 40\n",
      "116\t39\t2.2\t1\t9 30\n",
      "117\t32\t2.2\t1\t10 22\n",
      "118\t46\t2.2\t1\t15 31\n",
      "119\t41\t2.2\t1\t7 34\n",
      "120\t45\t2.2\t1\t14 31\n",
      "121\t34\t2.2\t1\t15 19\n",
      "122\t44\t2.2\t1\t4 40\n",
      "123\t43\t2.2\t1\t2 41\n",
      "124\t31\t2.2\t1\t0 31\n",
      "125\t49\t2.2\t1\t2 47\n",
      "126\t197\t2.2\t1\t3 194\n",
      "127\t445\t2.2\t1\t5 440\n",
      "128\t856\t2.2\t1\t3 853\n",
      "129\t508\t2.2\t1\t4 504\n",
      "130\t551\t2.2\t1\t17 534\n",
      "131\t1129\t2.2\t1\t5 1124\n",
      "132\t485\t2.2\t1\t1 484\n",
      "133\t300\t2.2\t1\t2 298\n",
      "134\t508\t2.2\t1\t5 503\n",
      "135\t1252\t2.2\t1\t15 1237\n",
      "136\t290\t2.2\t1\t4 286\n",
      "137\t811\t2.2\t1\t7 804\n",
      "138\t84\t2.2\t1\t5 79\n",
      "139\t93\t2.2\t1\t10 83\n",
      "140\t119\t2.2\t1\t0 119\n",
      "141\t279\t2.2\t1\t3 276\n",
      "142\t35\t2.2\t1\t1 34\n",
      "143\t58\t2.2\t1\t0 58\n",
      "144\t22\t2.2\t1\t0 22\n",
      "145\t130\t2.2\t1\t0 130\n",
      "146\t22\t2.2\t1\t0 22\n",
      "147\t31\t2.2\t1\t0 31\n",
      "148\t15\t2.2\t1\t0 15\n",
      "149\t13\t2.2\t1\t0 13\n",
      "150\t37\t2.2\t1\t0 37\n",
      "151\t40\t2.2\t1\t0 40\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/SRR6399484_1.fastq.gz\n",
      "=============================================\n",
      "37169462 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/trim_galore/SRR6399484_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/SRR6399484_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'TGGAATTCTCGG' (Illumina small RNA adapter; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Optional adapter 2 sequence (only used for read 2 of paired-end files): 'GATCGTCGGACT'\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 18 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399484_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'GATCGTCGGACT' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/SRR6399484_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a GATCGTCGGACT /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/SRR6399484_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 100.325 s (2.699 µs/read; 22.23 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              37,169,462\n",
      "Reads with adapters:                12,249,634 (33.0%)\n",
      "Reads written (passing filters):    37,169,462 (100.0%)\n",
      "\n",
      "Total basepairs processed: 5,412,545,684 bp\n",
      "Quality-trimmed:                  37,182 bp (0.0%)\n",
      "Total written (filtered):  5,394,019,488 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: GATCGTCGGACT; Type: regular 3'; Length: 12; Trimmed: 12249634 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 26.0%\n",
      "  C: 21.7%\n",
      "  G: 31.2%\n",
      "  T: 21.0%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t9804735\t9292365.5\t0\t9804735\n",
      "2\t1781643\t2323091.4\t0\t1781643\n",
      "3\t443024\t580772.8\t0\t443024\n",
      "4\t106564\t145193.2\t0\t106564\n",
      "5\t63007\t36298.3\t0\t63007\n",
      "6\t3908\t9074.6\t0\t3908\n",
      "7\t3133\t2268.6\t0\t3133\n",
      "8\t1542\t567.2\t0\t1542\n",
      "9\t693\t141.8\t0\t262 431\n",
      "10\t802\t35.4\t1\t1 801\n",
      "11\t253\t8.9\t1\t0 253\n",
      "12\t138\t2.2\t1\t0 138\n",
      "13\t80\t2.2\t1\t0 80\n",
      "14\t162\t2.2\t1\t0 162\n",
      "15\t157\t2.2\t1\t0 157\n",
      "16\t169\t2.2\t1\t0 169\n",
      "17\t194\t2.2\t1\t0 194\n",
      "18\t148\t2.2\t1\t0 148\n",
      "19\t107\t2.2\t1\t0 107\n",
      "20\t103\t2.2\t1\t1 102\n",
      "21\t177\t2.2\t1\t0 177\n",
      "22\t246\t2.2\t1\t0 246\n",
      "23\t343\t2.2\t1\t1 342\n",
      "24\t210\t2.2\t1\t0 210\n",
      "25\t184\t2.2\t1\t1 183\n",
      "26\t241\t2.2\t1\t0 241\n",
      "27\t176\t2.2\t1\t0 176\n",
      "28\t242\t2.2\t1\t0 242\n",
      "29\t100\t2.2\t1\t0 100\n",
      "30\t200\t2.2\t1\t0 200\n",
      "31\t176\t2.2\t1\t0 176\n",
      "32\t143\t2.2\t1\t0 143\n",
      "33\t150\t2.2\t1\t0 150\n",
      "34\t164\t2.2\t1\t0 164\n",
      "35\t229\t2.2\t1\t0 229\n",
      "36\t133\t2.2\t1\t0 133\n",
      "37\t212\t2.2\t1\t0 212\n",
      "38\t151\t2.2\t1\t1 150\n",
      "39\t178\t2.2\t1\t0 178\n",
      "40\t190\t2.2\t1\t0 190\n",
      "41\t214\t2.2\t1\t0 214\n",
      "42\t173\t2.2\t1\t0 173\n",
      "43\t194\t2.2\t1\t0 194\n",
      "44\t213\t2.2\t1\t0 213\n",
      "45\t221\t2.2\t1\t0 221\n",
      "46\t143\t2.2\t1\t0 143\n",
      "47\t170\t2.2\t1\t0 170\n",
      "48\t268\t2.2\t1\t1 267\n",
      "49\t239\t2.2\t1\t1 238\n",
      "50\t299\t2.2\t1\t1 298\n",
      "51\t306\t2.2\t1\t0 306\n",
      "52\t273\t2.2\t1\t0 273\n",
      "53\t254\t2.2\t1\t0 254\n",
      "54\t209\t2.2\t1\t0 209\n",
      "55\t340\t2.2\t1\t2 338\n",
      "56\t302\t2.2\t1\t0 302\n",
      "57\t294\t2.2\t1\t0 294\n",
      "58\t253\t2.2\t1\t0 253\n",
      "59\t277\t2.2\t1\t0 277\n",
      "60\t438\t2.2\t1\t0 438\n",
      "61\t340\t2.2\t1\t0 340\n",
      "62\t1679\t2.2\t1\t0 1679\n",
      "63\t1671\t2.2\t1\t0 1671\n",
      "64\t924\t2.2\t1\t0 924\n",
      "65\t1291\t2.2\t1\t0 1291\n",
      "66\t1696\t2.2\t1\t0 1696\n",
      "67\t713\t2.2\t1\t0 713\n",
      "68\t976\t2.2\t1\t0 976\n",
      "69\t1270\t2.2\t1\t0 1270\n",
      "70\t1928\t2.2\t1\t1 1927\n",
      "71\t1172\t2.2\t1\t0 1172\n",
      "72\t1957\t2.2\t1\t0 1957\n",
      "73\t422\t2.2\t1\t0 422\n",
      "74\t351\t2.2\t1\t0 351\n",
      "75\t190\t2.2\t1\t0 190\n",
      "76\t631\t2.2\t1\t0 631\n",
      "77\t378\t2.2\t1\t0 378\n",
      "78\t234\t2.2\t1\t0 234\n",
      "79\t325\t2.2\t1\t0 325\n",
      "80\t313\t2.2\t1\t0 313\n",
      "81\t203\t2.2\t1\t0 203\n",
      "82\t237\t2.2\t1\t1 236\n",
      "83\t277\t2.2\t1\t0 277\n",
      "84\t272\t2.2\t1\t0 272\n",
      "85\t134\t2.2\t1\t0 134\n",
      "86\t191\t2.2\t1\t0 191\n",
      "87\t238\t2.2\t1\t1 237\n",
      "88\t289\t2.2\t1\t0 289\n",
      "89\t151\t2.2\t1\t0 151\n",
      "90\t273\t2.2\t1\t0 273\n",
      "91\t169\t2.2\t1\t0 169\n",
      "92\t150\t2.2\t1\t0 150\n",
      "93\t195\t2.2\t1\t0 195\n",
      "94\t150\t2.2\t1\t0 150\n",
      "95\t197\t2.2\t1\t0 197\n",
      "96\t109\t2.2\t1\t0 109\n",
      "97\t211\t2.2\t1\t0 211\n",
      "98\t193\t2.2\t1\t0 193\n",
      "99\t175\t2.2\t1\t0 175\n",
      "100\t222\t2.2\t1\t0 222\n",
      "101\t305\t2.2\t1\t0 305\n",
      "102\t143\t2.2\t1\t0 143\n",
      "103\t183\t2.2\t1\t0 183\n",
      "104\t151\t2.2\t1\t0 151\n",
      "105\t136\t2.2\t1\t0 136\n",
      "106\t183\t2.2\t1\t0 183\n",
      "107\t154\t2.2\t1\t0 154\n",
      "108\t177\t2.2\t1\t0 177\n",
      "109\t130\t2.2\t1\t0 130\n",
      "110\t253\t2.2\t1\t0 253\n",
      "111\t151\t2.2\t1\t0 151\n",
      "112\t205\t2.2\t1\t0 205\n",
      "113\t129\t2.2\t1\t0 129\n",
      "114\t236\t2.2\t1\t0 236\n",
      "115\t139\t2.2\t1\t0 139\n",
      "116\t68\t2.2\t1\t0 68\n",
      "117\t107\t2.2\t1\t0 107\n",
      "118\t53\t2.2\t1\t0 53\n",
      "119\t98\t2.2\t1\t0 98\n",
      "120\t117\t2.2\t1\t0 117\n",
      "121\t90\t2.2\t1\t0 90\n",
      "122\t225\t2.2\t1\t0 225\n",
      "123\t260\t2.2\t1\t0 260\n",
      "124\t304\t2.2\t1\t0 304\n",
      "125\t174\t2.2\t1\t0 174\n",
      "126\t114\t2.2\t1\t0 114\n",
      "127\t63\t2.2\t1\t0 63\n",
      "128\t76\t2.2\t1\t0 76\n",
      "129\t210\t2.2\t1\t0 210\n",
      "130\t158\t2.2\t1\t0 158\n",
      "131\t49\t2.2\t1\t0 49\n",
      "132\t127\t2.2\t1\t0 127\n",
      "133\t296\t2.2\t1\t0 296\n",
      "134\t118\t2.2\t1\t0 118\n",
      "135\t158\t2.2\t1\t0 158\n",
      "136\t131\t2.2\t1\t0 131\n",
      "137\t213\t2.2\t1\t0 213\n",
      "138\t75\t2.2\t1\t0 75\n",
      "139\t140\t2.2\t1\t0 140\n",
      "140\t116\t2.2\t1\t0 116\n",
      "141\t121\t2.2\t1\t0 121\n",
      "142\t38\t2.2\t1\t0 38\n",
      "143\t111\t2.2\t1\t0 111\n",
      "144\t65\t2.2\t1\t0 65\n",
      "145\t198\t2.2\t1\t0 198\n",
      "146\t307\t2.2\t1\t0 307\n",
      "147\t147\t2.2\t1\t0 147\n",
      "148\t131\t2.2\t1\t0 131\n",
      "149\t84\t2.2\t1\t0 84\n",
      "150\t142\t2.2\t1\t0 142\n",
      "151\t91\t2.2\t1\t0 91\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484/SRR6399484_2.fastq.gz\n",
      "=============================================\n",
      "37169462 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399484_1_trimmed.fq.gz and SRR6399484_2_trimmed.fq.gz\n",
      "file_1: SRR6399484_1_trimmed.fq.gz, file_2: SRR6399484_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399484_1_trimmed.fq.gz and SRR6399484_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399484_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399484_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 37169462\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (18 bp): 7256 (0.02%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399484_1_trimmed.fq.gz and SRR6399484_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   4%|█▊                                       |  ETA: 16.94 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3878.535805 seconds (119.47 k allocations: 3.400 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 98,171,796\n",
      "reads read      : 196,343,592\n",
      "reads written   : 196,343,592\n",
      "\n",
      "Using an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/SRR6399485_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "Nextera\t1\tCTGTCTCTTATA\t1000000\t0.00\n",
      "smallRNA\t0\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Using Nextera adapter for trimming (count: 1). Second best hit was smallRNA (count: 0)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/trim_galore/SRR6399485_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/SRR6399485_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399485_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/SRR6399485_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "60000000 sequences processed\n",
      "70000000 sequences processed\n",
      "80000000 sequences processed\n",
      "90000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/SRR6399485_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 270.186 s (2.752 µs/read; 21.80 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              98,171,796\n",
      "Reads with adapters:                32,121,637 (32.7%)\n",
      "Reads written (passing filters):    98,171,796 (100.0%)\n",
      "\n",
      "Total basepairs processed: 14,720,560,822 bp\n",
      "Quality-trimmed:                  43,545 bp (0.0%)\n",
      "Total written (filtered):  14,670,407,680 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 32121637 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 22.3%\n",
      "  C: 25.9%\n",
      "  G: 23.9%\n",
      "  T: 27.9%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t24047355\t24542949.0\t0\t24047355\n",
      "2\t5878046\t6135737.2\t0\t5878046\n",
      "3\t1558069\t1533934.3\t0\t1558069\n",
      "4\t385580\t383483.6\t0\t385580\n",
      "5\t97244\t95870.9\t0\t97244\n",
      "6\t37023\t23967.7\t0\t37023\n",
      "7\t11402\t5991.9\t0\t11402\n",
      "8\t8328\t1498.0\t0\t8328\n",
      "9\t3804\t374.5\t0\t901 2903\n",
      "10\t4887\t93.6\t1\t63 4824\n",
      "11\t2140\t23.4\t1\t17 2123\n",
      "12\t594\t5.9\t1\t2 592\n",
      "13\t667\t5.9\t1\t2 665\n",
      "14\t559\t5.9\t1\t0 559\n",
      "15\t472\t5.9\t1\t1 471\n",
      "16\t748\t5.9\t1\t1 747\n",
      "17\t385\t5.9\t1\t0 385\n",
      "18\t515\t5.9\t1\t1 514\n",
      "19\t658\t5.9\t1\t0 658\n",
      "20\t564\t5.9\t1\t0 564\n",
      "21\t1051\t5.9\t1\t0 1051\n",
      "22\t686\t5.9\t1\t1 685\n",
      "23\t803\t5.9\t1\t1 802\n",
      "24\t564\t5.9\t1\t1 563\n",
      "25\t581\t5.9\t1\t3 578\n",
      "26\t644\t5.9\t1\t0 644\n",
      "27\t500\t5.9\t1\t0 500\n",
      "28\t534\t5.9\t1\t1 533\n",
      "29\t542\t5.9\t1\t0 542\n",
      "30\t549\t5.9\t1\t0 549\n",
      "31\t777\t5.9\t1\t0 777\n",
      "32\t836\t5.9\t1\t2 834\n",
      "33\t764\t5.9\t1\t1 763\n",
      "34\t532\t5.9\t1\t2 530\n",
      "35\t446\t5.9\t1\t0 446\n",
      "36\t499\t5.9\t1\t0 499\n",
      "37\t596\t5.9\t1\t0 596\n",
      "38\t525\t5.9\t1\t0 525\n",
      "39\t611\t5.9\t1\t1 610\n",
      "40\t830\t5.9\t1\t1 829\n",
      "41\t626\t5.9\t1\t0 626\n",
      "42\t648\t5.9\t1\t1 647\n",
      "43\t656\t5.9\t1\t1 655\n",
      "44\t533\t5.9\t1\t2 531\n",
      "45\t579\t5.9\t1\t2 577\n",
      "46\t499\t5.9\t1\t2 497\n",
      "47\t604\t5.9\t1\t5 599\n",
      "48\t776\t5.9\t1\t1 775\n",
      "49\t679\t5.9\t1\t1 678\n",
      "50\t708\t5.9\t1\t0 708\n",
      "51\t726\t5.9\t1\t0 726\n",
      "52\t521\t5.9\t1\t1 520\n",
      "53\t510\t5.9\t1\t1 509\n",
      "54\t527\t5.9\t1\t2 525\n",
      "55\t685\t5.9\t1\t1 684\n",
      "56\t515\t5.9\t1\t0 515\n",
      "57\t851\t5.9\t1\t2 849\n",
      "58\t652\t5.9\t1\t0 652\n",
      "59\t600\t5.9\t1\t0 600\n",
      "60\t717\t5.9\t1\t1 716\n",
      "61\t804\t5.9\t1\t1 803\n",
      "62\t870\t5.9\t1\t0 870\n",
      "63\t852\t5.9\t1\t2 850\n",
      "64\t558\t5.9\t1\t0 558\n",
      "65\t489\t5.9\t1\t1 488\n",
      "66\t590\t5.9\t1\t0 590\n",
      "67\t486\t5.9\t1\t1 485\n",
      "68\t637\t5.9\t1\t2 635\n",
      "69\t726\t5.9\t1\t2 724\n",
      "70\t1141\t5.9\t1\t1 1140\n",
      "71\t678\t5.9\t1\t2 676\n",
      "72\t792\t5.9\t1\t1 791\n",
      "73\t688\t5.9\t1\t1 687\n",
      "74\t650\t5.9\t1\t6 644\n",
      "75\t509\t5.9\t1\t1 508\n",
      "76\t603\t5.9\t1\t0 603\n",
      "77\t622\t5.9\t1\t1 621\n",
      "78\t577\t5.9\t1\t1 576\n",
      "79\t657\t5.9\t1\t0 657\n",
      "80\t540\t5.9\t1\t0 540\n",
      "81\t643\t5.9\t1\t0 643\n",
      "82\t546\t5.9\t1\t0 546\n",
      "83\t562\t5.9\t1\t1 561\n",
      "84\t702\t5.9\t1\t0 702\n",
      "85\t501\t5.9\t1\t1 500\n",
      "86\t640\t5.9\t1\t1 639\n",
      "87\t652\t5.9\t1\t1 651\n",
      "88\t475\t5.9\t1\t1 474\n",
      "89\t558\t5.9\t1\t0 558\n",
      "90\t430\t5.9\t1\t0 430\n",
      "91\t554\t5.9\t1\t0 554\n",
      "92\t526\t5.9\t1\t0 526\n",
      "93\t703\t5.9\t1\t0 703\n",
      "94\t722\t5.9\t1\t3 719\n",
      "95\t507\t5.9\t1\t0 507\n",
      "96\t467\t5.9\t1\t2 465\n",
      "97\t563\t5.9\t1\t1 562\n",
      "98\t668\t5.9\t1\t1 667\n",
      "99\t534\t5.9\t1\t0 534\n",
      "100\t676\t5.9\t1\t1 675\n",
      "101\t516\t5.9\t1\t0 516\n",
      "102\t678\t5.9\t1\t1 677\n",
      "103\t639\t5.9\t1\t0 639\n",
      "104\t580\t5.9\t1\t1 579\n",
      "105\t406\t5.9\t1\t0 406\n",
      "106\t662\t5.9\t1\t0 662\n",
      "107\t773\t5.9\t1\t1 772\n",
      "108\t702\t5.9\t1\t2 700\n",
      "109\t563\t5.9\t1\t0 563\n",
      "110\t663\t5.9\t1\t1 662\n",
      "111\t559\t5.9\t1\t3 556\n",
      "112\t615\t5.9\t1\t0 615\n",
      "113\t777\t5.9\t1\t0 777\n",
      "114\t697\t5.9\t1\t0 697\n",
      "115\t538\t5.9\t1\t0 538\n",
      "116\t856\t5.9\t1\t1 855\n",
      "117\t599\t5.9\t1\t1 598\n",
      "118\t457\t5.9\t1\t0 457\n",
      "119\t531\t5.9\t1\t0 531\n",
      "120\t847\t5.9\t1\t1 846\n",
      "121\t543\t5.9\t1\t2 541\n",
      "122\t705\t5.9\t1\t0 705\n",
      "123\t685\t5.9\t1\t0 685\n",
      "124\t826\t5.9\t1\t0 826\n",
      "125\t786\t5.9\t1\t1 785\n",
      "126\t833\t5.9\t1\t5 828\n",
      "127\t612\t5.9\t1\t0 612\n",
      "128\t692\t5.9\t1\t0 692\n",
      "129\t418\t5.9\t1\t1 417\n",
      "130\t616\t5.9\t1\t0 616\n",
      "131\t623\t5.9\t1\t2 621\n",
      "132\t582\t5.9\t1\t1 581\n",
      "133\t586\t5.9\t1\t1 585\n",
      "134\t617\t5.9\t1\t1 616\n",
      "135\t513\t5.9\t1\t1 512\n",
      "136\t748\t5.9\t1\t2 746\n",
      "137\t600\t5.9\t1\t1 599\n",
      "138\t835\t5.9\t1\t0 835\n",
      "139\t631\t5.9\t1\t0 631\n",
      "140\t648\t5.9\t1\t0 648\n",
      "141\t680\t5.9\t1\t3 677\n",
      "142\t331\t5.9\t1\t2 329\n",
      "143\t816\t5.9\t1\t2 814\n",
      "144\t555\t5.9\t1\t2 553\n",
      "145\t581\t5.9\t1\t0 581\n",
      "146\t664\t5.9\t1\t0 664\n",
      "147\t489\t5.9\t1\t1 488\n",
      "148\t465\t5.9\t1\t0 465\n",
      "149\t550\t5.9\t1\t1 549\n",
      "150\t469\t5.9\t1\t0 469\n",
      "151\t660\t5.9\t1\t0 660\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/SRR6399485_1.fastq.gz\n",
      "=============================================\n",
      "98171796 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/trim_galore/SRR6399485_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/SRR6399485_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399485_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'CTGTCTCTTATA' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/SRR6399485_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "50000000 sequences processed\n",
      "60000000 sequences processed\n",
      "70000000 sequences processed\n",
      "80000000 sequences processed\n",
      "90000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/SRR6399485_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 292.607 s (2.981 µs/read; 20.13 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              98,171,796\n",
      "Reads with adapters:                31,848,078 (32.4%)\n",
      "Reads written (passing filters):    98,171,796 (100.0%)\n",
      "\n",
      "Total basepairs processed: 14,519,666,965 bp\n",
      "Quality-trimmed:                 147,563 bp (0.0%)\n",
      "Total written (filtered):  14,470,367,680 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CTGTCTCTTATA; Type: regular 3'; Length: 12; Trimmed: 31848078 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 21.7%\n",
      "  C: 27.7%\n",
      "  G: 23.4%\n",
      "  T: 27.3%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t24411070\t24542949.0\t0\t24411070\n",
      "2\t5282531\t6135737.2\t0\t5282531\n",
      "3\t1529286\t1533934.3\t0\t1529286\n",
      "4\t373135\t383483.6\t0\t373135\n",
      "5\t96613\t95870.9\t0\t96613\n",
      "6\t35533\t23967.7\t0\t35533\n",
      "7\t12568\t5991.9\t0\t12568\n",
      "8\t7688\t1498.0\t0\t7688\n",
      "9\t3581\t374.5\t0\t1138 2443\n",
      "10\t4690\t93.6\t1\t125 4565\n",
      "11\t2211\t23.4\t1\t12 2199\n",
      "12\t587\t5.9\t1\t1 586\n",
      "13\t623\t5.9\t1\t3 620\n",
      "14\t576\t5.9\t1\t0 576\n",
      "15\t474\t5.9\t1\t4 470\n",
      "16\t733\t5.9\t1\t0 733\n",
      "17\t429\t5.9\t1\t1 428\n",
      "18\t622\t5.9\t1\t2 620\n",
      "19\t667\t5.9\t1\t1 666\n",
      "20\t753\t5.9\t1\t2 751\n",
      "21\t901\t5.9\t1\t1 900\n",
      "22\t686\t5.9\t1\t1 685\n",
      "23\t813\t5.9\t1\t3 810\n",
      "24\t562\t5.9\t1\t0 562\n",
      "25\t650\t5.9\t1\t3 647\n",
      "26\t664\t5.9\t1\t1 663\n",
      "27\t574\t5.9\t1\t1 573\n",
      "28\t605\t5.9\t1\t0 605\n",
      "29\t674\t5.9\t1\t0 674\n",
      "30\t709\t5.9\t1\t3 706\n",
      "31\t939\t5.9\t1\t3 936\n",
      "32\t1062\t5.9\t1\t0 1062\n",
      "33\t827\t5.9\t1\t2 825\n",
      "34\t723\t5.9\t1\t3 720\n",
      "35\t528\t5.9\t1\t1 527\n",
      "36\t462\t5.9\t1\t1 461\n",
      "37\t728\t5.9\t1\t2 726\n",
      "38\t558\t5.9\t1\t3 555\n",
      "39\t701\t5.9\t1\t0 701\n",
      "40\t701\t5.9\t1\t1 700\n",
      "41\t571\t5.9\t1\t1 570\n",
      "42\t683\t5.9\t1\t5 678\n",
      "43\t627\t5.9\t1\t3 624\n",
      "44\t493\t5.9\t1\t4 489\n",
      "45\t601\t5.9\t1\t0 601\n",
      "46\t490\t5.9\t1\t1 489\n",
      "47\t609\t5.9\t1\t0 609\n",
      "48\t751\t5.9\t1\t1 750\n",
      "49\t709\t5.9\t1\t3 706\n",
      "50\t611\t5.9\t1\t1 610\n",
      "51\t755\t5.9\t1\t3 752\n",
      "52\t612\t5.9\t1\t1 611\n",
      "53\t587\t5.9\t1\t1 586\n",
      "54\t492\t5.9\t1\t0 492\n",
      "55\t822\t5.9\t1\t0 822\n",
      "56\t566\t5.9\t1\t2 564\n",
      "57\t752\t5.9\t1\t2 750\n",
      "58\t649\t5.9\t1\t0 649\n",
      "59\t585\t5.9\t1\t1 584\n",
      "60\t737\t5.9\t1\t1 736\n",
      "61\t773\t5.9\t1\t1 772\n",
      "62\t912\t5.9\t1\t1 911\n",
      "63\t783\t5.9\t1\t4 779\n",
      "64\t567\t5.9\t1\t2 565\n",
      "65\t545\t5.9\t1\t0 545\n",
      "66\t565\t5.9\t1\t0 565\n",
      "67\t545\t5.9\t1\t0 545\n",
      "68\t660\t5.9\t1\t3 657\n",
      "69\t682\t5.9\t1\t3 679\n",
      "70\t1096\t5.9\t1\t3 1093\n",
      "71\t725\t5.9\t1\t1 724\n",
      "72\t791\t5.9\t1\t0 791\n",
      "73\t697\t5.9\t1\t0 697\n",
      "74\t570\t5.9\t1\t3 567\n",
      "75\t581\t5.9\t1\t1 580\n",
      "76\t568\t5.9\t1\t1 567\n",
      "77\t681\t5.9\t1\t1 680\n",
      "78\t557\t5.9\t1\t0 557\n",
      "79\t682\t5.9\t1\t0 682\n",
      "80\t502\t5.9\t1\t0 502\n",
      "81\t619\t5.9\t1\t1 618\n",
      "82\t639\t5.9\t1\t2 637\n",
      "83\t582\t5.9\t1\t0 582\n",
      "84\t642\t5.9\t1\t0 642\n",
      "85\t509\t5.9\t1\t1 508\n",
      "86\t663\t5.9\t1\t7 656\n",
      "87\t567\t5.9\t1\t1 566\n",
      "88\t481\t5.9\t1\t1 480\n",
      "89\t525\t5.9\t1\t0 525\n",
      "90\t446\t5.9\t1\t0 446\n",
      "91\t669\t5.9\t1\t1 668\n",
      "92\t530\t5.9\t1\t0 530\n",
      "93\t778\t5.9\t1\t2 776\n",
      "94\t707\t5.9\t1\t2 705\n",
      "95\t544\t5.9\t1\t0 544\n",
      "96\t482\t5.9\t1\t2 480\n",
      "97\t574\t5.9\t1\t1 573\n",
      "98\t642\t5.9\t1\t1 641\n",
      "99\t522\t5.9\t1\t0 522\n",
      "100\t628\t5.9\t1\t1 627\n",
      "101\t475\t5.9\t1\t1 474\n",
      "102\t701\t5.9\t1\t0 701\n",
      "103\t605\t5.9\t1\t2 603\n",
      "104\t620\t5.9\t1\t1 619\n",
      "105\t497\t5.9\t1\t1 496\n",
      "106\t587\t5.9\t1\t5 582\n",
      "107\t818\t5.9\t1\t2 816\n",
      "108\t595\t5.9\t1\t2 593\n",
      "109\t662\t5.9\t1\t2 660\n",
      "110\t588\t5.9\t1\t1 587\n",
      "111\t562\t5.9\t1\t1 561\n",
      "112\t540\t5.9\t1\t0 540\n",
      "113\t776\t5.9\t1\t0 776\n",
      "114\t689\t5.9\t1\t0 689\n",
      "115\t559\t5.9\t1\t1 558\n",
      "116\t821\t5.9\t1\t2 819\n",
      "117\t603\t5.9\t1\t1 602\n",
      "118\t415\t5.9\t1\t0 415\n",
      "119\t619\t5.9\t1\t1 618\n",
      "120\t760\t5.9\t1\t2 758\n",
      "121\t611\t5.9\t1\t2 609\n",
      "122\t660\t5.9\t1\t0 660\n",
      "123\t687\t5.9\t1\t0 687\n",
      "124\t840\t5.9\t1\t3 837\n",
      "125\t732\t5.9\t1\t0 732\n",
      "126\t858\t5.9\t1\t1 857\n",
      "127\t604\t5.9\t1\t0 604\n",
      "128\t676\t5.9\t1\t3 673\n",
      "129\t469\t5.9\t1\t2 467\n",
      "130\t693\t5.9\t1\t1 692\n",
      "131\t664\t5.9\t1\t1 663\n",
      "132\t593\t5.9\t1\t4 589\n",
      "133\t637\t5.9\t1\t0 637\n",
      "134\t584\t5.9\t1\t1 583\n",
      "135\t586\t5.9\t1\t0 586\n",
      "136\t718\t5.9\t1\t1 717\n",
      "137\t580\t5.9\t1\t0 580\n",
      "138\t765\t5.9\t1\t0 765\n",
      "139\t683\t5.9\t1\t1 682\n",
      "140\t708\t5.9\t1\t0 708\n",
      "141\t664\t5.9\t1\t3 661\n",
      "142\t335\t5.9\t1\t1 334\n",
      "143\t825\t5.9\t1\t0 825\n",
      "144\t532\t5.9\t1\t2 530\n",
      "145\t510\t5.9\t1\t0 510\n",
      "146\t602\t5.9\t1\t1 601\n",
      "147\t443\t5.9\t1\t3 440\n",
      "148\t467\t5.9\t1\t0 467\n",
      "149\t515\t5.9\t1\t4 511\n",
      "150\t437\t5.9\t1\t0 437\n",
      "151\t538\t5.9\t1\t0 538\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485/SRR6399485_2.fastq.gz\n",
      "=============================================\n",
      "98171796 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399485_1_trimmed.fq.gz and SRR6399485_2_trimmed.fq.gz\n",
      "file_1: SRR6399485_1_trimmed.fq.gz, file_2: SRR6399485_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399485_1_trimmed.fq.gz and SRR6399485_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399485_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399485_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 98171796\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (20 bp): 30761 (0.03%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399485_1_trimmed.fq.gz and SRR6399485_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   5%|█▉                                       |  ETA: 17.72 days\u001b[39mUsing an excessive number of cores has a diminishing return! It is recommended not to exceed 8 cores per trimming process (you asked for 48 cores). Please consider re-specifying\n",
      "Path to Cutadapt set as: 'cutadapt' (default)\n",
      "Cutadapt seems to be working fine (tested command 'cutadapt --version')\n",
      "Cutadapt version: 4.2\n",
      "Could not detect version of Python used by Cutadapt from the first line of Cutadapt (but found this: >>>#!/bin/sh<<<)\n",
      "Letting the (modified) Cutadapt deal with the Python version instead\n",
      "Parallel gzip (pigz) detected. Proceeding with multicore (de)compression using 48 cores\n",
      "\n",
      "igzip detected. Using igzip for decompressing\n",
      "\n",
      "No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)\n",
      "\n",
      "Output directory /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/trim_galore/ doesn't exist, creating it for you...\n",
      "\n",
      "Output will be written into the directory: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/trim_galore/\n",
      "\n",
      "\n",
      "AUTO-DETECTING ADAPTER TYPE\n",
      "===========================\n",
      "Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/SRR6399486_1.fastq.gz <<)\n",
      "\n",
      "Found perfect matches for the following adapter sequences:\n",
      "Adapter type\tCount\tSequence\tSequences analysed\tPercentage\n",
      "smallRNA\t6\tTGGAATTCTCGG\t1000000\t0.00\n",
      "Nextera\t0\tCTGTCTCTTATA\t1000000\t0.00\n",
      "Illumina\t0\tAGATCGGAAGAGC\t1000000\t0.00\n",
      "Using smallRNA adapter for trimming (count: 6). Second best hit was Nextera (count: 0)\n",
      "\n",
      "Reducing length cutoff to 18bp for small RNA-Seq reads because a cutoff of 20bp may remove some short species of small RNAs if they had been trimmed by 1,2 or 3bp\n",
      "Setting the Illumina smallRNA 5' adapter as adapter 2: 'GATCGTCGGACT'\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/trim_galore/SRR6399486_1.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/SRR6399486_1.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'TGGAATTCTCGG' (Illumina small RNA adapter; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Optional adapter 2 sequence (only used for read 2 of paired-end files): 'GATCGTCGGACT'\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 18 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399486_1_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'TGGAATTCTCGG' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/SRR6399486_1.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a TGGAATTCTCGG /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/SRR6399486_1.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 344.460 s (6.963 µs/read; 8.62 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              49,468,670\n",
      "Reads with adapters:                16,481,144 (33.3%)\n",
      "Reads written (passing filters):    49,468,670 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,407,417,690 bp\n",
      "Quality-trimmed:                  49,562 bp (0.0%)\n",
      "Total written (filtered):  7,383,623,586 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: TGGAATTCTCGG; Type: regular 3'; Length: 12; Trimmed: 16481144 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 25.4%\n",
      "  C: 23.4%\n",
      "  G: 20.6%\n",
      "  T: 30.6%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t11914235\t12367167.5\t0\t11914235\n",
      "2\t3336822\t3091791.9\t0\t3336822\n",
      "3\t968441\t772948.0\t0\t968441\n",
      "4\t168633\t193237.0\t0\t168633\n",
      "5\t59970\t48309.2\t0\t59970\n",
      "6\t10900\t12077.3\t0\t10900\n",
      "7\t3822\t3019.3\t0\t3822\n",
      "8\t1102\t754.8\t0\t1102\n",
      "9\t1043\t188.7\t0\t125 918\n",
      "10\t1756\t47.2\t1\t15 1741\n",
      "11\t435\t11.8\t1\t3 432\n",
      "12\t132\t2.9\t1\t1 131\n",
      "13\t87\t2.9\t1\t1 86\n",
      "14\t101\t2.9\t1\t0 101\n",
      "15\t101\t2.9\t1\t0 101\n",
      "16\t118\t2.9\t1\t0 118\n",
      "17\t104\t2.9\t1\t0 104\n",
      "18\t91\t2.9\t1\t1 90\n",
      "19\t100\t2.9\t1\t0 100\n",
      "20\t100\t2.9\t1\t1 99\n",
      "21\t109\t2.9\t1\t0 109\n",
      "22\t87\t2.9\t1\t0 87\n",
      "23\t81\t2.9\t1\t0 81\n",
      "24\t66\t2.9\t1\t0 66\n",
      "25\t70\t2.9\t1\t0 70\n",
      "26\t101\t2.9\t1\t1 100\n",
      "27\t80\t2.9\t1\t0 80\n",
      "28\t89\t2.9\t1\t0 89\n",
      "29\t77\t2.9\t1\t0 77\n",
      "30\t87\t2.9\t1\t3 84\n",
      "31\t102\t2.9\t1\t1 101\n",
      "32\t65\t2.9\t1\t2 63\n",
      "33\t100\t2.9\t1\t1 99\n",
      "34\t77\t2.9\t1\t0 77\n",
      "35\t91\t2.9\t1\t1 90\n",
      "36\t107\t2.9\t1\t0 107\n",
      "37\t92\t2.9\t1\t1 91\n",
      "38\t67\t2.9\t1\t0 67\n",
      "39\t109\t2.9\t1\t0 109\n",
      "40\t155\t2.9\t1\t0 155\n",
      "41\t104\t2.9\t1\t0 104\n",
      "42\t80\t2.9\t1\t0 80\n",
      "43\t65\t2.9\t1\t0 65\n",
      "44\t93\t2.9\t1\t0 93\n",
      "45\t112\t2.9\t1\t0 112\n",
      "46\t110\t2.9\t1\t0 110\n",
      "47\t121\t2.9\t1\t0 121\n",
      "48\t51\t2.9\t1\t0 51\n",
      "49\t87\t2.9\t1\t1 86\n",
      "50\t102\t2.9\t1\t0 102\n",
      "51\t115\t2.9\t1\t0 115\n",
      "52\t130\t2.9\t1\t0 130\n",
      "53\t72\t2.9\t1\t0 72\n",
      "54\t90\t2.9\t1\t2 88\n",
      "55\t111\t2.9\t1\t0 111\n",
      "56\t89\t2.9\t1\t1 88\n",
      "57\t85\t2.9\t1\t0 85\n",
      "58\t145\t2.9\t1\t0 145\n",
      "59\t112\t2.9\t1\t0 112\n",
      "60\t114\t2.9\t1\t1 113\n",
      "61\t70\t2.9\t1\t0 70\n",
      "62\t124\t2.9\t1\t0 124\n",
      "63\t120\t2.9\t1\t0 120\n",
      "64\t178\t2.9\t1\t0 178\n",
      "65\t91\t2.9\t1\t1 90\n",
      "66\t112\t2.9\t1\t1 111\n",
      "67\t87\t2.9\t1\t0 87\n",
      "68\t106\t2.9\t1\t0 106\n",
      "69\t91\t2.9\t1\t1 90\n",
      "70\t90\t2.9\t1\t0 90\n",
      "71\t72\t2.9\t1\t0 72\n",
      "72\t80\t2.9\t1\t0 80\n",
      "73\t133\t2.9\t1\t0 133\n",
      "74\t157\t2.9\t1\t0 157\n",
      "75\t118\t2.9\t1\t2 116\n",
      "76\t111\t2.9\t1\t0 111\n",
      "77\t148\t2.9\t1\t0 148\n",
      "78\t84\t2.9\t1\t0 84\n",
      "79\t151\t2.9\t1\t0 151\n",
      "80\t110\t2.9\t1\t0 110\n",
      "81\t103\t2.9\t1\t0 103\n",
      "82\t93\t2.9\t1\t1 92\n",
      "83\t112\t2.9\t1\t1 111\n",
      "84\t79\t2.9\t1\t0 79\n",
      "85\t86\t2.9\t1\t0 86\n",
      "86\t94\t2.9\t1\t0 94\n",
      "87\t75\t2.9\t1\t1 74\n",
      "88\t103\t2.9\t1\t0 103\n",
      "89\t88\t2.9\t1\t0 88\n",
      "90\t84\t2.9\t1\t0 84\n",
      "91\t69\t2.9\t1\t0 69\n",
      "92\t152\t2.9\t1\t1 151\n",
      "93\t108\t2.9\t1\t0 108\n",
      "94\t93\t2.9\t1\t1 92\n",
      "95\t90\t2.9\t1\t0 90\n",
      "96\t100\t2.9\t1\t1 99\n",
      "97\t84\t2.9\t1\t0 84\n",
      "98\t69\t2.9\t1\t1 68\n",
      "99\t115\t2.9\t1\t0 115\n",
      "100\t87\t2.9\t1\t0 87\n",
      "101\t113\t2.9\t1\t0 113\n",
      "102\t126\t2.9\t1\t1 125\n",
      "103\t99\t2.9\t1\t0 99\n",
      "104\t114\t2.9\t1\t1 113\n",
      "105\t155\t2.9\t1\t1 154\n",
      "106\t140\t2.9\t1\t1 139\n",
      "107\t91\t2.9\t1\t1 90\n",
      "108\t76\t2.9\t1\t1 75\n",
      "109\t101\t2.9\t1\t0 101\n",
      "110\t115\t2.9\t1\t0 115\n",
      "111\t68\t2.9\t1\t0 68\n",
      "112\t89\t2.9\t1\t1 88\n",
      "113\t77\t2.9\t1\t0 77\n",
      "114\t98\t2.9\t1\t0 98\n",
      "115\t87\t2.9\t1\t0 87\n",
      "116\t138\t2.9\t1\t0 138\n",
      "117\t107\t2.9\t1\t0 107\n",
      "118\t96\t2.9\t1\t0 96\n",
      "119\t72\t2.9\t1\t0 72\n",
      "120\t91\t2.9\t1\t1 90\n",
      "121\t110\t2.9\t1\t1 109\n",
      "122\t84\t2.9\t1\t0 84\n",
      "123\t100\t2.9\t1\t0 100\n",
      "124\t170\t2.9\t1\t0 170\n",
      "125\t125\t2.9\t1\t1 124\n",
      "126\t84\t2.9\t1\t0 84\n",
      "127\t111\t2.9\t1\t0 111\n",
      "128\t101\t2.9\t1\t0 101\n",
      "129\t121\t2.9\t1\t1 120\n",
      "130\t101\t2.9\t1\t0 101\n",
      "131\t67\t2.9\t1\t0 67\n",
      "132\t86\t2.9\t1\t2 84\n",
      "133\t89\t2.9\t1\t0 89\n",
      "134\t149\t2.9\t1\t0 149\n",
      "135\t67\t2.9\t1\t0 67\n",
      "136\t92\t2.9\t1\t1 91\n",
      "137\t114\t2.9\t1\t1 113\n",
      "138\t76\t2.9\t1\t0 76\n",
      "139\t87\t2.9\t1\t0 87\n",
      "140\t134\t2.9\t1\t0 134\n",
      "141\t59\t2.9\t1\t0 59\n",
      "142\t85\t2.9\t1\t0 85\n",
      "143\t67\t2.9\t1\t0 67\n",
      "144\t100\t2.9\t1\t1 99\n",
      "145\t53\t2.9\t1\t0 53\n",
      "146\t65\t2.9\t1\t0 65\n",
      "147\t89\t2.9\t1\t3 86\n",
      "148\t150\t2.9\t1\t0 150\n",
      "149\t159\t2.9\t1\t1 158\n",
      "150\t106\t2.9\t1\t0 106\n",
      "151\t80\t2.9\t1\t0 80\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/SRR6399486_1.fastq.gz\n",
      "=============================================\n",
      "49468670 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Writing report to '/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/trim_galore/SRR6399486_2.fastq.gz_trimming_report.txt'\n",
      "\n",
      "SUMMARISING RUN PARAMETERS\n",
      "==========================\n",
      "Input filename: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/SRR6399486_2.fastq.gz\n",
      "Trimming mode: paired-end\n",
      "Trim Galore version: 0.6.10\n",
      "Cutadapt version: 4.2\n",
      "Python version: could not detect\n",
      "Number of cores used for trimming: 48\n",
      "Quality Phred score cutoff: 20\n",
      "Quality encoding type selected: ASCII+33\n",
      "Adapter sequence: 'TGGAATTCTCGG' (Illumina small RNA adapter; auto-detected)\n",
      "Maximum trimming error rate: 0.1 (default)\n",
      "Optional adapter 2 sequence (only used for read 2 of paired-end files): 'GATCGTCGGACT'\n",
      "Minimum required adapter overlap (stringency): 1 bp\n",
      "Minimum required sequence length for both reads before a sequence pair gets removed: 18 bp\n",
      "Output file(s) will be GZIP compressed\n",
      "\n",
      "Cutadapt seems to be fairly up-to-date (version 4.2). Setting -j -j 48\n",
      "Writing final adapter and quality trimmed output to SRR6399486_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      "  >>> Now performing quality (cutoff '-q 20') and adapter trimming in a single pass for the adapter sequence: 'GATCGTCGGACT' from file /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/SRR6399486_2.fastq.gz <<< \n",
      "10000000 sequences processed\n",
      "20000000 sequences processed\n",
      "30000000 sequences processed\n",
      "40000000 sequences processed\n",
      "This is cutadapt 4.2 with Python 3.10.6\n",
      "Command line parameters: -j 48 -e 0.1 -q 20 -O 1 -a GATCGTCGGACT /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/SRR6399486_2.fastq.gz\n",
      "Processing single-end reads on 48 cores ...\n",
      "Finished in 384.638 s (7.775 µs/read; 7.72 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:              49,468,670\n",
      "Reads with adapters:                17,613,786 (35.6%)\n",
      "Reads written (passing filters):    49,468,670 (100.0%)\n",
      "\n",
      "Total basepairs processed: 7,162,725,698 bp\n",
      "Quality-trimmed:                 107,848 bp (0.0%)\n",
      "Total written (filtered):  7,139,248,079 bp (99.7%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: GATCGTCGGACT; Type: regular 3'; Length: 12; Trimmed: 17613786 times\n",
      "\n",
      "Minimum overlap: 1\n",
      "No. of allowed errors:\n",
      "1-9 bp: 0; 10-12 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 24.6%\n",
      "  C: 22.6%\n",
      "  G: 28.7%\n",
      "  T: 24.1%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "1\t14483547\t12367167.5\t0\t14483547\n",
      "2\t2346592\t3091791.9\t0\t2346592\n",
      "3\t514151\t772948.0\t0\t514151\n",
      "4\t168431\t193237.0\t0\t168431\n",
      "5\t68773\t48309.2\t0\t68773\n",
      "6\t5693\t12077.3\t0\t5693\n",
      "7\t3056\t3019.3\t0\t3056\n",
      "8\t1110\t754.8\t0\t1110\n",
      "9\t932\t188.7\t0\t196 736\n",
      "10\t1383\t47.2\t1\t10 1373\n",
      "11\t526\t11.8\t1\t1 525\n",
      "12\t159\t2.9\t1\t0 159\n",
      "13\t159\t2.9\t1\t0 159\n",
      "14\t148\t2.9\t1\t2 146\n",
      "15\t122\t2.9\t1\t0 122\n",
      "16\t116\t2.9\t1\t3 113\n",
      "17\t142\t2.9\t1\t0 142\n",
      "18\t107\t2.9\t1\t2 105\n",
      "19\t169\t2.9\t1\t0 169\n",
      "20\t133\t2.9\t1\t1 132\n",
      "21\t187\t2.9\t1\t2 185\n",
      "22\t142\t2.9\t1\t5 137\n",
      "23\t172\t2.9\t1\t3 169\n",
      "24\t124\t2.9\t1\t0 124\n",
      "25\t166\t2.9\t1\t0 166\n",
      "26\t142\t2.9\t1\t0 142\n",
      "27\t173\t2.9\t1\t6 167\n",
      "28\t142\t2.9\t1\t1 141\n",
      "29\t173\t2.9\t1\t1 172\n",
      "30\t136\t2.9\t1\t2 134\n",
      "31\t133\t2.9\t1\t0 133\n",
      "32\t114\t2.9\t1\t2 112\n",
      "33\t161\t2.9\t1\t1 160\n",
      "34\t123\t2.9\t1\t0 123\n",
      "35\t144\t2.9\t1\t3 141\n",
      "36\t169\t2.9\t1\t0 169\n",
      "37\t106\t2.9\t1\t2 104\n",
      "38\t175\t2.9\t1\t4 171\n",
      "39\t171\t2.9\t1\t2 169\n",
      "40\t186\t2.9\t1\t1 185\n",
      "41\t141\t2.9\t1\t1 140\n",
      "42\t163\t2.9\t1\t3 160\n",
      "43\t135\t2.9\t1\t2 133\n",
      "44\t158\t2.9\t1\t0 158\n",
      "45\t157\t2.9\t1\t2 155\n",
      "46\t193\t2.9\t1\t0 193\n",
      "47\t197\t2.9\t1\t1 196\n",
      "48\t112\t2.9\t1\t2 110\n",
      "49\t157\t2.9\t1\t0 157\n",
      "50\t122\t2.9\t1\t0 122\n",
      "51\t138\t2.9\t1\t6 132\n",
      "52\t196\t2.9\t1\t2 194\n",
      "53\t100\t2.9\t1\t2 98\n",
      "54\t164\t2.9\t1\t1 163\n",
      "55\t134\t2.9\t1\t3 131\n",
      "56\t140\t2.9\t1\t4 136\n",
      "57\t161\t2.9\t1\t1 160\n",
      "58\t160\t2.9\t1\t6 154\n",
      "59\t133\t2.9\t1\t0 133\n",
      "60\t172\t2.9\t1\t1 171\n",
      "61\t113\t2.9\t1\t3 110\n",
      "62\t121\t2.9\t1\t1 120\n",
      "63\t156\t2.9\t1\t2 154\n",
      "64\t172\t2.9\t1\t0 172\n",
      "65\t165\t2.9\t1\t0 165\n",
      "66\t155\t2.9\t1\t4 151\n",
      "67\t160\t2.9\t1\t1 159\n",
      "68\t180\t2.9\t1\t1 179\n",
      "69\t146\t2.9\t1\t2 144\n",
      "70\t172\t2.9\t1\t3 169\n",
      "71\t152\t2.9\t1\t2 150\n",
      "72\t151\t2.9\t1\t2 149\n",
      "73\t116\t2.9\t1\t1 115\n",
      "74\t101\t2.9\t1\t2 99\n",
      "75\t168\t2.9\t1\t1 167\n",
      "76\t120\t2.9\t1\t0 120\n",
      "77\t127\t2.9\t1\t1 126\n",
      "78\t184\t2.9\t1\t1 183\n",
      "79\t136\t2.9\t1\t1 135\n",
      "80\t120\t2.9\t1\t0 120\n",
      "81\t157\t2.9\t1\t1 156\n",
      "82\t140\t2.9\t1\t2 138\n",
      "83\t118\t2.9\t1\t2 116\n",
      "84\t151\t2.9\t1\t1 150\n",
      "85\t153\t2.9\t1\t2 151\n",
      "86\t126\t2.9\t1\t3 123\n",
      "87\t152\t2.9\t1\t0 152\n",
      "88\t134\t2.9\t1\t2 132\n",
      "89\t174\t2.9\t1\t0 174\n",
      "90\t113\t2.9\t1\t0 113\n",
      "91\t150\t2.9\t1\t1 149\n",
      "92\t181\t2.9\t1\t6 175\n",
      "93\t122\t2.9\t1\t2 120\n",
      "94\t174\t2.9\t1\t1 173\n",
      "95\t104\t2.9\t1\t0 104\n",
      "96\t143\t2.9\t1\t0 143\n",
      "97\t105\t2.9\t1\t1 104\n",
      "98\t174\t2.9\t1\t2 172\n",
      "99\t153\t2.9\t1\t2 151\n",
      "100\t147\t2.9\t1\t1 146\n",
      "101\t95\t2.9\t1\t1 94\n",
      "102\t147\t2.9\t1\t1 146\n",
      "103\t125\t2.9\t1\t2 123\n",
      "104\t150\t2.9\t1\t1 149\n",
      "105\t109\t2.9\t1\t1 108\n",
      "106\t127\t2.9\t1\t3 124\n",
      "107\t139\t2.9\t1\t5 134\n",
      "108\t151\t2.9\t1\t1 150\n",
      "109\t153\t2.9\t1\t1 152\n",
      "110\t148\t2.9\t1\t0 148\n",
      "111\t118\t2.9\t1\t1 117\n",
      "112\t129\t2.9\t1\t3 126\n",
      "113\t135\t2.9\t1\t2 133\n",
      "114\t124\t2.9\t1\t0 124\n",
      "115\t178\t2.9\t1\t1 177\n",
      "116\t130\t2.9\t1\t4 126\n",
      "117\t133\t2.9\t1\t0 133\n",
      "118\t178\t2.9\t1\t4 174\n",
      "119\t84\t2.9\t1\t2 82\n",
      "120\t137\t2.9\t1\t1 136\n",
      "121\t123\t2.9\t1\t0 123\n",
      "122\t126\t2.9\t1\t1 125\n",
      "123\t124\t2.9\t1\t3 121\n",
      "124\t150\t2.9\t1\t0 150\n",
      "125\t129\t2.9\t1\t3 126\n",
      "126\t129\t2.9\t1\t1 128\n",
      "127\t124\t2.9\t1\t1 123\n",
      "128\t192\t2.9\t1\t4 188\n",
      "129\t157\t2.9\t1\t4 153\n",
      "130\t97\t2.9\t1\t1 96\n",
      "131\t109\t2.9\t1\t5 104\n",
      "132\t160\t2.9\t1\t0 160\n",
      "133\t121\t2.9\t1\t2 119\n",
      "134\t105\t2.9\t1\t2 103\n",
      "135\t108\t2.9\t1\t0 108\n",
      "136\t155\t2.9\t1\t2 153\n",
      "137\t125\t2.9\t1\t4 121\n",
      "138\t141\t2.9\t1\t3 138\n",
      "139\t94\t2.9\t1\t0 94\n",
      "140\t106\t2.9\t1\t1 105\n",
      "141\t104\t2.9\t1\t1 103\n",
      "142\t108\t2.9\t1\t1 107\n",
      "143\t103\t2.9\t1\t1 102\n",
      "144\t122\t2.9\t1\t0 122\n",
      "145\t109\t2.9\t1\t2 107\n",
      "146\t143\t2.9\t1\t0 143\n",
      "147\t157\t2.9\t1\t2 155\n",
      "148\t92\t2.9\t1\t1 91\n",
      "149\t85\t2.9\t1\t1 84\n",
      "150\t102\t2.9\t1\t2 100\n",
      "151\t89\t2.9\t1\t0 89\n",
      "\n",
      "RUN STATISTICS FOR INPUT FILE: /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486/SRR6399486_2.fastq.gz\n",
      "=============================================\n",
      "49468670 sequences processed in total\n",
      "The length threshold of paired-end sequences gets evaluated later on (in the validation step)\n",
      "\n",
      "Validate paired-end files SRR6399486_1_trimmed.fq.gz and SRR6399486_2_trimmed.fq.gz\n",
      "file_1: SRR6399486_1_trimmed.fq.gz, file_2: SRR6399486_2_trimmed.fq.gz\n",
      "\n",
      "\n",
      ">>>>> Now validing the length of the 2 paired-end infiles: SRR6399486_1_trimmed.fq.gz and SRR6399486_2_trimmed.fq.gz <<<<<\n",
      "Writing validated paired-end Read 1 reads to SRR6399486_1_val_1.fq.gz\n",
      "Writing validated paired-end Read 2 reads to SRR6399486_2_val_2.fq.gz\n",
      "\n",
      "Total number of sequences analysed: 49468670\n",
      "\n",
      "Number of sequence pairs removed because at least one read was shorter than the length cutoff (18 bp): 4391 (0.01%)\n",
      "\n",
      "Deleting both intermediate output files SRR6399486_1_trimmed.fq.gz and SRR6399486_2_trimmed.fq.gz\n",
      "\n",
      "====================================================================================================\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pigz 2.6\n",
      "igzip command line interface 2.30.0\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   5%|█▉                                       |  ETA: 18.34 days\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2890.503111 seconds (89.06 k allocations: 2.535 MiB)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "spots read      : 70,466,130\n",
      "reads read      : 140,932,260\n",
      "reads written   : 140,932,260\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ProgressMeter.@showprogress for srr_identifier in srr_identifiers\n",
    "    sample_outdir = \"$(dirname(pwd()))/data/SRA/$(srr_identifier)\"\n",
    "    # fasterq_dump(outdir=sample_outdir, srr_identifier=srr_identifier)\n",
    "    download_and_filter_reads(outdir=sample_outdir, srr_identifier=srr_identifier)\n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.3",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
