{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d0594e70-4ab7-40f1-9b74-dd2897562648",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m  Activating\u001b[22m\u001b[39m environment at `~/work/Mycelia/projects/viral-pangenome-discovery/notebooks/Project.toml`\n",
      "\u001b[32m\u001b[1m    Updating\u001b[22m\u001b[39m registry at `/opt/julia/registries/General`\n",
      "\u001b[32m\u001b[1m   Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/work/Mycelia/projects/viral-pangenome-discovery/notebooks/Project.toml`\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/work/Mycelia/projects/viral-pangenome-discovery/notebooks/Manifest.toml`\n"
     ]
    }
   ],
   "source": [
    "import Pkg\n",
    "Pkg.activate(\".\")\n",
    "\n",
    "# not for the faint of heart!\n",
    "# Pkg.update()\n",
    "\n",
    "pkgs = [\n",
    "\"ArgParse\",\n",
    "\"Base64\",\n",
    "\"BioSequences\",\n",
    "\"DataFrames\",\n",
    "\"Dates\",\n",
    "\"DelimitedFiles\",\n",
    "\"FASTX\",\n",
    "\"GLM\",\n",
    "\"HTTP\",\n",
    "\"JSON\",\n",
    "\"Graphs\",\n",
    "\"MetaGraphs\",\n",
    "\"MD5\",\n",
    "\"Statistics\",\n",
    "\"StatsPlots\",\n",
    "\"uCSV\",\n",
    "\"CodecZlib\",\n",
    "\"YAML\",\n",
    "\"Revise\",\n",
    "\"Kmers\",\n",
    "\"StatsBase\",\n",
    "\"ProgressMeter\"\n",
    "]\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "16b78fcf-5a2f-40dc-b136-db19875d8861",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data\""
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir = joinpath(dirname(pwd()), \"data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b47aae98-1bca-4d5b-9dae-cd9698c5bb89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32-element Vector{String}:\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399459\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399460\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399461\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399463\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471\"\n",
       " ⋮\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399479\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399480\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399481\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399482\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399483\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399484\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399485\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399486\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399487\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399488\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399489\"\n",
       " \"/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399490\""
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRR_paths = filter(x -> !occursin(\".ipynb_checkpoints\", x), readdir(joinpath(data_dir, \"SRA\"), join=true))\n",
    "SRR_paths = filter(x -> \"trim_galore\" in readdir(x), SRR_paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f56746fd-6258-4912-b064-dde22bfdaa2a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR = \"SRR6399459\"\n",
      "SRR = \"SRR6399460\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   6%|██▌                                      |  ETA: 0:00:58\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR = \"SRR6399461\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-02-23 22:38:58 - MEGAHIT v1.2.9\n",
      "2023-02-23 22:38:58 - Using megahit_core with POPCNT and BMI2 support\n",
      "2023-02-23 22:38:58 - Convert reads to binary library\n",
      "2023-02-23 22:44:22 - b'INFO  sequence/io/sequence_lib.cpp  :   75 - Lib 0 (/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399461/trim_galore/SRR6399461_1_val_1.fq.gz,/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399461/trim_galore/SRR6399461_2_val_2.fq.gz): pe, 136163688 reads, 151 max length'\n",
      "2023-02-23 22:44:29 - b'INFO  utils/utils.h                 :  152 - Real: 330.9140\\tuser: 147.9503\\tsys: 10.6447\\tmaxrss: 254416'\n",
      "2023-02-23 22:44:29 - k-max reset to: 141 \n",
      "2023-02-23 22:44:29 - Start assembly. Number of CPU threads 48 \n",
      "2023-02-23 22:44:29 - k list: 21,29,39,59,79,99,119,141 \n",
      "2023-02-23 22:44:29 - Memory used: 121546591027\n",
      "2023-02-23 22:44:29 - Extract solid (k+1)-mers for k = 21 \n",
      "2023-02-23 22:54:23 - Build graph for k = 21 \n",
      "2023-02-23 22:56:14 - Assemble contigs from SdBG for k = 21\n",
      "2023-02-23 23:03:02 - Local assembly for k = 21\n",
      "2023-02-23 23:05:18 - Extract iterative edges from k = 21 to 29 \n",
      "2023-02-23 23:09:11 - Build graph for k = 29 \n",
      "2023-02-23 23:09:50 - Assemble contigs from SdBG for k = 29\n",
      "2023-02-23 23:10:26 - Local assembly for k = 29\n",
      "2023-02-23 23:12:23 - Extract iterative edges from k = 29 to 39 \n",
      "2023-02-23 23:12:54 - Build graph for k = 39 \n",
      "2023-02-23 23:13:03 - Assemble contigs from SdBG for k = 39\n",
      "2023-02-23 23:13:20 - Local assembly for k = 39\n",
      "2023-02-23 23:15:05 - Extract iterative edges from k = 39 to 59 \n",
      "2023-02-23 23:15:44 - Build graph for k = 59 \n",
      "2023-02-23 23:15:53 - Assemble contigs from SdBG for k = 59\n",
      "2023-02-23 23:16:10 - Local assembly for k = 59\n",
      "2023-02-23 23:18:03 - Extract iterative edges from k = 59 to 79 \n",
      "2023-02-23 23:18:26 - Build graph for k = 79 \n",
      "2023-02-23 23:18:36 - Assemble contigs from SdBG for k = 79\n",
      "2023-02-23 23:18:51 - Local assembly for k = 79\n",
      "2023-02-23 23:20:43 - Extract iterative edges from k = 79 to 99 \n",
      "2023-02-23 23:21:06 - Build graph for k = 99 \n",
      "2023-02-23 23:21:22 - Assemble contigs from SdBG for k = 99\n",
      "2023-02-23 23:21:36 - Local assembly for k = 99\n",
      "2023-02-23 23:23:37 - Extract iterative edges from k = 99 to 119 \n",
      "2023-02-23 23:23:59 - Build graph for k = 119 \n",
      "2023-02-23 23:24:08 - Assemble contigs from SdBG for k = 119\n",
      "2023-02-23 23:24:21 - Local assembly for k = 119\n",
      "2023-02-23 23:26:44 - Extract iterative edges from k = 119 to 141 \n",
      "2023-02-23 23:27:01 - Build graph for k = 141 \n",
      "2023-02-23 23:27:11 - Assemble contigs from SdBG for k = 141\n",
      "2023-02-23 23:27:22 - Merging to output final contigs \n",
      "2023-02-23 23:27:23 - 23833 contigs, total 31325615 bp, min 200 bp, max 33978 bp, avg 1314 bp, N50 3904 bp\n",
      "2023-02-23 23:27:24 - ALL DONE. Time elapsed: 2906.525506 seconds \n",
      "QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-jovyan'\n",
      "QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-jovyan'\n",
      "[bwa_index] Pack FASTA... 0.28 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=62651230, availableWord=16408252\n",
      "[BWTIncConstructFromPacked] 10 iterations done. 27065438 characters processed.\n",
      "[BWTIncConstructFromPacked] 20 iterations done. 49999342 characters processed.\n",
      "[bwt_gen] Finished constructing BWT in 27 iterations.\n",
      "[bwa_index] 11.82 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.32 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.19 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 5.83 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399461/megahit/final.contigs.fastg.gfa.fna\n",
      "[main] Real time: 22.415 sec; CPU: 18.450 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 3321764 sequences (480000077 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (20344, 437350, 5214, 19435)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (118, 311, 757)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2035)\n",
      "[M::mem_pestat] mean and std.dev: (458.01, 451.01)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2674)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (258, 309, 356)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (62, 552)\n",
      "[M::mem_pestat] mean and std.dev: (306.27, 73.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 650)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (141, 305, 1046)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2856)\n",
      "[M::mem_pestat] mean and std.dev: (577.76, 614.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3761)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (126, 340, 785)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2103)\n",
      "[M::mem_pestat] mean and std.dev: (482.44, 466.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2762)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3321764 reads in 1415.544 CPU sec, 37.366 real sec\n",
      "[M::process] read 3244816 sequences (480000095 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (23513, 634611, 5186, 22547)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (219, 494, 1019)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2619)\n",
      "[M::mem_pestat] mean and std.dev: (646.49, 563.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3419)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (302, 338, 376)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (154, 524)\n",
      "[M::mem_pestat] mean and std.dev: (339.05, 56.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (80, 598)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (517, 1124, 1968)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4870)\n",
      "[M::mem_pestat] mean and std.dev: (1227.18, 1053.31)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6321)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (223, 512, 1029)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2641)\n",
      "[M::mem_pestat] mean and std.dev: (660.88, 569.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3447)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::process] read 3229908 sequences (480000048 bp)...\n",
      "[M::mem_process_seqs] Processed 3244816 reads in 1136.709 CPU sec, 25.513 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (24062, 679883, 5556, 23693)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 515, 1047)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2671)\n",
      "[M::mem_pestat] mean and std.dev: (672.62, 570.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3483)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (343.23, 53.87)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (521, 1118, 1978)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4892)\n",
      "[M::mem_pestat] mean and std.dev: (1219.47, 1050.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6349)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 528, 1045)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2653)\n",
      "[M::mem_pestat] mean and std.dev: (675.69, 564.28)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3457)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3229908 reads in 1112.504 CPU sec, 25.074 real sec\n",
      "[M::process] read 3235548 sequences (480000029 bp)...\n",
      "[M::process] read 3237368 sequences (480000068 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (24703, 683759, 5650, 23955)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 518, 1040)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2648)\n",
      "[M::mem_pestat] mean and std.dev: (671.32, 566.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3452)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (343.26, 53.96)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (542, 1124, 2078)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5150)\n",
      "[M::mem_pestat] mean and std.dev: (1480.57, 1365.77)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6944)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 532, 1052)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2674)\n",
      "[M::mem_pestat] mean and std.dev: (681.35, 571.03)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3485)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3235548 reads in 1095.978 CPU sec, 24.450 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (24956, 685055, 5635, 24467)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 525, 1055)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2691)\n",
      "[M::mem_pestat] mean and std.dev: (677.25, 571.86)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3509)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (343.14, 53.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (507, 1123, 2103)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5295)\n",
      "[M::mem_pestat] mean and std.dev: (1499.04, 1401.28)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7104)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (240, 531, 1067)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2721)\n",
      "[M::mem_pestat] mean and std.dev: (688.56, 581.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3548)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3237368 reads in 1110.535 CPU sec, 25.187 real sec\n",
      "[M::process] read 3239074 sequences (480000049 bp)...\n",
      "[M::process] read 3242844 sequences (480000129 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (25224, 688772, 5738, 24433)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 518, 1063)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2723)\n",
      "[M::mem_pestat] mean and std.dev: (681.12, 583.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3553)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (168, 518)\n",
      "[M::mem_pestat] mean and std.dev: (343.64, 53.41)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (98, 588)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (513, 1114, 2013)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5013)\n",
      "[M::mem_pestat] mean and std.dev: (1376.01, 1262.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6513)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (238, 527, 1053)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2683)\n",
      "[M::mem_pestat] mean and std.dev: (679.49, 571.82)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3498)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3239074 reads in 1099.985 CPU sec, 24.406 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (25300, 690871, 5687, 24648)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 513, 1034)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2642)\n",
      "[M::mem_pestat] mean and std.dev: (665.04, 564.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3446)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (343.03, 54.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (505, 1121, 2076)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5218)\n",
      "[M::mem_pestat] mean and std.dev: (1471.51, 1373.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6967)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 530, 1038)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2636)\n",
      "[M::mem_pestat] mean and std.dev: (674.36, 565.25)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3435)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3242844 reads in 1112.667 CPU sec, 25.201 real sec\n",
      "[M::process] read 3247862 sequences (480000128 bp)...\n",
      "[M::process] read 3253510 sequences (480000039 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (25622, 692323, 5696, 24960)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 512, 1042)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2658)\n",
      "[M::mem_pestat] mean and std.dev: (668.26, 569.73)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3466)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (343.03, 53.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (501, 1121, 2068)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5202)\n",
      "[M::mem_pestat] mean and std.dev: (1473.60, 1386.33)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7019)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 531, 1056)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2690)\n",
      "[M::mem_pestat] mean and std.dev: (685.08, 576.67)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3507)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3247862 reads in 1103.854 CPU sec, 24.952 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (25870, 692605, 5672, 25060)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (232, 515, 1030)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2626)\n",
      "[M::mem_pestat] mean and std.dev: (661.79, 559.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3424)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 342, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (343.50, 53.61)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (506, 1127, 2127)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5369)\n",
      "[M::mem_pestat] mean and std.dev: (1537.78, 1438.02)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7290)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 522, 1052)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2682)\n",
      "[M::mem_pestat] mean and std.dev: (676.58, 574.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3497)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3253510 reads in 1118.208 CPU sec, 25.364 real sec\n",
      "[M::process] read 3257912 sequences (480000172 bp)...\n",
      "[M::process] read 3265164 sequences (480000266 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26239, 691660, 5882, 25264)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 504, 1033)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2639)\n",
      "[M::mem_pestat] mean and std.dev: (663.23, 570.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3442)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (168, 518)\n",
      "[M::mem_pestat] mean and std.dev: (343.59, 53.58)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (98, 588)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (509, 1117, 2104)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5294)\n",
      "[M::mem_pestat] mean and std.dev: (1503.28, 1405.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7126)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (240, 527, 1073)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2739)\n",
      "[M::mem_pestat] mean and std.dev: (691.12, 587.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3572)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3257912 reads in 1121.441 CPU sec, 25.470 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26241, 691872, 6032, 25574)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 505, 1039)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2655)\n",
      "[M::mem_pestat] mean and std.dev: (667.22, 569.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3463)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (309, 342, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (169, 519)\n",
      "[M::mem_pestat] mean and std.dev: (344.46, 52.81)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (99, 589)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (512, 1125, 2221)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5639)\n",
      "[M::mem_pestat] mean and std.dev: (1575.02, 1475.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7477)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 538, 1053)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2677)\n",
      "[M::mem_pestat] mean and std.dev: (679.57, 568.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3489)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3265164 reads in 1128.967 CPU sec, 25.371 real sec\n",
      "[M::process] read 3267542 sequences (480000251 bp)...\n",
      "[M::process] read 3234038 sequences (480000262 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (25901, 692288, 5808, 25285)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 508, 1032)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2628)\n",
      "[M::mem_pestat] mean and std.dev: (660.70, 558.91)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3426)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (342.74, 54.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (490, 1115, 2065)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5215)\n",
      "[M::mem_pestat] mean and std.dev: (1482.34, 1399.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7079)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 522, 1057)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2705)\n",
      "[M::mem_pestat] mean and std.dev: (679.71, 579.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3529)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3267542 reads in 1128.225 CPU sec, 25.500 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26299, 703575, 5821, 25814)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 524, 1060)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2702)\n",
      "[M::mem_pestat] mean and std.dev: (682.08, 579.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3523)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (166, 521)\n",
      "[M::mem_pestat] mean and std.dev: (344.39, 53.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (95, 592)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (490, 1126, 2182)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5566)\n",
      "[M::mem_pestat] mean and std.dev: (1560.52, 1472.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7450)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (243, 536, 1051)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2667)\n",
      "[M::mem_pestat] mean and std.dev: (683.25, 567.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3475)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3234038 reads in 1087.691 CPU sec, 24.644 real sec\n",
      "[M::process] read 3239188 sequences (480000200 bp)...\n",
      "[M::process] read 3241278 sequences (480000189 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26549, 703677, 5757, 25881)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 522, 1049)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2673)\n",
      "[M::mem_pestat] mean and std.dev: (677.57, 571.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3485)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (166, 521)\n",
      "[M::mem_pestat] mean and std.dev: (343.80, 53.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (95, 592)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (496, 1120, 2057)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5179)\n",
      "[M::mem_pestat] mean and std.dev: (1467.01, 1370.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6948)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 533, 1059)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2695)\n",
      "[M::mem_pestat] mean and std.dev: (683.51, 574.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3513)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3239188 reads in 1079.645 CPU sec, 24.109 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26657, 703913, 5858, 25999)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 518, 1039)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2649)\n",
      "[M::mem_pestat] mean and std.dev: (668.87, 565.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3454)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 342, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 523)\n",
      "[M::mem_pestat] mean and std.dev: (343.81, 54.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (91, 595)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (485, 1119, 2154)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5492)\n",
      "[M::mem_pestat] mean and std.dev: (1534.87, 1453.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7350)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (244, 532, 1062)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2698)\n",
      "[M::mem_pestat] mean and std.dev: (686.08, 574.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3516)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3241278 reads in 1094.763 CPU sec, 24.799 real sec\n",
      "[M::process] read 3244402 sequences (480000171 bp)...\n",
      "[M::process] read 3248860 sequences (480000169 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26574, 703443, 5969, 25594)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 517, 1036)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2636)\n",
      "[M::mem_pestat] mean and std.dev: (666.00, 562.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3436)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (342.98, 54.34)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (505, 1115, 2114)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5332)\n",
      "[M::mem_pestat] mean and std.dev: (1507.50, 1412.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7156)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 536, 1068)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2722)\n",
      "[M::mem_pestat] mean and std.dev: (691.54, 582.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3549)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3244402 reads in 1091.015 CPU sec, 24.554 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26643, 704620, 6014, 25735)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (232, 510, 1034)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2638)\n",
      "[M::mem_pestat] mean and std.dev: (661.97, 563.04)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3440)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (306, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (162, 522)\n",
      "[M::mem_pestat] mean and std.dev: (342.78, 54.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (90, 594)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (520, 1114, 2131)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5353)\n",
      "[M::mem_pestat] mean and std.dev: (1516.42, 1412.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7166)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (242, 535, 1070)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2726)\n",
      "[M::mem_pestat] mean and std.dev: (692.99, 584.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3554)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3248860 reads in 1107.497 CPU sec, 25.069 real sec\n",
      "[M::process] read 3251344 sequences (480000033 bp)...\n",
      "[M::process] read 3255378 sequences (480000104 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27038, 705471, 6028, 26218)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 506, 1033)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2639)\n",
      "[M::mem_pestat] mean and std.dev: (662.69, 566.21)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3442)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 523)\n",
      "[M::mem_pestat] mean and std.dev: (343.57, 54.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (91, 595)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (506, 1129, 2181)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5531)\n",
      "[M::mem_pestat] mean and std.dev: (1553.51, 1451.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7360)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 524, 1045)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2657)\n",
      "[M::mem_pestat] mean and std.dev: (676.09, 567.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3463)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3251344 reads in 1098.861 CPU sec, 24.595 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27424, 704393, 5935, 26175)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (232, 515, 1039)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2653)\n",
      "[M::mem_pestat] mean and std.dev: (668.42, 567.52)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3460)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (306, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (162, 522)\n",
      "[M::mem_pestat] mean and std.dev: (342.56, 54.67)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (90, 594)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (512, 1133, 2261)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5759)\n",
      "[M::mem_pestat] mean and std.dev: (1623.72, 1525.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7725)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 519, 1042)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2658)\n",
      "[M::mem_pestat] mean and std.dev: (672.50, 568.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3466)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3255378 reads in 1108.247 CPU sec, 25.158 real sec\n",
      "[M::process] read 3267908 sequences (480000248 bp)...\n",
      "[M::process] read 3273680 sequences (480000109 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27107, 698884, 5756, 26077)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (232, 510, 1050)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2686)\n",
      "[M::mem_pestat] mean and std.dev: (674.32, 578.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3504)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (306, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (162, 522)\n",
      "[M::mem_pestat] mean and std.dev: (342.70, 54.69)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (90, 594)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (516, 1126, 2177)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5499)\n",
      "[M::mem_pestat] mean and std.dev: (1568.25, 1464.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7428)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 520, 1055)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2699)\n",
      "[M::mem_pestat] mean and std.dev: (680.55, 579.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3521)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3267908 reads in 1106.346 CPU sec, 24.845 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26861, 700031, 6046, 26010)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (232, 515, 1054)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2698)\n",
      "[M::mem_pestat] mean and std.dev: (676.51, 578.67)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3520)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (306, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (162, 522)\n",
      "[M::mem_pestat] mean and std.dev: (342.53, 54.57)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (90, 594)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (494, 1122, 2175)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5537)\n",
      "[M::mem_pestat] mean and std.dev: (1562.77, 1488.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7516)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 514, 1030)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2622)\n",
      "[M::mem_pestat] mean and std.dev: (663.81, 560.26)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3418)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3273680 reads in 1119.237 CPU sec, 25.129 real sec\n",
      "[M::process] read 3279124 sequences (480000044 bp)...\n",
      "[M::process] read 3229560 sequences (480000104 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27383, 696832, 5982, 26162)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 509, 1032)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2636)\n",
      "[M::mem_pestat] mean and std.dev: (662.10, 564.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3438)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (343.51, 53.58)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (491, 1118, 2053)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5177)\n",
      "[M::mem_pestat] mean and std.dev: (1484.99, 1405.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7109)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 525, 1055)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2697)\n",
      "[M::mem_pestat] mean and std.dev: (681.20, 577.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3518)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3279124 reads in 1133.253 CPU sec, 25.660 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27384, 711512, 6080, 27006)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 520, 1043)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2651)\n",
      "[M::mem_pestat] mean and std.dev: (672.03, 565.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3455)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 342, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 526)\n",
      "[M::mem_pestat] mean and std.dev: (344.19, 55.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (88, 599)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (484, 1117, 2134)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5434)\n",
      "[M::mem_pestat] mean and std.dev: (1539.35, 1462.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7391)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (243, 531, 1055)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2679)\n",
      "[M::mem_pestat] mean and std.dev: (685.30, 573.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3491)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3229560 reads in 1073.060 CPU sec, 24.310 real sec\n",
      "[M::process] read 3230388 sequences (480000016 bp)...\n",
      "[M::process] read 3232096 sequences (480000049 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27874, 715109, 5975, 27092)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 512, 1037)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2645)\n",
      "[M::mem_pestat] mean and std.dev: (668.94, 569.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3449)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 342, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 526)\n",
      "[M::mem_pestat] mean and std.dev: (344.06, 55.41)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (88, 599)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (506, 1110, 2142)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5414)\n",
      "[M::mem_pestat] mean and std.dev: (1552.80, 1471.41)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7438)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (247, 540, 1066)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2704)\n",
      "[M::mem_pestat] mean and std.dev: (690.13, 576.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3523)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3230388 reads in 1048.450 CPU sec, 23.581 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (28640, 714724, 6116, 27465)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 517, 1049)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2673)\n",
      "[M::mem_pestat] mean and std.dev: (673.73, 569.72)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3485)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 343, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (164, 524)\n",
      "[M::mem_pestat] mean and std.dev: (344.75, 54.96)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (92, 596)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (509, 1116, 2141)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5405)\n",
      "[M::mem_pestat] mean and std.dev: (1543.78, 1445.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7324)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 526, 1054)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2680)\n",
      "[M::mem_pestat] mean and std.dev: (681.01, 572.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3493)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3232096 reads in 1058.069 CPU sec, 23.886 real sec\n",
      "[M::process] read 3237054 sequences (480000259 bp)...\n",
      "[M::process] read 3241796 sequences (480000249 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (28146, 714047, 6173, 26957)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (238, 516, 1039)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2641)\n",
      "[M::mem_pestat] mean and std.dev: (671.18, 565.07)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3442)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 343, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (164, 524)\n",
      "[M::mem_pestat] mean and std.dev: (344.74, 54.81)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (92, 596)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (464, 1103, 2148)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5516)\n",
      "[M::mem_pestat] mean and std.dev: (1524.15, 1459.21)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7361)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (242, 537, 1063)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2705)\n",
      "[M::mem_pestat] mean and std.dev: (687.75, 578.28)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3526)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3237054 reads in 1062.599 CPU sec, 23.657 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (28092, 713463, 6112, 27351)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 516, 1042)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2652)\n",
      "[M::mem_pestat] mean and std.dev: (669.68, 565.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3457)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (164, 524)\n",
      "[M::mem_pestat] mean and std.dev: (344.63, 54.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (92, 596)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (483, 1118, 2181)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5577)\n",
      "[M::mem_pestat] mean and std.dev: (1540.39, 1452.62)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7351)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 525, 1051)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2681)\n",
      "[M::mem_pestat] mean and std.dev: (677.52, 571.74)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3496)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3241796 reads in 1083.479 CPU sec, 24.588 real sec\n",
      "[M::process] read 3245728 sequences (480000187 bp)...\n",
      "[M::process] read 3256962 sequences (480000003 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (28281, 710462, 6070, 27307)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 517, 1052)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2684)\n",
      "[M::mem_pestat] mean and std.dev: (676.50, 576.98)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3500)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (164, 524)\n",
      "[M::mem_pestat] mean and std.dev: (344.58, 54.79)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (92, 596)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (486, 1107, 2067)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5229)\n",
      "[M::mem_pestat] mean and std.dev: (1458.80, 1367.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6929)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 531, 1065)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2713)\n",
      "[M::mem_pestat] mean and std.dev: (688.77, 582.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3537)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3245728 reads in 1085.185 CPU sec, 24.214 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (28184, 708620, 5987, 26977)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 502, 1044)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2666)\n",
      "[M::mem_pestat] mean and std.dev: (668.92, 572.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3477)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 342, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 526)\n",
      "[M::mem_pestat] mean and std.dev: (344.15, 55.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (88, 599)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (534, 1131, 2199)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5529)\n",
      "[M::mem_pestat] mean and std.dev: (1565.72, 1435.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7308)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (244, 539, 1065)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2707)\n",
      "[M::mem_pestat] mean and std.dev: (687.63, 577.03)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3528)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3256962 reads in 1100.016 CPU sec, 24.767 real sec\n",
      "[M::process] read 3257490 sequences (480000176 bp)...\n",
      "[M::process] read 3265632 sequences (480000002 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26991, 692244, 6043, 25598)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 508, 1048)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2678)\n",
      "[M::mem_pestat] mean and std.dev: (667.84, 568.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3493)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 342, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 523)\n",
      "[M::mem_pestat] mean and std.dev: (343.45, 54.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (91, 595)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (509, 1117, 2061)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5165)\n",
      "[M::mem_pestat] mean and std.dev: (1462.35, 1362.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6912)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 520, 1058)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2706)\n",
      "[M::mem_pestat] mean and std.dev: (682.31, 584.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3530)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3257490 reads in 1162.772 CPU sec, 25.803 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (28092, 703949, 6022, 27098)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (228, 503, 1030)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2634)\n",
      "[M::mem_pestat] mean and std.dev: (658.95, 562.62)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3436)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 342, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 526)\n",
      "[M::mem_pestat] mean and std.dev: (343.95, 55.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (88, 599)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (504, 1117, 2172)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5508)\n",
      "[M::mem_pestat] mean and std.dev: (1550.24, 1445.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7330)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 522, 1047)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2669)\n",
      "[M::mem_pestat] mean and std.dev: (678.17, 573.81)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3480)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3265632 reads in 1131.193 CPU sec, 36.040 real sec\n",
      "[M::process] read 3273238 sequences (480000117 bp)...\n",
      "[M::process] read 3236778 sequences (480000117 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27970, 707977, 6050, 26957)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 504, 1038)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2648)\n",
      "[M::mem_pestat] mean and std.dev: (662.55, 565.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3453)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 342, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 526)\n",
      "[M::mem_pestat] mean and std.dev: (344.15, 54.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (88, 599)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (504, 1117, 2136)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5400)\n",
      "[M::mem_pestat] mean and std.dev: (1538.19, 1441.03)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7302)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 527, 1054)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2688)\n",
      "[M::mem_pestat] mean and std.dev: (679.18, 573.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3505)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3273238 reads in 1098.792 CPU sec, 24.681 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27305, 707651, 6028, 26881)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 525, 1054)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2684)\n",
      "[M::mem_pestat] mean and std.dev: (681.22, 575.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3499)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 380)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (164, 524)\n",
      "[M::mem_pestat] mean and std.dev: (344.19, 54.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (92, 596)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (480, 1120, 2096)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5328)\n",
      "[M::mem_pestat] mean and std.dev: (1491.26, 1412.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7143)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (240, 527, 1055)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2685)\n",
      "[M::mem_pestat] mean and std.dev: (682.43, 572.90)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3500)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3236778 reads in 1098.859 CPU sec, 24.819 real sec\n",
      "[M::process] read 3238876 sequences (480000279 bp)...\n",
      "[M::process] read 3242250 sequences (480000010 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26655, 697140, 6142, 25596)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 518, 1051)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2683)\n",
      "[M::mem_pestat] mean and std.dev: (675.83, 573.38)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3499)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (166, 521)\n",
      "[M::mem_pestat] mean and std.dev: (344.61, 53.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (95, 592)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (497, 1113, 2073)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5225)\n",
      "[M::mem_pestat] mean and std.dev: (1469.78, 1380.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6991)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (243, 530, 1052)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2670)\n",
      "[M::mem_pestat] mean and std.dev: (680.76, 570.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3479)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3238876 reads in 1121.160 CPU sec, 24.925 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (25381, 683392, 5786, 25111)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 525, 1047)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2667)\n",
      "[M::mem_pestat] mean and std.dev: (677.23, 572.51)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3477)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (343.03, 54.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (525, 1136, 2131)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5343)\n",
      "[M::mem_pestat] mean and std.dev: (1527.55, 1401.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7136)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 533, 1065)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2717)\n",
      "[M::mem_pestat] mean and std.dev: (688.60, 578.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3543)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3242250 reads in 1188.489 CPU sec, 26.864 real sec\n",
      "[M::process] read 3244024 sequences (480000020 bp)...\n",
      "[M::process] read 3249786 sequences (480000120 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (25594, 685468, 5853, 24537)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 518, 1045)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2667)\n",
      "[M::mem_pestat] mean and std.dev: (672.85, 572.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3478)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (342.96, 54.14)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (501, 1120, 2064)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5190)\n",
      "[M::mem_pestat] mean and std.dev: (1466.82, 1372.04)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6955)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (240, 531, 1056)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2688)\n",
      "[M::mem_pestat] mean and std.dev: (684.86, 576.20)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3504)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3244024 reads in 1287.475 CPU sec, 28.565 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27124, 701044, 5979, 26316)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 520, 1058)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2696)\n",
      "[M::mem_pestat] mean and std.dev: (680.41, 578.57)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3515)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 523)\n",
      "[M::mem_pestat] mean and std.dev: (343.30, 54.85)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (91, 595)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (504, 1106, 1973)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4911)\n",
      "[M::mem_pestat] mean and std.dev: (1216.45, 1048.02)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6380)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 520, 1040)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2646)\n",
      "[M::mem_pestat] mean and std.dev: (673.83, 567.57)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3449)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3249786 reads in 1141.311 CPU sec, 25.694 real sec\n",
      "[M::process] read 3250034 sequences (480000067 bp)...\n",
      "[M::process] read 3259362 sequences (480000101 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (25373, 681350, 5991, 24904)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 519, 1055)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2691)\n",
      "[M::mem_pestat] mean and std.dev: (678.41, 575.59)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3509)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (342.94, 53.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (505, 1120, 2022)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5056)\n",
      "[M::mem_pestat] mean and std.dev: (1440.76, 1325.67)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6743)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (240, 527, 1048)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2664)\n",
      "[M::mem_pestat] mean and std.dev: (676.00, 569.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3472)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3250034 reads in 1196.639 CPU sec, 26.819 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27527, 699250, 5962, 26623)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (234, 517, 1049)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2679)\n",
      "[M::mem_pestat] mean and std.dev: (673.97, 573.59)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3494)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (166, 521)\n",
      "[M::mem_pestat] mean and std.dev: (344.06, 54.21)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (95, 592)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (502, 1119, 2089)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5263)\n",
      "[M::mem_pestat] mean and std.dev: (1475.65, 1389.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7033)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (242, 535, 1070)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2726)\n",
      "[M::mem_pestat] mean and std.dev: (691.99, 580.73)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3554)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3259362 reads in 1120.077 CPU sec, 25.169 real sec\n",
      "[M::process] read 3278274 sequences (480000210 bp)...\n",
      "[M::process] read 3281322 sequences (480000068 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (27501, 694345, 5892, 26247)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 520, 1049)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2675)\n",
      "[M::mem_pestat] mean and std.dev: (674.73, 574.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3488)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (308, 342, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (166, 521)\n",
      "[M::mem_pestat] mean and std.dev: (344.15, 54.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (95, 592)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (513, 1123, 2123)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5343)\n",
      "[M::mem_pestat] mean and std.dev: (1530.70, 1429.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7250)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (238, 524, 1050)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2674)\n",
      "[M::mem_pestat] mean and std.dev: (679.91, 576.25)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3486)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3278274 reads in 1128.366 CPU sec, 25.444 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (26455, 686415, 5940, 25510)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 503, 1029)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2627)\n",
      "[M::mem_pestat] mean and std.dev: (660.55, 561.09)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3426)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 378)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (165, 520)\n",
      "[M::mem_pestat] mean and std.dev: (343.16, 54.13)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (94, 591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (520, 1130, 2206)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5578)\n",
      "[M::mem_pestat] mean and std.dev: (1582.57, 1488.51)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7537)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 526, 1050)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2676)\n",
      "[M::mem_pestat] mean and std.dev: (676.41, 569.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3489)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3281322 reads in 1186.622 CPU sec, 26.716 real sec\n",
      "[M::process] read 2834526 sequences (414198377 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (23987, 601742, 5228, 22848)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 497, 1030)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2630)\n",
      "[M::mem_pestat] mean and std.dev: (658.06, 563.54)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3430)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (307, 341, 379)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 523)\n",
      "[M::mem_pestat] mean and std.dev: (343.21, 54.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (91, 595)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (507, 1129, 2141)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 5409)\n",
      "[M::mem_pestat] mean and std.dev: (1552.69, 1452.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 7364)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 517, 1049)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2681)\n",
      "[M::mem_pestat] mean and std.dev: (674.12, 572.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3497)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 2834526 reads in 982.277 CPU sec, 21.888 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -R @RG\\tID:SRR6399461\\tSM:bar -t 48 /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399461/megahit/final.contigs.fastg.gfa.fna /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399461/trim_galore/SRR6399461_1_val_1.fq.gz /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399461/trim_galore/SRR6399461_2_val_2.fq.gz\n",
      "[main] Real time: 1325.139 sec; CPU: 47180.204 sec\n",
      "[bam_sort_core] merging from 48 files and 48 in-memory blocks...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Java memory size is set to 4G\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.2.2-dev\n",
      "Built on 2019-11-11 14:05\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 37\n",
      "Max memory (Mb): 4294\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 24232\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 48\n",
      "Processed 2423 out of 24232 windows...\n",
      "Processed 4846 out of 24232 windows...\n",
      "Processed 7269 out of 24232 windows...\n",
      "Processed 9692 out of 24232 windows...\n",
      "Processed 12115 out of 24232 windows...\n",
      "Processed 14538 out of 24232 windows...\n",
      "Processed 16961 out of 24232 windows...\n",
      "Processed 19384 out of 24232 windows...\n",
      "Processed 21807 out of 24232 windows...\n",
      "Processed 24230 out of 24232 windows...\n",
      "Total processed windows:24232\n",
      "Number of reads: 136163688\n",
      "Number of valid reads: 139883903\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 132583918\n",
      "Num mapped first of pair: 66447504\n",
      "Num mapped second of pair: 66136414\n",
      "Num singletons: 1430636\n",
      "Time taken to analyze reads: 2337\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 19314636016\n",
      "referenceSize: 31325615\n",
      "numberOfSequencedBases: 19311169119\n",
      "numberOfAs: 5016396740\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 2376\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing PDF:HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "PDF file created successfully \n",
      "\n",
      "Finished\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:   9%|███▉                                     |  ETA: 23:41:23\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR = \"SRR6399462\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-02-24 01:05:56 - MEGAHIT v1.2.9\n",
      "2023-02-24 01:05:56 - Using megahit_core with POPCNT and BMI2 support\n",
      "2023-02-24 01:05:56 - Convert reads to binary library\n",
      "2023-02-24 01:10:17 - b'INFO  sequence/io/sequence_lib.cpp  :   75 - Lib 0 (/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/trim_galore/SRR6399462_1_val_1.fq.gz,/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/trim_galore/SRR6399462_2_val_2.fq.gz): pe, 124896854 reads, 151 max length'\n",
      "2023-02-24 01:10:21 - b'INFO  utils/utils.h                 :  152 - Real: 264.8290\\tuser: 140.5360\\tsys: 12.6366\\tmaxrss: 254076'\n",
      "2023-02-24 01:10:21 - k-max reset to: 141 \n",
      "2023-02-24 01:10:21 - Start assembly. Number of CPU threads 48 \n",
      "2023-02-24 01:10:21 - k list: 21,29,39,59,79,99,119,141 \n",
      "2023-02-24 01:10:21 - Memory used: 121546591027\n",
      "2023-02-24 01:10:21 - Extract solid (k+1)-mers for k = 21 \n",
      "2023-02-24 01:19:48 - Build graph for k = 21 \n",
      "2023-02-24 01:20:49 - Assemble contigs from SdBG for k = 21\n",
      "2023-02-24 01:25:27 - Local assembly for k = 21\n",
      "2023-02-24 01:26:07 - Extract iterative edges from k = 21 to 29 \n",
      "2023-02-24 01:29:18 - Build graph for k = 29 \n",
      "2023-02-24 01:29:32 - Assemble contigs from SdBG for k = 29\n",
      "2023-02-24 01:29:44 - Local assembly for k = 29\n",
      "2023-02-24 01:31:20 - Extract iterative edges from k = 29 to 39 \n",
      "2023-02-24 01:31:47 - Build graph for k = 39 \n",
      "2023-02-24 01:31:51 - Assemble contigs from SdBG for k = 39\n",
      "2023-02-24 01:31:53 - Local assembly for k = 39\n",
      "2023-02-24 01:32:31 - Extract iterative edges from k = 39 to 59 \n",
      "2023-02-24 01:32:56 - Build graph for k = 59 \n",
      "2023-02-24 01:33:02 - Assemble contigs from SdBG for k = 59\n",
      "2023-02-24 01:33:04 - Local assembly for k = 59\n",
      "2023-02-24 01:33:48 - Extract iterative edges from k = 59 to 79 \n",
      "2023-02-24 01:34:10 - Build graph for k = 79 \n",
      "2023-02-24 01:34:15 - Assemble contigs from SdBG for k = 79\n",
      "2023-02-24 01:34:16 - Local assembly for k = 79\n",
      "2023-02-24 01:35:11 - Extract iterative edges from k = 79 to 99 \n",
      "2023-02-24 01:35:46 - Build graph for k = 99 \n",
      "2023-02-24 01:35:51 - Assemble contigs from SdBG for k = 99\n",
      "2023-02-24 01:35:53 - Local assembly for k = 99\n",
      "2023-02-24 01:36:55 - Extract iterative edges from k = 99 to 119 \n",
      "2023-02-24 01:37:20 - Build graph for k = 119 \n",
      "2023-02-24 01:37:23 - Assemble contigs from SdBG for k = 119\n",
      "2023-02-24 01:37:24 - Local assembly for k = 119\n",
      "2023-02-24 01:38:44 - Extract iterative edges from k = 119 to 141 \n",
      "2023-02-24 01:39:00 - Build graph for k = 141 \n",
      "2023-02-24 01:39:03 - Assemble contigs from SdBG for k = 141\n",
      "2023-02-24 01:39:05 - Merging to output final contigs \n",
      "2023-02-24 01:39:05 - 6084 contigs, total 4459785 bp, min 200 bp, max 20612 bp, avg 733 bp, N50 1548 bp\n",
      "2023-02-24 01:39:06 - ALL DONE. Time elapsed: 1990.117793 seconds \n",
      "QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-jovyan'\n",
      "QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-jovyan'\n",
      "[bwa_index] Pack FASTA... 0.04 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 1.86 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.05 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.03 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.25 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/megahit/final.contigs.fastg.gfa.fna\n",
      "[main] Real time: 3.121 sec; CPU: 2.227 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 3259988 sequences (480000273 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7538, 160436, 1786, 7488)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (164, 319, 772)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1988)\n",
      "[M::mem_pestat] mean and std.dev: (489.85, 448.03)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2596)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (282, 348, 395)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (56, 621)\n",
      "[M::mem_pestat] mean and std.dev: (340.13, 82.54)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 734)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (245, 730, 1095)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2795)\n",
      "[M::mem_pestat] mean and std.dev: (745.96, 565.69)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3645)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (168, 341, 781)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2007)\n",
      "[M::mem_pestat] mean and std.dev: (502.05, 447.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2620)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::process] read 3259412 sequences (480000000 bp)...\n",
      "[M::mem_process_seqs] Processed 3259988 reads in 1585.244 CPU sec, 42.210 real sec\n",
      "[M::process] read 3360512 sequences (480000162 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7983, 138757, 1059, 7834)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (101, 178, 343)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 827)\n",
      "[M::mem_pestat] mean and std.dev: (215.87, 166.23)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1069)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (248, 309, 376)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 632)\n",
      "[M::mem_pestat] mean and std.dev: (305.09, 96.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 760)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (129, 505, 871)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2355)\n",
      "[M::mem_pestat] mean and std.dev: (573.54, 494.86)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3097)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (105, 179, 350)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 840)\n",
      "[M::mem_pestat] mean and std.dev: (221.91, 169.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1085)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 3259412 reads in 1909.662 CPU sec, 42.695 real sec\n",
      "[M::process] read 3222952 sequences (480000117 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (10928, 179902, 5092, 10642)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (194, 458, 1124)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2984)\n",
      "[M::mem_pestat] mean and std.dev: (611.09, 477.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3914)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (259, 332, 389)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 649)\n",
      "[M::mem_pestat] mean and std.dev: (318.35, 98.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 779)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (700, 1076, 1143)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2029)\n",
      "[M::mem_pestat] mean and std.dev: (883.97, 395.79)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 2472)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (198, 482, 1132)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3000)\n",
      "[M::mem_pestat] mean and std.dev: (621.95, 482.91)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3934)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 3360512 reads in 2078.221 CPU sec, 46.967 real sec\n",
      "[M::process] read 3227254 sequences (480000263 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (5635, 131825, 1748, 5591)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (240, 499, 966)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2418)\n",
      "[M::mem_pestat] mean and std.dev: (632.26, 510.89)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3144)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 366, 407)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (170, 565)\n",
      "[M::mem_pestat] mean and std.dev: (368.66, 59.04)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (91, 644)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (58, 507, 1409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4111)\n",
      "[M::mem_pestat] mean and std.dev: (786.17, 884.85)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5462)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (248, 502, 999)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2501)\n",
      "[M::mem_pestat] mean and std.dev: (653.60, 523.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3252)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3222952 reads in 2210.639 CPU sec, 49.951 real sec\n",
      "[M::process] read 3231468 sequences (480000210 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6116, 140999, 1843, 5945)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 478, 969)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2433)\n",
      "[M::mem_pestat] mean and std.dev: (619.90, 501.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3165)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 366, 407)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (170, 565)\n",
      "[M::mem_pestat] mean and std.dev: (368.63, 59.09)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (91, 644)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (63, 531, 1600)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4674)\n",
      "[M::mem_pestat] mean and std.dev: (880.04, 964.01)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6211)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (244, 503, 970)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2422)\n",
      "[M::mem_pestat] mean and std.dev: (631.14, 504.31)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3148)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3227254 reads in 2163.051 CPU sec, 48.417 real sec\n",
      "[M::process] read 3232640 sequences (480000263 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6321, 144677, 1917, 6235)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 496, 961)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2411)\n",
      "[M::mem_pestat] mean and std.dev: (627.15, 502.25)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3136)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 366, 408)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (168, 568)\n",
      "[M::mem_pestat] mean and std.dev: (368.80, 59.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (88, 648)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (62, 548, 1527)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4457)\n",
      "[M::mem_pestat] mean and std.dev: (847.95, 919.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5922)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 498, 966)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2426)\n",
      "[M::mem_pestat] mean and std.dev: (625.85, 495.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3156)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3231468 reads in 2155.228 CPU sec, 48.796 real sec\n",
      "[M::process] read 3236778 sequences (480000109 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6662, 148112, 1975, 6537)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 493, 968)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2430)\n",
      "[M::mem_pestat] mean and std.dev: (623.61, 502.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3161)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 367, 408)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (168, 568)\n",
      "[M::mem_pestat] mean and std.dev: (369.43, 59.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (88, 648)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (61, 533, 1483)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4327)\n",
      "[M::mem_pestat] mean and std.dev: (805.53, 890.65)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5749)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 510, 976)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2458)\n",
      "[M::mem_pestat] mean and std.dev: (635.40, 510.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3199)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3232640 reads in 2143.326 CPU sec, 48.441 real sec\n",
      "[M::process] read 3238698 sequences (480000073 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6737, 148739, 2032, 6897)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (232, 494, 953)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2395)\n",
      "[M::mem_pestat] mean and std.dev: (615.84, 494.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3116)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 366, 408)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (168, 568)\n",
      "[M::mem_pestat] mean and std.dev: (369.07, 59.80)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (88, 648)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (61, 475, 1581)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4621)\n",
      "[M::mem_pestat] mean and std.dev: (834.39, 936.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6141)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 502, 963)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2407)\n",
      "[M::mem_pestat] mean and std.dev: (621.05, 492.01)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3129)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3236778 reads in 2152.256 CPU sec, 48.713 real sec\n",
      "[M::process] read 3243830 sequences (480000268 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6918, 150558, 1986, 6870)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (228, 480, 942)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2370)\n",
      "[M::mem_pestat] mean and std.dev: (609.91, 491.74)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3084)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 367, 408)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (168, 568)\n",
      "[M::mem_pestat] mean and std.dev: (369.32, 59.48)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (88, 648)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (68, 594, 1547)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4505)\n",
      "[M::mem_pestat] mean and std.dev: (892.67, 944.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5984)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (242, 486, 949)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2363)\n",
      "[M::mem_pestat] mean and std.dev: (619.24, 493.75)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3070)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3238698 reads in 2164.610 CPU sec, 48.704 real sec\n",
      "[M::process] read 3246866 sequences (480000163 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7006, 149961, 2041, 6697)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 476, 947)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2379)\n",
      "[M::mem_pestat] mean and std.dev: (613.86, 502.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3095)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 367, 409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (166, 571)\n",
      "[M::mem_pestat] mean and std.dev: (369.65, 60.17)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (85, 652)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (56, 547, 1561)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4571)\n",
      "[M::mem_pestat] mean and std.dev: (854.11, 940.84)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6076)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (243, 484, 955)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2379)\n",
      "[M::mem_pestat] mean and std.dev: (616.37, 489.42)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3091)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3243830 reads in 2192.193 CPU sec, 49.673 real sec\n",
      "[M::process] read 3247618 sequences (480000199 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7092, 150125, 2034, 6763)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 486, 970)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2444)\n",
      "[M::mem_pestat] mean and std.dev: (623.45, 502.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3181)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 367, 409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (166, 571)\n",
      "[M::mem_pestat] mean and std.dev: (369.62, 60.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (85, 652)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (60, 539, 1561)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4563)\n",
      "[M::mem_pestat] mean and std.dev: (861.98, 946.06)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6064)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 490, 944)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2366)\n",
      "[M::mem_pestat] mean and std.dev: (617.87, 492.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3077)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3246866 reads in 2223.297 CPU sec, 50.369 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7017, 149505, 2011, 7046)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (226, 464, 930)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2338)\n",
      "[M::mem_pestat] mean and std.dev: (603.30, 496.77)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3042)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 368, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (164, 574)\n",
      "[M::mem_pestat] mean and std.dev: (370.24, 60.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (82, 656)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (66, 597, 1637)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4779)\n",
      "[M::mem_pestat] mean and std.dev: (912.32, 971.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6350)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 497, 983)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2479)\n",
      "[M::mem_pestat] mean and std.dev: (633.98, 516.03)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3227)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3252130 reads in 2239.228 CPU sec, 50.528 real sec\n",
      "[M::process] read 3246436 sequences (480000169 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7536, 158293, 2104, 7281)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 480, 971)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2451)\n",
      "[M::mem_pestat] mean and std.dev: (618.33, 503.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3191)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 366, 409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 573)\n",
      "[M::mem_pestat] mean and std.dev: (369.04, 60.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (81, 655)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (73, 625, 1678)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4888)\n",
      "[M::mem_pestat] mean and std.dev: (905.98, 952.34)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6493)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 495, 971)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2441)\n",
      "[M::mem_pestat] mean and std.dev: (629.88, 501.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3176)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3241472 reads in 2136.222 CPU sec, 48.355 real sec\n",
      "[M::process] read 3246334 sequences (480000246 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7416, 156923, 2126, 7278)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 491, 949)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2377)\n",
      "[M::mem_pestat] mean and std.dev: (619.12, 496.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3091)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 366, 409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 573)\n",
      "[M::mem_pestat] mean and std.dev: (369.34, 60.62)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (81, 655)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (68, 525, 1647)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4805)\n",
      "[M::mem_pestat] mean and std.dev: (886.44, 975.65)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6384)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (238, 491, 968)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2428)\n",
      "[M::mem_pestat] mean and std.dev: (630.36, 509.70)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3158)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3246436 reads in 2160.941 CPU sec, 49.021 real sec\n",
      "[M::process] read 3249090 sequences (480000045 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7387, 156923, 2156, 7341)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 491, 961)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2423)\n",
      "[M::mem_pestat] mean and std.dev: (627.02, 511.18)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3154)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 573)\n",
      "[M::mem_pestat] mean and std.dev: (369.54, 60.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (81, 655)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (74, 622, 1666)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4850)\n",
      "[M::mem_pestat] mean and std.dev: (910.25, 953.38)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6442)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 485, 951)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2391)\n",
      "[M::mem_pestat] mean and std.dev: (614.47, 494.25)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3111)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3246334 reads in 2166.342 CPU sec, 48.862 real sec\n",
      "[M::process] read 3250988 sequences (480000079 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7567, 156517, 2199, 7325)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 491, 953)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2385)\n",
      "[M::mem_pestat] mean and std.dev: (621.59, 503.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3101)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 367, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (164, 574)\n",
      "[M::mem_pestat] mean and std.dev: (369.96, 60.85)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (82, 656)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (79, 606, 1658)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4816)\n",
      "[M::mem_pestat] mean and std.dev: (919.99, 964.96)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6395)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (243, 507, 969)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2421)\n",
      "[M::mem_pestat] mean and std.dev: (632.44, 502.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3147)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3249090 reads in 2191.426 CPU sec, 49.323 real sec\n",
      "[M::process] read 3253696 sequences (480000029 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7552, 155561, 2160, 7363)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 491, 955)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2391)\n",
      "[M::mem_pestat] mean and std.dev: (617.55, 492.93)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3109)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 576)\n",
      "[M::mem_pestat] mean and std.dev: (369.84, 60.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (78, 659)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (73, 602, 1609)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4681)\n",
      "[M::mem_pestat] mean and std.dev: (890.58, 948.73)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6217)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 492, 950)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2372)\n",
      "[M::mem_pestat] mean and std.dev: (621.22, 491.59)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3083)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3250988 reads in 2197.716 CPU sec, 49.539 real sec\n",
      "[M::process] read 3259064 sequences (480000239 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7592, 155609, 2156, 7319)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 482, 957)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2411)\n",
      "[M::mem_pestat] mean and std.dev: (621.69, 508.59)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3138)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 576)\n",
      "[M::mem_pestat] mean and std.dev: (369.65, 60.96)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (78, 659)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (75, 594, 1536)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4458)\n",
      "[M::mem_pestat] mean and std.dev: (879.57, 908.56)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 5919)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 488, 951)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2383)\n",
      "[M::mem_pestat] mean and std.dev: (616.42, 495.01)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3099)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3253696 reads in 2207.269 CPU sec, 49.519 real sec\n",
      "[M::process] read 3262780 sequences (480000145 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7693, 152980, 2132, 7431)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 489, 952)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2394)\n",
      "[M::mem_pestat] mean and std.dev: (618.89, 499.31)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3115)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 368, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (164, 574)\n",
      "[M::mem_pestat] mean and std.dev: (370.21, 61.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (82, 656)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (69, 574, 1630)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4752)\n",
      "[M::mem_pestat] mean and std.dev: (887.35, 950.32)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6313)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (243, 505, 979)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2451)\n",
      "[M::mem_pestat] mean and std.dev: (636.45, 509.71)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3187)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3259064 reads in 2382.205 CPU sec, 53.766 real sec\n",
      "[M::process] read 3260656 sequences (480000026 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7578, 152758, 2122, 7397)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 482, 947)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2371)\n",
      "[M::mem_pestat] mean and std.dev: (613.59, 492.77)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3083)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 576)\n",
      "[M::mem_pestat] mean and std.dev: (370.12, 61.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (78, 659)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (74, 658, 1658)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4826)\n",
      "[M::mem_pestat] mean and std.dev: (926.08, 950.71)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6410)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 499, 962)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2416)\n",
      "[M::mem_pestat] mean and std.dev: (619.99, 496.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3143)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3262780 reads in 2272.067 CPU sec, 51.371 real sec\n",
      "[M::process] read 3213198 sequences (480000086 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7722, 153861, 2211, 7495)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (230, 472, 964)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2432)\n",
      "[M::mem_pestat] mean and std.dev: (619.45, 511.96)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3166)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (328, 368, 411)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (162, 577)\n",
      "[M::mem_pestat] mean and std.dev: (370.73, 61.57)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (79, 660)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (79, 649, 1667)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4843)\n",
      "[M::mem_pestat] mean and std.dev: (921.05, 959.51)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6431)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 491, 944)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2366)\n",
      "[M::mem_pestat] mean and std.dev: (618.94, 499.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3077)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3260656 reads in 2475.342 CPU sec, 61.282 real sec\n",
      "[M::process] read 3215756 sequences (480000219 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6949, 147147, 2024, 6765)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (240, 504, 955)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2385)\n",
      "[M::mem_pestat] mean and std.dev: (625.32, 496.48)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3100)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 573)\n",
      "[M::mem_pestat] mean and std.dev: (369.42, 61.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (81, 655)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (67, 591, 1610)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4696)\n",
      "[M::mem_pestat] mean and std.dev: (901.50, 979.23)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6239)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (241, 498, 953)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2377)\n",
      "[M::mem_pestat] mean and std.dev: (618.81, 487.92)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3089)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3213198 reads in 2373.047 CPU sec, 57.404 real sec\n",
      "[M::process] read 3213518 sequences (480000004 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7550, 155124, 2270, 7516)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (229, 491, 972)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2458)\n",
      "[M::mem_pestat] mean and std.dev: (626.51, 511.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3201)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 576)\n",
      "[M::mem_pestat] mean and std.dev: (369.97, 61.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (78, 659)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (73, 605, 1622)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4720)\n",
      "[M::mem_pestat] mean and std.dev: (904.64, 956.65)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6269)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (238, 500, 961)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2407)\n",
      "[M::mem_pestat] mean and std.dev: (622.64, 498.66)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3130)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3215756 reads in 2268.312 CPU sec, 58.712 real sec\n",
      "[M::process] read 3217380 sequences (480000184 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6692, 142395, 1947, 6624)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 503, 966)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2432)\n",
      "[M::mem_pestat] mean and std.dev: (628.66, 504.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3165)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 573)\n",
      "[M::mem_pestat] mean and std.dev: (369.57, 60.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (81, 655)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (67, 594, 1694)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4948)\n",
      "[M::mem_pestat] mean and std.dev: (901.81, 976.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6575)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 507, 978)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2462)\n",
      "[M::mem_pestat] mean and std.dev: (628.97, 499.27)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3204)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3213518 reads in 2661.544 CPU sec, 76.399 real sec\n",
      "[M::process] read 3216976 sequences (480000241 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6684, 142076, 1965, 6651)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 491, 952)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2386)\n",
      "[M::mem_pestat] mean and std.dev: (619.99, 496.65)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3103)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 573)\n",
      "[M::mem_pestat] mean and std.dev: (369.60, 61.01)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (81, 655)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (66, 544, 1557)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4539)\n",
      "[M::mem_pestat] mean and std.dev: (851.23, 937.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6030)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (235, 493, 959)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2407)\n",
      "[M::mem_pestat] mean and std.dev: (620.81, 500.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3131)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3217380 reads in 2633.200 CPU sec, 63.785 real sec\n",
      "[M::process] read 3218312 sequences (480000244 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7179, 148880, 2089, 7204)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (228, 471, 936)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2352)\n",
      "[M::mem_pestat] mean and std.dev: (603.94, 489.71)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3060)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 576)\n",
      "[M::mem_pestat] mean and std.dev: (369.50, 61.56)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (78, 659)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (71, 612, 1583)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4607)\n",
      "[M::mem_pestat] mean and std.dev: (891.07, 924.61)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6119)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 493, 955)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2403)\n",
      "[M::mem_pestat] mean and std.dev: (618.14, 492.73)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3127)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3216976 reads in 2590.765 CPU sec, 62.514 real sec\n",
      "[M::process] read 3220660 sequences (480000004 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7017, 145931, 1999, 6853)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (229, 479, 961)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2425)\n",
      "[M::mem_pestat] mean and std.dev: (615.29, 497.85)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3157)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 409)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (163, 573)\n",
      "[M::mem_pestat] mean and std.dev: (369.60, 61.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (81, 655)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (67, 588, 1599)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4663)\n",
      "[M::mem_pestat] mean and std.dev: (875.06, 928.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6195)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (233, 500, 950)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2384)\n",
      "[M::mem_pestat] mean and std.dev: (623.08, 495.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3101)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3218312 reads in 2467.206 CPU sec, 60.110 real sec\n",
      "[M::process] read 3219418 sequences (480000235 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (7638, 152564, 2136, 7489)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (228, 479, 945)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2379)\n",
      "[M::mem_pestat] mean and std.dev: (610.20, 494.15)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3096)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 411)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (159, 579)\n",
      "[M::mem_pestat] mean and std.dev: (370.15, 62.04)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (75, 663)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (85, 629, 1676)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4858)\n",
      "[M::mem_pestat] mean and std.dev: (931.67, 969.47)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6449)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (229, 487, 961)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2425)\n",
      "[M::mem_pestat] mean and std.dev: (624.00, 505.93)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3157)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3220660 reads in 2706.218 CPU sec, 63.678 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (6741, 141932, 1982, 6652)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (222, 470, 939)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2373)\n",
      "[M::mem_pestat] mean and std.dev: (610.26, 502.51)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3090)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 410)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (161, 576)\n",
      "[M::mem_pestat] mean and std.dev: (369.71, 61.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (78, 659)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (74, 612, 1598)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4646)\n",
      "[M::mem_pestat] mean and std.dev: (894.66, 925.60)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6170)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 498, 968)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2442)\n",
      "[M::mem_pestat] mean and std.dev: (625.59, 506.08)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3179)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3219418 reads in 2515.586 CPU sec, 67.972 real sec\n",
      "[M::process] read 3231708 sequences (480000079 bp)...\n",
      "[M::process] read 3229040 sequences (480000222 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8273, 155621, 2197, 7763)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (222, 461, 926)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2334)\n",
      "[M::mem_pestat] mean and std.dev: (599.99, 497.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3038)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 367, 411)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (159, 579)\n",
      "[M::mem_pestat] mean and std.dev: (370.19, 62.26)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (75, 663)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (81, 649, 1646)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4776)\n",
      "[M::mem_pestat] mean and std.dev: (930.28, 964.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6341)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (225, 469, 925)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2325)\n",
      "[M::mem_pestat] mean and std.dev: (598.60, 487.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3025)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_process_seqs] Processed 3231708 reads in 2401.828 CPU sec, 56.028 real sec\n",
      "[M::process] read 3233674 sequences (480000250 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8470, 163580, 2340, 8520)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (229, 481, 962)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2428)\n",
      "[M::mem_pestat] mean and std.dev: (619.25, 505.58)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3161)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (326, 367, 411)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (156, 581)\n",
      "[M::mem_pestat] mean and std.dev: (370.21, 62.46)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (71, 666)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (89, 648, 1596)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4610)\n",
      "[M::mem_pestat] mean and std.dev: (922.91, 948.34)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6117)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (237, 504, 968)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2430)\n",
      "[M::mem_pestat] mean and std.dev: (627.93, 501.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3161)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 3229040 reads in 2592.597 CPU sec, 61.550 real sec\n",
      "[M::process] read 3231540 sequences (480000074 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8906, 165737, 2260, 8578)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 472, 951)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2391)\n",
      "[M::mem_pestat] mean and std.dev: (613.61, 499.12)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3111)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 368, 412)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (157, 582)\n",
      "[M::mem_pestat] mean and std.dev: (370.66, 62.49)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (72, 667)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (99, 676, 1730)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4992)\n",
      "[M::mem_pestat] mean and std.dev: (1022.60, 1071.78)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6623)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (239, 502, 947)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2363)\n",
      "[M::mem_pestat] mean and std.dev: (619.84, 488.76)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3071)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 3233674 reads in 2456.689 CPU sec, 56.214 real sec\n",
      "[M::process] read 3234980 sequences (480000148 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8931, 167865, 2456, 8629)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 495, 953)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2397)\n",
      "[M::mem_pestat] mean and std.dev: (622.44, 501.37)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3119)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 368, 413)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (155, 585)\n",
      "[M::mem_pestat] mean and std.dev: (370.99, 63.26)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (69, 671)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (105, 673, 1669)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4797)\n",
      "[M::mem_pestat] mean and std.dev: (948.07, 949.81)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6361)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (240, 496, 961)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2403)\n",
      "[M::mem_pestat] mean and std.dev: (623.17, 495.04)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3124)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 3231540 reads in 2484.177 CPU sec, 57.347 real sec\n",
      "[M::process] read 3235280 sequences (480000272 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8815, 165622, 2421, 8676)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (229, 499, 968)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2446)\n",
      "[M::mem_pestat] mean and std.dev: (628.39, 507.30)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3185)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 368, 412)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (157, 582)\n",
      "[M::mem_pestat] mean and std.dev: (370.88, 62.94)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (72, 667)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (92, 625, 1632)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4712)\n",
      "[M::mem_pestat] mean and std.dev: (905.82, 923.07)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6252)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (231, 481, 932)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2334)\n",
      "[M::mem_pestat] mean and std.dev: (611.94, 494.74)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3035)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 3234980 reads in 2504.711 CPU sec, 61.947 real sec\n",
      "[M::process] read 3239196 sequences (480000075 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8629, 164564, 2444, 8675)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (227, 492, 976)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2474)\n",
      "[M::mem_pestat] mean and std.dev: (624.04, 509.44)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3223)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 368, 413)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (155, 585)\n",
      "[M::mem_pestat] mean and std.dev: (371.11, 63.26)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (69, 671)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (89, 665, 1689)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4889)\n",
      "[M::mem_pestat] mean and std.dev: (940.24, 962.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6489)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (232, 487, 963)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2425)\n",
      "[M::mem_pestat] mean and std.dev: (623.37, 505.23)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3156)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 3235280 reads in 2573.856 CPU sec, 69.710 real sec\n",
      "[M::process] read 3241358 sequences (480000230 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8888, 162333, 2419, 8609)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (232, 484, 958)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2410)\n",
      "[M::mem_pestat] mean and std.dev: (618.79, 499.82)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3136)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 368, 412)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (157, 582)\n",
      "[M::mem_pestat] mean and std.dev: (370.85, 62.93)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (72, 667)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (92, 672, 1621)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4679)\n",
      "[M::mem_pestat] mean and std.dev: (928.08, 924.79)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6208)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (236, 495, 959)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2405)\n",
      "[M::mem_pestat] mean and std.dev: (619.53, 494.99)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3128)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 3239196 reads in 2563.319 CPU sec, 63.694 real sec\n",
      "[M::process] read 1754198 sequences (259781433 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (8799, 161628, 2463, 8565)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (223, 464, 940)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2374)\n",
      "[M::mem_pestat] mean and std.dev: (604.60, 498.40)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3091)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 368, 412)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (157, 582)\n",
      "[M::mem_pestat] mean and std.dev: (370.89, 63.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (72, 667)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (102, 635, 1643)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4725)\n",
      "[M::mem_pestat] mean and std.dev: (924.31, 942.81)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6266)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (229, 479, 953)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2401)\n",
      "[M::mem_pestat] mean and std.dev: (616.42, 501.44)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3125)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 3241358 reads in 2577.554 CPU sec, 63.905 real sec\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (4895, 87354, 1284, 4830)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (227, 460, 947)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2387)\n",
      "[M::mem_pestat] mean and std.dev: (599.92, 489.02)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3107)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (327, 368, 413)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (155, 585)\n",
      "[M::mem_pestat] mean and std.dev: (371.30, 63.32)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (69, 671)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (111, 706, 1647)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 4719)\n",
      "[M::mem_pestat] mean and std.dev: (938.39, 935.61)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 6255)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (227, 472, 936)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2354)\n",
      "[M::mem_pestat] mean and std.dev: (602.24, 487.59)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3063)\n",
      "[M::mem_pestat] skip orientation RF\n",
      "[M::mem_process_seqs] Processed 1754198 reads in 1458.069 CPU sec, 36.750 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -R @RG\\tID:SRR6399462\\tSM:bar -t 48 /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/megahit/final.contigs.fastg.gfa.fna /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/trim_galore/SRR6399462_1_val_1.fq.gz /home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/trim_galore/SRR6399462_2_val_2.fq.gz\n",
      "[main] Real time: 2177.985 sec; CPU: 89475.048 sec\n",
      "[bam_sort_core] merging from 48 files and 48 in-memory blocks...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Java memory size is set to 4G\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.2.2-dev\n",
      "Built on 2019-11-11 14:05\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 37\n",
      "Max memory (Mb): 4294\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 6483\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 48\n",
      "Processed 648 out of 6483 windows...\n",
      "Processed 1296 out of 6483 windows...\n",
      "Processed 1944 out of 6483 windows...\n",
      "Processed 2592 out of 6483 windows...\n",
      "Processed 3240 out of 6483 windows...\n",
      "Processed 3888 out of 6483 windows...\n",
      "Processed 4536 out of 6483 windows...\n",
      "Processed 5184 out of 6483 windows...\n",
      "Processed 5832 out of 6483 windows...\n",
      "Processed 6480 out of 6483 windows...\n",
      "Total processed windows:6483\n",
      "Number of reads: 124896854\n",
      "Number of valid reads: 132348735\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 123797785\n",
      "Num mapped first of pair: 62000508\n",
      "Num mapped second of pair: 61797277\n",
      "Num singletons: 722557\n",
      "Time taken to analyze reads: 1196\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 18092985381\n",
      "referenceSize: 4459785\n",
      "numberOfSequencedBases: 18080902967\n",
      "numberOfAs: 4834337109\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 1224\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing PDF:HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "PDF file created successfully \n",
      "\n",
      "Finished\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32mProgress:  12%|█████▏                                   |  ETA: 1 days, 7:16:24\u001b[39m"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR = \"SRR6399463\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-02-24 03:06:57 - MEGAHIT v1.2.9\n",
      "2023-02-24 03:06:57 - Using megahit_core with POPCNT and BMI2 support\n",
      "2023-02-24 03:06:57 - Convert reads to binary library\n",
      "2023-02-24 03:10:59 - b'INFO  sequence/io/sequence_lib.cpp  :   75 - Lib 0 (/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399463/trim_galore/SRR6399463_1_val_1.fq.gz,/home/jovyan/work/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399463/trim_galore/SRR6399463_2_val_2.fq.gz): pe, 93012880 reads, 151 max length'\n",
      "2023-02-24 03:11:04 - b'INFO  utils/utils.h                 :  152 - Real: 246.6784\\tuser: 102.7976\\tsys: 10.4561\\tmaxrss: 254220'\n",
      "2023-02-24 03:11:04 - k-max reset to: 141 \n",
      "2023-02-24 03:11:04 - Start assembly. Number of CPU threads 48 \n",
      "2023-02-24 03:11:04 - k list: 21,29,39,59,79,99,119,141 \n",
      "2023-02-24 03:11:04 - Memory used: 121546591027\n",
      "2023-02-24 03:11:04 - Extract solid (k+1)-mers for k = 21 \n",
      "2023-02-24 03:19:46 - Build graph for k = 21 \n",
      "2023-02-24 03:22:47 - Assemble contigs from SdBG for k = 21\n",
      "2023-02-24 03:31:56 - Local assembly for k = 21\n",
      "2023-02-24 03:35:34 - Extract iterative edges from k = 21 to 29 \n",
      "2023-02-24 03:38:54 - Build graph for k = 29 \n",
      "2023-02-24 03:39:46 - Assemble contigs from SdBG for k = 29\n",
      "2023-02-24 03:40:58 - Local assembly for k = 29\n",
      "2023-02-24 03:48:07 - Extract iterative edges from k = 29 to 39 \n",
      "2023-02-24 03:48:34 - Build graph for k = 39 \n"
     ]
    }
   ],
   "source": [
    "ProgressMeter.@showprogress for SRR_path in SRR_paths\n",
    "    SRR = basename(SRR_path)\n",
    "    @show SRR\n",
    "\n",
    "    out_dir = joinpath(SRR_path, \"megahit\")\n",
    "\n",
    "    trimmed_forward_reads = joinpath(SRR_path, \"trim_galore\", \"$(SRR)_1_val_1.fq.gz\")\n",
    "    trimmed_reverse_reads = joinpath(SRR_path, \"trim_galore\", \"$(SRR)_2_val_2.fq.gz\")\n",
    "\n",
    "    # megahit: MEGAHIT v1.2.9\n",
    "\n",
    "    # contact: Dinghua Li <voutcn@gmail.com>\n",
    "\n",
    "    # Usage:\n",
    "    #   megahit [options] {-1 <pe1> -2 <pe2> | --12 <pe12> | -r <se>} [-o <out_dir>]\n",
    "\n",
    "    #   Input options that can be specified for multiple times (supporting plain text and gz/bz2 extensions)\n",
    "    #     -1                       <pe1>          comma-separated list of fasta/q paired-end #1 files, paired with files in <pe2>\n",
    "    #     -2                       <pe2>          comma-separated list of fasta/q paired-end #2 files, paired with files in <pe1>\n",
    "    #     --12                     <pe12>         comma-separated list of interleaved fasta/q paired-end files\n",
    "    #     -r/--read                <se>           comma-separated list of fasta/q single-end files\n",
    "\n",
    "    # Optional Arguments:\n",
    "    #   Basic assembly options:\n",
    "    #     --min-count              <int>          minimum multiplicity for filtering (k_min+1)-mers [2]\n",
    "    #     --k-list                 <int,int,..>   comma-separated list of kmer size\n",
    "    #                                             all must be odd, in the range 15-255, increment <= 28)\n",
    "    #                                             [21,29,39,59,79,99,119,141]\n",
    "\n",
    "    #   Another way to set --k-list (overrides --k-list if one of them set):\n",
    "    #     --k-min                  <int>          minimum kmer size (<= 255), must be odd number [21]\n",
    "    #     --k-max                  <int>          maximum kmer size (<= 255), must be odd number [141]\n",
    "    #     --k-step                 <int>          increment of kmer size of each iteration (<= 28), must be even number [12]\n",
    "\n",
    "    #   Advanced assembly options:\n",
    "    #     --no-mercy                              do not add mercy kmers\n",
    "    #     --bubble-level           <int>          intensity of bubble merging (0-2), 0 to disable [2]\n",
    "    #     --merge-level            <l,s>          merge complex bubbles of length <= l*kmer_size and similarity >= s [20,0.95]\n",
    "    #     --prune-level            <int>          strength of low depth pruning (0-3) [2]\n",
    "    #     --prune-depth            <int>          remove unitigs with avg kmer depth less than this value [2]\n",
    "    #     --disconnect-ratio       <float>        disconnect unitigs if its depth is less than this ratio times \n",
    "    #                                             the total depth of itself and its siblings [0.1]  \n",
    "    #     --low-local-ratio        <float>        remove unitigs if its depth is less than this ratio times\n",
    "    #                                             the average depth of the neighborhoods [0.2]\n",
    "    #     --max-tip-len            <int>          remove tips less than this value [2*k]\n",
    "    #     --cleaning-rounds        <int>          number of rounds for graph cleanning [5]\n",
    "    #     --no-local                              disable local assembly\n",
    "    #     --kmin-1pass                            use 1pass mode to build SdBG of k_min\n",
    "\n",
    "    #   Presets parameters:\n",
    "    #     --presets                <str>          override a group of parameters; possible values:\n",
    "    #                                             meta-sensitive: '--min-count 1 --k-list 21,29,39,49,...,129,141'\n",
    "    #                                             meta-large: '--k-min 27 --k-max 127 --k-step 10'\n",
    "    #                                             (large & complex metagenomes, like soil)\n",
    "\n",
    "    #   Hardware options:\n",
    "    #     -t/--num-cpu-threads     <int>          number of CPU threads [# of logical processors]\n",
    "\n",
    "    #   Output options:\n",
    "    #     -o/--out-dir             <string>       output directory [./megahit_out]\n",
    "    #     --out-prefix             <string>       output prefix (the contig file will be OUT_DIR/OUT_PREFIX.contigs.fa)\n",
    "    #     --min-contig-len         <int>          minimum length of contigs to output [200]\n",
    "\n",
    "    # median_read_length = Statistics.median(vcat(Mycelia.determine_read_lengths(trimmed_forward_reads), Mycelia.determine_read_lengths(trimmed_reverse_reads)))\n",
    "\n",
    "    # k_lengths = [i for i in [21,29,39,59,79,99,119,141] if i < joint_median_read_length]\n",
    "    # k_lengths_string = join(string.(k_lengths), ',')\n",
    "\n",
    "    # 2023-02-19 00:51:46 - ALL DONE. Time elapsed: 2031.380950 seconds \n",
    "\n",
    "    initial_assembled_fasta = \"$(out_dir)/final.contigs.fa\"\n",
    "    assembled_fastg = replace(initial_assembled_fasta, \".fa\" => \".fastg\")\n",
    "\n",
    "    if !isfile(initial_assembled_fasta)\n",
    "        run(`megahit --num-cpu-threads $(Sys.CPU_THREADS) -1 $trimmed_forward_reads -2 $trimmed_reverse_reads -o $(out_dir)`)\n",
    "        # run(`conda run --live-stream --no-capture-output -n viral-pangenome-discovery megahit --num-cpu-threads $(Sys.CPU_THREADS) -1 $trimmed_forward_reads -2 $trimmed_reverse_reads -o $(out_dir)`)\n",
    "    end\n",
    "\n",
    "    # read in the assembled fasta file and parse contig identifiers to get final k length\n",
    "    final_k_lengths = unique([replace(first(split(FASTX.identifier(record), '_')), r\"^k\" => \"\") for record in Mycelia.open_fastx(initial_assembled_fasta)])\n",
    "    @assert length(final_k_lengths) == 1\n",
    "    final_k_length = parse(Int, first(final_k_lengths))\n",
    "    if !isfile(assembled_fastg)\n",
    "        run(pipeline(`megahit_toolkit contig2fastg $(final_k_length) $(initial_assembled_fasta)`, assembled_fastg))\n",
    "    end\n",
    "\n",
    "    assembled_gfa = \"$(assembled_fastg).gfa\"\n",
    "    if !isfile(assembled_gfa)\n",
    "        run(`Bandage reduce $(assembled_fastg) $(assembled_gfa)`)\n",
    "    end\n",
    "\n",
    "    assembled_fasta = assembled_gfa * \".fna\"\n",
    "    open(assembled_fasta, \"w\") do io\n",
    "        fastx_io = FASTX.FASTA.Writer(io)\n",
    "        gfa_graph = Mycelia.parse_gfa(assembled_gfa)\n",
    "        for v in Graphs.vertices(gfa_graph)\n",
    "            record = FASTX.FASTA.Record(gfa_graph.vprops[v][:identifier], gfa_graph.vprops[v][:sequence])\n",
    "            write(fastx_io, record)\n",
    "        end\n",
    "        close(fastx_io)\n",
    "    end\n",
    "\n",
    "    # generate a bandage plot of the assembly graph\n",
    "    bandage_outfile = \"$(assembled_gfa).bandage.jpg\"\n",
    "    if !isfile(bandage_outfile)\n",
    "        run(`Bandage image $(assembled_gfa) $bandage_outfile`)\n",
    "    end\n",
    "\n",
    "    # map reads to the assembly and run qualimap QC\n",
    "    bwt_index = \"$(assembled_fasta).bwt\"\n",
    "    if !isfile(bwt_index)\n",
    "        run(`bwa index $(assembled_fasta)`)\n",
    "    end\n",
    "\n",
    "    # mapped_reads_bam = \"$(assembled_fasta).bwa.sorted.marked_duplicates.bam\"\n",
    "    mapped_reads_bam = \"$(assembled_fasta).bwa.bam\"\n",
    "    # sorting and deduping\n",
    "    # if !isfile(mapped_reads_bam)\n",
    "    #     run(pipeline(\n",
    "    #         `bwa mem -R \"@RG\\tID:$(SRR)\\tSM:bar\" -t $(Sys.CPU_THREADS) $(assembled_fasta) $(trimmed_forward_reads) $(trimmed_reverse_reads)`,\n",
    "    #         `samtools collate --threads $(Sys.CPU_THREADS) -O - -`,\n",
    "    #         `samtools fixmate --threads $(Sys.CPU_THREADS) -m - -`,\n",
    "    #         `samtools sort --threads $(Sys.CPU_THREADS)`,\n",
    "    #         `samtools markdup --threads $(Sys.CPU_THREADS) - -`,\n",
    "    #         `samtools view --threads $(Sys.CPU_THREADS) -buh`,\n",
    "    #         mapped_reads_bams))\n",
    "    # end\n",
    "    # sorting - 50Gb - 1594.823 sec\n",
    "    if !isfile(mapped_reads_bam)\n",
    "        run(pipeline(\n",
    "            `bwa mem -R \"@RG\\tID:$(SRR)\\tSM:bar\" -t $(Sys.CPU_THREADS) $(assembled_fasta) $(trimmed_forward_reads) $(trimmed_reverse_reads)`,\n",
    "            `samtools sort --threads $(Sys.CPU_THREADS)`,\n",
    "            `samtools view --threads $(Sys.CPU_THREADS) -buh`,\n",
    "            mapped_reads_bam))\n",
    "    end\n",
    "    # no sorting - 50Gb - 1762.476216 seconds\n",
    "    # if !isfile(mapped_reads_bam)\n",
    "    #     @time run(pipeline(\n",
    "    #         `bwa mem -R \"@RG\\tID:$(SRR)\\tSM:bar\" -t $(Sys.CPU_THREADS) $(assembled_fasta) $(trimmed_forward_reads) $(trimmed_reverse_reads)`,\n",
    "    #         `samtools view --threads $(Sys.CPU_THREADS) -buh`,\n",
    "    #         mapped_reads_bam))\n",
    "    # end\n",
    "\n",
    "    if !isfile(\"$(mapped_reads_bam).bai\")\n",
    "        run(`samtools index $(mapped_reads_bam)`)\n",
    "    end\n",
    "\n",
    "    qualimap_report_pdf = \"$(out_dir)/qualimap/report.pdf\"\n",
    "    qualimap_report_txt = \"$(out_dir)/qualimap/genome_results.txt\"\n",
    "\n",
    "    if !isfile(qualimap_report_pdf) || !isfile(qualimap_report_txt)\n",
    "        run(`\n",
    "            conda run --live-stream -n viral-pangenome-discovery\n",
    "            qualimap bamqc\n",
    "            -nt $(Sys.CPU_THREADS)\n",
    "            -bam $(mapped_reads_bam)\n",
    "            -outdir $(out_dir)/qualimap\n",
    "            -outformat PDF:HTML\n",
    "            --output-genome-coverage $(mapped_reads_bam).genome_coverage.txt\n",
    "            --java-mem-size=4G\n",
    "            `)\n",
    "    end\n",
    "end\n",
    "\n",
    "# qualimap_contig_coverage_table = Mycelia.parse_qualimap_contig_coverage(qualimap_report_txt)\n",
    "# qualimap_contig_coverage_table[!, \"% Mapped bases\"] = round.(qualimap_contig_coverage_table[!, \"Mapped bases\"] ./ sum(qualimap_contig_coverage_table[!, \"Mapped bases\"]) .* 100, digits=3);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2e6f4c6-10cc-4863-86df-aa88a146ae42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# consider deleting the bams after we get the qualimap reports - we don't need them anymore"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.3",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
