{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce59cd77-c0f0-4b90-8de8-3c5f071935ae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import Pkg\n",
    "Pkg.activate(\".\")\n",
    "\n",
    "# not for the faint of heart!\n",
    "# Pkg.update()\n",
    "\n",
    "pkgs = [\n",
    "\"ArgParse\",\n",
    "\"Base64\",\n",
    "\"BioSequences\",\n",
    "\"DataFrames\",\n",
    "\"Dates\",\n",
    "\"DelimitedFiles\",\n",
    "\"FASTX\",\n",
    "\"GLM\",\n",
    "\"HTTP\",\n",
    "\"JSON\",\n",
    "\"Graphs\",\n",
    "\"MetaGraphs\",\n",
    "\"MD5\",\n",
    "\"Statistics\",\n",
    "\"StatsPlots\",\n",
    "\"uCSV\",\n",
    "\"CodecZlib\",\n",
    "\"YAML\",\n",
    "\"Revise\",\n",
    "\"Kmers\",\n",
    "\"StatsBase\"\n",
    "]\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "909f13b4-7e9d-429b-8b83-33c99af653c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = joinpath(dirname(pwd()), \"data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e92373a-9477-4d55-8a80-75cc392470be",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# cpan -f App::cpan minus\n",
    "# cpanm List::MoreUtils\n",
    "# mamba install perl-list-moreutils\n",
    "\n",
    "blastdb_dir = \"$(homedir())/workspace/blastdb\"\n",
    "blast_db = \"nt\"\n",
    "if isdir(blastdb_dir) && !isempty(readdir(blastdb_dir))\n",
    "    @info \"blast db detected, using existing\"\n",
    "else\n",
    "    # Mycelia.download_blast_db(db=blast_db, outdir=blastdb_dir, source=\"ncbi\", conda_env=\"viral-pangenome-discovery\")\n",
    "    Mycelia.download_blast_db(db=blast_db, outdir=blastdb_dir, source=\"ncbi\")\n",
    "end\n",
    "\n",
    "# move me up to the initial download location\n",
    "taxdump_tar = joinpath(blastdb_dir, \"taxdump.tar.gz\")\n",
    "if !isfile(taxdump_tar)\n",
    "    run(`wget --quiet https://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz`)\n",
    "end\n",
    "taxdump_dir = mkpath(joinpath(blastdb_dir, \"taxdump\"))\n",
    "if isempty(readdir(taxdump_dir))\n",
    "    run(`tar -xvzf taxdump.tar.gz --directory $(taxdump_dir)`)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5685b842-daff-4e86-9235-4b8094979ca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "SRR_paths = filter(x -> !occursin(\".ipynb_checkpoints\", x), readdir(joinpath(data_dir, \"SRA\"), join=true))\n",
    "SRR_paths = filter(x -> isfile(joinpath(x, \"megahit\", \"final.contigs.fastg.gfa.fna\")), SRR_paths)\n",
    "SRR_paths = filter(x -> !isfile(joinpath(x, basename(x) * \".contig_info.tsv\")), SRR_paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a1a618d-77be-4274-8f98-e25e7703886e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for SRR_path in SRR_paths\n",
    "    SRR = basename(SRR_path)\n",
    "    @show SRR\n",
    "    contig_info_tsv = \"$(SRR_path)/$(SRR).contig_info.tsv\"\n",
    "    if isfile(contig_info_tsv)\n",
    "        @info \"$(contig_info_tsv) already present, skipping...\"\n",
    "    else\n",
    "        assembled_fasta = joinpath(SRR_path, \"megahit\", \"final.contigs.fastg.gfa.fna\")\n",
    "\n",
    "        blast_db = \"nt\"\n",
    "        blastdb_dir = \"$(homedir())/workspace/blastdb\"\n",
    "        db_path = joinpath(blastdb_dir, blast_db)\n",
    "        @time blast_report = Mycelia.run_blast(out_dir = SRR_path, fasta = assembled_fasta, blast_db = db_path, blast_command = \"blastn\")\n",
    "        # join the blast results to generate the contig info file\n",
    "        ncbi_blast_results = Mycelia.parse_blast_report(blast_report)\n",
    "        detected_tax_id_file = \"$(assembled_fasta).detected_tax_ids.txt\"\n",
    "        open(detected_tax_id_file, \"w\") do io\n",
    "            for taxid in unique(filter(!ismissing, ncbi_blast_results[!, \"subject tax id\"]))\n",
    "                println(io, taxid)\n",
    "            end\n",
    "        end\n",
    "        taxid_to_lineage_map = Dict(parse(Int, split_line[1]) => split_line[2] for split_line in split.(readlines(`taxonkit lineage --data-dir $(taxdump_dir) $(detected_tax_id_file)`), '\\t'))\n",
    "\n",
    "        qualimap_report_txt = joinpath(SRR_path, \"megahit\", \"qualimap\", \"genome_results.txt\")\n",
    "        qualimap_contig_coverage_table = Mycelia.parse_qualimap_contig_coverage(qualimap_report_txt)\n",
    "\n",
    "        ncbi_blast_results[!, \"lineage\"] = map(x -> get(taxid_to_lineage_map, x, \"\"), ncbi_blast_results[!, \"subject tax id\"])\n",
    "        ncbi_blast_results[!, \"% of subject length\"] = round.(ncbi_blast_results[!, \"query length\"] ./ ncbi_blast_results[!, \"subject length\"] * 100, digits=3)\n",
    "        contig_info_table = DataFrames.leftjoin(qualimap_contig_coverage_table, ncbi_blast_results, on=\"Contig\" => \"query id\")\n",
    "\n",
    "        # re-order columns based on utility\n",
    "        reordered_columns = [\n",
    "            \"Contig\",\n",
    "            \"Length\",\n",
    "            \"Mapped bases\",\n",
    "            \"Mean coverage\",\n",
    "            \"Standard Deviation\",\n",
    "            \"% Mapped bases\",\n",
    "            \"subject id\",\n",
    "            \"subject acc.\",\n",
    "            \"subject title\",\n",
    "            \"subject tax id\",\n",
    "            \"lineage\",\n",
    "            \"% identity\",\n",
    "            \"% of subject length\",\n",
    "            \"evalue\",\n",
    "            \"bit score\",\n",
    "            \"query length\",\n",
    "            \"subject length\",\n",
    "            \"alignment length\",\n",
    "            \"q. start\",\n",
    "            \"q. end\",\n",
    "            \"s. start\",\n",
    "            \"s. end\",\n",
    "            \"identical\",\n",
    "            \"mismatches\"\n",
    "        ]\n",
    "        contig_info_table = contig_info_table[!, reordered_columns]\n",
    "        sort!(contig_info_table, [\"% Mapped bases\", \"bit score\"], rev=true)\n",
    "\n",
    "        # consider gzipping for large files!\n",
    "        uCSV.write(contig_info_tsv, contig_info_table)\n",
    "    end\n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.3",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
