{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ce59cd77-c0f0-4b90-8de8-3c5f071935ae",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-07-10T20:41:49.893000Z",
     "iopub.status.busy": "2023-07-10T20:41:49.224000Z",
     "iopub.status.idle": "2023-07-10T20:42:31.795000Z",
     "shell.execute_reply": "2023-07-10T20:42:31.725000Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m  Activating\u001b[22m\u001b[39m environment at `~/workspace/Mycelia/projects/viral-pangenome-discovery/notebooks/Project.toml`\n",
      "\u001b[32m\u001b[1m    Updating\u001b[22m\u001b[39m registry at `~/.julia/registries/General`\n",
      "\u001b[32m\u001b[1m   Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/workspace/Mycelia/projects/viral-pangenome-discovery/notebooks/Project.toml`\n",
      "\u001b[32m\u001b[1m  No Changes\u001b[22m\u001b[39m to `~/workspace/Mycelia/projects/viral-pangenome-discovery/notebooks/Manifest.toml`\n"
     ]
    }
   ],
   "source": [
    "import Pkg\n",
    "Pkg.activate(\".\")\n",
    "\n",
    "# not for the faint of heart!\n",
    "# Pkg.update()\n",
    "\n",
    "pkgs = [\n",
    "\"ArgParse\",\n",
    "\"Base64\",\n",
    "\"BioSequences\",\n",
    "\"DataFrames\",\n",
    "\"Dates\",\n",
    "\"DelimitedFiles\",\n",
    "\"FASTX\",\n",
    "\"Statistics\",\n",
    "\"StatsPlots\",\n",
    "\"uCSV\",\n",
    "\"Revise\",\n",
    "\"StatsBase\"\n",
    "]\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "909f13b4-7e9d-429b-8b83-33c99af653c9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-07-10T20:42:33.360000Z",
     "iopub.status.busy": "2023-07-10T20:42:31.795000Z",
     "iopub.status.idle": "2023-07-10T20:42:34.680000Z",
     "shell.execute_reply": "2023-07-10T20:42:34.678000Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data\""
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_dir = joinpath(dirname(pwd()), \"data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1ed0cada-73ac-4c3c-8d8c-e50e1115e414",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-07-10T20:42:34.680000Z",
     "iopub.status.busy": "2023-07-10T20:42:34.680000Z",
     "iopub.status.idle": "2023-07-10T20:42:37.184000Z",
     "shell.execute_reply": "2023-07-10T20:42:37.184000Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "594-element Vector{String}:\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399459\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399460\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399461\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399463\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399464\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399465\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399466\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399467\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399468\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399469\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399470\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399471\"\n",
       " â‹®\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365474\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365475\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365476\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365477\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365478\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365479\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365480\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365481\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365482\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365483\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365484\"\n",
       " \"/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR7365485\""
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SRR_paths = filter(x -> !occursin(\".ipynb_checkpoints\", x), readdir(joinpath(data_dir, \"SRA\"), join=true))\n",
    "SRR_paths = filter(x -> isfile(joinpath(x, \"megahit\", \"final.contigs.fastg.gfa.fna\")), SRR_paths)\n",
    "# filter out samples that have already been processed\n",
    "# SRR_paths = filter(x -> !isfile(joinpath(x, basename(x) * \".contig_info.tsv\")), SRR_paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64305f91-470b-48b4-9a36-f61fe7a2c58d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-07-10T20:42:37.184000Z",
     "iopub.status.busy": "2023-07-10T20:42:37.184000Z",
     "iopub.status.idle": "2023-07-10T20:42:37.213000Z",
     "shell.execute_reply": "2023-07-10T20:42:37.213000Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# mamba install -c conda-forge -c bioconda virsorter=2\n",
    "# mamba create -n vs2 -c conda-forge -c bioconda virsorter=2\n",
    "# mamba activate vs2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0e3904dd-55e2-471e-a3ad-bb34e5881ae7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-07-10T20:42:37.214000Z",
     "iopub.status.busy": "2023-07-10T20:42:37.213000Z",
     "iopub.status.idle": "2023-07-10T20:42:37.241000Z",
     "shell.execute_reply": "2023-07-10T20:42:37.241000Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "virsorter_db_dir = mkpath(\"$(homedir())/workspace/virsorter2\")\n",
    "\n",
    "# got conda errors with this\n",
    "\n",
    "# (base) 2023-06-06T16:05:47 jovyan@76eefed4ebe2:~/workspace/Mycelia\n",
    "# $ virsorter setup -d /home/jovyan/workspace/virsorter2 -j 4\n",
    "# Traceback (most recent call last):\n",
    "#   File \"/opt/conda/bin/virsorter\", line 6, in <module>\n",
    "#     from virsorter.virsorter import cli\n",
    "#   File \"/opt/conda/lib/python3.9/site-packages/virsorter/virsorter.py\", line 11, in <module>\n",
    "#     from snakemake import load_configfile\n",
    "#   File \"/opt/conda/lib/python3.9/site-packages/snakemake/__init__.py\", line 21, in <module>\n",
    "#     from snakemake.workflow import Workflow\n",
    "#   File \"/opt/conda/lib/python3.9/site-packages/snakemake/workflow.py\", line 23, in <module>\n",
    "#     from snakemake.rules import Rule, Ruleorder, RuleProxy\n",
    "#   File \"/opt/conda/lib/python3.9/site-packages/snakemake/rules.py\", line 54, in <module>\n",
    "#     from snakemake.common import Mode, lazy_property, TBDInt\n",
    "# ImportError: cannot import name 'TBDInt' from 'snakemake.common' (/opt/conda/lib/python3.9/site-packages/snakemake/common/__init__.py)\n",
    "\n",
    "# run(`virsorter setup -d $(virsorter_db_dir) -j 4`)\n",
    "# run(`https://osf.io/v46sc/download`)\n",
    "\n",
    "db_tgz_path = joinpath(virsorter_db_dir, \"db.tgz\")\n",
    "if !isfile(db_tgz_path)\n",
    "    download(\"https://osf.io/v46sc/download\", joinpath(virsorter_db_dir, \"db.tgz\"))\n",
    "end\n",
    "db_dir = replace(db_tgz_path, \".tgz\" => \"\")\n",
    "if !isdir(db_dir)\n",
    "    run(`tar -xzf $(db_tgz_path) --directory=$(virsorter_db_dir)`)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "728c1331-c9b4-44f3-ab89-308f309d39f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# mamba create -n vs2 -c conda-forge -c bioconda virsorter=2\n",
    "# $  mamba activate vs2\n",
    "# (vs2) 2023-06-06T16:26:44 jovyan@76eefed4ebe2:~\n",
    "# $ virsorter config --init-source --db-dir=/home/jovyan/workspace/virsorter2/db\n",
    "# [2023-06-06 16:27 INFO] VirSorter 2.2.4\n",
    "# [2023-06-06 16:27 INFO] /opt/conda/envs/vs2/bin/virsorter config --init-source --db-dir=/home/jovyan/workspace/virsorter2/db\n",
    "# [2023-06-06 16:27 INFO] saving /home/jovyan/workspace/virsorter2/db as DBDIR to config file /opt/conda/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml\n",
    "\n",
    "# # run(`virsorter config --init-source --db-dir=$(db_dir)`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3950e040-b8d9-46c6-a106-bf23e5e947ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # fetch testing data\n",
    "# wget -O test.fa https://raw.githubusercontent.com/jiarong/VirSorter2/master/test/8seq.fa\n",
    "# # run classification with 4 threads (-j) and test-out as output diretory (-w)\n",
    "# virsorter run -w test.out -i test.fa --min-length 1500 -j 4 all\n",
    "# ls test.out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "632f4f05-b7a4-4a0b-a80b-4c03fee68129",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-07-10T20:42:42.892000Z",
     "iopub.status.busy": "2023-07-10T20:42:42.891000Z",
     "iopub.status.idle": "2023-07-10T20:42:42.904000Z",
     "shell.execute_reply": "2023-07-10T20:42:42.904000Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"/home/jovyan/workspace/virsorter2/db\""
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "virsorter_db_dir_path = joinpath(virsorter_db_dir, \"db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "443c110b-d791-4b18-a31b-e32dc6a89b97",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-07-10T20:45:44.395000Z",
     "iopub.status.busy": "2023-07-10T20:45:44.394000Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399459/virsorter/final-viral-combined.fa already exists, skipping...\n",
      "/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399460/virsorter/final-viral-combined.fa already exists, skipping...\n",
      "/home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399461/virsorter/final-viral-combined.fa already exists, skipping...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[2023-07-10 20:45 INFO] VirSorter 2.2.4\n",
      "[2023-07-10 20:45 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-10 20:45 INFO] \"template-config.yaml\" has not been initialized; initializing..\n",
      "[2023-07-10 20:45 INFO] saving /home/jovyan/workspace/virsorter2/db as DBDIR to config file /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml\n",
      "[2023-07-10 20:45 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/virsorter/config.yaml\n",
      "\n",
      "[2023-07-10 20:45 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399462/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-10 20:45 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-10 20:45 INFO] # of circular seqs: 90\n",
      "[2023-07-10 20:45 INFO] # of linear seqs  : 5994\n",
      "[2023-07-10 20:45 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-10 20:45 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-10 20:46 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-10 21:09 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-10 21:10 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t26\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-10 21:10 INFO] Step 3 - classify finished.\n",
      "[2023-07-10 21:10 INFO] VirSorter 2.2.4\n",
      "[2023-07-10 21:10 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399463/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399463/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-10 21:10 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-10 21:10 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399463/virsorter/config.yaml\n",
      "\n",
      "[2023-07-10 21:10 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399463/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399463/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-10 21:10 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-10 21:10 INFO] # of circular seqs: 304\n",
      "[2023-07-10 21:10 INFO] # of linear seqs  : 50153\n",
      "[2023-07-10 21:10 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-10 21:10 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-10 21:11 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-11 13:29 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-11 13:32 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t111\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-11 13:32 INFO] Step 3 - classify finished.\n",
      "[2023-07-11 13:32 INFO] VirSorter 2.2.4\n",
      "[2023-07-11 13:32 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399487/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399487/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-11 13:32 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-11 13:32 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399487/virsorter/config.yaml\n",
      "\n",
      "[2023-07-11 13:32 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399487/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399487/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-11 13:32 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-11 13:32 INFO] # of circular seqs: 385\n",
      "[2023-07-11 13:32 INFO] # of linear seqs  : 30574\n",
      "[2023-07-11 13:32 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-11 13:32 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-11 13:35 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 14:58 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 14:59 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t5\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 14:59 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 14:59 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 14:59 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399497/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399497/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 14:59 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 14:59 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399497/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 14:59 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399497/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399497/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 14:59 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 14:59 INFO] # of circular seqs: 154\n",
      "[2023-07-12 14:59 INFO] # of linear seqs  : 18673\n",
      "[2023-07-12 14:59 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 14:59 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 14:59 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 15:27 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 15:27 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t11\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 15:27 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 15:27 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 15:27 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399498/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399498/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 15:27 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 15:27 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399498/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 15:27 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399498/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399498/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 15:27 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 15:27 INFO] # of circular seqs: 100\n",
      "[2023-07-12 15:27 INFO] # of linear seqs  : 10177\n",
      "[2023-07-12 15:27 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 15:27 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 15:28 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 15:46 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 15:46 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t10\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 15:46 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 15:46 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 15:46 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399499/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399499/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 15:46 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 15:46 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399499/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 15:46 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399499/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399499/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 15:46 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 15:46 INFO] # of circular seqs: 121\n",
      "[2023-07-12 15:46 INFO] # of linear seqs  : 4643\n",
      "[2023-07-12 15:46 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 15:46 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 15:47 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 16:02 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 16:02 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t19\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 16:02 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 16:02 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 16:02 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399500/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399500/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 16:02 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 16:02 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399500/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 16:02 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399500/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399500/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 16:02 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 16:02 INFO] # of circular seqs: 127\n",
      "[2023-07-12 16:02 INFO] # of linear seqs  : 14160\n",
      "[2023-07-12 16:02 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 16:02 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 16:03 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 16:24 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 16:25 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t10\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 16:25 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 16:25 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 16:25 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399501/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399501/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 16:25 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 16:25 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399501/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 16:25 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399501/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399501/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 16:25 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 16:25 INFO] # of circular seqs: 174\n",
      "[2023-07-12 16:25 INFO] # of linear seqs  : 16767\n",
      "[2023-07-12 16:25 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 16:25 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 16:25 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 16:45 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 16:45 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t10\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 16:45 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 16:45 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 16:45 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399502/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399502/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 16:45 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 16:45 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399502/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 16:45 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399502/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399502/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 16:45 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 16:45 INFO] # of circular seqs: 56\n",
      "[2023-07-12 16:45 INFO] # of linear seqs  : 3075\n",
      "[2023-07-12 16:45 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 16:45 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 16:45 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 16:55 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 16:56 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t1\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 16:56 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 16:56 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 16:56 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399503/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399503/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 16:56 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 16:56 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399503/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 16:56 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399503/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399503/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 16:56 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 16:56 INFO] # of circular seqs: 108\n",
      "[2023-07-12 16:56 INFO] # of linear seqs  : 16312\n",
      "[2023-07-12 16:56 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 16:56 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 16:56 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 17:21 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 17:21 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t13\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 17:21 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 17:21 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 17:21 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399504/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399504/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 17:21 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 17:21 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399504/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 17:21 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399504/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399504/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 17:21 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 17:21 INFO] # of circular seqs: 100\n",
      "[2023-07-12 17:21 INFO] # of linear seqs  : 5192\n",
      "[2023-07-12 17:21 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 17:21 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 17:21 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 17:35 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 17:35 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t6\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 17:35 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 17:35 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 17:35 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399505/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399505/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 17:35 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 17:35 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399505/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 17:35 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399505/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399505/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 17:35 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 17:35 INFO] # of circular seqs: 182\n",
      "[2023-07-12 17:35 INFO] # of linear seqs  : 16931\n",
      "[2023-07-12 17:35 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 17:35 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 17:38 INFO] Step 1 - preprocess finished.\n",
      "[2023-07-12 19:08 INFO] Step 2 - extract-feature finished.\n",
      "[2023-07-12 19:11 INFO] \n",
      "            ====> VirSorter run (provirus mode) finished.\n",
      "            # of full    seqs (>=2 genes) as viral:\t107\n",
      "            # of partial seqs (>=2 genes) as viral:\t0\n",
      "            # of short   seqs (< 2 genes) as viral:\t0\n",
      "\n",
      "            Useful output files:\n",
      "                final-viral-score.tsv       ==> score table\n",
      "                final-viral-combined.fa     ==> all viral seqs\n",
      "                final-viral-boundary.tsv    ==> table with boundary info\n",
      "                \n",
      "            \n",
      "                Suffix is added to seq names in final-viral-combined.fa:\n",
      "                full    seqs (>=2 genes) as viral:\t||full\n",
      "                partial seqs (>=2 genes) as viral:\t||partial\n",
      "                short   seqs (< 2 genes) as viral:\t||lt2gene\n",
      "                \n",
      "                \n",
      "            NOTES:\n",
      "            Users can further screen the results based on the following \n",
      "                columns in final-viral-score.tsv:\n",
      "                - contig length (length) \n",
      "                - hallmark gene count (hallmark)\n",
      "                - viral gene % (viral) \n",
      "                - cellular gene % (cellular)\n",
      "            The group field in final-viral-score.tsv should NOT be used\n",
      "                as reliable taxonomy info\n",
      "            We recommend this SOP/tutorial for quality control \n",
      "                (make sure to use the lastest version):\n",
      "                https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
      "\n",
      "            <====\n",
      "            \n",
      "[2023-07-12 19:11 INFO] Step 3 - classify finished.\n",
      "[2023-07-12 19:11 INFO] VirSorter 2.2.4\n",
      "[2023-07-12 19:11 INFO] /opt/saturncloud/envs/saturn/envs/vs2/bin/virsorter run --db-dir /home/jovyan/workspace/virsorter2/db -w /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399506/virsorter -i /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399506/megahit/final.contigs.fastg.gfa.fna --min-length 1 -j 3 all\n",
      "[2023-07-12 19:11 INFO] Using /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/template-config.yaml as config template\n",
      "[2023-07-12 19:11 INFO] conig file written to /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399506/virsorter/config.yaml\n",
      "\n",
      "[2023-07-12 19:11 INFO] Executing: snakemake --snakefile /opt/saturncloud/envs/saturn/envs/vs2/lib/python3.10/site-packages/virsorter/Snakefile --directory /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399506/virsorter --jobs 3 --configfile /home/jovyan/workspace/Mycelia/projects/viral-pangenome-discovery/data/SRA/SRR6399506/virsorter/config.yaml --latency-wait 600 --rerun-incomplete --nolock  --conda-frontend mamba --conda-prefix /home/jovyan/workspace/virsorter2/db/conda_envs --use-conda    --quiet  all   \n",
      "Job counts:\n",
      "\tcount\tjobs\n",
      "\t1\tall\n",
      "\t1\tcheck_point_for_reclassify\n",
      "\t1\tcircular_linear_split\n",
      "\t1\tclassify\n",
      "\t2\tclassify_by_group\n",
      "\t2\tclassify_full_and_part_by_group\n",
      "\t1\tcombine_linear_circular\n",
      "\t2\tcombine_linear_circular_by_group\n",
      "\t1\textract_feature\n",
      "\t1\textract_provirus_seqs\n",
      "\t1\tfinalize\n",
      "\t1\tgff_feature\n",
      "\t2\tgff_feature_by_group\n",
      "\t2\thmm_features_by_group\n",
      "\t1\thmm_sort_to_best_hit_taxon\n",
      "\t2\thmm_sort_to_best_hit_taxon_by_group\n",
      "\t1\tmerge_classification\n",
      "\t1\tmerge_full_and_part_classification\n",
      "\t2\tmerge_hmm_gff_features_by_group\n",
      "\t2\tmerge_provirus_call_by_group_by_split\n",
      "\t1\tmerge_provirus_call_from_groups\n",
      "\t5\tmerge_split_hmmtbl\n",
      "\t10\tmerge_split_hmmtbl_by_group\n",
      "\t10\tmerge_split_hmmtbl_by_group_tmp\n",
      "\t1\tpick_viral_fullseq\n",
      "\t1\tpreprocess\n",
      "\t1\tsplit_faa\n",
      "\t2\tsplit_faa_by_group\n",
      "\t2\tsplit_gff_by_group\n",
      "\t61\n",
      "[2023-07-12 19:11 INFO] # of seqs < 1 bp and removed: 0\n",
      "[2023-07-12 19:11 INFO] # of circular seqs: 1446\n",
      "[2023-07-12 19:11 INFO] # of linear seqs  : 49078\n",
      "[2023-07-12 19:11 INFO] Finish spliting circular contig file with common rbs\n",
      "[2023-07-12 19:11 INFO] Finish spliting linear contig file with common rbs\n",
      "[2023-07-12 19:13 INFO] Step 1 - preprocess finished.\n"
     ]
    }
   ],
   "source": [
    "# SRR_path = SRR_paths[1]\n",
    "# SRR_path = SRR_paths[1]\n",
    "for SRR_path in SRR_paths\n",
    "    SRR = basename(SRR_path)\n",
    "    assembled_fasta = joinpath(SRR_path, \"megahit\", \"final.contigs.fastg.gfa.fna\")\n",
    "    virsorter_dir = mkpath(joinpath(SRR_path, \"virsorter\"))\n",
    "    outfile = joinpath(virsorter_dir, \"final-viral-combined.fa\")\n",
    "    if !isfile(outfile)\n",
    "        run(`conda run --no-capture-output -n vs2 virsorter run --db-dir $(virsorter_db_dir_path) -w $(virsorter_dir) -i $(assembled_fasta) --min-length 1 -j $(Sys.CPU_THREADS-1) all`)\n",
    "    else\n",
    "        println(\"$outfile already exists, skipping...\")\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7cbdf58-fa45-4e34-96ff-a50eef9ce778",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 30 minutes per run\n",
    "\n",
    "# works in standalone conda environment but not general installation\n",
    "# run(`virsorter run -w $(virsorter_dir) -i $(assembled_fasta) --min-length 1 -j $(Sys.CPU_THREADS) all`)\n",
    "\n",
    "virsorter_dir = mkpath(joinpath(SRR_path, \"virsorter\"))\n",
    "\n",
    "outfile = joinpath(virsorter_dir, \"final-viral-combined.fa\")\n",
    "if !isfile(outfile)\n",
    "    run(`conda run --no-capture-output -n vs2 virsorter run -w $(virsorter_dir) -i $(assembled_fasta) --min-length 1 -j $(Sys.CPU_THREADS-1) all`)\n",
    "else\n",
    "    println(\"$outfile already exists, skipping...\")\n",
    "end\n",
    "    \n",
    "# [2023-06-06 16:59 INFO] Step 2 - extract-feature finished.\n",
    "# [2023-06-06 17:00 INFO] \n",
    "#             ====> VirSorter run (provirus mode) finished.\n",
    "#             # of full    seqs (>=2 genes) as viral:     30\n",
    "#             # of partial seqs (>=2 genes) as viral:     0\n",
    "#             # of short   seqs (< 2 genes) as viral:     0\n",
    "\n",
    "#             Useful output files:\n",
    "#                 final-viral-score.tsv       ==> score table\n",
    "#                 final-viral-combined.fa     ==> all viral seqs\n",
    "#                 final-viral-boundary.tsv    ==> table with boundary info\n",
    "                \n",
    "            \n",
    "#                 Suffix is added to seq names in final-viral-combined.fa:\n",
    "#                 full    seqs (>=2 genes) as viral:      ||full\n",
    "#                 partial seqs (>=2 genes) as viral:      ||partial\n",
    "#                 short   seqs (< 2 genes) as viral:      ||lt2gene\n",
    "                \n",
    "                \n",
    "#             NOTES:\n",
    "#             Users can further screen the results based on the following \n",
    "#                 columns in final-viral-score.tsv:\n",
    "#                 - contig length (length) \n",
    "#                 - hallmark gene count (hallmark)\n",
    "#                 - viral gene % (viral) \n",
    "#                 - cellular gene % (cellular)\n",
    "#             The group field in final-viral-score.tsv should NOT be used\n",
    "#                 as reliable taxonomy info\n",
    "#             We recommend this SOP/tutorial for quality control \n",
    "#                 (make sure to use the lastest version):\n",
    "#                 https://dx.doi.org/10.17504/protocols.io.bwm5pc86\n",
    "\n",
    "#             <====\n",
    "            \n",
    "# [2023-06-06 17:00 INFO] Step 3 - classify finished."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c9a1ce9-d833-426c-a81c-1e2b18354da6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.7",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
