{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Neo4J Notes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Schema for a pan-meta-genome graph with annotations (multi-dataset)\n",
    "\n",
    "![pan-meta-genome-graph-with-annotations](pan-meta-genome-graph-with-annotations.png)\n",
    "\n",
    "```\n",
    "//pan-meta-genome-graph-with-annotations\n",
    "CREATE (dnamer:DNAmer {label: \"DNAmer\"})\n",
    "CREATE (sequence:Sequence {label: \"Sequence\"})\n",
    "CREATE (aamer:AAmer {label: \"AAmer\"})\n",
    "CREATE (source:Source {label: \"Source\"})\n",
    "CREATE (environment:Environment {label: \"Environment\"})\n",
    "CREATE (genome:Genome {label: \"Genome\"})\n",
    "CREATE (species:Species {label: \"Species\"})\n",
    "CREATE (genus:Genus {label: \"Genus\"})\n",
    "CREATE (family:Family {label: \"Family\"})\n",
    "CREATE (order:Order {label: \"Order\"})\n",
    "CREATE (subclass:Subclass {label: \"Subclass\"})\n",
    "CREATE (class:Class {label: \"Class\"})\n",
    "CREATE (subphylum:Subphylum {label: \"Subphylum\"})\n",
    "CREATE (phylum:Phylum {label: \"Phylum\"})\n",
    "CREATE (subkingdom:Subkingdom {label: \"Subkingdom\"})\n",
    "CREATE (kingdom:Kingdom {label: \"Kingdom\"})\n",
    "CREATE (clade:Clade {label: \"Clade\"})\n",
    "CREATE (superkingdom:Superkingdom {label: \"Superkingdom\"})\n",
    "CREATE (root:Root {label: \"Root\"})\n",
    "CREATE (annotation:Annotation {label: \"Annotation\"})\n",
    "CREATE (entity:Entity {label: \"Entity\"})\n",
    "CREATE (pathway:Pathway {label: \"Pathway\"})\n",
    "CREATE (annotationclass:AnnotationClass {label:\"AnnotationClass\"})\n",
    "CREATE (host:Entity {label: \"Host\"})\n",
    "\n",
    "CREATE (entity)-[entity_host:HOST_OF]->(host)\n",
    "CREATE (annotation)-[annotation_annotationclass:ISA]->(annotationclass)\n",
    "CREATE (host)-[host_species:ISA]->(species)\n",
    "CREATE (host)-[host_genome:GENOME_OF]->(genome)\n",
    "CREATE (dnamer)-[dnamer_annotation:CONTAINED_IN]->(annotation)\n",
    "CREATE (aamer)-[aamer_annotation:CONTAINED_IN]->(annotation)\n",
    "CREATE (genome)-[genome_annotation:CONTAINED_IN]->(annotation)\n",
    "CREATE (superkingdom)-[superkingdom_root:PARENT]->(root)\n",
    "CREATE (clade)-[clade_superkingdom:PARENT]->(superkingdom)\n",
    "CREATE (kingdom)-[kingdom_clade:PARENT]->(clade)\n",
    "CREATE (subkingdom)-[subkingdom_kingdom:PARENT]->(kingdom)\n",
    "CREATE (phylum)-[phylum_subkingdom:PARENT]->(subkingdom)\n",
    "CREATE (subphylum)-[subphylum_phylum:PARENT]->(phylum)\n",
    "CREATE (class)-[class_subphylum:PARENT]->(subphylum)\n",
    "CREATE (subclass)-[subclass_class:PARENT]->(class)\n",
    "CREATE (order)-[order_subclass:PARENT]->(subclass)\n",
    "CREATE (family)-[family_order:PARENT]->(order)\n",
    "CREATE (genus)-[genus_family:PARENT]->(family)\n",
    "CREATE (species)-[species_genus:PARENT]->(genus)\n",
    "CREATE (entity)-[entity_species:ISA]->(species)\n",
    "CREATE (dnamer)-[dnamer_genome:CONTAINED_IN]->(genome)\n",
    "CREATE (source)-[source_environment:SOURCED_FROM]->(environment)\n",
    "CREATE (sequence)-[sequence_source:CONTAINED_IN]->(source)\n",
    "CREATE (aamer)-[aamer_sequence:CONTAINED_IN]->(sequence)\n",
    "CREATE (dnamer)-[dnamer_sequence:CONTAINED_IN]->(sequence)\n",
    "CREATE (dnamer)-[dnamer_aamer:TRANSLATES_TO]->(aamer)\n",
    "CREATE (aamer)-[aamer_genome:CONTAINED_IN]->(genome)\n",
    "CREATE (genome)-[genome_entity:GENOME_OF]->(entity)\n",
    "CREATE (annotation)-[annotation_pathway:CONTAINED_IN]->(pathway)\n",
    "\n",
    "RETURN dnamer,\n",
    "sequence,\n",
    "aamer,\n",
    "source,\n",
    "environment,\n",
    "genome,\n",
    "species,\n",
    "genus,\n",
    "family,\n",
    "order,\n",
    "subclass,\n",
    "class,\n",
    "subphylum,\n",
    "phylum,\n",
    "subkingdom,\n",
    "kingdom,\n",
    "clade,\n",
    "superkingdom,\n",
    "root,\n",
    "annotation,\n",
    "entity,\n",
    "pathway,\n",
    "annotation_pathway,\n",
    "annotationclass,\n",
    "host,\n",
    "entity_host,\n",
    "annotation_annotationclass,\n",
    "host_species,\n",
    "host_genome\n",
    "```\n",
    "\n",
    "## Example Queries\n",
    "\n",
    "English\n",
    "```\n",
    "CYPHER\n",
    "```\n",
    "\n",
    "- Give me the genome of entity x\n",
    "- Give me the pangenome of species x\n",
    "- Give me the pangenome of order x\n",
    "- Give me all kmers in species x\n",
    "\n",
    "- Finding spacers:\n",
    "- Give me all kmers in species x matching pattern y\n",
    "- Give me all datasets containing species x\n",
    "- Give me all datasets containing pathway x\n",
    "- Give me all reads supporting path/variant x\n",
    "\n",
    "## Creating a blank database\n",
    "\n",
    "https://neo4j.com/developer/neo4j-desktop/#desktop-create-project\n",
    "\n",
    "## Interacting with a Database\n",
    "\n",
    "- username: neo4j\n",
    "- password: password\n",
    "    - I set this and yours may be different!\n",
    "   \n",
    "\n",
    "https://neo4j.com/developer/manage-multiple-databases/\n",
    "\n",
    "https://stackoverflow.com/a/29658062\n",
    "https://neo4j.com/docs/operations-manual/current/tools/cypher-shell/#cypher-shell-syntax\n",
    "https://contentaudience.com/guides/neo4j-cli-cypher-shell/\n",
    "https://neo4j.com/developer/kb/how-do-i-authenticate-with-cypher-shell-without-specifying-the-username-and-password-on-the-command-line/\n",
    "\n",
    "[Bulk import](https://neo4j.com/developer/guide-import-csv/)\n",
    "\n",
    "## Database\n",
    "\n",
    "- Basic Alphabets\n",
    "    - Canonical Nucleic Acids\n",
    "    - Canonical Amino Acids\n",
    "- Types of paths\n",
    "    - DNAmers\n",
    "        - DNAmers are canonical dna sequences of k-length\n",
    "        - we track all odd primes between 1 and 31 for DNA mers\n",
    "        - we track DNAmers 3, 9, 15, and 21 for interoperability between AAmers and the DNAmers they translate from\n",
    "        - kmers store both their full sequences AND their sequences as paths through shorter kmers\n",
    "    - AAmers\n",
    "        - we track all odd primes between 1 and 5 for AA mers\n",
    "        - kmers store both their full sequences AND their sequences as paths through shorter kmers\n",
    "    - Genomes\n",
    "        - genomes are paths through DNA kmers\n",
    "    - annotations\n",
    "        - annotations are paths through DNA kmers and possibly also through AA kmers\n",
    "\n",
    "## Learning to use Cyper CLI\n",
    "```\n",
    "usage: cypher-shell [-h] [-a ADDRESS] [-u USERNAME] [-p PASSWORD] [--encryption {true,false,default}] [-d DATABASE]\n",
    "                    [--format {auto,verbose,plain}] [-P PARAM] [--debug] [--non-interactive] [--sample-rows SAMPLE-ROWS]\n",
    "                    [--wrap {true,false}] [-v] [--driver-version] [-f FILE] [--fail-fast | --fail-at-end] [cypher]\n",
    "\n",
    "A command line shell where you can execute Cypher against an  instance  of  Neo4j.  By default the shell is interactive but you can use it\n",
    "for scripting by passing cypher directly on the command line or by piping a file with cypher statements (requires Powershell on Windows).\n",
    "\n",
    "example of piping a file:\n",
    "  cat some-cypher.txt | cypher-shell\n",
    "\n",
    "positional arguments:\n",
    "  cypher                 an optional string of cypher to execute and then exit\n",
    "\n",
    "optional arguments:\n",
    "  -h, --help             show this help message and exit\n",
    "  --fail-fast            exit and report failure on first error when reading from file (this is the default behavior)\n",
    "  --fail-at-end          exit and report failures at end of input when reading from file\n",
    "  --format {auto,verbose,plain}\n",
    "                         desired output format, verbose displays results  in  tabular  format  and  prints statistics, plain displays data\n",
    "                         with minimal formatting (default: auto)\n",
    "  -P PARAM, --param PARAM\n",
    "                         Add a parameter to this session. Example: `-P \"number => 3\"`. This argument can be specified multiple times.\n",
    "  --debug                print additional debug information (default: false)\n",
    "  --non-interactive      force non-interactive mode, only useful if auto-detection fails (like on Windows) (default: false)\n",
    "  --sample-rows SAMPLE-ROWS\n",
    "                         number of rows sampled to compute table widths (only for format=VERBOSE) (default: 1000)\n",
    "  --wrap {true,false}    wrap table column values if column is too narrow (only for format=VERBOSE) (default: true)\n",
    "  -v, --version          print version of cypher-shell and exit (default: false)\n",
    "  --driver-version       print version of the Neo4j Driver used and exit (default: false)\n",
    "  -f FILE, --file FILE   Pass a file with cypher statements to be executed.  After  the statements have been executed cypher-shell will be\n",
    "                         shutdown\n",
    "\n",
    "connection arguments:\n",
    "  -a ADDRESS, --address ADDRESS\n",
    "                         address and port to connect to (default: neo4j://localhost:7687)\n",
    "  -u USERNAME, --username USERNAME\n",
    "                         username to connect as. Can also be specified using environment variable NEO4J_USERNAME (default: )\n",
    "  -p PASSWORD, --password PASSWORD\n",
    "                         password to connect with. Can also be specified using environment variable NEO4J_PASSWORD (default: )\n",
    "  --encryption {true,false,default}\n",
    "                         whether the connection to Neo4j should  be  encrypted.  This  must  be  consistent with Neo4j's configuration. If\n",
    "                         choosing 'default' the encryption setting is  deduced  from  the  specified  address. For example the 'neo4j+ssc'\n",
    "                         protocol would use encryption. (default: default)\n",
    "  -d DATABASE, --database DATABASE\n",
    "                         database to connect to. Can also be specified using environment variable NEO4J_DATABASE (default: )\n",
    "```\n",
    "\n",
    "List databases\n",
    "```\n",
    "# verbose outputs mdtable with pretty printing\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database system --format verbose 'show databases'\n",
    "# plain outputs a CSV\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database system --format plain 'show databases'\n",
    "# auto uses verbose\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database system --format auto 'show databases'\n",
    "```\n",
    "\n",
    "Creating a database for a specific dataset\n",
    "```\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database system --format auto 'create database test'\n",
    "```\n",
    "\n",
    "Show schema\n",
    "```\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto 'CALL db.schema.visualization()'\n",
    "```\n",
    "\n",
    "Return all nodes\n",
    "```\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto 'MATCH (n) RETURN n'\n",
    "```\n",
    "\n",
    "Delete entire graph\n",
    "```\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto 'MATCH (n) DETACH DELETE n'\n",
    "```\n",
    "\n",
    "Add kmers and connections in one go\n",
    "```\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto 'MERGE (d1:DNAmer {sequence: \"A\"})-[d1d2:CONNECTION {orientations: [1,1]}]->(d2:DNAmer {sequence: \"T\"}) RETURN *'\n",
    "```\n",
    "\n",
    "Add kmers and connections in separate steps\n",
    "```\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto 'MERGE (:DNAmer {sequence: \"A\"})'\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto 'MERGE (:DNAmer {sequence: \"T\"})'\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto 'MERGE (d1:DNAmer {sequence: \"A\"})-[d1d2:CONNECTION {orientations: [1,1]}]->(d2:DNAmer {sequence: \"T\"}) RETURN *'\n",
    "```\n",
    "\n",
    "Find the import folder on \n",
    "```\n",
    "# https://neo4j.com/docs/operations-manual/current/configuration/file-locations/\n",
    "/Users/cameronprybol/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-8ab8baac-5dea-4137-bb24-e0b426447940/import\n",
    "```\n",
    "\n",
    "find all nmers\n",
    "```\n",
    "cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto 'MATCH (d:DNAmer) WHERE size(d.sequence) = 1 return d'\n",
    "```\n",
    "\n",
    "create import uniqueness constraints\n",
    "```\n",
    "CREATE CONSTRAINT ON (d:Dnamer) ASSERT d.sequence IS UNIQUE;\n",
    "CREATE CONSTRAINT ON (f:Fasta) ASSERT f.identifier IS UNIQUE;\n",
    "CREATE CONSTRAINT ON (f:Fastq) ASSERT f.identifier IS UNIQUE;\n",
    "```\n",
    "\n",
    "data import\n",
    "```\n",
    "USING PERIODIC COMMIT\n",
    "LOAD CSV WITH HEADERS FROM\n",
    "'file:///path/to/file' AS line\n",
    "with line\n",
    "\n",
    "CREATE (n:Node {id: line.`id`})\n",
    "```\n",
    "\n",
    "data conversions for loading CSV\n",
    "```\n",
    "TOINT\n",
    "TOFLOAT\n",
    "https://neo4j.com/docs/cypher-manual/current/functions/\n",
    "```\n",
    "\n",
    "\n",
    "merge nodes to avoid creating again\n",
    "\n",
    "!important!\n",
    "\n",
    "When merging new information into nodes, need to first match on unique key to get all existing fields.\n",
    "Merging matches on all fields, and merging on partial matches will duplicate nodes which we don't want"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.3",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
