{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TODAY=\"2021-11-13\"\n",
    "TASK = \"phylogenetic-determination\"\n",
    "DIR = \"$(homedir())/$(TODAY)-$(TASK)\"\n",
    "if !isdir(DIR)\n",
    "    mkdir(DIR)\n",
    "end\n",
    "cd(DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# alternative workflow would be to utilize ezAAI http://leb.snu.ac.kr/ezaai\n",
    "# or recreate ezAAI with diamond\n",
    "\n",
    "# use the genus assignment of the phage, and then assume the family level assignment from the genus classification\n",
    "# require that scale dependent, correlation, and probability based genus is consisent, otherwise flag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Pkg\n",
    "\n",
    "pkgs = [\n",
    "\"JSON\",\n",
    "\"HTTP\",\n",
    "\"Dates\",\n",
    "\"uCSV\",\n",
    "\"DelimitedFiles\",\n",
    "\"DataFrames\",\n",
    "\"ProgressMeter\",\n",
    "\"BioSequences\",\n",
    "\"FASTX\",\n",
    "\"Distances\",\n",
    "\"Plots\",\n",
    "\"StatsPlots\",\n",
    "\"StatsBase\",\n",
    "\"Statistics\",\n",
    "\"Mmap\",\n",
    "\"MultivariateStats\",\n",
    "\"PyCall\",\n",
    "\"Random\",\n",
    "\"Primes\",\n",
    "\"Revise\",\n",
    "\"SparseArrays\",\n",
    "\"SHA\",\n",
    "\"Mycelia\",\n",
    "\"GenomicAnnotations\",\n",
    "\"BioFetch\",\n",
    "\"Combinatorics\",\n",
    "\"StaticArrays\",\n",
    "\"BioSymbols\",\n",
    "\"RollingFunctions\",\n",
    "\"OrderedCollections\"\n",
    "]\n",
    "\n",
    "for pkg in pkgs\n",
    "    try\n",
    "        eval(Meta.parse(\"import $pkg\"))\n",
    "    catch\n",
    "        Pkg.add(pkg)\n",
    "        eval(Meta.parse(\"import $pkg\"))\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function generate_all_possible_kmers(k, alphabet)\n",
    "    kmer_iterator = Iterators.product([alphabet for i in 1:k]...)\n",
    "    kmer_vectors = collect.(vec(collect(kmer_iterator)))\n",
    "    if eltype(alphabet) == BioSymbols.AminoAcid\n",
    "        kmers = BioSequences.LongAminoAcidSeq.(kmer_vectors)\n",
    "    elseif eltype(alphabet) == BioSymbols.DNA\n",
    "        kmers = BioSequences.LongDNASeq.(kmer_vectors)\n",
    "    else\n",
    "        error()\n",
    "    end\n",
    "    return sort!(kmers)\n",
    "end\n",
    "\n",
    "function generate_all_possible_canonical_kmers(k, alphabet)\n",
    "    kmers = generate_all_possible_kmers(k, alphabet)\n",
    "    if eltype(alphabet) == BioSymbols.AminoAcid\n",
    "        return kmers\n",
    "    elseif eltype(alphabet) == BioSymbols.DNA\n",
    "        return BioSequences.DNAMer.(unique!(BioSequences.canonical.(kmers)))\n",
    "    else\n",
    "        error()\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function count_canonical_aamers(k, fasta_proteins)\n",
    "    aamer_counts = OrderedCollections.OrderedDict{BioSequences.LongAminoAcidSeq, Int64}()\n",
    "    for protein in fasta_proteins\n",
    "        s = FASTX.sequence(protein)\n",
    "        these_counts = sort(StatsBase.countmap([s[i:i+k-1] for i in 1:length(s)-k-1]))\n",
    "        merge!(+, aamer_counts, these_counts)\n",
    "    end\n",
    "    return sort(aamer_counts)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function update_counts_matrix!(matrix, sample_index, countmap, sorted_kmers)\n",
    "    for (i, kmer) in enumerate(sorted_kmers)\n",
    "        matrix[i, sample_index] = get(countmap, kmer, 0)\n",
    "    end\n",
    "    return matrix\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function accession_list_to_counts_table(;accession_list, k, alphabet, outfile=\"\")\n",
    "    if alphabet == :AA\n",
    "        canonical_mers = generate_all_possible_canonical_kmers(k, AA_ALPHABET)\n",
    "    elseif alphabet == :DNA\n",
    "        canonical_mers = generate_all_possible_canonical_kmers(k, DNA_ALPHABET)\n",
    "    else\n",
    "        error(\"invalid alphabet\")\n",
    "    end\n",
    "    if isempty(outfile)\n",
    "        outfile = joinpath(pwd(), \"$(hash(accession_list)).$(alphabet).k$(k).bin\")\n",
    "    end\n",
    "    if isfile(outfile)\n",
    "        println(\"$outfile found, loading into memory\")\n",
    "        mer_counts_matrix = Mmap.mmap(outfile, Array{Int, 2}, (length(canonical_mers), length(accession_list)))\n",
    "    else\n",
    "        println(\"creating new counts matrix $outfile\")\n",
    "        mer_counts_matrix = Mmap.mmap(outfile, Array{Int, 2}, (length(canonical_mers), length(accession_list)))\n",
    "        mer_counts_matrix .= 0\n",
    "        ProgressMeter.@showprogress for (entity_index, accession) in enumerate(accession_list)\n",
    "            fna_file = \"$(accession).fna\"\n",
    "            if !isfile(fna_file)\n",
    "                open(FASTX.FASTA.Writer, fna_file) do fastx_io\n",
    "                    for record in Mycelia.get_sequence(db=\"nuccore\", accession = accession)\n",
    "                        write(fastx_io, record)\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "            if alphabet == :DNA\n",
    "                entity_mer_counts = Mycelia.count_canonical_kmers(BioSequences.DNAMer{dna_k}, fna_file)\n",
    "            elseif alphabet == :AA\n",
    "                faa_file = \"$(accession).fna.faa\"\n",
    "                if !isfile(faa_file)\n",
    "                    run(pipeline(`prodigal -i $(fna_file) -o $(fna_file).genes -a $(faa_file) -p meta`, stderr=\"$(fna_file).prodigal.stderr\"))\n",
    "                end\n",
    "                entity_mer_counts = count_canonical_aamers(aa_k, collect(FASTX.FASTA.Reader(open(faa_file))))\n",
    "            end\n",
    "            update_counts_matrix!(mer_counts_matrix, entity_index, entity_mer_counts, canonical_mers)            \n",
    "        end\n",
    "    end\n",
    "    return mer_counts_matrix, outfile\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function normalize_distance_matrix(distance_matrix)\n",
    "    max_non_nan_value = maximum(filter(x -> !isnan(x) && !isnothing(x) && !ismissing(x), vec(distance_matrix)))\n",
    "    return distance_matrix ./ max_non_nan_value\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function count_matrix_to_probability_matrix(counts_matrix, counts_matrix_file)\n",
    "    probability_matrix_file = replace(counts_matrix_file, \".bin\" => \".probability_matrix.bin\")\n",
    "    already_there = isfile(probability_matrix_file)\n",
    "    probability_matrix = Mmap.mmap(probability_matrix_file, Array{Float64, 2}, size(counts_matrix))\n",
    "    if !already_there\n",
    "        println(\"creating new probability matrix $probability_matrix_file\")\n",
    "        for (i, col) in enumerate(eachcol(counts_matrix))\n",
    "            probability_matrix[:, i] .= col ./ sum(col)\n",
    "        end\n",
    "    else\n",
    "        println(\"probability matrix found $probability_matrix_file\")\n",
    "    end\n",
    "    return probability_matrix, probability_matrix_file\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MYCELIA_METADATA = joinpath(Pkg.dir(\"Mycelia\"), \"metadata\")\n",
    "MYCELIA_METADATA = joinpath(dirname(dirname(pathof(Mycelia))),  \"metadata\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# AA_ALPHABET = collect(filter(x -> x != BioSequences.AA_Term, Mycelia.AA_ALPHABET))\n",
    "AA_ALPHABET = Mycelia.AA_ALPHABET\n",
    "DNA_ALPHABET = Mycelia.DNA_ALPHABET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = \n",
    "DataFrames.DataFrame(\n",
    "    n = Int[],\n",
    "    grammar = String[],\n",
    "    distance_metric = [],\n",
    "    genus_accuracy = Float64[],\n",
    "    family_accuracy = Float64[],\n",
    "    min_distances = Vector{Float64}[]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cosine distance is slower than correlation distance\n",
    "# city block is faster than euclidean\n",
    "# total variation is fastest probability measure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# kill_js_divergence\n",
    "distance_matrix_map = Dict(\n",
    "    Distances.euclidean => \"count\",\n",
    "    Distances.cityblock => \"count\",\n",
    "    Distances.corr_dist => \"count\",\n",
    "    Distances.cosine_dist => \"count\",\n",
    "    Distances.totalvariation => \"probability\",\n",
    "#     Distances.js_divergence => \"probability\", too slow\n",
    "#     Distances.bhattacharyya => \"probability\",\n",
    "#     Distances.hellinger => \"probability\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# these are too small, all of the within vs between have some disagreement\n",
    "# dna_k = 5\n",
    "# aa_k = 2\n",
    "dna_k = 7\n",
    "aa_k = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide&VirusLineage_ss=Bacteriophage,%20all%20taxids&Completeness_s=complete\n",
    "entity_metadata_full = DataFrames.DataFrame(uCSV.read(\"$(MYCELIA_METADATA)/2021-11-13-ncbi-complete-bacteriophage.csv\", header=1, quotes='\"')...)\n",
    "sort!(entity_metadata_full, \"Accession\")\n",
    "entity_metadata_full = entity_metadata_full[entity_metadata_full[!, \"Nuc_Completeness\"] .== \"complete\", :]\n",
    "# filter down to only include labelled genera\n",
    "entity_metadata_full = entity_metadata_full[.!isempty.(entity_metadata_full[!, \"Genus\"]), :]\n",
    "# filter down to only include genera that are present at least once\n",
    "genera_counts = sort(collect(StatsBase.countmap(entity_metadata_full[!, \"Genus\"])), by=x->x[2], rev=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_2fold_coverage_genera = length(unique(filter(x -> x > 1, last.(genera_counts))))\n",
    "# min_threshold = 2\n",
    "# 2..87\n",
    "# min_threshold = unique(last.(genera_counts))[2]\n",
    "# min_threshold = unique(last.(genera_counts))[3]\n",
    "# min_threshold = unique(last.(genera_counts))[4]\n",
    "# min_threshold = unique(last.(genera_counts))[5]\n",
    "# min_threshold = unique(last.(genera_counts))[6]\n",
    "for n_genera in 2:max_2fold_coverage_genera\n",
    "    min_threshold = unique(last.(genera_counts))[n_genera]\n",
    "    repeat_genera = Set(first.(filter(x -> x[2] >= min_threshold, genera_counts)))\n",
    "    entity_metadata = entity_metadata_full[map(genus -> genus in repeat_genera, entity_metadata_full[!, \"Genus\"]), :]\n",
    "    accession_list = entity_metadata[!, \"Accession\"]\n",
    "\n",
    "    # dnamer_counts_matrix = accession_list_to_dnamer_counts_table(accession_list, dna_k)\n",
    "    # Progress: 100%|█████████████████████████████████████████| Time: 1:36:46m\n",
    "    mkpath(\"$(DIR)/data\")\n",
    "    cd(\"$(DIR)/data\")\n",
    "    # dnamer_counts_matrix = accession_list_to_counts_table(accession_list=accession_list, k=dna_k, alphabet=:DNA)\n",
    "    dnamer_counts_matrix, dnamer_counts_matrix_file = accession_list_to_counts_table(accession_list=accession_list, k=dna_k, alphabet=:DNA)\n",
    "    # dnamer_counts_matrix_file = \"/home/jupyter-cjprybol/2021-11-13-phylogenetic-determination/data/8861628634012713288.DNA.k7.bin\"\n",
    "    cd(DIR)\n",
    "\n",
    "    # aamer_counts_matrix = accession_list_to_aamer_counts_table(accession_list, aa_k, AA_ALPHABET)\n",
    "    # Progress: 100%|█████████████████████████████████████████| Time: 2:08:08\n",
    "    mkpath(\"$(DIR)/data\")\n",
    "    cd(\"$(DIR)/data\")\n",
    "    aamer_counts_matrix, aamer_counts_matrix_file = accession_list_to_counts_table(accession_list=accession_list, k=aa_k, alphabet=:AA)\n",
    "    # aamer_counts_matrix, aamer_counts_matrix_file = accession_list_to_counts_table(accession_list=accession_list, k=aa_k, alphabet=:AA)\n",
    "    # aamer_counts_matrix_file = \"/home/jupyter-cjprybol/2021-11-13-phylogenetic-determination/data/8861628634012713288.AA.k3.bin\"\n",
    "    cd(DIR)\n",
    "    dnamer_probility_matrix, dnamer_probility_matrix_file = count_matrix_to_probability_matrix(dnamer_counts_matrix, dnamer_counts_matrix_file)\n",
    "    aamer_probability_matrix, aamer_probability_matrix_file = count_matrix_to_probability_matrix(aamer_counts_matrix, aamer_counts_matrix_file)\n",
    "    \n",
    "    grammar_matrices_map = Dict(\n",
    "        \"AA\" => Dict(\n",
    "                    \"count\" => aamer_counts_matrix,\n",
    "                    \"probability\" => aamer_probability_matrix, \n",
    "                    ),\n",
    "        \"DNA\" => Dict(\n",
    "                    \"count\" => dnamer_counts_matrix,\n",
    "                    \"probability\" => dnamer_probility_matrix,\n",
    "                    )\n",
    "    )\n",
    "\n",
    "    for distance_metric in keys(distance_matrix_map)\n",
    "        for grammar in keys(grammar_matrices_map)\n",
    "\n",
    "            @show distance_metric, grammar\n",
    "\n",
    "            samples = 1:size(entity_metadata, 1)\n",
    "\n",
    "            size_assessed = results[!, \"n\"] .== length(samples)\n",
    "            grammar_assessed = results[!, \"grammar\"] .== grammar\n",
    "            distance_metric_assessed = map(metric -> metric == distance_metric, results[!, \"distance_metric\"])\n",
    "\n",
    "            sample_already_assessed = any(size_assessed .& grammar_assessed .& distance_metric_assessed)\n",
    "            if sample_already_assessed\n",
    "                @show \"already calculated...\"\n",
    "                continue\n",
    "            end\n",
    "\n",
    "            feature_matrix = grammar_matrices_map[grammar][distance_matrix_map[distance_metric]]\n",
    "            @time distance_matrix = normalize_distance_matrix(Distances.pairwise(distance_metric, feature_matrix, dims=2))\n",
    "            \n",
    "#             @show size(distance_matrix)\n",
    "#             @show size(entity_metadata)\n",
    "\n",
    "            correct_genus_hits = 0\n",
    "            correct_family_hits = 0\n",
    "\n",
    "            min_distances = zeros(length(samples))\n",
    "    #         \n",
    "            ProgressMeter.@showprogress for i in samples\n",
    "                other_indices = setdiff(1:size(entity_metadata, 1), i)\n",
    "                value, index = findmin(distance_matrix[i, other_indices])\n",
    "                min_distances[i] = value\n",
    "                index = other_indices[index]\n",
    "                if entity_metadata[i, \"Genus\"] == entity_metadata[index, \"Genus\"]\n",
    "                    correct_genus_hits += 1\n",
    "                end\n",
    "                if entity_metadata[i, \"Family\"] == entity_metadata[index, \"Family\"]\n",
    "                    correct_family_hits += 1\n",
    "                end\n",
    "            end\n",
    "            row = (\n",
    "                n = length(samples),\n",
    "                grammar = grammar,\n",
    "                distance_metric = distance_metric,\n",
    "                genus_accuracy = correct_genus_hits / length(samples),\n",
    "                family_accuracy = correct_family_hits / length(samples),\n",
    "                min_distances = min_distances\n",
    "            )\n",
    "            push!(results, row)\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for (grammar, k) in (\"AA\" => aa_k, \"DNA\" => dna_k)\n",
    "    is_right_grammar = results[!, \"grammar\"] .== grammar\n",
    "    xs = []\n",
    "    ys = []\n",
    "    for distance_metric in keys(distance_matrix_map)\n",
    "        is_right_distance_metric = map(dm -> dm == distance_metric, results[!, \"distance_metric\"])\n",
    "        indices = is_right_grammar .& is_right_distance_metric\n",
    "        xs_ = results[indices, \"n\"]\n",
    "        ys_ = [results[findfirst(indices .& (results[!, \"n\"] .== x)), \"genus_accuracy\"] for x in xs_]\n",
    "        push!(xs, xs_)\n",
    "        push!(ys, ys_)\n",
    "    end \n",
    "    labels = hcat(replace.(string.(keys(distance_matrix_map)), \"Distances.\" => \"\")...)\n",
    "    scale = 400\n",
    "    p = StatsPlots.scatter(\n",
    "        xs,\n",
    "        ys,\n",
    "#         ylims = (minimum(minimum.(ys)) * .9, 1.2),\n",
    "        ylims = (minimum(minimum.(ys)) * 0.999, maximum(maximum.(ys)) * 1.001),\n",
    "        xlims = (0, maximum(maximum.(xs)) * 1.1),\n",
    "        labels = labels,\n",
    "        legend = :outertopright,\n",
    "        size = (2 * scale, scale),\n",
    "        ylabel = \"accuracy\",\n",
    "        xlabel = \"n_phage\",\n",
    "        margins = StatsPlots.cm,\n",
    "        title = \"$(grammar) k=$k\"\n",
    "    )\n",
    "    display(p)\n",
    "    StatsPlots.savefig(p, \"$DIR/$(grammar)-k$(k)-distance-metric-accuracy.svg\")\n",
    "    StatsPlots.savefig(p, \"$DIR/$(grammar)-k$(k)-distance-metric-accuracy.png\")\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# euclidean/city_block for assignment and correlation/total_variation for outlier tresholding??"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for distance_metric in collect(keys(distance_matrix_map))\n",
    "    is_right_distance_metric = map(m -> m == distance_metric, results[!, \"distance_metric\"])\n",
    "    for (grammar, k) in (\"AA\" => aa_k, \"DNA\" => dna_k)\n",
    "        is_right_grammar = results[!, \"grammar\"] .== grammar\n",
    "        xs = []\n",
    "        ys = []\n",
    "        for x in sort(unique(results[!, \"n\"]))\n",
    "            is_right_n = results[!, \"n\"] .== x\n",
    "            i = findfirst(is_right_grammar .& is_right_distance_metric .& is_right_n)\n",
    "            if !isnothing(i)\n",
    "                ys_ = results[i, \"min_distances\"]\n",
    "                # drop the 0s since they aren't helpful\n",
    "                ys_ = filter(y -> y > 0, ys_)\n",
    "                ys_ = StatsBase.sample(ys_, min(100, length(ys_)), replace=false)\n",
    "                push!(ys, ys_)\n",
    "                push!(xs, fill(x, length(ys_)))\n",
    "            end\n",
    "        end\n",
    "#         ys = [r/esults[findfirst(indices .& ()), \"min_distances\"] for x in xs]\n",
    "        p = StatsPlots.scatter(\n",
    "            xs,\n",
    "            ys,\n",
    "            xlabel = \"n phage\",\n",
    "            ylabel = \"distance\",\n",
    "#             legend = false,\n",
    "            labels= \"\",\n",
    "            title = \"closest distance $(distance_metric)\\ngrammar=$(grammar) k=$k\"\n",
    "        )\n",
    "        m = Statistics.mean(reduce(vcat, ys))\n",
    "        σ = Statistics.std(reduce(vcat, ys))\n",
    "        σ3 = round(m+3σ, digits=3)\n",
    "        σ4 = round(m+4σ, digits=3)\n",
    "        σ5 = round(m+5σ, digits=3)\n",
    "        p = StatsPlots.hline!(p,\n",
    "            [[σ3], [σ4], [σ5]],\n",
    "            labels = [\"3σ @ $(σ3)\" \"4σ @ $(σ4)\" \"5σ @ $(σ5)\"]\n",
    "        )\n",
    "        display(p)\n",
    "        StatsPlots.savefig(p, \"$DIR/$(grammar)-k$(k)-$(distance_metric)-closest-neighbor.png\")\n",
    "        StatsPlots.savefig(p, \"$DIR/$(grammar)-k$(k)-$(distance_metric)-closest-neighbor.svg\")\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# AA 2\n",
    "# DNA 5\n",
    "\n",
    "# total variation\n",
    "# species 0.1\n",
    "# genus 0.2\n",
    "# family 0.3\n",
    "\n",
    "# correlation\n",
    "# species 0.2\n",
    "# genus 0.4\n",
    "# family 0.6\n",
    "\n",
    "# cosine\n",
    "# species 0.1\n",
    "# genus 0.2\n",
    "# family 0.3\n",
    "\n",
    "# euclidean\n",
    "# city block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# corr_distance = 0:2\n",
    "# cosine and total variation = 0:1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Distances.cosine_dist(a, b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Distances.cosine_dist(a ./ sum(a), b ./ sum(b))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Distances.totalvariation(a ./ sum(a), b ./ sum(b))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.3",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
