{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca88122e-f127-4334-a638-926203ac03da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ENV[\"LD_LIBRARY_PATH\"] = \"\"\n",
    "\n",
    "# import Pkg\n",
    "# Pkg.activate(\".\")\n",
    "\n",
    "# pkgs = [\n",
    "# # \"BioSequences\",\n",
    "# # \"DataFrames\",\n",
    "# # \"Dates\",\n",
    "# # \"DelimitedFiles\",\n",
    "# # \"FASTX\",\n",
    "# # \"GLM\",\n",
    "# # \"HTTP\",\n",
    "# # \"JSON\",\n",
    "# # \"Graphs\",\n",
    "# # \"MetaGraphs\",\n",
    "# # \"MD5\",\n",
    "# # \"Statistics\",\n",
    "# # \"StatsPlots\",\n",
    "# # \"uCSV\",\n",
    "# # \"CodecZlib\",\n",
    "# # \"YAML\",\n",
    "# \"Revise\",\n",
    "# # \"Kmers\",\n",
    "# # \"StatsBase\",\n",
    "# # \"ProgressMeter\"\n",
    "# ]\n",
    "# Pkg.add(pkgs)\n",
    "# for pkg in pkgs\n",
    "#     eval(Meta.parse(\"import $pkg\"))\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e93a6237-5cf6-4d6e-9169-3c966a481bfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_directory = joinpath(dirname(pwd()), \"data\")\n",
    "sample_directories = readdir(data_directory, join=true)\n",
    "trim_galore_directories = filter(x -> isdir(x) && !occursin(\".ipynb_checkpoints\", x), reduce(vcat, readdir.(sample_directories, join=true)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c7ba3db-0c63-4c2b-8b27-3c97d72f7d6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # QC filtered read sets\n",
    "# for trim_galore_directory in trim_galore_directories\n",
    "#     forward = first(filter(x -> occursin(r\"_1_val_1.fq.gz\",x), readdir(trim_galore_directory, join=true)))\n",
    "#     reverse = first(filter(x -> occursin(r\"_2_val_2.fq.gz\",x), readdir(trim_galore_directory, join=true)))\n",
    "#     out_dir = joinpath(trim_galore_directory, replace(basename(trim_galore_directory), \"_trimgalore\" => \"_megahit\"))\n",
    "\n",
    "#     if isfile(joinpath(out_dir, \"final.contigs.fa\")) && isfile(joinpath(out_dir, \"done\"))\n",
    "#         @show \"$(out_dir) is done\"\n",
    "#     else\n",
    "#         run(`sbatch 1.0.megahit.sh $(out_dir) $(forward) $(reverse)`)\n",
    "#     end\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b62bf66a-3097-4159-9198-c02238c6f4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# QC and human filtered read sets\n",
    "for trim_galore_directory in trim_galore_directories\n",
    "    forward = first(filter(x -> occursin(r\"unmapped\\.1\\.fq\\.gz\", x), readdir(trim_galore_directory, join=true)))\n",
    "    reverse = first(filter(x -> occursin(r\"unmapped\\.2\\.fq\\.gz\", x), readdir(trim_galore_directory, join=true)))\n",
    "    out_dir = replace(forward, \".1.fq.gz\" => \"\") * \"_megahit\"\n",
    "    if isfile(joinpath(out_dir, \"final.contigs.fa\")) && isfile(joinpath(out_dir, \"done\"))\n",
    "        @show \"$(out_dir) is done\"\n",
    "    else\n",
    "        if isdir(out_dir)\n",
    "            rm(out_dir, recursive=true)\n",
    "        end\n",
    "        run(`sbatch 1.0.megahit.sh $(out_dir) $(forward) $(reverse)`)\n",
    "        # print(`sbatch 1.0.megahit.sh $(out_dir) $(forward) $(reverse)`)\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e52dd7a3-e46f-40c6-a912-8d1776c4d4c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# isdir(out_dir) && rm(out_dir, recursive=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc1e3827-ae8a-42cf-9b5c-ac5ea79d0ce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run(`conda create -n megahit -c bioconda megahit`)\n",
    "# ncores = 2\n",
    "# run(`conda run --live-stream --no-capture-output -n megahit megahit --num-cpu-threads $(ncores) -1 $(forward) -2 $(reverse) -o $(out_dir)`)\n",
    "\n",
    "# ncores = 2\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94751fa5-8437-4d73-ac74-032d44056596",
   "metadata": {},
   "outputs": [],
   "source": [
    "#     initial_assembled_fasta = \"$(out_dir)/final.contigs.fa\"\n",
    "#     assembled_fastg = replace(initial_assembled_fasta, \".fa\" => \".fastg\")\n",
    "\n",
    "#     if !isfile(initial_assembled_fasta)\n",
    "#         run(`megahit --num-cpu-threads $(n_cores) -1 $trimmed_forward_reads -2 $trimmed_reverse_reads -o $(out_dir)`)\n",
    "#         # run(`conda run --live-stream --no-capture-output -n viral-pangenome-discovery megahit --num-cpu-threads $(n_cores) -1 $trimmed_forward_reads -2 $trimmed_reverse_reads -o $(out_dir)`)\n",
    "#     end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f56746fd-6258-4912-b064-dde22bfdaa2a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#     # read in the assembled fasta file and parse contig identifiers to get final k length\n",
    "#     final_k_lengths = unique([replace(first(split(FASTX.identifier(record), '_')), r\"^k\" => \"\") for record in Mycelia.open_fastx(initial_assembled_fasta)])\n",
    "#     @assert length(final_k_lengths) == 1\n",
    "#     final_k_length = parse(Int, first(final_k_lengths))\n",
    "#     if !isfile(assembled_fastg)\n",
    "#         run(pipeline(`megahit_toolkit contig2fastg $(final_k_length) $(initial_assembled_fasta)`, assembled_fastg))\n",
    "#     end\n",
    "\n",
    "#     assembled_gfa = \"$(assembled_fastg).gfa\"\n",
    "#     if !isfile(assembled_gfa)\n",
    "#         run(`Bandage reduce $(assembled_fastg) $(assembled_gfa)`)\n",
    "#     end\n",
    "\n",
    "#     assembled_fasta = assembled_gfa * \".fna\"\n",
    "#     if !isfile(assembled_fasta)\n",
    "#         open(assembled_fasta, \"w\") do io\n",
    "#             fastx_io = FASTX.FASTA.Writer(io)\n",
    "#             gfa_graph = Mycelia.parse_gfa(assembled_gfa)\n",
    "#             for v in Graphs.vertices(gfa_graph)\n",
    "#                 record = FASTX.FASTA.Record(gfa_graph.vprops[v][:identifier], gfa_graph.vprops[v][:sequence])\n",
    "#                 write(fastx_io, record)\n",
    "#             end\n",
    "#             close(fastx_io)\n",
    "#         end\n",
    "#     end\n",
    "\n",
    "#     # generate a bandage plot of the assembly graph\n",
    "#     bandage_outfile = \"$(assembled_gfa).bandage.jpg\"\n",
    "#     if !isfile(bandage_outfile)\n",
    "#         run(`Bandage image $(assembled_gfa) $bandage_outfile`)\n",
    "#     end\n",
    "\n",
    "#     # map reads to the assembly and run qualimap QC\n",
    "#     bwt_index = \"$(assembled_fasta).bwt\"\n",
    "#     if !isfile(bwt_index)\n",
    "#         run(`bwa index $(assembled_fasta)`)\n",
    "#     end\n",
    "\n",
    "#     # mapped_reads_bam = \"$(assembled_fasta).bwa.sorted.marked_duplicates.bam\"\n",
    "#     mapped_reads_bam = \"$(assembled_fasta).bwa.bam\"\n",
    "#     # sorting and deduping\n",
    "#     # if !isfile(mapped_reads_bam)\n",
    "#     #     run(pipeline(\n",
    "#     #         `bwa mem -R \"@RG\\tID:$(SRR)\\tSM:bar\" -t $(n_cores) $(assembled_fasta) $(trimmed_forward_reads) $(trimmed_reverse_reads)`,\n",
    "#     #         `samtools collate --threads $(n_cores) -O - -`,\n",
    "#     #         `samtools fixmate --threads $(n_cores) -m - -`,\n",
    "#     #         `samtools sort --threads $(n_cores)`,\n",
    "#     #         `samtools markdup --threads $(n_cores) - -`,\n",
    "#     #         `samtools view --threads $(n_cores) -buh`,\n",
    "#     #         mapped_reads_bams))\n",
    "#     # end\n",
    "#     # sorting - 50Gb - 1594.823 sec\n",
    "#     if !isfile(mapped_reads_bam) || (filesize(mapped_reads_bam) == 0)\n",
    "#         run(pipeline(\n",
    "#             `bwa mem -R \"@RG\\tID:$(SRR)\\tSM:bar\" -t $(n_cores) $(assembled_fasta) $(trimmed_forward_reads) $(trimmed_reverse_reads)`,\n",
    "#             `samtools sort --threads $(n_cores) -u -m 100M`,\n",
    "#             `samtools view --threads $(n_cores) -bh`,\n",
    "#             mapped_reads_bam))\n",
    "#     end\n",
    "\n",
    "#     # if !isfile(\"$(mapped_reads_bam).bai\")\n",
    "#     #     run(`samtools index $(mapped_reads_bam)`)\n",
    "#     # end\n",
    "\n",
    "#     qualimap_report_pdf = \"$(out_dir)/qualimap/report.pdf\"\n",
    "#     qualimap_report_txt = \"$(out_dir)/qualimap/genome_results.txt\"\n",
    "\n",
    "#     if !isfile(qualimap_report_pdf) || !isfile(qualimap_report_txt)\n",
    "#         run(`\n",
    "#             conda run --live-stream -n viral-pangenome-discovery\n",
    "#             qualimap bamqc\n",
    "#             -nt $(n_cores)\n",
    "#             -bam $(mapped_reads_bam)\n",
    "#             -outdir $(out_dir)/qualimap\n",
    "#             -outformat PDF:HTML\n",
    "#             --output-genome-coverage $(mapped_reads_bam).genome_coverage.txt\n",
    "#             --java-mem-size=4G\n",
    "#             `)\n",
    "#     end\n",
    "# end\n",
    "\n",
    "# # qualimap_contig_coverage_table = Mycelia.parse_qualimap_contig_coverage(qualimap_report_txt)\n",
    "# # qualimap_contig_coverage_table[!, \"% Mapped bases\"] = round.(qualimap_contig_coverage_table[!, \"Mapped bases\"] ./ sum(qualimap_contig_coverage_table[!, \"Mapped bases\"]) .* 100, digits=3);"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.7",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
