{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce23d7a1-3e01-467b-8a3b-3af4f2fa92a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if hit plotting library issues, try resetting LD path for julia\n",
    "# can set in ~/.local/share/jupyter/kernels/\n",
    "@assert ENV[\"LD_LIBRARY_PATH\"] == \"\"\n",
    "import Pkg\n",
    "Pkg.activate(;temp=true)\n",
    "Pkg.add(\"Revise\")\n",
    "import Revise\n",
    "\n",
    "# Pkg.develop(path=\"/global/cfs/projectdirs/m4269/cjprybol/Mycelia\")\n",
    "# Pkg.develop(path=\"../../..\")\n",
    "Pkg.develop(path=\"$(homedir())/workspace/Mycelia\")\n",
    "import Mycelia\n",
    "\n",
    "pkgs = String[\n",
    "    \"DataFrames\",\n",
    "    \"uCSV\",\n",
    "    \"OrderedCollections\",\n",
    "    \"CSV\"\n",
    "    # \"XAM\",\n",
    "    # \"CodecZlib\"\n",
    "]\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "457402c5-0c5a-4f91-880c-cc7ddad9b8f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = dirname(pwd())\n",
    "data_dir = mkpath(joinpath(project_dir, \"data\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feef2282-3f45-4442-9013-e6627d3577da",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "db = \"nt\"\n",
    "path_to_db = joinpath(homedir(), \"workspace\", \"blastdb\", db)\n",
    "# path_to_db = Mycelia.download_blast_db(db=db, source=\"ncbi\")\n",
    "# compressed_fasta_export = Mycelia.export_blast_db(path_to_db = path_to_db)\n",
    "compressed_fasta_export = path_to_db * \".fna.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f24576f-d8d8-42b6-8e83-57ee0e802e52",
   "metadata": {},
   "outputs": [],
   "source": [
    "threads = 12\n",
    "mem_gb = threads * 32\n",
    "# mem_gb = # Int(floor(mem_gb * .8))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ad68efc-ccfd-49d9-bafb-b0ec0709a8d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequencing_runs = readdir(joinpath(data_dir, \"Shared-ME-CFS\", \"destination-path\"), join=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "873348bf-cc94-49a5-80fe-2df35afefdae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO prioritize samples by top level cohort THEN by sample size\n",
    "# 1 FJ\n",
    "# 2 FJMZ\n",
    "# 3 E (external cohort)\n",
    "# 4 UK (validation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cbbdb7d-44cf-4ef3-9f19-c299451b05c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_metadata = DataFrames.DataFrame(forward = String[], reverse = String[], outdir = String[], joint_filesize = Int[])\n",
    "for sequencing_run in sequencing_runs\n",
    "    raw_data_dir = joinpath(sequencing_run, \"raw_data\")\n",
    "    if isdir(raw_data_dir)\n",
    "        sample_directories = readdir(raw_data_dir, join=true)\n",
    "    else\n",
    "        raw_data_dir = joinpath(sequencing_run, \"01.RawData\")\n",
    "        @assert isdir(raw_data_dir)\n",
    "        sample_directories = readdir(raw_data_dir, join=true)\n",
    "    end\n",
    "    for sample_directory in sample_directories\n",
    "        fastq_files = readdir(sample_directory, join=true)\n",
    "        lanes = unique(replace.(fastq_files, r\"_[12]\\.fq\\.gz\" => \"\"))\n",
    "        for lane in lanes\n",
    "            if !occursin(r\"d7c3CaC3$\", lane)\n",
    "                forward = lane * \"_1.fq.gz\"\n",
    "                @assert isfile(forward) forward\n",
    "                reverse = lane * \"_2.fq.gz\"\n",
    "                @assert isfile(reverse) reverse\n",
    "                outdir = replace(lane, \"Shared-ME-CFS/destination-path\" => \"local-ME-CFS\")\n",
    "                joint_filesize = filesize(forward) + filesize(reverse)\n",
    "                row = (;forward, reverse, outdir, joint_filesize)\n",
    "                push!(sample_metadata, row)\n",
    "            else\n",
    "                @show lane\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "sample_metadata\n",
    "\n",
    "# # sort SRA dirs by size so that smallest jobs will run first\n",
    "# sample_metadata = sort(sample_metadata, \"joint_filesize\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3b41ee4-df11-43e9-9c1c-9f347a2aad06",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_metadata[!, \"sample_ID\"] = map(x -> basename(dirname(x)), sample_metadata[!, \"outdir\"])\n",
    "\n",
    "sample_metadata[!, \"cohort_ID\"] = map(x -> first(match(r\"^([A-Z]+).*\", x).captures), sample_metadata[!, \"sample_ID\"])\n",
    "\n",
    "cohort_id_to_priority = Dict(\n",
    "    \"FJ\" => 1,\n",
    "    \"FJMZ\" => 2,\n",
    "    \"E\" => 3,\n",
    "    \"UK\" => 4\n",
    ")\n",
    "sample_metadata[!, \"cohort_priority\"] = map(x -> cohort_id_to_priority[x], sample_metadata[!, \"cohort_ID\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca8f5d66-bbea-4848-bb2f-d005f4b80648",
   "metadata": {},
   "outputs": [],
   "source": [
    "sort!(sample_metadata, [\"cohort_priority\", \"sample_ID\", \"joint_filesize\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd74ea40-d992-466c-b893-6586beed4dbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# i = 1\n",
    "# i = 2\n",
    "# i = 3\n",
    "# i = 4\n",
    "# i = 5\n",
    "# i = 6\n",
    "# i = 7\n",
    "# done through above\n",
    "# next to submit is below\n",
    "i = 8\n",
    "N=min(2^i, DataFrames.nrow(sample_metadata))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b98a8c1b-31aa-48a2-b338-4bddc6c218c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# done_count = 0\n",
    "# for row in DataFrames.eachrow(sample_metadata[1:N, :])\n",
    "#     forward = row[\"forward\"]\n",
    "#     reverse = row[\"reverse\"]\n",
    "#     outdir = row[\"outdir\"]\n",
    "    \n",
    "#     if !isdir(outdir)\n",
    "#         # @show \"making path $(outdir)\"\n",
    "#         mkpath(outdir)\n",
    "#     end\n",
    "    \n",
    "#     map_result = Mycelia.minimap_map_paired_end_with_index(\n",
    "#         fasta = compressed_fasta_export,\n",
    "#         outdir = outdir,\n",
    "#         forward = forward,\n",
    "#         reverse = reverse,\n",
    "#         mem_gb = Int(floor(mem_gb * .8)),\n",
    "#         threads = threads,\n",
    "#         denominator = 6,\n",
    "#         as_string = true\n",
    "#     )\n",
    "    \n",
    "#     outfile = map_result.outfile\n",
    "#     cmd = map_result.cmd\n",
    "    \n",
    "#     # Check if the output file exists\n",
    "#     if !isfile(outfile)\n",
    "#         @show done_count\n",
    "#     else\n",
    "#         done_count += 1\n",
    "#     end\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cba4e342-8238-461e-bbde-4a4bfdfc202e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# udpate to submit 8, 12, 16, etc or something per day\n",
    "N_submissions = 12"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b956df6-3ded-4876-8f56-339b297431fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# In this updated code:\n",
    "# 1. A counter `successful_submissions` is initialized to track the number of successful submissions.\n",
    "# 2. This counter is incremented each time a successful submission is detected (i.e., when the output file exists).\n",
    "# 3. The loop checks if the number of successful submissions has reached `N` and stops the loop if the condition is met.\n",
    "\n",
    "# This way, the code will continue to submit jobs until `N` successful submissions are achieved and then terminate.\n",
    "\n",
    "successful_submissions = 0  # Initialize counter for successful submissions\n",
    "\n",
    "for row in DataFrames.eachrow(sample_metadata[1:N, :])\n",
    "    # Stop the loop if we've reached N successful submissions\n",
    "    if successful_submissions >= N_submissions\n",
    "        println(\"Reached $(N_submissions) successful submissions. Stopping.\")\n",
    "        break\n",
    "    end\n",
    "    \n",
    "    forward = row[\"forward\"]\n",
    "    reverse = row[\"reverse\"]\n",
    "    outdir = row[\"outdir\"]\n",
    "    \n",
    "    if !isdir(outdir)\n",
    "        # @show \"making path $(outdir)\"\n",
    "        mkpath(outdir)\n",
    "    end\n",
    "    \n",
    "    map_result = Mycelia.minimap_map_paired_end_with_index(\n",
    "        fasta = compressed_fasta_export,\n",
    "        outdir = outdir,\n",
    "        forward = forward,\n",
    "        reverse = reverse,\n",
    "        mem_gb = Int(floor(mem_gb * .8)),\n",
    "        threads = threads,\n",
    "        denominator = 6,\n",
    "        as_string = true\n",
    "    )\n",
    "    \n",
    "    outfile = map_result.outfile\n",
    "    cmd = map_result.cmd\n",
    "    \n",
    "    # Check if the output file exists\n",
    "    if !isfile(outfile)\n",
    "        println(\"submitting job to produce $(outfile)\")\n",
    "        Mycelia.scg_sbatch(\n",
    "            job_name=basename(outfile),\n",
    "            mail_user=\"cameron.prybol@gmail.com\",\n",
    "            logdir=mkpath(\"$(homedir())/workspace/slurmlogs\"),\n",
    "            mem_gb=mem_gb,\n",
    "            cpus_per_task=threads,\n",
    "            partition=\"nih_s10\",\n",
    "            account=\"mpsnyder\",\n",
    "            time=\"7-00:00:00\",\n",
    "            cmd=cmd\n",
    "        )\n",
    "        sleep(60)  # Pause for 60 seconds\n",
    "        successful_submissions += 1\n",
    "    else\n",
    "        @show outfile\n",
    "        @show Base.format_bytes(filesize(outfile))\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e1ddcb6-1b7e-4bb3-8dbc-2fc2bd011393",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for row in DataFrames.eachrow(sample_metadata[1:N, :])\n",
    "#     forward = row[\"forward\"]\n",
    "#     reverse = row[\"reverse\"]\n",
    "#     outdir = row[\"outdir\"]\n",
    "#     if !isdir(outdir)\n",
    "#         @show \"making path $(outdir)\"\n",
    "#         mkpath(outdir)\n",
    "#     end\n",
    "#     map_result = Mycelia.minimap_map_paired_end_with_index(fasta = compressed_fasta_export, outdir = outdir, forward = forward, reverse =reverse, mem_gb = Int(floor(mem_gb * .8)), threads=threads, denominator = 6, as_string=true)\n",
    "#     outfile = map_result.outfile\n",
    "#     cmd = map_result.cmd\n",
    "#     # resubmit if we run out of time\n",
    "#     if !isfile(outfile)\n",
    "#         println(outfile)\n",
    "#         Mycelia.scg_sbatch(\n",
    "#             job_name=basename(outfile),\n",
    "#             mail_user=\"cameron.prybol@gmail.com\",\n",
    "#             logdir=mkpath(\"$(homedir())/workspace/slurmlogs\"),\n",
    "#             mem_gb=mem_gb,\n",
    "#             cpus_per_task=threads,\n",
    "#             partition=\"nih_s10\",\n",
    "#             account=\"mpsnyder\",\n",
    "#             time=\"7-00:00:00\",\n",
    "#             cmd=cmd)\n",
    "#         # make longer?\n",
    "#         sleep(60)\n",
    "#     else\n",
    "#         @show isfile(outfile)\n",
    "#         @show Base.format_bytes(filesize(outfile))\n",
    "#     end\n",
    "# end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.5 1.10.5",
   "language": "julia",
   "name": "julia-1.10.5-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
