{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73f9dc1f-1f89-4238-8740-488772fda204",
   "metadata": {},
   "outputs": [],
   "source": [
    "import Pkg\n",
    "Pkg.activate(;temp=true)\n",
    "Pkg.add(\"Revise\")\n",
    "import Revise\n",
    "\n",
    "# you can comment these all and uncomment them one-by-one as you need them\n",
    "# to avoid loading more packages than you need\n",
    "pkgs = [\n",
    "    \"AlgebraOfGraphics\",\n",
    "    \"Arrow\",\n",
    "    \"BioAlignments\",\n",
    "    \"BioSequences\",\n",
    "    \"BioSymbols\",\n",
    "    \"CairoMakie\",\n",
    "    \"Clustering\",\n",
    "    \"CodecZlib\",\n",
    "    \"Colors\",\n",
    "    \"ColorSchemes\",\n",
    "    \"Conda\",\n",
    "    \"CSV\",\n",
    "    \"DataFrames\",\n",
    "    \"DataStructures\",\n",
    "    \"Dates\",\n",
    "    \"DelimitedFiles\",\n",
    "    \"Distances\",\n",
    "    \"Distributions\",\n",
    "    \"DocStringExtensions\",\n",
    "    \"Downloads\",\n",
    "    \"FASTX\",\n",
    "    \"FileIO\",\n",
    "    \"GenomicAnnotations\",\n",
    "    \"GeoMakie\",\n",
    "    \"GFF3\",\n",
    "    \"GLM\",\n",
    "    \"GraphMakie\",\n",
    "    \"Graphs\",\n",
    "    \"HDF5\",\n",
    "    \"HTTP\",\n",
    "    \"JLD2\",\n",
    "    \"JSON\",\n",
    "    \"Karnak\",\n",
    "    \"Kmers\",\n",
    "    \"LsqFit\",\n",
    "    \"Luxor\",\n",
    "    \"Makie\",\n",
    "    \"MetaGraphs\",\n",
    "    \"Mmap\",\n",
    "    \"OrderedCollections\",\n",
    "    \"Plots\",\n",
    "    \"Primes\",\n",
    "    \"ProgressMeter\",\n",
    "    \"Random\",\n",
    "    \"SparseArrays\",\n",
    "    \"Statistics\",\n",
    "    \"StatsBase\",\n",
    "    \"StatsPlots\",\n",
    "    \"TopoPlots\",\n",
    "    \"uCSV\",\n",
    "    \"XAM\",\n",
    "    \"XMLDict\"\n",
    "]\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "\n",
    "Pkg.develop(path=\"$(homedir())/workspace/Mycelia\")\n",
    "import Mycelia\n",
    "Mycelia.system_overview()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb118c02-c216-4da0-b13e-cc39c5efd605",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_directory = joinpath(dirname(pwd()), \"data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f4613b4-0be3-49a5-943c-4cf3658b34a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_directories = readdir(joinpath(data_directory, \"samples\"), join=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7c9aed1-2ed4-44a6-ad0f-52000970e25f",
   "metadata": {},
   "outputs": [],
   "source": [
    "kraken_dbs = filter(x -> isdir(x), readdir(\"$(homedir())/workspace/kraken\", join=true))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69d4ac2b-3f7a-42f5-a19e-d1ab46230bf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # adjust to database size\n",
    "kraken_db = \"/home/cjprybol/workspace/kraken/k2_pluspfp_20231009\"\n",
    "# kraken_db = \"/home/cjprybol/workspace/kraken/k2_pluspfp_08gb_20231009\"\n",
    "memory = 256\n",
    "threads = 32"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6edbbbf5-eba5-460e-bac8-4252d74f97fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use me to auto-configure sbatch parameterization\n",
    "# host = first(readlines(`hostname`))\n",
    "# @assert occursin(r\"\\.stanford\\.edu\", host)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e26a23de-ea5c-4a04-8f18-046b219447a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample_directory in sample_directories[2:end]\n",
    "    fastq_files = filter(x -> occursin(r\"\\.fq\\.gz$\", x), readdir(sample_directory, join=true))\n",
    "    sample_ids = unique(replace.(basename.(fastq_files), r\"_[12]\\.fq\\.gz\" => \"\"))\n",
    "    for sample_id in sample_ids\n",
    "        trim_galore_directory = joinpath(sample_directory, sample_id * \"_trimgalore\")\n",
    "        # kraken_directories = filter(x -> occursin(r\"_kraken$\", x) && isdir(x), readdir(trim_galore_directory, join=true))\n",
    "        # @show kraken_directories\n",
    "        # for kraken_directory in kraken_directories\n",
    "        #     rm(kraken_directory, recursive=true)\n",
    "        # end\n",
    "        forward = first(filter(x -> occursin(r\".unmapped.1.fq.gz\", x), readdir(trim_galore_directory, join=true)))\n",
    "        reverse = first(filter(x -> occursin(r\".unmapped.2.fq.gz\", x), readdir(trim_galore_directory, join=true)))\n",
    "        # forward = joinpath(trim_galore_directory, sample_id * \"_1_val_1.fq.gz\")\n",
    "        # reverse = joinpath(trim_galore_directory, sample_id * \"_2_val_2.fq.gz\")\n",
    "        @assert isfile(forward) && isfile(reverse)\n",
    "        kraken_dir = first(filter(x -> occursin(\"kraken\", x), readdir(trim_galore_directory, join=true)))\n",
    "        # @show kraken_dir\n",
    "        kraken_reports = filter(x -> occursin(r\"kraken-report\\.tsv$\", x), readdir(kraken_dir))\n",
    "        matching_kraken_reports = filter(x -> occursin(basename(kraken_db), x), kraken_reports)\n",
    "        if isempty(matching_kraken_reports)\n",
    "            # cmd = `julia $(homedir())/workspace/Mycelia/scripts/classify-short-reads.jl --forward_reads $(forward) --reverse_reads $(reverse) --kraken_db $(kraken_db) --threads $(threads)`\n",
    "            # run(cmd)\n",
    "            cmd = \"julia $(homedir())/workspace/Mycelia/scripts/classify-short-reads.jl --forward_reads $(forward) --reverse_reads $(reverse) --kraken_db $(kraken_db) --threads $(threads)\"\n",
    "            submission = \n",
    "            `sbatch \n",
    "            --mail-user=cameron.prybol@gmail.com\n",
    "            --mail-type=ALL\n",
    "            --error=%x-%j.err\n",
    "            --output=%x-%j.out\n",
    "            --partition=batch\n",
    "            --account=mpsnyder\n",
    "            --nodes=1\n",
    "            --ntasks=1\n",
    "            --time=1-00:00:00    \n",
    "            --cpus-per-task=$(threads)\n",
    "            --mem=$(memory)G\n",
    "            --job-name=classify-short-reads\n",
    "            --wrap $cmd\n",
    "            `\n",
    "            run(submission)\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.7",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
