{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m   Updating\u001b[22m\u001b[39m registry at `~/.julia/registries/General`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?25l    "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m   Updating\u001b[22m\u001b[39m git-repo `https://github.com/JuliaRegistries/General.git`\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2K\u001b[?25h[1mFetching:\u001b[22m\u001b[39m [========================================>]  99.9 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [======>                                  ]  12.5 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [=======>                                 ]  15.7 %                         ]  35.7 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [================>                        ]  37.7 %                   ]  50.7 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [=====================>                   ]  52.5 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [=============================>           ]  70.1 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [=============================>           ]  71.4 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [==============================>          ]  73.0 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [===============================>         ]  75.4 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [===============================>         ]  77.3 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [================================>        ]  79.6 %  ]  92.6 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [======================================>  ]  93.6 %\u001b[36m\u001b[1mFetching:\u001b[22m\u001b[39m [=======================================> ]  96.8 %"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "┌ Info: Precompiling Eisenia [453d265d-8292-4a7b-a57c-dce3f9ae6acd]\n",
      "└ @ Base loading.jl:1278\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n",
      "\u001b[32m\u001b[1m  Resolving\u001b[22m\u001b[39m package versions...\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Project.toml`\n",
      "\u001b[32m\u001b[1mNo Changes\u001b[22m\u001b[39m to `~/.julia/environments/v1.5/Manifest.toml`\n"
     ]
    }
   ],
   "source": [
    "import Pkg\n",
    "pkgs = [\n",
    "    \"Eisenia\",\n",
    "    \"Random\",\n",
    "    \"Revise\",\n",
    "    \"BioSequences\",\n",
    "    \"Test\",\n",
    "    \"LightGraphs\",\n",
    "    \"Primes\",\n",
    "    \"Statistics\",\n",
    "    \"Plots\",\n",
    "    \"uCSV\",\n",
    "    \"DataFrames\",\n",
    "    \"HTTP\",\n",
    "    \"CodecZlib\",\n",
    "    \"FASTX\"\n",
    "]\n",
    "\n",
    "for pkg in pkgs\n",
    "    try\n",
    "        Pkg.add(pkg)\n",
    "    catch\n",
    "#         # tried to install an unregistered local package\n",
    "    end\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "TASK = \"graph-decomposition-of-related-genomes\"\n",
    "DATE = \"2021-02-14\"\n",
    "DIR = \"$(homedir())/$(DATE)-$(TASK)\"\n",
    "if !isdir(DIR)\n",
    "    mkdir(DIR)\n",
    "end\n",
    "cd(DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "get_genome_from_ftp_directory (generic function with 1 method)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function get_genome_from_ftp_directory(ftp_path)\n",
    "    fasta_url = \"$(ftp_path)/$(basename(ftp_path))_genomic.fna.gz\"\n",
    "    fasta_buffer = CodecZlib.GzipDecompressorStream(IOBuffer(HTTP.get(fasta_url).body))\n",
    "    return collect(FASTX.FASTA.Reader(fasta_buffer))\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th># assembly_accession</th><th>bioproject</th><th>biosample</th><th>wgs_master</th><th>refseq_category</th></tr><tr><th></th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th></tr></thead><tbody><p>952,203 rows × 22 columns (omitted printing of 17 columns)</p><tr><th>1</th><td>GCA_000001215.4</td><td>PRJNA13812</td><td>SAMN02803731</td><td></td><td>reference genome</td></tr><tr><th>2</th><td>GCA_000001405.28</td><td>PRJNA31257</td><td></td><td></td><td>reference genome</td></tr><tr><th>3</th><td>GCA_000001515.5</td><td>PRJNA13184</td><td>SAMN02981217</td><td>AACZ00000000.4</td><td>na</td></tr><tr><th>4</th><td>GCA_000001545.3</td><td>PRJNA20869</td><td>SAMN02981238</td><td>ABGA00000000.1</td><td>na</td></tr><tr><th>5</th><td>GCA_000001635.9</td><td>PRJNA20689</td><td></td><td></td><td>reference genome</td></tr><tr><th>6</th><td>GCA_000001735.2</td><td>PRJNA10719</td><td>SAMN03081427</td><td></td><td>reference genome</td></tr><tr><th>7</th><td>GCA_000001765.3</td><td>PRJNA10626</td><td>SAMN00779672</td><td>AADE00000000.2</td><td>na</td></tr><tr><th>8</th><td>GCA_000001895.4</td><td>PRJNA10629</td><td>SAMN02808228</td><td>AABR00000000.7</td><td>na</td></tr><tr><th>9</th><td>GCA_000001905.1</td><td>PRJNA12569</td><td>SAMN02953622</td><td>AAGU00000000.3</td><td>representative genome</td></tr><tr><th>10</th><td>GCA_000001985.1</td><td>PRJNA19555</td><td>SAMN02953685</td><td>ABAR00000000.1</td><td>representative genome</td></tr><tr><th>11</th><td>GCA_000002035.4</td><td>PRJNA11776</td><td>SAMN06930106</td><td></td><td>reference genome</td></tr><tr><th>12</th><td>GCA_000002075.2</td><td>PRJNA13635</td><td>SAMN02953658</td><td>AASC00000000.3</td><td>representative genome</td></tr><tr><th>13</th><td>GCA_000002115.2</td><td>PRJNA1431</td><td>SAMN02981219</td><td>AADD00000000.1</td><td>na</td></tr><tr><th>14</th><td>GCA_000002125.2</td><td>PRJNA19621</td><td>SAMN02981236</td><td>ABBA00000000.1</td><td>na</td></tr><tr><th>15</th><td>GCA_000002135.3</td><td>PRJNA10793</td><td></td><td></td><td>na</td></tr><tr><th>16</th><td>GCA_000002165.1</td><td>PRJNA11785</td><td>SAMN03004379</td><td>AAHY00000000.1</td><td>na</td></tr><tr><th>17</th><td>GCA_000002175.2</td><td>PRJNA16847</td><td>SAMD00067074</td><td></td><td>na</td></tr><tr><th>18</th><td>GCA_000002195.1</td><td>PRJNA10625</td><td>SAMN00002455</td><td>AADG00000000.6</td><td>na</td></tr><tr><th>19</th><td>GCA_000002235.4</td><td>PRJNA10736</td><td>SAMN00829422</td><td>AAGJ00000000.6</td><td>representative genome</td></tr><tr><th>20</th><td>GCA_000002255.2</td><td>PRJNA12537</td><td>SAMN02981228</td><td>AANU00000000.1</td><td>na</td></tr><tr><th>21</th><td>GCA_000002265.1</td><td>PRJNA13999</td><td>SAMN03000701</td><td>AAHX00000000.1</td><td>na</td></tr><tr><th>22</th><td>GCA_000002275.2</td><td>PRJNA12885</td><td>SAMN02953646</td><td>AAPN00000000.1</td><td>na</td></tr><tr><th>23</th><td>GCA_000002285.4</td><td>PRJNA13179</td><td>SAMN02953603</td><td>AAEX00000000.4</td><td>na</td></tr><tr><th>24</th><td>GCA_000002295.1</td><td>PRJNA12561</td><td>SAMN02953611</td><td>AAFR00000000.3</td><td>representative genome</td></tr><tr><th>25</th><td>GCA_000002305.1</td><td>PRJNA18661</td><td>SAMN02953672</td><td>AAWR00000000.2</td><td>na</td></tr><tr><th>26</th><td>GCA_000002315.5</td><td>PRJNA13342</td><td>SAMN02981218</td><td>AADN00000000.5</td><td>representative genome</td></tr><tr><th>27</th><td>GCA_000002325.2</td><td>PRJNA13660</td><td>SAMN02902141</td><td>AAZX00000000.1</td><td>na</td></tr><tr><th>28</th><td>GCA_000002335.3</td><td>PRJNA12540</td><td>SAMN02953634</td><td>AAJJ00000000.2</td><td>representative genome</td></tr><tr><th>29</th><td>GCA_000002415.2</td><td>PRJNA150</td><td>SAMN02953638</td><td>AAKM00000000.1</td><td>representative genome</td></tr><tr><th>30</th><td>GCA_000002425.2</td><td>PRJNA13064</td><td>SAMN00120126</td><td>ABEU00000000.2</td><td>representative genome</td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|cccccc}\n",
       "\t& \\# assembly\\_accession & bioproject & biosample & wgs\\_master & refseq\\_category & \\\\\n",
       "\t\\hline\n",
       "\t& String & String & String & String & String & \\\\\n",
       "\t\\hline\n",
       "\t1 & GCA\\_000001215.4 & PRJNA13812 & SAMN02803731 &  & reference genome & $\\dots$ \\\\\n",
       "\t2 & GCA\\_000001405.28 & PRJNA31257 &  &  & reference genome & $\\dots$ \\\\\n",
       "\t3 & GCA\\_000001515.5 & PRJNA13184 & SAMN02981217 & AACZ00000000.4 & na & $\\dots$ \\\\\n",
       "\t4 & GCA\\_000001545.3 & PRJNA20869 & SAMN02981238 & ABGA00000000.1 & na & $\\dots$ \\\\\n",
       "\t5 & GCA\\_000001635.9 & PRJNA20689 &  &  & reference genome & $\\dots$ \\\\\n",
       "\t6 & GCA\\_000001735.2 & PRJNA10719 & SAMN03081427 &  & reference genome & $\\dots$ \\\\\n",
       "\t7 & GCA\\_000001765.3 & PRJNA10626 & SAMN00779672 & AADE00000000.2 & na & $\\dots$ \\\\\n",
       "\t8 & GCA\\_000001895.4 & PRJNA10629 & SAMN02808228 & AABR00000000.7 & na & $\\dots$ \\\\\n",
       "\t9 & GCA\\_000001905.1 & PRJNA12569 & SAMN02953622 & AAGU00000000.3 & representative genome & $\\dots$ \\\\\n",
       "\t10 & GCA\\_000001985.1 & PRJNA19555 & SAMN02953685 & ABAR00000000.1 & representative genome & $\\dots$ \\\\\n",
       "\t11 & GCA\\_000002035.4 & PRJNA11776 & SAMN06930106 &  & reference genome & $\\dots$ \\\\\n",
       "\t12 & GCA\\_000002075.2 & PRJNA13635 & SAMN02953658 & AASC00000000.3 & representative genome & $\\dots$ \\\\\n",
       "\t13 & GCA\\_000002115.2 & PRJNA1431 & SAMN02981219 & AADD00000000.1 & na & $\\dots$ \\\\\n",
       "\t14 & GCA\\_000002125.2 & PRJNA19621 & SAMN02981236 & ABBA00000000.1 & na & $\\dots$ \\\\\n",
       "\t15 & GCA\\_000002135.3 & PRJNA10793 &  &  & na & $\\dots$ \\\\\n",
       "\t16 & GCA\\_000002165.1 & PRJNA11785 & SAMN03004379 & AAHY00000000.1 & na & $\\dots$ \\\\\n",
       "\t17 & GCA\\_000002175.2 & PRJNA16847 & SAMD00067074 &  & na & $\\dots$ \\\\\n",
       "\t18 & GCA\\_000002195.1 & PRJNA10625 & SAMN00002455 & AADG00000000.6 & na & $\\dots$ \\\\\n",
       "\t19 & GCA\\_000002235.4 & PRJNA10736 & SAMN00829422 & AAGJ00000000.6 & representative genome & $\\dots$ \\\\\n",
       "\t20 & GCA\\_000002255.2 & PRJNA12537 & SAMN02981228 & AANU00000000.1 & na & $\\dots$ \\\\\n",
       "\t21 & GCA\\_000002265.1 & PRJNA13999 & SAMN03000701 & AAHX00000000.1 & na & $\\dots$ \\\\\n",
       "\t22 & GCA\\_000002275.2 & PRJNA12885 & SAMN02953646 & AAPN00000000.1 & na & $\\dots$ \\\\\n",
       "\t23 & GCA\\_000002285.4 & PRJNA13179 & SAMN02953603 & AAEX00000000.4 & na & $\\dots$ \\\\\n",
       "\t24 & GCA\\_000002295.1 & PRJNA12561 & SAMN02953611 & AAFR00000000.3 & representative genome & $\\dots$ \\\\\n",
       "\t25 & GCA\\_000002305.1 & PRJNA18661 & SAMN02953672 & AAWR00000000.2 & na & $\\dots$ \\\\\n",
       "\t26 & GCA\\_000002315.5 & PRJNA13342 & SAMN02981218 & AADN00000000.5 & representative genome & $\\dots$ \\\\\n",
       "\t27 & GCA\\_000002325.2 & PRJNA13660 & SAMN02902141 & AAZX00000000.1 & na & $\\dots$ \\\\\n",
       "\t28 & GCA\\_000002335.3 & PRJNA12540 & SAMN02953634 & AAJJ00000000.2 & representative genome & $\\dots$ \\\\\n",
       "\t29 & GCA\\_000002415.2 & PRJNA150 & SAMN02953638 & AAKM00000000.1 & representative genome & $\\dots$ \\\\\n",
       "\t30 & GCA\\_000002425.2 & PRJNA13064 & SAMN00120126 & ABEU00000000.2 & representative genome & $\\dots$ \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ &  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "952203×22 DataFrame. Omitted printing of 19 columns\n",
       "│ Row    │ # assembly_accession │ bioproject │ biosample    │\n",
       "│        │ \u001b[90mString\u001b[39m               │ \u001b[90mString\u001b[39m     │ \u001b[90mString\u001b[39m       │\n",
       "├────────┼──────────────────────┼────────────┼──────────────┤\n",
       "│ 1      │ GCA_000001215.4      │ PRJNA13812 │ SAMN02803731 │\n",
       "│ 2      │ GCA_000001405.28     │ PRJNA31257 │              │\n",
       "│ 3      │ GCA_000001515.5      │ PRJNA13184 │ SAMN02981217 │\n",
       "│ 4      │ GCA_000001545.3      │ PRJNA20869 │ SAMN02981238 │\n",
       "│ 5      │ GCA_000001635.9      │ PRJNA20689 │              │\n",
       "│ 6      │ GCA_000001735.2      │ PRJNA10719 │ SAMN03081427 │\n",
       "│ 7      │ GCA_000001765.3      │ PRJNA10626 │ SAMN00779672 │\n",
       "│ 8      │ GCA_000001895.4      │ PRJNA10629 │ SAMN02808228 │\n",
       "│ 9      │ GCA_000001905.1      │ PRJNA12569 │ SAMN02953622 │\n",
       "│ 10     │ GCA_000001985.1      │ PRJNA19555 │ SAMN02953685 │\n",
       "⋮\n",
       "│ 952193 │ GCA_905232335.1      │ PRJEB42964 │ SAMEA7998009 │\n",
       "│ 952194 │ GCA_905232345.1      │ PRJEB42964 │ SAMEA7998003 │\n",
       "│ 952195 │ GCA_905232355.1      │ PRJEB42964 │ SAMEA7998005 │\n",
       "│ 952196 │ GCA_905232365.1      │ PRJEB42964 │ SAMEA7998006 │\n",
       "│ 952197 │ GCA_905236995.1      │ PRJEB28470 │ SAMEA7833113 │\n",
       "│ 952198 │ GCA_905237015.1      │ PRJEB28470 │ SAMEA7833114 │\n",
       "│ 952199 │ GCA_905237025.1      │ PRJEB28470 │ SAMEA7833116 │\n",
       "│ 952200 │ GCA_905237035.1      │ PRJEB28470 │ SAMEA7833115 │\n",
       "│ 952201 │ GCA_905237045.1      │ PRJEB28470 │ SAMEA7833117 │\n",
       "│ 952202 │ GCA_905237085.1      │ PRJEB33539 │ SAMEA5987284 │\n",
       "│ 952203 │ GCA_905237095.1      │ PRJEB33539 │ SAMEA5987285 │"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db = \"genbank\"\n",
    "assembly_summary_url = \"https://ftp.ncbi.nlm.nih.gov/genomes/$(db)/assembly_summary_$(db).txt\"\n",
    "assembly_summary_response = HTTP.get(assembly_summary_url)\n",
    "assembly_summary_response_buffer = IOBuffer(assembly_summary_response.body)\n",
    "assembly_summary_table = DataFrames.DataFrame(uCSV.read(assembly_summary_response_buffer, delim='\\t', header=2, skipmalformed=true)...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th>#species_taxid</th><th>min_ungapped_length</th><th>max_ungapped_length</th><th>expected_ungapped_length</th></tr><tr><th></th><th>Int64</th><th>Int64</th><th>Int64</th><th>Int64</th></tr></thead><tbody><p>2,797 rows × 5 columns (omitted printing of 1 columns)</p><tr><th>1</th><td>9</td><td>318000</td><td>956000</td><td>636876</td></tr><tr><th>2</th><td>24</td><td>2950000</td><td>6050000</td><td>4500062</td></tr><tr><th>3</th><td>34</td><td>7405000</td><td>11109000</td><td>9257167</td></tr><tr><th>4</th><td>56</td><td>5994000</td><td>15000000</td><td>11989871</td></tr><tr><th>5</th><td>69</td><td>4912000</td><td>7369000</td><td>6140343</td></tr><tr><th>6</th><td>114</td><td>7212000</td><td>10820000</td><td>9015893</td></tr><tr><th>7</th><td>122</td><td>6222000</td><td>9334000</td><td>7777997</td></tr><tr><th>8</th><td>124</td><td>5461000</td><td>9315000</td><td>7387897</td></tr><tr><th>9</th><td>139</td><td>813000</td><td>1786000</td><td>1299537</td></tr><tr><th>10</th><td>140</td><td>688000</td><td>2067000</td><td>1377933</td></tr><tr><th>11</th><td>158</td><td>2253000</td><td>3380000</td><td>2816290</td></tr><tr><th>12</th><td>159</td><td>2436000</td><td>3655000</td><td>3045648</td></tr><tr><th>13</th><td>160</td><td>911000</td><td>1367000</td><td>1139090</td></tr><tr><th>14</th><td>162</td><td>2302000</td><td>4443000</td><td>3372116</td></tr><tr><th>15</th><td>164</td><td>1779000</td><td>2669000</td><td>2224027</td></tr><tr><th>16</th><td>172</td><td>3103000</td><td>4655000</td><td>3879078</td></tr><tr><th>17</th><td>173</td><td>3700000</td><td>5552000</td><td>4626069</td></tr><tr><th>18</th><td>174</td><td>3138000</td><td>4708000</td><td>3922817</td></tr><tr><th>19</th><td>192</td><td>5672000</td><td>8563000</td><td>7117205</td></tr><tr><th>20</th><td>193</td><td>5968000</td><td>9403000</td><td>7685216</td></tr><tr><th>21</th><td>195</td><td>1408000</td><td>2114000</td><td>1760952</td></tr><tr><th>22</th><td>196</td><td>1600000</td><td>2200000</td><td>1900000</td></tr><tr><th>23</th><td>197</td><td>1376000</td><td>2065000</td><td>1720172</td></tr><tr><th>24</th><td>198</td><td>1427000</td><td>2141000</td><td>1783989</td></tr><tr><th>25</th><td>199</td><td>1340000</td><td>2534000</td><td>1937433</td></tr><tr><th>26</th><td>200</td><td>1573000</td><td>2361000</td><td>1966742</td></tr><tr><th>27</th><td>201</td><td>1206000</td><td>1856000</td><td>1531350</td></tr><tr><th>28</th><td>202</td><td>1425000</td><td>2182000</td><td>1803524</td></tr><tr><th>29</th><td>204</td><td>1566000</td><td>2781000</td><td>2173965</td></tr><tr><th>30</th><td>206</td><td>1402000</td><td>2105000</td><td>1753386</td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ccccc}\n",
       "\t& \\#species\\_taxid & min\\_ungapped\\_length & max\\_ungapped\\_length & expected\\_ungapped\\_length & \\\\\n",
       "\t\\hline\n",
       "\t& Int64 & Int64 & Int64 & Int64 & \\\\\n",
       "\t\\hline\n",
       "\t1 & 9 & 318000 & 956000 & 636876 & $\\dots$ \\\\\n",
       "\t2 & 24 & 2950000 & 6050000 & 4500062 & $\\dots$ \\\\\n",
       "\t3 & 34 & 7405000 & 11109000 & 9257167 & $\\dots$ \\\\\n",
       "\t4 & 56 & 5994000 & 15000000 & 11989871 & $\\dots$ \\\\\n",
       "\t5 & 69 & 4912000 & 7369000 & 6140343 & $\\dots$ \\\\\n",
       "\t6 & 114 & 7212000 & 10820000 & 9015893 & $\\dots$ \\\\\n",
       "\t7 & 122 & 6222000 & 9334000 & 7777997 & $\\dots$ \\\\\n",
       "\t8 & 124 & 5461000 & 9315000 & 7387897 & $\\dots$ \\\\\n",
       "\t9 & 139 & 813000 & 1786000 & 1299537 & $\\dots$ \\\\\n",
       "\t10 & 140 & 688000 & 2067000 & 1377933 & $\\dots$ \\\\\n",
       "\t11 & 158 & 2253000 & 3380000 & 2816290 & $\\dots$ \\\\\n",
       "\t12 & 159 & 2436000 & 3655000 & 3045648 & $\\dots$ \\\\\n",
       "\t13 & 160 & 911000 & 1367000 & 1139090 & $\\dots$ \\\\\n",
       "\t14 & 162 & 2302000 & 4443000 & 3372116 & $\\dots$ \\\\\n",
       "\t15 & 164 & 1779000 & 2669000 & 2224027 & $\\dots$ \\\\\n",
       "\t16 & 172 & 3103000 & 4655000 & 3879078 & $\\dots$ \\\\\n",
       "\t17 & 173 & 3700000 & 5552000 & 4626069 & $\\dots$ \\\\\n",
       "\t18 & 174 & 3138000 & 4708000 & 3922817 & $\\dots$ \\\\\n",
       "\t19 & 192 & 5672000 & 8563000 & 7117205 & $\\dots$ \\\\\n",
       "\t20 & 193 & 5968000 & 9403000 & 7685216 & $\\dots$ \\\\\n",
       "\t21 & 195 & 1408000 & 2114000 & 1760952 & $\\dots$ \\\\\n",
       "\t22 & 196 & 1600000 & 2200000 & 1900000 & $\\dots$ \\\\\n",
       "\t23 & 197 & 1376000 & 2065000 & 1720172 & $\\dots$ \\\\\n",
       "\t24 & 198 & 1427000 & 2141000 & 1783989 & $\\dots$ \\\\\n",
       "\t25 & 199 & 1340000 & 2534000 & 1937433 & $\\dots$ \\\\\n",
       "\t26 & 200 & 1573000 & 2361000 & 1966742 & $\\dots$ \\\\\n",
       "\t27 & 201 & 1206000 & 1856000 & 1531350 & $\\dots$ \\\\\n",
       "\t28 & 202 & 1425000 & 2182000 & 1803524 & $\\dots$ \\\\\n",
       "\t29 & 204 & 1566000 & 2781000 & 2173965 & $\\dots$ \\\\\n",
       "\t30 & 206 & 1402000 & 2105000 & 1753386 & $\\dots$ \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ &  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "2797×5 DataFrame. Omitted printing of 2 columns\n",
       "│ Row  │ #species_taxid │ min_ungapped_length │ max_ungapped_length │\n",
       "│      │ \u001b[90mInt64\u001b[39m          │ \u001b[90mInt64\u001b[39m               │ \u001b[90mInt64\u001b[39m               │\n",
       "├──────┼────────────────┼─────────────────────┼─────────────────────┤\n",
       "│ 1    │ 9              │ 318000              │ 956000              │\n",
       "│ 2    │ 24             │ 2950000             │ 6050000             │\n",
       "│ 3    │ 34             │ 7405000             │ 11109000            │\n",
       "│ 4    │ 56             │ 5994000             │ 15000000            │\n",
       "│ 5    │ 69             │ 4912000             │ 7369000             │\n",
       "│ 6    │ 114            │ 7212000             │ 10820000            │\n",
       "│ 7    │ 122            │ 6222000             │ 9334000             │\n",
       "│ 8    │ 124            │ 5461000             │ 9315000             │\n",
       "│ 9    │ 139            │ 813000              │ 1786000             │\n",
       "│ 10   │ 140            │ 688000              │ 2067000             │\n",
       "⋮\n",
       "│ 2787 │ 2723674        │ 17496000            │ 37645000            │\n",
       "│ 2788 │ 2741499        │ 4138000             │ 6209000             │\n",
       "│ 2789 │ 2742203        │ 2103000             │ 3156000             │\n",
       "│ 2790 │ 2754717        │ 850000              │ 2552000             │\n",
       "│ 2791 │ 2763006        │ 6549000             │ 10029000            │\n",
       "│ 2792 │ 2766470        │ 1325000             │ 2578000             │\n",
       "│ 2793 │ 2778550        │ 3291000             │ 5889000             │\n",
       "│ 2794 │ 2781243        │ 2321000             │ 3483000             │\n",
       "│ 2795 │ 2782701        │ 7171000             │ 10758000            │\n",
       "│ 2796 │ 2801451        │ 1594000             │ 4518000             │\n",
       "│ 2797 │ 2813651        │ 31916000            │ 47875000            │"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_size_table_url = \"ftp://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/species_genome_size.txt.gz\"\n",
    "genome_size_table_response = HTTP.get(genome_size_table_url)\n",
    "genome_size_table_buffer = CodecZlib.GzipDecompressorStream(IOBuffer(genome_size_table_response.body))\n",
    "genome_size_table = DataFrames.DataFrame(uCSV.read(genome_size_table_buffer, header=1, delim='\\t')...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th># assembly_accession</th><th>bioproject</th><th>biosample</th><th>wgs_master</th><th>refseq_category</th><th>taxid</th></tr><tr><th></th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>Int64</th></tr></thead><tbody><p>764,249 rows × 26 columns (omitted printing of 20 columns)</p><tr><th>1</th><td>GCA_002829805.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>2</th><td>GCA_003972145.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>3</th><td>GCA_003972165.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>4</th><td>GCA_003972185.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>5</th><td>GCA_003972205.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>6</th><td>GCA_003972225.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>7</th><td>GCA_003972245.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>8</th><td>GCA_003972265.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>9</th><td>GCA_000851485.1</td><td></td><td></td><td></td><td>reference genome</td><td>154834</td></tr><tr><th>10</th><td>GCA_000856565.1</td><td></td><td></td><td></td><td>na</td><td>12475</td></tr><tr><th>11</th><td>GCA_003033635.1</td><td></td><td></td><td></td><td>na</td><td>12475</td></tr><tr><th>12</th><td>GCA_003033645.1</td><td></td><td></td><td></td><td>na</td><td>12475</td></tr><tr><th>13</th><td>GCA_003047595.1</td><td></td><td></td><td></td><td>na</td><td>12475</td></tr><tr><th>14</th><td>GCA_003047615.1</td><td></td><td></td><td></td><td>na</td><td>12475</td></tr><tr><th>15</th><td>GCA_003047635.1</td><td></td><td></td><td></td><td>na</td><td>12475</td></tr><tr><th>16</th><td>GCA_003047655.1</td><td></td><td></td><td></td><td>na</td><td>12475</td></tr><tr><th>17</th><td>GCA_003047915.1</td><td></td><td></td><td></td><td>na</td><td>12475</td></tr><tr><th>18</th><td>GCA_003971865.1</td><td></td><td></td><td></td><td>na</td><td>12475</td></tr><tr><th>19</th><td>GCA_000919655.1</td><td></td><td></td><td></td><td>na</td><td>1234882</td></tr><tr><th>20</th><td>GCA_002819965.1</td><td></td><td></td><td></td><td>na</td><td>1234882</td></tr><tr><th>21</th><td>GCA_003985465.1</td><td></td><td></td><td></td><td>na</td><td>1072162</td></tr><tr><th>22</th><td>GCA_004369285.1</td><td></td><td></td><td></td><td>na</td><td>1072162</td></tr><tr><th>23</th><td>GCA_000918015.1</td><td></td><td></td><td></td><td>na</td><td>1475143</td></tr><tr><th>24</th><td>GCA_004046075.1</td><td></td><td></td><td></td><td>na</td><td>1475143</td></tr><tr><th>25</th><td>GCA_004046095.1</td><td></td><td></td><td></td><td>na</td><td>1475143</td></tr><tr><th>26</th><td>GCA_004046115.1</td><td></td><td></td><td></td><td>na</td><td>1475143</td></tr><tr><th>27</th><td>GCA_004046135.1</td><td></td><td></td><td></td><td>na</td><td>1475143</td></tr><tr><th>28</th><td>GCA_004046155.1</td><td></td><td></td><td></td><td>na</td><td>1475143</td></tr><tr><th>29</th><td>GCA_004046175.1</td><td></td><td></td><td></td><td>na</td><td>1475143</td></tr><tr><th>30</th><td>GCA_004051695.1</td><td></td><td></td><td></td><td>na</td><td>1475143</td></tr><tr><th>&vellip;</th><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td><td>&vellip;</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ccccccc}\n",
       "\t& \\# assembly\\_accession & bioproject & biosample & wgs\\_master & refseq\\_category & taxid & \\\\\n",
       "\t\\hline\n",
       "\t& String & String & String & String & String & Int64 & \\\\\n",
       "\t\\hline\n",
       "\t1 & GCA\\_002829805.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t2 & GCA\\_003972145.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t3 & GCA\\_003972165.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t4 & GCA\\_003972185.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t5 & GCA\\_003972205.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t6 & GCA\\_003972225.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t7 & GCA\\_003972245.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t8 & GCA\\_003972265.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t9 & GCA\\_000851485.1 &  &  &  & reference genome & 154834 & $\\dots$ \\\\\n",
       "\t10 & GCA\\_000856565.1 &  &  &  & na & 12475 & $\\dots$ \\\\\n",
       "\t11 & GCA\\_003033635.1 &  &  &  & na & 12475 & $\\dots$ \\\\\n",
       "\t12 & GCA\\_003033645.1 &  &  &  & na & 12475 & $\\dots$ \\\\\n",
       "\t13 & GCA\\_003047595.1 &  &  &  & na & 12475 & $\\dots$ \\\\\n",
       "\t14 & GCA\\_003047615.1 &  &  &  & na & 12475 & $\\dots$ \\\\\n",
       "\t15 & GCA\\_003047635.1 &  &  &  & na & 12475 & $\\dots$ \\\\\n",
       "\t16 & GCA\\_003047655.1 &  &  &  & na & 12475 & $\\dots$ \\\\\n",
       "\t17 & GCA\\_003047915.1 &  &  &  & na & 12475 & $\\dots$ \\\\\n",
       "\t18 & GCA\\_003971865.1 &  &  &  & na & 12475 & $\\dots$ \\\\\n",
       "\t19 & GCA\\_000919655.1 &  &  &  & na & 1234882 & $\\dots$ \\\\\n",
       "\t20 & GCA\\_002819965.1 &  &  &  & na & 1234882 & $\\dots$ \\\\\n",
       "\t21 & GCA\\_003985465.1 &  &  &  & na & 1072162 & $\\dots$ \\\\\n",
       "\t22 & GCA\\_004369285.1 &  &  &  & na & 1072162 & $\\dots$ \\\\\n",
       "\t23 & GCA\\_000918015.1 &  &  &  & na & 1475143 & $\\dots$ \\\\\n",
       "\t24 & GCA\\_004046075.1 &  &  &  & na & 1475143 & $\\dots$ \\\\\n",
       "\t25 & GCA\\_004046095.1 &  &  &  & na & 1475143 & $\\dots$ \\\\\n",
       "\t26 & GCA\\_004046115.1 &  &  &  & na & 1475143 & $\\dots$ \\\\\n",
       "\t27 & GCA\\_004046135.1 &  &  &  & na & 1475143 & $\\dots$ \\\\\n",
       "\t28 & GCA\\_004046155.1 &  &  &  & na & 1475143 & $\\dots$ \\\\\n",
       "\t29 & GCA\\_004046175.1 &  &  &  & na & 1475143 & $\\dots$ \\\\\n",
       "\t30 & GCA\\_004051695.1 &  &  &  & na & 1475143 & $\\dots$ \\\\\n",
       "\t$\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ & $\\dots$ &  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "764249×26 DataFrame. Omitted printing of 23 columns\n",
       "│ Row    │ # assembly_accession │ bioproject │ biosample    │\n",
       "│        │ \u001b[90mString\u001b[39m               │ \u001b[90mString\u001b[39m     │ \u001b[90mString\u001b[39m       │\n",
       "├────────┼──────────────────────┼────────────┼──────────────┤\n",
       "│ 1      │ GCA_002829805.1      │            │              │\n",
       "│ 2      │ GCA_003972145.1      │            │              │\n",
       "│ 3      │ GCA_003972165.1      │            │              │\n",
       "│ 4      │ GCA_003972185.1      │            │              │\n",
       "│ 5      │ GCA_003972205.1      │            │              │\n",
       "│ 6      │ GCA_003972225.1      │            │              │\n",
       "│ 7      │ GCA_003972245.1      │            │              │\n",
       "│ 8      │ GCA_003972265.1      │            │              │\n",
       "│ 9      │ GCA_000851485.1      │            │              │\n",
       "│ 10     │ GCA_000856565.1      │            │              │\n",
       "⋮\n",
       "│ 764239 │ GCA_902810675.1      │ PRJEB35709 │ SAMEA6374024 │\n",
       "│ 764240 │ GCA_902810685.1      │ PRJEB35709 │ SAMEA6374021 │\n",
       "│ 764241 │ GCA_903993795.1      │ PRJEB39553 │ SAMEA7089055 │\n",
       "│ 764242 │ GCA_903993975.1      │ PRJEB39555 │ SAMEA7089057 │\n",
       "│ 764243 │ GCA_903993985.1      │ PRJEB39552 │ SAMEA7089054 │\n",
       "│ 764244 │ GCA_903994155.1      │ PRJEB39560 │ SAMEA7089062 │\n",
       "│ 764245 │ GCA_903994175.1      │ PRJEB39557 │ SAMEA7089059 │\n",
       "│ 764246 │ GCA_903994185.1      │ PRJEB39558 │ SAMEA7089060 │\n",
       "│ 764247 │ GCA_903994195.1      │ PRJEB39554 │ SAMEA7089056 │\n",
       "│ 764248 │ GCA_903995565.1      │ PRJEB39556 │ SAMEA7089058 │\n",
       "│ 764249 │ GCA_904066035.1      │ PRJEB39982 │ SAMEA7202579 │"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "joint_table = DataFrames.innerjoin(assembly_summary_table, genome_size_table, on=[\"species_taxid\" => \"#species_taxid\"])\n",
    "sort!(joint_table, \"expected_ungapped_length\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "764249×26 DataFrame\n",
      "│ Row    │ # assembly_accession │ bioproject │ biosample    │\n",
      "│        │ \u001b[90mString\u001b[39m               │ \u001b[90mString\u001b[39m     │ \u001b[90mString\u001b[39m       │\n",
      "├────────┼──────────────────────┼────────────┼──────────────┤\n",
      "│ 1      │ GCA_002829805.1      │            │              │\n",
      "│ 2      │ GCA_003972145.1      │            │              │\n",
      "│ 3      │ GCA_003972165.1      │            │              │\n",
      "│ 4      │ GCA_003972185.1      │            │              │\n",
      "│ 5      │ GCA_003972205.1      │            │              │\n",
      "│ 6      │ GCA_003972225.1      │            │              │\n",
      "│ 7      │ GCA_003972245.1      │            │              │\n",
      "│ 8      │ GCA_003972265.1      │            │              │\n",
      "│ 9      │ GCA_000851485.1      │            │              │\n",
      "│ 10     │ GCA_000856565.1      │            │              │\n",
      "⋮\n",
      "│ 764239 │ GCA_902810675.1      │ PRJEB35709 │ SAMEA6374024 │\n",
      "│ 764240 │ GCA_902810685.1      │ PRJEB35709 │ SAMEA6374021 │\n",
      "│ 764241 │ GCA_903993795.1      │ PRJEB39553 │ SAMEA7089055 │\n",
      "│ 764242 │ GCA_903993975.1      │ PRJEB39555 │ SAMEA7089057 │\n",
      "│ 764243 │ GCA_903993985.1      │ PRJEB39552 │ SAMEA7089054 │\n",
      "│ 764244 │ GCA_903994155.1      │ PRJEB39560 │ SAMEA7089062 │\n",
      "│ 764245 │ GCA_903994175.1      │ PRJEB39557 │ SAMEA7089059 │\n",
      "│ 764246 │ GCA_903994185.1      │ PRJEB39558 │ SAMEA7089060 │\n",
      "│ 764247 │ GCA_903994195.1      │ PRJEB39554 │ SAMEA7089056 │\n",
      "│ 764248 │ GCA_903995565.1      │ PRJEB39556 │ SAMEA7089058 │\n",
      "│ 764249 │ GCA_904066035.1      │ PRJEB39982 │ SAMEA7202579 │\n",
      "\n",
      "│ Row    │ wgs_master        │ refseq_category  │ taxid  │ species_taxid │\n",
      "│        │ \u001b[90mString\u001b[39m            │ \u001b[90mString\u001b[39m           │ \u001b[90mInt64\u001b[39m  │ \u001b[90mInt64\u001b[39m         │\n",
      "├────────┼───────────────────┼──────────────────┼────────┼───────────────┤\n",
      "│ 1      │                   │ na               │ 230604 │ 1922242       │\n",
      "│ 2      │                   │ na               │ 230604 │ 1922242       │\n",
      "│ 3      │                   │ na               │ 230604 │ 1922242       │\n",
      "│ 4      │                   │ na               │ 230604 │ 1922242       │\n",
      "│ 5      │                   │ na               │ 230604 │ 1922242       │\n",
      "│ 6      │                   │ na               │ 230604 │ 1922242       │\n",
      "│ 7      │                   │ na               │ 230604 │ 1922242       │\n",
      "│ 8      │                   │ na               │ 230604 │ 1922242       │\n",
      "│ 9      │                   │ reference genome │ 154834 │ 1921432       │\n",
      "│ 10     │                   │ na               │ 12475  │ 12475         │\n",
      "⋮\n",
      "│ 764239 │ CADDYO000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764240 │ CADDYY000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764241 │ CAJEVU000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764242 │ CAJEVV000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764243 │ CAJEVW000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764244 │ CAJEWR000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764245 │ CAJEWO000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764246 │ CAJEWS000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764247 │ CAJEWQ000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764248 │ CAJFAH000000000.1 │ na               │ 4565   │ 4565          │\n",
      "│ 764249 │ CAJFCQ000000000.1 │ na               │ 4565   │ 4565          │\n",
      "\n",
      "│ Row    │ organism_name                  │ infraspecific_name        │\n",
      "│        │ \u001b[90mString\u001b[39m                         │ \u001b[90mString\u001b[39m                    │\n",
      "├────────┼────────────────────────────────┼───────────────────────────┤\n",
      "│ 1      │ Extra small virus              │                           │\n",
      "│ 2      │ Extra small virus              │                           │\n",
      "│ 3      │ Extra small virus              │                           │\n",
      "│ 4      │ Extra small virus              │                           │\n",
      "│ 5      │ Extra small virus              │                           │\n",
      "│ 6      │ Extra small virus              │                           │\n",
      "│ 7      │ Extra small virus              │                           │\n",
      "│ 8      │ Extra small virus              │                           │\n",
      "│ 9      │ Panicum mosaic satellite virus │                           │\n",
      "│ 10     │ Hepatitis delta virus          │                           │\n",
      "⋮\n",
      "│ 764239 │ Triticum aestivum              │                           │\n",
      "│ 764240 │ Triticum aestivum              │                           │\n",
      "│ 764241 │ Triticum aestivum              │ cultivar=Jagger           │\n",
      "│ 764242 │ Triticum aestivum              │ cultivar=LongReach Lancer │\n",
      "│ 764243 │ Triticum aestivum              │ cultivar=ArinaLrFor       │\n",
      "│ 764244 │ Triticum aestivum              │ cultivar=CDC Stanley      │\n",
      "│ 764245 │ Triticum aestivum              │ cultivar=Mace             │\n",
      "│ 764246 │ Triticum aestivum              │ cultivar=SY Mattis        │\n",
      "│ 764247 │ Triticum aestivum              │ cultivar=Julius           │\n",
      "│ 764248 │ Triticum aestivum              │ cultivar=CDC Landmark     │\n",
      "│ 764249 │ Triticum aestivum              │ cultivar=Norin 61         │\n",
      "\n",
      "│ Row    │ isolate │ version_status │ assembly_level  │ release_type │\n",
      "│        │ \u001b[90mString\u001b[39m  │ \u001b[90mString\u001b[39m         │ \u001b[90mString\u001b[39m          │ \u001b[90mString\u001b[39m       │\n",
      "├────────┼─────────┼────────────────┼─────────────────┼──────────────┤\n",
      "│ 1      │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 2      │         │ latest         │ Complete Genome │ Major        │\n",
      "│ 3      │         │ latest         │ Complete Genome │ Major        │\n",
      "│ 4      │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 5      │ M298    │ latest         │ Complete Genome │ Major        │\n",
      "│ 6      │ M299    │ latest         │ Complete Genome │ Major        │\n",
      "│ 7      │ M308    │ latest         │ Complete Genome │ Major        │\n",
      "│ 8      │ M23     │ latest         │ Complete Genome │ Major        │\n",
      "│ 9      │         │ latest         │ Complete Genome │ Major        │\n",
      "│ 10     │         │ latest         │ Complete Genome │ Major        │\n",
      "⋮\n",
      "│ 764239 │         │ latest         │ Scaffold        │ Major        │\n",
      "│ 764240 │         │ latest         │ Scaffold        │ Major        │\n",
      "│ 764241 │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 764242 │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 764243 │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 764244 │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 764245 │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 764246 │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 764247 │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 764248 │         │ latest         │ Chromosome      │ Major        │\n",
      "│ 764249 │         │ latest         │ Chromosome      │ Major        │\n",
      "\n",
      "│ Row    │ genome_rep │ seq_rel_date │ asm_name                         │\n",
      "│        │ \u001b[90mString\u001b[39m     │ \u001b[90mString\u001b[39m       │ \u001b[90mString\u001b[39m                           │\n",
      "├────────┼────────────┼──────────────┼──────────────────────────────────┤\n",
      "│ 1      │ Partial    │ 2004/03/01   │ ASM282980v1                      │\n",
      "│ 2      │ Full       │ 2005/09/11   │ ASM397214v1                      │\n",
      "│ 3      │ Full       │ 2006/06/20   │ ASM397216v1                      │\n",
      "│ 4      │ Partial    │ 2007/01/10   │ ASM397218v1                      │\n",
      "│ 5      │ Full       │ 2007/11/01   │ ASM397220v1                      │\n",
      "│ 6      │ Full       │ 2007/11/01   │ ASM397222v1                      │\n",
      "│ 7      │ Full       │ 2007/11/01   │ ASM397224v1                      │\n",
      "│ 8      │ Full       │ 2007/11/01   │ ASM397226v1                      │\n",
      "│ 9      │ Full       │ 1993/08/03   │ ViralProj14816                   │\n",
      "│ 10     │ Full       │ 1993/04/29   │ ViralProj15032                   │\n",
      "⋮\n",
      "│ 764239 │ Full       │ 2020/03/13   │ Triticum_aestivum_weebil_EIv1    │\n",
      "│ 764240 │ Full       │ 2020/03/13   │ Triticum_aestivum_Robigus_EIv1.1 │\n",
      "│ 764241 │ Full       │ 2020/08/06   │ 10wheat_assembly_jagger          │\n",
      "│ 764242 │ Full       │ 2020/08/06   │ 10wheat_assembly_lancer          │\n",
      "│ 764243 │ Full       │ 2020/08/06   │ 10wheat_assembly_arinaLrFor      │\n",
      "│ 764244 │ Full       │ 2020/08/20   │ 10wheat_assembly_stanley         │\n",
      "│ 764245 │ Full       │ 2020/08/14   │ 10wheat_assembly_mace            │\n",
      "│ 764246 │ Full       │ 2020/08/20   │ 10wheat_assembly_sy_mattis       │\n",
      "│ 764247 │ Full       │ 2020/08/15   │ 10wheat_assembly_julius          │\n",
      "│ 764248 │ Full       │ 2020/08/22   │ 10wheat_assembly_landmark1       │\n",
      "│ 764249 │ Full       │ 2020/09/03   │ 10wheat_assembly_norin61         │\n",
      "\n",
      "│ Row    │ submitter       │ gbrs_paired_asm │ paired_asm_comp │\n",
      "│        │ \u001b[90mString\u001b[39m          │ \u001b[90mString\u001b[39m          │ \u001b[90mString\u001b[39m          │\n",
      "├────────┼─────────────────┼─────────────────┼─────────────────┤\n",
      "│ 1      │                 │ GCF_002829805.1 │ identical       │\n",
      "│ 2      │                 │ GCF_003972145.1 │ identical       │\n",
      "│ 3      │                 │ GCF_003972165.1 │ identical       │\n",
      "│ 4      │                 │ GCF_003972185.1 │ identical       │\n",
      "│ 5      │                 │ GCF_003972205.1 │ identical       │\n",
      "│ 6      │                 │ GCF_003972225.1 │ identical       │\n",
      "│ 7      │                 │ GCF_003972245.1 │ identical       │\n",
      "│ 8      │                 │ GCF_003972265.1 │ identical       │\n",
      "│ 9      │                 │ GCF_000851485.1 │ identical       │\n",
      "│ 10     │                 │ GCF_000856565.1 │ identical       │\n",
      "⋮\n",
      "│ 764239 │ EI              │ na              │ na              │\n",
      "│ 764240 │ EI              │ na              │ na              │\n",
      "│ 764241 │ IPK-Gatersleben │ na              │ na              │\n",
      "│ 764242 │ IPK-Gatersleben │ na              │ na              │\n",
      "│ 764243 │ IPK-Gatersleben │ na              │ na              │\n",
      "│ 764244 │ IPK-Gatersleben │ na              │ na              │\n",
      "│ 764245 │ IPK-Gatersleben │ na              │ na              │\n",
      "│ 764246 │ IPK-Gatersleben │ na              │ na              │\n",
      "│ 764247 │ IPK-Gatersleben │ na              │ na              │\n",
      "│ 764248 │ IPK-Gatersleben │ na              │ na              │\n",
      "│ 764249 │ IPK-Gatersleben │ na              │ na              │\n",
      "\n",
      "│ Row    │ ftp_path                                                                                                │\n",
      "│        │ \u001b[90mString\u001b[39m                                                                                                  │\n",
      "├────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤\n",
      "│ 1      │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002/829/805/GCA_002829805.1_ASM282980v1                      │\n",
      "│ 2      │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/145/GCA_003972145.1_ASM397214v1                      │\n",
      "│ 3      │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/165/GCA_003972165.1_ASM397216v1                      │\n",
      "│ 4      │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/185/GCA_003972185.1_ASM397218v1                      │\n",
      "│ 5      │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/205/GCA_003972205.1_ASM397220v1                      │\n",
      "│ 6      │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/225/GCA_003972225.1_ASM397222v1                      │\n",
      "│ 7      │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/245/GCA_003972245.1_ASM397224v1                      │\n",
      "│ 8      │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/265/GCA_003972265.1_ASM397226v1                      │\n",
      "│ 9      │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/851/485/GCA_000851485.1_ViralProj14816                   │\n",
      "│ 10     │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/856/565/GCA_000856565.1_ViralProj15032                   │\n",
      "⋮\n",
      "│ 764239 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/902/810/675/GCA_902810675.1_Triticum_aestivum_weebil_EIv1    │\n",
      "│ 764240 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/902/810/685/GCA_902810685.1_Triticum_aestivum_Robigus_EIv1.1 │\n",
      "│ 764241 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/903/993/795/GCA_903993795.1_10wheat_assembly_jagger          │\n",
      "│ 764242 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/903/993/975/GCA_903993975.1_10wheat_assembly_lancer          │\n",
      "│ 764243 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/903/993/985/GCA_903993985.1_10wheat_assembly_arinaLrFor      │\n",
      "│ 764244 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/903/994/155/GCA_903994155.1_10wheat_assembly_stanley         │\n",
      "│ 764245 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/903/994/175/GCA_903994175.1_10wheat_assembly_mace            │\n",
      "│ 764246 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/903/994/185/GCA_903994185.1_10wheat_assembly_sy_mattis       │\n",
      "│ 764247 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/903/994/195/GCA_903994195.1_10wheat_assembly_julius          │\n",
      "│ 764248 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/903/995/565/GCA_903995565.1_10wheat_assembly_landmark1       │\n",
      "│ 764249 │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/904/066/035/GCA_904066035.1_10wheat_assembly_norin61         │\n",
      "\n",
      "│ Row    │ excluded_from_refseq │ relation_to_type_material │\n",
      "│        │ \u001b[90mString\u001b[39m               │ \u001b[90mString\u001b[39m                    │\n",
      "├────────┼──────────────────────┼───────────────────────────┤\n",
      "│ 1      │ partial              │ ICTV species exemplar     │\n",
      "│ 2      │                      │ ICTV additional isolate   │\n",
      "│ 3      │                      │ ICTV additional isolate   │\n",
      "│ 4      │ partial              │ ICTV additional isolate   │\n",
      "│ 5      │                      │ ICTV additional isolate   │\n",
      "│ 6      │                      │ ICTV additional isolate   │\n",
      "│ 7      │                      │ ICTV additional isolate   │\n",
      "│ 8      │                      │ ICTV additional isolate   │\n",
      "│ 9      │                      │ ICTV species exemplar     │\n",
      "│ 10     │                      │                           │\n",
      "⋮\n",
      "│ 764239 │                      │                           │\n",
      "│ 764240 │                      │                           │\n",
      "│ 764241 │                      │                           │\n",
      "│ 764242 │                      │                           │\n",
      "│ 764243 │                      │                           │\n",
      "│ 764244 │                      │                           │\n",
      "│ 764245 │                      │                           │\n",
      "│ 764246 │                      │                           │\n",
      "│ 764247 │                      │                           │\n",
      "│ 764248 │                      │                           │\n",
      "│ 764249 │                      │                           │\n",
      "\n",
      "│ Row    │ min_ungapped_length │ max_ungapped_length │\n",
      "│        │ \u001b[90mInt64\u001b[39m               │ \u001b[90mInt64\u001b[39m               │\n",
      "├────────┼─────────────────────┼─────────────────────┤\n",
      "│ 1      │ 636                 │ 954                 │\n",
      "│ 2      │ 636                 │ 954                 │\n",
      "│ 3      │ 636                 │ 954                 │\n",
      "│ 4      │ 636                 │ 954                 │\n",
      "│ 5      │ 636                 │ 954                 │\n",
      "│ 6      │ 636                 │ 954                 │\n",
      "│ 7      │ 636                 │ 954                 │\n",
      "│ 8      │ 636                 │ 954                 │\n",
      "│ 9      │ 660                 │ 991                 │\n",
      "│ 10     │ 1344                │ 2016                │\n",
      "⋮\n",
      "│ 764239 │ 7437553000          │ 20961289000         │\n",
      "│ 764240 │ 7437553000          │ 20961289000         │\n",
      "│ 764241 │ 7437553000          │ 20961289000         │\n",
      "│ 764242 │ 7437553000          │ 20961289000         │\n",
      "│ 764243 │ 7437553000          │ 20961289000         │\n",
      "│ 764244 │ 7437553000          │ 20961289000         │\n",
      "│ 764245 │ 7437553000          │ 20961289000         │\n",
      "│ 764246 │ 7437553000          │ 20961289000         │\n",
      "│ 764247 │ 7437553000          │ 20961289000         │\n",
      "│ 764248 │ 7437553000          │ 20961289000         │\n",
      "│ 764249 │ 7437553000          │ 20961289000         │\n",
      "\n",
      "│ Row    │ expected_ungapped_length │ number_of_genomes │\n",
      "│        │ \u001b[90mInt64\u001b[39m                    │ \u001b[90mInt64\u001b[39m             │\n",
      "├────────┼──────────────────────────┼───────────────────┤\n",
      "│ 1      │ 795                      │ 6                 │\n",
      "│ 2      │ 795                      │ 6                 │\n",
      "│ 3      │ 795                      │ 6                 │\n",
      "│ 4      │ 795                      │ 6                 │\n",
      "│ 5      │ 795                      │ 6                 │\n",
      "│ 6      │ 795                      │ 6                 │\n",
      "│ 7      │ 795                      │ 6                 │\n",
      "│ 8      │ 795                      │ 6                 │\n",
      "│ 9      │ 826                      │ 1                 │\n",
      "│ 10     │ 1680                     │ 9                 │\n",
      "⋮\n",
      "│ 764239 │ 14199421073              │ 22                │\n",
      "│ 764240 │ 14199421073              │ 22                │\n",
      "│ 764241 │ 14199421073              │ 22                │\n",
      "│ 764242 │ 14199421073              │ 22                │\n",
      "│ 764243 │ 14199421073              │ 22                │\n",
      "│ 764244 │ 14199421073              │ 22                │\n",
      "│ 764245 │ 14199421073              │ 22                │\n",
      "│ 764246 │ 14199421073              │ 22                │\n",
      "│ 764247 │ 14199421073              │ 22                │\n",
      "│ 764248 │ 14199421073              │ 22                │\n",
      "│ 764249 │ 14199421073              │ 22                │"
     ]
    }
   ],
   "source": [
    "show(joint_table, allcols=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 112903 on refseq\n",
    "# 795 on genbank!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"data-frame\"><thead><tr><th></th><th># assembly_accession</th><th>bioproject</th><th>biosample</th><th>wgs_master</th><th>refseq_category</th><th>taxid</th></tr><tr><th></th><th>String</th><th>String</th><th>String</th><th>String</th><th>String</th><th>Int64</th></tr></thead><tbody><p>8 rows × 26 columns (omitted printing of 20 columns)</p><tr><th>1</th><td>GCA_002829805.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>2</th><td>GCA_003972145.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>3</th><td>GCA_003972165.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>4</th><td>GCA_003972185.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>5</th><td>GCA_003972205.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>6</th><td>GCA_003972225.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>7</th><td>GCA_003972245.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr><tr><th>8</th><td>GCA_003972265.1</td><td></td><td></td><td></td><td>na</td><td>230604</td></tr></tbody></table>"
      ],
      "text/latex": [
       "\\begin{tabular}{r|ccccccc}\n",
       "\t& \\# assembly\\_accession & bioproject & biosample & wgs\\_master & refseq\\_category & taxid & \\\\\n",
       "\t\\hline\n",
       "\t& String & String & String & String & String & Int64 & \\\\\n",
       "\t\\hline\n",
       "\t1 & GCA\\_002829805.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t2 & GCA\\_003972145.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t3 & GCA\\_003972165.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t4 & GCA\\_003972185.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t5 & GCA\\_003972205.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t6 & GCA\\_003972225.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t7 & GCA\\_003972245.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\t8 & GCA\\_003972265.1 &  &  &  & na & 230604 & $\\dots$ \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/plain": [
       "8×26 SubDataFrame. Omitted printing of 22 columns\n",
       "│ Row │ # assembly_accession │ bioproject │ biosample │ wgs_master │\n",
       "│     │ \u001b[90mString\u001b[39m               │ \u001b[90mString\u001b[39m     │ \u001b[90mString\u001b[39m    │ \u001b[90mString\u001b[39m     │\n",
       "├─────┼──────────────────────┼────────────┼───────────┼────────────┤\n",
       "│ 1   │ GCA_002829805.1      │            │           │            │\n",
       "│ 2   │ GCA_003972145.1      │            │           │            │\n",
       "│ 3   │ GCA_003972165.1      │            │           │            │\n",
       "│ 4   │ GCA_003972185.1      │            │           │            │\n",
       "│ 5   │ GCA_003972205.1      │            │           │            │\n",
       "│ 6   │ GCA_003972225.1      │            │           │            │\n",
       "│ 7   │ GCA_003972245.1      │            │           │            │\n",
       "│ 8   │ GCA_003972265.1      │            │           │            │"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group = first(DataFrames.groupby(joint_table, \"species_taxid\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many groups are there?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8×26 SubDataFrame\n",
      "│ Row │ # assembly_accession │ bioproject │ biosample │ wgs_master │\n",
      "│     │ \u001b[90mString\u001b[39m               │ \u001b[90mString\u001b[39m     │ \u001b[90mString\u001b[39m    │ \u001b[90mString\u001b[39m     │\n",
      "├─────┼──────────────────────┼────────────┼───────────┼────────────┤\n",
      "│ 1   │ GCA_002829805.1      │            │           │            │\n",
      "│ 2   │ GCA_003972145.1      │            │           │            │\n",
      "│ 3   │ GCA_003972165.1      │            │           │            │\n",
      "│ 4   │ GCA_003972185.1      │            │           │            │\n",
      "│ 5   │ GCA_003972205.1      │            │           │            │\n",
      "│ 6   │ GCA_003972225.1      │            │           │            │\n",
      "│ 7   │ GCA_003972245.1      │            │           │            │\n",
      "│ 8   │ GCA_003972265.1      │            │           │            │\n",
      "\n",
      "│ Row │ refseq_category │ taxid  │ species_taxid │ organism_name     │\n",
      "│     │ \u001b[90mString\u001b[39m          │ \u001b[90mInt64\u001b[39m  │ \u001b[90mInt64\u001b[39m         │ \u001b[90mString\u001b[39m            │\n",
      "├─────┼─────────────────┼────────┼───────────────┼───────────────────┤\n",
      "│ 1   │ na              │ 230604 │ 1922242       │ Extra small virus │\n",
      "│ 2   │ na              │ 230604 │ 1922242       │ Extra small virus │\n",
      "│ 3   │ na              │ 230604 │ 1922242       │ Extra small virus │\n",
      "│ 4   │ na              │ 230604 │ 1922242       │ Extra small virus │\n",
      "│ 5   │ na              │ 230604 │ 1922242       │ Extra small virus │\n",
      "│ 6   │ na              │ 230604 │ 1922242       │ Extra small virus │\n",
      "│ 7   │ na              │ 230604 │ 1922242       │ Extra small virus │\n",
      "│ 8   │ na              │ 230604 │ 1922242       │ Extra small virus │\n",
      "\n",
      "│ Row │ infraspecific_name │ isolate │ version_status │ assembly_level  │\n",
      "│     │ \u001b[90mString\u001b[39m             │ \u001b[90mString\u001b[39m  │ \u001b[90mString\u001b[39m         │ \u001b[90mString\u001b[39m          │\n",
      "├─────┼────────────────────┼─────────┼────────────────┼─────────────────┤\n",
      "│ 1   │                    │         │ latest         │ Chromosome      │\n",
      "│ 2   │                    │         │ latest         │ Complete Genome │\n",
      "│ 3   │                    │         │ latest         │ Complete Genome │\n",
      "│ 4   │                    │         │ latest         │ Chromosome      │\n",
      "│ 5   │                    │ M298    │ latest         │ Complete Genome │\n",
      "│ 6   │                    │ M299    │ latest         │ Complete Genome │\n",
      "│ 7   │                    │ M308    │ latest         │ Complete Genome │\n",
      "│ 8   │                    │ M23     │ latest         │ Complete Genome │\n",
      "\n",
      "│ Row │ release_type │ genome_rep │ seq_rel_date │ asm_name    │ submitter │\n",
      "│     │ \u001b[90mString\u001b[39m       │ \u001b[90mString\u001b[39m     │ \u001b[90mString\u001b[39m       │ \u001b[90mString\u001b[39m      │ \u001b[90mString\u001b[39m    │\n",
      "├─────┼──────────────┼────────────┼──────────────┼─────────────┼───────────┤\n",
      "│ 1   │ Major        │ Partial    │ 2004/03/01   │ ASM282980v1 │           │\n",
      "│ 2   │ Major        │ Full       │ 2005/09/11   │ ASM397214v1 │           │\n",
      "│ 3   │ Major        │ Full       │ 2006/06/20   │ ASM397216v1 │           │\n",
      "│ 4   │ Major        │ Partial    │ 2007/01/10   │ ASM397218v1 │           │\n",
      "│ 5   │ Major        │ Full       │ 2007/11/01   │ ASM397220v1 │           │\n",
      "│ 6   │ Major        │ Full       │ 2007/11/01   │ ASM397222v1 │           │\n",
      "│ 7   │ Major        │ Full       │ 2007/11/01   │ ASM397224v1 │           │\n",
      "│ 8   │ Major        │ Full       │ 2007/11/01   │ ASM397226v1 │           │\n",
      "\n",
      "│ Row │ gbrs_paired_asm │ paired_asm_comp │\n",
      "│     │ \u001b[90mString\u001b[39m          │ \u001b[90mString\u001b[39m          │\n",
      "├─────┼─────────────────┼─────────────────┤\n",
      "│ 1   │ GCF_002829805.1 │ identical       │\n",
      "│ 2   │ GCF_003972145.1 │ identical       │\n",
      "│ 3   │ GCF_003972165.1 │ identical       │\n",
      "│ 4   │ GCF_003972185.1 │ identical       │\n",
      "│ 5   │ GCF_003972205.1 │ identical       │\n",
      "│ 6   │ GCF_003972225.1 │ identical       │\n",
      "│ 7   │ GCF_003972245.1 │ identical       │\n",
      "│ 8   │ GCF_003972265.1 │ identical       │\n",
      "\n",
      "│ Row │ ftp_path                                                                           │\n",
      "│     │ \u001b[90mString\u001b[39m                                                                             │\n",
      "├─────┼────────────────────────────────────────────────────────────────────────────────────┤\n",
      "│ 1   │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/002/829/805/GCA_002829805.1_ASM282980v1 │\n",
      "│ 2   │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/145/GCA_003972145.1_ASM397214v1 │\n",
      "│ 3   │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/165/GCA_003972165.1_ASM397216v1 │\n",
      "│ 4   │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/185/GCA_003972185.1_ASM397218v1 │\n",
      "│ 5   │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/205/GCA_003972205.1_ASM397220v1 │\n",
      "│ 6   │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/225/GCA_003972225.1_ASM397222v1 │\n",
      "│ 7   │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/245/GCA_003972245.1_ASM397224v1 │\n",
      "│ 8   │ ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/003/972/265/GCA_003972265.1_ASM397226v1 │\n",
      "\n",
      "│ Row │ excluded_from_refseq │ relation_to_type_material │ min_ungapped_length │\n",
      "│     │ \u001b[90mString\u001b[39m               │ \u001b[90mString\u001b[39m                    │ \u001b[90mInt64\u001b[39m               │\n",
      "├─────┼──────────────────────┼───────────────────────────┼─────────────────────┤\n",
      "│ 1   │ partial              │ ICTV species exemplar     │ 636                 │\n",
      "│ 2   │                      │ ICTV additional isolate   │ 636                 │\n",
      "│ 3   │                      │ ICTV additional isolate   │ 636                 │\n",
      "│ 4   │ partial              │ ICTV additional isolate   │ 636                 │\n",
      "│ 5   │                      │ ICTV additional isolate   │ 636                 │\n",
      "│ 6   │                      │ ICTV additional isolate   │ 636                 │\n",
      "│ 7   │                      │ ICTV additional isolate   │ 636                 │\n",
      "│ 8   │                      │ ICTV additional isolate   │ 636                 │\n",
      "\n",
      "│ Row │ max_ungapped_length │ expected_ungapped_length │ number_of_genomes │\n",
      "│     │ \u001b[90mInt64\u001b[39m               │ \u001b[90mInt64\u001b[39m                    │ \u001b[90mInt64\u001b[39m             │\n",
      "├─────┼─────────────────────┼──────────────────────────┼───────────────────┤\n",
      "│ 1   │ 954                 │ 795                      │ 6                 │\n",
      "│ 2   │ 954                 │ 795                      │ 6                 │\n",
      "│ 3   │ 954                 │ 795                      │ 6                 │\n",
      "│ 4   │ 954                 │ 795                      │ 6                 │\n",
      "│ 5   │ 954                 │ 795                      │ 6                 │\n",
      "│ 6   │ 954                 │ 795                      │ 6                 │\n",
      "│ 7   │ 954                 │ 795                      │ 6                 │\n",
      "│ 8   │ 954                 │ 795                      │ 6                 │"
     ]
    }
   ],
   "source": [
    "show(group, allcols=true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8-element Array{FASTX.FASTA.Record,1}:\n",
       " FASTX.FASTA.Record:\n",
       "   identifier: AY247793.1\n",
       "  description: Macrobrachium rosenbergii XSV virus capsid protein gene, complete cds\n",
       "     sequence: CCACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGTA…\n",
       " FASTX.FASTA.Record:\n",
       "   identifier: DQ174318.1\n",
       "  description: Macrobrachium rosenbergii XSV virus nucleocapsid protein CP17 (cp17) and CP16 (cp16) genes, complete cds\n",
       "     sequence: TTGTAAAACAAGCTTCGGCTTGTTTTACACGTCTAGCTG…\n",
       " FASTX.FASTA.Record:\n",
       "   identifier: DQ521573.1\n",
       "  description: Macrobrachium rosenbergii XSV virus nucleocapsid protein CP17 gene, complete cds\n",
       "     sequence: CCACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGAA…\n",
       " FASTX.FASTA.Record:\n",
       "   identifier: EF159731.1\n",
       "  description: Macrobrachium rosenbergii XSV virus capsid protein gene, partial cds\n",
       "     sequence: ATGAATAAGCGCATTAATAATAATCGGAGAACCATGAGA…\n",
       " FASTX.FASTA.Record:\n",
       "   identifier: EU150130.1\n",
       "  description: Macrobrachium rosenbergii XSV virus isolate M298 capsid protein gene, complete cds\n",
       "     sequence: TACACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGT…\n",
       " FASTX.FASTA.Record:\n",
       "   identifier: EU150131.1\n",
       "  description: Macrobrachium rosenbergii XSV virus isolate M299 capsid protein gene, complete cds\n",
       "     sequence: CGCGGAATTTGGTTACACGTCTAGCTGCTGACGTTAAAT…\n",
       " FASTX.FASTA.Record:\n",
       "   identifier: EU150132.1\n",
       "  description: Macrobrachium rosenbergii XSV virus isolate M308 capsid protein gene, complete cds\n",
       "     sequence: TACACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGT…\n",
       " FASTX.FASTA.Record:\n",
       "   identifier: EU150133.1\n",
       "  description: Macrobrachium rosenbergii XSV virus isolate M23 capsid protein gene, complete cds\n",
       "     sequence: TACACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGT…"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes = reduce(vcat, map(ftp_path -> get_genome_from_ftp_directory(ftp_path), group[!, \"ftp_path\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8-element Array{BioSequences.LongSequence{BioSequences.DNAAlphabet{4}},1}:\n",
       " CCACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGTA…TTAGGCATTATTCCGGGTGCGTCGATAAATAAAGACCTT\n",
       " TTGTAAAACAAGCTTCGGCTTGTTTTACACGTCTAGCTG…ACAATTAGTTCTCATGGTGGATTTACACATTGACGGTTG\n",
       " CCACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGAA…TTGTGGCAAGGGGAGACGTATTGGTTAGGCATTATTCCG\n",
       " ATGAATAAGCGCATTAATAATAATCGGAGAACCATGAGA…CTGCTAAAGATCCAGTTCCAGTGGTTCGTGCTTTGGTAA\n",
       " TACACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGT…GGTTAGGCATTATTCCGGGTGCGTCGATAAATAAAGACA\n",
       " CGCGGAATTTGGTTACACGTCTAGCTGCTGACGTTAAAT…GGTTAGGCATTATTCCGGGTGCGTCGATAAATAAAGACA\n",
       " TACACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGT…GGTTAGGCATTATTCCGGGTGCGTCGATAAATAAAGACA\n",
       " TACACGTCTAGCTGCTGACGTTAAATGCAGCCGGGTGGT…GGTTAGGCATTATTCCGGGTGCGTCGATAAATAAAGACA"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_sequences = FASTX.sequence.(genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Eisenia.KmerGraph{BioSequences.BigMer{BioSequences.DNAAlphabet{2},7}}({964, 1057} undirected simple Int64 graph, Dict{LightGraphs.SimpleGraphs.SimpleEdge{Int64},Array{Eisenia.EdgeEvidence,1}}(Edge 263 => 853 => [Eisenia.EdgeEvidence(1, 593), Eisenia.EdgeEvidence(2, 619), Eisenia.EdgeEvidence(3, 593), Eisenia.EdgeEvidence(5, 594), Eisenia.EdgeEvidence(6, 607), Eisenia.EdgeEvidence(7, 594), Eisenia.EdgeEvidence(8, 594)],Edge 87 => 88 => [Eisenia.EdgeEvidence(2, 9)],Edge 17 => 104 => [Eisenia.EdgeEvidence(1, 305), Eisenia.EdgeEvidence(2, 331), Eisenia.EdgeEvidence(3, 305), Eisenia.EdgeEvidence(4, 243), Eisenia.EdgeEvidence(5, 306), Eisenia.EdgeEvidence(6, 319), Eisenia.EdgeEvidence(7, 306), Eisenia.EdgeEvidence(8, 306)],Edge 38 => 473 => [Eisenia.EdgeEvidence(6, 4)],Edge 373 => 600 => [Eisenia.EdgeEvidence(1, 163), Eisenia.EdgeEvidence(2, 189), Eisenia.EdgeEvidence(3, 163), Eisenia.EdgeEvidence(4, 101), Eisenia.EdgeEvidence(5, 164), Eisenia.EdgeEvidence(6, 177), Eisenia.EdgeEvidence(7, 164), Eisenia.EdgeEvidence(8, 164)],Edge 420 => 768 => [Eisenia.EdgeEvidence(1, 312), Eisenia.EdgeEvidence(2, 338), Eisenia.EdgeEvidence(3, 312), Eisenia.EdgeEvidence(4, 250), Eisenia.EdgeEvidence(5, 313), Eisenia.EdgeEvidence(6, 326), Eisenia.EdgeEvidence(7, 313), Eisenia.EdgeEvidence(8, 313)],Edge 157 => 480 => [Eisenia.EdgeEvidence(1, 615), Eisenia.EdgeEvidence(2, 641), Eisenia.EdgeEvidence(3, 615), Eisenia.EdgeEvidence(5, 616), Eisenia.EdgeEvidence(6, 629), Eisenia.EdgeEvidence(7, 616), Eisenia.EdgeEvidence(8, 616)],Edge 388 => 877 => [Eisenia.EdgeEvidence(1, 459), Eisenia.EdgeEvidence(2, 485), Eisenia.EdgeEvidence(3, 459), Eisenia.EdgeEvidence(4, 397), Eisenia.EdgeEvidence(5, 460), Eisenia.EdgeEvidence(6, 473), Eisenia.EdgeEvidence(7, 460), Eisenia.EdgeEvidence(8, 460)],Edge 496 => 574 => [Eisenia.EdgeEvidence(4, 130)],Edge 639 => 810 => [Eisenia.EdgeEvidence(2, 372), Eisenia.EdgeEvidence(3, 346), Eisenia.EdgeEvidence(4, 284), Eisenia.EdgeEvidence(5, 347), Eisenia.EdgeEvidence(6, 360), Eisenia.EdgeEvidence(8, 347)]…), BioSequences.BigMer{BioSequences.DNAAlphabet{2},7}[AAAACAA, AAAACAT, AAAACCC, AAAACGA, AAAACTA, AAAACTG, AAAATAT, AAACAAG, AAACAAT, AAACATT  …  TGTGTAA, TGTTAAA, TGTTGAA, TTAACAA, TTAATAA, TTACCAA, TTACTAA, TTCTCAA, TTGTAAA, TTTAAAA], [9, 1, 7, 7, 8, 8, 9, 9, 1, 1  …  7, 1, 6, 1, 8, 15, 4, 16, 1, 8])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph = Eisenia.KmerGraph(BioSequences.BigDNAMer{7}, genome_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# graph = Eisenia.KmerGraph(BioSequences.BigDNAMer{7}, genome_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \tComment\n",
    "# H \tHeader\n",
    "# S \tSegment\n",
    "# L \tLink\n",
    "# C \tContainment\n",
    "# P \tPath\n",
    "\n",
    "# H\tVN:Z:1.0\n",
    "# S\t11\tACCTT\n",
    "# S\t12\tTCAAGG\n",
    "# S\t13\tCTTGATT\n",
    "# L\t11\t+\t12\t-\t4M\n",
    "# L\t12\t-\t13\t+\t5M\n",
    "# L\t11\t+\t13\t+\t3M\n",
    "# P\t14\t11+,12-,13+\t4M,5M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Edge 1 => 8"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge = first(LightGraphs.edges(graph.graph))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "orientations = assess_path_orientations(path, kmers, initial_orientation)\n",
    "if orientations == nothing\n",
    "    path_likelihood = 0.0\n",
    "    oriented_path = OrientedKmer[]\n",
    "else\n",
    "    oriented_path = orient_path(path, orientations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2-element Array{Eisenia.OrientedKmer,1}:\n",
       " Eisenia.OrientedKmer(1, true)\n",
       " Eisenia.OrientedKmer(8, true)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function kmer_pair_to_oriented_path(kmer_pair, graph)\n",
    "    path = [kmer_pair[1], kmer_pair[2]]\n",
    "    orientations = Eisenia.assess_path_orientations(path, graph.kmers, true)\n",
    "    if orientations == nothing\n",
    "        orientations = Eisenia.assess_path_orientations(path, graph.kmers, false)\n",
    "    end\n",
    "#     if orientations == nothing\n",
    "#         path = reverse(path)\n",
    "#         @show \"reversing path...\"\n",
    "#         orientations = Eisenia.assess_path_orientations(path, graph.kmers, true)\n",
    "#         if orientations == nothing\n",
    "#             orientations = Eisenia.assess_path_orientations(path, graph.kmers, false)\n",
    "#         end\n",
    "#     end\n",
    "    if orientations == nothing\n",
    "        @show graph.kmers[path]\n",
    "        error()\n",
    "    end\n",
    "    return Eisenia.orient_path(path, orientations)\n",
    "end\n",
    "\n",
    "kmer_pair_to_oriented_path(edge.src => edge.dst, graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "GFA_file = \"$(DIR)/resolving-related-genomes.k-7.gfa\"\n",
    "open(GFA_file, \"w\") do io\n",
    "    println(io, \"H\\tVN:Z:1.0\")\n",
    "    for (i, kmer) in enumerate(graph.kmers)\n",
    "        fields = [\"S\", \"$i\", string(kmer)]\n",
    "        line = join(fields, '\\t')\n",
    "        println(io, line)\n",
    "    end\n",
    "    for edge in LightGraphs.edges(graph.graph)\n",
    "        oriented_src, oriented_dst = kmer_pair_to_oriented_path(edge.src => edge.dst, graph)\n",
    "        overlap = length(graph.kmers[1]) - 1\n",
    "        link = [\"L\",\n",
    "                    oriented_src.index,\n",
    "                    oriented_src.orientation ? '+' : '-',\n",
    "                    oriented_dst.index,\n",
    "                    oriented_dst.orientation ? '+' : '-',\n",
    "                    \"$(overlap)M\"]\n",
    "        line = join(link, '\\t')\n",
    "        println(io, line)\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This worked great!!\n",
    "# try and run bandage from the command line and show the graph inline\n",
    "# show with all possible nodes merged and default"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.5.3",
   "language": "julia",
   "name": "julia-1.5"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
