<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>- ¬∑ Mycelia</title><meta name="title" content="- ¬∑ Mycelia"/><meta property="og:title" content="- ¬∑ Mycelia"/><meta property="twitter:title" content="- ¬∑ Mycelia"/><meta name="description" content="Documentation for Mycelia."/><meta property="og:description" content="Documentation for Mycelia."/><meta property="twitter:description" content="Documentation for Mycelia."/><script data-outdated-warner src="../../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../search_index.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../../">Mycelia</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../../">Home</a></li><li><a class="tocitem" href="../../../installation/">Installation</a></li><li><a class="tocitem" href="../../../getting-started/">Getting Started</a></li><li><a class="tocitem" href="../../../concepts/">Concepts</a></li><li><a class="tocitem" href="../../../probabilistic-assembly-hub/">Probabilistic Assembly</a></li><li><a class="tocitem" href="../../../tutorials/">Tutorials</a></li><li><span class="tocitem">Documentation</span><ul><li><a class="tocitem" href="../../../architecture/">Architecture Overview</a></li><li><a class="tocitem" href="../../../assembly-method-selection/">Assembly Method Selection</a></li><li><a class="tocitem" href="../../../workflow-map/">Workflow &amp; Tool Map</a></li><li><a class="tocitem" href="../../../api/function-coverage/">Function Coverage Audit</a></li><li><a class="tocitem" href="../../../performance/">Performance Guide</a></li><li><a class="tocitem" href="../../../faq/">FAQ</a></li></ul></li><li><span class="tocitem">API Reference</span><ul><li><a class="tocitem" href="../../../api-reference/">Complete API Reference</a></li><li><a class="tocitem" href="../../../api/all-functions/">Complete API Surface</a></li><li><input class="collapse-toggle" id="menuitem-8-3" type="checkbox"/><label class="tocitem" for="menuitem-8-3"><span class="docs-label">Workflows</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../api/workflows/assembly-suite/">Assembly Suite</a></li><li><a class="tocitem" href="../../../api/workflows/data-acquisition/">Data Acquisition</a></li><li><a class="tocitem" href="../../../api/workflows/quality-control/">Quality Control</a></li><li><a class="tocitem" href="../../../api/workflows/sequence-analysis/">Sequence Analysis</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-8-4" type="checkbox"/><label class="tocitem" for="menuitem-8-4"><span class="docs-label">Quick Reference</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../api/quick-reference/function-index/">Function Index</a></li><li><a class="tocitem" href="../../../api/quick-reference/parameter-guide/">Parameter Guide</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-8-5" type="checkbox"/><label class="tocitem" for="menuitem-8-5"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../api/examples/basic-workflows/">Basic Workflows</a></li></ul></li><li><a class="tocitem" href="../../../api/">Legacy API Documentation</a></li><li><a class="tocitem" href="../../../related-projects/">Related Projects</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>-</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>-</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/cjprybol/Mycelia" title="View the repository on GitHub"><span class="docs-icon fa-brands">ÔÇõ</span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/cjprybol/Mycelia/blob/master/tutorials/00_assembly_in_5_minutes.jl" title="Edit source on GitHub"><span class="docs-icon fa-solid">ÔÅÑ</span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><pre><code class="language-julia hljs">&quot;&quot;&quot;</code></pre><p>Assembly in 5 Minutes with Mycelia</p><pre><code class="language-julia hljs">This tutorial will take you from zero to assembled genome in under 5 minutes.
No prior experience with genome assembly required!

# What You&#39;ll Learn
- How to install and load Mycelia
- How to perform your first genome assembly
- How to interpret the results
- How to save your assembly

# Prerequisites
- Julia installed (version 1.10 or higher)
- About 1GB of free disk space
- Internet connection (for package installation)

Let&#39;s get started!
&quot;&quot;&quot;</code></pre><p>%% [markdown]</p><h2 id="Step-1:-Installation-(One-time-setup)"><a class="docs-heading-anchor" href="#Step-1:-Installation-(One-time-setup)">Step 1: Installation (One-time setup)</a><a id="Step-1:-Installation-(One-time-setup)-1"></a><a class="docs-heading-anchor-permalink" href="#Step-1:-Installation-(One-time-setup)" title="Permalink"></a></h2><p>If you haven&#39;t installed Mycelia yet, run this cell:</p><p>%% Uncomment the next two lines if you need to install Mycelia import Pkg Pkg.add(url=&quot;https://github.com/cjprybol/Mycelia.git&quot;)</p><p>%% [markdown]</p><h2 id="Step-2:-Load-Mycelia"><a class="docs-heading-anchor" href="#Step-2:-Load-Mycelia">Step 2: Load Mycelia</a><a id="Step-2:-Load-Mycelia-1"></a><a class="docs-heading-anchor-permalink" href="#Step-2:-Load-Mycelia" title="Permalink"></a></h2><p>Import the packages we&#39;ll need:</p><p>%%</p><pre><code class="language-julia hljs">import Mycelia
import FASTX
import Pkg
import Statistics
import Kmers

println(&quot;‚úÖ Mycelia loaded successfully!&quot;)</code></pre><p>%% [markdown]</p><h2 id="Step-3:-Get-Some-Data-to-Assemble"><a class="docs-heading-anchor" href="#Step-3:-Get-Some-Data-to-Assemble">Step 3: Get Some Data to Assemble</a><a id="Step-3:-Get-Some-Data-to-Assemble-1"></a><a class="docs-heading-anchor-permalink" href="#Step-3:-Get-Some-Data-to-Assemble" title="Permalink"></a></h2><p>We&#39;ll use a small virus genome (phiX174) as our test case. It&#39;s only 5,386 base pairs, so it assembles quickly!</p><p>%% Download the reference genome</p><pre><code class="language-julia hljs">println(&quot;üì• Downloading test genome (phiX174)...&quot;)
reference_file = Mycelia.download_genome_by_accession(accession=&quot;NC_001422.1&quot;)
println(&quot;‚úÖ Downloaded to: $reference_file&quot;)</code></pre><p>Let&#39;s see what we downloaded</p><pre><code class="language-julia hljs">ref_record = first(Mycelia.open_fastx(reference_file))
println(&quot;\nüìä Reference genome info:&quot;)
println(&quot;  Name: &quot;, FASTX.identifier(ref_record))
println(&quot;  Length: &quot;, length(FASTX.sequence(ref_record)), &quot; bp&quot;)</code></pre><p>%% [markdown]</p><h2 id="Step-4:-Simulate-Some-Sequencing-Reads"><a class="docs-heading-anchor" href="#Step-4:-Simulate-Some-Sequencing-Reads">Step 4: Simulate Some Sequencing Reads</a><a id="Step-4:-Simulate-Some-Sequencing-Reads-1"></a><a class="docs-heading-anchor-permalink" href="#Step-4:-Simulate-Some-Sequencing-Reads" title="Permalink"></a></h2><p>For this tutorial, we&#39;ll simulate some reads. In real life, you&#39;d use actual sequencing data from your experiment.</p><p>%% Simulate PacBio-style long reads with realistic error rates</p><pre><code class="language-julia hljs">println(&quot;\nüß¨ Simulating sequencing reads...&quot;)
reads_file = Mycelia.simulate_pacbio_reads(
    fasta=reference_file,
    quantity=&quot;20x&quot;      # 20x coverage
)
println(&quot;‚úÖ Created reads file: $reads_file&quot;)</code></pre><p>Check the reads</p><pre><code class="language-julia hljs">reads = collect(Mycelia.open_fastx(reads_file))
println(&quot;\nüìä Simulated reads info:&quot;)
println(&quot;  Number of reads: &quot;, length(reads))
println(&quot;  Average length: &quot;, round(Statistics.mean(length(FASTX.sequence(r)) for r in reads)), &quot; bp&quot;)
println(&quot;  Total bases: &quot;, sum(length(FASTX.sequence(r)) for r in reads), &quot; bp&quot;)</code></pre><p>%% [markdown]</p><h2 id="Step-5:-Run-the-Assembly!"><a class="docs-heading-anchor" href="#Step-5:-Run-the-Assembly!">Step 5: Run the Assembly! üöÄ</a><a id="Step-5:-Run-the-Assembly!-1"></a><a class="docs-heading-anchor-permalink" href="#Step-5:-Run-the-Assembly!" title="Permalink"></a></h2><p>This is where the magic happens. Mycelia will automatically:</p><ul><li>Detect that you have FASTQ data with quality scores</li><li>Choose optimal parameters</li><li>Build quality-aware graphs</li><li>Find the best assembly path</li></ul><p>%%</p><pre><code class="language-julia hljs">println(&quot;\nüîß Starting probabilistic assembly...&quot;)
println(&quot;  Using intelligent assembly mode (automatic parameter selection)&quot;)</code></pre><p>Record start time</p><pre><code class="language-julia hljs">start_time = time()</code></pre><p>Run the assembly</p><pre><code class="language-julia hljs">assembly_result = Mycelia.mycelia_assemble(
    reads_file;
    output_dir=&quot;my_first_assembly&quot;,
    max_k=51,  # Maximum k-mer size to try
    memory_limit=4_000_000_000  # 4GB memory limit
)</code></pre><p>Calculate elapsed time</p><pre><code class="language-julia hljs">elapsed = round(time() - start_time, digits=1)
println(&quot;\n‚úÖ Assembly completed in $elapsed seconds!&quot;)</code></pre><p>%% [markdown]</p><h2 id="Step-6:-Check-Your-Results"><a class="docs-heading-anchor" href="#Step-6:-Check-Your-Results">Step 6: Check Your Results</a><a id="Step-6:-Check-Your-Results-1"></a><a class="docs-heading-anchor-permalink" href="#Step-6:-Check-Your-Results" title="Permalink"></a></h2><p>Let&#39;s see how well we did:</p><p>%% Load the assembled contigs - check what files were created</p><pre><code class="language-julia hljs">println(&quot;üìÅ Assembly files created:&quot;)
for file in readdir(&quot;my_first_assembly&quot;)
    println(&quot;  - $file&quot;)
end</code></pre><p>Extract contigs from assembly result The assembly returns a dictionary with :final_assembly containing sequence strings</p><pre><code class="language-julia hljs">final_sequences = get(assembly_result, :final_assembly, String[])
println(&quot;‚úÖ Assembly produced $(length(final_sequences)) unique sequences&quot;)</code></pre><p>Convert strings to FASTA records for further analysis</p><pre><code class="language-julia hljs">contigs = FASTX.FASTA.Record[]
for (i, seq) in enumerate(final_sequences)
    if !isempty(seq)  # Skip empty sequences
        record = FASTX.FASTA.Record(&quot;contig_$i&quot;, seq)
        push!(contigs, record)
    end
end
println(&quot;‚úÖ Created $(length(contigs)) FASTA contigs from assembly&quot;)

println(&quot;\nüìä Assembly Results:&quot;)
println(&quot;  Number of contigs: &quot;, length(contigs))</code></pre><p>Calculate basic statistics</p><pre><code class="language-julia hljs">contig_lengths = [length(FASTX.sequence(c)) for c in contigs]
total_length = sum(contig_lengths)
println(&quot;  Total length: &quot;, total_length, &quot; bp&quot;)
println(&quot;  Longest contig: &quot;, maximum(contig_lengths), &quot; bp&quot;)</code></pre><p>Compare to reference</p><pre><code class="language-julia hljs">ref_length = length(FASTX.sequence(ref_record))
completeness = round(100 * total_length / ref_length, digits=1)
println(&quot;\nüéØ Assembly completeness: $completeness%&quot;)

if length(contigs) == 1 &amp;&amp; abs(total_length - ref_length) &lt; 100
    println(&quot;üåü Perfect assembly! The genome was assembled into a single contig!&quot;)
elseif completeness &gt; 95
    println(&quot;üéâ Excellent assembly! Nearly complete genome recovered.&quot;)
elseif completeness &gt; 80
    println(&quot;üëç Good assembly! Most of the genome was recovered.&quot;)
else
    println(&quot;üîß Assembly needs optimization. Try adjusting parameters.&quot;)
end</code></pre><p>%% [markdown]</p><h2 id="Step-7:-Save-Your-Assembly"><a class="docs-heading-anchor" href="#Step-7:-Save-Your-Assembly">Step 7: Save Your Assembly</a><a id="Step-7:-Save-Your-Assembly-1"></a><a class="docs-heading-anchor-permalink" href="#Step-7:-Save-Your-Assembly" title="Permalink"></a></h2><p>Save the assembly for further analysis:</p><p>%% Save to a new file with a descriptive name</p><pre><code class="language-julia hljs">output_file = &quot;phiX174_mycelia_assembly.fasta&quot;
if !isempty(contigs)
    open(output_file, &quot;w&quot;) do io
        writer = FASTX.FASTA.Writer(io)
        for contig in contigs
            write(writer, contig)
        end
    end
    println(&quot;\nüíæ Assembly saved to: $output_file&quot;)
else
    println(&quot;\n‚ö†Ô∏è  No contigs found in assembly result&quot;)
end</code></pre><p>%% [markdown]</p><h2 id="Step-8:-Visualize-Assembly-Quality-(Optional)"><a class="docs-heading-anchor" href="#Step-8:-Visualize-Assembly-Quality-(Optional)">Step 8: Visualize Assembly Quality (Optional)</a><a id="Step-8:-Visualize-Assembly-Quality-(Optional)-1"></a><a class="docs-heading-anchor-permalink" href="#Step-8:-Visualize-Assembly-Quality-(Optional)" title="Permalink"></a></h2><p>Let&#39;s create a simple visualization of our assembly:</p><p>%% Assembly quality analysis</p><pre><code class="language-julia hljs">println(&quot;\nüìà Assembly Quality Analysis:&quot;)</code></pre><p>Basic coverage check</p><pre><code class="language-julia hljs">println(&quot;üìä Coverage Analysis:&quot;)
println(&quot;  Input reads: $(length(reads)) reads&quot;)
println(&quot;  Output contigs: $(length(contigs)) contigs&quot;)

if !isempty(contigs)
    total_assembled = sum(length(FASTX.sequence(c)) for c in contigs)
    println(&quot;  Total assembled length: $total_assembled bp&quot;)</code></pre><p>Compare to reference length</p><pre><code class="language-julia hljs">    ref_length = length(FASTX.sequence(ref_record))
    recovery_rate = round(100 * total_assembled / ref_length, digits=1)
    println(&quot;  Genome recovery rate: $recovery_rate%&quot;)</code></pre><p>K-mer analysis comparison (if assembly file was saved successfully)</p><pre><code class="language-julia hljs">    if isfile(output_file)
        println(&quot;\nüß¨ K-mer Analysis:&quot;)</code></pre><p>Count k-mers in reference</p><pre><code class="language-julia hljs">        ref_kmers = Mycelia.count_canonical_kmers(Kmers.DNAKmer{21}, reference_file)
        println(&quot;  Reference unique 21-mers: &quot;, length(ref_kmers))</code></pre><p>Count k-mers in assembly</p><pre><code class="language-julia hljs">        asm_kmers = Mycelia.count_canonical_kmers(Kmers.DNAKmer{21}, output_file)
        println(&quot;  Assembly unique 21-mers: &quot;, length(asm_kmers))</code></pre><p>Calculate k-mer recovery</p><pre><code class="language-julia hljs">        shared_kmers = length(intersect(keys(ref_kmers), keys(asm_kmers)))
        kmer_recovery = round(100 * shared_kmers / length(ref_kmers), digits=1)
        println(&quot;  K-mer recovery rate: $kmer_recovery%&quot;)
    else
        println(&quot;  ‚ö†Ô∏è  K-mer analysis skipped - assembly file not available&quot;)
    end
else
    println(&quot;  ‚ö†Ô∏è  No contigs produced - assembly may need debugging&quot;)
end</code></pre><p>%% [markdown]</p><h2 id="Congratulations!"><a class="docs-heading-anchor" href="#Congratulations!">üéâ Congratulations!</a><a id="Congratulations!-1"></a><a class="docs-heading-anchor-permalink" href="#Congratulations!" title="Permalink"></a></h2><p>You&#39;ve just completed your first genome assembly with Mycelia!</p><h3 id="What-Just-Happened?"><a class="docs-heading-anchor" href="#What-Just-Happened?">What Just Happened?</a><a id="What-Just-Happened?-1"></a><a class="docs-heading-anchor-permalink" href="#What-Just-Happened?" title="Permalink"></a></h3><p>Mycelia&#39;s intelligent assembly:</p><ol><li><strong>Analyzed your reads</strong> to determine optimal parameters</li><li><strong>Built quality-aware graphs</strong> preserving base quality information</li><li><strong>Used probabilistic algorithms</strong> to find the best assembly path</li><li><strong>Self-optimized</strong> through multiple k-mer sizes</li><li><strong>Produced a high-quality assembly</strong> automatically</li></ol><h3 id="Next-Steps"><a class="docs-heading-anchor" href="#Next-Steps">Next Steps</a><a id="Next-Steps-1"></a><a class="docs-heading-anchor-permalink" href="#Next-Steps" title="Permalink"></a></h3><p>Now that you&#39;ve seen Mycelia in action, you can:</p><ol><li><strong>Try with your own data</strong> - Replace the simulated reads with your FASTQ file</li><li><strong>Explore different modes</strong>:<ul><li><code>method=:iterative</code> for maximum accuracy</li><li><code>method=:quality_aware</code> for speed with good quality data</li></ul></li><li><strong>Learn more</strong>:<ul><li><a href="../../../theoretical-foundations/">Understanding the algorithms</a></li><li><a href="../../../assembly-method-selection/">Choosing assembly methods</a></li><li><a href="../../../tutorials/">Advanced tutorials</a></li></ul></li></ol><h3 id="Using-Your-Own-Data"><a class="docs-heading-anchor" href="#Using-Your-Own-Data">Using Your Own Data</a><a id="Using-Your-Own-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Using-Your-Own-Data" title="Permalink"></a></h3><p>To use your own sequencing data, simply replace the read simulation step:</p><pre><code class="language-julia hljs"># Instead of simulating reads:
reads_file = &quot;path/to/your/reads.fastq&quot;

# Then run assembly as before:
assembly_result = Mycelia.mycelia_assemble(reads_file)</code></pre><h3 id="Getting-Help"><a class="docs-heading-anchor" href="#Getting-Help">Getting Help</a><a id="Getting-Help-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-Help" title="Permalink"></a></h3><ul><li>Check the <a href="../../../faq/">FAQ</a> for common questions and issues</li><li>Ask questions in <a href="https://github.com/cjprybol/Mycelia/discussions">Discussions</a></li><li>Report issues on <a href="https://github.com/cjprybol/Mycelia/issues">GitHub</a></li></ul><p>Happy assembling! üß¨</p><p>%% [markdown]</p><h2 id="Appendix:-Understanding-the-Output"><a class="docs-heading-anchor" href="#Appendix:-Understanding-the-Output">Appendix: Understanding the Output</a><a id="Appendix:-Understanding-the-Output-1"></a><a class="docs-heading-anchor-permalink" href="#Appendix:-Understanding-the-Output" title="Permalink"></a></h2><p>The assembly process creates several files in the output directory:</p><ul><li><code>final_assembly.fasta</code> - Your assembled contigs</li><li><code>assembly_summary.txt</code> - Statistics about the assembly</li><li><code>kmer_progression.txt</code> - How the assembly progressed through k-mer sizes</li><li><code>graphs/</code> - Intermediate graph files (for debugging)</li></ul><p>The intelligent assembly typically tries several k-mer sizes (e.g., 21, 31, 41) and automatically stops when it finds the optimal assembly.</p></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Saturday 13 December 2025 15:26">Saturday 13 December 2025</span>. Using Julia version 1.10.10.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
