{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335bf35f-c55e-4857-83f0-ed03dc666f4e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# if hit plotting library issues, try resetting LD path for julia\n",
    "# can set in ~/.local/share/jupyter/kernels/\n",
    "haskey(ENV, \"LD_LIBRARY_PATH\") && @assert ENV[\"LD_LIBRARY_PATH\"] == \"\"\n",
    "import Pkg\n",
    "pkgs = [\n",
    "    \"Revise\",\n",
    "    \"FASTX\",\n",
    "    \"BioSequences\",\n",
    "    \"Kmers\",\n",
    "    \"Graphs\",\n",
    "    \"MetaGraphs\",\n",
    "    \"SparseArrays\",\n",
    "    \"ProgressMeter\",\n",
    "    \"Distributions\",\n",
    "    \"HiddenMarkovModels\",\n",
    "    \"BioAlignments\",\n",
    "    \"StatsBase\",\n",
    "    \"Random\",\n",
    "    \"StatsPlots\",\n",
    "    \"Statistics\",\n",
    "    # \"GraphMakie\",\n",
    "    \"IterTools\",\n",
    "    \"Primes\",\n",
    "    \"OnlineStats\",\n",
    "    \"IteratorSampling\",\n",
    "    \"HypothesisTests\",\n",
    "    \"Clustering\",\n",
    "    \"Distances\",\n",
    "    \"BioAlignments\",\n",
    "    \"Statistics\"\n",
    "]\n",
    "# Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "# Pkg.develop(path=\"/global/cfs/projectdirs/m4269/cjprybol/Mycelia\")\n",
    "# Pkg.develop(path=\"../../..\")\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "128cc1e1-929a-42e1-8376-4c81832fd40c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "PROJECT_BASEDIR = dirname(pwd())\n",
    "data_dir = joinpath(PROJECT_BASEDIR, \"data\")\n",
    "genome_dir = mkpath(joinpath(data_dir, \"genomes\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00736937-92d0-4cb6-a933-0cb068457ad1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "working_dir = joinpath(data_dir, \"test\")\n",
    "mkpath(working_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e294dc9-dc3d-4a15-8e15-70e649200874",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# short_read_sets = unique(map(x -> match(r\"^(.+\\.\\d+x)\\.\", x).captures[1], filter(x -> occursin(r\"\\.fna\\.art\", x) && occursin(r\"\\.fq\\.gz\", x) && !occursin(\"trimming_report\", x) && !occursin(\"_val_\", x), sort(readdir(genome_dir, join=true), by=x->filesize(x)))))\n",
    "# # forward = short_read_set * \".1_val_1.fq.gz\"\n",
    "# # reverse = short_read_set * \".2_val_2.fq.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4ab3b27-b540-4f4d-8bd3-093e58e0304d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "long_read_fastqs = sort(filter(x -> occursin(r\"\\.filtlong\\.fq\\.gz$\", x), readdir(genome_dir, join=true)), by=x->filesize(x))\n",
    "fastq = long_read_fastqs[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9de34fec-79aa-419f-a1a1-dd7dfe664a7b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# reference_fasta = replace(fastq, r\"\\.badread.*\" => \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61eb94ad-bbf5-4d29-aaa6-e96941303ff6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "function find_resampling_stretches(;record_kmer_solidity, solid_branching_kmer_indices)\n",
    "    indices = findall(.!record_kmer_solidity)  # Find the indices of false values\n",
    "    if isempty(indices)\n",
    "        return UnitRange{Int64}[]\n",
    "    end\n",
    "    \n",
    "    diffs = diff(indices)  # Calculate the differences between consecutive indices\n",
    "    # @show diffs\n",
    "    range_starts = [indices[1]]  # Start with the first false index\n",
    "    range_ends = Int[]\n",
    "    \n",
    "    for (i, d) in enumerate(diffs)\n",
    "        if d > 1\n",
    "            push!(range_ends, indices[i])\n",
    "            push!(range_starts, indices[i+1])\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    push!(range_ends, indices[end])  # Add the last false index as a range end\n",
    "    \n",
    "    low_quality_runs = [(start, stop) for (start, stop) in zip(range_starts, range_ends)]\n",
    "    \n",
    "    resampling_stretches = UnitRange{Int64}[]\n",
    "    \n",
    "    for low_quality_run in low_quality_runs\n",
    "        unders = filter(solid_branching_kmer -> solid_branching_kmer < first(low_quality_run), solid_branching_kmer_indices)\n",
    "        overs = filter(solid_branching_kmer -> solid_branching_kmer > last(low_quality_run), solid_branching_kmer_indices)\n",
    "        if isempty(overs) || isempty(unders)\n",
    "            continue\n",
    "        else\n",
    "            nearest_under = maximum(unders)\n",
    "            nearest_over = minimum(overs)\n",
    "            push!(resampling_stretches, nearest_under:nearest_over)\n",
    "        end\n",
    "    end\n",
    "    if !allunique(resampling_stretches)\n",
    "        resampling_stretches = unique!(resampling_stretches)\n",
    "    end\n",
    "    return resampling_stretches\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5e75667-f28c-470a-922c-34cf64227eb2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "function fastq_record(;identifier, sequence, quality_scores)\n",
    "    # Fastx wont parse anything higher than 93\n",
    "    quality_scores = min.(quality_scores, 93)\n",
    "    record_string = join([\"@\" * identifier, sequence, \"+\", join([Char(x+33) for x in quality_scores])], \"\\n\")\n",
    "    return FASTX.parse(FASTX.FASTQRecord, record_string)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f2671df-62ab-4f6a-be06-a70dfc1f4438",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "function process_fastq_record(;record, graph, likely_valid_kmers, kmer_to_index_map, branching_nodes_set, assembly_k, transition_likelihoods, kmer_mean_quality, yen_k_shortest_paths_and_weights, kmer_total_quality, ordered_kmers, yen_k=7)\n",
    "    new_record_identifier = FASTX.identifier(record) * \".k$(assembly_k)\"\n",
    "    record_sequence = BioSequences.LongDNA{4}(FASTX.sequence(record))\n",
    "\n",
    "    kmer_type = Kmers.DNAKmer{assembly_k}\n",
    "    record_kmers = last.(collect(Kmers.EveryKmer{kmer_type}(record_sequence)))\n",
    "    record_quality_scores = collect(FASTX.quality_scores(record))\n",
    "    record_kmer_quality_scores = [record_quality_scores[i:i+assembly_k-1] for i in 1:length(record_quality_scores)-assembly_k+1]\n",
    "    \n",
    "    record_kmer_solidity = map(kmer -> kmer in likely_valid_kmers, record_kmers)\n",
    "    record_branching_kmers = [kmer_to_index_map[kmer] in branching_nodes_set for kmer in record_kmers]\n",
    "    record_solid_branching_kmers = record_kmer_solidity .& record_branching_kmers\n",
    "    \n",
    "    # trim beginning of fastq\n",
    "    initial_solid_kmer = findfirst(record_kmer_solidity)\n",
    "    if isnothing(initial_solid_kmer)\n",
    "        return record\n",
    "    elseif initial_solid_kmer > 1\n",
    "        record_kmers = record_kmers[initial_solid_kmer:end]\n",
    "        record_kmer_quality_scores = record_kmer_quality_scores[initial_solid_kmer:end]\n",
    "        record_kmer_solidity = map(kmer -> kmer in likely_valid_kmers, record_kmers)\n",
    "        record_branching_kmers = [kmer_to_index_map[kmer] in branching_nodes_set for kmer in record_kmers]\n",
    "        record_solid_branching_kmers = record_kmer_solidity .& record_branching_kmers\n",
    "    end\n",
    "    initial_solid_kmer = 1\n",
    "    \n",
    "    # trim end of fastq\n",
    "    last_solid_kmer = findlast(record_kmer_solidity)\n",
    "    if last_solid_kmer != length(record_kmer_solidity)\n",
    "        record_kmers = record_kmers[1:last_solid_kmer]\n",
    "        record_kmer_quality_scores = record_kmer_quality_scores[1:last_solid_kmer]\n",
    "        record_kmer_solidity = map(kmer -> kmer in likely_valid_kmers, record_kmers)\n",
    "        record_branching_kmers = [kmer_to_index_map[kmer] in branching_nodes_set for kmer in record_kmers]\n",
    "        record_solid_branching_kmers = record_kmer_solidity .& record_branching_kmers\n",
    "    end\n",
    "    \n",
    "    # identify low quality runs and the solid branchpoints we will use for resampling\n",
    "    solid_branching_kmer_indices = findall(record_solid_branching_kmers)\n",
    "    resampling_stretches = find_resampling_stretches(;record_kmer_solidity, solid_branching_kmer_indices)\n",
    "\n",
    "    # nothing to do\n",
    "    if isempty(resampling_stretches)\n",
    "        return record\n",
    "    end\n",
    "    trusted_range = 1:max(first(first(resampling_stretches))-1, 1)\n",
    "    \n",
    "    new_record_kmers = record_kmers[trusted_range]\n",
    "    new_record_kmer_qualities = record_kmer_quality_scores[trusted_range]\n",
    "    \n",
    "    \n",
    "    for (i, resampling_stretch) in enumerate(resampling_stretches)\n",
    "        starting_solid_kmer = record_kmers[first(resampling_stretch)]\n",
    "        ending_solid_kmer = record_kmers[last(resampling_stretch)]\n",
    "        \n",
    "        current_quality_scores = record_quality_scores[resampling_stretch]\n",
    "        u = kmer_to_index_map[starting_solid_kmer]\n",
    "        v = kmer_to_index_map[ending_solid_kmer]\n",
    "        if !haskey(yen_k_shortest_paths_and_weights, u => v)\n",
    "            yen_k_result = Graphs.yen_k_shortest_paths(graph, u, v, Graphs.weights(graph), yen_k)\n",
    "            yen_k_shortest_paths_and_weights[u => v] = Vector{Pair{Vector{Int}, Float64}}()\n",
    "            for path in yen_k_result.paths\n",
    "                path_weight = Statistics.mean([kmer_total_quality[ordered_kmers[node]] for node in path])\n",
    "                path_transition_likelihoods = 1.0\n",
    "                for (a, b) in zip(path[1:end-1], path[2:end])\n",
    "                    path_transition_likelihoods *= transition_likelihoods[a, b]\n",
    "                end\n",
    "                joint_weight = path_weight * path_transition_likelihoods\n",
    "                push!(yen_k_shortest_paths_and_weights[u => v], path => joint_weight)\n",
    "            end\n",
    "        end\n",
    "        yen_k_path_weights = yen_k_shortest_paths_and_weights[u => v]      \n",
    "        if length(yen_k_path_weights) > 1\n",
    "            current_distance = length(resampling_stretch)\n",
    "            initial_weights = last.(yen_k_path_weights)\n",
    "            path_lengths = length.(first.(yen_k_path_weights))\n",
    "            deltas = map(l -> abs(l-current_distance), path_lengths)\n",
    "            adjusted_weights = initial_weights .* map(d -> exp(-d * log(2)), deltas)\n",
    "            # make it more severe?\n",
    "            # adjusted_weights = adjusted_weights.^2\n",
    "            \n",
    "            # and a bonus for usually being correct\n",
    "            \n",
    "            selected_path_index = StatsBase.sample(StatsBase.weights(adjusted_weights))\n",
    "            selected_path, selected_path_weights = yen_k_path_weights[selected_path_index]\n",
    "            selected_path_kmers = [ordered_kmers[kmer_index] for kmer_index in selected_path]\n",
    "            \n",
    "            if last(new_record_kmers) == first(selected_path_kmers)\n",
    "                selected_path_kmers = selected_path_kmers[2:end]\n",
    "            end\n",
    "            append!(new_record_kmers, selected_path_kmers)\n",
    "            selected_kmer_qualities = [Int8.(min.(typemax(Int8), floor.(kmer_mean_quality[kmer]))) for kmer in selected_path_kmers]\n",
    "            append!(new_record_kmer_qualities, selected_kmer_qualities)\n",
    "        else\n",
    "            selected_path_kmers = record_kmers[resampling_stretch]\n",
    "            if last(new_record_kmers) == first(selected_path_kmers)\n",
    "                selected_path_kmers = selected_path_kmers[2:end]\n",
    "            end\n",
    "            append!(new_record_kmers, selected_path_kmers)\n",
    "            selected_kmer_qualities = [Int8.(min.(typemax(Int8), floor.(kmer_mean_quality[kmer]))) for kmer in selected_path_kmers]\n",
    "            append!(new_record_kmer_qualities, selected_kmer_qualities)\n",
    "        end\n",
    "        if i < length(resampling_stretches) # append high quality gap\n",
    "            next_solid_start = last(resampling_stretch)+1\n",
    "            next_resampling_stretch = resampling_stretches[i+1]\n",
    "            next_solid_stop = first(next_resampling_stretch)-1\n",
    "            if !isempty(next_solid_start:next_solid_stop)\n",
    "                selected_path_kmers = record_kmers[next_solid_start:next_solid_stop]\n",
    "                append!(new_record_kmers, selected_path_kmers)\n",
    "                selected_kmer_qualities = record_kmer_quality_scores[next_solid_start:next_solid_stop]\n",
    "                append!(new_record_kmer_qualities, selected_kmer_qualities)\n",
    "            end\n",
    "        else # append remainder of sequence\n",
    "            @assert i == length(resampling_stretches)\n",
    "            next_solid_start = last(resampling_stretch)+1\n",
    "            if next_solid_start < length(record_kmers)\n",
    "                selected_path_kmers = record_kmers[next_solid_start:end]\n",
    "                append!(new_record_kmers, selected_path_kmers)\n",
    "                selected_kmer_qualities = record_kmer_quality_scores[next_solid_start:end]\n",
    "                append!(new_record_kmer_qualities, selected_kmer_qualities)\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "    for (a, b) in zip(new_record_kmers[1:end-1], new_record_kmers[2:end])\n",
    "        @assert a != b\n",
    "    end\n",
    "    new_record_sequence = Mycelia.kmer_path_to_sequence(new_record_kmers)\n",
    "    new_record_quality_scores = new_record_kmer_qualities[1]\n",
    "    for new_record_kmer_quality in new_record_kmer_qualities[2:end]\n",
    "        push!(new_record_quality_scores, last(new_record_kmer_quality))\n",
    "    end\n",
    "    new_record = fastq_record(identifier=new_record_identifier, sequence=new_record_sequence, quality_scores=new_record_quality_scores)\n",
    "    return new_record\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee32804e-a25e-4e8c-a9c2-d2ca407a447d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "function polish_fastq(;fastq, k=1)\n",
    "    if k == 1\n",
    "        assembly_k = Mycelia.assess_dnamer_saturation([fastq])\n",
    "    else\n",
    "        assembly_k = k\n",
    "    end\n",
    "    fastq_out = replace(fastq, Mycelia.FASTQ_REGEX => \".k$(assembly_k).fq\")\n",
    "    kmer_type = Kmers.DNAKmer{assembly_k}\n",
    "    records = collect(Mycelia.open_fastx(fastq))\n",
    "\n",
    "    read_quality_scores = [collect(FASTX.quality_scores(record)) for record in records]\n",
    "\n",
    "    # make a dictionary associating all kmers with their quality scores\n",
    "    all_kmer_quality_support = Dict{kmer_type, Vector{Float64}}()\n",
    "    for record in records\n",
    "        record_quality_scores = collect(FASTX.quality_scores(record))\n",
    "        record_quality_score_slices = [record_quality_scores[i:i+assembly_k-1] for i in 1:length(record_quality_scores)-assembly_k+1]\n",
    "        sequence = BioSequences.LongDNA{2}(FASTX.sequence(record))\n",
    "        for ((i, kmer), kmer_base_qualities) in zip(Kmers.EveryKmer{kmer_type}(sequence), record_quality_score_slices)\n",
    "            if haskey(all_kmer_quality_support, kmer)\n",
    "                all_kmer_quality_support[kmer] = all_kmer_quality_support[kmer] .+ kmer_base_qualities\n",
    "            else\n",
    "                all_kmer_quality_support[kmer] = kmer_base_qualities\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    kmer_counts = Mycelia.count_kmers(kmer_type, fastq)\n",
    "    kmer_indices = Dict(kmer => i for (i, kmer) in enumerate(keys(kmer_counts)))\n",
    "    canonical_kmer_counts = Mycelia.count_canonical_kmers(kmer_type, fastq)\n",
    "    canonical_kmer_indices = Dict(kmer => i for (i, kmer) in enumerate(keys(canonical_kmer_counts)))\n",
    "\n",
    "    strand_normalized_quality_support = Dict{kmer_type, Vector{Float64}}()\n",
    "    for (kmer, support) in all_kmer_quality_support\n",
    "        strand_normalized_quality_support[kmer] = support\n",
    "        if haskey(all_kmer_quality_support, BioSequences.reverse_complement(kmer))\n",
    "            strand_normalized_quality_support[kmer] .+= all_kmer_quality_support[BioSequences.reverse_complement(kmer)]\n",
    "        end\n",
    "    end\n",
    "    strand_normalized_quality_support\n",
    "    \n",
    "    ordered_kmers = collect(keys(kmer_counts))\n",
    "    \n",
    "    kmer_mean_quality = Dict(kmer => strand_normalized_quality_support[kmer] ./ canonical_kmer_counts[BioSequences.canonical(kmer)] for kmer in ordered_kmers)\n",
    "    \n",
    "    kmer_total_quality = Dict(kmer => sum(quality_values) for (kmer, quality_values) in strand_normalized_quality_support)\n",
    "    state_likelihoods = Dict(kmer => total_quality / sum(values(kmer_total_quality)) for (kmer, total_quality) in kmer_total_quality)\n",
    "\n",
    "    total_states = length(state_likelihoods)\n",
    "\n",
    "    transition_likelihoods = SparseArrays.spzeros(total_states, total_states)\n",
    "    for record in records\n",
    "        sequence = BioSequences.LongDNA{4}(FASTX.sequence(record))\n",
    "        sources = Kmers.EveryKmer{kmer_type}(sequence[1:end-1])\n",
    "        destinations = Kmers.EveryKmer{kmer_type}(sequence[2:end])\n",
    "        for ((source_i, source), (destination_i, destination)) in zip(sources, destinations)\n",
    "            source_index = kmer_indices[source]\n",
    "            destination_index = kmer_indices[destination]\n",
    "            transition_likelihoods[source_index, destination_index] += 1\n",
    "        end\n",
    "    end\n",
    "    for source in 1:total_states\n",
    "        outgoing_transition_counts = transition_likelihoods[source, :]\n",
    "        if sum(outgoing_transition_counts) > 0\n",
    "            transition_likelihoods[source, :] .= transition_likelihoods[source, :] ./ sum(transition_likelihoods[source, :]) \n",
    "        end\n",
    "    end\n",
    "    transition_likelihoods\n",
    "\n",
    "    graph = Graphs.SimpleDiGraph(total_states)\n",
    "    row_indices, column_indices, cell_values = SparseArrays.findnz(transition_likelihoods)\n",
    "    for (row, col) in zip(row_indices, column_indices)\n",
    "        Graphs.add_edge!(graph, row, col)\n",
    "    end\n",
    "\n",
    "    unbranching_nodes = Set(Int[])\n",
    "    for node in Graphs.vertices(graph)\n",
    "        if (Graphs.indegree(graph, node) <= 1) && (Graphs.outdegree(graph, node) <= 1)\n",
    "            push!(unbranching_nodes, node)\n",
    "        end\n",
    "    end\n",
    "    branching_nodes = setdiff(Graphs.vertices(graph), unbranching_nodes)\n",
    "    branching_nodes_set = Set(branching_nodes)\n",
    "    \n",
    "    total_strand_normalized_quality_support = sum.(collect(values(strand_normalized_quality_support)))\n",
    "    mean_total_support = Statistics.mean(total_strand_normalized_quality_support)\n",
    "    # Statistics.std(total_strand_normalized_quality_support)\n",
    "    \n",
    "    sorted_kmer_total_quality = sort(kmer_total_quality)\n",
    "    sorted_kmer_total_quality_values = collect(values(sorted_kmer_total_quality))\n",
    "    \n",
    "    mean_quality_value = Statistics.mean(sorted_kmer_total_quality_values)\n",
    "    threshold = mean_quality_value\n",
    "\n",
    "    xs = [\n",
    "        [i for (i, y) in enumerate(sorted_kmer_total_quality_values) if y > threshold],\n",
    "        [i for (i, y) in enumerate(sorted_kmer_total_quality_values) if y <= threshold]\n",
    "        ]\n",
    "    ys = [\n",
    "        [y for y in sorted_kmer_total_quality_values if y > threshold],\n",
    "        [y for y in sorted_kmer_total_quality_values if y <= threshold]\n",
    "    ]\n",
    "    \n",
    "    p = StatsPlots.scatter(\n",
    "        xs,\n",
    "        ys,\n",
    "        title = \"kmer qualities\",\n",
    "        ylabel = \"canonical kmer cumulative QUAL value\",\n",
    "        label = [\"above\" \"below\"],\n",
    "        legend = :outertopright,\n",
    "        size = (900, 500),\n",
    "        margins=10StatsPlots.Plots.PlotMeasures.mm,\n",
    "        xticks = false\n",
    "    )\n",
    "    p = StatsPlots.hline!(p, [mean_quality_value], label=\"mean\")\n",
    "    display(p)\n",
    "    \n",
    "    likely_valid_kmer_indices = xs[1]\n",
    "    likely_sequencing_artifact_indices = xs[2]\n",
    "    likely_sequencing_artifact_kmers = Set(ordered_kmers[likely_sequencing_artifact_indices])\n",
    "    likely_valid_kmers = Set(ordered_kmers[likely_valid_kmer_indices])\n",
    "    kmer_to_index_map = Dict(kmer => i for (i, kmer) in enumerate(ordered_kmers))\n",
    "    \n",
    "    revised_records = []\n",
    "    ProgressMeter.@showprogress for record in records\n",
    "        yen_k_shortest_paths_and_weights = Dict{Pair{Int, Int}, Vector{Pair{Vector{Int}, Float64}}}()\n",
    "        revised_record = process_fastq_record(;record, graph, likely_valid_kmers, kmer_to_index_map, branching_nodes_set, assembly_k, kmer_mean_quality, transition_likelihoods, yen_k_shortest_paths_and_weights, kmer_total_quality, ordered_kmers)\n",
    "        push!(revised_records, revised_record)\n",
    "    end\n",
    "    \n",
    "    open(fastq_out, \"w\") do io\n",
    "        fastx_io = FASTX.FASTQ.Writer(io)\n",
    "        for record in revised_records\n",
    "            write(fastx_io, record)\n",
    "        end\n",
    "        close(fastx_io)\n",
    "    end\n",
    "    run(`gzip --force $(fastq_out)`)\n",
    "    return (fastq = fastq_out * \".gz\", k=assembly_k)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e305a34d-7490-432f-9aa9-029ebd10dad3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Mycelia.ks()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1ee2189-7f79-4861-a2f2-476ae507f430",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# max_k = maximum(Mycelia.ks())\n",
    "# this is polishing up to the ~ strain level separation boundary\n",
    "\n",
    "# max_k = 53\n",
    "# w/o increment\n",
    "# {870, 869}\n",
    "# {1325, 1329}\n",
    "# {888, 887}\n",
    "# {1371, 1375}\n",
    "# {1264, 1267}\n",
    "\n",
    "# w/ increment\n",
    "# {834, 833}\n",
    "# {1392, 1393}\n",
    "# {848, 847}\n",
    "# {889, 888}\n",
    "# {1298, 1298}\n",
    "\n",
    "\n",
    "# {1109, 1109}\n",
    "# {769, 768}\n",
    "# {802, 801}\n",
    "# {819, 818}\n",
    "# {715, 714}\n",
    "max_k = 89\n",
    "\n",
    "# don't increment assembly beyond 89\n",
    "\n",
    "\n",
    "polishing_results = [polish_fastq(;fastq, k=11)]\n",
    "# @show last(polishing_results)\n",
    "while (!ismissing(last(polishing_results).k)) && (last(polishing_results).k < max_k)\n",
    "    next_k = first(filter(k -> k > last(polishing_results).k, Mycelia.ks()))\n",
    "    @show next_k\n",
    "    push!(polishing_results, polish_fastq(fastq=last(polishing_results).fastq, k=next_k))\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0a0ac1b-2586-4fdb-8fa5-8d8667e1798a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# finish updating all reads for all k rounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41a55c25-97c4-4a39-9f57-7ff5629c73d7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "polishing_k = last(polishing_results).k\n",
    "assembly_fastq = last(polishing_results).fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fae5748-8bdc-4308-81e7-9b897fd7ec35",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "assembly_k = polishing_k\n",
    "kmer_graph = Mycelia.fastx_to_kmer_graph(Kmers.DNAKmer{assembly_k}, assembly_fastq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8feb470c-2951-4fab-9beb-461a7ff00104",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "kmer_graph.gprops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d14d7c1-9b37-48fb-922d-66847db19291",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "initial_connected_components = length(Graphs.connected_components(kmer_graph))\n",
    "coverage_threshold = 1\n",
    "candidate_subgraph, vertex_map = Graphs.induced_subgraph(kmer_graph, [v for v in Graphs.vertices(kmer_graph) if kmer_graph.vprops[v][:count] > coverage_threshold])\n",
    "while (length(Graphs.connected_components(candidate_subgraph)) == initial_connected_components) && (coverage_threshold < 5)\n",
    "    coverage_threshold += 1\n",
    "    candidate_subgraph, vertex_map = Graphs.induced_subgraph(kmer_graph, [v for v in Graphs.vertices(kmer_graph) if kmer_graph.vprops[v][:count] > coverage_threshold])\n",
    "end\n",
    "@show coverage_threshold\n",
    "candidate_subgraph, vertex_map = Graphs.induced_subgraph(kmer_graph, [v for v in Graphs.vertices(kmer_graph) if kmer_graph.vprops[v][:count] >= coverage_threshold])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5fc4953-977e-40ce-b708-5a2127d627c1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Graphs.connected_components(candidate_subgraph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44b33ac5-5611-4e4b-a97b-0cfcaac49fa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out final assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b530abc-7146-4439-872f-12f4aca53ed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# call variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30f38987-1550-4fee-bd1d-135dedc4ef6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assess accuracy"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.2",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
