{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335bf35f-c55e-4857-83f0-ed03dc666f4e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# if hit plotting library issues, try resetting LD path for julia\n",
    "# can set in ~/.local/share/jupyter/kernels/\n",
    "haskey(ENV, \"LD_LIBRARY_PATH\") && @assert ENV[\"LD_LIBRARY_PATH\"] == \"\"\n",
    "import Pkg\n",
    "pkgs = [\n",
    "    \"Revise\",\n",
    "    \"FASTX\",\n",
    "    \"ProgressMeter\",\n",
    "    \"DataFrames\",\n",
    "    \"StatsBase\",\n",
    "    \"StatsPlots\",\n",
    "    \"OrderedCollections\",\n",
    "    \"ColorSchemes\",\n",
    "    \"uCSV\"\n",
    "]\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "# Pkg.develop(path=\"/global/cfs/projectdirs/m4269/cjprybol/Mycelia\")\n",
    "# Pkg.develop(path=\"../../..\")\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caab51db-26ef-4314-b4ac-a37928aaa991",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "PROJECT_BASEDIR = dirname(pwd())\n",
    "data_dir = joinpath(PROJECT_BASEDIR, \"data\")\n",
    "genome_dir = mkpath(joinpath(data_dir, \"genomes\"))\n",
    "results_dir = joinpath(PROJECT_BASEDIR, \"results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c7ad3ae-d0dc-4fd6-a119-fe784fd9a403",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "original_fastas = sort(filter(x -> occursin(r\"\\.fna$\", x) && !occursin(\"normalized\", x), readdir(genome_dir, join=true)), by=x->filesize(x))\n",
    "\n",
    "reference_variant_fastas = original_fastas .* \".normalized.vcf.fna\"\n",
    "\n",
    "@assert all(isfile.(reference_variant_fastas))\n",
    "\n",
    "fastqs = filter(x -> occursin(r\"\\.fq\\.gz$\", x), readdir(genome_dir, join=true))\n",
    "\n",
    "forward_fastqs = filter(x -> occursin(\"1_val_1.fq.gz\", x), fastqs)\n",
    "\n",
    "reverse_fastqs = filter(x -> occursin(\"2_val_2.fq.gz\", x), fastqs)\n",
    "\n",
    "long_read_fastqs = filter(x -> occursin(\"filtlong.fq.gz\", x), fastqs)\n",
    "\n",
    "assembly_table = DataFrames.DataFrame(\n",
    "    original_fasta = String[],\n",
    "    reference_variant_fasta = String[],\n",
    "    coverage = String[],\n",
    "    fastqs = Vector{String}[],\n",
    "    assembler = String[],\n",
    "    assembled_variant_fasta = String[]\n",
    ")\n",
    "\n",
    "assemblers = [\n",
    "    \"megahit\",\n",
    "    \"spades_isolate\",\n",
    "    \"flye\",\n",
    "    \"raven\",\n",
    "    \"hifiasm\",\n",
    "    \"mycelia-lr\",\n",
    "    \"mycelia-sr\"\n",
    "]\n",
    "coverages = [\"10x\", \"100x\", \"1000x\"]\n",
    "\n",
    "for (original_fasta, reference_variant_fasta) in zip(original_fastas, reference_variant_fastas)\n",
    "    reference_variant_fasta_matches = filter(x -> occursin(reference_variant_fasta, x), readdir(genome_dir, join=true))\n",
    "    for coverage in coverages\n",
    "        coverage_matches = filter(x -> occursin(coverage, x), reference_variant_fasta_matches)\n",
    "        for assembler in assemblers\n",
    "            # @show coverage_matches\n",
    "            if assembler == \"mycelia-lr\"\n",
    "                assembly_matches = filter(x -> occursin(r\"\\.fna$\", x) && occursin(\"badread\", x) && !occursin(\"joint\", x), coverage_matches)\n",
    "                # @show assembly_matches\n",
    "                @assert 0 <= length(assembly_matches) <= 1\n",
    "                if isempty(assembly_matches)\n",
    "                    continue\n",
    "                else\n",
    "                    assembled_variant_fasta = first(assembly_matches)\n",
    "                    matching_fastqs = filter(x -> occursin(coverage, x) && occursin(reference_variant_fasta, x), long_read_fastqs)\n",
    "                    # @show assembled_variant_fasta\n",
    "                    # @show matching_fastqs\n",
    "                end\n",
    "            elseif assembler == \"mycelia-sr\"\n",
    "                assembly_matches = filter(x -> occursin(r\"\\.fna$\", x) && occursin(\"art\", x) && !occursin(r\"\\.joint\\.fna$\", x), coverage_matches)\n",
    "                if isempty(assembly_matches)\n",
    "                    continue\n",
    "                else\n",
    "                    assembled_variant_fasta = first(assembly_matches)\n",
    "                    matching_fastqs = filter(x -> occursin(coverage, x) && occursin(reference_variant_fasta, x), vcat(forward_fastqs, reverse_fastqs))\n",
    "                    # @show assembled_variant_fasta\n",
    "                    # @show matching_fastqs\n",
    "                end\n",
    "            else\n",
    "                directories = filter(x -> isdir(x) && occursin(assembler, x), coverage_matches)\n",
    "                if !isempty(directories)\n",
    "                    # directory = missing\n",
    "                    @assert length(directories) == 1 \"$(reference_variant_fasta) $(assembler) $(coverage)\"\n",
    "                    directory = first(directories)\n",
    "                else\n",
    "                    continue\n",
    "                end\n",
    "                if assembler == \"megahit\"\n",
    "                    assembled_variant_fasta = joinpath(directory, \"final.contigs.fa\")\n",
    "                    matching_fastqs = filter(x -> occursin(coverage, x) && occursin(reference_variant_fasta, x), vcat(forward_fastqs, reverse_fastqs))\n",
    "                elseif assembler == \"spades_isolate\"\n",
    "                    assembled_variant_fasta = joinpath(directory, \"scaffolds.fasta\")\n",
    "                    matching_fastqs = filter(x -> occursin(coverage, x) && occursin(reference_variant_fasta, x), vcat(forward_fastqs, reverse_fastqs))\n",
    "                elseif assembler == \"flye\"\n",
    "                    assembled_variant_fasta = joinpath(directory, \"assembly.fasta\")\n",
    "                    matching_fastqs = filter(x -> occursin(coverage, x) && occursin(reference_variant_fasta, x), long_read_fastqs)\n",
    "                elseif assembler == \"raven\"\n",
    "                    assembled_variant_fasta = joinpath(directory, \"assembly.fasta\")\n",
    "                    matching_fastqs = filter(x -> occursin(coverage, x) && occursin(reference_variant_fasta, x), long_read_fastqs)\n",
    "                elseif assembler == \"hifiasm\"\n",
    "                    directory_fastas = filter(x -> occursin(r\"\\.hifiasm\\.p_ctg\\.gfa\\.fna$\", x), readdir(directory))\n",
    "                    # @show directory_fastas\n",
    "                    @assert length(directory_fastas) <= 1\n",
    "                    if length(directory_fastas) == 0\n",
    "                        continue\n",
    "                    else\n",
    "                        assembled_variant_fasta = joinpath(directory, \"assembly.fasta\")\n",
    "                        if !isfile(assembled_variant_fasta)\n",
    "                            cp(joinpath(directory, first(directory_fastas)), assembled_variant_fasta)\n",
    "                        end\n",
    "                        matching_fastqs = filter(x -> occursin(coverage, x) && occursin(original_fasta, x), long_read_fastqs)\n",
    "                    end\n",
    "                end\n",
    "            end\n",
    "            if !isnothing(assembled_variant_fasta)\n",
    "                push!(assembly_table, (original_fasta=original_fasta, reference_variant_fasta=reference_variant_fasta, assembler=assembler, coverage=coverage, fastqs=matching_fastqs, assembled_variant_fasta=assembled_variant_fasta))\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "assembly_table\n",
    "\n",
    "assembly_table = assembly_table[map(x -> isfile(x) && filesize(x) > 0, assembly_table[!, \"assembled_variant_fasta\"]), :]\n",
    "\n",
    "# assembly_table_to_visualize[!, \"reference_variant_calls\"] = assembly_table_to_visualize[!, \"reference_variant_calls\"] .* \".gz\"\n",
    "assembly_table[!, \"reference_variant_calls\"] .= replace.(assembly_table[!, \"reference_variant_fasta\"], r\"\\.fna$\" => \".gz\")\n",
    "\n",
    "mycelia_subset = assembly_table[map(x -> occursin(\"mycelia\", x), assembly_table[!, \"assembler\"]), [\"coverage\", \"fastqs\", \"original_fasta\"]]\n",
    "\n",
    "assembly_table = DataFrames.innerjoin(assembly_table, mycelia_subset, on=[\"coverage\", \"fastqs\", \"original_fasta\"])\n",
    "\n",
    "# assembly_table[!, \"reference_assembly\"] = replace.(assembly_table[!, \"reference_assembly\"], \"$(genome_dir)/\" => \"\")\n",
    "# assembly_table[!, \"fasta\"] = replace.(assembly_table[!, \"fasta\"], \"$(genome_dir)/\" => \"\")\n",
    "genome_coverages_to_visualize = [x[1] for x in filter(x -> x[2] > 1, StatsBase.countmap(collect(eachrow(Matrix(unique(assembly_table[!, [\"original_fasta\", \"coverage\", \"fastqs\"]])[!, [\"original_fasta\", \"coverage\"]])))))]\n",
    "assembly_table_to_visualize = DataFrames.innerjoin(\n",
    "    assembly_table,\n",
    "    DataFrames.DataFrame(original_fasta = [x[1] for x in genome_coverages_to_visualize], coverage = [x[2] for x in genome_coverages_to_visualize]),\n",
    "    on=[\"original_fasta\", \"coverage\"]\n",
    ")\n",
    "\n",
    "assembly_qv_results = DataFrames.DataFrame()\n",
    "for row in DataFrames.eachrow(assembly_table_to_visualize)\n",
    "    assembly_result_table = Mycelia.assess_assembly_quality(assembly=row[\"assembled_variant_fasta\"], observations=row[\"fastqs\"])\n",
    "    assembly_result_table[!, \"assembled_variant_fasta\"] .= row[\"assembled_variant_fasta\"]\n",
    "    append!(assembly_qv_results, assembly_result_table)\n",
    "end\n",
    "assembly_qv_results_to_visualize = DataFrames.innerjoin(assembly_table_to_visualize, assembly_qv_results, on=\"assembled_variant_fasta\")\n",
    "# max_qv_value = (ceil(maximum(filter(x -> x != Inf, assembly_qv_results_to_visualize[!, \"qv\"])) / 10) + 2) * 10\n",
    "max_qv_value = 80\n",
    "assembly_qv_results_to_visualize[!, \"qv\"] = map(x -> x == Inf ? max_qv_value : x, assembly_qv_results_to_visualize[!, \"qv\"])\n",
    "assembly_qv_results_to_visualize\n",
    "\n",
    "ks = sort(unique(assembly_qv_results_to_visualize[!, \"k\"]))\n",
    "assemblers = sort(append!(unique(assembly_qv_results_to_visualize[!, \"assembler\"]), [\"hifiasm\", \"raven\"]))\n",
    "reference_variant_fasta_coverage_groups = DataFrames.groupby(assembly_qv_results_to_visualize, [\"reference_variant_fasta\", \"coverage\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83879b8a-030a-4554-ae41-3197873d6f03",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "reference_variant_fasta_coverage_group = reference_variant_fasta_coverage_groups[1]\n",
    "this_fasta = basename(reference_variant_fasta_coverage_group[1, \"original_fasta\"])\n",
    "this_coverage = reference_variant_fasta_coverage_group[1, \"coverage\"]\n",
    "# group_assembly_qv_scores = OrderedCollections.OrderedDict(assembler => Vector{Union{Float64, Missing}}(missing, length(ks)) for assembler in assemblers)\n",
    "group_assembly_qv_scores = OrderedCollections.OrderedDict(assembler => zeros(length(ks)) for assembler in assemblers)\n",
    "for assembler_group in DataFrames.groupby(reference_variant_fasta_coverage_group, \"assembler\")\n",
    "    assembler = assembler_group[1, \"assembler\"]\n",
    "    sag = sort(assembler_group)\n",
    "    scores = Float64[]\n",
    "    for (i, (row, k)) in enumerate(zip(DataFrames.eachrow(sag), ks))\n",
    "        @assert row[\"k\"] == k\n",
    "        push!(scores, row[\"qv\"])\n",
    "    end\n",
    "    group_assembly_qv_scores[assembler] = scores\n",
    "end\n",
    "p = StatsPlots.heatmap(\n",
    "    reduce(hcat, [group_assembly_qv_scores[assembler] for assembler in assemblers]),\n",
    "    xlabel = \"Assembler\",\n",
    "    ylabel = \"k-length\",\n",
    "    xticks = (1:length(assemblers), assemblers),\n",
    "    yticks = (1:length(ks), ks),\n",
    "    title = \"Merqury QV values\\n$(this_fasta) @ $(this_coverage)\",\n",
    "    colorbar_title = \"Quality Score\",\n",
    "    size = (600, 400),\n",
    "    clims=(0,80),\n",
    "    topmargin=5StatsPlots.Plots.PlotMeasures.mm\n",
    ")\n",
    "display(p)\n",
    "StatsPlots.savefig(p, joinpath(results_dir, \"merqury-qv.$(this_fasta).$(this_coverage).svg\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92a68f07-527a-42f1-84b5-101a14567f0b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "reference_variant_fasta_coverage_group = reference_variant_fasta_coverage_groups[2]\n",
    "this_fasta = basename(reference_variant_fasta_coverage_group[1, \"original_fasta\"])\n",
    "this_coverage = reference_variant_fasta_coverage_group[1, \"coverage\"]\n",
    "# group_assembly_qv_scores = OrderedCollections.OrderedDict(assembler => Vector{Union{Float64, Missing}}(missing, length(ks)) for assembler in assemblers)\n",
    "group_assembly_qv_scores = OrderedCollections.OrderedDict(assembler => zeros(length(ks)) for assembler in assemblers)\n",
    "for assembler_group in DataFrames.groupby(reference_variant_fasta_coverage_group, \"assembler\")\n",
    "    assembler = assembler_group[1, \"assembler\"]\n",
    "    sag = sort(assembler_group)\n",
    "    scores = Float64[]\n",
    "    for (i, (row, k)) in enumerate(zip(DataFrames.eachrow(sag), ks))\n",
    "        @assert row[\"k\"] == k\n",
    "        push!(scores, row[\"qv\"])\n",
    "    end\n",
    "    group_assembly_qv_scores[assembler] = scores\n",
    "end\n",
    "p = StatsPlots.heatmap(\n",
    "    reduce(hcat, [group_assembly_qv_scores[assembler] for assembler in assemblers]),\n",
    "    xlabel = \"Assembler\",\n",
    "    ylabel = \"k-length\",\n",
    "    xticks = (1:length(assemblers), assemblers),\n",
    "    yticks = (1:length(ks), ks),\n",
    "    title = \"Merqury QV values\\n$(this_fasta) @ $(this_coverage)\",\n",
    "    colorbar_title = \"Quality Score\",\n",
    "    size = (600, 400),\n",
    "    clims=(0,80),\n",
    "    topmargin=5StatsPlots.Plots.PlotMeasures.mm\n",
    ")\n",
    "display(p)\n",
    "StatsPlots.savefig(p, joinpath(results_dir, \"merqury-qv.$(this_fasta).$(this_coverage).svg\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e59fe6b-54d1-499e-894e-d39c2aa8df9e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "reference_variant_fasta_coverage_group = reference_variant_fasta_coverage_groups[4]\n",
    "this_fasta = basename(reference_variant_fasta_coverage_group[1, \"original_fasta\"])\n",
    "this_coverage = reference_variant_fasta_coverage_group[1, \"coverage\"]\n",
    "# group_assembly_qv_scores = OrderedCollections.OrderedDict(assembler => Vector{Union{Float64, Missing}}(missing, length(ks)) for assembler in assemblers)\n",
    "group_assembly_qv_scores = OrderedCollections.OrderedDict(assembler => zeros(length(ks)) for assembler in assemblers)\n",
    "for assembler_group in DataFrames.groupby(reference_variant_fasta_coverage_group, \"assembler\")\n",
    "    assembler = assembler_group[1, \"assembler\"]\n",
    "    sag = sort(assembler_group)\n",
    "    scores = Float64[]\n",
    "    for (i, (row, k)) in enumerate(zip(DataFrames.eachrow(sag), ks))\n",
    "        @assert row[\"k\"] == k\n",
    "        push!(scores, row[\"qv\"])\n",
    "    end\n",
    "    group_assembly_qv_scores[assembler] = scores\n",
    "end\n",
    "p = StatsPlots.heatmap(\n",
    "    reduce(hcat, [group_assembly_qv_scores[assembler] for assembler in assemblers]),\n",
    "    xlabel = \"Assembler\",\n",
    "    ylabel = \"k-length\",\n",
    "    xticks = (1:length(assemblers), assemblers),\n",
    "    yticks = (1:length(ks), ks),\n",
    "    title = \"Merqury QV values\\n$(this_fasta) @ $(this_coverage)\",\n",
    "    colorbar_title = \"Quality Score\",\n",
    "    size = (600, 400),\n",
    "    clims=(0,80),\n",
    "    topmargin=5StatsPlots.Plots.PlotMeasures.mm\n",
    ")\n",
    "display(p)\n",
    "StatsPlots.savefig(p, joinpath(results_dir, \"merqury-qv.$(this_fasta).$(this_coverage).svg\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c2c3728-732b-46f2-aae0-d55a770bb289",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pggb_vcfs = String[]\n",
    "for row in DataFrames.eachrow(assembly_table_to_visualize)\n",
    "    genomes = [row[\"original_fasta\"], row[\"assembled_variant_fasta\"]]\n",
    "    joint_fasta = row[\"assembled_variant_fasta\"] * \".\" * \"joint.fna\"\n",
    "    @show joint_fasta\n",
    "    outdir = joint_fasta * \"__PGGB\"\n",
    "    gfa_file = first(filter(x -> occursin(r\"\\.gfa\", x), readdir(outdir, join=true)))\n",
    "    final_vcf = gfa_file * \".vcf\"\n",
    "    push!(pggb_vcfs, final_vcf)\n",
    "end\n",
    "pggb_vcfs\n",
    "\n",
    "cactus_vcfs = String[]\n",
    "ProgressMeter.@showprogress for row in DataFrames.eachrow(assembly_table_to_visualize)\n",
    "    outdir = joinpath(genome_dir, row[\"assembled_variant_fasta\"] * \"-cactus\")\n",
    "    out = replace(outdir, \"$(genome_dir)/\" => \"\")\n",
    "    vcf = joinpath(outdir, out * \".vcf.gz\")\n",
    "    push!(cactus_vcfs, vcf)\n",
    "end\n",
    "cactus_vcfs\n",
    "\n",
    "vcf_table = DataFrames.DataFrame(\n",
    "    vcf = vcat(pggb_vcfs, cactus_vcfs),\n",
    "    caller = vcat(fill(\"pggb\", length(pggb_vcfs)), fill(\"cactus\", length(cactus_vcfs))))\n",
    "vcf_table[!, \"original_fasta\"] = [first(filter(x -> occursin(x, row[\"vcf\"]), original_fastas)) for row in DataFrames.eachrow(vcf_table)]\n",
    "vcf_table = vcf_table[isfile.(vcf_table[!, \"vcf\"]), :]\n",
    "normalized_vcfs = String[]\n",
    "for row in DataFrames.eachrow(vcf_table)\n",
    "    push!(normalized_vcfs, Mycelia.normalize_vcf(reference_fasta=row[\"original_fasta\"], vcf_file=row[\"vcf\"]))\n",
    "end\n",
    "vcf_table[!, \"normalized_vcf\"] = normalized_vcfs\n",
    "vcf_table = DataFrames.innerjoin(vcf_table, unique(assembly_table_to_visualize[!, [\"original_fasta\", \"reference_variant_calls\"]]), on=\"original_fasta\")\n",
    "\n",
    "pangenomic_variant_call_accuracy_table = DataFrames.DataFrame()\n",
    "for row in DataFrames.eachrow(vcf_table)\n",
    "    normalized_vcf = row[\"normalized_vcf\"]\n",
    "    reference_vcf = row[\"reference_variant_calls\"]\n",
    "    reference_fasta = row[\"original_fasta\"]\n",
    "    outdir = normalized_vcf * \"_RTG_eval\"\n",
    "    if isdir(outdir) && (\"done\" in readdir(outdir))\n",
    "        row = (\n",
    "            weighted_roc=joinpath(outdir, \"weighted_roc.tsv.gz\"),\n",
    "            snp_roc = joinpath(outdir, \"snp_roc.tsv.gz\"),\n",
    "            non_snp_roc = joinpath(outdir, \"non_snp_roc.tsv.gz\"),\n",
    "            normalized_vcf=normalized_vcf,\n",
    "            outdir = outdir\n",
    "        )\n",
    "        push!(pangenomic_variant_call_accuracy_table, row)\n",
    "    end\n",
    "end\n",
    "pangenomic_variant_call_accuracy_table\n",
    "pangenomic_variant_call_accuracy_table = DataFrames.innerjoin(vcf_table, pangenomic_variant_call_accuracy_table, on=\"normalized_vcf\")\n",
    "\n",
    "# need to get assembly out of this - caller is already broken out! pangenomic_variant_call_accuracy_table\n",
    "\n",
    "assembled_fastas_to_visualize = assembly_table_to_visualize[!, \"assembled_variant_fasta\"]\n",
    "\n",
    "pangenomic_variant_call_accuracy_table[!, \"assembled_variant_fasta\"] .= \"\"\n",
    "for (i, vcf) in enumerate(pangenomic_variant_call_accuracy_table[!, \"vcf\"])\n",
    "    matching_fasta_assemblies = filter(x -> occursin(x, vcf), assembled_fastas_to_visualize)\n",
    "    # display(matching_fasta_assemblies)\n",
    "    @assert length(matching_fasta_assemblies) == 1\n",
    "    pangenomic_variant_call_accuracy_table[i, \"assembled_variant_fasta\"] = first(matching_fasta_assemblies)\n",
    "end\n",
    "pangenomic_variant_call_accuracy_table[!, \"assembled_variant_fasta\"]\n",
    "\n",
    "pangenomic_variant_call_accuracy_table = DataFrames.innerjoin(pangenomic_variant_call_accuracy_table, assembly_table_to_visualize, on=[\"assembled_variant_fasta\", \"original_fasta\", \"reference_variant_calls\"], makeunique=true)\n",
    "\n",
    "weighted_roc_table = DataFrames.DataFrame()\n",
    "for row in DataFrames.eachrow(pangenomic_variant_call_accuracy_table)\n",
    "    result_table = Mycelia.parse_rtg_eval_output(row[\"weighted_roc\"])\n",
    "    if !isempty(result_table)\n",
    "        result_table[!, \"weighted_roc\"] .= row[\"weighted_roc\"]\n",
    "        append!(weighted_roc_table, result_table)\n",
    "        # row = (weighted_roc = row[\"weighted_roc\"], )\n",
    "    end\n",
    "end\n",
    "weighted_roc_table = DataFrames.innerjoin(pangenomic_variant_call_accuracy_table, weighted_roc_table, on=\"weighted_roc\")\n",
    "weighted_roc_table[!, \"assembler-caller\"] .= weighted_roc_table[!, \"assembler\"] .* \" - \" .* weighted_roc_table[!, \"caller\"]\n",
    "variant_call_accuracy_visualization_groups = DataFrames.groupby(weighted_roc_table, [\"original_fasta\", \"coverage\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dec846d-8e12-491a-ba21-4ab50adee23a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "variant_call_accuracy_visualization_group = variant_call_accuracy_visualization_groups[2]\n",
    "sort!(variant_call_accuracy_visualization_group, \"assembler-caller\")\n",
    "this_fasta = basename(first(variant_call_accuracy_visualization_group[!, \"original_fasta\"]))\n",
    "this_coverage = first(variant_call_accuracy_visualization_group[!, \"coverage\"])\n",
    "title = \"$(this_fasta) @ $(this_coverage)\"\n",
    "unique_assembler_callers = variant_call_accuracy_visualization_group[!, \"assembler-caller\"]\n",
    "# Precision plot\n",
    "p = StatsPlots.scatter(\n",
    "    collect(1:length(unique_assembler_callers)) .- 0.1,\n",
    "    variant_call_accuracy_visualization_group.precision,\n",
    "    xticks = (1:length(unique_assembler_callers), unique_assembler_callers),\n",
    "    xlabel = \"Assembler-Caller\",\n",
    "    title = title,\n",
    "    label = \"Precision\",\n",
    "    ylims = (-0.1, 1.1),\n",
    "    margins = 10StatsPlots.Plots.PlotMeasures.mm,\n",
    "    xrotation=45,\n",
    "    legend=:bottomleft\n",
    ")\n",
    "\n",
    "# Sensitivity plot\n",
    "StatsPlots.scatter!(p,\n",
    "    collect(1:length(unique_assembler_callers)),\n",
    "    variant_call_accuracy_visualization_group.sensitivity,\n",
    "    label = \"Sensitivity\"\n",
    ")\n",
    "\n",
    "# F1 score plot\n",
    "StatsPlots.scatter!(p,\n",
    "    collect(1:length(unique_assembler_callers)) .+ 0.1,\n",
    "    variant_call_accuracy_visualization_group.f_measure,\n",
    "    label = \"F1 Score\"\n",
    ")\n",
    "\n",
    "display(p)\n",
    "StatsPlots.savefig(p, joinpath(results_dir, \"rtg-eval.$(this_fasta).$(this_coverage).svg\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.2",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
