{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335bf35f-c55e-4857-83f0-ed03dc666f4e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# if hit plotting library issues, try resetting LD path for julia\n",
    "# can set in ~/.local/share/jupyter/kernels/\n",
    "@assert ENV[\"LD_LIBRARY_PATH\"] == \"\"\n",
    "import Pkg\n",
    "Pkg.activate(;temp=true)\n",
    "Pkg.add(\"Revise\")\n",
    "import Revise\n",
    "\n",
    "# Pkg.develop(path=\"/global/cfs/projectdirs/m4269/cjprybol/Mycelia\")\n",
    "# Pkg.develop(path=\"../../..\")\n",
    "Pkg.develop(path=\"$(homedir())/workspace/Mycelia\")\n",
    "import Mycelia\n",
    "\n",
    "pkgs = String[\n",
    "    \"DataFrames\",\n",
    "    \"uCSV\",\n",
    "    \"Kmers\",\n",
    "    \"CodecZlib\",\n",
    "    \"StatsPlots\",\n",
    "    \"GLM\",\n",
    "    \"StatsBase\",\n",
    "    \"CSV\"\n",
    "]\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "128cc1e1-929a-42e1-8376-4c81832fd40c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "PROJECT_BASEDIR = dirname(pwd())\n",
    "data_dir = joinpath(PROJECT_BASEDIR, \"data\")\n",
    "genome_dir = mkpath(joinpath(data_dir, \"genomes\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c2889b8-d3a3-4134-aeea-59480a0417f8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "refseq_metadata = Mycelia.load_refseq_metadata();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9c09fc6-0573-4cbb-8914-f0cc22fbe7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function plot_kmer_frequency_spectra_with_trendline(counts; log_scale = log2, kwargs...)\n",
    "#     kmer_counts_hist = StatsBase.countmap(c for c in counts)\n",
    "#     xs = collect(keys(kmer_counts_hist))\n",
    "#     ys = collect(values(kmer_counts_hist))\n",
    "#     if isa(log_scale, Function)\n",
    "#         xs = log_scale.(xs)\n",
    "#         ys = log_scale.(ys)\n",
    "#     end\n",
    "    \n",
    "#     p = StatsPlots.plot(\n",
    "#         xs,\n",
    "#         ys,\n",
    "#         xlims = (0, maximum(xs) + max(1, ceil(0.1 * maximum(xs)))),\n",
    "#         ylims = (0, maximum(ys) + max(1, ceil(0.1 * maximum(ys)))),\n",
    "#         seriestype = :scatter,\n",
    "#         legend = false,\n",
    "#         xlabel = isa(log_scale, Function) ? \"$(log_scale)(observed frequency)\" : \"observed frequency\",\n",
    "#         ylabel = isa(log_scale, Function) ? \"$(log_scale)(# of kmers)\" : \"observed frequency\",\n",
    "#         ;kwargs...\n",
    "#     )\n",
    "#     return p\n",
    "# end\n",
    "\n",
    "# function plot_kmer_frequency_spectra(counts; log_scale = log2, kwargs...)\n",
    "#     kmer_counts_hist = StatsBase.countmap(c for c in counts)\n",
    "#     xs = collect(keys(kmer_counts_hist))\n",
    "#     ys = collect(values(kmer_counts_hist))\n",
    "#     if isa(log_scale, Function)\n",
    "#         xs = log_scale.(xs)\n",
    "#         ys = log_scale.(ys)\n",
    "#     end\n",
    "    \n",
    "#     p = StatsPlots.plot(\n",
    "#         xs,\n",
    "#         ys,\n",
    "#         xlims = (0, maximum(xs) + max(1, ceil(0.1 * maximum(xs)))),\n",
    "#         ylims = (0, maximum(ys) + max(1, ceil(0.1 * maximum(ys)))),\n",
    "#         seriestype = :scatter,\n",
    "#         legend = false,\n",
    "#         xlabel = isa(log_scale, Function) ? \"$(log_scale)(observed frequency)\" : \"observed frequency\",\n",
    "#         ylabel = isa(log_scale, Function) ? \"$(log_scale)(# of kmers)\" : \"observed frequency\",\n",
    "#         ;kwargs...\n",
    "#     )\n",
    "#     return p\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e2a3124-bd38-4fae-a238-2101815f1c2a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a048062-2947-4f12-946d-2567145afc0f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Danio rerio\n",
    "# 7955\n",
    "# GCF_000002035.6\n",
    "\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   2.059938 seconds (617 allocations: 309.625 KiB)\n",
    "#   0.000201 seconds (116 allocations: 9.531 KiB)\n",
    "# [ Info: /global/cfs/cdirs/m4269/cjprybol/Mycelia/projects/variant-calling-benchmarking/data/genomes/GCF_000002035.6/GCF_000002035.6_GRCz11_genomic.fna.gz.k7.canonical.jf.count_histogram.tsv already exists\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   2.051873 seconds (617 allocations: 309.641 KiB)\n",
    "#   1.875662 seconds (181.84 k allocations: 11.496 MiB)\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   1.962617 seconds (613 allocations: 309.562 KiB)\n",
    "#  21.618736 seconds (116.74 k allocations: 9.253 MiB)\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   1.977223 seconds (734 allocations: 320.039 KiB)\n",
    "# 479.764688 seconds (100.25 k allocations: 8.412 MiB)\n",
    "\n",
    "accession = \"GCF_000002035.6\"\n",
    "outdir=mkpath(joinpath(genome_dir, accession))\n",
    "ftp = first(refseq_metadata[(refseq_metadata[!, \"refseq_category\"] .== \"reference genome\") .& map(x -> x == accession, refseq_metadata[!, \"#assembly_accession\"]), \"ftp_path\"])\n",
    "fasta = Mycelia.download_genome_by_ftp(ftp=ftp, outdir=outdir)\n",
    "kmer_counts_dict = Dict()\n",
    "for k in [7, 11, 13, 17]\n",
    "    @time jellyfish_counts_file = Mycelia.jellyfish_count(fastx=fasta, k = k, canonical=true)\n",
    "    @time count_histogram_file = Mycelia.jellyfish_counts_to_kmer_frequency_histogram(jellyfish_counts_file)\n",
    "    kmer_counts_dict[k] = CSV.read(count_histogram_file, DataFrames.DataFrame, delim='\\t')\n",
    "end\n",
    "kmer_counts_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be3de209-91c2-4a12-9f19-3b9288e1ebe7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   2.707734 seconds (632 allocations: 310.141 KiB)\n",
    "#   0.462371 seconds (101.25 k allocations: 8.673 MiB)\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   2.008665 seconds (615 allocations: 309.594 KiB)\n",
    "#   2.110626 seconds (229.84 k allocations: 13.199 MiB)\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   2.025346 seconds (615 allocations: 309.594 KiB)\n",
    "#  21.925132 seconds (175.54 k allocations: 11.459 MiB)\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   2.024590 seconds (615 allocations: 309.594 KiB)\n",
    "# 832.700463 seconds (170.55 k allocations: 11.121 MiB)\n",
    "\n",
    "# Mus musculus\n",
    "# 10090\n",
    "# GCF_000001635.27\n",
    "accession = \"GCF_000001635.27\"\n",
    "outdir=mkpath(joinpath(genome_dir, accession))\n",
    "ftp = first(refseq_metadata[(refseq_metadata[!, \"refseq_category\"] .== \"reference genome\") .& map(x -> x == accession, refseq_metadata[!, \"#assembly_accession\"]), \"ftp_path\"])\n",
    "fasta = Mycelia.download_genome_by_ftp(ftp=ftp, outdir=outdir)\n",
    "kmer_counts_dict = Dict()\n",
    "for k in [7, 11, 13, 17]\n",
    "    @time jellyfish_counts_file = Mycelia.jellyfish_count(fastx=fasta, k = k, canonical=true)\n",
    "    @time count_histogram_file = Mycelia.jellyfish_counts_to_kmer_frequency_histogram(jellyfish_counts_file)\n",
    "    kmer_counts_dict[k] = CSV.read(count_histogram_file, DataFrames.DataFrame, delim='\\t')\n",
    "end\n",
    "kmer_counts_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f873132-fd8c-4c1e-83bd-0183592dd5ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Homo sapiens\n",
    "# 9606\n",
    "# GCF_000001405.40\n",
    "\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   2.123828 seconds (620 allocations: 309.953 KiB)\n",
    "#   0.425395 seconds (101.41 k allocations: 8.880 MiB)\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   1.656025 seconds (606 allocations: 309.172 KiB)\n",
    "#   1.583714 seconds (251.21 k allocations: 15.355 MiB)\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   1.745315 seconds (609 allocations: 309.500 KiB)\n",
    "#  20.587222 seconds (181.83 k allocations: 11.732 MiB)\n",
    "# [ Info: conda environment kmer-jellyfish already present; set force=true to update/re-install\n",
    "#   1.816905 seconds (608 allocations: 309.203 KiB)\n",
    "# 991.531925 seconds (176.60 k allocations: 11.345 MiB)\n",
    "\n",
    "accession = \"GCF_000001405.40\"\n",
    "outdir=mkpath(joinpath(genome_dir, accession))\n",
    "ftp = first(refseq_metadata[(refseq_metadata[!, \"refseq_category\"] .== \"reference genome\") .& map(x -> x == accession, refseq_metadata[!, \"#assembly_accession\"]), \"ftp_path\"])\n",
    "fasta = Mycelia.download_genome_by_ftp(ftp=ftp, outdir=outdir)\n",
    "kmer_counts_dict = Dict()\n",
    "for k in [7, 11, 13, 17]\n",
    "    @time jellyfish_counts_file = Mycelia.jellyfish_count(fastx=fasta, k = k, canonical=true)\n",
    "    @time count_histogram_file = Mycelia.jellyfish_counts_to_kmer_frequency_histogram(jellyfish_counts_file)\n",
    "    kmer_counts_dict[k] = CSV.read(count_histogram_file, DataFrames.DataFrame, delim='\\t')\n",
    "end\n",
    "kmer_counts_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e760535f-c92f-43b9-9a62-bfe2e7ba9a10",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# # Add back R-squared analysis\n",
    "# # write out as PDF & PNG\n",
    "# # counts = canonical_kmer_counts[!, \"count\"]\n",
    "# kmer_counts_hist = StatsBase.countmap(counts)\n",
    "# xs = log2.(collect(keys(kmer_counts_hist)))\n",
    "# ys = log2.(collect(values(kmer_counts_hist)))\n",
    "# # Fit the model\n",
    "# model = GLM.lm(GLM.@formula(_ys ~ _xs), DataFrames.DataFrame(_xs=xs, _ys=ys))\n",
    "# r2_value = GLM.r2(model)\n",
    "# p = Mycelia.plot_kmer_frequency_spectra(counts)\n",
    "# StatsPlots.plot!(p, title = basename(jellyfish_counts_file))\n",
    "# StatsPlots.plot!(p, xs, GLM.predict(model))\n",
    "# StatsPlots.annotate!(p, 0.5, 0.5, StatsPlots.text(\"r = $(round(r2_value, digits=2))\", 10, :left))\n",
    "\n",
    "# # Create sample data\n",
    "# x = 0:0.1:10\n",
    "# y1 = sin.(x)\n",
    "# y2 = cos.(x)\n",
    "# y3 = tan.(x)\n",
    "\n",
    "# # Create a multi-panel plot\n",
    "# p = StatsPlots.plot(layout = (3, 1), size = (600, 800), xlabel = \"Shared X-axis\", title = accession)\n",
    "\n",
    "# # First subplot\n",
    "# StatsPlots.plot!(p, x, y1, subplot = 1, ylabel = \"Y1\", title = \"k=7\", xlabel=\"\")\n",
    "\n",
    "# StatsPlots.plot!(p, x, y2, subplot = 2, ylabel = \"Y2\", title = \"k=11\", xlabel=\"\")\n",
    "\n",
    "# StatsPlots.plot!(p, x, y3, subplot = 2, ylabel = \"Y2\", title = \"k=13\", xlabel=\"\")\n",
    "\n",
    "# StatsPlots.plot!(p, x, y4, subplot = 2, ylabel = \"Y2\", title = \"k=17\", xlabel=\"\")\n",
    "\n",
    "# p"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.10.2",
   "language": "julia",
   "name": "julia-1.10"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
