{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335bf35f-c55e-4857-83f0-ed03dc666f4e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# if hit plotting library issues, try resetting LD path for julia\n",
    "# can set in ~/.local/share/jupyter/kernels/\n",
    "# @assert ENV[\"LD_LIBRARY_PATH\"] == \"\"\n",
    "import Pkg\n",
    "pkgs = [\n",
    "    \"Revise\",\n",
    "    \"FASTX\",\n",
    "    \"BioSequences\",\n",
    "    \"Kmers\",\n",
    "    \"Graphs\",\n",
    "    \"MetaGraphs\",\n",
    "    \"SparseArrays\",\n",
    "    \"ProgressMeter\",\n",
    "    \"Distributions\",\n",
    "    \"HiddenMarkovModels\",\n",
    "    \"BioAlignments\",\n",
    "    \"StatsBase\"\n",
    "]\n",
    "# Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "# Pkg.develop(path=\"/global/cfs/projectdirs/m4269/cjprybol/Mycelia\")\n",
    "# Pkg.develop(path=\"../../..\")\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d2ad095-a433-4185-82fb-cdf61aa6328e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "function find_contiguous_stretches_as_ranges(indices)\n",
    "    # Sort the indices to ensure they are in ascending order\n",
    "    sorted_indices = sort(indices)\n",
    "    \n",
    "    # Initialize the result list\n",
    "    contiguous_stretches = UnitRange{Int64}[]\n",
    "    \n",
    "    # Start with the first index as both the start and end of the current stretch\n",
    "    start_of_current_stretch = sorted_indices[1]\n",
    "    end_of_current_stretch = sorted_indices[1]\n",
    "\n",
    "    # Iterate through the sorted indices starting from the second element\n",
    "    for i in 2:length(sorted_indices)\n",
    "        # Check if the current index is consecutive to the end of the current stretch\n",
    "        if sorted_indices[i] == end_of_current_stretch + 1\n",
    "            end_of_current_stretch = sorted_indices[i]\n",
    "        else\n",
    "            # Current index is not consecutive, so the current stretch ends here\n",
    "            # Add the current stretch as a range to the result list\n",
    "            push!(contiguous_stretches, start_of_current_stretch:end_of_current_stretch)\n",
    "            \n",
    "            # Update the start and end of the new current stretch\n",
    "            start_of_current_stretch = sorted_indices[i]\n",
    "            end_of_current_stretch = sorted_indices[i]\n",
    "        end\n",
    "    end\n",
    "\n",
    "    # Add the last stretch to the result list as a range\n",
    "    push!(contiguous_stretches, start_of_current_stretch:end_of_current_stretch)\n",
    "\n",
    "    return contiguous_stretches\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56fe180b-4889-4a89-b262-d494a2e5e4a5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "function emission_likelihood(;error_rate, accuracy, observed_sequence, alternative_sequence, fast=false)\n",
    "    likelihood = 1.0\n",
    "    if fast\n",
    "        for (a, b) in zip(observed_sequence, alternative_sequence)\n",
    "            if a == b\n",
    "                likelihood *= accuracy\n",
    "            else\n",
    "                likelihood *= error_rate\n",
    "            end\n",
    "        end\n",
    "    else\n",
    "        alignment_result = BioAlignments.pairalign(BioAlignments.LevenshteinDistance(), observed_sequence, alternative_sequence)\n",
    "        distance = BioAlignments.distance(alignment_result)\n",
    "        matches = length(observed_sequence) - distance\n",
    "        for i in 1:matches\n",
    "            likelihood *= accuracy\n",
    "        end\n",
    "        for i in 1:distance\n",
    "            likelihood *= error_rate\n",
    "        end\n",
    "    end\n",
    "    return likelihood\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "128cc1e1-929a-42e1-8376-4c81832fd40c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "PROJECT_BASEDIR = dirname(pwd())\n",
    "data_dir = joinpath(PROJECT_BASEDIR, \"data\")\n",
    "genome_dir = mkpath(joinpath(data_dir, \"genomes\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01ad975e-4976-45af-8157-a8f8656cdd1a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "reference_fastas = sort(filter(x -> occursin(r\"\\.fna$\", x) && !occursin(\"normalized\", x), readdir(genome_dir, join=true)), by=x->filesize(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8bdbaf8-d1a0-471c-a077-eab949fda355",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "reference_fasta = first(reference_fastas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea39a3f1-b82e-4ed8-a99c-4c9ed1642836",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "filter(x -> occursin(reference_fasta, x) && occursin(\"10x\", x) && occursin(r\"\\.fq\\.gz$\", x), readdir(genome_dir, join=true))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ffc059b-cec6-4165-b446-b44f6ca88c0d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fastq = filter(x -> occursin(reference_fasta, x) && occursin(\"10x\", x) && occursin(\"filtlong\", x) && occursin(r\"\\.fq\\.gz$\", x), readdir(genome_dir, join=true))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebdc9d43-8c88-4ae4-827b-e387e09784af",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "initial_k = Mycelia.assess_dnamer_saturation(fastq, plot=false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c335dce-e4e1-48bd-a0c6-3bae14493904",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "initial_kmer_type = Kmers.DNAKmer{initial_k}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eae4ac2b-e629-4601-b1c4-e24b1ed73e88",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# kmer_iterator = Kmers.EveryKmer{initial_kmer_type}(BioSequences.LongDNA{2}(FASTX.sequence(first(collect(Mycelia.open_fastx(first(fastq)))))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74ce46f6-d2ca-45bd-bf21-96991572177f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "raw_graph = Mycelia.fastx_to_kmer_graph(initial_kmer_type, fastq)\n",
    "solid_kmers = findall(v -> MetaGraphs.get_prop(graph, v, :count) >= 3, Graphs.vertices(graph))\n",
    "graph, vertex_map = Graphs.induced_subgraph(graph, solid_kmers)\n",
    "MetaGraphs.set_indexing_prop!(graph, :kmer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d86f398-1cf6-4434-bbf6-d405a9c34e07",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# build a directed, simple graph from undirected metagraph\n",
    "canonical_kmers = (MetaGraphs.get_prop(graph, v, :kmer) for v in Graphs.vertices(graph))\n",
    "unique_stranded_kmers = unique(sort(reduce(vcat, ([canonical_kmer, BioSequences.reverse_complement(canonical_kmer)] for canonical_kmer in canonical_kmers))))\n",
    "stranded_kmer_graph = MetaGraphs.MetaDiGraph(length(unique_stranded_kmers))\n",
    "MetaGraphs.set_prop!(stranded_kmer_graph, :k, MetaGraphs.get_prop(graph, :k))\n",
    "@info \"adding kmers\"\n",
    "for (i, kmer) in enumerate(unique_stranded_kmers)\n",
    "    MetaGraphs.set_prop!(stranded_kmer_graph, i, :kmer, kmer)\n",
    "end\n",
    "@info \"indexing kmers\"\n",
    "# allow graph[kmer, :kmer] to dict-lookup the index of a kmer\n",
    "MetaGraphs.set_indexing_prop!(stranded_kmer_graph, :kmer)\n",
    "\n",
    "for vertex in Graphs.vertices(graph)\n",
    "    observation_count = MetaGraphs.get_prop(graph, vertex, :count)\n",
    "    canonical_kmer = MetaGraphs.get_prop(graph, vertex, :kmer)\n",
    "    MetaGraphs.set_prop!(stranded_kmer_graph, stranded_kmer_graph[canonical_kmer, :kmer], :count, observation_count)\n",
    "    MetaGraphs.set_prop!(stranded_kmer_graph, stranded_kmer_graph[BioSequences.reverse_complement(canonical_kmer), :kmer], :count, observation_count)\n",
    "end\n",
    "\n",
    "# iterate through edges in graph and use edge evidence to build these stranded edges and their weights\n",
    "for edge in Graphs.edges(graph)\n",
    "    for edge_observation in MetaGraphs.get_prop(graph, edge, :observations)\n",
    "        \n",
    "        ##### forward\n",
    "        source = (index=edge.src, orientation=first(edge_observation.orientation))\n",
    "        oriented_source = graph[source.index, :kmer]\n",
    "        if !source.orientation\n",
    "            oriented_source = BioSequences.reverse_complement(oriented_source)\n",
    "        end\n",
    "        \n",
    "        dest = (index=edge.dst, orientation=last(edge_observation.orientation))\n",
    "        oriented_dest = graph[dest.index, :kmer]\n",
    "        if !dest.orientation\n",
    "            oriented_dest = BioSequences.reverse_complement(oriented_dest)\n",
    "        end\n",
    "        \n",
    "        stranded_edge = Graphs.Edge(stranded_kmer_graph[oriented_source, :kmer] => stranded_kmer_graph[oriented_dest, :kmer])\n",
    "        stranded_observation = (record_identifier = edge_observation.record_identifier, index = edge_observation.index)\n",
    "        if !Graphs.has_edge(stranded_kmer_graph, stranded_edge)\n",
    "            Graphs.add_edge!(stranded_kmer_graph, stranded_edge)\n",
    "            MetaGraphs.set_prop!(stranded_kmer_graph, stranded_edge, :observations, Set([stranded_observation]))\n",
    "        else\n",
    "            stranded_observations = push!(MetaGraphs.get_prop(stranded_kmer_graph, stranded_edge, :observations), stranded_observation)\n",
    "            MetaGraphs.set_prop!(stranded_kmer_graph, stranded_edge, :observations, stranded_observations)\n",
    "        end\n",
    "        \n",
    "        \n",
    "        ##### reverse_complement\n",
    "        source = (index=edge.dst, orientation=!last(edge_observation.orientation))\n",
    "        oriented_source = graph[source.index, :kmer]\n",
    "        if !source.orientation\n",
    "            oriented_source = BioSequences.reverse_complement(oriented_source)\n",
    "        end\n",
    "        \n",
    "        dest = (index=edge.src, orientation=!first(edge_observation.orientation))\n",
    "        oriented_dest = graph[dest.index, :kmer]\n",
    "        if !dest.orientation\n",
    "            oriented_dest = BioSequences.reverse_complement(oriented_dest)\n",
    "        end\n",
    "        \n",
    "        stranded_edge = Graphs.Edge(stranded_kmer_graph[oriented_source, :kmer] => stranded_kmer_graph[oriented_dest, :kmer])\n",
    "        stranded_observation = (record_identifier = edge_observation.record_identifier, index = edge_observation.index)\n",
    "        if !Graphs.has_edge(stranded_kmer_graph, stranded_edge)\n",
    "            Graphs.add_edge!(stranded_kmer_graph, stranded_edge)\n",
    "            MetaGraphs.set_prop!(stranded_kmer_graph, stranded_edge, :observations, Set([stranded_observation]))\n",
    "        else\n",
    "            stranded_observations = push!(MetaGraphs.get_prop(stranded_kmer_graph, stranded_edge, :observations), stranded_observation)\n",
    "            MetaGraphs.set_prop!(stranded_kmer_graph, stranded_edge, :observations, stranded_observations)\n",
    "        end\n",
    "        \n",
    "    end\n",
    "end\n",
    "stranded_kmer_graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7d1dda6-0903-419b-9d45-b95c1a75a830",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "kmer_counts = map(v -> MetaGraphs.get_prop(stranded_kmer_graph, v, :count), Graphs.vertices(stranded_kmer_graph))\n",
    "state_likelihoods = kmer_counts ./ sum(kmer_counts)\n",
    "ranked_state_likelihoods = sortperm(state_likelihoods, rev=true)\n",
    "\n",
    "transition_likelihoods = SparseArrays.spzeros(Graphs.nv(stranded_kmer_graph), Graphs.nv(stranded_kmer_graph))\n",
    "ProgressMeter.@showprogress for v in Graphs.vertices(stranded_kmer_graph)\n",
    "    dijkstra = Graphs.dijkstra_shortest_paths(stranded_kmer_graph, v)\n",
    "    # 0 distance is a deletion, but let's make that a \n",
    "    distances = map(x -> x == 0 ? 2 : x, dijkstra.dists)\n",
    "    inverse_distance_based_transition_likelihoods = 1 ./ distances\n",
    "    transition_likelihoods[v, :] .= inverse_distance_based_transition_likelihoods\n",
    "end\n",
    "ProgressMeter.@showprogress for v in Graphs.vertices(stranded_kmer_graph)\n",
    "    transition_likelihoods[v, :] .= transition_likelihoods[v, :] ./ sum(transition_likelihoods[v, :])\n",
    "end\n",
    "transition_likelihoods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5742b732-8f75-4543-99ec-50ef46e57199",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "assumed_error_rate = 1 / (MetaGraphs.get_prop(graph, :k) + 1)\n",
    "assumed_accuracy = 1 - assumed_error_rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6227f477-9e66-4429-b004-8640c3761ce2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "kmer_indices = Dict(kmer => i for (i, kmer) in enumerate(map(v -> MetaGraphs.get_prop(stranded_kmer_graph, v, :kmer), Graphs.vertices(stranded_kmer_graph))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03d8433c-8be9-4799-902a-5e144a39dd24",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "records = collect(values(MetaGraphs.get_prop(graph, :records)))\n",
    "record = first(records)\n",
    "sequence = BioSequences.LongDNA{4}(FASTX.sequence(record))\n",
    "\n",
    "\n",
    "# observed_kmer_states = collect(Kmers.EveryKmer{Kmers.DNAKmer{MetaGraphs.get_prop(graph, :k)}}(sequence))\n",
    "# observed_kmer_indices = [stranded_kmer_graph[kmer, :kmer] for kmer in last.(observed_kmer_states)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18596ac3-0b0d-40e4-bb27-1ca53ebbe639",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "observed_kmers = last.(Kmers.EveryKmer{Kmers.DNAKmer{MetaGraphs.get_prop(graph, :k)}}(sequence))\n",
    "observed_kmer_indices = [get(kmer_indices, observed_kmer, missing) for observed_kmer in observed_kmers]\n",
    "\n",
    "contiguous_stretches = find_contiguous_stretches_as_ranges(findall(ismissing, observed_kmer_indices))\n",
    "\n",
    "for (i, contiguous_stretch) in enumerate(contiguous_stretches)\n",
    "    if length(contiguous_stretch) < MetaGraphs.get_prop(graph, :k)\n",
    "        n_short = MetaGraphs.get_prop(graph, :k) - length(contiguous_stretch)\n",
    "        n_to_prune_on_each_side = Int(ceil(n_short / 2))\n",
    "        for i in first(contiguous_stretch)-n_to_prune_on_each_side:first(contiguous_stretch)\n",
    "            observed_kmer_indices[i] = missing\n",
    "        end\n",
    "        for i in last(contiguous_stretch):last(contiguous_stretch)+n_to_prune_on_each_side\n",
    "            observed_kmer_indices[i] = missing\n",
    "        end        \n",
    "    end\n",
    "end\n",
    "contiguous_stretches\n",
    "observed_kmer_indices\n",
    "# contiguous_stretches = find_contiguous_stretches_as_ranges(findall(ismissing, observed_kmer_indices))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14d97f8a-246d-4c97-b66d-f3009fdbeb31",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "contiguous_stretches = find_contiguous_stretches_as_ranges(findall(ismissing, observed_kmer_indices))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eab640a8-3cb6-4546-b6a2-36bb2f5665b8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "448138e9-0a08-41cb-b11e-27cf19f6875a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for contiguous_stretch in contiguous_stretches\n",
    "    # @show contiguous_stretch\n",
    "    replacement_path_length = 0\n",
    "    alternate_path = []\n",
    "    prior_solid_kmer_index = first(contiguous_stretch)\n",
    "    posterior_solid_kmer_index = last(contiguous_stretch)\n",
    "    while (replacement_path_length == 0) && (prior_solid_kmer_index > 1) && (posterior_solid_kmer_index < length(observed_kmer_indices))\n",
    "        prior_solid_kmer_index -= 1\n",
    "        posterior_solid_kmer_index += 1\n",
    "        alternate_path = Graphs.a_star(stranded_kmer_graph, observed_kmer_indices[prior_solid_kmer_index], observed_kmer_indices[posterior_solid_kmer_index])\n",
    "        replacement_path_length = length(alternate_path)\n",
    "    end\n",
    "    @show length(alternate_path), length(contiguous_stretch), replacement_path_length\n",
    "    @show prior_solid_kmer_index, posterior_solid_kmer_index\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20e674d6-7ca3-4919-9368-fd8cddadf4d9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdc34309-e1f4-4a48-8f24-03314bb0a834",
   "metadata": {},
   "outputs": [],
   "source": [
    "for contigu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f32694b9-b9e3-49a3-a170-eaa5e50b5887",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # emission_probability_distributions = Distributions.Categorical{Float64, Vector{Float64}}[]\n",
    "# # ProgressMeter.@showprogress for v in Graphs.vertices(stranded_kmer_graph)\n",
    "# #     # @show v\n",
    "# #     state_emission_likelihoods = ones(Graphs.nv(stranded_kmer_graph))\n",
    "# #     for (i, v2) in enumerate(Graphs.vertices(stranded_kmer_graph))\n",
    "# #         state_emission_likelihoods[i] *= emission_likelihood(\n",
    "# #                 error_rate=assumed_error_rate,\n",
    "# #                 accuracy=assumed_accuracy,\n",
    "# #                 observed_sequence=stranded_kmer_graph[v, :kmer],\n",
    "# #                 alternative_sequence=stranded_kmer_graph[v2, :kmer])\n",
    "        \n",
    "# #     end\n",
    "# #     state_emission_likelihoods = state_emission_likelihoods ./ sum(state_emission_likelihoods)\n",
    "# #     push!(emission_probability_distributions, Distributions.Categorical(state_emission_likelihoods))\n",
    "# # end\n",
    "# # emission_probability_distributions\n",
    "# # hmm = HiddenMarkovModels.HMM(state_likelihoods, transition_likelihoods, emission_probability_distributions)\n",
    "\n",
    "# # Forward\n",
    "\n",
    "# # Viterbri\n",
    "\n",
    "# viterbi(hmm, obs_seq)\n",
    "\n",
    "# records = collect(values(MetaGraphs.get_prop(graph, :records)))\n",
    "# record = first(records)\n",
    "# sequence = BioSequences.LongDNA{4}(FASTX.sequence(record))\n",
    "# observed_kmer_states = collect(Kmers.EveryKmer{Kmers.DNAKmer{MetaGraphs.get_prop(graph, :k)}}(sequence))\n",
    "# observed_kmer_indices = [stranded_kmer_graph[kmer, :kmer] for kmer in last.(observed_kmer_states)]\n",
    "# log_likelihood_observation = last(HiddenMarkovModels.forward(hmm, observed_kmer_indices))\n",
    "\n",
    "# HiddenMarkovModels.viterbi(hmm, observed_kmer_indices)\n",
    "\n",
    "# maximum_likelihood_observation = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3975036b-5cee-4ce9-9a96-143ee2524d5e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# arrivals = Dict{NamedTuple{(:record_index, :state_index), Tuple{Int64, Int64}}, Graphs.SimpleGraphs.SimpleEdge{Int64}}()\n",
    "# state_evaluation_matrix = SparseArrays.spzeros(BigFloat, Graphs.nv(stranded_kmer_graph), length(observed_kmer_states))\n",
    "# (current_record_index, observed_kmer_state) = first(observed_kmer_states)\n",
    "# prior_record_index = current_record_index - 1\n",
    "# ProgressMeter.@showprogress for (state_index, alt_kmer) in enumerate(unique_stranded_kmers)\n",
    "#     state_evaluation_matrix[state_index, current_record_index] = emission_likelihood(\n",
    "#         error_rate=assumed_error_rate,\n",
    "#         accuracy=assumed_accuracy,\n",
    "#         observed_sequence=observed_kmer_state,\n",
    "#         alternative_sequence=alt_kmer)\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a0dfa08-037c-4cc7-9cda-a62e80dfae18",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ProgressMeter.@showprogress for (current_record_index, observed_kmer_state) in observed_kmer_states[2:end]\n",
    "#     prior_record_index = current_record_index - 1\n",
    "#     ranked_prior_states = sortperm(state_evaluation_matrix[:, prior_record_index], rev=true)\n",
    "#     matching_kmer_index = stranded_kmer_graph[observed_kmer_state, :kmer]\n",
    "#     ranked_alternative_kmer_indices = filter(x -> x != matching_kmer_index, ranked_state_likelihoods)\n",
    "#     ranked_next_states = vcat([matching_kmer_index], ranked_alternative_kmer_indices)\n",
    "\n",
    "#     for source_index in ranked_prior_states\n",
    "#         source_likelihood = state_evaluation_matrix[source_index, prior_record_index]\n",
    "#         if source_likelihood > 0\n",
    "#             for destination_index in ranked_next_states\n",
    "#                 current_destination_likelihood = state_evaluation_matrix[destination_index, current_record_index]\n",
    "#                 # we have a chance to find a higher likelihood path\n",
    "#                 if source_likelihood > current_destination_likelihood\n",
    "#                     considered_destination_likelihood = source_likelihood * transition_likelihoods[source_index, destination_index] * state_likelihoods[destination_index]\n",
    "#                     if considered_destination_likelihood > current_destination_likelihood\n",
    "#                         considered_destination_likelihood *= emission_likelihood(\n",
    "#                             error_rate=assumed_error_rate,\n",
    "#                             accuracy=assumed_accuracy,\n",
    "#                             observed_sequence=observed_kmer_state,\n",
    "#                             alternative_sequence=stranded_kmer_graph[destination_index, :kmer])\n",
    "#                         if considered_destination_likelihood > current_destination_likelihood\n",
    "#                             state_evaluation_matrix[destination_index, current_record_index] = considered_destination_likelihood\n",
    "#                             arrivals[(record_index = current_record_index, state_index = destination_index)] = Graphs.Edge(source_index, destination_index)\n",
    "#                         end\n",
    "#                     end\n",
    "#                 end\n",
    "#             end\n",
    "#         end\n",
    "#     end\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06c7099b-3619-473a-8e31-4d3e0a5f0d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# using Distributions, HiddenMarkovModels\n",
    "# init = [0.4, 0.6]\n",
    "# trans = [0.9 0.1; 0.2 0.8]\n",
    "# dists = [Normal(-1.0), Normal(1.0)]\n",
    "# hmm = HMM(init, trans, dists)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76be58b4-29a4-4a20-a38b-2f9120d8a738",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# shortest_paths = Dict{Graphs.Edge, NamedTuple{(:path, :likelihood), Tuple{Vector{Graphs.SimpleGraphs.SimpleEdge{Int64}}, Float64}}}()\n",
    "# # 5 hours to solve - too long!\n",
    "# ProgressMeter.@showprogress for row in 1:size(transition_likelihoods, 1)\n",
    "#     for col in 1:size(transition_likelihoods, 2)\n",
    "#         path_likelihood = transition_likelihoods[row, col]\n",
    "#         if 1 > path_likelihood > 0\n",
    "#             distance = Int(1 / path_likelihood) - 1\n",
    "#             edge = Graphs.Edge(row, col)\n",
    "#             if distance == 1\n",
    "#                 path = [edge]\n",
    "#             elseif distance > 1\n",
    "#                 path = Graphs.a_star(stranded_kmer_graph, row, col)\n",
    "#             end\n",
    "#             for edge in path\n",
    "#                 outneighbors = Graphs.outneighbors(stranded_kmer_graph, edge.src)\n",
    "#                 outneighbor_weights = [length(MetaGraphs.get_prop(stranded_kmer_graph, Graphs.Edge(edge.src, outneighbor), :observations)) for outneighbor in outneighbors]\n",
    "#                 target_index = findfirst(x -> x == edge.dst, outneighbors)\n",
    "#                 outneighbor_likelihoods = outneighbor_weights ./ sum(outneighbor_weights)\n",
    "#                 path_likelihood *= outneighbor_likelihoods[target_index]\n",
    "#             end\n",
    "#             shortest_paths[edge] = (path = path, likelihood = path_likelihood)\n",
    "#         end\n",
    "#         transition_likelihoods[row, col] = path_likelihood\n",
    "#     end\n",
    "# end\n",
    "# shortest_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acdc5aa3-3963-459c-86a5-148e5c1f23bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35bb9f42-9a46-4ece-a525-992354d467cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# solve the "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68ac031c-513c-4824-b9f4-0c9bf07f2e38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function initialize_transition_probabilities(kmer_graph)\n",
    "    \n",
    "#     total_kmers = Graphs.nv(kmer_graph)\n",
    "#     transition_likelihoods = Dict(\n",
    "#         true => SparseArrays.spzeros(total_kmers, total_kmers),\n",
    "#         false => SparseArrays.spzeros(total_kmers, total_kmers)\n",
    "#     )\n",
    "\n",
    "#     for edge in collect(Graphs.edges(kmer_graph))\n",
    "# #         weight = length(kmer_graph.eprops[edge][:evidence])\n",
    "#         weight = kmer_graph.eprops[edge][:weight]\n",
    "#         for o in kmer_graph.eprops[edge][:orientations]\n",
    "#             transition_likelihoods[o.source_orientation][edge.src, edge.dst] = weight\n",
    "#         end\n",
    "#     end\n",
    "\n",
    "#     for source_orientation in (true, false)\n",
    "#         for src in 1:total_kmers\n",
    "#             transition_weights = transition_likelihoods[source_orientation][src, :]\n",
    "#             total_weight = sum(transition_weights)\n",
    "#             dsts, vals = SparseArrays.findnz(transition_weights)\n",
    "#             for (dst, val) in zip(dsts, vals) \n",
    "#                 transition_likelihoods[source_orientation][src, dst] = val / total_weight\n",
    "#             end\n",
    "#             normalized_probability = sum(transition_likelihoods[source_orientation][src, :])\n",
    "#             @assert isapprox(normalized_probability, 0) || isapprox(normalized_probability, 1)\n",
    "#         end\n",
    "#     end\n",
    "#     return transition_likelihoods\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6619f432-1ae0-4409-8ad5-f64dfbbab650",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function run_viterbi!(\n",
    "#         current_state,\n",
    "#         prior_state,\n",
    "#         observed_nucleotide,\n",
    "#         observed_quality_score,\n",
    "#         observed_error_rate,\n",
    "#         current_vertex,\n",
    "#         prior_vertex,\n",
    "#         state_likelihoods,\n",
    "#         transition_likelihoods,\n",
    "#         shortest_paths,\n",
    "#         arrival_paths,\n",
    "#         kmer_graph,\n",
    "#         kmer_likelihoods\n",
    "#         )\n",
    "#     # if probability of prior state is lower than current probability, skip\n",
    "\n",
    "# #     @show current_state\n",
    "# #     @show prior_state\n",
    "# #     @show current_vertex\n",
    "# #     @show prior_vertex\n",
    "    \n",
    "    \n",
    "#     current_state_likelihood = state_likelihoods[current_vertex, current_state]\n",
    "#     prior_state_likelihood = state_likelihoods[prior_vertex, prior_state]\n",
    "\n",
    "#     # if we already have a better possible path, skip calculating anything\n",
    "#     if prior_state_likelihood < current_state_likelihood\n",
    "# #         @show prior_state_likelihood < current_state_likelihood\n",
    "#         return\n",
    "#     end\n",
    "\n",
    "#     # take shortest path and assume it's the maximum likelihood path\n",
    "#     # this assumption seems fair because in an ideal situation\n",
    "#     # we're just moving to an adjacent kmer\n",
    "#     # and the shortest path and most likely path should be the same\n",
    "#     shortest_path = shortest_paths[prior_vertex][current_vertex]\n",
    "    \n",
    "# #     no path & not considering insertion\n",
    "#     if isempty(shortest_path) && (prior_vertex != current_vertex)\n",
    "# #         @show \"no path, skipping\"\n",
    "#         return\n",
    "#     end\n",
    "    \n",
    "#     # if shortest path isn't viable, exit\n",
    "#     if !isempty(shortest_path)\n",
    "# #         @show \"checking if path is viable\"\n",
    "\n",
    "#         terminal_orientation_prior_state = last(last(arrival_paths[prior_vertex, prior_state]))\n",
    "# #         @show arrival_paths[prior_vertex, prior_state]\n",
    "# #         @show \"we were at vertex $(prior_vertex) in orientation $(terminal_orientation_prior_state)\"\n",
    "#         candidate_edge = Graphs.Edge(shortest_path[1], shortest_path[2])\n",
    "                \n",
    "#         if !ismissing(terminal_orientation_prior_state) && \n",
    "#             !any(o -> o.source_orientation == terminal_orientation_prior_state, kmer_graph.eprops[candidate_edge][:orientations])\n",
    "            \n",
    "# #             @show \"no viable orientation matching edges detected between $(candidate_edge)\"\n",
    "# #             @show \"full candidate path was $(shortest_path)\"\n",
    "# #             @show \"orientation options were:\"\n",
    "# #             @show kmer_graph.eprops[candidate_edge][:orientations]\n",
    "#             return\n",
    "#         end\n",
    "#     end\n",
    "    \n",
    "#     # zero step path - insertion in observed sequence relative to kmer graph\n",
    "#     is_same_vertex = (current_vertex == prior_vertex)\n",
    "#     has_edge = Graphs.has_edge(kmer_graph, Graphs.Edge(prior_vertex, current_vertex))\n",
    "#     if is_same_vertex && has_edge\n",
    "#         shortest_path = [prior_vertex, current_vertex]\n",
    "#     end\n",
    "    \n",
    "#     if is_same_vertex\n",
    "# #         @show \"same vertex, considering insertion potential\"\n",
    "#         emission_likelihood = observed_error_rate\n",
    "#         transition_likelihood = observed_error_rate\n",
    "#         state_likelihood = kmer_likelihoods[current_vertex]\n",
    "#         path_likelihood = prior_state_likelihood * emission_likelihood * transition_likelihood * state_likelihood\n",
    "#         path = [last(arrival_paths[prior_vertex, prior_state])]\n",
    "\n",
    "#         if current_state_likelihood > state_likelihoods[current_vertex, current_state]\n",
    "# #             @show \"selecting path\"\n",
    "# #             @show path\n",
    "# #             @show path_likelihood\n",
    "#             state_likelihoods[current_vertex, current_state] = path_likelihood\n",
    "#             arrival_paths[current_vertex, current_state] = path\n",
    "#         end\n",
    "#     # one or more step path - match, mismatch, or deletion in observed sequence relative to kmer graph\n",
    "#     elseif !isempty(shortest_path)\n",
    "# #         @show \"path is viable!\"\n",
    "# #         @show \"considering shortest path: $(shortest_path)\"\n",
    "\n",
    "#         initial_path_state = last(arrival_paths[prior_vertex, prior_state])\n",
    "\n",
    "#         path = Vector{typeof(initial_path_state)}(undef, length(shortest_path))\n",
    "#         path[1] = initial_path_state\n",
    "\n",
    "#         path_likelihood::Float64 = state_likelihoods[prior_vertex, prior_state]\n",
    "\n",
    "#         for i in 2:length(shortest_path)\n",
    "\n",
    "#             this_vertex = shortest_path[i]\n",
    "#             prior_vertex, prior_orientation = path[i-1]\n",
    "#             edge = Graphs.Edge(prior_vertex, this_vertex)\n",
    "\n",
    "#             possible_edge_orientations::Set{NamedTuple{(:source_orientation, :destination_orientation), Tuple{Bool, Bool}}} = kmer_graph.eprops[edge][:orientations]\n",
    "            \n",
    "# #             @show possible_edge_orientations\n",
    "            \n",
    "#             if !ismissing(prior_orientation)\n",
    "#                 possible_edge_orientations = filter(o -> o.source_orientation == prior_orientation, possible_edge_orientations)\n",
    "#             end\n",
    "            \n",
    "# #             @show possible_edge_orientations\n",
    "            \n",
    "#             if isempty(possible_edge_orientations)\n",
    "#                 path_likelihood *= 0.0\n",
    "#                 path = Vector{eltype(path)}()\n",
    "# #                 @show \"no possible orientations, bailing early\"\n",
    "#                 break\n",
    "#             end\n",
    "\n",
    "# #             @show prior_orientation\n",
    "#             if ismissing(prior_orientation)\n",
    "#                 if transition_likelihoods[true][prior_vertex, this_vertex] > transition_likelihoods[false][prior_vertex, this_vertex]\n",
    "#                     prior_orientation = true\n",
    "#                     transition_likelihood = transition_likelihoods[true][prior_vertex, this_vertex]::Float64\n",
    "#                 elseif transition_likelihoods[true][prior_vertex, this_vertex] < transition_likelihoods[false][prior_vertex, this_vertex]\n",
    "#                     prior_orientation = false\n",
    "#                     transition_likelihood = transition_likelihoods[false][prior_vertex, this_vertex]::Float64\n",
    "#                 else transition_likelihoods[true][prior_vertex, this_vertex] == transition_likelihoods[false][prior_vertex, this_vertex]\n",
    "#                     prior_orientation = missing\n",
    "#                     transition_likelihood = transition_likelihoods[true][prior_vertex, this_vertex]::Float64\n",
    "#                 end\n",
    "#             else\n",
    "#                 transition_likelihood = transition_likelihoods[prior_orientation][prior_vertex, this_vertex]::Float64\n",
    "#             end\n",
    "#             state_likelihood::Float64 = kmer_likelihoods[this_vertex]\n",
    "#             path_likelihood *= transition_likelihood * state_likelihood\n",
    "            \n",
    "#             if length(possible_edge_orientations) == 1\n",
    "#                 orientation = first(possible_edge_orientations).destination_orientation\n",
    "#                 path[i] = this_vertex => orientation\n",
    "#             else\n",
    "#                 path[i] = this_vertex => missing\n",
    "#             end\n",
    "#         end\n",
    "\n",
    "#         # see if new nucleotide is a match or mismatch to terminal kmer in path\n",
    "#         if !isempty(path) && path_likelihood > 0\n",
    "#             terminal_kmer_index, terminal_kmer_orientation = last(path)\n",
    "#             terminal_kmer = BioSequences.LongDNASeq(kmer_graph.vprops[terminal_kmer_index][:kmer])::BioSequences.LongDNASeq\n",
    "#             if ismissing(terminal_kmer_orientation)\n",
    "#                 fw_is_match = observed_nucleotide == last(terminal_kmer)\n",
    "#                 bw_is_match = observed_nucleotide == last(BioSequences.reverse_complement!(terminal_kmer))\n",
    "#                 if fw_ismatch && !bw_is_match\n",
    "#                     path[end] = terminal_kmer_index => true\n",
    "#                     path_likelihood *= 1 - observed_error_rate\n",
    "#                 elseif !fw_ismatch && bw_is_match\n",
    "#                     path[end] = terminal_kmer_index => false\n",
    "#                     path_likelihood *= 1 - observed_error_rate\n",
    "#                 elseif fw_ismatch && bw_is_match\n",
    "#                     path_likelihood *= 1 - observed_error_rate\n",
    "#                 elseif !fw_ismatch && !bw_is_match\n",
    "#                     path_likelihood *= observed_error_rate\n",
    "#                 end\n",
    "#             elseif terminal_kmer_orientation\n",
    "#                 is_match = observed_nucleotide == last(terminal_kmer)\n",
    "#                 if is_match\n",
    "#                     path_likelihood *= 1 - observed_error_rate\n",
    "#                 else\n",
    "#                     path_likelihood *= observed_error_rate\n",
    "#                 end\n",
    "#             else\n",
    "#                 terminal_kmer = BioSequences.reverse_complement!(terminal_kmer)\n",
    "#                 is_match = observed_nucleotide == last(terminal_kmer)\n",
    "#                 if is_match\n",
    "#                     path_likelihood *= 1 - observed_error_rate\n",
    "#                 else\n",
    "#                     path_likelihood *= observed_error_rate\n",
    "#                 end\n",
    "#             end\n",
    "#         end\n",
    "\n",
    "#         if path_likelihood > state_likelihoods[current_vertex, current_state]\n",
    "# #             @show \"selecting path\"\n",
    "# #             @show path\n",
    "# #             @show path_likelihood\n",
    "#             state_likelihoods[current_vertex, current_state] = path_likelihood\n",
    "#             arrival_paths[current_vertex, current_state] = path\n",
    "#         end\n",
    "#     end\n",
    "#     return\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4d91e9c-b12c-43f2-9681-3a85edd018ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function polish_fastq(kmer_graph, fastq_file)\n",
    "\n",
    "# #     @info \"Assessing kmer likelihoods\"\n",
    "#     kmers = [kmer_graph.vprops[v][:kmer] for v in Graphs.vertices(kmer_graph)]\n",
    "# #     kmer_counts = [length(kmer_graph.vprops[v][:evidence]) for v in Graphs.vertices(kmer_graph)]\n",
    "#     kmer_counts = [kmer_graph.vprops[v][:weight] for v in Graphs.vertices(kmer_graph)]\n",
    "#     kmer_likelihoods = kmer_counts ./ sum(kmer_counts)\n",
    "#     k = kmer_graph.gprops[:k]\n",
    "#     kmer_type = BioSequences.BigDNAMer{k}\n",
    "#     total_kmers = length(kmers)\n",
    "    \n",
    "# #     @info \"determining shortest paths between kmers\"\n",
    "#     shortest_paths = Graphs.enumerate_paths(Graphs.floyd_warshall_shortest_paths(kmer_graph));\n",
    "    \n",
    "#     @info \"counting the number of records to establish runtime estimate\"\n",
    "#     number_of_records = 0\n",
    "#     for fastq_record in FASTX.FASTQ.Reader(open(fastq_file))\n",
    "#         number_of_records += 1\n",
    "#     end\n",
    "#     progress_bar = ProgressMeter.Progress(number_of_records, 1)\n",
    "    \n",
    "#     output_fastq_file = replace(fastq_file, \".fastq\" => \".k$(kmer_graph.gprops[:k]).fastq\")\n",
    "#     fastq_writer = FASTX.FASTQ.Writer(open(output_fastq_file, \"w\"))\n",
    "#     for fastq_record in FASTX.FASTQ.Reader(open(fastq_file))\n",
    "#         ProgressMeter.next!(progress_bar)\n",
    "        \n",
    "# #         @info \"Initializing matrices\"\n",
    "#         total_states = length(FASTX.sequence(fastq_record))-k+1\n",
    "#         transition_likelihoods = initialize_transition_probabilities(kmer_graph)\n",
    "#         state_likelihoods = zeros(total_kmers, total_states)\n",
    "#         arrival_paths = fill(Pair{Int, Union{Bool, Missing}}[], total_kmers, total_states)\n",
    "\n",
    "# #         @info \"Determining Likelihoods of initial states\"\n",
    "#         initial_state = first(BioSequences.each(kmer_type, FASTX.sequence(fastq_record)))\n",
    "#         current_state = 1\n",
    "#         # note this is a place for potential improvement, use the q value at each base to guide probability rather than median\n",
    "#         median_q_value = Statistics.median(Int.(FASTX.quality(fastq_record)[1:k]))\n",
    "#         current_error_rate = q_value_to_error_rate(median_q_value)\n",
    "#         # canonical_kmer = BioSequences.canonical(initial_state.fw)\n",
    "#         set_initial_state_likelihoods!(\n",
    "#                 kmer_graph,\n",
    "#                 initial_state,\n",
    "#                 kmer_likelihoods,\n",
    "#                 current_error_rate,\n",
    "#                 state_likelihoods,\n",
    "#                 arrival_paths\n",
    "#             )\n",
    "\n",
    "# #         @info \"Determining likelihood of downstream states\"\n",
    "\n",
    "# #         non_singleton_states = findall(kmer_counts .> 1)\n",
    "\n",
    "#         for current_state in 2:total_states\n",
    "#             prior_state = current_state - 1\n",
    "\n",
    "#         #     observed_kmer = BioSequences.BigDNAMer{k}(FASTX.sequence(fastq_record)[current_state:current_state+k-1])\n",
    "\n",
    "#         #     @assert observed_kmer == collect(BioSequences.each(kmer_type, FASTX.sequence(fastq_record)))[current_state].fw\n",
    "\n",
    "#         #     canonical_kmer = BioSequences.canonical(observed_kmer)\n",
    "\n",
    "#             observed_nucleotide = FASTX.sequence(fastq_record)[k-1+current_state]\n",
    "#         #     observed_nucleotide = last(observed_kmer)\n",
    "#             observed_quality_score = FASTX.quality(fastq_record)[k-1+current_state]\n",
    "#             observed_error_rate = q_value_to_error_rate(observed_quality_score)\n",
    "\n",
    "#             # we'll assess prior states in order of decreasing likelihood\n",
    "#             # such that we maximize how frequently we are able to utilize the\n",
    "#             # current_state_likelihood > candidate prior state\n",
    "#             # break that won't bother evaluating lower likelihood possibilities\n",
    "#             prior_states_in_decreasing_likelihood = sortperm(state_likelihoods[:, prior_state], rev=true)\n",
    "\n",
    "#             # and skip all prior states with zero probability\n",
    "\n",
    "#             for current_vertex in total_states\n",
    "#                 for prior_vertex in prior_states_in_decreasing_likelihood\n",
    "#                     if state_likelihoods[prior_vertex, prior_state] > 0\n",
    "#                         run_viterbi!(\n",
    "#                                 current_state,\n",
    "#                                 prior_state,\n",
    "#                                 observed_nucleotide,\n",
    "#                                 observed_quality_score,\n",
    "#                                 observed_error_rate,\n",
    "#                                 current_vertex,\n",
    "#                                 prior_vertex,\n",
    "#                                 state_likelihoods,\n",
    "#                                 transition_likelihoods,\n",
    "#                                 shortest_paths,\n",
    "#                                 arrival_paths,\n",
    "#                                 kmer_graph,\n",
    "#                                 kmer_likelihoods\n",
    "#                                 )\n",
    "#                     end\n",
    "#                 end\n",
    "#             end\n",
    "#         end\n",
    "\n",
    "# #         try\n",
    "#         maximum_likelihood_path, maximum_likelihood_value = \n",
    "#             determine_maximum_likelihood_path(\n",
    "#                 state_likelihoods,\n",
    "#                 arrival_paths\n",
    "#                 )\n",
    "# #         catch\n",
    "# #             return state_likelihoods, arrival_paths\n",
    "# #         end\n",
    "\n",
    "#         sequence = oriented_path_to_sequence(kmer_graph, maximum_likelihood_path)\n",
    "\n",
    "# #         @info \"comparing to original path\"\n",
    "#         original_sequence_likelihood = oriented_path_to_likelihood(kmer_graph, kmers, kmer_likelihoods, transition_likelihoods, fastq_record)\n",
    "#         relative_likelihood = maximum_likelihood_value / original_sequence_likelihood\n",
    "# #         relative_likelihood_formatted = NumericIO.formatted(relative_likelihood, ndigits=1, charset=:ASCII)\n",
    "# #         println(\"relative likelihood of new path to old path is $(relative_likelihood_formatted)\")\n",
    "\n",
    "# #         @info \"writing updated record\"\n",
    "#         identifier = FASTX.identifier(fastq_record) * \"_k$(k)\"\n",
    "#         description = string(relative_likelihood)\n",
    "#         # because the sequences won't always be the same length, we take an ordered sampling with replacement\n",
    "#         # which introduces some random error but preserves overall patterns and areas of high/low accuracy\n",
    "#         quality_scores = StatsBase.sample(FASTX.quality(fastq_record), length(sequence), ordered=true)\n",
    "\n",
    "#         new_fastq_record = FASTX.FASTQ.Record(\n",
    "#             identifier,\n",
    "#             description,\n",
    "#             sequence,\n",
    "#             quality_scores\n",
    "#         )\n",
    "#         write(fastq_writer, new_fastq_record)\n",
    "#     end\n",
    "#     close(fastq_writer)\n",
    "#     return output_fastq_file\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58463f07-3726-4716-b8ae-b1fca9d0f5e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function determine_maximum_likelihood_path(\n",
    "#     state_likelihoods,\n",
    "#     arrival_paths\n",
    "#     )\n",
    "#     maximum_likelihood_value = maximum(state_likelihoods[:, end])\n",
    "\n",
    "#     maximum_likelihood_path_indices = findall(state_likelihoods[:, end] .== maximum_likelihood_value)\n",
    "\n",
    "#     # if multiple paths are tied, randomly choose one\n",
    "#     maximum_likelihood_path_index = rand(maximum_likelihood_path_indices)\n",
    "\n",
    "#     maximum_likelihood_path = arrival_paths[maximum_likelihood_path_index, end]\n",
    "\n",
    "#     for state_index in size(state_likelihoods, 2)-1:-1:1\n",
    "#         next_kmer, next_orientation = first(maximum_likelihood_path)\n",
    "#         maximum_likelihood_arrival_path = arrival_paths[next_kmer, state_index]\n",
    "        \n",
    "#         is_match = last(maximum_likelihood_arrival_path) == (next_kmer => next_orientation)\n",
    "#         if !ismissing(is_match) && !is_match\n",
    "#             error(\"breaking\")\n",
    "#         end\n",
    "#         maximum_likelihood_path = vcat(maximum_likelihood_arrival_path[1:end-1], maximum_likelihood_path)\n",
    "#     end\n",
    "#     return maximum_likelihood_path, maximum_likelihood_value\n",
    "# end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e681bdd-4799-4687-a81f-04bc51422310",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Mycelia.polish_fastq(graph, fastq)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.7",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
