{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335bf35f-c55e-4857-83f0-ed03dc666f4e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# if hit plotting library issues, try resetting LD path for julia\n",
    "# can set in ~/.local/share/jupyter/kernels/\n",
    "# @assert ENV[\"LD_LIBRARY_PATH\"] == \"\"\n",
    "import Pkg\n",
    "pkgs = [\n",
    "    \"Revise\",\n",
    "    \"FASTX\",\n",
    "    \"VariantCallFormat\",\n",
    "    \"StatsBase\",\n",
    "    \"Distributions\",\n",
    "    \"StatsPlots\",\n",
    "    \"Random\",\n",
    "    \"Dates\",\n",
    "    \"DataFrames\",\n",
    "    \"BioSequences\"\n",
    "]\n",
    "# Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "# Pkg.develop(path=\"/global/cfs/projectdirs/m4269/cjprybol/Mycelia\")\n",
    "# Pkg.develop(path=\"../../..\")\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "128cc1e1-929a-42e1-8376-4c81832fd40c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "PROJECT_BASEDIR = dirname(pwd())\n",
    "data_dir = joinpath(PROJECT_BASEDIR, \"data\")\n",
    "genome_dir = mkpath(joinpath(data_dir, \"genomes\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e81dfc48-a014-4b52-b6b5-8b695eff7b18",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# seed = parse(Int, replace(string(Dates.today()), \"-\" => \"\"))\n",
    "seed = 20240121\n",
    "genome_size = 1_000_000\n",
    "n_variants = 1000\n",
    "window_size = Int(round(genome_size / n_variants))\n",
    "Random.seed!(seed)\n",
    "\n",
    "# make the original genome, or read it in\n",
    "original_record = Mycelia.random_fasta_record(seed=seed, L=genome_size)\n",
    "reference_id = FASTX.description(original_record)\n",
    "fasta_file = joinpath(genome_dir, reference_id * \".fna\")\n",
    "vcf_file = fasta_file * \".vcf\"\n",
    "modified_fasta_file = vcf_file * \".fna\"\n",
    "Mycelia.write_fasta(outfile = fasta_file, records = [original_record])\n",
    "\n",
    "original_sequence = FASTX.sequence(original_record)\n",
    "\n",
    "# chosen empirically to find the right distribution size\n",
    "n = Base.MathConstants.e^4 # ~ 1-500\n",
    "variant_size_disbribution = Distributions.Geometric(1/n)\n",
    "p = StatsPlots.plot(\n",
    "    variant_size_disbribution,\n",
    "    ylabel = \"probability density\",\n",
    "    xlabel = \"variant size\",\n",
    "    title = \"Geometric Distriubtion (1/â„¯^4)\",\n",
    "    legend=false\n",
    ")\n",
    "display(p)\n",
    "variant_sizes = rand(variant_size_disbribution, 1000) .+ 1\n",
    "p = StatsPlots.histogram(\n",
    "    variant_sizes,\n",
    "    ylabel = \"# of variants\",\n",
    "    xlabel = \"variant size\",\n",
    "    title = \"Actual samples drawn\",\n",
    "    nbins = length(unique(variant_sizes)),\n",
    "    legend=false\n",
    ")\n",
    "display(p)\n",
    "\n",
    "@assert all(variant_sizes .>= 1)\n",
    "\n",
    "variant_type_likelihoods = [\n",
    "    :substitution => 10^-1,\n",
    "    :insertion => 10^-2,\n",
    "    :deletion => 10^-2,\n",
    "    :inversion => 10^-3,\n",
    "    # special case insertion/deletions, skipping\n",
    "    # :translocations => 10^-3,\n",
    "    # :duplication => 10^-3,\n",
    "]\n",
    "\n",
    "# if we sample 1k variants, we generally hit a few or more of each\n",
    "variant_types = StatsBase.sample(first.(variant_type_likelihoods), StatsBase.weights(last.(variant_type_likelihoods)), 1000)\n",
    "StatsBase.countmap(variant_types)\n",
    "\n",
    "# we'll use 1k windows of 1k basepairs\n",
    "step = 1000\n",
    "window_starts = 1:step:length(original_sequence)\n",
    "window_ends = step:step:length(original_sequence)\n",
    "windows = zip(window_starts, window_ends)\n",
    "variant_type_sizes = zip(variant_types, variant_sizes)\n",
    "\n",
    "vcf_table = DataFrames.DataFrame(\n",
    "    \"#CHROM\" => String[],\n",
    "    \"POS\" => Int[],\n",
    "    \"ID\" => String[],\n",
    "    \"REF\" => String[],\n",
    "    \"ALT\" => String[],\n",
    "    \"QUAL\" => Int[],\n",
    "    \"FILTER\" => String[],\n",
    "    \"INFO\" => String[],\n",
    "    \"FORMAT\" => String[]\n",
    ")\n",
    "\n",
    "for ((variant_type, variant_size), (start, stop)) in collect(zip(variant_type_sizes, windows))\n",
    "    selected_start = rand(start:stop-variant_size)\n",
    "    @assert selected_start <= stop\n",
    "    original_subsequence = original_sequence[selected_start:selected_start+variant_size-1]\n",
    "    @assert length(original_subsequence) == variant_size\n",
    "    if variant_type == :substitution\n",
    "        ref = original_subsequence\n",
    "        alt = BioSequences.randdnaseq(variant_size)\n",
    "    elseif variant_type == :insertion\n",
    "        ref = \".\"\n",
    "        alt = BioSequences.randdnaseq(variant_size)\n",
    "    elseif variant_type == :deletion\n",
    "        ref = original_subsequence\n",
    "        alt = \".\"\n",
    "    elseif variant_type == :inversion\n",
    "        ref = original_subsequence\n",
    "        alt = reverse(original_subsequence)\n",
    "    end \n",
    "    # @show selected_start\n",
    "    row = Dict(\n",
    "        \"#CHROM\" => reference_id,\n",
    "        \"POS\" => selected_start,\n",
    "        \"ID\" => \".\",\n",
    "        \"REF\" => string(ref),\n",
    "        \"ALT\" => string(alt),\n",
    "        \"QUAL\" => 60,\n",
    "        \"FILTER\" => string(variant_type),\n",
    "        \"INFO\" => \"\",\n",
    "        \"FORMAT\" => \"\"\n",
    "    )\n",
    "    push!(vcf_table, row)\n",
    "end\n",
    "vcf_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "887bf591-0865-4f18-bdd9-6af03d78d34e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "open(vcf_file, \"w\") do io\n",
    "    VCF_HEADER = \n",
    "    \"\"\"\n",
    "    ##fileformat=VCFv4.3\n",
    "    ##fileDate=$(Dates.today())\n",
    "    ##source=simulated-variants\n",
    "    ##reference=$(reference_id)\n",
    "    \"\"\"\n",
    "    print(io, VCF_HEADER)\n",
    "    println(io, join(names(vcf_table), '\\t'))\n",
    "    for row in DataFrames.eachrow(vcf_table)\n",
    "        println(io, join([row[col] for col in names(vcf_table)], '\\t'))\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f607b4c5-1878-4790-b618-fda4b585fcdb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "run(`conda create -c conda-forge -c bioconda -c defaults --strict-channel-priority -n htslib htslib -y`)\n",
    "run(`conda create -c conda-forge -c bioconda -c defaults --strict-channel-priority -n tabix tabix -y`)\n",
    "run(`conda create -c conda-forge -c bioconda -c defaults --strict-channel-priority -n bcftools bcftools -y`)\n",
    "run(`conda create -c conda-forge -c bioconda -c defaults --strict-channel-priority -n vcftools vcftools -y`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99910ad2-de11-4afb-893f-71555f112bcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#     !mamba run --live-stream --no-capture-output -n htslib bgzip {short_read_vcf}.filtered.vcf\n",
    "#     !mamba run --live-stream --no-capture-output -n tabix tabix -p vcf {short_read_vcf}.filtered.vcf.gz\n",
    "#     !mamba run --live-stream --no-capture-output -n bcftools bcftools norm --fasta-ref {local_outdir}/flye/medaka/consensus.fasta {short_read_vcf}.filtered.vcf.gz > {short_read_vcf}.filtered.normalized.vcf\n",
    "\n",
    "#     !mamba run --live-stream --no-capture-output -n htslib bgzip {short_read_vcf}.filtered.normalized.vcf\n",
    "#     !mamba run --live-stream --no-capture-output -n tabix tabix -p vcf {short_read_vcf}.filtered.normalized.vcf.gz\n",
    "#     !mamba run --live-stream --no-capture-output -n bcftools bcftools consensus -f {local_outdir}/flye/medaka/consen"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.9.4",
   "language": "julia",
   "name": "julia-1.9.4"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
