<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Tutorial 7: Comparative Genomics and Pangenome Analysis · Mycelia</title><meta name="title" content="Tutorial 7: Comparative Genomics and Pangenome Analysis · Mycelia"/><meta property="og:title" content="Tutorial 7: Comparative Genomics and Pangenome Analysis · Mycelia"/><meta property="twitter:title" content="Tutorial 7: Comparative Genomics and Pangenome Analysis · Mycelia"/><meta name="description" content="Documentation for Mycelia."/><meta property="og:description" content="Documentation for Mycelia."/><meta property="twitter:description" content="Documentation for Mycelia."/><script data-outdated-warner src="../../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../search_index.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../../">Mycelia</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../../">Home</a></li><li><a class="tocitem" href="../../../installation/">Installation</a></li><li><a class="tocitem" href="../../../getting-started/">Getting Started</a></li><li><a class="tocitem" href="../../../concepts/">Concepts</a></li><li><a class="tocitem" href="../../../probabilistic-assembly-hub/">Probabilistic Assembly</a></li><li><span class="tocitem">Documentation</span><ul><li><a class="tocitem" href="../../../architecture/">Architecture Overview</a></li><li><a class="tocitem" href="../../../assembly-method-selection/">Assembly Method Selection</a></li><li><a class="tocitem" href="../../../performance/">Performance Guide</a></li><li><a class="tocitem" href="../../../faq/">FAQ</a></li></ul></li><li><span class="tocitem">API Reference</span><ul><li><a class="tocitem" href="../../../api-reference/">Complete API Reference</a></li><li><input class="collapse-toggle" id="menuitem-7-2" type="checkbox"/><label class="tocitem" for="menuitem-7-2"><span class="docs-label">Workflows</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../api/workflows/assembly-suite/">Assembly Suite</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-7-3" type="checkbox"/><label class="tocitem" for="menuitem-7-3"><span class="docs-label">Quick Reference</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../api/quick-reference/parameter-guide/">Parameter Guide</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-7-4" type="checkbox"/><label class="tocitem" for="menuitem-7-4"><span class="docs-label">Examples</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../../api/examples/basic-workflows/">Basic Workflows</a></li></ul></li><li><a class="tocitem" href="../../../api/">Legacy API Documentation</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Tutorial 7: Comparative Genomics and Pangenome Analysis</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Tutorial 7: Comparative Genomics and Pangenome Analysis</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/cjprybol/Mycelia/blob/master/tutorials/07_comparative_genomics.jl#L" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Tutorial-7:-Comparative-Genomics-and-Pangenome-Analysis"><a class="docs-heading-anchor" href="#Tutorial-7:-Comparative-Genomics-and-Pangenome-Analysis">Tutorial 7: Comparative Genomics and Pangenome Analysis</a><a id="Tutorial-7:-Comparative-Genomics-and-Pangenome-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial-7:-Comparative-Genomics-and-Pangenome-Analysis" title="Permalink"></a></h1><p>This tutorial covers comparative genomics approaches, including pangenome construction, phylogenetic analysis, and evolutionary genomics.</p><h2 id="Learning-Objectives"><a class="docs-heading-anchor" href="#Learning-Objectives">Learning Objectives</a><a id="Learning-Objectives-1"></a><a class="docs-heading-anchor-permalink" href="#Learning-Objectives" title="Permalink"></a></h2><p>By the end of this tutorial, you will understand:</p><ul><li>Pangenome concepts and construction methods</li><li>Core, accessory, and unique gene identification</li><li>Phylogenetic tree construction and interpretation</li><li>Synteny analysis and chromosomal rearrangements</li><li>Population genomics and evolutionary analysis</li><li>Graph-based genome representation</li></ul><h2 id="Setup"><a class="docs-heading-anchor" href="#Setup">Setup</a><a id="Setup-1"></a><a class="docs-heading-anchor-permalink" href="#Setup" title="Permalink"></a></h2><pre><code class="language-julia hljs">import Pkg
if isinteractive()
    Pkg.activate(&quot;..&quot;)
end

import Test
import Mycelia
import FASTX
import Random
import Statistics
import Graphs
import Kmers

Random.seed!(42)</code></pre><h2 id="Part-1:-Pangenome-Concepts"><a class="docs-heading-anchor" href="#Part-1:-Pangenome-Concepts">Part 1: Pangenome Concepts</a><a id="Part-1:-Pangenome-Concepts-1"></a><a class="docs-heading-anchor-permalink" href="#Part-1:-Pangenome-Concepts" title="Permalink"></a></h2><p>Pangenomes represent the complete set of genes within a species or closely related group of organisms.</p><pre><code class="language-julia hljs">println(&quot;=== Comparative Genomics Tutorial ===&quot;)

println(&quot;Pangenome Components:&quot;)
println(&quot;- Core genome: Genes present in all individuals&quot;)
println(&quot;- Accessory genome: Genes present in some individuals&quot;)
println(&quot;- Unique genes: Genes present in single individuals&quot;)
println(&quot;- Variable genes: Genes with presence/absence variation&quot;)
println()
println(&quot;Pangenome Types:&quot;)
println(&quot;- Open pangenome: Continues to expand with new genomes&quot;)
println(&quot;- Closed pangenome: Reaches saturation quickly&quot;)
println(&quot;- Mixed pangenome: Shows both open and closed characteristics&quot;)</code></pre><h2 id="Part-2:-Data-Preparation-for-Pangenome-Analysis"><a class="docs-heading-anchor" href="#Part-2:-Data-Preparation-for-Pangenome-Analysis">Part 2: Data Preparation for Pangenome Analysis</a><a id="Part-2:-Data-Preparation-for-Pangenome-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Part-2:-Data-Preparation-for-Pangenome-Analysis" title="Permalink"></a></h2><p>Prepare multiple genomes for comparative analysis</p><pre><code class="language-julia hljs">println(&quot;\n=== Data Preparation ===&quot;)</code></pre><h3 id="Simulating-Multiple-Genomes"><a class="docs-heading-anchor" href="#Simulating-Multiple-Genomes">Simulating Multiple Genomes</a><a id="Simulating-Multiple-Genomes-1"></a><a class="docs-heading-anchor-permalink" href="#Simulating-Multiple-Genomes" title="Permalink"></a></h3><p>Create a set of related genomes for pangenome analysis</p><pre><code class="language-julia hljs">println(&quot;--- Simulating Related Genomes ---&quot;)</code></pre><p>Generate core genes present in all genomes</p><pre><code class="language-julia hljs">n_genomes = 5
core_genes = 200
accessory_genes = 100
unique_genes = 50</code></pre><p>Generate simulated genomes with different characteristics</p><pre><code class="language-julia hljs">genome_files = String[]
for i in 1:n_genomes
    genome_size = core_genes * 1000 + rand(0:accessory_genes) * 1000 + rand(0:unique_genes) * 1000
    genome = Mycelia.random_fasta_record(moltype=:DNA, seed=i, L=genome_size)
    filename = &quot;genome_$i.fasta&quot;
    Mycelia.write_fasta(outfile=filename, records=[genome])
    push!(genome_files, filename)
end

println(&quot;Generated $(n_genomes) genomes:&quot;)
for (i, file) in enumerate(genome_files)
    size = filesize(file)
    println(&quot;  Genome $i: $file ($(size) bytes)&quot;)
end</code></pre><h2 id="Part-3:-Gene-Clustering-and-Ortholog-Identification"><a class="docs-heading-anchor" href="#Part-3:-Gene-Clustering-and-Ortholog-Identification">Part 3: Gene Clustering and Ortholog Identification</a><a id="Part-3:-Gene-Clustering-and-Ortholog-Identification-1"></a><a class="docs-heading-anchor-permalink" href="#Part-3:-Gene-Clustering-and-Ortholog-Identification" title="Permalink"></a></h2><p>Identify orthologous genes across genomes</p><pre><code class="language-julia hljs">println(&quot;\n=== Gene Clustering ===&quot;)</code></pre><h3 id="All-vs-All-Sequence-Comparison"><a class="docs-heading-anchor" href="#All-vs-All-Sequence-Comparison">All-vs-All Sequence Comparison</a><a id="All-vs-All-Sequence-Comparison-1"></a><a class="docs-heading-anchor-permalink" href="#All-vs-All-Sequence-Comparison" title="Permalink"></a></h3><p>Compare all genes against all genes</p><pre><code class="language-julia hljs">println(&quot;--- All-vs-All Comparison ---&quot;)</code></pre><p>Perform k-mer based pangenome analysis using implemented functions</p><pre><code class="language-julia hljs">println(&quot;Performing k-mer based pangenome analysis...&quot;)</code></pre><p>Use implemented pangenome analysis function</p><pre><code class="language-julia hljs">pangenome_result = Mycelia.analyze_pangenome_kmers(
    genome_files,
    kmer_type=Kmers.DNAKmer{21}
)

println(&quot;Pangenome Analysis Results:&quot;)
println(&quot;  Core k-mers: $(length(pangenome_result.core_kmers))&quot;)
println(&quot;  Accessory k-mers: $(length(pangenome_result.accessory_kmers))&quot;)
println(&quot;  Total pangenome size: $(pangenome_result.similarity_stats.pangenome_size)&quot;)</code></pre><h3 id="Genome-Similarity-Analysis"><a class="docs-heading-anchor" href="#Genome-Similarity-Analysis">Genome Similarity Analysis</a><a id="Genome-Similarity-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Genome-Similarity-Analysis" title="Permalink"></a></h3><p>Compare genome similarity using multiple metrics</p><pre><code class="language-julia hljs">println(&quot;--- Genome Similarity Analysis ---&quot;)</code></pre><p>Build distance matrix using different metrics</p><pre><code class="language-julia hljs">js_matrix = Mycelia.build_genome_distance_matrix(
    genome_files,
    kmer_type=Kmers.DNAKmer{21},
    metric=:js_divergence
)

println(&quot;Distance Matrix (JS Divergence):&quot;)
for i in 1:n_genomes
    for j in 1:n_genomes
        print(&quot;$(round(js_matrix.distance_matrix[i,j], digits=3)) &quot;)
    end
    println()
end</code></pre><h2 id="Part-4:-Pangenome-Construction"><a class="docs-heading-anchor" href="#Part-4:-Pangenome-Construction">Part 4: Pangenome Construction</a><a id="Part-4:-Pangenome-Construction-1"></a><a class="docs-heading-anchor-permalink" href="#Part-4:-Pangenome-Construction" title="Permalink"></a></h2><p>Build pangenome from ortholog clusters</p><pre><code class="language-julia hljs">println(&quot;\n=== Pangenome Construction ===&quot;)</code></pre><h3 id="Core-Genome-Analysis"><a class="docs-heading-anchor" href="#Core-Genome-Analysis">Core Genome Analysis</a><a id="Core-Genome-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Genome-Analysis" title="Permalink"></a></h3><p>Identify genes present in all genomes</p><pre><code class="language-julia hljs">println(&quot;--- Core Genome Analysis ---&quot;)</code></pre><p>Use results from pangenome analysis</p><pre><code class="language-julia hljs">core_genome_size = length(pangenome_result.core_kmers)
core_genome_percentage = pangenome_result.similarity_stats.core_percentage

println(&quot;Core Genome:&quot;)
println(&quot;  Size: $core_genome_size k-mers&quot;)
println(&quot;  Percentage: $(round(core_genome_percentage, digits=1))%&quot;)</code></pre><p>Core genome represents k-mers present in ALL genomes</p><pre><code class="language-julia hljs">println(&quot;Core genome represents conserved genomic content across all $(n_genomes) genomes&quot;)</code></pre><h3 id="Accessory-Genome-Analysis"><a class="docs-heading-anchor" href="#Accessory-Genome-Analysis">Accessory Genome Analysis</a><a id="Accessory-Genome-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Accessory-Genome-Analysis" title="Permalink"></a></h3><p>Analyze k-mers with variable presence</p><pre><code class="language-julia hljs">println(&quot;--- Accessory Genome Analysis ---&quot;)

accessory_genome_size = length(pangenome_result.accessory_kmers)
unique_genome_total = sum(length(kmers) for kmers in values(pangenome_result.unique_kmers_by_genome))

println(&quot;Accessory Genome:&quot;)
println(&quot;  Variable k-mers: $accessory_genome_size&quot;)
println(&quot;  Unique k-mers: $unique_genome_total&quot;)
println(&quot;  Total accessory: $(accessory_genome_size + unique_genome_total)&quot;)</code></pre><p>Analyze unique k-mers per genome</p><pre><code class="language-julia hljs">println(&quot;Unique k-mers per genome:&quot;)
for (genome, unique_kmers) in pangenome_result.unique_kmers_by_genome
    println(&quot;  $genome: $(length(unique_kmers)) unique k-mers&quot;)
end</code></pre><h3 id="Pangenome-Curves"><a class="docs-heading-anchor" href="#Pangenome-Curves">Pangenome Curves</a><a id="Pangenome-Curves-1"></a><a class="docs-heading-anchor-permalink" href="#Pangenome-Curves" title="Permalink"></a></h3><p>Model pangenome size vs number of genomes</p><pre><code class="language-julia hljs">println(&quot;--- Pangenome Curves ---&quot;)</code></pre><p>TODO: Implement pangenome curve analysis</p><ul><li>Calculate pangenome size for subsets</li><li>Fit mathematical models (power law, exponential)</li><li>Predict pangenome size for larger collections</li><li>Classify as open/closed pangenome</li></ul><h2 id="Part-5:-Graph-Based-Pangenome-Representation"><a class="docs-heading-anchor" href="#Part-5:-Graph-Based-Pangenome-Representation">Part 5: Graph-Based Pangenome Representation</a><a id="Part-5:-Graph-Based-Pangenome-Representation-1"></a><a class="docs-heading-anchor-permalink" href="#Part-5:-Graph-Based-Pangenome-Representation" title="Permalink"></a></h2><p>Represent pangenomes as graphs</p><pre><code class="language-julia hljs">println(&quot;\n=== Graph-Based Pangenomes ===&quot;)</code></pre><h3 id="Sequence-Graphs"><a class="docs-heading-anchor" href="#Sequence-Graphs">Sequence Graphs</a><a id="Sequence-Graphs-1"></a><a class="docs-heading-anchor-permalink" href="#Sequence-Graphs" title="Permalink"></a></h3><p>Build graphs from sequence overlaps</p><pre><code class="language-julia hljs">println(&quot;--- Sequence Graphs ---&quot;)</code></pre><p>TODO: Implement sequence graph construction</p><ul><li>Build k-mer graphs from sequences</li><li>Identify shared and unique paths</li><li>Handle graph bubbles and loops</li><li>Compress redundant sequences</li></ul><p>Use existing pangenome graph construction</p><pre><code class="language-julia hljs">k = 3
sequences = [FASTX.sequence(Mycelia.random_fasta_record(moltype=:DNA, seed=i, L=100)) for i in 1:5]</code></pre><p>graph = Mycelia.build<em>stranded</em>kmer_graph(Kmers.DNAKmer{k}, sequences)</p><pre><code class="language-julia hljs">println(&quot;Sequence Graph Construction:&quot;)
println(&quot;  K-mer size: $k&quot;)
println(&quot;  Sequences: $(length(sequences))&quot;)</code></pre><p>println(&quot;  Graph nodes: Graphs.nv(graph)&quot;) println(&quot;  Graph edges: Graphs.ne(graph)&quot;)</p><h3 id="Variation-Graphs"><a class="docs-heading-anchor" href="#Variation-Graphs">Variation Graphs</a><a id="Variation-Graphs-1"></a><a class="docs-heading-anchor-permalink" href="#Variation-Graphs" title="Permalink"></a></h3><p>Represent genetic variation as graphs</p><pre><code class="language-julia hljs">println(&quot;--- Variation Graphs ---&quot;)</code></pre><p>TODO: Implement variation graph construction</p><ul><li>Build graphs from multiple alignments</li><li>Represent SNPs and indels as bubbles</li><li>Handle complex structural variants</li><li>Optimize graph topology</li></ul><h2 id="Part-6:-Phylogenetic-Analysis"><a class="docs-heading-anchor" href="#Part-6:-Phylogenetic-Analysis">Part 6: Phylogenetic Analysis</a><a id="Part-6:-Phylogenetic-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Part-6:-Phylogenetic-Analysis" title="Permalink"></a></h2><p>Construct phylogenetic trees from pangenome data</p><pre><code class="language-julia hljs">println(&quot;\n=== Phylogenetic Analysis ===&quot;)</code></pre><h3 id="Core-Genome-Phylogeny"><a class="docs-heading-anchor" href="#Core-Genome-Phylogeny">Core Genome Phylogeny</a><a id="Core-Genome-Phylogeny-1"></a><a class="docs-heading-anchor-permalink" href="#Core-Genome-Phylogeny" title="Permalink"></a></h3><p>Build trees from core genes</p><pre><code class="language-julia hljs">println(&quot;--- Core Genome Phylogeny ---&quot;)</code></pre><p>TODO: Implement core genome phylogeny</p><ul><li>Concatenate core gene alignments</li><li>Calculate phylogenetic trees</li><li>Assess branch support</li><li>Compare with individual gene trees</li></ul><h3 id="Accessory-Genome-Phylogeny"><a class="docs-heading-anchor" href="#Accessory-Genome-Phylogeny">Accessory Genome Phylogeny</a><a id="Accessory-Genome-Phylogeny-1"></a><a class="docs-heading-anchor-permalink" href="#Accessory-Genome-Phylogeny" title="Permalink"></a></h3><p>Analyze evolution of accessory genes</p><pre><code class="language-julia hljs">println(&quot;--- Accessory Genome Phylogeny ---&quot;)</code></pre><p>TODO: Implement accessory genome phylogeny</p><ul><li>Build trees from gene presence/absence</li><li>Use binary character evolution models</li><li>Identify horizontal gene transfer events</li><li>Analyze gene gain/loss patterns</li></ul><h2 id="Part-7:-Synteny-Analysis"><a class="docs-heading-anchor" href="#Part-7:-Synteny-Analysis">Part 7: Synteny Analysis</a><a id="Part-7:-Synteny-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Part-7:-Synteny-Analysis" title="Permalink"></a></h2><p>Analyze conserved gene order</p><pre><code class="language-julia hljs">println(&quot;\n=== Synteny Analysis ===&quot;)</code></pre><h3 id="Synteny-Detection"><a class="docs-heading-anchor" href="#Synteny-Detection">Synteny Detection</a><a id="Synteny-Detection-1"></a><a class="docs-heading-anchor-permalink" href="#Synteny-Detection" title="Permalink"></a></h3><p>Identify conserved gene order</p><pre><code class="language-julia hljs">println(&quot;--- Synteny Detection ---&quot;)</code></pre><p>TODO: Implement synteny detection</p><ul><li>Identify colinear gene blocks</li><li>Calculate synteny conservation</li><li>Detect chromosomal rearrangements</li><li>Visualize synteny relationships</li></ul><h3 id="Rearrangement-Analysis"><a class="docs-heading-anchor" href="#Rearrangement-Analysis">Rearrangement Analysis</a><a id="Rearrangement-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Rearrangement-Analysis" title="Permalink"></a></h3><p>Analyze chromosomal rearrangements</p><pre><code class="language-julia hljs">println(&quot;--- Rearrangement Analysis ---&quot;)</code></pre><p>TODO: Implement rearrangement analysis</p><ul><li>Identify inversions, translocations, duplications</li><li>Calculate rearrangement distances</li><li>Reconstruct ancestral gene orders</li><li>Analyze rearrangement hotspots</li></ul><h2 id="Part-8:-Population-Genomics"><a class="docs-heading-anchor" href="#Part-8:-Population-Genomics">Part 8: Population Genomics</a><a id="Part-8:-Population-Genomics-1"></a><a class="docs-heading-anchor-permalink" href="#Part-8:-Population-Genomics" title="Permalink"></a></h2><p>Analyze genetic diversity within populations</p><pre><code class="language-julia hljs">println(&quot;\n=== Population Genomics ===&quot;)</code></pre><h3 id="Genetic-Diversity"><a class="docs-heading-anchor" href="#Genetic-Diversity">Genetic Diversity</a><a id="Genetic-Diversity-1"></a><a class="docs-heading-anchor-permalink" href="#Genetic-Diversity" title="Permalink"></a></h3><p>Calculate diversity metrics</p><pre><code class="language-julia hljs">println(&quot;--- Genetic Diversity ---&quot;)</code></pre><p>TODO: Implement diversity analysis</p><ul><li>Calculate nucleotide diversity (π)</li><li>Estimate effective population size</li><li>Analyze allele frequency spectra</li><li>Identify population structure</li></ul><h3 id="Selection-Analysis"><a class="docs-heading-anchor" href="#Selection-Analysis">Selection Analysis</a><a id="Selection-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Selection-Analysis" title="Permalink"></a></h3><p>Detect selection pressure</p><pre><code class="language-julia hljs">println(&quot;--- Selection Analysis ---&quot;)</code></pre><p>TODO: Implement selection analysis</p><ul><li>Calculate dN/dS ratios</li><li>Identify positively selected genes</li><li>Detect balancing selection</li><li>Analyze codon usage bias</li></ul><h2 id="Part-9:-Functional-Analysis"><a class="docs-heading-anchor" href="#Part-9:-Functional-Analysis">Part 9: Functional Analysis</a><a id="Part-9:-Functional-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Part-9:-Functional-Analysis" title="Permalink"></a></h2><p>Analyze functional aspects of pangenomes</p><pre><code class="language-julia hljs">println(&quot;\n=== Functional Analysis ===&quot;)</code></pre><h3 id="Functional-Enrichment"><a class="docs-heading-anchor" href="#Functional-Enrichment">Functional Enrichment</a><a id="Functional-Enrichment-1"></a><a class="docs-heading-anchor-permalink" href="#Functional-Enrichment" title="Permalink"></a></h3><p>Identify enriched functional categories</p><pre><code class="language-julia hljs">println(&quot;--- Functional Enrichment ---&quot;)</code></pre><p>TODO: Implement functional enrichment analysis</p><ul><li>GO term enrichment analysis</li><li>KEGG pathway enrichment</li><li>Protein domain analysis</li><li>Metabolic pathway reconstruction</li></ul><h3 id="Adaptive-Evolution"><a class="docs-heading-anchor" href="#Adaptive-Evolution">Adaptive Evolution</a><a id="Adaptive-Evolution-1"></a><a class="docs-heading-anchor-permalink" href="#Adaptive-Evolution" title="Permalink"></a></h3><p>Identify adaptively evolving genes</p><pre><code class="language-julia hljs">println(&quot;--- Adaptive Evolution ---&quot;)</code></pre><p>TODO: Implement adaptive evolution analysis</p><ul><li>Identify rapidly evolving genes</li><li>Analyze gene family expansions</li><li>Detect horizontal gene transfer</li><li>Correlate with environmental factors</li></ul><h2 id="Part-10:-Visualization-and-Interpretation"><a class="docs-heading-anchor" href="#Part-10:-Visualization-and-Interpretation">Part 10: Visualization and Interpretation</a><a id="Part-10:-Visualization-and-Interpretation-1"></a><a class="docs-heading-anchor-permalink" href="#Part-10:-Visualization-and-Interpretation" title="Permalink"></a></h2><p>Create visualizations for comparative genomics</p><pre><code class="language-julia hljs">println(&quot;\n=== Visualization ===&quot;)</code></pre><h3 id="Pangenome-Plots"><a class="docs-heading-anchor" href="#Pangenome-Plots">Pangenome Plots</a><a id="Pangenome-Plots-1"></a><a class="docs-heading-anchor-permalink" href="#Pangenome-Plots" title="Permalink"></a></h3><p>Visualize pangenome structure</p><pre><code class="language-julia hljs">println(&quot;--- Pangenome Plots ---&quot;)</code></pre><p>TODO: Implement pangenome visualization</p><ul><li>Pangenome accumulation curves</li><li>Gene presence/absence heatmaps</li><li>Phylogenetic trees with annotations</li><li>Synteny dot plots</li></ul><h3 id="Interactive-Visualization"><a class="docs-heading-anchor" href="#Interactive-Visualization">Interactive Visualization</a><a id="Interactive-Visualization-1"></a><a class="docs-heading-anchor-permalink" href="#Interactive-Visualization" title="Permalink"></a></h3><p>Create interactive pangenome browsers</p><pre><code class="language-julia hljs">println(&quot;--- Interactive Visualization ---&quot;)</code></pre><p>TODO: Implement interactive visualization</p><ul><li>Web-based pangenome browser</li><li>Interactive phylogenetic trees</li><li>Searchable gene catalogs</li><li>Comparative genome viewers</li></ul><h2 id="Part-11:-Applications-and-Case-Studies"><a class="docs-heading-anchor" href="#Part-11:-Applications-and-Case-Studies">Part 11: Applications and Case Studies</a><a id="Part-11:-Applications-and-Case-Studies-1"></a><a class="docs-heading-anchor-permalink" href="#Part-11:-Applications-and-Case-Studies" title="Permalink"></a></h2><p>Real-world applications of comparative genomics</p><pre><code class="language-julia hljs">println(&quot;\n=== Applications ===&quot;)</code></pre><h3 id="Pathogen-Genomics"><a class="docs-heading-anchor" href="#Pathogen-Genomics">Pathogen Genomics</a><a id="Pathogen-Genomics-1"></a><a class="docs-heading-anchor-permalink" href="#Pathogen-Genomics" title="Permalink"></a></h3><p>Applications in infectious disease research</p><pre><code class="language-julia hljs">println(&quot;--- Pathogen Genomics ---&quot;)

println(&quot;Pathogen Genomics Applications:&quot;)
println(&quot;- Outbreak investigation and source tracking&quot;)
println(&quot;- Drug resistance evolution&quot;)
println(&quot;- Vaccine target identification&quot;)
println(&quot;- Virulence factor discovery&quot;)</code></pre><h3 id="Agricultural-Genomics"><a class="docs-heading-anchor" href="#Agricultural-Genomics">Agricultural Genomics</a><a id="Agricultural-Genomics-1"></a><a class="docs-heading-anchor-permalink" href="#Agricultural-Genomics" title="Permalink"></a></h3><p>Applications in crop improvement</p><pre><code class="language-julia hljs">println(&quot;--- Agricultural Genomics ---&quot;)

println(&quot;Agricultural Genomics Applications:&quot;)
println(&quot;- Crop diversity assessment&quot;)
println(&quot;- Disease resistance breeding&quot;)
println(&quot;- Stress tolerance identification&quot;)
println(&quot;- Nutritional quality improvement&quot;)</code></pre><h3 id="Environmental-Genomics"><a class="docs-heading-anchor" href="#Environmental-Genomics">Environmental Genomics</a><a id="Environmental-Genomics-1"></a><a class="docs-heading-anchor-permalink" href="#Environmental-Genomics" title="Permalink"></a></h3><p>Applications in environmental microbiology</p><pre><code class="language-julia hljs">println(&quot;--- Environmental Genomics ---&quot;)

println(&quot;Environmental Genomics Applications:&quot;)
println(&quot;- Microbial community analysis&quot;)
println(&quot;- Ecosystem function prediction&quot;)
println(&quot;- Biogeochemical cycling&quot;)
println(&quot;- Climate change adaptation&quot;)</code></pre><h2 id="Part-12:-Best-Practices-and-Guidelines"><a class="docs-heading-anchor" href="#Part-12:-Best-Practices-and-Guidelines">Part 12: Best Practices and Guidelines</a><a id="Part-12:-Best-Practices-and-Guidelines-1"></a><a class="docs-heading-anchor-permalink" href="#Part-12:-Best-Practices-and-Guidelines" title="Permalink"></a></h2><p>Recommendations for comparative genomics analysis</p><pre><code class="language-julia hljs">println(&quot;\n=== Best Practices ===&quot;)

println(&quot;Data Quality:&quot;)
println(&quot;- Use high-quality genome assemblies&quot;)
println(&quot;- Ensure consistent annotation standards&quot;)
println(&quot;- Validate ortholog assignments&quot;)
println(&quot;- Check for contamination and artifacts&quot;)
println()
println(&quot;Analysis Strategy:&quot;)
println(&quot;- Start with closely related genomes&quot;)
println(&quot;- Use multiple clustering methods&quot;)
println(&quot;- Validate results with independent approaches&quot;)
println(&quot;- Consider biological context in interpretation&quot;)
println()
println(&quot;Computational Considerations:&quot;)
println(&quot;- Plan for large memory requirements&quot;)
println(&quot;- Use parallel processing when possible&quot;)
println(&quot;- Implement checkpointing for long analyses&quot;)
println(&quot;- Archive intermediate results&quot;)</code></pre><h2 id="Summary"><a class="docs-heading-anchor" href="#Summary">Summary</a><a id="Summary-1"></a><a class="docs-heading-anchor-permalink" href="#Summary" title="Permalink"></a></h2><pre><code class="language-julia hljs">println(&quot;\n=== Comparative Genomics Summary ===&quot;)
println(&quot;✓ Understanding pangenome concepts and construction&quot;)
println(&quot;✓ Implementing gene clustering and ortholog identification&quot;)
println(&quot;✓ Building core and accessory genome catalogs&quot;)
println(&quot;✓ Constructing graph-based pangenome representations&quot;)
println(&quot;✓ Performing phylogenetic analysis&quot;)
println(&quot;✓ Analyzing synteny and chromosomal rearrangements&quot;)
println(&quot;✓ Applying population genomics approaches&quot;)
println(&quot;✓ Creating comprehensive visualizations&quot;)
println(&quot;✓ Understanding real-world applications&quot;)</code></pre><p>Cleanup</p><pre><code class="language-julia hljs">for file in genome_files
    if isfile(file)
        rm(file, force=true)
    end
end

println(&quot;\nNext: Tutorial 8 - Tool Integration&quot;)

nothing</code></pre></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Thursday 11 September 2025 20:10">Thursday 11 September 2025</span>. Using Julia version 1.10.10.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
