{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "504b2382-d3aa-43f7-b02d-73f283cb5c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ENV[\"LD_LIBRARY_PATH\"] = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0594e70-4ab7-40f1-9b74-dd2897562648",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import Pkg\n",
    "Pkg.activate(\".\")\n",
    "\n",
    "pkgs = [\n",
    "\"BioSequences\",\n",
    "\"DataFrames\",\n",
    "\"Dates\",\n",
    "\"DelimitedFiles\",\n",
    "\"FASTX\",\n",
    "\"GLM\",\n",
    "\"HTTP\",\n",
    "\"JSON\",\n",
    "\"Graphs\",\n",
    "\"MetaGraphs\",\n",
    "\"MD5\",\n",
    "\"Statistics\",\n",
    "\"StatsPlots\",\n",
    "\"uCSV\",\n",
    "\"CodecZlib\",\n",
    "\"YAML\",\n",
    "\"Revise\",\n",
    "\"Kmers\",\n",
    "\"StatsBase\",\n",
    "\"ProgressMeter\"\n",
    "]\n",
    "Pkg.add(pkgs)\n",
    "for pkg in pkgs\n",
    "    eval(Meta.parse(\"import $pkg\"))\n",
    "end\n",
    "\n",
    "import Mycelia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b78fcf-5a2f-40dc-b136-db19875d8861",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = joinpath(dirname(pwd()), \"data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b47aae98-1bca-4d5b-9dae-cd9698c5bb89",
   "metadata": {},
   "outputs": [],
   "source": [
    "SRR_paths = filter(x -> !occursin(\".ipynb_checkpoints\", x), readdir(joinpath(data_dir, \"SRA\"), join=true))\n",
    "SRR_paths = filter(x -> \"trim_galore\" in readdir(x), SRR_paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e59e02f-3679-4e0f-b29a-ff4c798b5d62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# n_cores = max(Int(round(Sys.CPU_THREADS / 2)), 1)\n",
    "n_cores = Sys.CPU_THREADS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f56746fd-6258-4912-b064-dde22bfdaa2a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ProgressMeter.@showprogress for SRR_path in SRR_paths\n",
    "    SRR = basename(SRR_path)\n",
    "\n",
    "    out_dir = joinpath(SRR_path, \"megahit\")\n",
    "\n",
    "    trimmed_forward_reads = joinpath(SRR_path, \"trim_galore\", \"$(SRR)_1_val_1.fq.gz\")\n",
    "    trimmed_reverse_reads = joinpath(SRR_path, \"trim_galore\", \"$(SRR)_2_val_2.fq.gz\")\n",
    "\n",
    "    # megahit: MEGAHIT v1.2.9\n",
    "\n",
    "    # contact: Dinghua Li <voutcn@gmail.com>\n",
    "\n",
    "    # Usage:\n",
    "    #   megahit [options] {-1 <pe1> -2 <pe2> | --12 <pe12> | -r <se>} [-o <out_dir>]\n",
    "\n",
    "    #   Input options that can be specified for multiple times (supporting plain text and gz/bz2 extensions)\n",
    "    #     -1                       <pe1>          comma-separated list of fasta/q paired-end #1 files, paired with files in <pe2>\n",
    "    #     -2                       <pe2>          comma-separated list of fasta/q paired-end #2 files, paired with files in <pe1>\n",
    "    #     --12                     <pe12>         comma-separated list of interleaved fasta/q paired-end files\n",
    "    #     -r/--read                <se>           comma-separated list of fasta/q single-end files\n",
    "\n",
    "    # Optional Arguments:\n",
    "    #   Basic assembly options:\n",
    "    #     --min-count              <int>          minimum multiplicity for filtering (k_min+1)-mers [2]\n",
    "    #     --k-list                 <int,int,..>   comma-separated list of kmer size\n",
    "    #                                             all must be odd, in the range 15-255, increment <= 28)\n",
    "    #                                             [21,29,39,59,79,99,119,141]\n",
    "\n",
    "    #   Another way to set --k-list (overrides --k-list if one of them set):\n",
    "    #     --k-min                  <int>          minimum kmer size (<= 255), must be odd number [21]\n",
    "    #     --k-max                  <int>          maximum kmer size (<= 255), must be odd number [141]\n",
    "    #     --k-step                 <int>          increment of kmer size of each iteration (<= 28), must be even number [12]\n",
    "\n",
    "    #   Advanced assembly options:\n",
    "    #     --no-mercy                              do not add mercy kmers\n",
    "    #     --bubble-level           <int>          intensity of bubble merging (0-2), 0 to disable [2]\n",
    "    #     --merge-level            <l,s>          merge complex bubbles of length <= l*kmer_size and similarity >= s [20,0.95]\n",
    "    #     --prune-level            <int>          strength of low depth pruning (0-3) [2]\n",
    "    #     --prune-depth            <int>          remove unitigs with avg kmer depth less than this value [2]\n",
    "    #     --disconnect-ratio       <float>        disconnect unitigs if its depth is less than this ratio times \n",
    "    #                                             the total depth of itself and its siblings [0.1]  \n",
    "    #     --low-local-ratio        <float>        remove unitigs if its depth is less than this ratio times\n",
    "    #                                             the average depth of the neighborhoods [0.2]\n",
    "    #     --max-tip-len            <int>          remove tips less than this value [2*k]\n",
    "    #     --cleaning-rounds        <int>          number of rounds for graph cleanning [5]\n",
    "    #     --no-local                              disable local assembly\n",
    "    #     --kmin-1pass                            use 1pass mode to build SdBG of k_min\n",
    "\n",
    "    #   Presets parameters:\n",
    "    #     --presets                <str>          override a group of parameters; possible values:\n",
    "    #                                             meta-sensitive: '--min-count 1 --k-list 21,29,39,49,...,129,141'\n",
    "    #                                             meta-large: '--k-min 27 --k-max 127 --k-step 10'\n",
    "    #                                             (large & complex metagenomes, like soil)\n",
    "\n",
    "    #   Hardware options:\n",
    "    #     -t/--num-cpu-threads     <int>          number of CPU threads [# of logical processors]\n",
    "\n",
    "    #   Output options:\n",
    "    #     -o/--out-dir             <string>       output directory [./megahit_out]\n",
    "    #     --out-prefix             <string>       output prefix (the contig file will be OUT_DIR/OUT_PREFIX.contigs.fa)\n",
    "    #     --min-contig-len         <int>          minimum length of contigs to output [200]\n",
    "\n",
    "    # median_read_length = Statistics.median(vcat(Mycelia.determine_read_lengths(trimmed_forward_reads), Mycelia.determine_read_lengths(trimmed_reverse_reads)))\n",
    "\n",
    "    # k_lengths = [i for i in [21,29,39,59,79,99,119,141] if i < joint_median_read_length]\n",
    "    # k_lengths_string = join(string.(k_lengths), ',')\n",
    "\n",
    "    # 2023-02-19 00:51:46 - ALL DONE. Time elapsed: 2031.380950 seconds \n",
    "\n",
    "    initial_assembled_fasta = \"$(out_dir)/final.contigs.fa\"\n",
    "    assembled_fastg = replace(initial_assembled_fasta, \".fa\" => \".fastg\")\n",
    "\n",
    "    if !isfile(initial_assembled_fasta)\n",
    "        run(`megahit --num-cpu-threads $(n_cores) -1 $trimmed_forward_reads -2 $trimmed_reverse_reads -o $(out_dir)`)\n",
    "        # run(`conda run --live-stream --no-capture-output -n viral-pangenome-discovery megahit --num-cpu-threads $(n_cores) -1 $trimmed_forward_reads -2 $trimmed_reverse_reads -o $(out_dir)`)\n",
    "    end\n",
    "\n",
    "    # read in the assembled fasta file and parse contig identifiers to get final k length\n",
    "    final_k_lengths = unique([replace(first(split(FASTX.identifier(record), '_')), r\"^k\" => \"\") for record in Mycelia.open_fastx(initial_assembled_fasta)])\n",
    "    @assert length(final_k_lengths) == 1\n",
    "    final_k_length = parse(Int, first(final_k_lengths))\n",
    "    if !isfile(assembled_fastg)\n",
    "        run(pipeline(`megahit_toolkit contig2fastg $(final_k_length) $(initial_assembled_fasta)`, assembled_fastg))\n",
    "    end\n",
    "\n",
    "    assembled_gfa = \"$(assembled_fastg).gfa\"\n",
    "    if !isfile(assembled_gfa)\n",
    "        run(`Bandage reduce $(assembled_fastg) $(assembled_gfa)`)\n",
    "    end\n",
    "\n",
    "    assembled_fasta = assembled_gfa * \".fna\"\n",
    "    if !isfile(assembled_fasta)\n",
    "        open(assembled_fasta, \"w\") do io\n",
    "            fastx_io = FASTX.FASTA.Writer(io)\n",
    "            gfa_graph = Mycelia.parse_gfa(assembled_gfa)\n",
    "            for v in Graphs.vertices(gfa_graph)\n",
    "                record = FASTX.FASTA.Record(gfa_graph.vprops[v][:identifier], gfa_graph.vprops[v][:sequence])\n",
    "                write(fastx_io, record)\n",
    "            end\n",
    "            close(fastx_io)\n",
    "        end\n",
    "    end\n",
    "\n",
    "    # generate a bandage plot of the assembly graph\n",
    "    bandage_outfile = \"$(assembled_gfa).bandage.jpg\"\n",
    "    if !isfile(bandage_outfile)\n",
    "        run(`Bandage image $(assembled_gfa) $bandage_outfile`)\n",
    "    end\n",
    "\n",
    "    # map reads to the assembly and run qualimap QC\n",
    "    bwt_index = \"$(assembled_fasta).bwt\"\n",
    "    if !isfile(bwt_index)\n",
    "        run(`bwa index $(assembled_fasta)`)\n",
    "    end\n",
    "\n",
    "    # mapped_reads_bam = \"$(assembled_fasta).bwa.sorted.marked_duplicates.bam\"\n",
    "    mapped_reads_bam = \"$(assembled_fasta).bwa.bam\"\n",
    "    # sorting and deduping\n",
    "    # if !isfile(mapped_reads_bam)\n",
    "    #     run(pipeline(\n",
    "    #         `bwa mem -R \"@RG\\tID:$(SRR)\\tSM:bar\" -t $(n_cores) $(assembled_fasta) $(trimmed_forward_reads) $(trimmed_reverse_reads)`,\n",
    "    #         `samtools collate --threads $(n_cores) -O - -`,\n",
    "    #         `samtools fixmate --threads $(n_cores) -m - -`,\n",
    "    #         `samtools sort --threads $(n_cores)`,\n",
    "    #         `samtools markdup --threads $(n_cores) - -`,\n",
    "    #         `samtools view --threads $(n_cores) -buh`,\n",
    "    #         mapped_reads_bams))\n",
    "    # end\n",
    "    # sorting - 50Gb - 1594.823 sec\n",
    "    if !isfile(mapped_reads_bam) || (filesize(mapped_reads_bam) == 0)\n",
    "        run(pipeline(\n",
    "            `bwa mem -R \"@RG\\tID:$(SRR)\\tSM:bar\" -t $(n_cores) $(assembled_fasta) $(trimmed_forward_reads) $(trimmed_reverse_reads)`,\n",
    "            `samtools sort --threads $(n_cores) -u -m 100M`,\n",
    "            `samtools view --threads $(n_cores) -bh`,\n",
    "            mapped_reads_bam))\n",
    "    end\n",
    "\n",
    "    # if !isfile(\"$(mapped_reads_bam).bai\")\n",
    "    #     run(`samtools index $(mapped_reads_bam)`)\n",
    "    # end\n",
    "\n",
    "    qualimap_report_pdf = \"$(out_dir)/qualimap/report.pdf\"\n",
    "    qualimap_report_txt = \"$(out_dir)/qualimap/genome_results.txt\"\n",
    "\n",
    "    if !isfile(qualimap_report_pdf) || !isfile(qualimap_report_txt)\n",
    "        run(`\n",
    "            conda run --live-stream -n viral-pangenome-discovery\n",
    "            qualimap bamqc\n",
    "            -nt $(n_cores)\n",
    "            -bam $(mapped_reads_bam)\n",
    "            -outdir $(out_dir)/qualimap\n",
    "            -outformat PDF:HTML\n",
    "            --output-genome-coverage $(mapped_reads_bam).genome_coverage.txt\n",
    "            --java-mem-size=4G\n",
    "            `)\n",
    "    end\n",
    "end\n",
    "\n",
    "# qualimap_contig_coverage_table = Mycelia.parse_qualimap_contig_coverage(qualimap_report_txt)\n",
    "# qualimap_contig_coverage_table[!, \"% Mapped bases\"] = round.(qualimap_contig_coverage_table[!, \"Mapped bases\"] ./ sum(qualimap_contig_coverage_table[!, \"Mapped bases\"]) .* 100, digits=3);"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.6.7",
   "language": "julia",
   "name": "julia-1.6"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
