<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Step 6: Gene Annotation · Mycelia</title><meta name="title" content="Step 6: Gene Annotation · Mycelia"/><meta property="og:title" content="Step 6: Gene Annotation · Mycelia"/><meta property="twitter:title" content="Step 6: Gene Annotation · Mycelia"/><meta name="description" content="Documentation for Mycelia."/><meta property="og:description" content="Documentation for Mycelia."/><meta property="twitter:description" content="Documentation for Mycelia."/><script data-outdated-warner src="../../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../search_index.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../../">Mycelia</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../../">Home</a></li><li><a class="tocitem" href="../../../getting-started/">Getting Started</a></li><li><a class="tocitem" href="../../../installation/">Installation</a></li><li><a class="tocitem" href="../../../concepts/">Concepts</a></li><li><a class="tocitem" href="../../../workflow-map/">Workflow Map</a></li><li><a class="tocitem" href="../../../metagenomic-workflow/">Metagenomic Workflow</a></li><li><a class="tocitem" href="../../../microbiome-visualization/">Microbiome Visualization</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../../../tutorials/">Overview</a></li><li><a class="tocitem" href="../01_data_acquisition/">Step 1: Data Acquisition</a></li><li><a class="tocitem" href="../02_quality_control/">Step 2: Quality Control</a></li><li><a class="tocitem" href="../03_kmer_analysis/">Step 3: K-mer Analysis</a></li><li><a class="tocitem" href="../04_genome_assembly/">Step 4: Genome Assembly</a></li><li><a class="tocitem" href="../04_graph_type_tutorials/">Step 4b: Graph Type Tutorials</a></li><li><a class="tocitem" href="../05_assembly_validation/">Step 5: Assembly Validation</a></li><li class="is-active"><a class="tocitem" href>Step 6: Gene Annotation</a><ul class="internal"><li><a class="tocitem" href="#Learning-Objectives"><span>Learning Objectives</span></a></li><li><a class="tocitem" href="#Setup"><span>Setup</span></a></li><li><a class="tocitem" href="#Part-1:-Gene-Annotation-Overview"><span>Part 1: Gene Annotation Overview</span></a></li><li><a class="tocitem" href="#Part-2:-Structural-Annotation"><span>Part 2: Structural Annotation</span></a></li><li><a class="tocitem" href="#Part-3:-Functional-Annotation"><span>Part 3: Functional Annotation</span></a></li><li><a class="tocitem" href="#Part-4:-Annotation-Quality-Assessment"><span>Part 4: Annotation Quality Assessment</span></a></li><li><a class="tocitem" href="#Part-5:-Comparative-Annotation"><span>Part 5: Comparative Annotation</span></a></li><li><a class="tocitem" href="#Part-6:-Specialized-Annotation-Types"><span>Part 6: Specialized Annotation Types</span></a></li><li><a class="tocitem" href="#Part-7:-Annotation-Visualization"><span>Part 7: Annotation Visualization</span></a></li><li><a class="tocitem" href="#Part-8:-Annotation-File-Formats"><span>Part 8: Annotation File Formats</span></a></li><li><a class="tocitem" href="#Part-9:-Annotation-Pipelines"><span>Part 9: Annotation Pipelines</span></a></li><li><a class="tocitem" href="#Part-10:-Best-Practices-and-Guidelines"><span>Part 10: Best Practices and Guidelines</span></a></li><li><a class="tocitem" href="#Summary"><span>Summary</span></a></li></ul></li><li><a class="tocitem" href="../07_comparative_genomics/">Step 7: Comparative Genomics</a></li><li><a class="tocitem" href="../08_tool_integration/">Step 8: Tool Integration</a></li><li><a class="tocitem" href="../09_round_trip_01_string_graphs/">Round-Trip 01: String Graphs</a></li><li><a class="tocitem" href="../09_round_trip_02_ngram_to_string/">Round-Trip 02: N-gram to String</a></li><li><a class="tocitem" href="../09_round_trip_03_fasta_sequences/">Round-Trip 03: FASTA Sequences</a></li><li><a class="tocitem" href="../09_round_trip_04_kmer_to_sequence/">Round-Trip 04: K-mer to Sequence</a></li><li><a class="tocitem" href="../09_round_trip_05_fastq_graphs/">Round-Trip 05: FASTQ Graphs</a></li><li><a class="tocitem" href="../09_round_trip_06_qualmer_graphs/">Round-Trip 06: Qualmer Graphs</a></li><li><a class="tocitem" href="../10_viroid_assembly_workflow/">Step 10: Viroid Assembly Workflow</a></li><li><a class="tocitem" href="../11_reduced_amino_acid_alphabets/">Step 11: Reduced Amino Acid Alphabets</a></li><li><a class="tocitem" href="../12_coverm_coverage/">Step 12: CoverM Coverage</a></li><li><a class="tocitem" href="../13_rhizomorph_assembly/">Step 13: Rhizomorph Assembly</a></li><li><a class="tocitem" href="../14_binning_workflow/">Step 14a: Binning Workflow</a></li><li><a class="tocitem" href="../14_mash_classification/">Step 14b: Mash Classification</a></li><li><a class="tocitem" href="../15_round_trip_benchmarking/">Step 15: Round Trip Benchmarking</a></li><li><a class="tocitem" href="../16_un_corpus_ngram_vs_token_graphs/">Step 16: UN Corpus Ngram vs Token Graphs</a></li><li><a class="tocitem" href="../17_viroid_sketch_round_trip/">Step 17: Viroid Sketch Round Trip</a></li><li><a class="tocitem" href="../18_advanced_assembly_theory_and_practice/">Step 18: Advanced Assembly Theory and Practice</a></li><li><a class="tocitem" href="../19_relational_clustering/">Step 19: Relational Clustering</a></li><li><a class="tocitem" href="../20_microbiome_visualization/">Step 20: Microbiome Visualization</a></li><li><a class="tocitem" href="../21_cluster_comparison/">Cluster Comparison</a></li><li><a class="tocitem" href="../conceptual_figures/">Conceptual Figures</a></li><li><a class="tocitem" href="../run_all_tutorials/">Run All Tutorials</a></li></ul></li><li><span class="tocitem">API Reference</span><ul><li><a class="tocitem" href="../../../api/all-functions/">Complete API Surface</a></li></ul></li><li><a class="tocitem" href="../../../benchmarks/">Benchmarks</a></li><li><a class="tocitem" href="../../../references/">References</a></li><li><a class="tocitem" href="../../../contributing/">Contributing</a></li><li><a class="tocitem" href="../../../related-projects/">Related Projects</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorials</a></li><li class="is-active"><a href>Step 6: Gene Annotation</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Step 6: Gene Annotation</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/cjprybol/Mycelia" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/cjprybol/Mycelia/blob/master/tutorials/06_gene_annotation.jl" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Tutorial-6:-Gene-Annotation"><a class="docs-heading-anchor" href="#Tutorial-6:-Gene-Annotation">Tutorial 6: Gene Annotation</a><a id="Tutorial-6:-Gene-Annotation-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial-6:-Gene-Annotation" title="Permalink"></a></h1><p>This tutorial covers comprehensive gene annotation techniques, from basic gene prediction to functional annotation and annotation quality assessment.</p><h2 id="Learning-Objectives"><a class="docs-heading-anchor" href="#Learning-Objectives">Learning Objectives</a><a id="Learning-Objectives-1"></a><a class="docs-heading-anchor-permalink" href="#Learning-Objectives" title="Permalink"></a></h2><p>By the end of this tutorial, you will understand:</p><ul><li>Different gene prediction approaches and their applications</li><li>Structural annotation techniques for identifying genes and features</li><li>Functional annotation methods for assigning biological roles</li><li>Annotation quality assessment and validation</li><li>Comparative annotation approaches</li><li>Best practices for different organism types</li></ul><h2 id="Setup"><a class="docs-heading-anchor" href="#Setup">Setup</a><a id="Setup-1"></a><a class="docs-heading-anchor-permalink" href="#Setup" title="Permalink"></a></h2><p>From the Mycelia base directory, convert this tutorial to a notebook:</p><pre><code class="language-bash hljs">julia --project=. -e &#39;import Literate; Literate.notebook(&quot;tutorials/06_gene_annotation.jl&quot;, &quot;tutorials/notebooks&quot;, execute=false)&#39;</code></pre><pre><code class="language-julia hljs">import Pkg
if isinteractive()
    Pkg.activate(&quot;..&quot;)
end

import Test
import Mycelia
import FASTX
import Random
import Statistics

Random.seed!(42)</code></pre><h2 id="Part-1:-Gene-Annotation-Overview"><a class="docs-heading-anchor" href="#Part-1:-Gene-Annotation-Overview">Part 1: Gene Annotation Overview</a><a id="Part-1:-Gene-Annotation-Overview-1"></a><a class="docs-heading-anchor-permalink" href="#Part-1:-Gene-Annotation-Overview" title="Permalink"></a></h2><p>Gene annotation is a multi-step process that identifies genes and assigns functional information to genomic sequences.</p><pre><code class="language-julia hljs">println(&quot;=== Gene Annotation Tutorial ===&quot;)

println(&quot;Annotation Pipeline Overview:&quot;)
println(&quot;1. Structural Annotation&quot;)
println(&quot;   - Gene prediction&quot;)
println(&quot;   - Exon-intron structure&quot;)
println(&quot;   - Regulatory elements&quot;)
println()
println(&quot;2. Functional Annotation&quot;)
println(&quot;   - Protein function prediction&quot;)
println(&quot;   - Pathway assignment&quot;)
println(&quot;   - GO term annotation&quot;)
println()
println(&quot;3. Comparative Annotation&quot;)
println(&quot;   - Ortholog identification&quot;)
println(&quot;   - Synteny analysis&quot;)
println(&quot;   - Evolutionary analysis&quot;)</code></pre><h2 id="Part-2:-Structural-Annotation"><a class="docs-heading-anchor" href="#Part-2:-Structural-Annotation">Part 2: Structural Annotation</a><a id="Part-2:-Structural-Annotation-1"></a><a class="docs-heading-anchor-permalink" href="#Part-2:-Structural-Annotation" title="Permalink"></a></h2><p>Structural annotation identifies the physical structure of genes and other functional elements.</p><pre><code class="language-julia hljs">println(&quot;\n=== Structural Annotation ===&quot;)</code></pre><h3 id="Preparing-Test-Genome"><a class="docs-heading-anchor" href="#Preparing-Test-Genome">Preparing Test Genome</a><a id="Preparing-Test-Genome-1"></a><a class="docs-heading-anchor-permalink" href="#Preparing-Test-Genome" title="Permalink"></a></h3><p>Create a test genome for annotation demonstration</p><pre><code class="language-julia hljs">println(&quot;--- Preparing Test Genome ---&quot;)</code></pre><p>Generate test genome with gene-like features</p><pre><code class="language-julia hljs">genome_size = 50000
test_genome = Mycelia.random_fasta_record(moltype=:DNA, seed=1, L=genome_size)
genome_file = &quot;test_genome.fasta&quot;
Mycelia.write_fasta(outfile=genome_file, records=[test_genome])

println(&quot;Test genome prepared: $(genome_size) bp&quot;)</code></pre><h3 id="Ab-Initio-Gene-Prediction"><a class="docs-heading-anchor" href="#Ab-Initio-Gene-Prediction">Ab Initio Gene Prediction</a><a id="Ab-Initio-Gene-Prediction-1"></a><a class="docs-heading-anchor-permalink" href="#Ab-Initio-Gene-Prediction" title="Permalink"></a></h3><p>Predict genes using sequence signals alone</p><pre><code class="language-julia hljs">println(&quot;--- Ab Initio Gene Prediction ---&quot;)</code></pre><p>TODO: Implement ab initio gene prediction</p><ul><li>Use Prodigal for prokaryotic genes</li><li>Identify start/stop codons</li><li>Predict coding sequences</li><li>Handle overlapping genes</li></ul><p>Simulate gene prediction results</p><pre><code class="language-julia hljs">predicted_genes = [
    Dict(&quot;start&quot; =&gt; 1000, &quot;end&quot; =&gt; 2500, &quot;strand&quot; =&gt; &quot;+&quot;, &quot;confidence&quot; =&gt; 0.95),
    Dict(&quot;start&quot; =&gt; 3000, &quot;end&quot; =&gt; 4200, &quot;strand&quot; =&gt; &quot;-&quot;, &quot;confidence&quot; =&gt; 0.88),
    Dict(&quot;start&quot; =&gt; 5500, &quot;end&quot; =&gt; 7000, &quot;strand&quot; =&gt; &quot;+&quot;, &quot;confidence&quot; =&gt; 0.92),
    Dict(&quot;start&quot; =&gt; 8000, &quot;end&quot; =&gt; 9500, &quot;strand&quot; =&gt; &quot;+&quot;, &quot;confidence&quot; =&gt; 0.85),
    Dict(&quot;start&quot; =&gt; 10000, &quot;end&quot; =&gt; 11800, &quot;strand&quot; =&gt; &quot;-&quot;, &quot;confidence&quot; =&gt; 0.90)
]

println(&quot;Ab Initio Gene Prediction Results:&quot;)
println(&quot;  Predicted genes: $(length(predicted_genes))&quot;)
println(&quot;  Mean confidence: $(round(Statistics.mean([g[&quot;confidence&quot;] for g in predicted_genes]), digits=3))&quot;)
println(&quot;  Coding density: $(round(sum([g[&quot;end&quot;] - g[&quot;start&quot;] + 1 for g in predicted_genes]) / genome_size * 100, digits=1))%&quot;)</code></pre><h3 id="Taxonomy-Aware-ORF-Calling"><a class="docs-heading-anchor" href="#Taxonomy-Aware-ORF-Calling">Taxonomy-Aware ORF Calling</a><a id="Taxonomy-Aware-ORF-Calling-1"></a><a class="docs-heading-anchor-permalink" href="#Taxonomy-Aware-ORF-Calling" title="Permalink"></a></h3><p>When contigs are classified against NCBI, select the translation table before ORF calling so start/stop codons match the expected genetic code.</p><pre><code class="language-julia hljs">ncbi_taxonomy = Mycelia.load_ncbi_taxonomy()
tax_id = 562
table_id = Mycelia.get_ncbi_genetic_code(ncbi_taxonomy, tax_id)

orf_calls = Mycelia.run_pyrodigal(
    fasta_file=genome_file,
    out_dir=&quot;orf_calls&quot;,
    translation_table=table_id
)</code></pre><p>Use <code>type=:mitochondrial</code> for mitochondrial contigs.</p><h3 id="Homology-Based-Gene-Prediction"><a class="docs-heading-anchor" href="#Homology-Based-Gene-Prediction">Homology-Based Gene Prediction</a><a id="Homology-Based-Gene-Prediction-1"></a><a class="docs-heading-anchor-permalink" href="#Homology-Based-Gene-Prediction" title="Permalink"></a></h3><p>Use similarity to known genes for prediction</p><pre><code class="language-julia hljs">println(&quot;--- Homology-Based Gene Prediction ---&quot;)</code></pre><p>TODO: Implement homology-based prediction</p><ul><li>BLAST against protein databases</li><li>Identify homologous sequences</li><li>Transfer annotations from homologs</li><li>Handle partial matches</li></ul><p>Simulate homology search results</p><pre><code class="language-julia hljs">homology_results = [
    Dict(&quot;query&quot; =&gt; &quot;gene_1&quot;, &quot;subject&quot; =&gt; &quot;protein_X&quot;, &quot;identity&quot; =&gt; 85.2, &quot;coverage&quot; =&gt; 95.0),
    Dict(&quot;query&quot; =&gt; &quot;gene_2&quot;, &quot;subject&quot; =&gt; &quot;protein_Y&quot;, &quot;identity&quot; =&gt; 78.9, &quot;coverage&quot; =&gt; 88.0),
    Dict(&quot;query&quot; =&gt; &quot;gene_3&quot;, &quot;subject&quot; =&gt; &quot;protein_Z&quot;, &quot;identity&quot; =&gt; 92.1, &quot;coverage&quot; =&gt; 97.0),
    Dict(&quot;query&quot; =&gt; &quot;gene_4&quot;, &quot;subject&quot; =&gt; &quot;protein_W&quot;, &quot;identity&quot; =&gt; 65.4, &quot;coverage&quot; =&gt; 82.0),
    Dict(&quot;query&quot; =&gt; &quot;gene_5&quot;, &quot;subject&quot; =&gt; &quot;protein_V&quot;, &quot;identity&quot; =&gt; 88.7, &quot;coverage&quot; =&gt; 91.0)
]

println(&quot;Homology Search Results:&quot;)
for result in homology_results
    println(&quot;  $(result[&quot;query&quot;]) -&gt; $(result[&quot;subject&quot;]): $(result[&quot;identity&quot;])% identity, $(result[&quot;coverage&quot;])% coverage&quot;)
end</code></pre><h3 id="RNA-seq-Guided-Prediction"><a class="docs-heading-anchor" href="#RNA-seq-Guided-Prediction">RNA-seq Guided Prediction</a><a id="RNA-seq-Guided-Prediction-1"></a><a class="docs-heading-anchor-permalink" href="#RNA-seq-Guided-Prediction" title="Permalink"></a></h3><p>Use transcriptome data to improve gene prediction</p><pre><code class="language-julia hljs">println(&quot;--- RNA-seq Guided Prediction ---&quot;)</code></pre><p>TODO: Implement RNA-seq guided prediction</p><ul><li>Map RNA-seq reads to genome</li><li>Identify transcribed regions</li><li>Predict exon-intron structure</li><li>Handle alternative splicing</li></ul><h3 id="Eukaryotic-Gene-Prediction"><a class="docs-heading-anchor" href="#Eukaryotic-Gene-Prediction">Eukaryotic Gene Prediction</a><a id="Eukaryotic-Gene-Prediction-1"></a><a class="docs-heading-anchor-permalink" href="#Eukaryotic-Gene-Prediction" title="Permalink"></a></h3><p>Use Augustus for ab initio eukaryotic gene prediction or MetaEuk for metagenomic contigs with reference databases.</p><pre><code class="language-julia hljs">augustus = Mycelia.run_augustus(fasta_file=genome_file, species=&quot;human&quot;)
metaeuk = Mycelia.run_metaeuk(fasta_file=genome_file, db_file=&quot;uniref50_db&quot;)</code></pre><h3 id="Regulatory-Element-Prediction"><a class="docs-heading-anchor" href="#Regulatory-Element-Prediction">Regulatory Element Prediction</a><a id="Regulatory-Element-Prediction-1"></a><a class="docs-heading-anchor-permalink" href="#Regulatory-Element-Prediction" title="Permalink"></a></h3><p>Identify promoters, enhancers, and other regulatory sequences</p><pre><code class="language-julia hljs">println(&quot;--- Regulatory Element Prediction ---&quot;)</code></pre><p>TODO: Implement regulatory element prediction</p><ul><li>Promoter prediction</li><li>Transcription factor binding sites</li><li>CpG islands</li><li>Repetitive elements</li></ul><h2 id="Part-3:-Functional-Annotation"><a class="docs-heading-anchor" href="#Part-3:-Functional-Annotation">Part 3: Functional Annotation</a><a id="Part-3:-Functional-Annotation-1"></a><a class="docs-heading-anchor-permalink" href="#Part-3:-Functional-Annotation" title="Permalink"></a></h2><p>Functional annotation assigns biological roles to predicted genes</p><pre><code class="language-julia hljs">println(&quot;\n=== Functional Annotation ===&quot;)</code></pre><h3 id="Protein-Function-Prediction"><a class="docs-heading-anchor" href="#Protein-Function-Prediction">Protein Function Prediction</a><a id="Protein-Function-Prediction-1"></a><a class="docs-heading-anchor-permalink" href="#Protein-Function-Prediction" title="Permalink"></a></h3><p>Predict protein functions using various approaches</p><pre><code class="language-julia hljs">println(&quot;--- Protein Function Prediction ---&quot;)</code></pre><p>TODO: Implement protein function prediction</p><ul><li>Domain identification (Pfam, InterPro)</li><li>Enzyme classification (EC numbers)</li><li>Pathway assignment (KEGG, MetaCyc)</li><li>GO term annotation</li></ul><p>Simulate functional annotation results</p><pre><code class="language-julia hljs">functional_annotations = [
    Dict(&quot;gene&quot; =&gt; &quot;gene_1&quot;, &quot;function&quot; =&gt; &quot;DNA helicase&quot;, &quot;ec&quot; =&gt; &quot;3.6.4.12&quot;, &quot;confidence&quot; =&gt; 0.92),
    Dict(&quot;gene&quot; =&gt; &quot;gene_2&quot;, &quot;function&quot; =&gt; &quot;Transcriptional regulator&quot;, &quot;ec&quot; =&gt; &quot;&quot;, &quot;confidence&quot; =&gt; 0.78),
    Dict(&quot;gene&quot; =&gt; &quot;gene_3&quot;, &quot;function&quot; =&gt; &quot;Ribosomal protein L1&quot;, &quot;ec&quot; =&gt; &quot;&quot;, &quot;confidence&quot; =&gt; 0.95),
    Dict(&quot;gene&quot; =&gt; &quot;gene_4&quot;, &quot;function&quot; =&gt; &quot;Hypothetical protein&quot;, &quot;ec&quot; =&gt; &quot;&quot;, &quot;confidence&quot; =&gt; 0.45),
    Dict(&quot;gene&quot; =&gt; &quot;gene_5&quot;, &quot;function&quot; =&gt; &quot;ATP synthase subunit&quot;, &quot;ec&quot; =&gt; &quot;3.6.3.14&quot;, &quot;confidence&quot; =&gt; 0.88)
]

println(&quot;Functional Annotation Results:&quot;)
for annot in functional_annotations
    ec_str = annot[&quot;ec&quot;] != &quot;&quot; ? &quot; (EC: $(annot[&quot;ec&quot;]))&quot; : &quot;&quot;
    println(&quot;  $(annot[&quot;gene&quot;]): $(annot[&quot;function&quot;])$ec_str [$(annot[&quot;confidence&quot;])]&quot;)
end</code></pre><h3 id="Database-Annotation"><a class="docs-heading-anchor" href="#Database-Annotation">Database Annotation</a><a id="Database-Annotation-1"></a><a class="docs-heading-anchor-permalink" href="#Database-Annotation" title="Permalink"></a></h3><p>Annotate genes using specialized databases</p><pre><code class="language-julia hljs">println(&quot;--- Database Annotation ---&quot;)</code></pre><p>TODO: Implement database annotation</p><ul><li>BLAST against Swiss-Prot</li><li>Search COG database</li><li>Query KEGG pathways</li><li>Check specialized databases</li></ul><h3 id="Gene-Ontology-Annotation"><a class="docs-heading-anchor" href="#Gene-Ontology-Annotation">Gene Ontology Annotation</a><a id="Gene-Ontology-Annotation-1"></a><a class="docs-heading-anchor-permalink" href="#Gene-Ontology-Annotation" title="Permalink"></a></h3><p>Assign GO terms for standardized functional description</p><pre><code class="language-julia hljs">println(&quot;--- Gene Ontology Annotation ---&quot;)</code></pre><p>TODO: Implement GO annotation</p><ul><li>Assign GO terms</li><li>Validate GO term relationships</li><li>Calculate GO term confidence</li><li>Generate GO term summaries</li></ul><p>Simulate GO annotation results</p><pre><code class="language-julia hljs">go_annotations = [
    Dict(&quot;gene&quot; =&gt; &quot;gene_1&quot;, &quot;go_terms&quot; =&gt; [&quot;GO:0003678&quot;, &quot;GO:0006310&quot;], &quot;aspect&quot; =&gt; [&quot;MF&quot;, &quot;BP&quot;]),
    Dict(&quot;gene&quot; =&gt; &quot;gene_2&quot;, &quot;go_terms&quot; =&gt; [&quot;GO:0003677&quot;, &quot;GO:0006355&quot;], &quot;aspect&quot; =&gt; [&quot;MF&quot;, &quot;BP&quot;]),
    Dict(&quot;gene&quot; =&gt; &quot;gene_3&quot;, &quot;go_terms&quot; =&gt; [&quot;GO:0003735&quot;, &quot;GO:0006412&quot;], &quot;aspect&quot; =&gt; [&quot;MF&quot;, &quot;BP&quot;]),
    Dict(&quot;gene&quot; =&gt; &quot;gene_4&quot;, &quot;go_terms&quot; =&gt; [], &quot;aspect&quot; =&gt; []),
    Dict(&quot;gene&quot; =&gt; &quot;gene_5&quot;, &quot;go_terms&quot; =&gt; [&quot;GO:0046933&quot;, &quot;GO:0015986&quot;], &quot;aspect&quot; =&gt; [&quot;MF&quot;, &quot;BP&quot;])
]

println(&quot;GO Annotation Results:&quot;)
for annot in go_annotations
    if !isempty(annot[&quot;go_terms&quot;])
        println(&quot;  $(annot[&quot;gene&quot;]): $(join(annot[&quot;go_terms&quot;], &quot;, &quot;))&quot;)
    else
        println(&quot;  $(annot[&quot;gene&quot;]): No GO terms assigned&quot;)
    end
end</code></pre><h2 id="Part-4:-Annotation-Quality-Assessment"><a class="docs-heading-anchor" href="#Part-4:-Annotation-Quality-Assessment">Part 4: Annotation Quality Assessment</a><a id="Part-4:-Annotation-Quality-Assessment-1"></a><a class="docs-heading-anchor-permalink" href="#Part-4:-Annotation-Quality-Assessment" title="Permalink"></a></h2><p>Evaluate the quality and completeness of annotations</p><pre><code class="language-julia hljs">println(&quot;\n=== Annotation Quality Assessment ===&quot;)</code></pre><h3 id="Annotation-Completeness"><a class="docs-heading-anchor" href="#Annotation-Completeness">Annotation Completeness</a><a id="Annotation-Completeness-1"></a><a class="docs-heading-anchor-permalink" href="#Annotation-Completeness" title="Permalink"></a></h3><p>Assess what fraction of genes have functional annotations</p><pre><code class="language-julia hljs">println(&quot;--- Annotation Completeness ---&quot;)</code></pre><p>Calculate annotation statistics</p><pre><code class="language-julia hljs">total_genes = length(predicted_genes)
functionally_annotated = sum([annot[&quot;function&quot;] != &quot;Hypothetical protein&quot; for annot in functional_annotations])
ec_annotated = sum([annot[&quot;ec&quot;] != &quot;&quot; for annot in functional_annotations])
go_annotated = sum([!isempty(annot[&quot;go_terms&quot;]) for annot in go_annotations])

annotation_stats = Dict(
    &quot;total_genes&quot; =&gt; total_genes,
    &quot;functionally_annotated&quot; =&gt; functionally_annotated,
    &quot;ec_annotated&quot; =&gt; ec_annotated,
    &quot;go_annotated&quot; =&gt; go_annotated,
    &quot;functional_coverage&quot; =&gt; functionally_annotated / total_genes * 100,
    &quot;ec_coverage&quot; =&gt; ec_annotated / total_genes * 100,
    &quot;go_coverage&quot; =&gt; go_annotated / total_genes * 100
)

println(&quot;Annotation Completeness:&quot;)
for (metric, value) in annotation_stats
    println(&quot;  $metric: $value&quot;)
end</code></pre><h3 id="Annotation-Consistency"><a class="docs-heading-anchor" href="#Annotation-Consistency">Annotation Consistency</a><a id="Annotation-Consistency-1"></a><a class="docs-heading-anchor-permalink" href="#Annotation-Consistency" title="Permalink"></a></h3><p>Check for consistency between different annotation methods</p><pre><code class="language-julia hljs">println(&quot;--- Annotation Consistency ---&quot;)</code></pre><p>TODO: Implement annotation consistency checks</p><ul><li>Compare ab initio vs homology predictions</li><li>Validate functional annotations</li><li>Check for conflicting annotations</li><li>Assess annotation confidence</li></ul><h3 id="Annotation-Validation"><a class="docs-heading-anchor" href="#Annotation-Validation">Annotation Validation</a><a id="Annotation-Validation-1"></a><a class="docs-heading-anchor-permalink" href="#Annotation-Validation" title="Permalink"></a></h3><p>Validate annotations using external evidence</p><pre><code class="language-julia hljs">println(&quot;--- Annotation Validation ---&quot;)</code></pre><p>TODO: Implement annotation validation</p><ul><li>Cross-reference with literature</li><li>Validate with experimental data</li><li>Check annotation standards compliance</li><li>Assess annotation quality scores</li></ul><h2 id="Part-5:-Comparative-Annotation"><a class="docs-heading-anchor" href="#Part-5:-Comparative-Annotation">Part 5: Comparative Annotation</a><a id="Part-5:-Comparative-Annotation-1"></a><a class="docs-heading-anchor-permalink" href="#Part-5:-Comparative-Annotation" title="Permalink"></a></h2><p>Use comparative genomics to improve annotation quality</p><pre><code class="language-julia hljs">println(&quot;\n=== Comparative Annotation ===&quot;)</code></pre><h3 id="Ortholog-Identification"><a class="docs-heading-anchor" href="#Ortholog-Identification">Ortholog Identification</a><a id="Ortholog-Identification-1"></a><a class="docs-heading-anchor-permalink" href="#Ortholog-Identification" title="Permalink"></a></h3><p>Identify corresponding genes in related species</p><pre><code class="language-julia hljs">println(&quot;--- Ortholog Identification ---&quot;)</code></pre><p>TODO: Implement ortholog identification</p><ul><li>Bidirectional best hits</li><li>Ortholog clustering</li><li>Phylogenetic analysis</li><li>Synteny-based validation</li></ul><h3 id="Synteny-Analysis"><a class="docs-heading-anchor" href="#Synteny-Analysis">Synteny Analysis</a><a id="Synteny-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Synteny-Analysis" title="Permalink"></a></h3><p>Analyze conserved gene order for annotation validation</p><pre><code class="language-julia hljs">println(&quot;--- Synteny Analysis ---&quot;)</code></pre><p>TODO: Implement synteny analysis</p><ul><li>Identify syntenic blocks</li><li>Validate gene annotations</li><li>Detect gene duplications</li><li>Analyze evolutionary events</li></ul><h3 id="Evolutionary-Analysis"><a class="docs-heading-anchor" href="#Evolutionary-Analysis">Evolutionary Analysis</a><a id="Evolutionary-Analysis-1"></a><a class="docs-heading-anchor-permalink" href="#Evolutionary-Analysis" title="Permalink"></a></h3><p>Analyze gene evolution patterns</p><pre><code class="language-julia hljs">println(&quot;--- Evolutionary Analysis ---&quot;)</code></pre><p>TODO: Implement evolutionary analysis</p><ul><li>Selection pressure analysis</li><li>Gene family evolution</li><li>Horizontal gene transfer detection</li><li>Pseudogene identification</li></ul><h2 id="Part-6:-Specialized-Annotation-Types"><a class="docs-heading-anchor" href="#Part-6:-Specialized-Annotation-Types">Part 6: Specialized Annotation Types</a><a id="Part-6:-Specialized-Annotation-Types-1"></a><a class="docs-heading-anchor-permalink" href="#Part-6:-Specialized-Annotation-Types" title="Permalink"></a></h2><p>Handle organism-specific annotation challenges</p><pre><code class="language-julia hljs">println(&quot;\n=== Specialized Annotation ===&quot;)</code></pre><h3 id="Prokaryotic-Annotation"><a class="docs-heading-anchor" href="#Prokaryotic-Annotation">Prokaryotic Annotation</a><a id="Prokaryotic-Annotation-1"></a><a class="docs-heading-anchor-permalink" href="#Prokaryotic-Annotation" title="Permalink"></a></h3><p>Features specific to bacterial and archaeal genomes</p><pre><code class="language-julia hljs">println(&quot;--- Prokaryotic Annotation ---&quot;)</code></pre><p>TODO: Implement prokaryotic-specific annotation</p><ul><li>Operon prediction</li><li>Sigma factor binding sites</li><li>Ribosome binding sites</li><li>CRISPR arrays</li></ul><pre><code class="language-julia hljs">prokaryotic_features = Dict(
    &quot;operons&quot; =&gt; 3,
    &quot;sigma_sites&quot; =&gt; 12,
    &quot;ribosome_binding_sites&quot; =&gt; 5,
    &quot;crispr_arrays&quot; =&gt; 1
)

println(&quot;Prokaryotic Features:&quot;)
for (feature, count) in prokaryotic_features
    println(&quot;  $feature: $count&quot;)
end</code></pre><h3 id="Eukaryotic-Annotation"><a class="docs-heading-anchor" href="#Eukaryotic-Annotation">Eukaryotic Annotation</a><a id="Eukaryotic-Annotation-1"></a><a class="docs-heading-anchor-permalink" href="#Eukaryotic-Annotation" title="Permalink"></a></h3><p>Features specific to eukaryotic genomes</p><pre><code class="language-julia hljs">println(&quot;--- Eukaryotic Annotation ---&quot;)</code></pre><p>TODO: Implement eukaryotic-specific annotation</p><ul><li>Intron-exon structure</li><li>Alternative splicing</li><li>Pseudogenes</li><li>Non-coding RNAs</li></ul><h3 id="Viral-Annotation"><a class="docs-heading-anchor" href="#Viral-Annotation">Viral Annotation</a><a id="Viral-Annotation-1"></a><a class="docs-heading-anchor-permalink" href="#Viral-Annotation" title="Permalink"></a></h3><p>Features specific to viral genomes</p><pre><code class="language-julia hljs">println(&quot;--- Viral Annotation ---&quot;)</code></pre><p>TODO: Implement viral-specific annotation</p><ul><li>Overlapping genes</li><li>Frameshift elements</li><li>Regulatory sequences</li><li>Host interaction factors</li></ul><h2 id="Part-7:-Annotation-Visualization"><a class="docs-heading-anchor" href="#Part-7:-Annotation-Visualization">Part 7: Annotation Visualization</a><a id="Part-7:-Annotation-Visualization-1"></a><a class="docs-heading-anchor-permalink" href="#Part-7:-Annotation-Visualization" title="Permalink"></a></h2><p>Create visualizations for annotation results</p><pre><code class="language-julia hljs">println(&quot;\n=== Annotation Visualization ===&quot;)</code></pre><h3 id="Genome-Browser-Tracks"><a class="docs-heading-anchor" href="#Genome-Browser-Tracks">Genome Browser Tracks</a><a id="Genome-Browser-Tracks-1"></a><a class="docs-heading-anchor-permalink" href="#Genome-Browser-Tracks" title="Permalink"></a></h3><p>Generate tracks for genome browsers</p><pre><code class="language-julia hljs">println(&quot;--- Genome Browser Tracks ---&quot;)</code></pre><p>TODO: Implement genome browser track generation</p><ul><li>Gene structure tracks</li><li>Functional annotation tracks</li><li>Comparative annotation tracks</li><li>Quality assessment tracks</li></ul><h3 id="Annotation-Summary-Plots"><a class="docs-heading-anchor" href="#Annotation-Summary-Plots">Annotation Summary Plots</a><a id="Annotation-Summary-Plots-1"></a><a class="docs-heading-anchor-permalink" href="#Annotation-Summary-Plots" title="Permalink"></a></h3><p>Create summary visualizations</p><pre><code class="language-julia hljs">println(&quot;--- Annotation Summary Plots ---&quot;)</code></pre><p>TODO: Implement annotation visualization</p><ul><li>Functional category pie charts</li><li>Annotation quality histograms</li><li>Comparative annotation plots</li><li>Pathway enrichment plots</li></ul><h2 id="Part-8:-Annotation-File-Formats"><a class="docs-heading-anchor" href="#Part-8:-Annotation-File-Formats">Part 8: Annotation File Formats</a><a id="Part-8:-Annotation-File-Formats-1"></a><a class="docs-heading-anchor-permalink" href="#Part-8:-Annotation-File-Formats" title="Permalink"></a></h2><p>Work with standard annotation file formats</p><pre><code class="language-julia hljs">println(&quot;\n=== Annotation File Formats ===&quot;)</code></pre><h3 id="GFF3-Format"><a class="docs-heading-anchor" href="#GFF3-Format">GFF3 Format</a><a id="GFF3-Format-1"></a><a class="docs-heading-anchor-permalink" href="#GFF3-Format" title="Permalink"></a></h3><p>Generate and manipulate GFF3 files</p><pre><code class="language-julia hljs">println(&quot;--- GFF3 Format ---&quot;)</code></pre><p>TODO: Implement GFF3 handling</p><ul><li>Generate GFF3 from predictions</li><li>Validate GFF3 format</li><li>Convert between formats</li><li>Merge annotation files</li></ul><p>Generate example GFF3 content</p><pre><code class="language-julia hljs">gff3_content = &quot;##gff-version 3\n&quot; *
    &quot;##sequence-region test_genome 1 50000\n&quot; *
    &quot;test_genome\tprodigal\tgene\t1000\t2500\t0.95\t+\t.\tID=gene_1;Name=gene_1\n&quot; *
    &quot;test_genome\tprodigal\tCDS\t1000\t2500\t0.95\t+\t0\tID=cds_1;Parent=gene_1\n&quot; *
    &quot;test_genome\tprodigal\tgene\t3000\t4200\t0.88\t-\t.\tID=gene_2;Name=gene_2\n&quot; *
    &quot;test_genome\tprodigal\tCDS\t3000\t4200\t0.88\t-\t0\tID=cds_2;Parent=gene_2\n&quot;

gff3_file = &quot;annotations.gff3&quot;
open(gff3_file, &quot;w&quot;) do io
    write(io, gff3_content)
end

println(&quot;GFF3 file generated: $gff3_file&quot;)</code></pre><h3 id="GenBank-Format"><a class="docs-heading-anchor" href="#GenBank-Format">GenBank Format</a><a id="GenBank-Format-1"></a><a class="docs-heading-anchor-permalink" href="#GenBank-Format" title="Permalink"></a></h3><p>Generate GenBank format files</p><pre><code class="language-julia hljs">println(&quot;--- GenBank Format ---&quot;)</code></pre><p>TODO: Implement GenBank format handling</p><ul><li>Generate GenBank files</li><li>Include sequence and annotations</li><li>Validate format compliance</li><li>Extract annotations from GenBank</li></ul><h2 id="Part-9:-Annotation-Pipelines"><a class="docs-heading-anchor" href="#Part-9:-Annotation-Pipelines">Part 9: Annotation Pipelines</a><a id="Part-9:-Annotation-Pipelines-1"></a><a class="docs-heading-anchor-permalink" href="#Part-9:-Annotation-Pipelines" title="Permalink"></a></h2><p>Integrate annotation steps into comprehensive pipelines</p><pre><code class="language-julia hljs">println(&quot;\n=== Annotation Pipelines ===&quot;)</code></pre><h3 id="Automated-Annotation-Pipeline"><a class="docs-heading-anchor" href="#Automated-Annotation-Pipeline">Automated Annotation Pipeline</a><a id="Automated-Annotation-Pipeline-1"></a><a class="docs-heading-anchor-permalink" href="#Automated-Annotation-Pipeline" title="Permalink"></a></h3><p>Create automated annotation workflows</p><pre><code class="language-julia hljs">println(&quot;--- Automated Pipeline ---&quot;)</code></pre><p>TODO: Implement automated annotation pipeline</p><ul><li>Combine multiple prediction methods</li><li>Automated quality control</li><li>Standardized output formats</li><li>Batch processing capabilities</li></ul><h3 id="Manual-Curation-Interface"><a class="docs-heading-anchor" href="#Manual-Curation-Interface">Manual Curation Interface</a><a id="Manual-Curation-Interface-1"></a><a class="docs-heading-anchor-permalink" href="#Manual-Curation-Interface" title="Permalink"></a></h3><p>Tools for manual annotation improvement</p><pre><code class="language-julia hljs">println(&quot;--- Manual Curation ---&quot;)</code></pre><p>TODO: Implement manual curation tools</p><ul><li>Interactive annotation editor</li><li>Evidence integration interface</li><li>Collaborative annotation platform</li><li>Version control for annotations</li></ul><h2 id="Part-10:-Best-Practices-and-Guidelines"><a class="docs-heading-anchor" href="#Part-10:-Best-Practices-and-Guidelines">Part 10: Best Practices and Guidelines</a><a id="Part-10:-Best-Practices-and-Guidelines-1"></a><a class="docs-heading-anchor-permalink" href="#Part-10:-Best-Practices-and-Guidelines" title="Permalink"></a></h2><p>Recommendations for high-quality annotation</p><pre><code class="language-julia hljs">println(&quot;\n=== Annotation Best Practices ===&quot;)

println(&quot;General Principles:&quot;)
println(&quot;- Use multiple complementary approaches&quot;)
println(&quot;- Validate predictions with experimental data&quot;)
println(&quot;- Maintain annotation standards compliance&quot;)
println(&quot;- Document annotation procedures and evidence&quot;)
println()
println(&quot;Quality Thresholds:&quot;)
println(&quot;- Functional annotation: &gt;80% of genes&quot;)
println(&quot;- Homology confidence: &gt;70% identity, &gt;80% coverage&quot;)
println(&quot;- GO term coverage: &gt;60% of genes&quot;)
println(&quot;- Manual validation: High-confidence predictions&quot;)
println()
println(&quot;Organism-Specific Considerations:&quot;)
println(&quot;- Prokaryotes: Focus on operons and regulation&quot;)
println(&quot;- Eukaryotes: Handle alternative splicing carefully&quot;)
println(&quot;- Viruses: Account for overlapping genes&quot;)
println(&quot;- Metagenomes: Use taxonomic context&quot;)</code></pre><h2 id="Summary"><a class="docs-heading-anchor" href="#Summary">Summary</a><a id="Summary-1"></a><a class="docs-heading-anchor-permalink" href="#Summary" title="Permalink"></a></h2><pre><code class="language-julia hljs">println(&quot;\n=== Gene Annotation Summary ===&quot;)
println(&quot;✓ Understanding structural and functional annotation approaches&quot;)
println(&quot;✓ Implementing ab initio and homology-based gene prediction&quot;)
println(&quot;✓ Assigning functional annotations and GO terms&quot;)
println(&quot;✓ Evaluating annotation quality and completeness&quot;)
println(&quot;✓ Applying comparative annotation techniques&quot;)
println(&quot;✓ Handling organism-specific annotation challenges&quot;)
println(&quot;✓ Working with standard annotation file formats&quot;)
println(&quot;✓ Creating comprehensive annotation pipelines&quot;)</code></pre><p>Cleanup</p><pre><code class="language-julia hljs">cleanup_files = [genome_file, gff3_file]
for file in cleanup_files
    if isfile(file)
        rm(file, force=true)
    end
end

println(&quot;\nNext: Tutorial 7 - Comparative Genomics&quot;)

nothing</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../05_assembly_validation/">« Step 5: Assembly Validation</a><a class="docs-footer-nextpage" href="../07_comparative_genomics/">Step 7: Comparative Genomics »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Friday 30 January 2026 04:41">Friday 30 January 2026</span>. Using Julia version 1.10.10.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
